--- a/build/ac-macros/macosx.m4
+++ b/build/ac-macros/macosx.m4
@@ -19,80 +19,3 @@ dnl ====================================
 dnl
 dnl  Mac OS X specific checks
 
-dnl SVN_LIB_MACHO_ITERATE
-dnl Check for _dyld_image_name and _dyld_image_header availability
-AC_DEFUN(SVN_LIB_MACHO_ITERATE,
-[
-  AC_MSG_CHECKING([for Mach-O dynamic module iteration functions])
-  AC_RUN_IFELSE([AC_LANG_PROGRAM([[
-    #include <mach-o/dyld.h>
-    #include <mach-o/loader.h>
-  ]],[[
-    const struct mach_header *header = _dyld_get_image_header(0);
-    const char *name = _dyld_get_image_name(0);
-    if (name && header) return 0;
-    return 1;
-  ]])],[
-    AC_DEFINE([SVN_HAVE_MACHO_ITERATE], [1],
-              [Is Mach-O low-level _dyld API available?])
-    AC_MSG_RESULT([yes])
-  ],[
-    AC_MSG_RESULT([no])
-  ])
-])
-
-dnl SVN_LIB_MACOS_PLIST
-dnl Assign variables for Mac OS property list support
-AC_DEFUN(SVN_LIB_MACOS_PLIST,
-[
-  AC_MSG_CHECKING([for Mac OS property list utilities])
-
-  AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[
-    #include <AvailabilityMacros.h>
-    #if !defined(MAC_OS_X_VERSION_MAX_ALLOWED) \
-     || !defined(MAC_OS_X_VERSION_10_0) \
-     || (MAC_OS_X_VERSION_MAX_ALLOWED <= MAC_OS_X_VERSION_10_0)
-    #error ProperyList API unavailable.
-    #endif
-  ]],[[]])],[
-    SVN_MACOS_PLIST_LIBS="-framework CoreFoundation"
-    AC_SUBST(SVN_MACOS_PLIST_LIBS)
-    AC_DEFINE([SVN_HAVE_MACOS_PLIST], [1],
-              [Is Mac OS property list API available?])
-    AC_MSG_RESULT([yes])
-  ],[
-    AC_MSG_RESULT([no])
-  ])
-])
-
-dnl SVN_LIB_MACOS_KEYCHAIN
-dnl Check configure options and assign variables related to Keychain support
-
-AC_DEFUN(SVN_LIB_MACOS_KEYCHAIN,
-[
-  AC_ARG_ENABLE(keychain,
-    AS_HELP_STRING([--disable-keychain],
-    [Disable use of Mac OS KeyChain for auth credentials]),
-    [enable_keychain=$enableval],[enable_keychain=yes])
-
-  AC_MSG_CHECKING([for Mac OS KeyChain Services])
-
-  if test "$enable_keychain" = "yes"; then
-    AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[
-      #include <AvailabilityMacros.h>
-      #if !defined(MAC_OS_X_VERSION_MAX_ALLOWED) \
-       || !defined(MAC_OS_X_VERSION_10_2) \
-       || (MAC_OS_X_VERSION_MAX_ALLOWED <= MAC_OS_X_VERSION_10_2)
-      #error KeyChain API unavailable.
-      #endif
-    ]],[[]])],[
-      SVN_MACOS_KEYCHAIN_LIBS="-framework Security -framework CoreServices"
-      AC_SUBST(SVN_MACOS_KEYCHAIN_LIBS)
-      AC_DEFINE([SVN_HAVE_KEYCHAIN_SERVICES], [1], [Is Mac OS KeyChain support enabled?])
-      AC_MSG_RESULT([yes])
-    ],[
-      enable_keychain=no
-      AC_MSG_RESULT([no])
-    ])
-  fi
-])
--- a/Makefile.in
+++ b/Makefile.in
@@ -56,8 +56,8 @@ SVN_XML_LIBS = @SVN_XML_LIBS@
 SVN_ZLIB_LIBS = @SVN_ZLIB_LIBS@
 SVN_LZ4_LIBS = @SVN_LZ4_LIBS@
 SVN_UTF8PROC_LIBS = @SVN_UTF8PROC_LIBS@
-SVN_MACOS_PLIST_LIBS = @SVN_MACOS_PLIST_LIBS@
-SVN_MACOS_KEYCHAIN_LIBS = @SVN_MACOS_KEYCHAIN_LIBS@
+SVN_MACOS_PLIST_LIBS =
+SVN_MACOS_KEYCHAIN_LIBS =
 
 LIBS = @LIBS@
 
