--- a/src/osflags
+++ b/src/osflags
@@ -16,12 +16,6 @@ link)
 		windows32)
 			echo '-lws2_32 -liphlpapi';
 		;;
-		Linux)
-			FLAGS="";
-			[ -e /usr/include/selinux/selinux.h ] && FLAGS="$FLAGS -lselinux";
-			[ -e /usr/include/systemd/sd-daemon.h ] && FLAGS="$FLAGS -lsystemd-daemon";
-			echo $FLAGS;
-		;;
 	esac
 	;;
 cflags)
@@ -34,8 +28,6 @@ cflags)
 		;;
 		Linux)
 			FLAGS="-D_GNU_SOURCE"
-			[ -e /usr/include/selinux/selinux.h ] && FLAGS="$FLAGS -DHAVE_SETCON";
-			[ -e /usr/include/systemd/sd-daemon.h ] && FLAGS="$FLAGS -DHAVE_SYSTEMD";
 			echo $FLAGS;
 		;;
 	esac
