--- a/conf/nginx.conf
+++ b/conf/nginx.conf
@@ -1,5 +1,5 @@
 
-#user  nobody;
+user nobody nogroup;
 worker_processes  1;
 
 #error_log  logs/error.log;
@@ -16,7 +16,7 @@ events {
 
 http {
     include       mime.types;
-    default_type  application/octet-stream;
+    #default_type  application/octet-stream;
 
     #log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
     #                  '$status $body_bytes_sent "$http_referer" '
