Index: v4.25-9656/src/Mayaqua/Network.c
===================================================================
--- v4.25-9656.orig/src/Mayaqua/Network.c
+++ v4.25-9656/src/Mayaqua/Network.c
@@ -13025,7 +13025,7 @@ bool StartSSLEx(SOCK *sock, X *x, K *pri
 		{
 			if (client_tls == false)
 			{
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#ifndef SSL_OP_NO_SSLv3
 				SSL_CTX_set_ssl_version(ssl_ctx, SSLv3_method());
 #else
 				SSL_CTX_set_ssl_version(ssl_ctx, SSLv23_method());
