--- a/build/mkversion
+++ b/build/mkversion
@@ -50,12 +50,6 @@ if test $# != 1 ; then
 fi
 
 APPLICATION=$1
-# Reproducible builds set SOURCE_DATE_EPOCH, want constant strings
-if [ -n "${SOURCE_DATE_EPOCH}" ]; then
-   WHOWHERE="openldap"
-else
-   WHOWHERE="$USER@$(uname -n):$(pwd)"
-fi
 
 cat << __EOF__
 /* This work is part of OpenLDAP Software <http://www.openldap.org/>.
@@ -77,7 +71,6 @@ static const char copyright[] =
 "COPYING RESTRICTIONS APPLY\n";
 
 $static $const char $SYMBOL[] =
-"@(#) \$$PACKAGE: $APPLICATION $VERSION (" __DATE__ " " __TIME__ ") \$\n"
-"\t$WHOWHERE\n";
+"@(#) \$$PACKAGE: $APPLICATION $VERSION\$\n";
 
 __EOF__
