--- a/clients/tools/Makefile.in
+++ b/clients/tools/Makefile.in
@@ -122,7 +122,7 @@ install-local:	FORCE
 	-$(MKDIR) $(DESTDIR)$(bindir)
 	@(								\
 	    for prg in $(PROGRAMS); do					\
-		$(LTINSTALL) $(INSTALLFLAGS) $(STRIP) -m 755 $$prg$(EXEEXT)	\
+		$(LTINSTALL) $(INSTALLFLAGS) -m 755 $$prg$(EXEEXT)	\
 		    $(DESTDIR)$(bindir);				\
 	    done							\
 	)
--- a/servers/slapd/Makefile.in
+++ b/servers/slapd/Makefile.in
@@ -380,7 +380,7 @@ install-local-srv: install-slapd install
 install-slapd: FORCE
 	-$(MKDIR) $(DESTDIR)$(libexecdir)
 	-$(MKDIR) $(DESTDIR)$(localstatedir)/run
-	$(LTINSTALL) $(INSTALLFLAGS) $(STRIP) -m 755 \
+	$(LTINSTALL) $(INSTALLFLAGS) -m 755 \
 		slapd$(EXEEXT) $(DESTDIR)$(libexecdir)
 	@for i in $(SUBDIRS); do \
 	    if test -d $$i && test -f $$i/Makefile ; then \
