Index: u-boot-2017.01/configs/omap3_beagle_defconfig
===================================================================
--- u-boot-2017.01.orig/configs/omap3_beagle_defconfig
+++ u-boot-2017.01/configs/omap3_beagle_defconfig
@@ -7,7 +7,7 @@ CONFIG_SYS_CONSOLE_INFO_QUIET=y
 CONFIG_VERSION_VARIABLE=y
 CONFIG_SPL=y
 CONFIG_SPL_MTD_SUPPORT=y
-CONFIG_SPL_OS_BOOT=y
+# CONFIG_SPL_EXT_SUPPORT is not set
 # CONFIG_CMD_IMLS is not set
 CONFIG_CMD_ASKENV=y
 # CONFIG_CMD_FLASH is not set
Index: u-boot-2017.01/common/Makefile
===================================================================
--- u-boot-2017.01.orig/common/Makefile
+++ u-boot-2017.01/common/Makefile
@@ -174,8 +174,11 @@ obj-$(CONFIG_CMDLINE) += cli_simple.o
 obj-y += cli.o
 obj-$(CONFIG_CMDLINE) += cli_readline.o
 obj-$(CONFIG_CMD_DFU) += dfu.o
+
+ifndef CONFIG_SPL_BUILD
 obj-y += command.o
 obj-y += s_record.o
 obj-y += xyzModem.o
+endif
 
 CFLAGS_env_embedded.o := -Wa,--no-warn -DENV_CRC=$(shell tools/envcrc 2>/dev/null)
Index: u-boot-2017.01/cmd/Makefile
===================================================================
--- u-boot-2017.01.orig/cmd/Makefile
+++ u-boot-2017.01/cmd/Makefile
@@ -161,7 +161,10 @@ endif # !CONFIG_SPL_BUILD
 
 obj-$(CONFIG_CMD_BLOB) += blob.o
 
+
+# ifndef CONFIG_SPL_BUILD
 # core command
 obj-y += nvedit.o
+# endif # !CONFIG_SPL_BUILD
 
 obj-$(CONFIG_ARCH_MVEBU) += mvebu/
