Index: u-boot-2017.01/include/configs/omap3_overo.h
===================================================================
--- u-boot-2017.01.orig/include/configs/omap3_overo.h
+++ u-boot-2017.01/include/configs/omap3_overo.h
@@ -11,6 +11,13 @@
 #define CONFIG_NAND
 
 #include <configs/ti_omap3_common.h>
+
+/* try to enable thumb - unknown if it's working
+ * but otherwise it's too big for spl */
+#ifndef CONFIG_SYS_THUMB_BUILD
+#define CONFIG_SYS_THUMB_BUILD
+#endif
+
 /*
  * We are only ever GP parts and will utilize all of the "downloaded image"
  * area in SRAM which starts at 0x40200000 and ends at 0x4020FFFF (64KB).
