--- a/libraries/liblmdb/Makefile
+++ b/libraries/liblmdb/Makefile
@@ -34,6 +34,7 @@ libdir = $(exec_prefix)/lib
 includedir = $(prefix)/include
 datarootdir = $(prefix)/share
 mandir = $(datarootdir)/man
+FPIC ?= -fPIC
 
 ########################################################################
 
@@ -86,10 +87,10 @@ midl.o: midl.c midl.h
 	$(CC) $(CFLAGS) $(CPPFLAGS) -c midl.c
 
 mdb.lo: mdb.c lmdb.h midl.h
-	$(CC) $(CFLAGS) -fPIC $(CPPFLAGS) -c mdb.c -o $@
+	$(CC) $(CFLAGS) $(FPIC) $(CPPFLAGS) -c mdb.c -o $@
 
 midl.lo: midl.c midl.h
-	$(CC) $(CFLAGS) -fPIC $(CPPFLAGS) -c midl.c -o $@
+	$(CC) $(CFLAGS) $(FPIC) $(CPPFLAGS) -c midl.c -o $@
 
 %:	%.o
 	$(CC) $(CFLAGS) $(LDFLAGS) $^ $(LDLIBS) -o $@
