--- a/usr/tgtd.h
+++ b/usr/tgtd.h
@@ -9,6 +9,10 @@
 #include <systemd/sd-daemon.h>
 #endif
 
+#ifndef __WORDSIZE
+#include <sys/reg.h>
+#endif
+
 struct concat_buf;
 
 #define NR_SCSI_OPCODES		256
--- a/usr/util.h
+++ b/usr/util.h
@@ -16,6 +16,10 @@
 #include <limits.h>
 #include <linux/types.h>
 
+#ifndef __WORDSIZE
+#include <sys/reg.h>
+#endif
+
 #include "be_byteshift.h"
 
 #define roundup(x, y) ((((x) + ((y) - 1)) / (y)) * (y))
