--- a/src/router.c
+++ b/src/router.c
@@ -871,11 +871,11 @@
 	} else {
 		adv.h.nd_ra_router_lifetime = 0;
 
-		if (default_route)
+		/* if (default_route)
 			warn("A default route is present but there is no public prefix "
 			     "on %s thus we announce no default route by setting ra_lifetime to 0!", iface->name);
 		else
-			warn("No default route present, setting ra_lifetime to 0!");
+			warn("No default route present, setting ra_lifetime to 0!"); */
 	}
 
 	debug("Using a RA lifetime of %d seconds on %s", ntohs(adv.h.nd_ra_router_lifetime), iface->name);
