diff -Nuar a/src/infocenter.c b/src/infocenter.c
--- a/src/infocenter.c	2017-05-02 19:49:08.000000000 +0800
+++ b/src/infocenter.c	2021-12-29 17:37:00.837529076 +0800
@@ -116,9 +116,10 @@
      * be malformatted and the results are not reliable.
      */
     const char *stopped_at = tokenize_and_store(out, '\n', stores, store_len);
+    int ret = *stopped_at == 0 ? SUCCESS : FAILURE;
+    
     free((void *)out);
-
-    return *stopped_at == 0 ? SUCCESS : FAILURE;
+    return ret;
 }
 
 BASIC_INFO g_basic_info;
