--- a/newfs_hfs.tproj/makehfs.c
+++ b/newfs_hfs.tproj/makehfs.c
@@ -2119,10 +2119,12 @@ void GenerateVolumeUUID(VolumeUUID *newV
 		sysctl(mib, 2, sysctlstring, &datalen, NULL, 0);
 		SHA1_Update(&context, sysctlstring, datalen);
 #endif
+#ifndef __UCLIBC__
 		/* The system's load average: */
 		datalen = sizeof(sysloadavg);
 		getloadavg(sysloadavg, 3);
 		SHA1_Update(&context, &sysloadavg, datalen);
+#endif
 
 		/* The system's VM statistics: */
 #if !LINUX
