--- a/fsck_hfs.tproj/dfalib/SVerify1.c
+++ b/fsck_hfs.tproj/dfalib/SVerify1.c
@@ -2848,7 +2848,7 @@ OSErr	VLockedChk( SGlobPtr GPtr )
 	}
 	else		//	Because we don't have the unicode converters, just fill it with a dummy name.
 	{
-		CopyMemory( "\x0dPure HFS Plus", calculatedVCB->vcbVN, sizeof(Str27) );
+		CopyMemory( u"\x0dPure HFS Plus", calculatedVCB->vcbVN, sizeof(Str27) );
 	}
 		
 	GPtr->TarBlock = hint;
