--- a/fsck_hfs.tproj/fsck_hfs.h
+++ b/fsck_hfs.tproj/fsck_hfs.h
@@ -24,6 +24,7 @@
 
 #include "cache.h"
 
+#include <sys/cdefs.h>
 
 const extern char	*cdevname;		/* name of device being checked */
 extern char	*progname;
--- a/newfs_hfs.tproj/makehfs.c
+++ b/newfs_hfs.tproj/makehfs.c
@@ -28,6 +28,7 @@
 
 */
 
+#include <sys/cdefs.h>
 #include <sys/param.h>
 #include <sys/types.h>
 #include <sys/time.h>
--- a/newfs_hfs.tproj/newfs_hfs.c
+++ b/newfs_hfs.tproj/newfs_hfs.c
@@ -34,6 +34,7 @@
 #include <syslog.h>
 #include <unistd.h>
 
+#include <sys/cdefs.h>
 #include <sys/ioctl.h>
 #include <sys/mount.h>
 #include <sys/param.h>
