From 972fb2bbfdb7118551da45c5828c9e74be581fc4 Mon Sep 17 00:00:00 2001
From: Daniel Kamil Kozar <dkk089@gmail.com>
Date: Sat, 15 Dec 2018 21:55:52 +0100
Subject: [PATCH] What about David Bowie's "Heroes"?

---
 utils.c | 7 -------
 1 file changed, 7 deletions(-)

--- a/utils.c
+++ b/utils.c
@@ -87,13 +87,6 @@ trim(char *str)
 		len--;
 	}
 
-	if (str[0] == '"' && str[len-1] == '"')
-	{
-		str[0] = '\0';
-		str[len-1] = '\0';
-		str++;
-	}
-
 	return str;
 }
 
