From ce34745bb19a649c8570ef0a3da9ffd5e941548e Mon Sep 17 00:00:00 2001
From: Daniel Kamil Kozar <dkk089@gmail.com>
Date: Sun, 30 Dec 2018 22:49:50 +0100
Subject: [PATCH] No return in a function returning non-void

---
 upnpdescgen.c | 1 +
 1 file changed, 1 insertion(+)

--- a/upnpdescgen.c
+++ b/upnpdescgen.c
@@ -525,6 +525,7 @@ genXML(char * str, int * len, int * tmpl
 			pile[top].eltname = eltname;
 		}
 	}
+	return str;
 }
 
 /* genRootDesc() :
