--- a/hs20/server/spp_server.c
+++ b/hs20/server/spp_server.c
@@ -1329,7 +1329,7 @@ static xml_node_t * spp_get_mo(struct hs
 		return NULL;
 	}
 
-	snprintf(fname, sizeof(fname), "%s/spp/dm_ddf-v1_2.dtd", ctx->root_dir);
+	snprintf(fname, sizeof(fname), "%s/spp/dm_ddf-v1_3.dtd", ctx->root_dir);
 	if (xml_validate_dtd(ctx->xml, tnds, fname, ret_err) == 0)
 		*valid = 1;
 	else if (ret_err && *ret_err &&
