--- a/hs20/server/www/spp.php
+++ b/hs20/server/www/spp.php
@@ -141,7 +141,7 @@ $addr = $_SERVER["REMOTE_ADDR"];
 putenv("HS20ADDR=$addr");
 putenv("HS20TEST=$test");
 
-$last = exec("$osu_root/spp/hs20_spp_server -r$osu_root -f/tmp/hs20_spp_server.log", $output, $ret);
+$last = exec("/usr/sbin/hs20_spp_server -r$osu_root -f/tmp/hs20_spp_server.log", $output, $ret);
 
 if ($ret == 2) {
   if (empty($_SERVER['PHP_AUTH_DIGEST'])) {
