--- a/hs20/server/www/config.php
+++ b/hs20/server/www/config.php
@@ -1,7 +1,11 @@
 <?php
-$osu_root = "/home/user/hs20-server";
+$osu_root = "/etc/hs20";
 $osu_db = "sqlite:$osu_root/AS/DB/eap_user.db";
 $t_c_file = "$osu_root/terms-and-conditions";
-$t_c_timestamp = 123456789;
-$hostapd_ctrl = "udg:///home/user/hs20-server/AS/ctrl/as"
+if (file_exists($t_c_file)) {
+	$t_c_timestamp = filemtime($t_c_file);
+} else {
+	$t_c_timestamp = filemtime($osu_root);
+}
+$hostapd_ctrl = "udg:///var/run/hostapd/hs20-radius/lo"
 ?>
