From: coolsnowwolf <coolsnowwolf@gmail.com>
Subject: [PATCH] phytium: vpu drm fops unsigned offset

Signed-off-by: coolsnowwolf <coolsnowwolf@gmail.com>

--- a/vpu_driver/ftv310_vpu_drm.c
+++ b/vpu_driver/ftv310_vpu_drm.c
@@ -1778,6 +1778,7 @@
 /* VFS methods */
 static const struct file_operations ftv310_vpu_fops = {
 	.owner = THIS_MODULE,
+	.fop_flags = FOP_UNSIGNED_OFFSET,
 	.open = ftv310_vpu_device_open,
 	.mmap = ftv310_vpu_mmap,
 	.release = ftv310_vpu_device_release,
