From: coolsnowwolf <coolsnowwolf@gmail.com>
Subject: [PATCH] phytium: npu import dma buf heap namespace

Signed-off-by: coolsnowwolf <coolsnowwolf@gmail.com>

--- a/npu_driver/heap/phytium_npu_heap.c
+++ b/npu_driver/heap/phytium_npu_heap.c
@@ -396,5 +396,6 @@
 
 MODULE_LICENSE("GPL");
 MODULE_IMPORT_NS("DMA_BUF");
+MODULE_IMPORT_NS("DMA_BUF_HEAP");
 MODULE_AUTHOR("Cheng Quan <chengquan@phytium.com.cn>");
 MODULE_DESCRIPTION("Phytium NPU DMA BUF heap driver");
