From: coolsnowwolf <coolsnowwolf@gmail.com>
Subject: [PATCH] phytium: vpu avoid unsupported drm clflush

Signed-off-by: coolsnowwolf <coolsnowwolf@gmail.com>

--- a/vpu_driver/ftv310_vpu_drm.c
+++ b/vpu_driver/ftv310_vpu_drm.c
@@ -2163,6 +2163,8 @@
 		pages[i] = (vmalloc_to_page(address));
 		address += PAGE_SIZE;
 	}
+#if defined(CONFIG_X86) || defined(CONFIG_PPC)
 	drm_clflush_pages(pages, n_pages);
+#endif
 	return 0;
 }
--- a/vpu_driver/vgem/vgem_drv.c
+++ b/vpu_driver/vgem/vgem_drv.c
@@ -343,7 +343,9 @@
 	/* Flush the object from the CPU cache so that importers can rely
 	 * on coherent indirect access via the exported dma-address.
 	 */
+#if defined(CONFIG_X86) || defined(CONFIG_PPC)
 	drm_clflush_pages(pages, n_pages);
+#endif
 
 	return 0;
 }
