From 06bd6af97055071be5ed0f83a808032688a503db Mon Sep 17 00:00:00 2001
From: Yousong Zhou <yszhou4tech@gmail.com>
Date: Thu, 12 Nov 2020 14:33:05 +0800
Subject: [PATCH] xl2tpd-control: fix typo in usage

---
 xl2tpd-control.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/xl2tpd-control.c
+++ b/xl2tpd-control.c
@@ -96,7 +96,7 @@ void usage()
             "   --help  show this help message\n\n"
             "List of supported commands:\n"
             "add-lac, status-lac, remove-lac, connect-lac, disconnect-lac\n"
-            "add-lns, status-lns, remove-lns, avaliable\n\n"
+            "add-lns, status-lns, remove-lns, available\n\n"
             "See xl2tpd-control(8) man page for more details.\n");
 }
 
