--- a/mt7925/main.c
+++ b/mt7925/main.c
@@ -260,7 +260,8 @@ __mt7925_set_stream_he_eht_caps(struct m
 		n++;
 	}
 
-	_ieee80211_set_sband_iftype_data(sband, data, n);
+	sband->iftype_data = data;
+	sband->n_iftype_data = n;
 }
 
 void mt7925_set_stream_he_eht_caps(struct mt792x_phy *phy)
--- a/mt7925/mcu.c
+++ b/mt7925/mcu.c
@@ -2209,7 +2209,7 @@ mt7925_mcu_bss_mld_tlv(struct sk_buff *s
 		       struct ieee80211_sta *sta)
 {
 	struct mt792x_vif *mvif = (struct mt792x_vif *)vif->drv_priv;
-	bool is_mld = ieee80211_vif_is_mld(vif);
+	bool is_mld = (vif->valid_links != 0);
 	struct bss_mld_tlv *mld;
 	struct tlv *tlv;
 
