--- a/Makefile
+++ b/Makefile
@@ -1,5 +1,5 @@
 # SPDX-License-Identifier: GPL-2.0-only
-EXTRA_CFLAGS += -Werror -DCONFIG_MT76_LEDS
+EXTRA_CFLAGS += -Werror -Wno-unused-variable -DCONFIG_MT76_LEDS
 obj-m := mt76.o
 obj-$(CONFIG_MT76_USB) += mt76-usb.o
 obj-$(CONFIG_MT76_SDIO) += mt76-sdio.o
--- a/mac80211.c
+++ b/mac80211.c
@@ -582,7 +582,11 @@ int mt76_create_page_pool(struct mt76_de
 {
 	struct page_pool_params pp_params = {
 		.order = 0,
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,7,0)
+		.flags = 0,
+#else
 		.flags = PP_FLAG_PAGE_FRAG,
+#endif
 		.nid = NUMA_NO_NODE,
 		.dev = dev->dma_dev,
 	};
