From 8088940ffd047a6d282a95af829562e8038f5b2d Mon Sep 17 00:00:00 2001
From: Yukari Yakumo <mistelinn@gmail.com>
Date: Wed, 5 Feb 2025 09:33:47 +0300
Subject: [PATCH] add tx_power check during eeprom loading for BE14

---
 mt7996/eeprom.c | 7 +++++++
 1 file changed, 7 insertions(+)

--- a/mt7996/eeprom.c
+++ b/mt7996/eeprom.c
@@ -206,6 +206,13 @@ static int mt7996_eeprom_load(struct mt7
 				goto out;
 			}
 		}
+
+		/* read tx_power values from fw */
+		u8 *eeprom = dev->mt76.eeprom.data;
+		if (!eeprom[MT_EE_TX0_POWER_2G] || !eeprom[MT_EE_TX0_POWER_5G] || !eeprom[MT_EE_TX0_POWER_6G] ) {
+			use_default = true;
+			goto out;
+		}
 	}
 
 out:
