diff --git a/drivers/net/wireless/ath/ath10k/htt_rx.c b/drivers/net/wireless/ath/ath10k/htt_rx.c
index 7d295ee71534..50670d651cff 100644
--- a/drivers/net/wireless/ath/ath10k/htt_rx.c
+++ b/drivers/net/wireless/ath/ath10k/htt_rx.c
@@ -2387,7 +2387,7 @@ static void ath10k_htt_fetch_peer_stats(
 	ppdu_len = resp->peer_tx_stats.ppdu_len * sizeof(__le32);
 
 	if (skb->len < sizeof(struct htt_resp_hdr) + num_ppdu * ppdu_len) {
-		ath10k_warn(ar, "Invalid peer stats buf length %d\n", skb->len);
+
 		return;
 	}
 
@@ -2399,8 +2399,8 @@ static void ath10k_htt_fetch_peer_stats(
 	spin_lock_bh(&ar->data_lock);
 	peer = ath10k_peer_find_by_id(ar, peer_id);
 	if (!peer) {
-		ath10k_warn(ar, "Invalid peer id %d peer stats buffer\n",
-			    peer_id);
+
+
 		goto out;
 	}
 
