--- a/arch/arm/boot/dts/kirkwood-iomega_ix2_200.dts
+++ b/arch/arm/boot/dts/kirkwood-iomega_ix2_200.dts
@@ -186,18 +186,18 @@
 	};
 
 	partition@a0000 {
-		label = "env";
+		label = "u-boot environment";
 		reg = <0xa0000 0x20000>;
 		read-only;
 	};
 
 	partition@100000 {
-		label = "uImage";
+		label = "kernel";
 		reg = <0x100000 0x300000>;
 	};
 
 	partition@400000 {
-		label = "rootfs";
+		label = "ubi";
 		reg = <0x400000 0x1C00000>;
 	};
 };
@@ -211,7 +211,7 @@
 };
 
 &eth0 {
-	status = "okay";
+	status = "disabled";
 	ethernet0-port@0 {
 		speed = <1000>;
 		duplex = <1>;
