--- a/arch/mips/Makefile
+++ b/arch/mips/Makefile
@@ -227,6 +227,7 @@ cflags-$(toolchain-virt)		+= -DTOOLCHAIN
 #
 libs-$(CONFIG_FW_ARC)		+= arch/mips/fw/arc/
 libs-$(CONFIG_FW_CFE)		+= arch/mips/fw/cfe/
+libs-$(CONFIG_MYLOADER)		+= arch/mips/fw/myloader/
 libs-$(CONFIG_FW_SNIPROM)	+= arch/mips/fw/sni/
 libs-y				+= arch/mips/fw/lib/
 
--- a/arch/mips/Kconfig
+++ b/arch/mips/Kconfig
@@ -1151,6 +1151,9 @@ config MIPS_MSC
 config MIPS_NILE4
 	bool
 
+config MYLOADER
+	bool
+
 config SYNC_R4K
 	bool
 
