--- a/arch/mips/ath79/dev-wmac.c
+++ b/arch/mips/ath79/dev-wmac.c
@@ -406,6 +406,11 @@ void __init ath79_wmac_set_ext_lna_gpio(
 		ar934x_set_ext_lna_gpio(chain, gpio);
 }
 
+void __init ath79_wmac_set_led_pin(int gpio)
+{
+	ath79_wmac_data.led_pin = gpio;
+}
+
 void __init ath79_register_wmac(u8 *cal_data, u8 *mac_addr)
 {
 	if (soc_is_ar913x())
--- a/arch/mips/ath79/dev-wmac.h
+++ b/arch/mips/ath79/dev-wmac.h
@@ -18,6 +18,7 @@ void ath79_wmac_disable_2ghz(void);
 void ath79_wmac_disable_5ghz(void);
 void ath79_wmac_set_tx_gain_buffalo(void);
 void ath79_wmac_set_ext_lna_gpio(unsigned chain, int gpio);
+void ath79_wmac_set_led_pin(int gpio);
 
 bool ar93xx_wmac_read_mac_address(u8 *dest);
 
