--- a/arch/mips/ath79/prom.c
+++ b/arch/mips/ath79/prom.c
@@ -136,6 +136,32 @@ void __init prom_init(void)
 		initrd_end = initrd_start + fw_getenvl("initrd_size");
 	}
 #endif
+
+	if (strstr(arcs_cmdline, "board=750Gr3") ||
+	    strstr(arcs_cmdline, "board=750i") ||
+	    strstr(arcs_cmdline, "board=750-hb") ||
+	    strstr(arcs_cmdline, "board=411") ||
+	    strstr(arcs_cmdline, "board=433") ||
+	    strstr(arcs_cmdline, "board=435") ||
+	    strstr(arcs_cmdline, "board=450") ||
+	    strstr(arcs_cmdline, "board=493") ||
+	    strstr(arcs_cmdline, "board=951G") ||
+	    strstr(arcs_cmdline, "board=H951L") ||
+	    strstr(arcs_cmdline, "board=952-hb") ||
+	    strstr(arcs_cmdline, "board=953gs") ||
+	    strstr(arcs_cmdline, "board=962") ||
+	    strstr(arcs_cmdline, "board=lhg") ||
+	    strstr(arcs_cmdline, "board=map-hb") ||
+	    strstr(arcs_cmdline, "board=map2-hb") ||
+	    strstr(arcs_cmdline, "board=wap-hb") ||
+	    strstr(arcs_cmdline, "board=wap-lte") ||
+	    strstr(arcs_cmdline, "board=wapg-sc") ||
+	    strstr(arcs_cmdline, "board=2011L") ||
+	    strstr(arcs_cmdline, "board=2011r") ||
+	    strstr(arcs_cmdline, "board=711Gr100") ||
+	    strstr(arcs_cmdline, "board=911L") ||
+	    strstr(arcs_cmdline, "board=922gs"))
+		ath79_prom_append_cmdline("console", "ttyS0,115200");
 }
 
 void __init prom_free_prom_memory(void)
