--- a/include/linux/switch.h
+++ b/include/linux/switch.h
@@ -66,6 +66,11 @@ struct switch_port_stats {
 	unsigned long long rx_bytes;
 };
 
+struct switch_ext {
+	const char *option_name;
+	const char *option_value;
+	struct switch_ext *next;
+};
 /**
  * struct switch_dev_ops - switch driver operations
  *
@@ -157,6 +162,7 @@ struct switch_val {
 		const char *s;
 		u32 i;
 		struct switch_port *ports;
+		struct switch_ext *ext_val;
 		struct switch_port_link *link;
 	} value;
 };
--- a/include/uapi/linux/switch.h
+++ b/include/uapi/linux/switch.h
@@ -89,6 +89,7 @@ enum switch_val_type {
 	SWITCH_TYPE_PORTS,
 	SWITCH_TYPE_LINK,
 	SWITCH_TYPE_NOVAL,
+	SWITCH_TYPE_EXT,
 };
 
 /* port nested attributes */
