--- a/Makefile
+++ b/Makefile
@@ -76,7 +76,7 @@ all: sslh $(MAN) echosrv $(CONDITIONAL_T
 version.h:
 	./genver.sh >version.h
 
-sslh: sslh-fork sslh-select
+sslh: sslh-fork
 
 $(OBJS): version.h
 
@@ -87,10 +87,6 @@ sslh-fork: version.h $(OBJS) sslh-fork.o
 	$(CC) $(CFLAGS) $(LDFLAGS) -o sslh-fork sslh-fork.o $(OBJS) $(LIBS)
 	#strip sslh-fork
 
-sslh-select: version.h $(OBJS) sslh-select.o Makefile common.h
-	$(CC) $(CFLAGS) $(LDFLAGS) -o sslh-select sslh-select.o $(OBJS) $(LIBS)
-	#strip sslh-select
-
 systemd-sslh-generator: systemd-sslh-generator.o
 	$(CC) $(CFLAGS) $(LDFLAGS) -o systemd-sslh-generator systemd-sslh-generator.o -lconfig
 
@@ -127,7 +123,7 @@ distclean: clean
 	rm -f tags sslh-conf.c sslh-conf.h cscope.*
 
 clean:
-	rm -f sslh-fork sslh-select echosrv version.h $(MAN) systemd-sslh-generator *.o *.gcov *.gcno *.gcda *.png *.html *.css *.info
+	rm -f sslh-fork echosrv version.h $(MAN) systemd-sslh-generator *.o *.gcov *.gcno *.gcda *.png *.html *.css *.info
 
 tags:
 	ctags --globals -T *.[ch]
