--- a/common/Event.h
+++ b/common/Event.h
@@ -9,7 +9,7 @@
 
 
 #ifndef _WINDOWS
-	#include <tr1/memory>
+	#include <memory>
 	typedef void* EVENT_T;
 #else
 	#include <windows.h>
@@ -32,7 +32,7 @@ namespace TelldusCore {
 		virtual bool isValid() const;
 	};
 
-	typedef std::tr1::shared_ptr<EventData> EventDataRef;
+	typedef std::shared_ptr<EventData> EventDataRef;
 
 	class EventBase {
 	public:
@@ -74,7 +74,7 @@ namespace TelldusCore {
 		friend class EventHandler;
 	};
 
-	typedef std::tr1::shared_ptr<Event> EventRef;
+	typedef std::shared_ptr<Event> EventRef;
 }
 
 #endif  // TELLDUS_CORE_COMMON_EVENT_H_
