From 00adcdb64a5487f0ea6b400d912dcf7c58637696 Mon Sep 17 00:00:00 2001
From: Denys Vlasenko <vda.linux@googlemail.com>
Date: Fri, 20 Aug 2021 13:33:50 +0200
Subject: udhcp: fix build breakage on MIPS

Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
 networking/udhcp/common.h | 12 ------------
 1 file changed, 12 deletions(-)

diff --git a/networking/udhcp/common.h b/networking/udhcp/common.h
index 8c678dd32..ca778dab8 100644
--- a/networking/udhcp/common.h
+++ b/networking/udhcp/common.h
@@ -304,18 +304,6 @@ void udhcp_dump_packet(struct dhcp_packet *packet) FAST_FUNC;
 # define log3s(msg) ((void)0)
 #endif
 
-#if defined(__mips__)
-/*
- * The 'simple' message functions have a negative impact on the size of the
- * DHCP code when compiled for MIPS, so don't use them in this case.
- */
-#define bb_simple_info_msg bb_info_msg
-#define bb_simple_error_msg bb_error_msg
-#define bb_simple_perror_msg_and_die bb_perror_msg_and_die
-#undef log1s
-#define log1s log1
-#endif
-
 /*** Other shared functions ***/
 
 /* 2nd param is "uint32_t*" */
-- 
cgit v1.2.3


