--- a/util-linux/mount.c
+++ b/util-linux/mount.c
@@ -237,9 +237,6 @@
 #if ENABLE_FEATURE_MOUNT_NFS
 /* This is just a warning of a common mistake.  Possibly this should be a
  * uclibc faq entry rather than in busybox... */
-# if defined(__UCLIBC__) && ! defined(__UCLIBC_HAS_RPC__)
-#  error "You need to build uClibc with UCLIBC_HAS_RPC for NFS support"
-# endif
 # include <rpc/rpc.h>
 # include <rpc/pmap_prot.h>
 # include <rpc/pmap_clnt.h>
