--- a/scripts/Kbuild.include
+++ b/scripts/Kbuild.include
@@ -131,7 +131,7 @@ make-cmd = $(subst \#,\\\#,$(subst $$,$$
 #
 if_changed = $(if $(strip $(filter-out $(PHONY),$?)          \
 		$(call arg-check, $(cmd_$(1)), $(cmd_$@)) ), \
-	@set -e; \
+	+@set -e; \
 	$(echo-cmd) $(cmd_$(1)); \
 	echo 'cmd_$@ := $(make-cmd)' > $(@D)/.$(@F).cmd)
 
@@ -140,7 +140,7 @@ if_changed = $(if $(strip $(filter-out $
 if_changed_dep = $(if $(strip $(filter-out $(PHONY),$?)  \
 		$(filter-out FORCE $(wildcard $^),$^)    \
 	$(call arg-check, $(cmd_$(1)), $(cmd_$@)) ),     \
-	@set -e; \
+	+@set -e; \
 	$(echo-cmd) $(cmd_$(1)); \
 	scripts/basic/fixdep $(depfile) $@ '$(make-cmd)' > $(@D)/.$(@F).tmp; \
 	rm -f $(depfile); \
@@ -151,5 +151,5 @@ if_changed_dep = $(if $(strip $(filter-o
 # and if so will execute $(rule_foo)
 if_changed_rule = $(if $(strip $(filter-out $(PHONY),$?)            \
 			$(call arg-check, $(cmd_$(1)), $(cmd_$@)) ),\
-			@set -e; \
+			+@set -e; \
 			$(rule_$(1)))
