--- a/networking/traceroute.c
+++ b/networking/traceroute.c
@@ -239,8 +239,8 @@
 //config:	  Add option -I to use ICMP ECHO instead of UDP datagrams.
 
 /* Needs socket(AF_INET, SOCK_RAW, IPPROTO_ICMP), therefore BB_SUID_MAYBE: */
-//applet:IF_TRACEROUTE(APPLET(traceroute, BB_DIR_USR_BIN, BB_SUID_MAYBE))
-//applet:IF_TRACEROUTE6(APPLET(traceroute6, BB_DIR_USR_BIN, BB_SUID_MAYBE))
+//applet:IF_TRACEROUTE(APPLET(traceroute, BB_DIR_BIN, BB_SUID_MAYBE))
+//applet:IF_TRACEROUTE6(APPLET(traceroute6, BB_DIR_BIN, BB_SUID_MAYBE))
 
 //kbuild:lib-$(CONFIG_TRACEROUTE) += traceroute.o
 //kbuild:lib-$(CONFIG_TRACEROUTE6) += traceroute.o
