From: Giuseppe Lippolis <giu.lippolis@gmail.com>
Subject: Add the linux,spidev compatible in spidev Several device in ramips have this binding in the dts

Signed-off-by: Giuseppe Lippolis <giu.lippolis@gmail.com>
---
 drivers/spi/spidev.c | 1 +
 1 file changed, 1 insertion(+)

--- a/drivers/spi/spidev.c
+++ b/drivers/spi/spidev.c
@@ -672,6 +672,7 @@ static const struct of_device_id spidev_
 	{ .compatible = "lineartechnology,ltc2488" },
 	{ .compatible = "ge,achc" },
 	{ .compatible = "semtech,sx1301" },
+	{ .compatible = "siliconlabs,si3210" },
 	{},
 };
 MODULE_DEVICE_TABLE(of, spidev_dt_ids);
