--- a/backport-include/linux/random.h
+++ b/backport-include/linux/random.h
@@ -23,7 +23,7 @@ static inline u16 get_random_u16(void)
 }
 #endif
 
-#if LINUX_VERSION_IS_LESS(6,2,0)
+#if LINUX_VERSION_IS_LESS(6,1,4)
 static inline u32 __get_random_u32_below(u32 ceil)
 {
 	/*
--- a/backport-include/net/dropreason.h
+++ b/backport-include/net/dropreason.h
@@ -3,10 +3,9 @@
 
 #include <linux/version.h>
 
+#include <net/dropreason-core.h>
 #if LINUX_VERSION_IS_GEQ(6,0,0)
 #include_next <net/dropreason.h>
-#else
-#include <net/dropreason-core.h>
 #endif
 
 #if LINUX_VERSION_IS_LESS(6,4,0)
