--- a/backport-include/linux/etherdevice.h
+++ b/backport-include/linux/etherdevice.h
@@ -39,7 +39,11 @@ static inline void u64_to_ether_addr(u64
 }
 #endif /* LINUX_VERSION_IS_LESS(4,11,0) */
 
-#if LINUX_VERSION_IS_LESS(5,15,0)
+/* This was backported to 4.19.291, but we do not support such high minor numbers use 255 instead. */
+#if LINUX_VERSION_IS_LESS(5,15,0) &&			\
+	!LINUX_VERSION_IN_RANGE(5,10,188, 5,11,0) &&	\
+	!LINUX_VERSION_IN_RANGE(5,4,251, 5,5,0) &&	\
+	!LINUX_VERSION_IN_RANGE(4,19,255, 4,20,0)
 /**
  * eth_hw_addr_set - Assign Ethernet address to a net_device
  * @dev: pointer to net_device structure
