--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -16466,8 +16466,7 @@ static u32 nl80211_internal_flags[] = {
 #undef SELECTOR
 };
 
-static int nl80211_pre_doit(const struct genl_split_ops *ops,
-			    struct sk_buff *skb,
+static int nl80211_pre_doit(const struct genl_ops *ops, struct sk_buff *skb,
 			    struct genl_info *info)
 {
 	struct cfg80211_registered_device *rdev = NULL;
@@ -16568,8 +16567,7 @@ out_unlock:
 	return err;
 }
 
-static void nl80211_post_doit(const struct genl_split_ops *ops,
-			      struct sk_buff *skb,
+static void nl80211_post_doit(const struct genl_ops *ops, struct sk_buff *skb,
 			      struct genl_info *info)
 {
 	u32 internal_flags = nl80211_internal_flags[ops->internal_flags];
