--- a/drivers/staging/rtl8723bs/include/osdep_service_linux.h
+++ b/drivers/staging/rtl8723bs/include/osdep_service_linux.h
@@ -45,7 +45,11 @@
 		spinlock_t	lock;
 	};
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 17, 0))
+	#define thread_exit() kthread_complete_and_exit(NULL, 0)
+#else
 	#define thread_exit() complete_and_exit(NULL, 0)
+#endif
 
 static inline struct list_head *get_next(struct list_head	*list)
 {
