--- a/drivers/net/wireless/intel/iwlwifi/mvm/tt.c
+++ b/drivers/net/wireless/intel/iwlwifi/mvm/tt.c
@@ -679,8 +679,13 @@
 }
 #endif
 
+#if LINUX_VERSION_IS_LESS(6,12,0)
 static int iwl_mvm_tzone_set_trip_temp(struct thermal_zone_device *device,
 				       int trip, int temp)
+#else
+static int iwl_mvm_tzone_set_trip_temp(struct thermal_zone_device *device,
+				       const struct thermal_trip *trip, int temp)
+#endif
 {
 	struct iwl_mvm *mvm = thermal_zone_device_priv(device);
 	struct iwl_mvm_thermal_device *tzone;
@@ -754,8 +759,10 @@
 	.set_trip_temp = iwl_mvm_tzone_set_trip_temp,
 };
 
+#if LINUX_VERSION_IS_LESS(6,12,0)
 /* make all trips writable */
 #define IWL_WRITABLE_TRIPS_MSK (BIT(IWL_MAX_DTS_TRIPS) - 1)
+#endif
 
 static void iwl_mvm_thermal_zone_register(struct iwl_mvm *mvm)
 {
@@ -779,7 +786,9 @@
 							mvm->tz_device.trips,
 #endif
 							IWL_MAX_DTS_TRIPS,
+#if LINUX_VERSION_IS_LESS(6,12,0)
 							IWL_WRITABLE_TRIPS_MSK,
+#endif
 							mvm, &tzone_ops,
 							NULL, 0, 0);
 	if (IS_ERR(mvm->tz_device.tzone)) {
