--- a/drivers/irqchip/irq-gic-v3-its.c
+++ b/drivers/irqchip/irq-gic-v3-its.c
@@ -4752,11 +4752,13 @@ static bool __maybe_unused its_enable_qu
 	return true;
 }
 
-static bool __maybe_unused its_enable_rk3588001(void *data)
+static bool __maybe_unused its_enable_rk35xx(void *data)
 {
 	struct its_node *its = data;
 
-	if (!of_machine_is_compatible("rockchip,rk3588") &&
+	if (!of_machine_is_compatible("rockchip,rk3566") &&
+	    !of_machine_is_compatible("rockchip,rk3568") &&
+	    !of_machine_is_compatible("rockchip,rk3588") &&
 	    !of_machine_is_compatible("rockchip,rk3588s"))
 		return false;
 
@@ -4822,10 +4824,10 @@ static const struct gic_quirk its_quirks
 #endif
 #ifdef CONFIG_ROCKCHIP_ERRATUM_3588001
 	{
-		.desc   = "ITS: Rockchip erratum RK3588001",
+		.desc   = "ITS: Rockchip erratum RK35XX",
 		.iidr   = 0x0201743b,
 		.mask   = 0xffffffff,
-		.init   = its_enable_rk3588001,
+		.init   = its_enable_rk35xx,
 	},
 #endif
 	{
