--- a/arch/arm64/boot/dts/rockchip/rk3308-rock-pi-s.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3308-rock-pi-s.dts
@@ -18,6 +18,10 @@
 		mmc0 = &emmc;
 		mmc1 = &sdmmc;
 		mmc2 = &sdio;
+		led-boot = &blue_led;
+		led-failsafe = &blue_led;
+		led-running = &blue_led;
+		led-upgrade = &blue_led;
 	};
 
 	chosen {
@@ -29,22 +33,19 @@
 		pinctrl-names = "default";
 		pinctrl-0 = <&green_led>, <&heartbeat_led>;
 
-		green-led {
+		led-0 {
 			color = <LED_COLOR_ID_GREEN>;
 			default-state = "on";
 			function = LED_FUNCTION_POWER;
 			gpios = <&gpio0 RK_PA6 GPIO_ACTIVE_HIGH>;
-			label = "rockpis:green:power";
 			linux,default-trigger = "default-on";
 		};
 
-		blue-led {
+		blue_led: led-1 {
 			color = <LED_COLOR_ID_BLUE>;
 			default-state = "on";
 			function = LED_FUNCTION_HEARTBEAT;
 			gpios = <&gpio0 RK_PA5 GPIO_ACTIVE_HIGH>;
-			label = "rockpis:blue:user";
-			linux,default-trigger = "heartbeat";
 		};
 	};
 
