--- imx-bootlets-src-10.05.02.orig/Makefile	2010-05-14 06:56:28.000000000 +0200
+++ imx-bootlets-src-10.05.02/Makefile	2012-10-24 21:41:44.000000000 +0200
@@ -32,10 +32,11 @@
 	sed -i 's,[^ *]image.*;,\timage="$(DFT_UBOOT)";,' uboot.db
 	elftosb2 -z -c ./uboot.db -o i$(ARCH)_uboot.sb
 else
-	@echo "by using the pre-built kernel"
-	elftosb2 -z -c ./linux_prebuilt.db -o i$(ARCH)_linux.sb
-	@echo "generating U-Boot boot stream image"
-	elftosb2 -z -c ./uboot_prebuilt.db -o i$(ARCH)_uboot.sb
+	@echo "... not generating any image for now."
+	#@echo "by using the pre-built kernel"
+	#elftosb2 -z -c ./linux_prebuilt.db -o i$(ARCH)_linux.sb
+	#@echo "generating U-Boot boot stream image"
+	#elftosb2 -z -c ./uboot_prebuilt.db -o i$(ARCH)_uboot.sb
 endif
 	#@echo "generating kernel bootstream file sd_mmc_bootstream.raw"
 	#Please use cfimager to burn xxx_linux.sb. The below way will no
