--- imx-bootlets-10.05.02.orig/linux_prebuilt.db	2010-05-14 06:56:28.000000000 +0200
+++ imx-bootlets-10.05.02/linux_prebuilt.db	2012-10-24 22:04:37.000000000 +0200
@@ -4,10 +4,10 @@
 	flags = 0x01;
 }
 sources {
-	power_prep="./power_prep/power_prep";
-	sdram_prep="./boot_prep/boot_prep";
-	linux_prep="./linux_prep/output-target/linux_prep";
-	zImage = "./zImage";
+	power_prep="./power_prep";
+	sdram_prep="./boot_prep";
+	linux_prep="./linux_prep";
+	zImage = "./zImage_dtb";
 }
 
 section (0) {
