--- a/src/Makefile
+++ b/src/Makefile
@@ -5,25 +5,25 @@
 TOP=..
 include $(TOP)/Makefile.common
 
-INCLUDE= -I/usr/include
-LIBS= -L/usr/lib -lc 
-CFLAGS= -g3
-LDFLAGS= -Xlinker -Map -Xlinker $(PROG).map #,--stack,16Mb
+#INCLUDE= -I/usr/include
+#LIBS= -L/usr/lib -lc 
+#CFLAGS= -g3
+#LDFLAGS= -Xlinker -Map -Xlinker $(PROG).map #,--stack,16Mb
 OBJ= main.o command.o parse.o
 SCRIPTS= *.sh
 
 build: $(PROG) 
 	
 $(PROG): $(OBJ) 
-	gcc -o $(PROG) $(OBJ) $(LIBS) $(LDFLAGS)
+	$(GCC) -o $(PROG) $(OBJ) $(LIBS) $(LDFLAGS)
 	@echo "==============================================="
 	@echo "edit $(PROG).conf file to set default preferences"
 
 %.o : %.c 
-	gcc $(CFLAGS) -D__COPYLEFT__='$(COPYLEFT)' -c $<
+	$(GCC) $(CFLAGS) -D__COPYLEFT__='$(COPYLEFT)' -c $<
 
 %.o : %.c %.h
-	gcc $(CFLAGS) -D__COPYLEFT__='$(COPYLEFT)' -c $<
+	$(GCC) $(CFLAGS) -D__COPYLEFT__='$(COPYLEFT)' -c $<
 
 distclean clean:
 	rm $(PROG) *~ *.o -vf
