--- a/src/parse.c
+++ b/src/parse.c
@@ -289,6 +289,12 @@ int ParseCommand( char * pchCommandLine)
         pchEventDevice = strdup( pchValue) ;
         return 1 ;
       }
+      if( (pchValue != NULL) && 
+          (strncmp( pchValue, "/dev/event", 6) == 0) )
+      {
+        pchEventDevice = strdup( pchValue) ;
+        return 1 ;
+      }
       printf( "Option 'device' expects a /dev/input/eventX argument\n");
       return -1 ;
     }
