--- a/src/command.c
+++ b/src/command.c
@@ -43,6 +43,7 @@
 
 #include <linux/input.h>
 #include <linux/ioctl.h>
+#include <sys/wait.h>
 #include <stdlib.h>
 #include <fcntl.h>
 #include <unistd.h>
--- a/src/main.c
+++ b/src/main.c
@@ -54,6 +54,7 @@
 
 #include "debug.h"
 #include "command.h"
+#include "parse.h"
 
 #define DEBUGNAME         "MAIN: "
 
--- a/src/parse.c
+++ b/src/parse.c
@@ -58,7 +58,7 @@
 extern char * pchProgramName ;
 extern char * pchEventDevice ;
 
-inline void ltrim( char * pchText)
+static void ltrim( char * pchText)
 {
 	if( pchText) {
 		char * pchTxt = pchText ;
@@ -67,7 +67,7 @@ inline void ltrim( char * pchText)
   	}
 }
 
-inline void rtrim( char * pchText)
+static void rtrim( char * pchText)
 {
   	if( pchText)
   	{
@@ -77,7 +77,7 @@ inline void rtrim( char * pchText)
  	}
 }
 
-inline void trim( char * pchText) 
+static void trim( char * pchText)
 {
   	ltrim( pchText) ;
   	rtrim( pchText) ;
