--- a/net/mac80211/sta_info.c
+++ b/net/mac80211/sta_info.c
@@ -2490,10 +2490,7 @@ void ieee80211_sta_update_pending_airtim
 	atomic_sub(tx_airtime, &local->aql_total_pending_airtime);
 	tx_pending = atomic_sub_return(tx_airtime,
 				       &local->aql_ac_pending_airtime[ac]);
-	if (WARN_ONCE(tx_pending < 0,
-		      "Device %s AC %d pending airtime underflow: %u, %u",
-		      wiphy_name(local->hw.wiphy), ac, tx_pending,
-		      tx_airtime)) {
+	if (tx_pending < 0) {
 		atomic_cmpxchg(&local->aql_ac_pending_airtime[ac],
 			       tx_pending, 0);
 		atomic_sub(tx_pending, &local->aql_total_pending_airtime);
