--- a/net/mac80211/iface.c
+++ b/net/mac80211/iface.c
@@ -1304,7 +1304,7 @@ void ieee80211_del_virtual_monitor(struc
 /* This callback is registered for nss redirect to receive packet exceptioned from nss in Rx path.
  * When packet does not match any of the ecm rules is redirected back here.
  */
-void receive_from_nss(struct net_device *dev, struct sk_buff *sk_buff, struct napi_struct *napi)
+static void receive_from_nss(struct net_device *dev, struct sk_buff *sk_buff, struct napi_struct *napi)
 {
 	struct net_device *netdev;
 	struct sk_buff *skb;
--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -4584,7 +4584,7 @@ static void ieee80211_mlo_multicast_tx(s
 }
 
 #ifdef CPTCFG_MAC80211_NSS_SUPPORT
-void ieee80211_xmit_nss_fixup(struct sk_buff *skb,
+static void ieee80211_xmit_nss_fixup(struct sk_buff *skb,
 			      struct net_device *dev)
 {
 	struct ieee80211_sub_if_data *sdata = IEEE80211_DEV_TO_SUB_IF(dev);
@@ -4830,7 +4830,7 @@ out_free:
 	kfree_skb(skb);
 }
 
-void ieee80211_8023_xmit_ap(struct ieee80211_sub_if_data *sdata,
+static void ieee80211_8023_xmit_ap(struct ieee80211_sub_if_data *sdata,
 			    struct net_device *dev, struct sta_info *sta,
 			    struct ieee80211_key *key, struct sk_buff *skb,
 			    u32 info_flags, u32 ctrl_flags, u64 *cookie)
