--- a/Makefile.in
+++ b/Makefile.in
@@ -3225,7 +3225,7 @@ unit_tests = $(am__append_54) unit/test-
 @CLIENT_TRUE@					client/gatt.h client/gatt.c
 
 @CLIENT_TRUE@client_bluetoothctl_LDADD = gdbus/libgdbus-internal.la src/libshared-glib.la \
-@CLIENT_TRUE@				$(GLIB_LIBS) $(DBUS_LIBS) -lreadline
+@CLIENT_TRUE@				$(GLIB_LIBS) $(DBUS_LIBS) -lreadline -lncurses
 
 @ZSH_COMPLETIONS_TRUE@zshcompletiondir = $(ZSH_COMPLETIONDIR)
 @ZSH_COMPLETIONS_TRUE@dist_zshcompletion_DATA = completion/zsh/_bluetoothctl
@@ -3457,7 +3457,7 @@ unit_tests = $(am__append_54) unit/test-
 
 @MESH_TRUE@@TOOLS_TRUE@tools_meshctl_LDADD = gdbus/libgdbus-internal.la src/libshared-glib.la \
 @MESH_TRUE@@TOOLS_TRUE@				lib/libbluetooth-internal.la \
-@MESH_TRUE@@TOOLS_TRUE@				$(GLIB_LIBS) $(DBUS_LIBS) -ljson-c -lreadline
+@MESH_TRUE@@TOOLS_TRUE@				$(GLIB_LIBS) $(DBUS_LIBS) -ljson-c -lreadline -lncurses
 
 @MESH_TRUE@@TOOLS_TRUE@tools_mesh_cfgclient_SOURCES = tools/mesh-cfgclient.c \
 @MESH_TRUE@@TOOLS_TRUE@				tools/mesh/model.h tools/mesh/config-model.h \
@@ -3524,7 +3524,7 @@ unit_tests = $(am__append_54) unit/test-
 @READLINE_TRUE@						tools/obex-client-tool.c
 
 @READLINE_TRUE@tools_obex_client_tool_LDADD = lib/libbluetooth-internal.la \
-@READLINE_TRUE@						$(GLIB_LIBS) -lreadline
+@READLINE_TRUE@						$(GLIB_LIBS) -lreadline -lncurses
 
 @READLINE_TRUE@tools_obex_server_tool_SOURCES = $(gobex_sources) $(btio_sources) \
 @READLINE_TRUE@						tools/obex-server-tool.c
@@ -3533,15 +3533,15 @@ unit_tests = $(am__append_54) unit/test-
 @READLINE_TRUE@tools_bluetooth_player_SOURCES = tools/bluetooth-player.c
 @READLINE_TRUE@tools_bluetooth_player_LDADD = gdbus/libgdbus-internal.la \
 @READLINE_TRUE@				src/libshared-glib.la \
-@READLINE_TRUE@				$(GLIB_LIBS) $(DBUS_LIBS) -lreadline
+@READLINE_TRUE@				$(GLIB_LIBS) $(DBUS_LIBS) -lreadline -lncurses
 
 @READLINE_TRUE@tools_obexctl_SOURCES = tools/obexctl.c
 @READLINE_TRUE@tools_obexctl_LDADD = gdbus/libgdbus-internal.la src/libshared-glib.la \
-@READLINE_TRUE@				$(GLIB_LIBS) $(DBUS_LIBS) -lreadline
+@READLINE_TRUE@				$(GLIB_LIBS) $(DBUS_LIBS) -lreadline -lncurses
 
 @READLINE_TRUE@tools_btmgmt_SOURCES = tools/btmgmt.c src/uuid-helper.c client/display.c
 @READLINE_TRUE@tools_btmgmt_LDADD = lib/libbluetooth-internal.la src/libshared-mainloop.la \
-@READLINE_TRUE@				-lreadline
+@READLINE_TRUE@				-lreadline -lncurses
 
 @DEPRECATED_TRUE@@READLINE_TRUE@attrib_gatttool_SOURCES = attrib/gatttool.c attrib/att.c attrib/gatt.c \
 @DEPRECATED_TRUE@@READLINE_TRUE@				attrib/gattrib.c btio/btio.c \
@@ -3550,7 +3550,7 @@ unit_tests = $(am__append_54) unit/test-
 @DEPRECATED_TRUE@@READLINE_TRUE@				client/display.h
 
 @DEPRECATED_TRUE@@READLINE_TRUE@attrib_gatttool_LDADD = lib/libbluetooth-internal.la \
-@DEPRECATED_TRUE@@READLINE_TRUE@			src/libshared-glib.la $(GLIB_LIBS) -lreadline
+@DEPRECATED_TRUE@@READLINE_TRUE@			src/libshared-glib.la $(GLIB_LIBS) -lreadline -lncurses
 
 @CUPS_TRUE@cupsdir = $(libdir)/cups/backend
 @CUPS_TRUE@profiles_cups_bluetooth_SOURCES = profiles/cups/main.c \
