--- a/arch/arm64/boot/dts/rockchip/rk356x.dtsi
+++ b/arch/arm64/boot/dts/rockchip/rk356x.dtsi
@@ -53,7 +53,7 @@
 			device_type = "cpu";
 			compatible = "arm,cortex-a55";
 			reg = <0x0 0x0>;
-			clocks = <&scmi_clk 0>;
+			clocks = <&cru ARMCLK>;
 			#cooling-cells = <2>;
 			enable-method = "psci";
 			operating-points-v2 = <&cpu0_opp_table>;
@@ -63,6 +63,7 @@
 			device_type = "cpu";
 			compatible = "arm,cortex-a55";
 			reg = <0x0 0x100>;
+			clocks = <&cru ARMCLK>;
 			#cooling-cells = <2>;
 			enable-method = "psci";
 			operating-points-v2 = <&cpu0_opp_table>;
@@ -72,6 +73,7 @@
 			device_type = "cpu";
 			compatible = "arm,cortex-a55";
 			reg = <0x0 0x200>;
+			clocks = <&cru ARMCLK>;
 			#cooling-cells = <2>;
 			enable-method = "psci";
 			operating-points-v2 = <&cpu0_opp_table>;
@@ -81,6 +83,7 @@
 			device_type = "cpu";
 			compatible = "arm,cortex-a55";
 			reg = <0x0 0x300>;
+			clocks = <&cru ARMCLK>;
 			#cooling-cells = <2>;
 			enable-method = "psci";
 			operating-points-v2 = <&cpu0_opp_table>;
@@ -136,6 +139,7 @@
 			shmem = <&scmi_shmem>;
 			#address-cells = <1>;
 			#size-cells = <0>;
+			status = "disabled";
 
 			scmi_clk: protocol@14 {
 				reg = <0x14>;
@@ -193,6 +197,7 @@
 		scmi_shmem: sram@0 {
 			compatible = "arm,scmi-shmem";
 			reg = <0x0 0x100>;
+			status = "disabled";
 		};
 	};
 
--- a/drivers/clk/rockchip/clk-rk3568.c
+++ b/drivers/clk/rockchip/clk-rk3568.c
@@ -150,6 +150,12 @@ static struct rockchip_pll_rate_table rk3568_pll_rates[] = {
 }
 
 static struct rockchip_cpuclk_rate_table rk3568_cpuclk_rates[] __initdata = {
+	RK3568_CPUCLK_RATE(2208000000, 0, 1, 11, 11, 11, 11),
+	RK3568_CPUCLK_RATE(2184000000, 0, 1, 11, 11, 11, 11),
+	RK3568_CPUCLK_RATE(2088000000, 0, 1, 9, 9, 9, 9),
+	RK3568_CPUCLK_RATE(2016000000, 0, 1, 9, 9, 9, 9),
+	RK3568_CPUCLK_RATE(1992000000, 0, 1, 9, 9, 9, 9),
+	RK3568_CPUCLK_RATE(1896000000, 0, 1, 7, 7, 7, 7),
 	RK3568_CPUCLK_RATE(1800000000, 0, 1, 7, 7, 7, 7),
 	RK3568_CPUCLK_RATE(1704000000, 0, 1, 7, 7, 7, 7),
 	RK3568_CPUCLK_RATE(1608000000, 0, 1, 5, 5, 5, 5),
