--- a/arch/arm64/boot/dts/rockchip/rk3568-rock-3a.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3568-rock-3a.dts
@@ -22,6 +22,7 @@
 		stdout-path = "serial2:1500000n8";
 	};
 
+#ifdef DTS_NO_LEGACY
 	hdmi-con {
 		compatible = "hdmi-connector";
 		type = "a";
@@ -32,6 +33,7 @@
 			};
 		};
 	};
+#endif
 
 	gmac1_clkin: external-gmac1-clock {
 		compatible = "fixed-clock";
@@ -272,6 +274,7 @@
 	status = "okay";
 };
 
+#ifdef DTS_NO_LEGACY
 &gpu {
 	mali-supply = <&vdd_gpu>;
 	status = "okay";
@@ -300,6 +303,7 @@
 &hdmi_sound {
 	status = "okay";
 };
+#endif
 
 &i2c0 {
 	status = "okay";
@@ -562,9 +566,11 @@
 	};
 };
 
+#ifdef DTS_NO_LEGACY
 &i2s0_8ch {
 	status = "okay";
 };
+#endif
 
 &i2s1_8ch {
 	pinctrl-names = "default";
@@ -851,6 +857,7 @@
 	status = "okay";
 };
 
+#ifdef DTS_NO_LEGACY
 &vop {
 	assigned-clocks = <&cru DCLK_VOP0>, <&cru DCLK_VOP1>;
 	assigned-clock-parents = <&pmucru PLL_HPLL>, <&cru PLL_VPLL>;
@@ -867,3 +874,4 @@
 		remote-endpoint = <&hdmi_in_vp0>;
 	};
 };
+#endif
