From: coolsnowwolf <coolsnowwolf@gmail.com>
Subject: [PATCH] arm64: add Phytium platform and peripheral support

Signed-off-by: coolsnowwolf <coolsnowwolf@gmail.com>

--- a/arch/arm64/Kconfig.platforms
+++ b/arch/arm64/Kconfig.platforms
@@ -454,3 +454,9 @@
 	  This enables support for Xilinx ZynqMP Family
 
 endmenu # "Platform selection"
+
+config ARCH_PHYTIUM
+    bool "Phytium SoC Family"
+    help
+      This enables support for Phytium ARMv8 SoC family.
+
--- a/arch/arm64/boot/dts/Makefile
+++ b/arch/arm64/boot/dts/Makefile
@@ -38,3 +38,5 @@
 subdir-y += ti
 subdir-y += toshiba
 subdir-y += xilinx
+
+subdir-y += phytium
--- a/drivers/char/hw_random/Kconfig
+++ b/drivers/char/hw_random/Kconfig
@@ -631,3 +631,16 @@
 	  entropy pool.
 
 	  If unsure, say Y.
+
+config HW_RANDOM_PHYTIUM
+	tristate "Phytium Random Number Generator support"
+	depends on ARCH_PHYTIUM || COMPILE_TEST
+	help
+	  This driver provides kernel-side support for the Random Number
+	  Generator hardware found on Phytium SoCs.
+
+	  To compile this driver as a module, choose M here: the
+	  module will be called phytium-rng.
+
+	  If unsure, say Y.
+
--- a/drivers/char/hw_random/Makefile
+++ b/drivers/char/hw_random/Makefile
@@ -52,3 +52,6 @@
 obj-$(CONFIG_HW_RANDOM_POLARFIRE_SOC) += mpfs-rng.o
 obj-$(CONFIG_HW_RANDOM_ROCKCHIP) += rockchip-rng.o
 obj-$(CONFIG_HW_RANDOM_JH7110) += jh7110-trng.o
+
+
+obj-$(CONFIG_HW_RANDOM_PHYTIUM) += phytium-rng.o
--- a/drivers/char/ipmi/Kconfig
+++ b/drivers/char/ipmi/Kconfig
@@ -195,3 +195,25 @@
 	  Provides a driver for a device (Satellite MC) to
 	  receive requests and send responses back to the BMC via
 	  the IPMB interface. This module requires I2C support.
+
+config PHYTIUM_KCS_IPMI_BMC
+	depends on ARCH_PHYTIUM
+	select IPMI_KCS_BMC
+	select REGMAP_MMIO
+	tristate "PHYTIUM KCS IPMI BMC driver"
+	help
+	  Provides a driver for the KCS (Kerboard Controller Style) IPMI
+	  interface found on Phytium SOCs.
+
+	  The driver implements the BMC side of the KCS controller, it
+	  provides the access of KCS IO space for BMC side.
+
+config PHYTIUM_BT_IPMI_BMC
+	depends on ARCH_PHYTIUM
+	depends on REGMAP && REGMAP_MMIO && MFD_SYSCON
+	tristate "PHYTIUM BT BMC driver"
+	help
+	  Provides a driver for the BT (Block Transfer) IPMI interface
+	  found on Phytium SOCs. The driver implements the BMC
+	  side of the BT interface.
+
--- a/drivers/char/ipmi/Makefile
+++ b/drivers/char/ipmi/Makefile
@@ -29,3 +29,6 @@
 obj-$(CONFIG_NPCM7XX_KCS_IPMI_BMC) += kcs_bmc_npcm7xx.o
 obj-$(CONFIG_IPMB_DEVICE_INTERFACE) += ipmb_dev_int.o
 obj-$(CONFIG_SSIF_IPMI_BMC) += ssif_bmc.o
+
+obj-$(CONFIG_PHYTIUM_KCS_IPMI_BMC) += kcs_bmc_phytium.o
+obj-$(CONFIG_PHYTIUM_BT_IPMI_BMC) += bt_bmc_phytium.o
--- a/drivers/dma/Kconfig
+++ b/drivers/dma/Kconfig
@@ -803,3 +803,11 @@
 	bool
 
 endif
+
+config PHYTIUM_DDMA
+	bool "Phytium PE220x DDMA support"
+	depends on (ARCH_PHYTIUM || COMPILE_TEST)
+	select DMA_ENGINE
+	select DMA_VIRTUAL_CHANNELS
+	help
+	  Enable support for Phytium PE220x DDMA controller.
--- a/drivers/dma/Makefile
+++ b/drivers/dma/Makefile
@@ -92,3 +92,5 @@
 obj-y += stm32/
 obj-y += ti/
 obj-y += xilinx/
+
+obj-y += phytium/
--- a/drivers/edac/Kconfig
+++ b/drivers/edac/Kconfig
@@ -593,3 +593,11 @@
 	  HNICX, PL on the AMD Versal NET DDR memory controller.
 
 endif # EDAC
+
+config EDAC_PHYTIUM
+	tristate "Phytium Pe220x SoC"
+	depends on (ARM64)
+	help
+	  Support for error detection and correction on the
+	  Phytium Pe220x family of SOCs.
+
--- a/drivers/edac/Makefile
+++ b/drivers/edac/Makefile
@@ -90,3 +90,5 @@
 obj-$(CONFIG_EDAC_LOONGSON)		+= loongson_edac.o
 obj-$(CONFIG_EDAC_VERSALNET)		+= versalnet_edac.o
 obj-$(CONFIG_EDAC_CORTEX_A72)		+= a72_edac.o
+
+obj-$(CONFIG_EDAC_PHYTIUM)		+= phytium_edac.o
--- a/drivers/gpio/Kconfig
+++ b/drivers/gpio/Kconfig
@@ -2053,3 +2053,43 @@
 
 config DEV_SYNC_PROBE
 	tristate
+
+# This symbol is selected by both MMIO and PCI expanders
+config GPIO_PHYTIUM_CORE
+	tristate
+
+
+config GPIO_PHYTIUM_PLAT
+	tristate "Phytium GPIO Platform support"
+	default y if ARCH_PHYTIUM
+	depends on ARM64
+	select GPIO_PHYTIUM_CORE
+	select IRQ_DOMAIN
+	select GENERIC_IRQ_CHIP
+	select GPIOLIB_IRQCHIP
+	help
+	  Say yes here to enable GPIO support for Phytium SoCs.
+
+config GPIO_PHYTIUM_SGPIO
+	tristate "Phytium SGPIO support"
+	default y if ARCH_PHYTIUM
+	depends on ARM64
+	select IRQ_DOMAIN
+	select GENERIC_IRQ_CHIP
+	help
+	  Say yes here to enable SGPIO support for Phytium SoCs.
+
+
+config GPIO_PHYTIUM_PCI
+	tristate "Phytium GPIO PCI support"
+	select GPIO_PHYTIUM_CORE
+	select IRQ_DOMAIN
+	select GENERIC_IRQ_CHIP
+	select GPIOLIB_IRQCHIP
+	help
+	  Say Y here to support Phytium PCI GPIO controller on px210 chipset.
+	  An interrupt is generated when any of the inputs change state
+	  (low to high or high to low).
+
+	  This driver can be used for Phytium px210.
+
--- a/drivers/gpio/Makefile
+++ b/drivers/gpio/Makefile
@@ -217,3 +217,9 @@
 obj-$(CONFIG_GPIO_ZEVIO)		+= gpio-zevio.o
 obj-$(CONFIG_GPIO_ZYNQ)			+= gpio-zynq.o
 obj-$(CONFIG_GPIO_ZYNQMP_MODEPIN)	+= gpio-zynqmp-modepin.o
+
+
+obj-$(CONFIG_GPIO_PHYTIUM_CORE)		+= gpio-phytium-core.o
+obj-$(CONFIG_GPIO_PHYTIUM_PCI)		+= gpio-phytium-pci.o
+obj-$(CONFIG_GPIO_PHYTIUM_PLAT)		+= gpio-phytium-platform.o
+obj-$(CONFIG_GPIO_PHYTIUM_SGPIO)	+= gpio-phytium-sgpio.o
--- a/drivers/gpu/drm/Kconfig
+++ b/drivers/gpu/drm/Kconfig
@@ -426,3 +426,6 @@
 # Separate option because drm_panel_orientation_quirks.c is shared with fbdev
 config DRM_PANEL_ORIENTATION_QUIRKS
 	tristate
+
+source "drivers/gpu/drm/phytium/Kconfig"
+
--- a/drivers/gpu/drm/Makefile
+++ b/drivers/gpu/drm/Makefile
@@ -252,3 +252,5 @@
 
 $(obj)/%.hdrtest: $(src)/%.h FORCE
 	$(call if_changed_dep,hdrtest)
+
+obj-$(CONFIG_DRM_PHYTIUM) += phytium/
--- a/drivers/hwmon/Kconfig
+++ b/drivers/hwmon/Kconfig
@@ -2730,3 +2730,14 @@
 endif # ACPI
 
 endif # HWMON
+
+
+config SENSORS_PHYTIUM
+	tristate "Phytium Fan tach and capture counter driver"
+	help
+	  This driver provides support for Phytium Fan Tacho and capture
+	  counter controllers.
+
+	  This driver can also be built as a module. If so, the module
+	  will be called tacho-phytium.
+
--- a/drivers/hwmon/Makefile
+++ b/drivers/hwmon/Makefile
@@ -250,3 +250,6 @@
 obj-$(CONFIG_PMBUS)		+= pmbus/
 
 ccflags-$(CONFIG_HWMON_DEBUG_CHIP) := -DDEBUG
+
+obj-$(CONFIG_SENSORS_PHYTIUM)	+= tacho-phytium.o
+
--- a/drivers/hwspinlock/Kconfig
+++ b/drivers/hwspinlock/Kconfig
@@ -64,3 +64,13 @@
 	  If unsure, say N.
 
 endif # HWSPINLOCK
+
+config HWSPINLOCK_PHYTIUM
+	tristate "Phytium Hardware Spinlock device"
+	depends on HWSPINLOCK
+	depends on ARCH_PHYTIUM
+	help
+	  Say y here to support the Phytium Hardware Spinlock device.
+
+	  If unsure, say N.
+
--- a/drivers/hwspinlock/Makefile
+++ b/drivers/hwspinlock/Makefile
@@ -10,3 +10,6 @@
 obj-$(CONFIG_HWSPINLOCK_STM32)		+= stm32_hwspinlock.o
 obj-$(CONFIG_HWSPINLOCK_SUN6I)		+= sun6i_hwspinlock.o
 obj-$(CONFIG_HSEM_U8500)		+= u8500_hsem.o
+
+
+obj-$(CONFIG_HWSPINLOCK_PHYTIUM)	+= phytium_hwspinlock.o
--- a/drivers/i2c/busses/Kconfig
+++ b/drivers/i2c/busses/Kconfig
@@ -1597,3 +1597,32 @@
           will be called i2c-virtio.
 
 endmenu
+
+config I2C_PHYTIUM_CORE
+	tristate
+
+config I2C_PHYTIUM_PCI
+	tristate "Phytium I2C PCI"
+	depends on PCI && ARCH_PHYTIUM
+	select I2C_PHYTIUM_CORE
+	select I2C_SMBUS
+	help
+	  If you say yes to this option, support will be included for the
+	  Phytium I2C adapter. Only master mode is supported.
+
+	  This driver can also be built as a module.  If so, the module
+	  will be called i2c-phytium-pci.
+
+config I2C_PHYTIUM_PLATFORM
+	tristate "Phytium I2C Platform"
+	depends on (ACPI && COMMON_CLK) || !ACPI
+	select I2C_SLAVE
+	select I2C_PHYTIUM_CORE
+	select I2C_SMBUS
+	help
+	  If you say yes to this option, support will be included for the
+	  Phytium I2C adapter. Only master mode is supported.
+
+	  This driver can also be built as a module.  If so, the module
+	  will be called i2c-phytium-platform.
+
--- a/drivers/i2c/busses/Makefile
+++ b/drivers/i2c/busses/Makefile
@@ -163,3 +163,8 @@
 obj-$(CONFIG_I2C_VIRTIO)	+= i2c-virtio.o
 
 ccflags-$(CONFIG_I2C_DEBUG_BUS) := -DDEBUG
+
+obj-$(CONFIG_I2C_PHYTIUM_CORE)      += i2c-phytium-core.o
+i2c-phytium-core-objs := i2c-phytium-common.o i2c-phytium-mix.o
+obj-$(CONFIG_I2C_PHYTIUM_PCI)       += i2c-phytium-pci.o
+obj-$(CONFIG_I2C_PHYTIUM_PLATFORM)  += i2c-phytium-platform.o
--- a/drivers/iio/adc/Kconfig
+++ b/drivers/iio/adc/Kconfig
@@ -1880,3 +1880,16 @@
 	  xilinx-ams.
 
 endmenu
+
+config PHYTIUM_ADC
+	tristate "Phytium ADC driver"
+	depends on ARCH_PHYTIUM || COMPILE_TEST
+	select IIO_BUFFER
+	select IIO_TRIGGERED_BUFFER
+	help
+	  Say yes here to build support for Phytium analog to digital
+	  converters (ADC).
+
+	  To compile this driver as a module, choose M here: the module
+	  will be called phytium-adc.
+
--- a/drivers/iio/adc/Makefile
+++ b/drivers/iio/adc/Makefile
@@ -163,3 +163,5 @@
 obj-$(CONFIG_XILINX_AMS) += xilinx-ams.o
 xilinx-xadc-y := xilinx-xadc-core.o xilinx-xadc-events.o
 obj-$(CONFIG_XILINX_XADC) += xilinx-xadc.o
+
+obj-$(CONFIG_PHYTIUM_ADC) += phytium-adc.o
--- a/drivers/input/keyboard/Kconfig
+++ b/drivers/input/keyboard/Kconfig
@@ -793,3 +793,15 @@
 	  module will be called cypress-sf.
 
 endif
+
+config KEYBOARD_PHYTIUM
+	tristate "Phytium keypad support"
+	depends on ARCH_PHYTIUM
+	select INPUT_MATRIXKMAP
+	help
+	  Say Y here if you want to enable support for Phytium keypad
+	  port.
+
+	  To compile this driver as a module, choose M here: the
+	  module will be called phytium_keypad.
+
--- a/drivers/input/keyboard/Makefile
+++ b/drivers/input/keyboard/Makefile
@@ -70,3 +70,5 @@
 obj-$(CONFIG_KEYBOARD_TM2_TOUCHKEY)	+= tm2-touchkey.o
 obj-$(CONFIG_KEYBOARD_TWL4030)		+= twl4030_keypad.o
 obj-$(CONFIG_KEYBOARD_XTKBD)		+= xtkbd.o
+
+obj-$(CONFIG_KEYBOARD_PHYTIUM)		+= phytium-keypad.o
--- a/drivers/input/serio/Kconfig
+++ b/drivers/input/serio/Kconfig
@@ -320,3 +320,16 @@
 	  If you are unsure, say N.
 
 endif
+
+config SERIO_PHYTIUM_PS2
+	depends on SERIO
+	tristate "PHYTIUM PS/2 (keyboard and mouse)"
+	default y if ARCH_PHYTIUM
+	depends on PCI
+	help
+	  This selects support for the PS/2 Host Controller on
+	  Phytium SoCs.
+
+	  To compile this driver as a module, choose M here: the
+	  module will be called phytium-ps2.
+
--- a/drivers/input/serio/Makefile
+++ b/drivers/input/serio/Makefile
@@ -33,3 +33,5 @@
 obj-$(CONFIG_SERIO_SUN4I_PS2)	+= sun4i-ps2.o
 obj-$(CONFIG_SERIO_GPIO_PS2)	+= ps2-gpio.o
 obj-$(CONFIG_USERIO)		+= userio.o
+
+obj-$(CONFIG_SERIO_PHYTIUM_PS2)	+= phytium-ps2.o
--- a/drivers/irqchip/Kconfig
+++ b/drivers/irqchip/Kconfig
@@ -824,3 +824,12 @@
 	  the primary controller on C-Chip.
 
 endmenu
+
+config PHYTIUM_IXIC
+	bool "Phytium SoC PCI Legacy Interrupt Controller"
+	depends on ARCH_PHYTIUM
+	select IRQ_DOMAIN
+	select IRQ_DOMAIN_HIERARCHY
+	help
+	  This enables support PCI Legacy Interrupt on Phytium SoC.
+
--- a/drivers/irqchip/Makefile
+++ b/drivers/irqchip/Makefile
@@ -136,3 +136,5 @@
 obj-$(CONFIG_MCHP_EIC)			+= irq-mchp-eic.o
 obj-$(CONFIG_SOPHGO_SG2042_MSI)		+= irq-sg2042-msi.o
 obj-$(CONFIG_SUNPLUS_SP7021_INTC)	+= irq-sp7021-intc.o
+
+obj-$(CONFIG_PHYTIUM_IXIC)		+= irq-phytium-ixic.o
--- a/drivers/mailbox/Kconfig
+++ b/drivers/mailbox/Kconfig
@@ -391,3 +391,13 @@
 	  If unsure say N.
 
 endif
+
+config PHYTIUM_MBOX
+	tristate "Phytium SoC Mailbox Support"
+	depends on ARCH_PHYTIUM || COMPILE_TEST
+	help
+	  Mailbox driver implementation for the Phytium platform. It is used
+	  to send message between application processors and on-chip management
+	  firmware. Say Y here if you want to build this mailbox controller
+	  driver.
+
--- a/drivers/mailbox/Makefile
+++ b/drivers/mailbox/Makefile
@@ -82,3 +82,6 @@
 obj-$(CONFIG_BCM74110_MAILBOX)	+= bcm74110-mailbox.o
 
 obj-$(CONFIG_RISCV_SBI_MPXY_MBOX)	+= riscv-sbi-mpxy-mbox.o
+
+
+obj-$(CONFIG_PHYTIUM_MBOX)	+= phytium-mailbox.o
--- a/drivers/media/platform/Kconfig
+++ b/drivers/media/platform/Kconfig
@@ -92,3 +92,13 @@
 source "drivers/media/platform/xilinx/Kconfig"
 
 endif # MEDIA_PLATFORM_DRIVERS
+
+config VIDEO_PHYTIUM_JPEG
+        tristate "Phytium JPEG Encoder driver"
+        depends on VIDEO_V4L2
+        select VIDEOBUF2_DMA_CONTIG
+	help
+          Support for the Phytium JPEG Encoder Engine embedded
+          in the Phytium SOCs. The engine can capture and
+          compress video data from digital or analog sources.
+
--- a/drivers/media/platform/Makefile
+++ b/drivers/media/platform/Makefile
@@ -39,3 +39,5 @@
 # (e. g. LC_ALL=C sort Makefile)
 obj-$(CONFIG_VIDEO_MEM2MEM_DEINTERLACE)	+= m2m-deinterlace.o
 obj-$(CONFIG_VIDEO_MUX)			+= video-mux.o
+
+obj-$(CONFIG_VIDEO_PHYTIUM_JPEG)	+= phytium-jpeg/
--- a/drivers/mfd/Kconfig
+++ b/drivers/mfd/Kconfig
@@ -2573,3 +2573,19 @@
 
 endmenu
 endif
+
+config MFD_PHYTIUM_I2S_LSD
+	tristate "PHYTIUM px210 I2S LSD MFD driver"
+	depends on (PCI && ARCH_PHYTIUM)
+	select MFD_CORE
+	help
+	  This enables support for the Phytium px210 LSD I2S controller.
+
+config MFD_PHYTIUM_I2S_MMD
+	tristate "PHYTIUM px210 I2S MMD MFD driver"
+	depends on (PCI && ARCH_PHYTIUM)
+	select MFD_CORE
+	help
+	  This enables support for the Phytium px210 MMD I2S controllers
+	  for Display Port.
+
--- a/drivers/mfd/Makefile
+++ b/drivers/mfd/Makefile
@@ -303,3 +303,7 @@
 obj-$(CONFIG_MFD_UPBOARD_FPGA)	+= upboard-fpga.o
 
 obj-$(CONFIG_MFD_LOONGSON_SE)	+= loongson-se.o
+
+
+obj-$(CONFIG_MFD_PHYTIUM_I2S_LSD) += phytium_px210_i2s_lsd.o
+obj-$(CONFIG_MFD_PHYTIUM_I2S_MMD) += phytium_px210_i2s_mmd.o
--- a/drivers/mmc/host/Kconfig
+++ b/drivers/mmc/host/Kconfig
@@ -1141,3 +1141,38 @@
 	  module will be called litex_mmc.
 
 	  If unsure, say N.
+
+
+config MMC_PHYTIUM_SDCI
+	tristate "Phytium SD Host Controller support"
+	depends on ARCH_PHYTIUM || COMPILE_TEST
+	default y if ARCH_PHYTIUM
+	help
+	  This selects support for the SD/MMC Host Controller on
+	  Phytium SoC family.
+
+	  If you have a controller with this interface, say Y or M here.
+
+	  If unsure, say N.
+
+config MMC_PHYTIUM_MCI_PCI
+	tristate "Phytium octopus PCI MultiMedia Card Interface support"
+	depends on ARCH_PHYTIUM
+	default y if ARCH_PHYTIUM
+	help
+	  This selects support for the PCI MultiMedia Card Interface on Phytium
+	  px210 chipset.
+
+	  If you have a controller with this interface, say Y or M here.
+
+	  If unsure, say N.
+
+config MMC_PHYTIUM_MCI_PLTFM
+	tristate "Phytium MultiMedia Card Interface support"
+	depends on ARCH_PHYTIUM && OF
+	default y if ARCH_PHYTIUM
+        help
+          This selects support for the MultiMedia Card Interface on Phytium SoCs.
+	  If you have a controller with this interface, say Y or M here.
+
+	  If unsure, say N.
--- a/drivers/mmc/host/Makefile
+++ b/drivers/mmc/host/Makefile
@@ -114,3 +114,8 @@
 
 obj-$(CONFIG_MMC_SDHCI_XENON)	+= sdhci-xenon-driver.o
 sdhci-xenon-driver-y		+= sdhci-xenon.o sdhci-xenon-phy.o
+
+obj-$(CONFIG_MMC_PHYTIUM_SDCI)	+= phytium-sdci.o
+
+obj-$(CONFIG_MMC_PHYTIUM_MCI_PCI)	+= phytium-mci-pci.o phytium-mci.o
+obj-$(CONFIG_MMC_PHYTIUM_MCI_PLTFM)	+= phytium-mci-plat.o phytium-mci.o
--- a/drivers/mtd/nand/raw/Kconfig
+++ b/drivers/mtd/nand/raw/Kconfig
@@ -538,3 +538,23 @@
 	  parameter "inftl_bbt_write=1".
 
 endif # MTD_RAW_NAND
+
+config MTD_NAND_PHYTIUM
+	tristate
+
+config MTD_NAND_PHYTIUM_PCI
+	tristate "Support Phytium NAND controller as a PCI device"
+	select MTD_NAND_PHYTIUM
+	depends on PCI
+	help
+	  Enable the driver for NAND flash controller of Phytium Px210 chipset,
+	  using the Phytium NAND controller core.
+
+config MTD_NAND_PHYTIUM_PLAT
+	tristate "Support Phytium NAND controller as a platform device"
+	select MTD_NAND_PHYTIUM
+	depends on ARCH_PHYTIUM
+	help
+	  Enable the driver for NAND flash controller of Phytium CPU chipset,
+	  using the Phytium NAND controller core.
+
--- a/drivers/mtd/nand/raw/Makefile
+++ b/drivers/mtd/nand/raw/Makefile
@@ -71,3 +71,8 @@
 nand-objs += nand_sandisk.o
 nand-objs += nand_samsung.o
 nand-objs += nand_toshiba.o
+
+obj-$(CONFIG_MTD_NAND_PHYTIUM)		+= phytium_nand.o
+obj-$(CONFIG_MTD_NAND_PHYTIUM_PCI)	+= phytium_nand_pci.o
+obj-$(CONFIG_MTD_NAND_PHYTIUM_PLAT)	+= phytium_nand_plat.o
+
--- a/drivers/mtd/spi-nor/controllers/Kconfig
+++ b/drivers/mtd/spi-nor/controllers/Kconfig
@@ -16,3 +16,13 @@
 	  SPIFI is a specialized controller for connecting serial SPI
 	  Flash. Enable this option if you have a device with a SPIFI
 	  controller and want to access the Flash as a mtd device.
+
+
+config SPI_PHYTIUM_QUADSPI
+    tristate "Phytium Quad SPI Controller"
+    depends on ARCH_PHYTIUM || ARM
+    depends on OF && HAS_IOMEM
+    help
+      This enables support for the Quad SPI controller in master mode.
+      This driver does not support generic SPI. The implementation only
+      supports SPI NOR.
--- a/drivers/mtd/spi-nor/controllers/Makefile
+++ b/drivers/mtd/spi-nor/controllers/Makefile
@@ -1,3 +1,5 @@
 # SPDX-License-Identifier: GPL-2.0
 obj-$(CONFIG_SPI_HISI_SFC)	+= hisi-sfc.o
 obj-$(CONFIG_SPI_NXP_SPIFI)	+= nxp-spifi.o
+
+obj-$(CONFIG_SPI_PHYTIUM_QUADSPI)   += phytium-quadspi.o
--- a/drivers/net/can/Kconfig
+++ b/drivers/net/can/Kconfig
@@ -243,3 +243,6 @@
 	  on.
 
 endif #CAN_DEV
+
+source "drivers/net/can/phytium/Kconfig"
+
--- a/drivers/net/can/Makefile
+++ b/drivers/net/can/Makefile
@@ -35,3 +35,5 @@
 obj-$(CONFIG_CAN_XILINXCAN)	+= xilinx_can.o
 
 subdir-ccflags-$(CONFIG_CAN_DEBUG_DEVICES) += -DDEBUG
+
+obj-$(CONFIG_CAN_PHYTIUM)	+= phytium/
--- a/drivers/net/ethernet/stmicro/stmmac/Kconfig
+++ b/drivers/net/ethernet/stmicro/stmmac/Kconfig
@@ -371,3 +371,16 @@
 
 	  If unsure, say N.
 endif
+
+config DWMAC_PHYTIUM
+	depends on STMMAC_ETH
+	select STMMAC_PLATFORM
+	tristate "Phytium dwmac support"
+	default ARCH_PHYTIUM
+	depends on (OF || ACPI) && (ARCH_PHYTIUM || COMPILE_TEST)
+	help
+	  Support for GMAC controller on Phytium SoCs.
+
+	  This selects the Phytium GMAC glue layer support for the
+	  stmmac device driver.
+
--- a/drivers/net/ethernet/stmicro/stmmac/Makefile
+++ b/drivers/net/ethernet/stmicro/stmmac/Makefile
@@ -47,3 +47,5 @@
 obj-$(CONFIG_DWMAC_INTEL)	+= dwmac-intel.o
 obj-$(CONFIG_DWMAC_LOONGSON)	+= dwmac-loongson.o
 stmmac-pci-objs:= stmmac_pci.o
+
+obj-$(CONFIG_DWMAC_PHYTIUM)	+= dwmac-phytium.o
--- a/drivers/pwm/Kconfig
+++ b/drivers/pwm/Kconfig
@@ -820,3 +820,14 @@
 	  will be called pwm-xilinx.
 
 endif
+
+config PWM_PHYTIUM
+	tristate "Phytium PWM support"
+	depends on ARCH_PHYTIUM
+	help
+	  Generic PWM framework driver for the PWM controller found on
+	  Phytium SoCs.
+
+	  To compile this driver as a module, choose M here: the module
+	  will be called pwm-phytium.
+
--- a/drivers/pwm/Makefile
+++ b/drivers/pwm/Makefile
@@ -75,3 +75,6 @@
 obj-$(CONFIG_PWM_VISCONTI)	+= pwm-visconti.o
 obj-$(CONFIG_PWM_VT8500)	+= pwm-vt8500.o
 obj-$(CONFIG_PWM_XILINX)	+= pwm-xilinx.o
+
+
+obj-$(CONFIG_PWM_PHYTIUM)       += pwm-phytium.o
--- a/drivers/remoteproc/Kconfig
+++ b/drivers/remoteproc/Kconfig
@@ -381,3 +381,10 @@
 endif # REMOTEPROC
 
 endmenu
+
+config HOMO_REMOTEPROC
+	bool "homogeneous remoteproc support"
+	select RPMSG_VIRTIO
+	help
+	  Say y here to support homogeneous processors via the remote processor framework.
+
--- a/drivers/remoteproc/Makefile
+++ b/drivers/remoteproc/Makefile
@@ -40,3 +40,5 @@
 obj-$(CONFIG_TI_K3_M4_REMOTEPROC)	+= ti_k3_m4_remoteproc.o ti_k3_common.o
 obj-$(CONFIG_TI_K3_R5_REMOTEPROC)	+= ti_k3_r5_remoteproc.o ti_k3_common.o
 obj-$(CONFIG_XLNX_R5_REMOTEPROC)	+= xlnx_r5_remoteproc.o
+
+obj-$(CONFIG_HOMO_REMOTEPROC)  += homo_remoteproc.o
--- a/drivers/spi/Kconfig
+++ b/drivers/spi/Kconfig
@@ -1409,3 +1409,42 @@
 endif # SPI_OFFLOAD
 
 endif # SPI
+
+config SPI_PHYTIUM
+	tristate
+	depends on ARCH_PHYTIUM || COMPILE_TEST
+
+config SPI_PHYTIUM_PLAT
+	tristate "Phytium SPI controller platform support"
+	select SPI_PHYTIUM
+	help
+	  This selects a platform driver for Phytium SPI controller.
+
+	  If you say yes to this option, support will be included for
+	  Pd1904 and pd2008 families of SPI controller.
+
+config SPI_PHYTIUM_PCI
+	tristate "Phytium SPI controller PCI support"
+	depends on PCI
+	select SPI_PHYTIUM
+	help
+	  This selects a PCI driver for Phytium SPI controller.
+
+	  If you say yes to this option, support will be included for
+	  Phytium px210 chipset of SPI controller.
+
+	  If unsure, say N.
+
+config SPI_PHYTIUM_QSPI
+	tristate "Phytium Quad SPI controller"
+	depends on ARCH_PHYTIUM || COMPILE_TEST
+	depends on OF
+	depends on SPI_MEM
+	help
+	  This enables support for Phytium Quad SPI flash controller.
+
+	  This driver does not support generic SPI. The implementation only
+	  supports spi-mem interface.
+
+	  If unsure, say N.
+
--- a/drivers/spi/Makefile
+++ b/drivers/spi/Makefile
@@ -175,3 +175,9 @@
 # SPI offload triggers
 obj-$(CONFIG_SPI_OFFLOAD_TRIGGER_PWM)	+= spi-offload-trigger-pwm.o
 obj-$(CONFIG_SPI_OFFLOAD_TRIGGER_ADI_UTIL_SD) += spi-offload-trigger-adi-util-sigma-delta.o
+
+obj-$(CONFIG_SPI_PHYTIUM)		+= spi-phytium.o
+obj-$(CONFIG_SPI_PHYTIUM_PLAT)		+= spi-phytium-plat.o
+obj-$(CONFIG_SPI_PHYTIUM_PCI)		+= spi-phytium-pci.o
+obj-$(CONFIG_SPI_PHYTIUM_QSPI)		+= spi-phytium-qspi.o
+obj-$(CONFIG_SPI_PHYTIUM)       += spi-phytium-dma.o
--- a/drivers/tty/serial/Kconfig
+++ b/drivers/tty/serial/Kconfig
@@ -1623,3 +1623,15 @@
 
 config SERIAL_MCTRL_GPIO
 	tristate
+
+config SERIAL_PHYTIUM_PCI
+	tristate "Phytium PCI serial port support"
+	depends on PCI
+	select SERIAL_CORE
+	help
+	  This driver supports the Phytium UART controller on PCI/PCIe adapters.
+	  If you want to compile this driver into the kernel, say Y here.  To
+	  compile this driver as a module, choose M here.
+
+	  If unsure, say N.
+
--- a/drivers/tty/serial/Makefile
+++ b/drivers/tty/serial/Makefile
@@ -100,3 +100,6 @@
 
 obj-$(CONFIG_KGDB_SERIAL_CONSOLE) += kgdboc.o
 obj-$(CONFIG_SERIAL_NUVOTON_MA35D1) += ma35d1_serial.o
+
+obj-$(CONFIG_SERIAL_PHYTIUM_PCI) += phytium-uart.o
+
--- a/drivers/usb/Kconfig
+++ b/drivers/usb/Kconfig
@@ -160,3 +160,6 @@
 source "drivers/usb/roles/Kconfig"
 
 endif # USB_SUPPORT
+
+source "drivers/usb/phytium/Kconfig"
+
--- a/drivers/usb/Makefile
+++ b/drivers/usb/Makefile
@@ -68,3 +68,6 @@
 obj-$(CONFIG_TYPEC)		+= typec/
 
 obj-$(CONFIG_USB_ROLE_SWITCH)	+= roles/
+
+
+obj-$(CONFIG_USB_PHYTIUM)   += phytium/
--- a/drivers/w1/masters/Kconfig
+++ b/drivers/w1/masters/Kconfig
@@ -90,3 +90,14 @@
 
 endmenu
 
+
+config W1_MASTER_PHYTIUM
+	tristate "Phytium 1-wire driver"
+	depends on ARCH_PHYTIUM || COMPILE_TEST
+	help
+	  Say Y here if you want to get support for the 1-wire interface
+	  on an Phytium SoC.
+
+	  This driver can also be built as a module.  If so, the module
+	  will be called phytium-w1.
+
--- a/drivers/w1/masters/Makefile
+++ b/drivers/w1/masters/Makefile
@@ -13,3 +13,6 @@
 obj-$(CONFIG_HDQ_MASTER_OMAP)		+= omap_hdq.o
 obj-$(CONFIG_W1_MASTER_SGI)		+= sgi_w1.o
 obj-$(CONFIG_W1_MASTER_UART)		+= w1-uart.o
+
+
+obj-$(CONFIG_W1_MASTER_PHYTIUM)		+= phytium_w1.o
--- a/sound/hda/controllers/Kconfig
+++ b/sound/hda/controllers/Kconfig
@@ -40,3 +40,20 @@
 
 	  To compile this driver as a module, choose M here: the module
 	  will be called snd-hda-acpi.
+
+config SND_HDA_PHYTIUM
+	tristate "PHYTIUM HD Audio"
+	depends on SOUND
+	select SND_HDA
+	select SND_HDA_ALIGNED_MMIO
+	help
+	  Say Y here to support the HDA controller present in PHYTIUM
+	  SoCs
+
+	  This options enables support for the HD Audio controller
+	  present in some PHYTIUM SoCs, used to communicate audio
+	  to the "High Definition Audio" codec.
+
+	  To compile this driver as a module, choose M here: the module
+	  will be called snd-hda-phytium.
+
--- a/sound/hda/controllers/Makefile
+++ b/sound/hda/controllers/Makefile
@@ -11,3 +11,6 @@
 obj-$(CONFIG_SND_HDA_INTEL) += snd-hda-intel.o
 obj-$(CONFIG_SND_HDA_TEGRA) += snd-hda-tegra.o
 obj-$(CONFIG_SND_HDA_ACPI) += snd-hda-acpi.o
+
+snd-hda-phytium-objs := hda_phytium.o
+obj-$(CONFIG_SND_HDA_PHYTIUM) += snd-hda-phytium.o
--- a/sound/soc/Kconfig
+++ b/sound/soc/Kconfig
@@ -153,3 +153,5 @@
 
 endif	# SND_SOC
 
+
+source "sound/soc/phytium/Kconfig"
--- a/sound/soc/Makefile
+++ b/sound/soc/Makefile
@@ -83,3 +83,5 @@
 obj-$(CONFIG_SND_SOC)	+= xilinx/
 obj-$(CONFIG_SND_SOC)	+= xtensa/
 obj-$(CONFIG_SND_SOC)	+= sdw_utils/
+
+obj-$(CONFIG_SND_SOC)	+= phytium/
