--- a/src/minicom.c
+++ b/src/minicom.c
@@ -1161,7 +1161,7 @@ int main(int argc, char **argv)
       switch(c) {
 	case 'v':
 	  printf(_("%s version %s"), PACKAGE, VERSION);
-#ifdef __DATE__
+#if 0 
 	  printf(_(" (compiled %s)"), __DATE__);
 #endif
 	  printf("\n");
@@ -1464,7 +1464,7 @@ int main(int argc, char **argv)
 
   mc_wprintf(us, "\n%s %s\r\n", _("Welcome to minicom"), VERSION);
   mc_wprintf(us, "\n%s: %s\r\n", _("OPTIONS"), option_string);
-#if defined (__DATE__) && defined (__TIME__)
+#if 0
   mc_wprintf(us, "%s %s, %s.\r\n",_("Compiled on"), __DATE__,__TIME__);
 #endif
   {
