--- a/gettext-runtime/man/Makefile.am
+++ b/gettext-runtime/man/Makefile.am
@@ -176,8 +176,7 @@ textdomain.3.html: textdomain.3.in
 bindtextdomain.3.html: bindtextdomain.3.in
 bind_textdomain_codeset.3.html: bind_textdomain_codeset.3.in
 
-install-html-local:
-	$(MKDIR_P) $(DESTDIR)$(htmldir)
+install-html: installdirs-html
 	for file in $(man_HTML); do \
 	  if test -f $$file; then dir=.; else dir=$(srcdir); fi; \
 	  $(INSTALL_DATA) $$dir/$$file $(DESTDIR)$(htmldir)/$$file; \
--- a/gettext-tools/man/Makefile.am
+++ b/gettext-tools/man/Makefile.am
@@ -158,8 +158,7 @@ recode-sr-latin.1.html: recode-sr-latin.
 gettextize.1.html: gettextize.1
 autopoint.1.html: autopoint.1
 
-install-html-local:
-	$(MKDIR_P) $(DESTDIR)$(htmldir)
+install-html: installdirs-html
 	for file in $(man_HTML); do \
 	  if test -f $$file; then dir=.; else dir=$(srcdir); fi; \
 	  $(INSTALL_DATA) $$dir/$$file $(DESTDIR)$(htmldir)/$$file; \
