--- a/autogen.sh
+++ b/autogen.sh
@@ -81,6 +81,7 @@ if ! $skip_gnulib; then
     getopt-gnu
     gettext-h
     havelib
+    libunistring-optional
     memmove
     noreturn
     progname
--- a/gettext-runtime/src/Makefile.am
+++ b/gettext-runtime/src/Makefile.am
@@ -43,7 +43,7 @@ envsubst_SOURCES = envsubst.c
 
 # Link dependencies.
 # Need @LTLIBICONV@ because striconv.c uses iconv().
-LDADD = ../gnulib-lib/libgrt.a @LTLIBINTL@ @LTLIBICONV@ $(WOE32_LDADD)
+LDADD = ../gnulib-lib/libgrt.a $(LTLIBUNISTRING) @LTLIBINTL@ @LTLIBICONV@ $(WOE32_LDADD)
 
 # Specify installation directory, for --enable-relocatable.
 gettext_CFLAGS = -DINSTALLDIR=\"$(bindir)\"
