--- a/arch/mips/ar7/Platform
+++ b/arch/mips/ar7/Platform
@@ -3,4 +3,9 @@
 #
 platform-$(CONFIG_AR7)		+= ar7/
 cflags-$(CONFIG_AR7)		+= -I$(srctree)/arch/mips/include/asm/mach-ar7
-load-$(CONFIG_AR7)		+= 0xffffffff94100000
+load-$(CONFIG_AR7_TI)		+= 0xffffffff94100000
+
+#
+# AudioCodes AC49x
+#
+load-$(CONFIG_AR7_AC49X)	+= 0xffffffff945ca000
--- a/arch/mips/ar7/setup.c
+++ b/arch/mips/ar7/setup.c
@@ -69,6 +69,10 @@ const char *get_system_type(void)
 			return "TI AR7 (TNETV1056)";
 		case TITAN_CHIP_1060:
 			return "TI AR7 (TNETV1060)";
+		case TITAN_CHIP_AC495:
+			return "AudioCodes AC495";
+		case TITAN_CHIP_AC496:
+			return "AudioCodes AC496";
 		}
 	default:
 		return "TI AR7 (unknown)";
--- a/arch/mips/include/asm/mach-ar7/ar7.h
+++ b/arch/mips/include/asm/mach-ar7/ar7.h
@@ -92,6 +92,8 @@
 #define TITAN_CHIP_1055 0x0e
 #define TITAN_CHIP_1056 0x0d
 #define TITAN_CHIP_1060 0x07
+#define TITAN_CHIP_AC495	0x00
+#define TITAN_CHIP_AC496	0x02
 
 /* Interrupts */
 #define AR7_IRQ_UART0	15
--- a/arch/mips/Kconfig
+++ b/arch/mips/Kconfig
@@ -94,7 +94,7 @@ config AR7
 	select HAVE_CLK
 	help
 	  Support for the Texas Instruments AR7 System-on-a-Chip
-	  family: TNETD7100, 7200 and 7300.
+	  family: TI TNETD7100, 7200, 7300 and AudioCodes AC49x.
 
 config ATH79
 	bool "Atheros AR71XX/AR724X/AR913X based boards"
@@ -835,6 +835,7 @@ config MIPS_PARAVIRT
 endchoice
 
 source "arch/mips/alchemy/Kconfig"
+source "arch/mips/ar7/Kconfig"
 source "arch/mips/ath79/Kconfig"
 source "arch/mips/bcm47xx/Kconfig"
 source "arch/mips/bcm63xx/Kconfig"
--- /dev/null
+++ b/arch/mips/ar7/Kconfig
@@ -0,0 +1,26 @@
+if AR7
+
+config AR7_TI
+	bool
+
+config AR7_AC49X
+	bool
+
+choice
+	prompt "AR7 SoC family selection"
+	default AR7_TYPE_TI
+	depends on AR7
+	help
+	  Select AR7 MIPS SoC implementation.
+
+	config AR7_TYPE_TI
+		bool "Texas Instruments AR7"
+		select AR7_TI
+
+	config AR7_TYPE_AC49X
+		bool "AudioCodes AC49X"
+		select AR7_AC49X
+
+endchoice
+
+endif
