--- a/arch/mips/ar7/prom.c
+++ b/arch/mips/ar7/prom.c
@@ -70,6 +70,7 @@ struct psbl_rec {
 };
 
 static const char psp_env_version[] __initconst = "TIENV0.8";
+static const char psp_env_version_ac49x[] __initconst = "MaxENV0.2";
 
 struct psp_env_chunk {
 	u8	num;
@@ -186,7 +187,8 @@ static void __init ar7_init_env(struct e
 	struct psbl_rec *psbl = (struct psbl_rec *)(KSEG1ADDR(0x14000300));
 	void *psp_env = (void *)KSEG1ADDR(psbl->env_base);
 
-	if (strcmp(psp_env, psp_env_version) == 0) {
+	if (strcmp(psp_env, psp_env_version) == 0 ||
+	    strcmp(psp_env, psp_env_version_ac49x) == 0) {
 		parse_psp_env(psp_env);
 	} else {
 		for (i = 0; i < MAX_ENTRY; i++, env++)
