--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -200,15 +200,7 @@ option(NO_PROTOCHAIN "Disable protochain
 #
 set(PCAP_TYPE "" CACHE STRING "Packet capture type")
 
-#
-# Default to having remote capture support on Windows and, for now, to
-# not having it on UN*X.
-#
-if(WIN32)
-    option(ENABLE_REMOTE "Enable remote capture" ON)
-else()
-    option(ENABLE_REMOTE "Enable remote capture" OFF)
-endif(WIN32)
+option(ENABLE_REMOTE "Enable remote capture" ON)
 
 if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
     option(BUILD_WITH_LIBNL "Build with libnl" ON)
