From: Felix Fietkau <nbd@nbd.name>
Date: Fri, 14 Aug 2020 16:13:45 +0200
Subject: [PATCH] backports: add netif_receive_skb_list

It will be needed by pending mac80211 changes

Signed-off-by: Felix Fietkau <nbd@nbd.name>
---

--- a/backport-include/linux/netdevice.h
+++ b/backport-include/linux/netdevice.h
@@ -372,6 +372,18 @@ static inline int _bp_netdev_upper_dev_l
 	macro_dispatcher(netdev_upper_dev_link, __VA_ARGS__)(__VA_ARGS__)
 #endif
 
+#if LINUX_VERSION_IS_LESS(4,19,0)
+static inline void netif_receive_skb_list(struct list_head *head)
+{
+	struct sk_buff *skb, *next;
+
+	list_for_each_entry_safe(skb, next, head, list) {
+		skb_list_del_init(skb);
+		netif_receive_skb(skb);
+	}
+}
+#endif
+
 #if LINUX_VERSION_IS_LESS(5,0,0)
 static inline int backport_dev_open(struct net_device *dev, struct netlink_ext_ack *extack)
 {
