From: Felix Fietkau <nbd@nbd.name>
Date: Fri, 14 Aug 2020 16:13:55 +0200
Subject: [PATCH] backports: add skb_list_del_init

It will be needed by pending mac80211 changes

Signed-off-by: Felix Fietkau <nbd@nbd.name>
---

--- a/backport-include/linux/skbuff.h
+++ b/backport-include/linux/skbuff.h
@@ -384,6 +384,12 @@ static inline void skb_mark_not_on_list(
 {
 	skb->next = NULL;
 }
+
+static inline void skb_list_del_init(struct sk_buff *skb)
+{
+	__list_del_entry(&skb->list);
+	skb_mark_not_on_list(skb);
+}
 #endif /* 4.19.10 <= x < 4.20 */
 #endif
 
