From: Felix Fietkau <nbd@nbd.name>
Date: Sun, 26 Jul 2020 14:42:58 +0200
Subject: [PATCH] mac80211: calculcate skb hash early when using itxq

This avoids flow separation issues when using software encryption

Signed-off-by: Felix Fietkau <nbd@nbd.name>
---

--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -3937,6 +3937,7 @@ void __ieee80211_subif_start_xmit(struct
 	if (local->ops->wake_tx_queue) {
 		u16 queue = __ieee80211_select_queue(sdata, sta, skb);
 		skb_set_queue_mapping(skb, queue);
+		skb_get_hash(skb);
 	}
 
 	if (sta) {
