diff --git a/net/mac80211/rx.c b/net/mac80211/rx.c
index bd5988e..df6bd18 100644
--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -4774,12 +4774,25 @@ void ieee80211_rx_napi(struct ieee80211_hw *hw, struct ieee80211_sta *pubsta,
 	rcu_read_unlock();
 
 	if (!napi) {
+#if LINUX_VERSION_IS_GEQ(4,19,0)
 		netif_receive_skb_list(&list);
+#else
+		list_for_each_entry_safe(skb, tmp, &list, list) {
+			__list_del_entry(&skb->list);
+			skb->next = NULL;
+			netif_receive_skb(skb);
+		}
+#endif
 		return;
 	}
 
 	list_for_each_entry_safe(skb, tmp, &list, list) {
+#if LINUX_VERSION_IS_GEQ(4,19,0)
 		skb_list_del_init(skb);
+#else
+		__list_del_entry(&skb->list);
+		skb->next = NULL;
+#endif
 		napi_gro_receive(napi, skb);
 	}
 }
