--- a/net/mac80211/ethtool.c
+++ b/net/mac80211/ethtool.c
@@ -14,7 +14,13 @@
 #include "driver-ops.h"
 
 static int ieee80211_set_ringparam(struct net_device *dev,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5,18,0)
 				   struct ethtool_ringparam *rp)
+#else
+				   struct ethtool_ringparam *rp,
+				   struct kernel_ethtool_ringparam *kernel_rp,
+				   struct netlink_ext_ack *extack)
+#endif
 {
 	struct ieee80211_local *local = wiphy_priv(dev->ieee80211_ptr->wiphy);
 
@@ -25,7 +31,13 @@ static int ieee80211_set_ringparam(struc
 }
 
 static void ieee80211_get_ringparam(struct net_device *dev,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5,18,0)
 				    struct ethtool_ringparam *rp)
+#else
+				    struct ethtool_ringparam *rp,
+				    struct kernel_ethtool_ringparam *kernel_rp,
+				    struct netlink_ext_ack *extack)
+#endif
 {
 	struct ieee80211_local *local = wiphy_priv(dev->ieee80211_ptr->wiphy);
 
