--- a/backport-include/linux/of_net.h
+++ b/backport-include/linux/of_net.h
@@ -7,7 +7,7 @@
 /* The behavior of of_get_mac_address() changed in kernel 5.2, it now
  * returns an error code and not NULL in case of an error.
  */
-#if LINUX_VERSION_IS_LESS(5,13,0)
+#if LINUX_VERSION_IS_LESS(5,2,0)
 static inline int backport_of_get_mac_address(struct device_node *np, u8 *mac_out)
 {
 	const void *mac = of_get_mac_address(np);

