--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -49,7 +49,11 @@
 
 	if (present_fcs_len)
 		__pskb_trim(skb, skb->len - present_fcs_len);
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6,1,0)
 	__pskb_pull(skb, rtap_space);
+#else
+	pskb_pull(skb, rtap_space);
+#endif
 
 	hdr = (void *)skb->data;
 	fc = hdr->frame_control;
@@ -74,8 +78,11 @@
 
 	memmove(skb->data + IEEE80211_HT_CTL_LEN, skb->data,
 		hdrlen - IEEE80211_HT_CTL_LEN);
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6,1,0)
 	__pskb_pull(skb, IEEE80211_HT_CTL_LEN);
-
+#else
+	pskb_pull(skb, IEEE80211_HT_CTL_LEN);
+#endif
 	return skb;
 }
 
