--- a/arch/arm/boot/dts/armada-388-helios4.dts
+++ b/arch/arm/boot/dts/armada-388-helios4.dts
@@ -15,6 +15,13 @@
 	model = "Helios4";
 	compatible = "kobol,helios4", "marvell,armada388",
 		"marvell,armada385", "marvell,armada380";
+		
+	aliases {
+		led-boot = &led_status;
+		led-failsafe = &led_status;
+		led-running = &led_status;
+		led-upgrade = &led_status;
+	};
 
 	memory {
 		device_type = "memory";
@@ -73,10 +80,9 @@
 		pinctrl-names = "default";
 		pinctrl-0 = <&helios_system_led_pins>;
 
-		status-led {
+		led_status: status-led {
 			label = "helios4:green:status";
 			gpios = <&gpio0 24 GPIO_ACTIVE_LOW>;
-			linux,default-trigger = "heartbeat";
 			default-state = "on";
 		};
 
