From dcd966fa7ca63f38cf7147e1184d13d66e2ca340 Mon Sep 17 00:00:00 2001
From: OpenWrt community <openwrt-devel@lists.openwrt.org>
Date: Wed, 13 Jul 2022 13:33:30 +0200
Subject: [PATCH] Kconfig: add tristate for OID and ASNI string

---
 init/Kconfig | 2 +-
 lib/Kconfig  | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

--- a/init/Kconfig
+++ b/init/Kconfig
@@ -2002,7 +2002,7 @@ config PADATA
 	bool
 
 config ASN1
-	tristate
+	tristate "ASN1"
 	help
 	  Build a simple ASN.1 grammar compiler that produces a bytecode output
 	  that can be interpreted by the ASN.1 stream decoder and used to
--- a/lib/Kconfig
+++ b/lib/Kconfig
@@ -641,7 +641,7 @@ config LIBFDT
 	bool
 
 config OID_REGISTRY
-	tristate
+	tristate "OID"
 	help
 	  Enable fast lookup object identifier registry.
 
