--- a/drivers/virtio/Kconfig
+++ b/drivers/virtio/Kconfig
@@ -3,7 +3,7 @@ config VIRTIO_ANCHOR
 	bool
 
 config VIRTIO
-	tristate
+	tristate "VIRTIO Support"
 	select VIRTIO_ANCHOR
 	help
 	  This option is selected by any driver which implements the virtio
@@ -11,7 +11,7 @@ config VIRTIO
 	  or CONFIG_S390_GUEST.
 
 config VIRTIO_PCI_LIB
-	tristate
+	tristate "VIRTIO_PCI_LIB Support"
 	help
 	  Modern PCI device implementation. This module implements the
 	  basic probe and control for devices which are based on modern
--- a/crypto/Kconfig
+++ b/crypto/Kconfig
@@ -842,7 +842,7 @@ config CRYPTO_GCM
 	  This is required for IPSec ESP (XFRM_ESP).
 
 config CRYPTO_GENIV
-	tristate
+	tristate "IV Generation for dm-crypt"
 	select CRYPTO_AEAD
 	select CRYPTO_NULL
 	select CRYPTO_MANAGER
