--- a/squashfs-tools/mksquashfs.c
+++ b/squashfs-tools/mksquashfs.c
@@ -32,6 +32,9 @@
 #include <stdio.h>
 #include <stddef.h>
 #include <sys/time.h>
+#if defined(__OpenBSD__)
+#include <sys/param.h>
+#endif
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <fcntl.h>
--- a/squashfs-tools/unsquashfs.h
+++ b/squashfs-tools/unsquashfs.h
@@ -25,6 +25,9 @@
 #define TRUE 1
 #define FALSE 0
 #include <stdio.h>
+#if defined(__OpenBSD__)
+#include <sys/param.h>
+#endif
 #include <sys/types.h>
 #include <unistd.h>
 #include <stdlib.h>
