--- a/source3/lib/messages.c
+++ b/source3/lib/messages.c
@@ -518,7 +518,7 @@ static NTSTATUS messaging_init_internal(
 		return NT_STATUS_ACCESS_DENIED;
 	}
 
-	priv_path = private_path("msg.sock");
+	priv_path = lock_path(talloc_tos(), "msg.sock");
 	if (priv_path == NULL) {
 		return NT_STATUS_NO_MEMORY;
 	}
@@ -662,7 +662,7 @@ NTSTATUS messaging_reinit(struct messagi
 
 	msg_ctx->msg_dgm_ref = messaging_dgm_ref(
 		msg_ctx, msg_ctx->event_ctx, &msg_ctx->id.unique_id,
-		private_path("msg.sock"), lck_path,
+		lock_path(talloc_tos(), "msg.sock"), lck_path,
 		messaging_recv_cb, msg_ctx, &ret);
 
 	if (msg_ctx->msg_dgm_ref == NULL) {
