Don't check xsltproc manpages

Signed-off-by: Bian Naimeng <biannm@cn.fujitsu.com>

Index: samba-4.4.2/lib/ldb/wscript
===================================================================
--- samba-4.4.2.orig/lib/ldb/wscript
+++ samba-4.4.2/lib/ldb/wscript
@@ -65,7 +65,7 @@ def configure(conf):
                 conf.define('USING_SYSTEM_LDB', 1)
 
     if conf.env.standalone_ldb:
-        conf.CHECK_XSLTPROC_MANPAGES()
+        #conf.CHECK_XSLTPROC_MANPAGES()
 
         # we need this for the ldap backend
         if conf.CHECK_FUNCS_IN('ber_flush ldap_open ldap_initialize', 'lber ldap', headers='lber.h ldap.h'):
Index: samba-4.4.2/lib/talloc/wscript
===================================================================
--- samba-4.4.2.orig/lib/talloc/wscript
+++ samba-4.4.2/lib/talloc/wscript
@@ -56,7 +56,7 @@ def configure(conf):
     if conf.env.standalone_talloc:
         conf.env.TALLOC_COMPAT1 = Options.options.TALLOC_COMPAT1
 
-    conf.CHECK_XSLTPROC_MANPAGES()
+    #conf.CHECK_XSLTPROC_MANPAGES()
 
     if not conf.env.disable_python:
         # also disable if we don't have the python libs installed
Index: samba-4.4.2/lib/tdb/wscript
===================================================================
--- samba-4.4.2.orig/lib/tdb/wscript
+++ samba-4.4.2/lib/tdb/wscript
@@ -92,7 +92,7 @@ def configure(conf):
         not conf.env.disable_tdb_mutex_locking):
         conf.define('USE_TDB_MUTEX_LOCKING', 1)
 
-    conf.CHECK_XSLTPROC_MANPAGES()
+    #conf.CHECK_XSLTPROC_MANPAGES()
 
     if not conf.env.disable_python:
         # also disable if we don't have the python libs installed
