--- a/libcli/smbreadline/wscript_configure
+++ b/libcli/smbreadline/wscript_configure
@@ -1,11 +1,13 @@
 #!/usr/bin/env python
 
 
-conf.CHECK_HEADERS('readline.h history.h readline/readline.h readline/history.h')
-for termlib in ['ncurses', 'curses', 'termcap', 'terminfo', 'termlib', 'tinfo']:
-    if conf.CHECK_FUNCS_IN('tgetent', termlib):
-        conf.env['READLINE_TERMLIB'] = termlib
-        break
+#conf.CHECK_HEADERS('readline.h history.h readline/readline.h readline/history.h')
+#for termlib in ['ncurses', 'curses', 'termcap', 'terminfo', 'termlib', 'tinfo']:
+#    if conf.CHECK_FUNCS_IN('tgetent', termlib):
+#        conf.env['READLINE_TERMLIB'] = termlib
+#        break
+
+conf.undefine('HAVE_READLINE_READLINE_H')
 
 #
 # Check if we need to work around readline/readline.h
