--- a/lib/crypto/wscript_configure	2018-07-26
+++ b/lib/crypto/wscript_configure	2018-08-27
@@ -2,11 +2,11 @@
 import Options
 import Utils
 
-if not conf.CHECK_FUNCS_IN('MD5Init', 'bsd', headers='bsd/md5.h',
-    checklibc=True):
-    conf.CHECK_FUNCS_IN('MD5Init', 'md5', headers='sys/md5.h',
-                        checklibc=True)
-    conf.CHECK_FUNCS_IN('MD5Init', 'md', headers='sys/md5.h',
+# if not conf.CHECK_FUNCS_IN('MD5Init', 'bsd', headers='bsd/md5.h',
+    # checklibc=True):
+conf.CHECK_FUNCS_IN('MD5Init', 'md5', headers='sys/md5.h',
+					checklibc=True)
+conf.CHECK_FUNCS_IN('MD5Init', 'md', headers='sys/md5.h',
                         checklibc=True)
 conf.CHECK_FUNCS_IN('CC_MD5_Init', '', headers='CommonCrypto/CommonDigest.h',
     checklibc=True)
--- a/lib/replace/wscript
+++ b/lib/replace/wscript
@@ -302,22 +302,13 @@ def configure(conf):
 
     conf.CHECK_FUNCS('prctl dirname basename')
 
-    strlcpy_in_bsd = False
-
-    # libbsd on some platforms provides strlcpy and strlcat
-    if not conf.CHECK_FUNCS('strlcpy strlcat'):
-        if conf.CHECK_FUNCS_IN('strlcpy strlcat', 'bsd', headers='bsd/string.h',
-                               checklibc=True):
-            strlcpy_in_bsd = True
-    if not conf.CHECK_FUNCS('getpeereid'):
-        conf.CHECK_FUNCS_IN('getpeereid', 'bsd', headers='sys/types.h bsd/unistd.h')
-    if not conf.CHECK_FUNCS_IN('setproctitle', 'setproctitle', headers='setproctitle.h'):
-        conf.CHECK_FUNCS_IN('setproctitle', 'bsd', headers='sys/types.h bsd/unistd.h')
-    if not conf.CHECK_FUNCS('setproctitle_init'):
-        conf.CHECK_FUNCS_IN('setproctitle_init', 'bsd', headers='sys/types.h bsd/unistd.h')
-
-    if not conf.CHECK_FUNCS('closefrom'):
-        conf.CHECK_FUNCS_IN('closefrom', 'bsd', headers='bsd/unistd.h')
+    # Not checking for libbsd
+    conf.CHECK_FUNCS('strlcpy strlcat')
+    conf.CHECK_FUNCS('getpeereid')
+    conf.CHECK_FUNCS_IN('setproctitle', 'setproctitle', headers='setproctitle.h')
+    conf.CHECK_FUNCS('setproctitle_init')
+
+    conf.CHECK_FUNCS('closefrom')
 
     conf.CHECK_CODE('''
                 struct ucred cred;
@@ -667,9 +658,6 @@ removeea setea
 
     # look for a method of finding the list of network interfaces
     for method in ['HAVE_IFACE_GETIFADDRS', 'HAVE_IFACE_AIX', 'HAVE_IFACE_IFCONF', 'HAVE_IFACE_IFREQ']:
-        bsd_for_strlcpy = ''
-        if strlcpy_in_bsd:
-            bsd_for_strlcpy = ' bsd'
         if conf.CHECK_CODE('''
                            #define %s 1
                            #define NO_CONFIG_H 1
@@ -682,7 +670,7 @@ removeea setea
                            #include "test/getifaddrs.c"
                            ''' % method,
                            method,
-                           lib='nsl socket' + bsd_for_strlcpy,
+                           lib='nsl socket',
                            addmain=False,
                            execute=True):
             break
@@ -730,7 +718,6 @@ def build(bld):
                 break
 
     extra_libs = ''
-    if bld.CONFIG_SET('HAVE_LIBBSD'): extra_libs += ' bsd'
 
     bld.SAMBA_SUBSYSTEM('LIBREPLACE_HOSTCC',
         REPLACE_HOSTCC_SOURCE,
