--- a/source3/modules/vfs_fruit.c	2019-07-09
+++ b/source3/modules/vfs_fruit.c	2019-07-09
@@ -6995,12 +6995,12 @@ static bool fruit_tmsize_do_dirent(vfs_h
 		return true;
 	}
 
-	if (bandsize > SIZE_MAX/nbands) {
-		DBG_ERR("tmsize overflow: bandsize [%zu] nbands [%zu]\n",
-			bandsize, nbands);
-		return false;
-	}
-	tm_size = bandsize * nbands;
+	// if (bandsize > SIZE_MAX/nbands) {
+		// DBG_ERR("tmsize overflow: bandsize [%zu] nbands [%zu]\n",
+			// bandsize, nbands);
+		// return false;
+	// }
+	tm_size = (off_t)bandsize * (off_t)nbands;
 
 	if (state->total_size + tm_size < state->total_size) {
 		DBG_ERR("tmsize overflow: bandsize [%zu] nbands [%zu]\n",
