From 9d4a6615dec47638f73d5f5b84120f210dcb1ba2 Mon Sep 17 00:00:00 2001
From: YooLc <lchxdev@gmail.com>
Date: Sun, 10 Aug 2025 14:41:56 +0800
Subject: [PATCH] PCI: dwc: Add a device tree property to skip PCIe version

---
 drivers/pci/controller/dwc/pcie-designware.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/pci/controller/dwc/pcie-designware.c b/drivers/pci/controller/dwc/pcie-designware.c
index 6d6cbc8b5..84d21aa12 100644
--- a/drivers/pci/controller/dwc/pcie-designware.c
+++ b/drivers/pci/controller/dwc/pcie-designware.c
@@ -183,6 +183,15 @@ void dw_pcie_version_detect(struct dw_pcie *pci)
 {
 	u32 ver;
 
+	/*
+	 * Some boards with Amlogic A311D SoC's AHCI controller breaks
+	 * if the version register is read.
+	 * Skip detection for some boards marked with skip-version-detect
+	 * in the device tree.
+	 */
+	if (of_property_read_bool(pci->dev->of_node, "skip-version-detect"))
+		return;
+
 	/* The content of the CSR is zero on DWC PCIe older than v4.70a */
 	ver = dw_pcie_readl_dbi(pci, PCIE_VERSION_NUMBER);
 	if (!ver)
-- 
2.50.1

