--- a/drivers/net/ethernet/stmicro/stmmac/dwmac-meson8b.c
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac-meson8b.c
@@ -427,6 +427,8 @@
 	if (IS_ERR(plat_dat))
 		return PTR_ERR(plat_dat);
 
+	plat_dat->flags |= STMMAC_FLAG_NO_HW_VLAN_FILTER;
+
 	dwmac = devm_kzalloc(&pdev->dev, sizeof(*dwmac), GFP_KERNEL);
 	if (!dwmac)
 		return -ENOMEM;
--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
@@ -6670,6 +6670,9 @@
 
 static void stmmac_vlan_restore(struct stmmac_priv *priv)
 {
+	if (priv->plat->flags & STMMAC_FLAG_NO_HW_VLAN_FILTER)
+		return;
+
 	if (!(priv->dev->features & NETIF_F_VLAN_FEATURES))
 		return;
 
@@ -7582,17 +7585,19 @@
 	ndev->watchdog_timeo = msecs_to_jiffies(watchdog);
 #ifdef STMMAC_VLAN_TAG_USED
 	/* Both mac100 and gmac support receive VLAN tag detection */
-	ndev->features |= NETIF_F_HW_VLAN_CTAG_RX | NETIF_F_HW_VLAN_STAG_RX;
-	if (dwmac_is_xmac(priv->plat->core_type)) {
-		ndev->hw_features |= NETIF_F_HW_VLAN_CTAG_RX;
-		priv->hw->hw_vlan_en = true;
-	}
-	if (priv->dma_cap.vlhash) {
-		ndev->features |= NETIF_F_HW_VLAN_CTAG_FILTER;
-		ndev->features |= NETIF_F_HW_VLAN_STAG_FILTER;
+	if (!(priv->plat->flags & STMMAC_FLAG_NO_HW_VLAN_FILTER)) {
+		ndev->features |= NETIF_F_HW_VLAN_CTAG_RX | NETIF_F_HW_VLAN_STAG_RX;
+		if (dwmac_is_xmac(priv->plat->core_type)) {
+			ndev->hw_features |= NETIF_F_HW_VLAN_CTAG_RX;
+			priv->hw->hw_vlan_en = true;
+		}
+		if (priv->dma_cap.vlhash) {
+			ndev->features |= NETIF_F_HW_VLAN_CTAG_FILTER;
+			ndev->features |= NETIF_F_HW_VLAN_STAG_FILTER;
+		}
+		if (priv->dma_cap.vlins)
+			ndev->features |= NETIF_F_HW_VLAN_CTAG_TX;
 	}
-	if (priv->dma_cap.vlins)
-		ndev->features |= NETIF_F_HW_VLAN_CTAG_TX;
 #endif
 	priv->msg_enable = netif_msg_init(debug, default_msg_level);
 
--- a/include/linux/stmmac.h
+++ b/include/linux/stmmac.h
@@ -191,6 +191,7 @@
 #define STMMAC_FLAG_EN_TX_LPI_CLOCKGATING	BIT(11)
 #define STMMAC_FLAG_EN_TX_LPI_CLK_PHY_CAP	BIT(12)
 #define STMMAC_FLAG_HWTSTAMP_CORRECT_LATENCY	BIT(13)
+#define STMMAC_FLAG_NO_HW_VLAN_FILTER		BIT(14)
 
 struct plat_stmmacenet_data {
 	enum dwmac_core_type core_type;
