--- a/drivers/net/wireless/virtual/mac80211_hwsim.c
+++ b/drivers/net/wireless/virtual/mac80211_hwsim.c
@@ -6402,6 +6402,7 @@
 
 static int init_vqs(struct virtio_device *vdev)
 {
+#if LINUX_VERSION_IS_LESS(6,11,0)
 	vq_callback_t *callbacks[HWSIM_NUM_VQS] = {
 		[HWSIM_VQ_TX] = hwsim_virtio_tx_done,
 		[HWSIM_VQ_RX] = hwsim_virtio_rx_done,
@@ -6413,6 +6414,15 @@
 
 	return virtio_find_vqs(vdev, HWSIM_NUM_VQS,
 			       hwsim_vqs, callbacks, names, NULL);
+#else
+	struct virtqueue_info vqs_info[HWSIM_NUM_VQS] = {
+		[HWSIM_VQ_TX] = { "tx", hwsim_virtio_tx_done },
+		[HWSIM_VQ_RX] = { "rx", hwsim_virtio_rx_done },
+	};
+	
+	return virtio_find_vqs(vdev, HWSIM_NUM_VQS,
+			       hwsim_vqs, vqs_info, NULL);
+#endif
 }
 
 static int fill_vq(struct virtqueue *vq)
