--- a/ip/ipxfrm.c
+++ b/ip/ipxfrm.c
@@ -470,7 +470,6 @@ void xfrm_selector_print(struct xfrm_sel
 	switch (sel->proto) {
 	case IPPROTO_TCP:
 	case IPPROTO_UDP:
-	case IPPROTO_SCTP:
 	case IPPROTO_DCCP:
 	default: /* XXX */
 		if (sel->sport_mask)
@@ -1341,7 +1340,6 @@ static int xfrm_selector_upspec_parse(st
 		switch (sel->proto) {
 		case IPPROTO_TCP:
 		case IPPROTO_UDP:
-		case IPPROTO_SCTP:
 		case IPPROTO_DCCP:
 		case IPPROTO_IP: /* to allow shared SA for different protocols */
 			break;
