--- a/configure
+++ b/configure
@@ -291,7 +291,7 @@ int main(int argc, char **argv) {
 }
 EOF
 
-    $CC -o $TMPDIR/libbpf_test $TMPDIR/libbpf_test.c $LIBBPF_CFLAGS $LIBBPF_LDLIBS >/dev/null 2>&1
+    $CC -o $TMPDIR/libbpf_test $TMPDIR/libbpf_test.c $LIBBPF_CFLAGS $LIBBPF_LDLIBS $LDFLAGS >/dev/null 2>&1
     local ret=$?
 
     rm -f $TMPDIR/libbpf_test.c $TMPDIR/libbpf_test
@@ -309,7 +309,7 @@ int main(int argc, char **argv) {
 }
 EOF
 
-    $CC -o $TMPDIR/libbpf_sec_test $TMPDIR/libbpf_sec_test.c $LIBBPF_CFLAGS $LIBBPF_LDLIBS >/dev/null 2>&1
+    $CC -o $TMPDIR/libbpf_sec_test $TMPDIR/libbpf_sec_test.c $LIBBPF_CFLAGS $LIBBPF_LDLIBS $LDFLAGS >/dev/null 2>&1
     local ret=$?
 
     rm -f $TMPDIR/libbpf_sec_test.c $TMPDIR/libbpf_sec_test
