--- a/src/Mayaqua/Internat.c
+++ b/src/Mayaqua/Internat.c
@@ -123,7 +123,7 @@
 #include <Mayaqua/Mayaqua.h>
 
 extern LOCK *token_lock;
-static char charset[MAX_SIZE] = "EUCJP";
+static char charset[MAX_SIZE] = "utf-8";
 static LOCK *iconv_lock = NULL;
 void *iconv_cache_wide_to_str = 0;
 void *iconv_cache_str_to_wide = 0;
@@ -938,7 +938,7 @@ void InitInternational()
 #ifdef	UNIX_MACOS
 		StrCpy(charset, sizeof(charset), "utf-8");
 #else	// UNIX_MACOS
-		StrCpy(charset, sizeof(charset), "EUCJP");
+		StrCpy(charset, sizeof(charset), "utf-8");
 #endif	// UNIX_MACOS
 		d = IconvWideToStrInternal();
 		if (d == (void *)-1)
@@ -1198,7 +1198,7 @@ void GetCurrentCharSet(char *name, UINT
 		}
 		else
 		{
-			StrCpy(name, size, "eucJP");
+			StrCpy(name, size, "utf-8");
 		}
 	}
 	FreeToken(t);
