--- a/arch/arm64/boot/dts/rockchip/rk3588-nanopc-t6.dtsi
+++ b/arch/arm64/boot/dts/rockchip/rk3588-nanopc-t6.dtsi
@@ -20,6 +20,10 @@
 	aliases {
 		mmc0 = &sdhci;
 		mmc1 = &sdmmc;
+		led-boot = &sys_led;
+		led-failsafe = &sys_led;
+		led-running = &sys_led;
+		led-upgrade = &sys_led;
 	};
 
 	adc-keys-0 {
@@ -53,7 +57,7 @@
 		sys_led: led-0 {
 			gpios = <&gpio2 RK_PB7 GPIO_ACTIVE_HIGH>;
 			label = "system-led";
-			linux,default-trigger = "heartbeat";
+			default-state = "on";
 			pinctrl-names = "default";
 			pinctrl-0 = <&sys_led_pin>;
 		};
