From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tianling Shen <cnsztl@gmail.com>
Date: Mon Sep 23 13:22:56 2024 +0800
Subject: [PATCH] arm64: dts: rockchip: Update LED properties for ArmSom
 Sige7

Add OpenWrt's LED aliases for showing system status.

Signed-off-by: Tianling Shen <cnsztl@gmail.com>
---

--- a/arch/arm64/boot/dts/rockchip/rk3588-armsom-sige7.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3588-armsom-sige7.dts
@@ -14,6 +14,11 @@
 	aliases {
 		mmc0 = &sdhci;
 		mmc1 = &sdmmc;
+
+		led-boot = &led_red;
+		led-failsafe = &led_red;
+		led-running = &led_red;
+		led-upgrade = &led_red;
 	};
 
 	chosen {
