From 8bb576d8640fdf896650a4d4a1b2e60254d75eb2 Mon Sep 17 00:00:00 2001
From: Samuel Holland <samuel@sholland.org>
Date: Sun, 6 Jun 2021 10:56:25 -0500
Subject: [PATCH 062/117] pwm: sun8i-v536: Add support for the Allwinner D1

Signed-off-by: Samuel Holland <samuel@sholland.org>
---
 drivers/pwm/pwm-sun8i-v536.c | 7 +++++++
 1 file changed, 7 insertions(+)

--- a/drivers/pwm/pwm-sun8i-v536.c
+++ b/drivers/pwm/pwm-sun8i-v536.c
@@ -285,6 +285,10 @@ static const struct sun8i_pwm_data sun8i
 	.npwm = 9,
 };
 
+static const struct sun8i_pwm_data sun20i_pwm_data_c8 = {
+	.npwm = 8,
+};
+
 static const struct sun8i_pwm_data sun50i_pwm_data_c16 = {
 	.npwm = 16,
 };
@@ -294,6 +298,9 @@ static const struct of_device_id sun8i_p
 		.compatible = "allwinner,sun8i-v536-pwm",
 		.data = &sun8i_pwm_data_c9,
 	}, {
+		.compatible = "allwinner,sun20i-d1-pwm",
+		.data = &sun20i_pwm_data_c8,
+	}, {
 		.compatible = "allwinner,sun50i-r818-pwm",
 		.data = &sun50i_pwm_data_c16,
 	}, {
