From c2b3f2c723e1b558afe5661bb91669e3b68154f7 Mon Sep 17 00:00:00 2001
From: Samuel Holland <samuel@sholland.org>
Date: Sun, 13 Jun 2021 23:52:47 -0500
Subject: [PATCH 073/117] ASoC: sun4i-spdif: Add support for separate resets

Signed-off-by: Samuel Holland <samuel@sholland.org>
---
 sound/soc/sunxi/sun4i-spdif.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

--- a/sound/soc/sunxi/sun4i-spdif.c
+++ b/sound/soc/sunxi/sun4i-spdif.c
@@ -28,10 +28,11 @@
 #include <sound/soc.h>
 
 #define	SUN4I_SPDIF_CTL		(0x00)
+	#define SUN4I_SPDIF_CTL_RST_RX			BIT(12)
 	#define SUN4I_SPDIF_CTL_MCLKDIV(v)		((v) << 4) /* v even */
 	#define SUN4I_SPDIF_CTL_MCLKOUTEN		BIT(2)
 	#define SUN4I_SPDIF_CTL_GEN			BIT(1)
-	#define SUN4I_SPDIF_CTL_RESET			BIT(0)
+	#define SUN4I_SPDIF_CTL_RST_TX			BIT(0)
 
 #define SUN4I_SPDIF_TXCFG	(0x04)
 	#define SUN4I_SPDIF_TXCFG_SINGLEMOD		BIT(31)
@@ -196,7 +197,7 @@ static void sun4i_spdif_configure(struct
 	const struct sun4i_spdif_quirks *quirks = host->quirks;
 
 	/* soft reset SPDIF */
-	regmap_write(host->regmap, SUN4I_SPDIF_CTL, SUN4I_SPDIF_CTL_RESET);
+	regmap_write(host->regmap, SUN4I_SPDIF_CTL, SUN4I_SPDIF_CTL_RST_TX);
 
 	/* flush TX FIFO */
 	regmap_update_bits(host->regmap, SUN4I_SPDIF_FCTL,
