From 87a77f803f5038e3fc64f45d5142ea402512029a Mon Sep 17 00:00:00 2001
From: Samuel Holland <samuel@sholland.org>
Date: Wed, 23 Jun 2021 21:18:47 -0500
Subject: [PATCH 082/117] ASoC: sun20i-codec: What is this ramp thing?

Signed-off-by: Samuel Holland <samuel@sholland.org>
---
 sound/soc/sunxi/sun20i-codec.c | 4 ++++
 1 file changed, 4 insertions(+)

--- a/sound/soc/sunxi/sun20i-codec.c
+++ b/sound/soc/sunxi/sun20i-codec.c
@@ -709,6 +709,10 @@ static int sun20i_codec_component_probe(
 				      0x3 << SUN20I_CODEC_ADC_DIG_CTRL_ADC_VOL_EN,
 				      0x3 << SUN20I_CODEC_ADC_DIG_CTRL_ADC_VOL_EN);
 
+	/* Maaagic... */
+	snd_soc_component_update_bits(component, SUN20I_CODEC_RAMP,
+				      BIT(1) | BIT(0), BIT(0));
+
 	return 0;
 
 err_assert_reset:
