From 03dbb926f6d65f75af902e421c44aeaaf84be66a Mon Sep 17 00:00:00 2001
From: Samuel Holland <samuel@sholland.org>
Date: Thu, 11 Aug 2022 22:46:28 -0500
Subject: [PATCH 089/117] riscv: dts: allwinner: lichee-rv-86-panel-480p: Add
 panel

Signed-off-by: Samuel Holland <samuel@sholland.org>
---
 .../sun20i-d1-lichee-rv-86-panel-480p.dts     | 51 +++++++++++++++++++
 1 file changed, 51 insertions(+)

--- a/arch/riscv/boot/dts/allwinner/sun20i-d1-lichee-rv-86-panel-480p.dts
+++ b/arch/riscv/boot/dts/allwinner/sun20i-d1-lichee-rv-86-panel-480p.dts
@@ -7,6 +7,40 @@
 	model = "Sipeed Lichee RV 86 Panel (480p)";
 	compatible = "sipeed,lichee-rv-86-panel-480p", "sipeed,lichee-rv",
 		     "allwinner,sun20i-d1";
+
+	backlight: backlight {
+		compatible = "pwm-backlight";
+		power-supply = <&reg_vcc>;
+		pwms = <&pwm 7 50000 0>;
+	};
+
+	spi {
+		compatible = "spi-gpio";
+		cs-gpios = <&pio 4 14 GPIO_ACTIVE_LOW>; /* PE14 */
+		mosi-gpios = <&pio 4 12 GPIO_ACTIVE_HIGH>; /* PE12 */
+		sck-gpios = <&pio 4 15 GPIO_ACTIVE_HIGH>; /* PE15 */
+		num-chipselects = <1>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		panel@0 {
+			compatible = "sitronix,st7701s";
+			reg = <0>;
+			backlight = <&backlight>;
+			reset-gpios = <&pio 6 13 GPIO_ACTIVE_LOW>; /* PG13 */
+			spi-3wire;
+
+			port {
+				panel_in_tcon_lcd0: endpoint {
+					remote-endpoint = <&tcon_lcd0_out_panel>;
+				};
+			};
+		};
+	};
+};
+
+&de {
+	status = "okay";
 };
 
 &i2c2 {
@@ -27,3 +61,20 @@
 		wakeup-source;
 	};
 };
+
+&pwm {
+	pinctrl-0 = <&pwm7_pd22_pin>;
+	pinctrl-names = "default";
+	status = "okay";
+};
+
+&tcon_lcd0 {
+	pinctrl-0 = <&lcd_rgb666_pins>;
+	pinctrl-names = "default";
+};
+
+&tcon_lcd0_out {
+	tcon_lcd0_out_panel: endpoint {
+		remote-endpoint = <&panel_in_tcon_lcd0>;
+	};
+};
