From 1a4ffe62c729f9c4f63cdfd26151e3724bc23f86 Mon Sep 17 00:00:00 2001
From: Moritz Warning <moritzwarning@web.de>
Date: Wed, 2 May 2018 16:06:46 +0200
Subject: [PATCH 4/8] accept external linker flags

---
 make-linux.mk | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/make-linux.mk
+++ b/make-linux.mk
@@ -73,7 +73,7 @@ else
 	override CFLAGS+=-Wall -Wno-deprecated -pthread $(INCLUDES) -DNDEBUG $(DEFS)
 	CXXFLAGS?=-O3 -fstack-protector
 	override CXXFLAGS+=-Wall -Wno-deprecated -std=c++11 -pthread $(INCLUDES) -DNDEBUG $(DEFS)
-	LDFLAGS=-Wl,-z,relro,-z,now
+	LDFLAGS+=-Wl,-z,relro,-z,now
 	STRIP?=strip
 	STRIP+=--strip-all
 endif
