--- a/configure.ac
+++ b/configure.ac
@@ -813,10 +813,12 @@ AC_ARG_WITH(libpulse,
             LDFLAGS="-L$with_libpulse/lib $LDFLAGS"
             CPPFLAGS="-I$with_libpulse/include $CPPFLAGS"
             AC_CHECK_HEADERS(pulse/simple.h,
-                             AC_CHECK_LIB([pulse],
-                                          [pa_stream_peek],
-                                          EXT_LIB_PATH="-L$with_libpulse/lib $EXT_LIB_PATH"
-                                          pulse=1))
+                             AC_CHECK_DECL([pa_stream_peek],
+                                       [pulse=1],
+                                       [],
+                                       [[
+                                       #include <pulse/stream.h>
+                                       ]]))
         ])
     ],
     [AC_MSG_RESULT([--with-libpulse not specified])
@@ -856,10 +858,12 @@ AC_ARG_WITH(libopus,
         LDFLAGS="-L$with_libopus/lib $LDFLAGS"
         CPPFLAGS="-I$with_libopus/include $CPPFLAGS"
         AC_CHECK_HEADERS(opus/opus.h,
-                         AC_CHECK_LIB([opus],
-                                      [OPUS_SET_GAIN],
-                                      EXT_LIB_PATH="-L$with_libopus/lib $EXT_LIB_PATH"
-                                      opus=1))
+                         AC_CHECK_DECL([OPUS_SET_GAIN],
+                                       [opus=1],
+                                       [],
+                                       [[
+                                       #include <opus/opus.h>
+                                       ]]))
       ])
     ],
     [AC_MSG_RESULT([--with-libopus not specified])
