--- a/configure.ac
+++ b/configure.ac
@@ -114,7 +114,6 @@ dnl###################################
 
 AC_CONFIG_FILES([
 Makefile
-strpool/Makefile
 libatasmart.pc
 ])
 AC_OUTPUT
--- a/Makefile.am
+++ b/Makefile.am
@@ -22,7 +22,6 @@ AM_LDFLAGS = $(GCLDFLAGS)
 dist_doc_DATA = README
 
 EXTRA_DIST = \
-	autogen.sh \
 	LGPL \
 	README \
 	atasmart.c \
@@ -47,9 +46,6 @@ EXTRA_DIST = \
 	blob-examples/WDC_WD5000AAKS--00TMA0-12.01C01 \
 	vala/atasmart.vapi
 
-# build the strpool tool first
-SUBDIRS = strpool .
-
 CLEANFILES = atasmart.strpool.c
 
 MAINTAINERCLEANFILES =
@@ -94,10 +90,7 @@ libatasmart_la_CFLAGS = \
 BUILT_SOURCES = \
 	atasmart.strpool.c
 
-strpool/strpool:
-	$(MAKE) -C strpool strpool
-
-atasmart.strpool.c: atasmart.c strpool/strpool
-	$(top_builddir)/strpool/strpool $< $@
+atasmart.strpool.c: atasmart.c
+	$(STAGING_DIR_HOSTPKG)/bin/strpool $< $@
 
 ACLOCAL_AMFLAGS = -I m4
