--- a/include/configs/clearfog.h
+++ b/include/configs/clearfog.h
@@ -107,7 +107,18 @@
 /* Keep device tree and initrd in lower memory so the kernel can access them */
 #define CONFIG_EXTRA_ENV_SETTINGS	\
 	"fdt_high=0x10000000\0"		\
-	"initrd_high=0x10000000\0"
+	"initrd_high=0x10000000\0"	\
+	"script=boot.scr\0"		\
+	"loadaddr=0x01000000\0"		\
+	"mmcdev=0\0"			\
+	"mmcpart=1\0"			\
+	"loadbootscript=fatload mmc ${mmcdev}:${mmcpart} ${loadaddr} ${script};\0" \
+	"bootscript=echo Running bootscript from mmc ...; source ${loadaddr}\0"
+
+#define CONFIG_BOOTCOMMAND		\
+	"if run loadbootscript; then "	\
+		"run bootscript; "	\
+	"fi"
 
 /* SPL */
 /*
--- a/configs/clearfog_defconfig
+++ b/configs/clearfog_defconfig
@@ -3,6 +3,7 @@ CONFIG_ARCH_MVEBU=y
 CONFIG_SYS_MALLOC_F_LEN=0x2000
 CONFIG_TARGET_CLEARFOG=y
 CONFIG_DEFAULT_DEVICE_TREE="armada-388-clearfog"
+CONFIG_HUSH_PARSER=y
 CONFIG_SPL=y
 # CONFIG_CMD_IMLS is not set
 # CONFIG_CMD_FLASH is not set
