--- a/libelf/libelfP.h
+++ b/libelf/libelfP.h
@@ -39,6 +39,9 @@
 #include <stdio.h>
 #include <string.h>
 
+#ifdef _ /* fix libintl-stub */
+#undef _
+#endif
 /* gettext helper macros.  */
 #define _(Str) dgettext ("elfutils", Str)
 
--- a/libdw/libdwP.h
+++ b/libdw/libdwP.h
@@ -35,7 +35,9 @@
 #include <libdw.h>
 #include <dwarf.h>
 
-
+#ifdef _ /* fix libintl-stub */
+#undef _
+#endif
 /* gettext helper macros.  */
 #define _(Str) dgettext ("elfutils", Str)
 
--- a/libdwfl/libdwflP.h
+++ b/libdwfl/libdwflP.h
@@ -43,6 +43,9 @@
 
 typedef struct Dwfl_Process Dwfl_Process;
 
+#ifdef _ /* fix libintl-stub */
+#undef _
+#endif
 /* gettext helper macros.  */
 #define _(Str) dgettext ("elfutils", Str)
 
--- a/libasm/libasmP.h
+++ b/libasm/libasmP.h
@@ -35,6 +35,9 @@
 
 #include "libdwelf.h"
 
+#ifdef _ /* fix libintl-stub */
+#undef _
+#endif
 /* gettext helper macros.  */
 #define _(Str) dgettext ("elfutils", Str)
 
