--- a/libebl/libebl.h
+++ b/libebl/libebl.h
@@ -245,7 +245,7 @@ extern ssize_t ebl_register_info (Ebl *e
    Each of these is filled with the DWARF register number corresponding,
    or -1 if there is none.  Returns zero when the information is available.  */
 extern int ebl_syscall_abi (Ebl *ebl, int *sp, int *pc,
-			    int *callno, int args[6]);
+			    int *callno, int args[]);
 
 /* Supply the ABI-specified state of DWARF CFI before CIE initial programs.
 
--- a/libdw/libdw.h
+++ b/libdw/libdw.h
@@ -1065,7 +1065,7 @@ extern int dwarf_frame_cfa (Dwarf_Frame
    expressions in the CFI, *OPS is an internal pointer that can be used as
    long as the Dwarf_CFI used to create FRAME remains alive.  */
 extern int dwarf_frame_register (Dwarf_Frame *frame, int regno,
-				 Dwarf_Op ops_mem[3],
+				 Dwarf_Op ops_mem[],
 				 Dwarf_Op **ops, size_t *nops)
   __nonnull_attribute__ (3, 4, 5);
 
