--- a/drivers/net/wireless/ralink/rt2x00/rt2x00usb.c
+++ b/drivers/net/wireless/ralink/rt2x00/rt2x00usb.c
@@ -586,10 +586,18 @@ static void rt2x00usb_assign_endpoint(st
 
 	if (queue->qid == QID_RX) {
 		pipe = usb_rcvbulkpipe(usb_dev, queue->usb_endpoint);
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5,19,0)
 		queue->usb_maxpacket = usb_maxpacket(usb_dev, pipe, 0);
+#else
+		queue->usb_maxpacket = usb_maxpacket(usb_dev, pipe);
+#endif
 	} else {
 		pipe = usb_sndbulkpipe(usb_dev, queue->usb_endpoint);
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5,19,0)
 		queue->usb_maxpacket = usb_maxpacket(usb_dev, pipe, 1);
+#else
+		queue->usb_maxpacket = usb_maxpacket(usb_dev, pipe);
+#endif
 	}
 
 	if (!queue->usb_maxpacket)
