--- a/drivers/pci/probe.c
+++ b/drivers/pci/probe.c
@@ -2619,7 +2619,8 @@ static void pcie_write_mrrs(struct pci_d
 	 * In the "safe" case, do not configure the MRRS.  There appear to be
 	 * issues with setting MRRS to 0 on a number of devices.
 	 */
-	if (pcie_bus_config != PCIE_BUS_PERFORMANCE)
+	if (pcie_bus_config != PCIE_BUS_PERFORMANCE &&
+	    pcie_bus_config != PCIE_BUS_PEER2PEER)
 		return;
 
 	/*
--- a/include/linux/pci.h
+++ b/include/linux/pci.h
@@ -866,7 +866,7 @@ enum pcie_bus_config_types {
 	PCIE_BUS_DEFAULT,	/* Ensure MPS matches upstream bridge */
 	PCIE_BUS_SAFE,		/* Use largest MPS boot-time devices support */
 	PCIE_BUS_PERFORMANCE,	/* Use MPS and MRRS for best performance */
-	PCIE_BUS_PEER2PEER,	/* Set MPS = 128 for all devices */
+	PCIE_BUS_PEER2PEER,	/* set MPS and MRSS to 128 for all devices */
 };
 
 extern enum pcie_bus_config_types pcie_bus_config;
