--- a/drivers/net/phy/broadcom.c
+++ b/drivers/net/phy/broadcom.c
@@ -429,7 +429,11 @@ static int bcm5481_config_aneg(struct ph
 	ret = genphy_config_aneg(phydev);
 
 	/* Then we can set up the delay. */
-	bcm5481x_config(phydev);
+	//bcm5481x_config(phydev);
+
+	phy_write(phydev, 0x18, 0xf1e7);
+	phy_write(phydev, 0x1c, 0x8e00);
+	phy_write(phydev, 0x1c, 0xa41f);
 
 	if (of_property_read_bool(np, "enet-phy-lane-swap")) {
 		/* Lane Swap - Undocumented register...magic! */
