--- a/src/subshell/common.c
+++ b/src/subshell/common.c
@@ -1128,7 +1128,7 @@ init_subshell_precmd (char *precmd, size
                     "else "
                     "[ \"${PWD##$HOME/}\" = \"$PWD\" ] && MC_PWD=\"$PWD\" || MC_PWD=\"~/${PWD##$HOME/}\"; "
                     "fi; "
-                    "echo \"$USER@$(hostname -s):$MC_PWD\"; "
+                    "echo \"$USER@$HOSTNAME:$MC_PWD\"; "
                     "pwd>&%d; "
                     "kill -STOP $$; "
                     "}; " "PRECMD=precmd; " "PS1='$($PRECMD)$ '\n", subshell_pipe[WRITE]);
