From: Felix Fietkau <nbd@nbd.name>
Subject: jffs2: use .rename2 and add RENAME_WHITEOUT support

It is required for renames on overlayfs

Signed-off-by: Felix Fietkau <nbd@nbd.name>
---

--- a/fs/jffs2/dir.c
+++ b/fs/jffs2/dir.c
@@ -609,7 +609,8 @@ static int jffs2_rmdir (struct inode *di
 	return ret;
 }
 
-static int jffs2_mknod (struct inode *dir_i, struct dentry *dentry, umode_t mode, dev_t rdev)
+static int __jffs2_mknod (struct inode *dir_i, struct dentry *dentry,
+			  umode_t mode, dev_t rdev, bool whiteout)
 {
 	struct jffs2_inode_info *f, *dir_f;
 	struct jffs2_sb_info *c;
@@ -748,7 +749,11 @@ static int jffs2_mknod (struct inode *di
 	mutex_unlock(&dir_f->sem);
 	jffs2_complete_reservation(c);
 
-	d_instantiate_new(dentry, inode);
+	if (!whiteout)
+		d_instantiate_new(dentry, inode);
+	else
+		unlock_new_inode(inode);
+
 	return 0;
 
  fail:
@@ -756,6 +761,17 @@ static int jffs2_mknod (struct inode *di
 	return ret;
 }
 
+static int jffs2_mknod (struct inode *dir_i, struct dentry *dentry, umode_t mode, dev_t rdev)
+{
+	return __jffs2_mknod(dir_i, dentry, mode, rdev, false);
+}
+
+static int jffs2_whiteout (struct inode *old_dir, struct dentry *old_dentry)
+{
+	return __jffs2_mknod(old_dir, old_dentry, S_IFCHR | WHITEOUT_MODE,
+			     WHITEOUT_DEV, true);
+}
+
 static int jffs2_rename (struct inode *old_dir_i, struct dentry *old_dentry,
 			 struct inode *new_dir_i, struct dentry *new_dentry,
 			 unsigned int flags)
@@ -766,7 +782,7 @@ static int jffs2_rename (struct inode *o
 	uint8_t type;
 	uint32_t now;
 
-	if (flags & ~RENAME_NOREPLACE)
+	if (flags & ~(RENAME_NOREPLACE|RENAME_WHITEOUT))
 		return -EINVAL;
 
 	/* The VFS will check for us and prevent trying to rename a
@@ -832,9 +848,14 @@ static int jffs2_rename (struct inode *o
 	if (d_is_dir(old_dentry) && !victim_f)
 		inc_nlink(new_dir_i);
 
-	/* Unlink the original */
-	ret = jffs2_do_unlink(c, JFFS2_INODE_INFO(old_dir_i),
-			      old_dentry->d_name.name, old_dentry->d_name.len, NULL, now);
+	if (flags & RENAME_WHITEOUT)
+		/* Replace with whiteout */
+		ret = jffs2_whiteout(old_dir_i, old_dentry);
+	else
+		/* Unlink the original */
+		ret = jffs2_do_unlink(c, JFFS2_INODE_INFO(old_dir_i),
+				      old_dentry->d_name.name,
+				      old_dentry->d_name.len, NULL, now);
 
 	/* We don't touch inode->i_nlink */
 
