From: Felix Fietkau <nbd@nbd.name>
Subject: Disable software protection bits for Macronix flashes.

Signed-off-by: Felix Fietkau <nbd@nbd.name>
---
 drivers/mtd/spi-nor/spi-nor.c | 1 +
 1 file changed, 1 insertion(+)

--- a/drivers/mtd/spi-nor/macronix.c
+++ b/drivers/mtd/spi-nor/macronix.c
@@ -93,6 +93,7 @@ static void macronix_default_init(struct
 {
 	nor->params->quad_enable = spi_nor_sr1_bit6_quad_enable;
 	nor->params->set_4byte_addr_mode = spi_nor_set_4byte_addr_mode;
+	nor->flags |= SNOR_F_HAS_LOCK;
 }
 
 static const struct spi_nor_fixups macronix_fixups = {
