--- a/configure.ac
+++ b/configure.ac
@@ -1001,7 +1001,8 @@ AC_RUN_IFELSE(
 		FILE *f = fopen(".", "r");
 		return f != NULL;]])],
 	[ac_cv_fread_reads_directories=no],
-	[ac_cv_fread_reads_directories=yes])
+	[ac_cv_fread_reads_directories=yes],
+	[ac_cv_fread_reads_directories=no])
 ])
 if test $ac_cv_fread_reads_directories = yes; then
 	FREAD_READS_DIRECTORIES=UnfortunatelyYes
@@ -1035,7 +1036,8 @@ AC_RUN_IFELSE(
 		  if (snprintf(buf, 3, "%s", "12345") != 5
 		      || strcmp(buf, "12")) return 1]])],
 	[ac_cv_snprintf_returns_bogus=no],
-	[ac_cv_snprintf_returns_bogus=yes])
+	[ac_cv_snprintf_returns_bogus=yes],
+	[ac_cv_snprintf_returns_bogus=no])
 ])
 if test $ac_cv_snprintf_returns_bogus = yes; then
 	SNPRINTF_RETURNS_BOGUS=UnfortunatelyYes
@@ -1058,7 +1060,8 @@ yippeeyeswehaveit
 #endif
 ]),
 	[ac_cv_sane_mode_bits=yes],
-	[ac_cv_sane_mode_bits=no])
+	[ac_cv_sane_mode_bits=no],
+	[ac_cv_sane_mode_bits=yes])
 ])
 if test $ac_cv_sane_mode_bits = yes; then
 	NEEDS_MODE_TRANSLATION=
