--- a/Makefile
+++ b/Makefile
@@ -1,4 +1,4 @@
-GLIBC=$(shell grep -s -c __GLIBC__ /usr/include/features.h)
+GLIBC=$(shell grep -s -c __GLIBC__ ${STAGING_DIR}/usr/include/features.h)
 
 # @(#) Makefile 1.23 97/03/21 19:27:20
 
@@ -146,9 +146,11 @@ freebsd:
 	LIBS= RANLIB=ranlib ARFLAGS=rv AUX_OBJ= NETGROUP= TLI= \
 	EXTRA_CFLAGS=-DSYS_ERRLIST_DEFINED VSYSLOG= all
 
+ifneq ($(GLIBC),)
 ifneq ($(GLIBC),0)
 MYLIB=-lnsl
 endif
+endif
 
 linux:
 	@make REAL_DAEMON_DIR=$(REAL_DAEMON_DIR) STYLE=$(STYLE) \
