--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -192,7 +192,6 @@ endif()
 # libicu is highly recommended for RSCALE support
 #  libicu can be found at http://www.icu-project.org
 #  RSCALE info at https://tools.ietf.org/html/rfc7529
-find_package(ICU)
 set_package_properties(ICU PROPERTIES
   TYPE RECOMMENDED
   PURPOSE "For RSCALE (RFC7529) support"
@@ -216,7 +215,6 @@ if(ICU_I18N_FOUND)
 endif()
 
 # compile in Berkeley DB support
-find_package(BDB)
 set_package_properties(BDB PROPERTIES
   TYPE OPTIONAL
   PURPOSE "For Berkeley DB storage support"
