From a887fe76be459d32a638d0abde96cc715632c8d5 Mon Sep 17 00:00:00 2001
From: Daniel Golle <daniel@makrotopia.org>
Date: Fri, 20 May 2016 06:31:15 +0200
Subject: [PATCH] device: include goto lable as well
To: lvm-devel@redhat.com

commit b5314c2a6ae5fe4f802e82a4f31cf2fad398ded9
device:  Retry open without O_NOATIME if it fails.

makes use of goto lable 'opened:' but that might not be defined, e.g.
on standard C libraries without O_DIRECT_SUPPORT.
---
 lib/device/dev-io.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/lib/device/dev-io.c
+++ b/lib/device/dev-io.c
@@ -380,7 +380,7 @@ int dev_open_flags(struct device *dev, i
 		return 0;
 	}
 
-#ifdef O_DIRECT_SUPPORT
+#if defined(O_DIRECT_SUPPORT) || defined(O_NOATIME)
       opened:
 	if (direct)
 		dev->flags |= DEV_O_DIRECT_TESTED;
