--- a/Makefile.am
+++ b/Makefile.am
@@ -1,4 +1,4 @@
-SUBDIRS = gphoto-m4 contrib doc gphoto2 packaging po tests
+SUBDIRS = gphoto-m4 contrib doc gphoto2 packaging tests
 
 ACLOCAL_AMFLAGS = -I auto-m4 -I gphoto-m4
 EXTRA_DIST = README.md
--- a/Makefile.in
+++ b/Makefile.in
@@ -389,7 +389,7 @@ target_alias = @target_alias@
 top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
-SUBDIRS = gphoto-m4 contrib doc gphoto2 packaging po tests
+SUBDIRS = gphoto-m4 contrib doc gphoto2 packaging tests
 ACLOCAL_AMFLAGS = -I auto-m4 -I gphoto-m4
 EXTRA_DIST = README.md
 all: config.h
--- a/configure.ac
+++ b/configure.ac
@@ -58,10 +58,7 @@ dnl ------------------------------------
 GP_GETTEXT_HACK([],[Lutz Müller and others],[${MAIL_GPHOTO_TRANSLATION}])
 ALL_LINGUAS="az cs da de en_GB es eu fi fr hu id is it ja nl pa pl pt_BR ro ru rw sk sr sv uk vi zh_CN zh_TW"
 AM_GNU_GETTEXT_VERSION([0.14.1])
-AM_GNU_GETTEXT([external])
-AM_PO_SUBDIRS()
 AM_ICONV()
-GP_GETTEXT_FLAGS()
 
 dnl We cannot use AC_DEFINE_UNQUOTED() for these definitions, as
 dnl we require make to do insert the proper $(datadir) value
@@ -406,7 +403,6 @@ AC_SUBST([AM_LDFLAGS])
 # Create output files
 # ---------------------------------------------------------------------------
 AC_CONFIG_FILES([
-po/Makefile.in 
 Makefile
 gphoto2/Makefile
 gphoto-m4/Makefile
