From c4dd9e1e752b604aa78b6e182fc4ef2e9f3f7347 Mon Sep 17 00:00:00 2001
From: Jian Chang <aa65535@live.com>
Date: Thu, 12 Nov 2015 14:41:50 +0800
Subject: [PATCH 1/2] refine chnroute list

---
 src/chinadns.c | 53 ++++++++++++++++++++++++++++-------------------------
 1 file changed, 28 insertions(+), 25 deletions(-)

diff --git a/src/chinadns.c b/src/chinadns.c
index 6716554..b0d29ad 100644
--- a/src/chinadns.c
+++ b/src/chinadns.c
@@ -97,6 +97,7 @@ static char *ip_list_file = NULL;
 static ip_list_t ip_list;
 static int parse_ip_list();
 
+#define NETMASK_MIN 0
 static char *chnroute_file = NULL;
 static net_list_t chnroute_list;
 static int parse_chnroute();
@@ -436,7 +437,6 @@ static int cmp_net_mask(const void *a, const void *b) {
   net_mask_t *netb = (net_mask_t *)b;
   if (neta->net.s_addr == netb->net.s_addr)
     return 0;
-  // TODO: pre ntohl
   if (ntohl(neta->net.s_addr) > ntohl(netb->net.s_addr))
     return 1;
   return -1;
@@ -451,6 +451,7 @@ static int parse_chnroute() {
   char net[32];
   chnroute_list.entries = 0;
   int i = 0;
+  int cidr;
 
   if (chnroute_file == NULL) {
     VERR("CHNROUTE_FILE not specified, CHNRoute is disabled\n");
@@ -481,9 +482,14 @@ static int parse_chnroute() {
     sp_pos = strchr(line, '/');
     if (sp_pos) {
       *sp_pos = 0;
-      chnroute_list.nets[i].mask = (1 << (32 - atoi(sp_pos + 1))) - 1;
+      cidr = atoi(sp_pos + 1);
+      if (cidr > 0) {
+        chnroute_list.nets[i].mask = (1 << (32 - cidr)) - 1;
+      } else {
+        chnroute_list.nets[i].mask = UINT32_MAX;
+      }
     } else {
-      chnroute_list.nets[i].mask = UINT32_MAX;
+      chnroute_list.nets[i].mask = NETMASK_MIN;
     }
     if (0 == inet_aton(line, &chnroute_list.nets[i].net)) {
       VERR("invalid addr %s in %s:%d\n", line, chnroute_file, i + 1);
@@ -507,29 +513,26 @@ static int test_ip_in_list(struct in_addr ip, const net_list_t *netlist) {
     return 0;
   net_mask_t ip_net;
   ip_net.net = ip;
-  while (l != r) {
-    m = (l + r) / 2;
-    cmp = cmp_net_mask(&ip_net, &netlist->nets[m]);
-    if (cmp == -1) {
-      if (r != m)
-        r = m;
-      else
-        break;
-    } else {
-      if (l != m)
-        l = m;
-      else
-        break;
-    }
+  while (l <= r) {
+    m = (l + r) >> 1;
+    cmp = cmp_net_mask(&netlist->nets[m], &ip_net);
+    if (cmp < 0)
+      l = m + 1;
+    else if (cmp > 0)
+      r = m - 1;
+    else
+      return 1;
+#ifdef DEBUG
     DLOG("l=%d, r=%d\n", l, r);
-    DLOG("%s, %d\n", inet_ntoa(netlist->nets[m].net),
-         netlist->nets[m].mask);
-  }
-  DLOG("result: %x\n",
-       (ntohl(netlist->nets[l].net.s_addr) ^ ntohl(ip.s_addr)));
-  DLOG("mask: %x\n", (UINT32_MAX - netlist->nets[l].mask));
-  if ((ntohl(netlist->nets[l].net.s_addr) ^ ntohl(ip.s_addr)) &
-      (UINT32_MAX ^ netlist->nets[l].mask)) {
+    DLOG("%s, %d\n", inet_ntoa(netlist->nets[m].net), netlist->nets[m].mask);
+#endif
+  }
+#ifdef DEBUG
+  DLOG("nets: %x <-> %x\n", ntohl(netlist->nets[l - 1].net.s_addr, ntohl(ip.s_addr));
+  DLOG("mask: %x\n", netlist->nets[l - 1].mask);
+#endif
+  if (0 == l || (ntohl(ip.s_addr) > (ntohl(netlist->nets[l - 1].net.s_addr)
+      | netlist->nets[l - 1].mask))) {
     return 0;
   }
   return 1;
-- 
2.6.2.windows.1

