--- a/rfs_cm.c
+++ b/rfs_cm.c
@@ -709,7 +709,7 @@ int rfs_cm_start(void)
 
 	RFS_DEBUG("RFS cm start\n");
 #ifdef CONFIG_NF_CONNTRACK_EVENTS
-	ret = nf_conntrack_register_notifier(&init_net, &rfs_cm_conntrack_notifier);
+	ret = nf_conntrack_register_chain_notifier(&init_net, &rfs_cm_conntrack_notifier);
 	if (ret < 0) {
 		RFS_ERROR("can't register nf notifier hook: %d\n", ret);
 		return -1;
@@ -740,7 +740,7 @@ int rfs_cm_stop(void)
 #endif
 
 #ifdef CONFIG_NF_CONNTRACK_EVENTS
-	nf_conntrack_unregister_notifier(&init_net, &rfs_cm_conntrack_notifier);
+	nf_conntrack_unregister_chain_notifier(&init_net, &rfs_cm_conntrack_notifier);
 #endif
 
 	rfs_cm_connection_destroy_all();
