From b05a93c48fdbb50f0c464310dc2ce45777d32ea2 Mon Sep 17 00:00:00 2001
From: Paul Spooren <mail@aparcar.org>
Date: Fri, 2 Oct 2020 14:08:52 -1000
Subject: [PATCH] remove doc generation

Signed-off-by: Paul Spooren <mail@aparcar.org>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/Makefile
+++ b/Makefile
@@ -25,7 +25,7 @@ export DESTDIR SBINDIR LIBDIR CONFDIR MA
 ##
 # Top-level subdirs
 
-subdirs		:= libfetch/ src/ doc/
+subdirs		:= libfetch/ src/
 
 ##
 # Include all rules and stuff
