--- a/usr/bin/byobu-launch.in
+++ b/usr/bin/byobu-launch.in
@@ -30,7 +30,7 @@
 #    or edit your sshd_config, ssh_config, and set:
 #      $HOME/.bashrc:  export BYOBU_DISABLE=1
 
-_tty=$(tty)
+_tty=$(readlink /proc/$$/fd/0)
 if [ "${_tty#/dev/ttyS}" != "$_tty" ] && [ "${_tty#/dev/ttyAMA}" != "$_tty" ]; then
 	# Don't autolaunch byobu on serial consoles
 	# You can certainly run 'byobu' manually, though
