--- a/usr/lib/byobu/include/constants
+++ b/usr/lib/byobu/include/constants
@@ -58,6 +58,7 @@ $BYOBU_SED --follow-symlinks "s///" /dev
 eval $BYOBU_TEST ulimit >/dev/null 2>&1 && export BYOBU_ULIMIT="ulimit" || export BYOBU_ULIMIT="false"
 
 # Find a suitable python interpreter, if undefined
+export BYOBU_PYTHON="python3"
 if [ -z "$BYOBU_PYTHON" ]; then
 	if python3 -c "import snack" >/dev/null 2>&1; then
 		export BYOBU_PYTHON="python3"
