--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -86,7 +86,6 @@ POST_INSTALL = :
 NORMAL_UNINSTALL = :
 PRE_UNINSTALL = :
 POST_UNINSTALL = :
-@USE_LUA_TRUE@noinst_PROGRAMS = lua2c$(EXEEXT)
 @INCLUDE_LUASHELL_TRUE@@USE_LUA_TRUE@am__append_1 = h_lua.c h_lua.h
 @INCLUDE_LUACSHELL_TRUE@@USE_LUA_TRUE@am__append_2 = h_luac.c h_luac.h
 bin_PROGRAMS = haserl$(EXEEXT)
@@ -124,9 +123,6 @@ am__DEPENDENCIES_1 =
 haserl_DEPENDENCIES = $(am__DEPENDENCIES_1)
 haserl_LINK = $(CCLD) $(haserl_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) \
 	$(LDFLAGS) -o $@
-lua2c_SOURCES = lua2c.c
-lua2c_OBJECTS = lua2c.$(OBJEXT)
-lua2c_LDADD = $(LDADD)
 AM_V_P = $(am__v_P_@AM_V@)
 am__v_P_ = $(am__v_P_@AM_DEFAULT_V@)
 am__v_P_0 = false
@@ -165,8 +161,8 @@ AM_V_CCLD = $(am__v_CCLD_@AM_V@)
 am__v_CCLD_ = $(am__v_CCLD_@AM_DEFAULT_V@)
 am__v_CCLD_0 = @echo "  CCLD    " $@;
 am__v_CCLD_1 = 
-SOURCES = $(haserl_SOURCES) $(nodist_haserl_SOURCES) lua2c.c
-DIST_SOURCES = $(am__haserl_SOURCES_DIST) lua2c.c
+SOURCES = $(haserl_SOURCES) $(nodist_haserl_SOURCES)
+DIST_SOURCES = $(am__haserl_SOURCES_DIST)
 am__can_run_installinfo = \
   case $$AM_UPDATE_INFO_DIR in \
     n|no|NO) false;; \
@@ -291,7 +287,7 @@ target_alias = @target_alias@
 top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
-EXTRA_DIST = haserl_lualib.lua lua2c.c
+EXTRA_DIST = haserl_lualib.lua
 
 # Not strictly required, but we explicitly do not distribute the generated inc file
 nodist_haserl_SOURCES = haserl_lualib.inc
@@ -403,10 +399,6 @@ haserl$(EXEEXT): $(haserl_OBJECTS) $(has
 	@rm -f haserl$(EXEEXT)
 	$(AM_V_CCLD)$(haserl_LINK) $(haserl_OBJECTS) $(haserl_LDADD) $(LIBS)
 
-@USE_LUA_FALSE@lua2c$(EXEEXT): $(lua2c_OBJECTS) $(lua2c_DEPENDENCIES) $(EXTRA_lua2c_DEPENDENCIES) 
-@USE_LUA_FALSE@	@rm -f lua2c$(EXEEXT)
-@USE_LUA_FALSE@	$(AM_V_CCLD)$(LINK) $(lua2c_OBJECTS) $(lua2c_LDADD) $(LIBS)
-
 mostlyclean-compile:
 	-rm -f *.$(OBJEXT)
 
@@ -423,7 +415,6 @@ distclean-compile:
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/haserl-haserl.Po@am__quote@ # am--include-marker
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/haserl-rfc2388.Po@am__quote@ # am--include-marker
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/haserl-sliding_buffer.Po@am__quote@ # am--include-marker
-@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/lua2c.Po@am__quote@ # am--include-marker
 
 $(am__depfiles_remade):
 	@$(MKDIR_P) $(@D)
@@ -806,17 +797,12 @@ uninstall-am: uninstall-binPROGRAMS
 .PRECIOUS: Makefile
 
 
-# lua2c_LIBS ?= -llua -ldl -lm
-@USE_LUA_TRUE@lua2c_LDFLAGS ?= -Wl,-E
-@USE_LUA_TRUE@lua2c:	lua2c.c
-@USE_LUA_TRUE@	$(CC_FOR_BUILD) $(CFLAGS) $(LUA_CFLAGS) $(LDFLAGS) $(lua2c_LDFLAGS) \
-@USE_LUA_TRUE@		-o $@ $^ $(LIBS) $(LUA_LIBS)
-
-@USE_LUA_TRUE@haserl_lualib.inc : haserl_lualib.lua lua2c
-@USE_LUA_TRUE@	if ! ./lua2c haserl_lualib haserl_lualib.lua >$@; then \
+@USE_LUA_TRUE@haserl_lualib.inc : haserl_lualib.lua
+@USE_LUA_TRUE@	if ! sed -re '/--\[\[/,/\]\]--/d; s/[[:space:]]*--.*//g; s/^[[:space:]]*//g; /^$$/d; s/ (=+) /\1/g; s/,[[:space:]]*/,/g; s/ ([()])/\1/g; s/key/k/g; s/value/v/g; s/^/"/; s/$$/\\n"/' haserl_lualib.lua >$@.tmp; then \
 @USE_LUA_TRUE@		rm $@; \
 @USE_LUA_TRUE@		exit 1 ;\
 @USE_LUA_TRUE@	fi
+@USE_LUA_TRUE@	{ echo "static const char haserl_lualib[] = "; cat $@.tmp; echo ";"; rm $@.tmp; } >$@
 
 @USE_LUA_TRUE@h_lua_common.c : haserl_lualib.inc
 
