--- a/src/h_lua_common.c
+++ b/src/h_lua_common.c
@@ -79,7 +79,7 @@ lua_common_setup (char *shell, list_t *
 
   /* and load our haserl library */
   if (luaL_loadbuffer
-      (lua_vm, (const char *) &haserl_lualib, sizeof (haserl_lualib),
+      (lua_vm, (const char *) haserl_lualib, sizeof(haserl_lualib)-1,
        "luascript.lua") || lua_pcall (lua_vm, 0, 0, 0))
     {
       die_with_message (NULL, NULL,
