--- a/drivers/char/hw_random/Kconfig
+++ b/drivers/char/hw_random/Kconfig
@@ -229,6 +229,18 @@ config HW_RANDOM_TX4939
 
 	  If unsure, say Y.
 
+config HW_RANDOM_MCS814X
+	tristate "Moschip MCS814x Random Number Generator"
+	depends on HW_RANDOM && ARCH_MCS814X
+	---help---
+	  This driver provides kernel-side support for the Random Number
+	  Generator hardware found on Moschip MCS814x processors.
+
+	  To compile this driver as a module, choose M here: the
+	  module will be called mcs814x-rng.
+
+	  If unusure, say Y.
+
 config HW_RANDOM_MXC_RNGA
 	tristate "Freescale i.MX RNGA Random Number Generator"
 	depends on ARCH_HAS_RNGA
--- a/drivers/char/hw_random/Makefile
+++ b/drivers/char/hw_random/Makefile
@@ -19,6 +19,7 @@ obj-$(CONFIG_HW_RANDOM_OMAP3_ROM) += oma
 obj-$(CONFIG_HW_RANDOM_PASEMI) += pasemi-rng.o
 obj-$(CONFIG_HW_RANDOM_VIRTIO) += virtio-rng.o
 obj-$(CONFIG_HW_RANDOM_TX4939) += tx4939-rng.o
+obj-$(CONFIG_HW_RANDOM_MCS814X) += mcs814x-rng.o
 obj-$(CONFIG_HW_RANDOM_MXC_RNGA) += mxc-rnga.o
 obj-$(CONFIG_HW_RANDOM_OCTEON) += octeon-rng.o
 obj-$(CONFIG_HW_RANDOM_NOMADIK) += nomadik-rng.o
