--- a/arch/arm/mach-mcs814x/include/mach/debug-macro.S
+++ /dev/null
@@ -1,11 +0,0 @@
-#include <mach/mcs814x.h>
-
-                .macro  addruart, rp, rv, tmp
-		ldr	\rp, =MCS814X_PHYS_BASE
-		ldr	\rv, =MCS814X_VIRT_BASE
-		orr	\rp, \rp, #MCS814X_UART
-		orr	\rv, \rv, #MCS814X_UART
-                .endm
-
-#define UART_SHIFT	2
-#include <asm/hardware/debug-8250.S>
--- a/arch/arm/Kconfig.debug
+++ b/arch/arm/Kconfig.debug
@@ -1089,7 +1089,7 @@ config DEBUG_UART_8250
 		(FOOTBRIDGE && !DEBUG_DC21285_PORT) || \
 		ARCH_GEMINI || ARCH_IOP13XX || ARCH_IOP32X || \
 		ARCH_IOP33X || ARCH_IXP4XX || \
-		ARCH_LPC32XX || ARCH_MV78XX0 || ARCH_ORION5X || ARCH_RPC
+		ARCH_LPC32XX || ARCH_MCS814X || ARCH_MV78XX0 || ARCH_ORION5X || ARCH_RPC
 
 # Compatibility options for BCM63xx
 config DEBUG_UART_BCM63XX
@@ -1124,6 +1124,7 @@ config DEBUG_UART_PHYS
 	default 0x3e000000 if DEBUG_BCM_KONA_UART
 	default 0x4000e400 if DEBUG_LL_UART_EFM32
 	default 0x40090000 if ARCH_LPC32XX
+	default 0x400dc000 if ARCH_MCS814X
 	default 0x40100000 if DEBUG_PXA_UART1
 	default 0x42000000 if ARCH_GEMINI
 	default 0x50000000 if DEBUG_S3C24XX_UART && (DEBUG_S3C_UART0 || \
@@ -1178,6 +1179,7 @@ config DEBUG_UART_VIRT
 	default 0xe0010fe0 if ARCH_RPC
 	default 0xe1000000 if DEBUG_MSM_UART
 	default 0xf0000be0 if ARCH_EBSA110
+	default 0xf00dc000 if ARCH_MCS814X
 	default 0xf01fb000 if DEBUG_NOMADIK_UART
 	default 0xf0201000 if DEBUG_BCM2835
 	default 0xf1000300 if DEBUG_BCM_5301X
