diff -ur u-boot-2010.03/drivers/spi/Makefile u-boot-nbg/drivers/spi/Makefile
--- u-boot-2010.03/drivers/spi/Makefile	2010-03-31 23:54:39.000000000 +0200
+++ u-boot-nbg/drivers/spi/Makefile	2010-04-15 19:31:27.000000000 +0200
@@ -25,6 +25,7 @@
 
 LIB	:= $(obj)libspi.a
 
+COBJS-$(CONFIG_AR71XX_SPI) += ar71xx_spi.o
 COBJS-$(CONFIG_ATMEL_DATAFLASH_SPI) += atmel_dataflash_spi.o
 COBJS-$(CONFIG_ATMEL_SPI) += atmel_spi.o
 COBJS-$(CONFIG_BFIN_SPI) += bfin_spi.o
