diff -ur u-boot-2010.03/drivers/net/Makefile u-boot-nbg/drivers/net/Makefile
--- u-boot-2010.03/drivers/net/Makefile	2010-03-31 23:54:39.000000000 +0200
+++ u-boot-nbg/drivers/net/Makefile	2010-04-19 23:30:01.000000000 +0200
@@ -65,6 +65,7 @@
 COBJS-$(CONFIG_DRIVER_RTL8019) += rtl8019.o
 COBJS-$(CONFIG_RTL8139) += rtl8139.o
 COBJS-$(CONFIG_RTL8169) += rtl8169.o
+COBJS-$(CONFIG_RTL8366_MII) += phy/rtl8366_mii.o
 COBJS-$(CONFIG_DRIVER_S3C4510_ETH) += s3c4510b_eth.o
 COBJS-$(CONFIG_SH_ETHER) += sh_eth.o
 COBJS-$(CONFIG_SMC91111) += smc91111.o
diff -ur u-boot-2010.03/include/netdev.h u-boot-nbg/include/netdev.h
--- u-boot-2010.03/include/netdev.h	2010-03-31 23:54:39.000000000 +0200
+++ u-boot-nbg/include/netdev.h	2010-05-02 11:30:58.000000000 +0200
@@ -175,5 +175,13 @@
 
 int mv88e61xx_switch_initialize(struct mv88e61xx_config *swconfig);
 #endif /* CONFIG_MV88E61XX_SWITCH */
+
+#if defined(CONFIG_RTL8366_MII)
+#define RTL8366_DEVNAME         "rtl8366"
+#define RTL8366_WANPHY_ID       4
+#define RTL8366_LANPHY_ID       -1
+int rtl8366_mii_register(bd_t *bis);
+int rtl8366s_initialize(void);
+#endif
 
 #endif /* _NETDEV_H_ */
