--- a/net/bridge/br_private.h
+++ b/net/bridge/br_private.h
@@ -193,7 +193,12 @@
 	struct net_bridge_fdb_key	key;
 	struct hlist_node		fdb_node;
 	unsigned long			flags;
-	unsigned char			offloaded:1;
+	unsigned char			is_local:1,
+                   is_static:1,
+                   is_sticky:1,
+                   added_by_user:1,
+                   added_by_external_learn:1,
+                   offloaded:1;
 
 	/* write-heavy members should not affect lookups */
 	unsigned long			updated ____cacheline_aligned_in_smp;
