--- a/src/ds2438.c
+++ b/src/ds2438.c
@@ -5,6 +5,7 @@
    Licensed under GPL v2
    ----------------------------------------------------------------------- */
 #include <stdio.h>
+#include <sys/types.h>
 #include "ownet.h"
 #include "ad26.h"
 
--- a/userial/ad26.c
+++ b/userial/ad26.c
@@ -33,7 +33,7 @@
 #include "ownet.h"
 #include "ad26.h"
 #include "owproto.h"
-
+#include "sys/types.h"
 
 extern int   owBlock(int,int,uchar *,int);
 extern void  setcrc8(int,uchar);
--- a/userial/cnt1d.c
+++ b/userial/cnt1d.c
@@ -29,6 +29,7 @@
 //  Version: 2.00
 //
 //
+#include "sys/types.h"
 #include "ownet.h"
 
 // external One Wire functions from nework layer
--- a/userial/crcutil.c
+++ b/userial/crcutil.c
@@ -29,6 +29,7 @@
 
 // Include files
 #include "ownet.h"
+#include "sys/types.h"
 
 // Local subroutines
 void setcrc16(int,ushort);
--- a/userial/ds9097u/owtrnu.c
+++ b/userial/ds9097u/owtrnu.c
@@ -39,6 +39,7 @@
 //                        Added file I/O operations
 //
 
+#include "sys/types.h"
 #include "ownet.h"
 #include "ds2480.h"
 
--- a/userial/ioutil.c
+++ b/userial/ioutil.c
@@ -36,6 +36,7 @@
 #include <string.h>
 #include <ctype.h>
 #include "ownet.h"
+#include "sys/types.h"
 
 #ifdef __MC68K__
 #include <PalmOS.h>
--- a/userial/owproto.h
+++ b/userial/owproto.h
@@ -1,3 +1,4 @@
+#include "sys/types.h"
 /* Prototypes for userial driver functions */
 
 /* From other low level userial files */
