From 4a1d8af9737868f6494f491e9a1efb0fb348588e Mon Sep 17 00:00:00 2001
From: Maxime Ripard <maxime@cerno.tech>
Date: Wed, 10 Jun 2020 16:29:08 +0200
Subject: [PATCH] clk: rpi: Fix compatible indentation

Signed-off-by: Maxime Ripard <maxime@cerno.tech>
---
 drivers/clk/bcm/clk-raspberrypi.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/drivers/clk/bcm/clk-raspberrypi.c
+++ b/drivers/clk/bcm/clk-raspberrypi.c
@@ -359,8 +359,8 @@ static int raspberrypi_clk_remove(struct
 }
 
 static const struct of_device_id raspberrypi_clk_match[] = {
-        { .compatible = "raspberrypi,firmware-clocks" },
-        { },
+	{ .compatible = "raspberrypi,firmware-clocks" },
+	{ },
 };
 MODULE_DEVICE_TABLE(of, raspberrypi_clk_match);
 
