From 3effc5ad6a017d4523158e04d2543a42def1744f Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Mon, 21 Sep 2020 14:02:44 +0100
Subject: [PATCH] rpivid_h265: Fix width/height typo

Signed-off-by: popcornmix <popcornmix@gmail.com>
---
 drivers/staging/media/rpivid/rpivid_h265.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/staging/media/rpivid/rpivid_h265.c
+++ b/drivers/staging/media/rpivid/rpivid_h265.c
@@ -2178,7 +2178,7 @@ static int rpivid_h265_start(struct rpiv
 	if (w > 4096)
 		w = 4096;
 	if (h == 0)
-		w = 1088;
+		h = 1088;
 	if (h > 4096)
 		h = 4096;
 	wxh = w * h;
