From e2882043cba45fd9f027dfc102aaaaf1208a65d0 Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.com>
Date: Wed, 6 May 2020 17:02:26 +0100
Subject: [PATCH] ARM: dts: Enable firmware-clocks on all Pis

Signed-off-by: Phil Elwell <phil@raspberrypi.com>
---
 arch/arm/boot/dts/bcm270x.dtsi | 6 ++++++
 1 file changed, 6 insertions(+)

--- a/arch/arm/boot/dts/bcm270x.dtsi
+++ b/arch/arm/boot/dts/bcm270x.dtsi
@@ -7,6 +7,12 @@
 		/delete-property/ stdout-path;
 	};
 
+	firmware_clocks: firmware-clocks {
+		compatible = "raspberrypi,firmware-clocks";
+		raspberrypi,firmware = <&firmware>;
+		#clock-cells = <1>;
+	};
+
 	soc: soc {
 
 		watchdog: watchdog@7e100000 {
