From 7c74b86873fd5252becee33d092d8317ceff5f5b Mon Sep 17 00:00:00 2001
From: Dave Stevenson <dave.stevenson@raspberrypi.com>
Date: Tue, 23 Jun 2020 10:35:24 +0100
Subject: [PATCH] include: media: Add vfl_devnode_type of
 VFL_TYPE_VIDEO

Upsstream are renaming VFL_TYPE_GRABBER to VFL_TYPE_VIDEO.
To make backporting the upstream Unicam driver easier, add an
extra enum entry (same as VFL_TYPE_GRABBER) to match that.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
---
 include/media/v4l2-dev.h | 1 +
 1 file changed, 1 insertion(+)

--- a/include/media/v4l2-dev.h
+++ b/include/media/v4l2-dev.h
@@ -34,6 +34,7 @@
  */
 enum vfl_devnode_type {
 	VFL_TYPE_GRABBER	= 0,
+	VFL_TYPE_VIDEO = VFL_TYPE_GRABBER,
 	VFL_TYPE_VBI,
 	VFL_TYPE_RADIO,
 	VFL_TYPE_SUBDEV,
