From 867a447dbe461b6cfdfae2fcd71c449dced9cd8a Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.com>
Date: Mon, 20 Jul 2020 09:35:46 +0100
Subject: [PATCH] SQUASH: leds: actpwr - delete unused variable

Signed-off-by: Phil Elwell <phil@raspberrypi.com>
---
 drivers/leds/trigger/ledtrig-actpwr.c | 1 -
 1 file changed, 1 deletion(-)

--- a/drivers/leds/trigger/ledtrig-actpwr.c
+++ b/drivers/leds/trigger/ledtrig-actpwr.c
@@ -94,7 +94,6 @@ static void actpwr_trig_cycle(struct tim
 {
 	struct actpwr_trig_data *trig  = &actpwr_data;
 	struct actpwr_vled *active;
-	enum led_brightness value;
 
 	active = &trig->virt_leds[trig->next_active];
 	trig->active = active;
