--- a/drivers/net/phy/Kconfig
+++ b/drivers/net/phy/Kconfig
@@ -95,6 +95,10 @@ config IP17XX_PHY
 	tristate "Driver for IC+ IP17xx switches"
 	select SWCONFIG
 
+config MVSWITCH_PHY
+	tristate "Driver for Marvell 88E6060 switches"
+	select ETHERNET_PACKET_MANGLE
+
 config PSB6970_PHY
 	tristate "Lantiq XWAY Tantos (PSB6970) Ethernet switch"
 	select SWCONFIG
--- a/drivers/net/phy/Makefile
+++ b/drivers/net/phy/Makefile
@@ -29,6 +29,7 @@ obj-$(CONFIG_ADM6996_PHY)	+= adm6996.o
 obj-$(CONFIG_AR8216_PHY)	+= ar8216.o ar8327.o
 obj-$(CONFIG_SWCONFIG_B53)	+= b53/
 obj-$(CONFIG_IP17XX_PHY)	+= ip17xx.o
+obj-$(CONFIG_MVSWITCH_PHY)	+= mvswitch.o
 obj-$(CONFIG_PSB6970_PHY)	+= psb6970.o
 obj-$(CONFIG_RTL8306_PHY)	+= rtl8306.o
 obj-$(CONFIG_RTL8366_SMI)	+= rtl8366_smi.o
