--- a/arch/arm/boot/dts/qcom-ipq8064-ap148.dts
+++ b/arch/arm/boot/dts/qcom-ipq8064-ap148.dts
@@ -77,15 +77,7 @@
 					spi-max-frequency = <50000000>;
 					reg = <0>;
 
-					partition@0 {
-						label = "rootfs";
-						reg = <0x0 0x1000000>;
-					};
-
-					partition@1 {
-						label = "scratch";
-						reg = <0x1000000 0x1000000>;
-					};
+					linux,part-probe = "qcom-smem";
 				};
 			};
 		};
