--- a/drivers/mtd/qcom_smem_part.c
+++ b/drivers/mtd/qcom_smem_part.c
@@ -189,6 +189,10 @@ static int parse_qcom_smem_partitions(st
 		m_part->size = le32_to_cpu(s_part->size) * (*smem_blksz);
 		m_part->offset = le32_to_cpu(s_part->start) * (*smem_blksz);
 
+		/* "rootfs" conflicts with OpenWrt auto mounting */
+		if (mtd_type_is_nand(master) && !strcmp(m_part->name, "rootfs"))
+			m_part->name = "ubi";
+
 		/*
 		 * The last SMEM partition may have its size marked as
 		 * something like 0xffffffff, which means "until the end of the
