--- a/arch/arm/boot/dts/qcom-ipq8064.dtsi
+++ b/arch/arm/boot/dts/qcom-ipq8064.dtsi
@@ -678,6 +678,8 @@
 			clocks = <&gcc USB30_0_MASTER_CLK>;
 			clock-names = "core";
 
+			syscon-tcsr = <&tcsr 0xb0 1>;
+
 			ranges;
 
 			status = "disabled";
@@ -689,6 +691,7 @@
 				phys = <&hs_phy_0>, <&ss_phy_0>;
 				phy-names = "usb2-phy", "usb3-phy";
 				dr_mode = "host";
+				snps,dis_u3_susphy_quirk;
 			};
 		};
 
@@ -699,6 +702,8 @@
 			clocks = <&gcc USB30_1_MASTER_CLK>;
 			clock-names = "core";
 
+			syscon-tcsr = <&tcsr 0xb0 0>;
+
 			ranges;
 
 			status = "disabled";
@@ -710,6 +715,7 @@
 				phys = <&hs_phy_1>, <&ss_phy_1>;
 				phy-names = "usb2-phy", "usb3-phy";
 				dr_mode = "host";
+				snps,dis_u3_susphy_quirk;
 			};
 		};
 
