--- a/makefile.rules
+++ b/makefile.rules
@@ -131,19 +131,20 @@ exec_always:
 	@echo "LIBS = ${LIBS}"
 	@echo "EXEC_FILE = ${EXEC_FILE}"
 	@echo "BUILD_DIR = ${BUILD_DIR}"
+	@echo "LDFLAGS = ${LDFLAGS}"
 
 clean:
 	rm -f ${OBJ_FILES_ELFTOSB2} ${OBJ_FILES_SBTOOL} ${OBJ_FILES_KEYGEN} \
 		${EXEC_FILE_ELFTOSB2} ${EXEC_FILE_SBTOOL} ${EXEC_FILE_KEYGEN}
 
 elftosb: ${OBJ_FILES_ELFTOSB2}
-	gcc ${OBJ_FILES_ELFTOSB2} ${LIBS} -o ${EXEC_FILE_ELFTOSB2}
+	gcc ${OBJ_FILES_ELFTOSB2} ${LIBS} ${LDFLAGS} -o ${EXEC_FILE_ELFTOSB2}
 
 sbtool: ${OBJ_FILES_SBTOOL}
-	gcc ${OBJ_FILES_SBTOOL} ${LIBS} -o ${EXEC_FILE_SBTOOL}
+	gcc ${OBJ_FILES_SBTOOL} ${LIBS} ${LDFLAGS} -o ${EXEC_FILE_SBTOOL}
 
 keygen: ${OBJ_FILES_KEYGEN}
-	gcc ${OBJ_FILES_KEYGEN} ${LIBS} -o ${EXEC_FILE_KEYGEN}
+	gcc ${OBJ_FILES_KEYGEN} ${LIBS} ${LDFLAGS} -o ${EXEC_FILE_KEYGEN}
 
 
 #ifeq ("${UNAMES}", "Linux")
