From 461322f8b15d66dc8860c4f940c3f8f4f36bb3a3 Mon Sep 17 00:00:00 2001
From: Hector Martin <marcan@marcan.st>
Date: Sat, 12 Mar 2022 21:00:22 +0900
Subject: [PATCH 014/171] arm64: dts: apple: t8103-j313: Also disable speakers

Signed-off-by: Hector Martin <marcan@marcan.st>
---
 arch/arm64/boot/dts/apple/t8103-j313.dts | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/arch/arm64/boot/dts/apple/t8103-j313.dts b/arch/arm64/boot/dts/apple/t8103-j313.dts
index ad3a6edeb651..d1f65e48699f 100644
--- a/arch/arm64/boot/dts/apple/t8103-j313.dts
+++ b/arch/arm64/boot/dts/apple/t8103-j313.dts
@@ -112,6 +112,15 @@ sound {
 		model = "MacBook Air J313 integrated audio";
 
 		dai-link@0 {
+			/*
+			 * DANGER ZONE: You can blow your speakers!
+			 *
+			 * The drivers are not ready, and unless you are careful
+			 * to attenuate the audio stream, you run the risk of
+			 * blowing your speakers.
+			 */
+			status = "disabled";
+
 			link-name = "Speakers";
 			mclk-fs = <64>;
 
-- 
2.34.1

