From fa4bcbf275b4c864bdf5506e44369e9a982d8bc4 Mon Sep 17 00:00:00 2001
From: Hector Martin <marcan@marcan.st>
Date: Thu, 3 Mar 2022 03:24:36 +0900
Subject: [PATCH 030/171] arm64: dts: apple: Add backlight node to j314/j316

Signed-off-by: Hector Martin <marcan@marcan.st>
---
 arch/arm64/boot/dts/apple/t600x-j314-j316.dtsi | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/arch/arm64/boot/dts/apple/t600x-j314-j316.dtsi b/arch/arm64/boot/dts/apple/t600x-j314-j316.dtsi
index 21608ca46a55..c5241d5515cf 100644
--- a/arch/arm64/boot/dts/apple/t600x-j314-j316.dtsi
+++ b/arch/arm64/boot/dts/apple/t600x-j314-j316.dtsi
@@ -26,6 +26,7 @@ framebuffer0: framebuffer@0 {
 			compatible = "apple,simple-framebuffer", "simple-framebuffer";
 			reg = <0 0 0 0>; /* To be filled by loader */
 			/* Format properties will be added by loader */
+			backlight = <&backlight>;
 			status = "disabled";
 		};
 	};
@@ -236,3 +237,11 @@ &dwc3_3_dart_1 {
 &dwc3_3 {
 	status = "disabled";
 };
+
+/ {
+	backlight: gpio-bl {
+		compatible = "gpio-backlight";
+		gpios = <&smc_gpio 19 GPIO_ACTIVE_HIGH>;
+		default-on;
+	};
+};
-- 
2.34.1

