From b65315774f730e0b8553230d6cff9912a0341ca9 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Martin=20Povi=C5=A1er?= <povik+lin@cutebit.org>
Date: Sat, 19 Feb 2022 09:49:45 +0100
Subject: [PATCH 074/171] HACK: ASoC: tas2770: Set no of channels to 1

---
 sound/soc/codecs/tas2770.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/sound/soc/codecs/tas2770.c b/sound/soc/codecs/tas2770.c
index 2b214223265e..753e3f8d459a 100644
--- a/sound/soc/codecs/tas2770.c
+++ b/sound/soc/codecs/tas2770.c
@@ -508,8 +508,8 @@ static struct snd_soc_dai_driver tas2770_dai_driver[] = {
 		.id = 0,
 		.playback = {
 			.stream_name    = "ASI1 Playback",
-			.channels_min   = 2,
-			.channels_max   = 2,
+			.channels_min   = 1,
+			.channels_max   = 1,
 			.rates      = TAS2770_RATES,
 			.formats    = TAS2770_FORMATS,
 		},
-- 
2.34.1

