From 3153617d401d69c1d4f52eb3ca69f5b4b8d2086a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Martin=20Povi=C5=A1er?= <povik+lin@cutebit.org>
Date: Thu, 24 Feb 2022 14:50:34 +0100
Subject: [PATCH 078/171] ASoC: Add asoc_pcm_to_rtd utility function
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Martin Povišer <povik+lin@cutebit.org>
---
 include/sound/soc.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/sound/soc.h b/include/sound/soc.h
index 7ef168a8c759..73cc83d73163 100644
--- a/include/sound/soc.h
+++ b/include/sound/soc.h
@@ -1101,6 +1101,8 @@ struct snd_soc_pcm_runtime {
 #define asoc_rtd_to_codec(rtd, n) (rtd)->dais[n + (rtd)->num_cpus]
 #define asoc_substream_to_rtd(substream) \
 	(struct snd_soc_pcm_runtime *)snd_pcm_substream_chip(substream)
+#define asoc_pcm_to_rtd(pcm) \
+	((struct snd_soc_pcm_runtime *)pcm->private_data)
 
 #define for_each_rtd_components(rtd, i, component)			\
 	for ((i) = 0, component = NULL;					\
-- 
2.34.1

