From ddf6983d9229819248d0373854d32b4e4a0e123d Mon Sep 17 00:00:00 2001
From: Hector Martin <marcan@marcan.st>
Date: Sat, 19 Mar 2022 06:06:20 +0900
Subject: [PATCH 084/171] macaudio: Unbork jack volume

Signed-off-by: Hector Martin <marcan@marcan.st>
---
 sound/soc/apple/macaudio.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/sound/soc/apple/macaudio.c b/sound/soc/apple/macaudio.c
index a1f2b1fcea1f..d39ef638ea95 100644
--- a/sound/soc/apple/macaudio.c
+++ b/sound/soc/apple/macaudio.c
@@ -305,7 +305,6 @@ struct fixed_kctl {
 	{"Left VSENSE Switch", "Off"},
 	{"Right ISENSE Switch", "Off"},
 	{"Right VSENSE Switch", "Off"},
-	{"Jack Mixer Volume", "63"},
 	{ }
 };
 
-- 
2.34.1

