diff --color -uNr a/src/corelib/global/qendian.h b/src/corelib/global/qendian.h
--- a/src/corelib/global/qendian.h	2020-10-27 16:02:11.000000000 +0800
+++ b/src/corelib/global/qendian.h	2021-07-19 06:21:23.185297425 +0800
@@ -43,7 +43,7 @@
 
 #include <QtCore/qfloat16.h>
 #include <QtCore/qglobal.h>
-
+#include <limits>
 // include stdlib.h and hope that it defines __GLIBC__ for glibc-based systems
 #include <stdlib.h>
 #include <string.h>
diff --color -uNr a/src/corelib/global/qfloat16.h b/src/corelib/global/qfloat16.h
--- a/src/corelib/global/qfloat16.h	2020-10-27 16:02:11.000000000 +0800
+++ b/src/corelib/global/qfloat16.h	2021-07-19 06:22:10.387349829 +0800
@@ -44,7 +44,7 @@
 #include <QtCore/qglobal.h>
 #include <QtCore/qmetatype.h>
 #include <string.h>
-
+#include <limits>
 #if defined(QT_COMPILER_SUPPORTS_F16C) && defined(__AVX2__) && !defined(__F16C__)
 // All processors that support AVX2 do support F16C too. That doesn't mean
 // we're allowed to use the intrinsics directly, so we'll do it only for
diff --color -uNr a/src/corelib/text/qbytearraymatcher.h b/src/corelib/text/qbytearraymatcher.h
--- a/src/corelib/text/qbytearraymatcher.h	2020-10-27 16:02:11.000000000 +0800
+++ b/src/corelib/text/qbytearraymatcher.h	2021-07-19 06:22:30.139372013 +0800
@@ -41,7 +41,7 @@
 #define QBYTEARRAYMATCHER_H
 
 #include <QtCore/qbytearray.h>
-
+#include <limits>
 QT_BEGIN_NAMESPACE
 
 
