--- a/gcmdline.c
+++ b/gcmdline.c
@@ -29,6 +29,7 @@
 */
 
 #include <stdio.h>
+#include <stdlib.h>
 #include "gkermit.h"
 
 /* Externals */
@@ -53,7 +54,7 @@ _MYPROTOTYPE( VOID fatal, (char *) );
 _MYPROTOTYPE( VOID usage, (void) );
 
 #ifndef NOGETENV
-_MYPROTOTYPE( char * getenv, (char *) );
+_MYPROTOTYPE( char * getenv, (const char *) );
 #define GARGC 32
 #define GBUFSIZ 256
 static char gbuf[GBUFSIZ], *gargs[GARGC], *gptr = NULL;
--- a/gunixio.c
+++ b/gunixio.c
@@ -58,6 +58,7 @@
 */
 
 #include <stdio.h>			/* Standard input/output */
+#include <stdlib.h>
 
 #ifdef POSIX
 #include <termios.h>			/* Terminal modes */
