From 8210ec212fb21dfd85af109af7d5ab39065fb16a Mon Sep 17 00:00:00 2001
From: W_Y_CPP <383152993@qq.com>
Date: Mon, 8 Apr 2024 00:02:11 -0400
Subject: [PATCH] remove print message info

---
 src/r8125_rss.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/r8125_rss.c b/src/r8125_rss.c
index 8980ba7..a27eee2 100755
--- a/src/r8125_rss.c
+++ b/src/r8125_rss.c
@@ -91,7 +91,7 @@ int rtl8125_get_rxnfc(struct net_device *dev, struct ethtool_rxnfc *cmd,
         struct rtl8125_private *tp = netdev_priv(dev);
         int ret = -EOPNOTSUPP;
 
-        netif_info(tp, drv, tp->dev, "rss get rxnfc\n");
+       // netif_info(tp, drv, tp->dev, "rss get rxnfc\n");
 
         if (!(dev->features & NETIF_F_RXHASH))
                 return ret;
-- 
2.17.1

