diff --git a/Makefile b/Makefile
index 4949577..908027d 100644
--- a/Makefile
+++ b/Makefile
@@ -52,6 +52,7 @@ DISABLE_MULTI_MSIX_VECTOR = n
 ENABLE_DOUBLE_VLAN = n
 ENABLE_PAGE_REUSE = n
 ENABLE_RX_PACKET_FRAGMENT = n
+ENABLE_LEGACY_2500baseX = y
 
 ifneq ($(KERNELRELEASE),)
 	obj-m := r8125.o
@@ -131,6 +132,9 @@ ifneq ($(KERNELRELEASE),)
 	ifeq ($(ENABLE_RX_PACKET_FRAGMENT), y)
 		EXTRA_CFLAGS += -DENABLE_RX_PACKET_FRAGMENT
 	endif
+	ifeq ($(ENABLE_LEGACY_2500baseX), y)
+		EXTRA_CFLAGS += -DENABLE_LEGACY_2500baseX
+	endif
 else
 	BASEDIR := /lib/modules/$(shell uname -r)
 	KERNELDIR ?= $(BASEDIR)/build
diff --git a/r8125_n.c b/r8125_n.c
index e1e97db..37ed2f3 100644
--- a/r8125_n.c
+++ b/r8125_n.c
@@ -5378,6 +5378,7 @@ static void rtl8125_gset_xmii(struct net_device *dev,
         ethtool_convert_legacy_u32_to_link_mode(cmd->link_modes.lp_advertising,
                                                 lpa_adv);
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(5,0,0)
+#ifndef ENABLE_LEGACY_2500baseX
         if (supported & SUPPORTED_2500baseX_Full) {
                 linkmode_mod_bit(ETHTOOL_LINK_MODE_2500baseX_Full_BIT,
                                  cmd->link_modes.supported, 0);
@@ -5390,6 +5391,7 @@ static void rtl8125_gset_xmii(struct net_device *dev,
                 linkmode_mod_bit(ETHTOOL_LINK_MODE_2500baseT_Full_BIT,
                                  cmd->link_modes.advertising, 1);
         }
+#endif
         if (HW_SUPP_PHY_LINK_SPEED_5000M(tp)) {
                 linkmode_mod_bit(ETHTOOL_LINK_MODE_5000baseT_Full_BIT,
                                  cmd->link_modes.supported, 1);
@@ -5397,12 +5399,14 @@ static void rtl8125_gset_xmii(struct net_device *dev,
                                  cmd->link_modes.advertising, tp->phy_2500_ctrl_reg & RTK_ADVERTISE_5000FULL);
         }
         if (report_lpa) {
+#ifndef ENABLE_LEGACY_2500baseX
                 if (lpa_adv & ADVERTISED_2500baseX_Full) {
                         linkmode_mod_bit(ETHTOOL_LINK_MODE_2500baseX_Full_BIT,
                                          cmd->link_modes.lp_advertising, 0);
                         linkmode_mod_bit(ETHTOOL_LINK_MODE_2500baseT_Full_BIT,
                                          cmd->link_modes.lp_advertising, 1);
                 }
+#endif
                 if (status_2500 & RTK_LPA_ADVERTISE_5000FULL)
                         linkmode_mod_bit(ETHTOOL_LINK_MODE_5000baseT_Full_BIT,
                                          cmd->link_modes.lp_advertising, 1);
