From 071caeaca82d555eb8560612dfad16832d0d3066 Mon Sep 17 00:00:00 2001
From: anusha <anusharao@codeaurora.org>
Date: Wed, 5 Aug 2020 11:01:41 +0530
Subject: [PATCH] remoteproc: wcss: Add ipq6018 support in remoteproc

Signed-off-by: anusha <anusharao@codeaurora.org>
---
 drivers/remoteproc/qcom_q6v5_wcss.c   | 21 +++++++++++++++++++--
 1 files changed, 20 insertions(+), 3 deletions(-)

--- a/drivers/remoteproc/qcom_q6v5_wcss.c
+++ b/drivers/remoteproc/qcom_q6v5_wcss.c
@@ -91,7 +91,7 @@
 #define WCNSS_PAS_ID		6
 
 enum {
-	WCSS_IPQ8074,
+	WCSS_IPQ,
 	WCSS_QCS404,
 };
 
@@ -906,7 +906,7 @@ static int q6v5_wcss_init_mmio(struct q6
 	if (!wcss->reg_base)
 		return -ENOMEM;
 
-	if (wcss->version == WCSS_IPQ8074) {
+	if (wcss->version == WCSS_IPQ) {
 		res = platform_get_resource_byname(pdev, IORESOURCE_MEM, "rmb");
 		wcss->rmb_base = devm_ioremap_resource(&pdev->dev, res);
 		if (IS_ERR(wcss->rmb_base))
@@ -1193,6 +1193,21 @@ static const struct wcss_data wcss_ipq80
 	.need_auto_boot = false,
 };
 
+static const struct wcss_data wcss_ipq6018_res_init = {
+	.init_clock = ipq8074_init_clock,
+	.q6_firmware_name = "IPQ6018/q6_fw.mdt",
+	.m3_firmware_name = "IPQ6018/m3_fw.mdt",
+	.crash_reason_smem = WCSS_CRASH_REASON,
+	.aon_reset_required = true,
+	.wcss_q6_reset_required = true,
+	.bcr_reset_required = false,
+	.ssr_name = "q6wcss",
+	.ops = &q6v5_wcss_ipq8074_ops,
+	.requires_force_stop = true,
+	.need_mem_protection = true,
+	.need_auto_boot = false,
+};
+
 static const struct wcss_data wcss_qcs404_res_init = {
 	.init_clock = qcs404_init_clock,
 	.init_regulator = qcs404_init_regulator,
@@ -1212,6 +1227,7 @@ static const struct wcss_data wcss_qcs40
 
 static const struct of_device_id q6v5_wcss_of_match[] = {
 	{ .compatible = "qcom,ipq8074-wcss-pil", .data = &wcss_ipq8074_res_init },
+	{ .compatible = "qcom,ipq6018-wcss-pil", .data = &wcss_ipq6018_res_init },
 	{ .compatible = "qcom,qcs404-wcss-pil", .data = &wcss_qcs404_res_init },
 	{ },
 };
