--- a/net/ipv6/ip6_tunnel.c
+++ b/net/ipv6/ip6_tunnel.c
@@ -2421,7 +2421,7 @@ nla_put_failure:
  */
 void ip6_update_offload_stats(struct net_device *dev, void *ptr)
 {
-	struct pcpu_sw_netstats *tstats = per_cpu_ptr(dev->tstats, 0);
+	struct pcpu_sw_netstats *tstats = this_cpu_ptr(dev->tstats);
 	const struct pcpu_sw_netstats *offload_stats =
 					(struct pcpu_sw_netstats *)ptr;
 
--- a/net/ipv6/sit.c
+++ b/net/ipv6/sit.c
@@ -1736,7 +1736,7 @@ nla_put_failure:
 /* QCA NSS Clients Support - Start */
 void ipip6_update_offload_stats(struct net_device *dev, void *ptr)
 {
-	struct pcpu_sw_netstats *tstats = per_cpu_ptr(dev->tstats, 0);
+	struct pcpu_sw_netstats *tstats = this_cpu_ptr(dev->tstats);
 	const struct pcpu_sw_netstats *offload_stats =
 					(struct pcpu_sw_netstats *)ptr;
 
