--- a/Makefile	2024-03-25 19:34:08.989713592 +0800
+++ b/Makefile	2024-03-25 19:33:52.584640195 +0800
@@ -834,6 +834,9 @@
 KBUILD_RUSTFLAGS += -Copt-level=s
 endif
 
+KBUILD_CFLAGS += -mcpu=cortex-a73+crypto+aes+sha2 -mtune=cortex-a73 
+KBUILD_AFLAGS += -mcpu=cortex-a73+crypto+aes+sha2 -mtune=cortex-a73 
+
 # Always set `debug-assertions` and `overflow-checks` because their default
 # depends on `opt-level` and `debug-assertions`, respectively.
 KBUILD_RUSTFLAGS += -Cdebug-assertions=$(if $(CONFIG_RUST_DEBUG_ASSERTIONS),y,n)

