From d857a83272d921929ae6ccf1fa70d85768840e84 Mon Sep 17 00:00:00 2001
From: Nikias Bassen <nikias@gmx.li>
Date: Mon, 10 Aug 2020 15:39:56 +0200
Subject: [PATCH] screenshotr: Set DeviceLink version to 400 to support iOS
 14b4+

---
 src/screenshotr.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/src/screenshotr.c
+++ b/src/screenshotr.c
@@ -30,7 +30,7 @@
 #include "device_link_service.h"
 #include "common/debug.h"
 
-#define SCREENSHOTR_VERSION_INT1 300
+#define SCREENSHOTR_VERSION_INT1 400
 #define SCREENSHOTR_VERSION_INT2 0
 
 /**
