--- a/squashfs-tools/version.sh
+++ b/squashfs-tools/version.sh
@@ -27,13 +27,11 @@ if [ -z "$OUTPUT" ] ; then
 fi
 
 our_date() {
-case $(uname) in
-NetBSD|OpenBSD|DragonFly|FreeBSD|Darwin)
-        date -r "$1" "$2"
-        ;;
-*)
-        date -d "@$1" "$2"
-esac
+	if date --version 2>&1 | grep -q "GNU coreutils"; then
+		date -d "@$1" "$2"
+	else
+		date -r "$1" "$2"
+	fi
 }
 
 try_version() {
