diff --git a/include/configs/gumstix.h b/include/configs/gumstix.h
index 319da63..5483993 100644
--- a/include/configs/gumstix.h
+++ b/include/configs/gumstix.h
@@ -136,7 +136,7 @@
 #define CONFIG_MISC_INIT_R	/* misc_init_r function in gumstix sets board serial number */
 
 #define CONFIG_BOOTFILE		boot/uImage
-#define CONFIG_BOOTARGS		"console=ttyS0,115200n8 root=1f01 rootfstype=jffs2 reboot=cold,hard"
+#define CONFIG_BOOTARGS		"console=ttyS0,115200n8 root=1f01 rootfstype=squashfs,jffs2 reboot=cold,hard"
 #define CONFIG_BOOTCOMMAND	"icache on; setenv stderr nulldev; setenv stdout nulldev; if pinit on && fatload ide 0 a2000000 gumstix-factory.script; then setenv stdout serial; setenv stderr serial; echo Found gumstix-factory.script on CF...; autoscr; else if mmcinit && fatload mmc 0 a2000000 gumstix-factory.script; then setenv stdout serial; setenv stderr serial; echo Found gumstix-factory.script on MMC...; autoscr; else setenv stdout serial; setenv stderr serial; katload 100000 && bootm; fi; fi"
 #define CONFIG_BOOTDELAY	2		/* in seconds */
 #define CONFIG_EXTRA_ENV_SETTINGS	"verify=no"
