--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -72,7 +72,7 @@ pppoe-sniff: pppoe-sniff.o if.o common.o
 	@CC@ -o $@ $^ $(LDFLAGS)
 
 pppoe-server: pppoe-server.o if.o debug.o common.o md5.o libevent/libevent.a @PPPOE_SERVER_DEPS@
-	@CC@ -o $@ @RDYNAMIC@ $^ $(LDFLAGS) $(PPPOE_SERVER_LIBS) -Llibevent -levent
+	@CC@ -o $@ @RDYNAMIC@ $^ $(LDFLAGS) $(PPPOE_SERVER_LIBS)
 
 # Experimental code from Savoir Faire Linux.  I do not consider it
 # production-ready, so not part of the official distribution.
