--- a/src/if.c
+++ b/src/if.c
@@ -33,9 +33,9 @@
 #include <linux/if_packet.h>
 #endif
 
-#ifdef HAVE_NET_ETHERNET_H
-#include <net/ethernet.h>
-#endif
+//#ifdef HAVE_NET_ETHERNET_H
+//#include <net/ethernet.h>
+//#endif
 
 #ifdef HAVE_ASM_TYPES_H
 #include <asm/types.h>
--- a/src/plugin.c
+++ b/src/plugin.c
@@ -52,7 +52,7 @@
 #include <unistd.h>
 #include <fcntl.h>
 #include <signal.h>
-#include <net/ethernet.h>
+//#include <net/ethernet.h>
 #include <net/if_arp.h>
 #include <linux/ppp_defs.h>
 #include <linux/if_pppox.h>
--- a/src/pppoe.h
+++ b/src/pppoe.h
@@ -129,9 +129,9 @@ typedef unsigned long UINT32_t;
 #ifdef HAVE_SYS_SOCKET_H
 #include <sys/socket.h>
 #endif
-#ifndef HAVE_SYS_DLPI_H
-#include <netinet/if_ether.h>
-#endif
+//#ifndef HAVE_SYS_DLPI_H
+//#include <netinet/if_ether.h>
+//#endif
 #endif
 
 
