--- a/src/drv_mei_cpe_api.h
+++ b/src/drv_mei_cpe_api.h
@@ -1000,7 +1000,7 @@ typedef struct MEI_dev_s
 #if ( defined(MEI_DRVOS_HAVE_DRV_SELECT) && (MEI_DRVOS_HAVE_DRV_SELECT == 1) )
    /** support for select() */
    IFX_boolean_t         bNfcNeedWakeUp;
-   MEI_DRVOS_event_t     selNfcWakeupList;
+   IFXOS_drvSelectQueue_t     selNfcWakeupList;
 #endif
 
    /** list of all open instances which can receive NFC's, EVT's ALM's */
--- a/src/drv_mei_cpe_linux.c
+++ b/src/drv_mei_cpe_linux.c
@@ -1302,9 +1302,9 @@ static unsigned int MEI_Poll (struct fil
    }
 
    MEI_DRVOS_SelectQueueAddTask(
-               (MEI_DRVOS_select_OSArg_t*) filp,
-               (MEI_DRVOS_selectQueue_t*)  &(pMeiDev->selNfcWakeupList),
-               (MEI_DRVOS_selectTable_t*)  wait);
+               filp,
+               &(pMeiDev->selNfcWakeupList),
+               wait);
 
    if (pDynNfc->pRecvDataCntrl[pDynNfc->rdIdxRd].bufCtrl != MEI_RECV_BUF_CTRL_FREE)      /* buffer in use */
    {
