--- a/src/drv_mei_cpe_linux.c
+++ b/src/drv_mei_cpe_linux.c
@@ -1501,8 +1501,8 @@ struct proc_entry {
 static void MEI_GetVersionProc(struct seq_file *s)
 {
    seq_printf(s, "%s" MEI_DRV_CRLF, &MEI_WHATVERSION[4]);
-   seq_printf(s, "Compiled on %s, %s for Linux kernel %s (jiffies: %ld)" MEI_DRV_CRLF,
-                                    __DATE__, __TIME__, UTS_RELEASE, jiffies);
+   seq_printf(s, "Compiled for Linux kernel %s (jiffies: %ld)" MEI_DRV_CRLF,
+                                    UTS_RELEASE, jiffies);
 }
 
 /**
@@ -4229,7 +4229,6 @@ module_exit (MEI_module_exit);
 #ifdef MODULE
 MODULE_AUTHOR("www.lantiq.com");
 MODULE_DESCRIPTION("MEI CPE Driver - www.lantiq.com");
-MODULE_SUPPORTED_DEVICE("MEI CPE Interface");
 MODULE_LICENSE ("GPL");
 #endif /* #ifdef MODULE*/
 
