--- a/src/drv_mei_cpe_api_atm_ptm_intern.c
+++ b/src/drv_mei_cpe_api_atm_ptm_intern.c
@@ -124,6 +124,9 @@ IFX_int32_t MEI_InternalXtmSwhowtimeExit
    /* Get line number*/
    dslLineNum = pMeiDynCntrl->pMeiDev->meiDrvCntrl.dslLineNum;
 
+   g_tx_link_rate[dslLineNum][0] = 0;
+   g_tx_link_rate[dslLineNum][1] = 0;
+
    /* get NULL or function pointer */
    mei_showtime_exit =
         (ltq_mei_atm_showtime_exit_t)ppa_callback_get(LTQ_MEI_SHOWTIME_EXIT);
