--- a/src/drv_mei_cpe_linux.c
+++ b/src/drv_mei_cpe_linux.c
@@ -1869,7 +1869,7 @@ static int mei_seq_single_show(struct se
 
 static int mei_proc_single_open(struct inode *inode, struct file *file)
 {
-   return single_open(file, mei_seq_single_show, PDE_DATA(inode));
+   return single_open(file, mei_seq_single_show, pde_data(inode));
 }
 
 static void mei_proc_entry_create(struct proc_dir_entry *parent_node,
--- a/src/drv_mei_cpe_linux_proc_config.c
+++ b/src/drv_mei_cpe_linux_proc_config.c
@@ -1036,7 +1036,7 @@ static int mei_seq_single_show(struct se
 
 static int mei_proc_single_open(struct inode *inode, struct file *file)
 {
-	return single_open(file, mei_seq_single_show, PDE_DATA(inode));
+	return single_open(file, mei_seq_single_show, pde_data(inode));
 }
 
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(5,6,0))
