From: Shiji Yang <yangshiji66@outlook.com>
Date: Tue, 14 Apr 2026 07:59:03 +0800
Subject: [PATCH] convert Makefile EXTRA_CFLAGS to ccflags-y

EXTRA_CFLAGS was removed since Linux 6.15 kernel.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>

--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -353,7 +353,7 @@ drv_mei_cpe.ko: $(drv_mei_cpe_SOURCES)
 	@echo -e "# drv_mei_cpe: Generated to build Linux 2.6.x kernel object" > $(PWD)/Kbuild
 	@echo -e "obj-m := $(subst .ko,.o,$@)"			>> $(PWD)/Kbuild
 	@echo -e "$(subst .ko,,$@)-y := $(drv_mei_cpe_OBJS)"	>> $(PWD)/Kbuild
-	@echo -e "EXTRA_CFLAGS := $(drv_mei_cpe_CFLAGS) -I@abs_srcdir@ -I@abs_srcdir@/auto_header $(IFXOS_INCLUDE_PATH)" >> $(PWD)/Kbuild
+	@echo -e "ccflags-y := $(drv_mei_cpe_CFLAGS) -I@abs_srcdir@ -I@abs_srcdir@/auto_header $(IFXOS_INCLUDE_PATH)" >> $(PWD)/Kbuild
 	$(MAKE) ARCH=@KERNEL_ARCH@ -C @KERNEL_BUILD_PATH@ O=@KERNEL_BUILD_PATH@ M=$(PWD) modules
 
 clean-generic:
