From: =?UTF-8?q?Rafa=C5=82=20Mi=C5=82ecki?= <rafal@milecki.pl>
Date: Mon, 15 Feb 2021 22:01:03 +0100
Subject: [PATCH] arm64: dts: broadcom: bcm4908: limit amount of GPIOs
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Linux driver can't handle more than 64 GPIOs

Signed-off-by: Rafał Miłecki <rafal@milecki.pl>
---

--- a/arch/arm64/boot/dts/broadcom/bcm4908/bcm4908.dtsi
+++ b/arch/arm64/boot/dts/broadcom/bcm4908/bcm4908.dtsi
@@ -292,7 +292,7 @@
 		gpio0: gpio-controller@500 {
 			compatible = "brcm,bcm6345-gpio";
 			reg-names = "dirout", "dat";
-			reg = <0x500 0x28>, <0x528 0x28>;
+			reg = <0x500 0x8>, <0x528 0x8>;
 
 			#gpio-cells = <2>;
 			gpio-controller;
