From: =?UTF-8?q?Rafa=C5=82=20Mi=C5=82ecki?= <rafal@milecki.pl>
Date: Mon, 22 Mar 2021 07:15:17 +0100
Subject: [PATCH] mtd: support BLKRRPART
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Rescan MTD subpartitions on request.

Signed-off-by: Rafał Miłecki <rafal@milecki.pl>
---

--- a/drivers/mtd/mtdchar.c
+++ b/drivers/mtd/mtdchar.c
@@ -1015,8 +1015,11 @@ static int mtdchar_ioctl(struct file *fi
 
 	case BLKRRPART:
 	{
-		/* No reread partition feature. Just return ok */
-		ret = 0;
+		int nr_parts;
+
+		del_mtd_partitions(mtd);
+		nr_parts = parse_mtd_partitions(mtd, NULL, NULL);
+		ret = nr_parts < 0 ? nr_parts : 0;
 		break;
 	}
 	}
