From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Petr=20=C5=A0tetiar?= <ynezz@true.cz>
Date: Thu, 26 Mar 2020 10:09:19 +0100
Subject: [PATCH] arm64: dts: allwinner: a64: olinuxino: add status LED aliases
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Petr Štetiar <ynezz@true.cz>

--- a/arch/arm64/boot/dts/allwinner/sun50i-a64-olinuxino.dts
+++ b/arch/arm64/boot/dts/allwinner/sun50i-a64-olinuxino.dts
@@ -53,6 +53,10 @@
 	aliases {
 		ethernet0 = &emac;
 		serial0 = &uart0;
+		led-boot = &led_user;
+		led-failsafe = &led_user;
+		led-running = &led_user;
+		led-upgrade = &led_user;
 	};
 
 	chosen {
@@ -73,7 +77,7 @@
 	leds {
 		compatible = "gpio-leds";
 
-		user {
+		led_user: user {
 			label = "a64-olinuxino:red:user";
 			gpios = <&pio 4 17 GPIO_ACTIVE_HIGH>; /* PE17 */
 		};
