--- a/configure.ac
+++ b/configure.ac
@@ -428,8 +428,7 @@ ACX_PTHREAD([
              LIBS="$PTHREAD_LIBS $LIBS"
              CFLAGS="$CFLAGS $PTHREAD_CFLAGS"
              CXXFLAGS="$CXXFLAGS $PTHREAD_CXXFLAGS"
-             CC="$PTHREAD_CC"
-             CXX="$PTHREAD_CXX"],
+             CC="$PTHREAD_CC"],
             [AC_MSG_ERROR([Missing pthread, please install it])])
 
 # pthread_setname_np can take either 1 or 2 arguments.
