--- a/configure.ac
+++ b/configure.ac
@@ -112,7 +112,6 @@ AS_IF([test "x$ac_cv_gnu_plus_plus_11" =
       [AS_IF([test "x$ac_cv_gnu_plus_plus_98" = xyes],
              [AS_IF([test "x$ac_cv_header_random_98" = xyes],
                     [AS_IF([test "x$enable_unittests" = xno],
-                           [CXXFLAGS="$CXXFLAGS -std=gnu++98"],
                            [PKG_CHECK_MODULES([CPPUNIT1], [cppunit < 1.14.0],
                                               [CXXFLAGS="$CXXFLAGS -std=gnu++98"],
                                               [PKG_CHECK_MODULES([CPPUNIT2], [cppunit >= 1.14.0],
