--- a/pppd/plugins/radius/config.c
+++ b/pppd/plugins/radius/config.c
@@ -369,31 +369,37 @@ static int test_config(char *filename)
 	}
 #endif
 
+#if 0
 	if (rc_conf_int("login_tries") <= 0)
 	{
 		error("%s: login_tries <= 0 is illegal", filename);
 		return (-1);
 	}
+#endif
 	if (rc_conf_str("seqfile") == NULL)
 	{
 		error("%s: seqfile not specified", filename);
 		return (-1);
 	}
+#if 0
 	if (rc_conf_int("login_timeout") <= 0)
 	{
 		error("%s: login_timeout <= 0 is illegal", filename);
 		return (-1);
 	}
+#endif
 	if (rc_conf_str("mapfile") == NULL)
 	{
 		error("%s: mapfile not specified", filename);
 		return (-1);
 	}
+#if 0
 	if (rc_conf_str("nologin") == NULL)
 	{
 		error("%s: nologin not specified", filename);
 		return (-1);
 	}
+#endif
 
 	return 0;
 }
--- a/pppd/plugins/radius/options.h
+++ b/pppd/plugins/radius/options.h
@@ -31,24 +31,21 @@ typedef struct _option {
 static SERVER acctserver = {0};
 static SERVER authserver = {0};
 
-int default_tries = 4;
-int default_timeout = 60;
-
 static OPTION config_options[] = {
 /* internally used options */
 {"config_file",		OT_STR, ST_UNDEF, NULL},
 /* General options */
 {"auth_order",	 	OT_AUO, ST_UNDEF, NULL},
-{"login_tries",	 	OT_INT, ST_UNDEF, &default_tries},
-{"login_timeout",	OT_INT, ST_UNDEF, &default_timeout},
-{"nologin",		OT_STR, ST_UNDEF, "/etc/nologin"},
-{"issue",		OT_STR, ST_UNDEF, "/etc/radiusclient/issue"},
+{"login_tries",	 	OT_INT, ST_UNDEF, NULL},
+{"login_timeout",	OT_INT, ST_UNDEF, NULL},
+{"nologin",		OT_STR, ST_UNDEF, NULL},
+{"issue",		OT_STR, ST_UNDEF, NULL},
 /* RADIUS specific options */
 {"authserver",		OT_SRV, ST_UNDEF, &authserver},
 {"acctserver",		OT_SRV, ST_UNDEF, &acctserver},
 {"servers",		OT_STR, ST_UNDEF, NULL},
 {"dictionary",		OT_STR, ST_UNDEF, NULL},
-{"login_radius",	OT_STR, ST_UNDEF, "/usr/sbin/login.radius"},
+{"login_radius",	OT_STR, ST_UNDEF, NULL},
 {"seqfile",		OT_STR, ST_UNDEF, NULL},
 {"mapfile",		OT_STR, ST_UNDEF, NULL},
 {"default_realm",	OT_STR, ST_UNDEF, NULL},
