From 2e3e25454a165d55cbcd4fc78539454d4e0d8d51 Mon Sep 17 00:00:00 2001
From: Rosen Penev <rosenp@gmail.com>
Date: Mon, 22 Feb 2021 16:07:02 -0800
Subject: [PATCH] add missing cast

Fixes -Wformat

Signed-off-by: Rosen Penev <rosenp@gmail.com>
---
 bon_time.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/bon_time.cpp
+++ b/bon_time.cpp
@@ -146,7 +146,7 @@ int BonTimer::print_stat(tests_t test, i
         }
         else
         {
-          fprintf(m_fp, " %5d", stat);
+          fprintf(m_fp, " %5d", int(stat));
         }
       }
       else
