From f409f1cbfbeedc7b35b6651de9873e597f2f78e1 Mon Sep 17 00:00:00 2001
From: W_Y_CPP <383152993@qq.com>
Date: Thu, 10 Aug 2023 23:15:47 +0900
Subject: [PATCH] rework eth hw addr set for kernel 5.4

---
 src/compatibility.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/compatibility.h b/src/compatibility.h
index 7738d17..25ffc39 100644
--- a/src/compatibility.h
+++ b/src/compatibility.h
@@ -608,10 +608,12 @@
 #endif /* LINUX_VERSION_CODE < KERNEL_VERSION(5,8,0) */
 #endif /* LINUX_VERSION_CODE < KERNEL_VERSION(5,9,0) */
 #endif /* LINUX_VERSION_CODE < KERNEL_VERSION(5,12,0) */
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5,4,251)
 	static inline void eth_hw_addr_set(struct net_device *dev, const u8 *addr)
 	{
 		memcpy(dev->dev_addr, addr, 6);
 	}
+#endif /* LINUX_VERSION_CODE < KERNEL_VERSION(5,4,251) */
 #endif /* LINUX_VERSION_CODE < KERNEL_VERSION(5,15,0) */
 
 #ifndef FALSE
-- 
2.17.1

