From b61b90b9aea30b944fc51b891dab3ab5eb30980c Mon Sep 17 00:00:00 2001
From: Tianling Shen <cnsztl@immortalwrt.org>
Date: Sat, 29 May 2021 00:46:25 +0800
Subject: [PATCH] compat: skb_mark_not_on_list was backported since kernel
 4.14.217 and 4.19.10

Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>
---
 compatibility.h | 2 ++
 1 file changed, 2 insertions(+)

--- a/compatibility.h
+++ b/compatibility.h
@@ -548,10 +548,12 @@
 			linkmode_clear_bit(nr, addr);
 	}
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,14,217) || LINUX_VERSION_CODE < KERNEL_VERSION(4,19,10)
 	static inline void skb_mark_not_on_list(struct sk_buff *skb)
 	{
 		skb->next = NULL;
 	}
+#endif /* LINUX_VERSION_CODE < KERNEL_VERSION(4,14,217) || LINUX_VERSION_CODE < KERNEL_VERSION(4,19,10) */
 #endif /* LINUX_VERSION_CODE < KERNEL_VERSION(5,0,0) */
 #endif /* LINUX_VERSION_CODE < KERNEL_VERSION(5,1,0) */
 #endif /* LINUX_VERSION_CODE < KERNEL_VERSION(5,4,0) */
