--- a/mDNSShared/PlatformCommon.c
+++ b/mDNSShared/PlatformCommon.c
@@ -43,6 +43,10 @@
 typedef unsigned int socklen_t;
 #endif
 
+#ifndef TCP_NOTSENT_LOWAT
+#define TCP_NOTSENT_LOWAT 25
+#endif
+
 #if MDNS_MALLOC_DEBUGGING
 // We ONLY want this for malloc debugging--on a running production system we want to deal with
 // malloc failures, not just die.   There is a small performance penalty for enabling these options
