--- a/drivers/clocksource/Kconfig
+++ b/drivers/clocksource/Kconfig
@@ -222,6 +222,12 @@ config VF_PIT_TIMER
 	help
 	  Support for Period Interrupt Timer on Freescale Vybrid Family SoCs.
 
+config CLKSRC_RPS_TIMER
+	def_bool y if ARCH_OXNAS
+	select CLKSRC_MMIO
+	help
+	  This option enables support for the oxnas rps timers.
+
 config SYS_SUPPORTS_SH_CMT
         bool
 
--- a/drivers/clocksource/Makefile
+++ b/drivers/clocksource/Makefile
@@ -41,6 +41,7 @@ obj-$(CONFIG_CLKSRC_EXYNOS_MCT)	+= exyno
 obj-$(CONFIG_CLKSRC_LPC32XX)	+= time-lpc32xx.o
 obj-$(CONFIG_CLKSRC_SAMSUNG_PWM)	+= samsung_pwm_timer.o
 obj-$(CONFIG_FSL_FTM_TIMER)	+= fsl_ftm_timer.o
+obj-$(CONFIG_CLKSRC_RPS_TIMER)	+= oxnas_rps_timer.o
 obj-$(CONFIG_VF_PIT_TIMER)	+= vf_pit_timer.o
 obj-$(CONFIG_CLKSRC_QCOM)	+= qcom-timer.o
 obj-$(CONFIG_MTK_TIMER)		+= mtk_timer.o
