--- a/arch/arm/boot/dts/marvell/Makefile
+++ b/arch/arm/boot/dts/marvell/Makefile
@@ -92,6 +92,7 @@ dtb-$(CONFIG_MACH_KIRKWOOD) += \
 	kirkwood-db-88f6282.dtb \
 	kirkwood-dir665.dtb \
 	kirkwood-dns320.dtb \
+	kirkwood-dns320l.dtb \
 	kirkwood-dns325.dtb \
 	kirkwood-dockstar.dtb \
 	kirkwood-dreamplug.dtb \
--- a/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts
+++ b/arch/arm/boot/dts/marvell/kirkwood-dns320l.dts
@@ -32,6 +32,13 @@
 		reg = <0x00000000 0x10000000>;
 	};
 
+	aliases {
+		led-boot = &led_orange_usb;
+		led-failsafe = &led_orange_usb;
+		led-running = &led_orange_usb;
+		led-upgrade = &led_orange_usb;
+	};
+
 	chosen {
 		bootargs = "console=ttyS0,115200n8 earlyprintk";
 		stdout-path = &uart0;
@@ -78,7 +85,7 @@
 			linux,default-trigger = "usbport";
 		};
 
-		orange-usb {
+		led_orange_usb: orange-usb {
 			label = "dns320l:usb:orange";
 			gpios = <&gpio0 26 GPIO_ACTIVE_HIGH>;
 		};
