--- a/Makefile.am
+++ b/Makefile.am
@@ -23,10 +23,6 @@ else
 endif
 endif
 
-if USE_APPLE_ALAC
-  shairport_sync_SOURCES += apple_alac.cpp
-endif
-
 if USE_CUSTOMPIDDIR
 AM_CFLAGS+= \
 	-DPIDDIR=\"$(CUSTOM_PID_DIR)\"
@@ -80,11 +76,6 @@ if USE_PA
 shairport_sync_SOURCES += audio_pa.c
 endif
 
-if USE_CONVOLUTION
-shairport_sync_SOURCES += FFTConvolver/AudioFFT.cpp FFTConvolver/FFTConvolver.cpp FFTConvolver/Utilities.cpp FFTConvolver/convolver.cpp
-AM_CXXFLAGS += -std=c++11
-endif
-
 if USE_DNS_SD
 shairport_sync_SOURCES += mdns_dns_sd.c
 endif
--- a/configure.ac
+++ b/configure.ac
@@ -23,7 +23,6 @@ fi
 
 # Checks for programs.
 AC_PROG_CC
-AC_PROG_CXX
 AC_PROG_INSTALL
 
 PKG_PROG_PKG_CONFIG([0.9.0])
