--- a/meson.build
+++ b/meson.build
@@ -2220,16 +2220,10 @@ endif
 
 glib_conf.set('HAVE_PROC_SELF_CMDLINE', have_proc_self_cmdline)
 
-python = import('python').find_installation('python3')
+python = 'python3'
 # used for '#!/usr/bin/env <name>'
 python_name = 'python3'
 
-python_version = python.language_version()
-python_version_req = '>=3.5'
-if not python_version.version_compare(python_version_req)
-  error('Requires Python @0@, @1@ found.'.format(python_version_req, python_version))
-endif
-
 # Determine which user environment-dependent files that we want to install
 have_bash = find_program('bash', required : false).found() # For completion scripts
 bash_comp_dep = dependency('bash-completion', version: '>=2.0', required: false)
