--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -311,6 +311,21 @@ static struct board_info __initdata boar
 
 	.has_ohci0 = 1,
 };
+
+static struct board_info __initdata board_96348_D4PW = {
+	.name				= "D-4P-W",
+	.expected_cpu_id		= 0x6348,
+
+	.has_enet1			= 1,
+	.has_pci			= 1,
+
+	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
+		.force_speed_100	= 1,
+		.force_duplex_full	= 1,
+	},
+};
 #endif /* CONFIG_BCM63XX_CPU_6348 */
 
 /*
@@ -447,6 +462,7 @@ static const struct board_info __initcon
 	&board_DV201AMR,
 	&board_96348gw_a,
 	&board_rta1025w_16,
+	&board_96348_D4PW,
 #endif
 
 #ifdef CONFIG_BCM63XX_CPU_6358
@@ -478,6 +494,7 @@ static struct of_device_id const bcm963x
 	{ .compatible = "brcm,bcm96348gw-10", .data = &board_96348gw_10, },
 	{ .compatible = "brcm,bcm96348gw-11", .data = &board_96348gw_11, },
 	{ .compatible = "brcm,bcm96348gw-a", .data = &board_96348gw_a, },
+	{ .compatible = "d-link,dsl-2640b-b", .data = &board_96348_D4PW, },
 	{ .compatible = "davolink,dv-201amr", .data = &board_DV201AMR, },
 	{ .compatible = "dynalink,rta1025w", .data = &board_rta1025w_16, },
 	{ .compatible = "netgear,dg834gtpn", .data = &board_96348gw_10, },
