--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -233,6 +233,64 @@ static struct board_info __initdata boar
 	},
 };
 
+static struct board_info __initdata board_gw6200 = {
+	.name				= "GW6200",
+	.expected_cpu_id		= 0x6348,
+
+	.has_enet0			= 1,
+	.has_enet1			= 1,
+	.has_pci			= 1,
+
+	.enet0 = {
+		.has_phy		= 1,
+		.use_internal_phy	= 1,
+	},
+	.enet1 = {
+		.force_speed_100	= 1,
+		.force_duplex_full	= 1,
+	},
+
+	.has_ohci0 = 1,
+
+	.has_dsp			= 1,
+		.dsp = {
+		.gpio_rst		= 8, /* FIXME: What is real GPIO here? */
+		.gpio_int		= 34,
+		.ext_irq		= 2,
+		.cs			= 2,
+	},
+};
+
+static struct board_info __initdata board_gw6000 = {
+	.name				= "GW6000",
+	.expected_cpu_id		= 0x6348,
+
+	.has_enet0			= 1,
+	.has_enet1			= 1,
+	.has_pci			= 1,
+
+	.enet0 = {
+		.has_phy		= 1,
+		.use_internal_phy	= 1,
+	},
+	.enet1 = {
+		.force_speed_100	= 1,
+		.force_duplex_full	= 1,
+	},
+
+	.has_ohci0 = 1,
+
+	.has_dsp			= 1,
+	.dsp = {
+		.gpio_rst		= 6,
+		.gpio_int		= 34,
+		.ext_irq		= 2,
+		.cs			= 2,
+	},
+};
+
+
+
 static struct board_info __initdata board_FAST2404 = {
 	.name				= "F@ST2404",
 	.expected_cpu_id		= 0x6348,
@@ -598,6 +656,8 @@ static const struct board_info __initcon
 #ifdef CONFIG_BCM63XX_CPU_6348
 	&board_96348r,
 	&board_96348gw,
+	&board_gw6000,
+	&board_gw6200,
 	&board_96348gw_10,
 	&board_96348gw_11,
 	&board_FAST2404,
@@ -648,6 +708,8 @@ static struct of_device_id const bcm963x
 	{ .compatible = "netgear,dg834gtpn", .data = &board_96348gw_10, },
 	{ .compatible = "sagem,f@st2404", .data = &board_FAST2404, },
 	{ .compatible = "t-com,spw500v", .data = &board_spw500v, },
+	{ .compatible = "tecom,gw6000", .data = &board_gw6000, },
+	{ .compatible = "tecom,gw6200", .data = &board_gw6200, },
 	{ .compatible = "tp-link,td-w8900gb", .data = &board_96348gw_11, },
 	{ .compatible = "usr,9108", .data = &board_96348gw_a, },
 #endif
