--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -93,6 +93,60 @@ static struct board_info __initdata boar
 	},
 };
 
+static struct sprom_fixup __initdata ar5381u_fixups[] = {
+	{ .offset = 97, .value = 0xfee5 },
+	{ .offset = 98, .value = 0x157c },
+	{ .offset = 99, .value = 0xfae7 },
+	{ .offset = 113, .value = 0xfefa },
+	{ .offset = 114, .value = 0x15d6 },
+	{ .offset = 115, .value = 0xfaf8 },
+};
+
+static struct board_info __initdata board_AR5381u = {
+	.name					= "96328A-1241N",
+	.expected_cpu_id			= 0x6328,
+
+	.has_pci				= 1,
+	.use_fallback_sprom			= 1,
+	.has_ohci0				= 1,
+	.has_ehci0				= 1,
+	.num_usbh_ports				= 1,
+	.has_enetsw				= 1,
+
+	.enetsw = {
+		.used_ports = {
+			[0] = {
+				.used		= 1,
+				.phy_id		= 1,
+				.name		= "Port 1",
+			},
+			[1] = {
+				.used		= 1,
+				.phy_id		= 2,
+				.name		= "Port 2",
+			},
+			[2] = {
+				.used		= 1,
+				.phy_id		= 3,
+				.name		= "Port 3",
+			},
+			[3] = {
+				.used		= 1,
+				.phy_id		= 4,
+				.name		= "Port 4",
+			},
+		},
+	},
+
+	.fallback_sprom = {
+		.type 				= SPROM_BCM43225,
+		.pci_bus			= 1,
+		.pci_dev			= 0,
+		.board_fixups			= ar5381u_fixups,
+		.num_board_fixups		= ARRAY_SIZE(ar5381u_fixups),
+	},
+};
+
 static struct sprom_fixup __initdata ar5387un_fixups[] = {
 	{ .offset = 2, .value = 0x05bb },
 	{ .offset = 65, .value = 0x1204 },
@@ -1273,6 +1327,7 @@ static const struct board_info __initcon
 #endif
 #ifdef CONFIG_BCM63XX_CPU_6328
 	&board_96328avng,
+	&board_AR5381u,
 	&board_AR5387un,
 	&board_963281TAN,
 	&board_A4001N1,
@@ -1344,6 +1399,7 @@ static struct of_device_id const bcm963x
 	{ .compatible = "adb,a4001n1", .data = &board_A4001N1, },
 	{ .compatible = "brcm,bcm963281TAN", .data = &board_963281TAN, },
 	{ .compatible = "brcm,bcm96328avng", .data = &board_96328avng, },
+	{ .compatible = "comtrend,ar-5381u", .data = &board_AR5381u, },
 	{ .compatible = "comtrend,ar-5387un", .data = &board_AR5387un, },
 	{ .compatible = "d-link,dsl-274xb-f", .data = &board_dsl_274xb_f1, },
 #endif
