--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -13,6 +13,8 @@
 #include <linux/kernel.h>
 #include <linux/string.h>
 #include <linux/pci_ids.h>
+#include <linux/platform_data/b53.h>
+#include <linux/spi/spi.h>
 #include <asm/addrspace.h>
 #include <bcm63xx_board.h>
 #include <bcm63xx_cpu.h>
@@ -1616,6 +1618,47 @@ static struct board_info __initdata boar
 	.has_ohci0 = 1,
 	.has_ehci0 = 1,
 };
+
+static struct sprom_fixup __initdata wap5813n_fixups[] = {
+	{ .offset = 97, .value = 0xfeed },
+	{ .offset = 98, .value = 0x15d1 },
+	{ .offset = 99, .value = 0xfb0d },
+	{ .offset = 113, .value = 0xfef7 },
+	{ .offset = 114, .value = 0x15f7 },
+	{ .offset = 115, .value = 0xfb1a },
+};
+
+static struct board_info __initdata board_WAP5813n = {
+	.name				= "96369R-1231N",
+	.expected_cpu_id		= 0x6368,
+
+	.has_pci			= 1,
+	.use_fallback_sprom		= 1,
+	.has_ohci0			= 1,
+	.has_ehci0			= 1,
+
+	.has_enetsw			= 1,
+	.enetsw = {
+		.used_ports = {
+			[4] = {
+				.used = 1,
+				.phy_id = 0xff,
+				.bypass_link = 1,
+				.force_speed = 1000,
+				.force_duplex_full = 1,
+				.name = "RGMII",
+			},
+		},
+	},
+
+	.fallback_sprom = {
+		.type 				= SPROM_BCM43222,
+		.pci_bus			= 0,
+		.pci_dev			= 1,
+		.board_fixups			= wap5813n_fixups,
+		.num_board_fixups		= ARRAY_SIZE(wap5813n_fixups),
+	},
+};
 #endif /* CONFIG_BCM63XX_CPU_6368 */
 
 /*
@@ -1846,6 +1889,7 @@ static const struct board_info __initcon
 #ifdef CONFIG_BCM63XX_CPU_6368
 	&board_96368mvwg,
 	&board_96368mvngr,
+	&board_WAP5813n,
 #endif
 #ifdef CONFIG_BCM63XX_CPU_63268
 	&board_963268bu_p300,
@@ -1938,6 +1982,7 @@ static struct of_device_id const bcm963x
 #ifdef CONFIG_BCM63XX_CPU_6368
 	{ .compatible = "brcm,bcm96368mvngr", .data = &board_96368mvngr, },
 	{ .compatible = "brcm,bcm96368mvwg", .data = &board_96368mvwg, },
+	{ .compatible = "comtrend,wap-5813n", .data = &board_WAP5813n, },
 #endif
 #ifdef CONFIG_BCM63XX_CPU_63268
 	{ .compatible = "brcm,bcm963268bu_p300", .data = &board_963268bu_p300, },
