--- a/arch/powerpc/platforms/44x/Kconfig
+++ b/arch/powerpc/platforms/44x/Kconfig
@@ -41,6 +41,19 @@ config EBONY
 	help
 	  This option enables support for the IBM PPC440GP evaluation board.
 
+config IKAREM
+	bool "Ikarem"
+	depends on 44x
+	default n
+	select PPC44x_SIMPLE
+	select APM821xx
+	select PCI_MSI
+	select PPC4xx_MSI
+	select PPC4xx_PCI_EXPRESS
+	select IBM_EMAC_RGMII
+	help
+	  This option enables support for the Cisco Meraki MR24 (Ikarem) Access Point.
+
 config SAM440EP
         bool "Sam440ep"
 	depends on 44x
--- a/arch/powerpc/platforms/44x/ppc44x_simple.c
+++ b/arch/powerpc/platforms/44x/ppc44x_simple.c
@@ -62,6 +62,7 @@ static char *board[] __initdata = {
 	"amcc,sequoia",
 	"amcc,taishan",
 	"amcc,yosemite",
+	"meraki,ikarem",
 	"mosaixtech,icon"
 };
 
