From 019ce2b273a27409664b52ae73a26f06dc696239 Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.com>
Date: Mon, 6 Mar 2023 20:04:34 +0000
Subject: [PATCH] iio: adc: mcp3422: Add correct compatible strings

Signed-off-by: Phil Elwell <phil@raspberrypi.com>
---
 drivers/iio/adc/mcp3422.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

--- a/drivers/iio/adc/mcp3422.c
+++ b/drivers/iio/adc/mcp3422.c
@@ -407,7 +407,14 @@ static const struct i2c_device_id mcp342
 MODULE_DEVICE_TABLE(i2c, mcp3422_id);
 
 static const struct of_device_id mcp3422_of_match[] = {
-	{ .compatible = "mcp3422" },
+	{ .compatible = "microchip,mcp3421" },
+	{ .compatible = "microchip,mcp3422" },
+	{ .compatible = "microchip,mcp3423" },
+	{ .compatible = "microchip,mcp3424" },
+	{ .compatible = "microchip,mcp3425" },
+	{ .compatible = "microchip,mcp3426" },
+	{ .compatible = "microchip,mcp3427" },
+	{ .compatible = "microchip,mcp3428" },
 	{ }
 };
 MODULE_DEVICE_TABLE(of, mcp3422_of_match);
