From 282819aead0166af415b780241dc2def4caee7f4 Mon Sep 17 00:00:00 2001
From: Alexander Winkowski <dereference23@outlook.com>
Date: Mon, 3 Jul 2023 18:12:01 +0000
Subject: [PATCH] char: broadcom: vc_mem: Fix preprocessor conditional

Signed-off-by: Alexander Winkowski <dereference23@outlook.com>
---
 drivers/char/broadcom/vc_mem.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/char/broadcom/vc_mem.c
+++ b/drivers/char/broadcom/vc_mem.c
@@ -353,7 +353,7 @@ vc_mem_exit(void)
 	pr_debug("%s: called\n", __func__);
 
 	if (vc_mem_inited) {
-#if CONFIG_DEBUG_FS
+#ifdef CONFIG_DEBUG_FS
 		vc_mem_debugfs_deinit();
 #endif
 		device_destroy(vc_mem_class, vc_mem_devnum);
