From 4ae8587685a3def5b87881f43940ffe0fd794208 Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.com>
Date: Tue, 7 Mar 2023 21:47:58 +0000
Subject: [PATCH] hwmon: (sht3x) Add DT compatible string

Signed-off-by: Phil Elwell <phil@raspberrypi.com>
---
 drivers/hwmon/sht3x.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

--- a/drivers/hwmon/sht3x.c
+++ b/drivers/hwmon/sht3x.c
@@ -741,8 +741,18 @@ static const struct i2c_device_id sht3x_
 
 MODULE_DEVICE_TABLE(i2c, sht3x_ids);
 
+static const struct of_device_id sht3x_of_ids[] = {
+	{ .compatible = "sensirion,sht3x" },
+	{ .compatible = "sensirion,sts3x" },
+	{}
+};
+MODULE_DEVICE_TABLE(of, sht3x_of_ids);
+
 static struct i2c_driver sht3x_i2c_driver = {
-	.driver.name = "sht3x",
+	.driver = {
+		.name = "sht3x",
+		.of_match_table = sht3x_of_ids,
+	},
 	.probe_new   = sht3x_probe,
 	.id_table    = sht3x_ids,
 };
