From 267601d68d4c5184727b4abe1fe64e89157ee8d7 Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.com>
Date: Tue, 7 Mar 2023 21:48:54 +0000
Subject: [PATCH] iio: light: tsl4531: Add DT compatible string

Signed-off-by: Phil Elwell <phil@raspberrypi.com>
---
 drivers/iio/light/tsl4531.c | 7 +++++++
 1 file changed, 7 insertions(+)

--- a/drivers/iio/light/tsl4531.c
+++ b/drivers/iio/light/tsl4531.c
@@ -233,9 +233,16 @@ static const struct i2c_device_id tsl453
 };
 MODULE_DEVICE_TABLE(i2c, tsl4531_id);
 
+static const struct of_device_id tsl4531_of_id[] = {
+	{ .compatible = "amstaos,tsl4531" },
+	{ }
+};
+MODULE_DEVICE_TABLE(of, tsl4531_of_id);
+
 static struct i2c_driver tsl4531_driver = {
 	.driver = {
 		.name   = TSL4531_DRV_NAME,
+		.of_match_table = tsl4531_of_id,
 		.pm	= pm_sleep_ptr(&tsl4531_pm_ops),
 	},
 	.probe  = tsl4531_probe,
