From 67c4dc83131e335f92a142d3511b94db4a1dc57c Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.com>
Date: Tue, 27 Oct 2020 12:12:22 +0000
Subject: [PATCH] bcm2708_fb: Fix a build warning

Signed-off-by: Phil Elwell <phil@raspberrypi.com>
---
 drivers/video/fbdev/bcm2708_fb.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/drivers/video/fbdev/bcm2708_fb.c
+++ b/drivers/video/fbdev/bcm2708_fb.c
@@ -693,7 +693,8 @@ static long vc_mem_copy(struct bcm2708_f
 		u8 *q = (u8 *)ioparam->dst + offset;
 
 		dma_memcpy(fb, bus_addr,
-			   INTALIAS_L1L2_NONALLOCATING((dma_addr_t)p), size);
+			   INTALIAS_L1L2_NONALLOCATING((u32)(uintptr_t)p),
+						       size);
 		if (copy_to_user(q, buf, s) != 0) {
 			pr_err("[%s]: failed to copy-to-user\n", __func__);
 			rc = -EFAULT;
