--- a/src/mach-ixp42x/slugos-fsg3-armeb_config
+++ b/src/mach-ixp42x/slugos-fsg3-armeb_config
@@ -17,7 +17,7 @@ CONFIG_EXPERIMENTAL=y
 #
 # General Setup
 #
-CONFIG_TARGET_DESCRIPTION="SlugOS FSG3/BE"
+CONFIG_TARGET_DESCRIPTION="OpenWrt FSG3"
 CONFIG_CROSS_COMPILE=""
 CONFIG_AEABI=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
@@ -148,9 +148,9 @@ CONFIG_ENV_REGION_KERNEL_ALT="fis://kern
 #    Overrides
 #
 CONFIG_ENV_DEFAULT_CMDLINE_OVERRIDE=y
-CONFIG_ENV_DEFAULT_CMDLINE="root=/dev/sda1 rootdelay=10 console=ttyS0,115200"
+CONFIG_ENV_DEFAULT_CMDLINE="root=/dev/sda1 rootdelay=10 console=ttyS0,115200 init=/etc/preinit noinitrd"
 CONFIG_ENV_DEFAULT_CMDLINE_ALT_P=y
-CONFIG_ENV_DEFAULT_CMDLINE_ALT="root=/dev/sda2 rootdelay=10 console=ttyS0,115200"
+CONFIG_ENV_DEFAULT_CMDLINE_ALT="root=/dev/mtdblock2 rootfstype=squashfs console=ttyS0,115200 init=/etc/preinit noinitrd"
 # CONFIG_ENV_DEFAULT_STARTUP_OVERRIDE is not set
 # CONFIG_ENV_DEFAULT_STARTUP_ALT_P is not set
 CONFIG_USES_NOR_BOOTFLASH=y
