--- a/src/mach-ixp42x/slugos-nslu2-armeb_config
+++ b/src/mach-ixp42x/slugos-nslu2-armeb_config
@@ -137,7 +137,7 @@ CONFIG_AUTOBOOT_DELAY=10
 CONFIG_ENV_STARTUP_KERNEL_COPY=y
 # CONFIG_ENV_REGION_KERNEL_SWAP is not set
 CONFIG_ENV_STARTUP_PREFIX_P=y
-CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+256m"
+CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+64m"
 
 #
 #    Regions
--- a/src/mach-ixp42x/slugos-nslu2-16mb-armeb_config
+++ b/src/mach-ixp42x/slugos-nslu2-16mb-armeb_config
@@ -137,7 +137,7 @@ CONFIG_AUTOBOOT_DELAY=10
 CONFIG_ENV_STARTUP_KERNEL_COPY=y
 # CONFIG_ENV_REGION_KERNEL_SWAP is not set
 CONFIG_ENV_STARTUP_PREFIX_P=y
-CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+256m"
+CONFIG_ENV_STARTUP_PREFIX="sdram-init; memscan -u 0+64m"
 
 #
 #    Regions
