--- a/src/mach-ixp42x/slugos-nas100d-armeb_config
+++ b/src/mach-ixp42x/slugos-nas100d-armeb_config
@@ -19,7 +19,7 @@ CONFIG_EXPERIMENTAL=y
 #
 # General Setup
 #
-CONFIG_TARGET_DESCRIPTION="SlugOS NAS100D/BE"
+CONFIG_TARGET_DESCRIPTION="OpenWrt NAS100D"
 CONFIG_CROSS_COMPILE=""
 CONFIG_AEABI=y
 # CONFIG_DRIVER_LONG_LONG_SIZE is not set
@@ -158,7 +158,7 @@ CONFIG_ENV_REGION_KERNEL="fis://kernel"
 # Overrides
 #
 CONFIG_ENV_DEFAULT_CMDLINE_OVERRIDE=y
-CONFIG_ENV_DEFAULT_CMDLINE="root=/dev/mtdblock2 rootfstype=jffs2 console=ttyS0,115200 init=/linuxrc"
+CONFIG_ENV_DEFAULT_CMDLINE="root=/dev/mtdblock2 rootfstype=squashfs,jffs2 console=ttyS0,115200 init=/etc/preinit noinitrd"
 # CONFIG_ENV_DEFAULT_STARTUP_OVERRIDE is not set
 CONFIG_USES_NOR_BOOTFLASH=y
 CONFIG_RELOCATE_SIMPLE=y
