--- a/rx.c
+++ b/rx.c
@@ -176,7 +176,7 @@ void rtw_update_rx_freq_from_ie(struct r
 	else
 		goto fill_rx_status;
 
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 4, 0))
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 4, 0)) && !defined(BUILD_OPENWRT)
 	channel_number = cfg80211_get_ies_channel_number(variable, ielen,
 							 NL80211_BAND_2GHZ);
 #elif (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 17, 0))
