--- a/Makefile
+++ b/Makefile
@@ -30,10 +30,6 @@ NO_SKIP_SIGN := y
 endif
 
 EXTRA_CFLAGS += -O2
-EXTRA_CFLAGS += -DCONFIG_RTW88_8822BE=1
-EXTRA_CFLAGS += -DCONFIG_RTW88_8821CE=1
-EXTRA_CFLAGS += -DCONFIG_RTW88_8822CE=1
-EXTRA_CFLAGS += -DCONFIG_RTW88_8723DE=1
 EXTRA_CFLAGS += -DCONFIG_RTW88_DEBUG=1
 EXTRA_CFLAGS += -DCONFIG_RTW88_DEBUGFS=1
 #EXTRA_CFLAGS += -DCONFIG_RTW88_REGD_USER_REG_HINTS
@@ -60,9 +56,6 @@ rtw_core-objs += main.o \
 obj-m       += rtw_8822b.o
 rtw_8822b-objs                := rtw8822b.o rtw8822b_table.o
 
-obj-m      += rtw_8822be.o
-rtw_8822be-objs               := rtw8822be.o
-
 obj-m      += rtw_8822bu.o
 rtw_8822bu-objs		:= rtw8822bu.o
 
@@ -72,9 +65,6 @@ rtw_8822bs-objs		:= rtw8822bs.o
 obj-m       += rtw_8822c.o
 rtw_8822c-objs                := rtw8822c.o rtw8822c_table.o
 
-obj-m      += rtw_8822ce.o
-rtw_8822ce-objs               := rtw8822ce.o
-
 obj-m      += rtw_8822cu.o
 rtw_8822cu-objs		:= rtw8822cu.o
 
@@ -102,9 +92,6 @@ rtw_8723cs-objs		:= rtw8723cs.o
 obj-m       += rtw_8723d.o
 rtw_8723d-objs          := rtw8723d.o rtw8723d_table.o
 
-obj-m      += rtw_8723de.o
-rtw_8723de-objs               := rtw8723de.o
-
 obj-m      += rtw_8723du.o
 rtw_8723du-objs		:= rtw8723du.o
 
@@ -114,15 +101,9 @@ rtw_8723ds-objs		:= rtw8723ds.o
 obj-m	+= rtw_8821c.o
 rtw_8821c-objs		:= rtw8821c.o rtw8821c_table.o
 
-obj-m	+= rtw_8821ce.o
-rtw_8821ce-objs		:= rtw8821ce.o
-
 obj-m	+= rtw_8821a.o
 rtw_8821a-objs		:= rtw8821a.o rtw8821a_table.o
 
-obj-m	   += rtw_8821ae.o
-rtw_8821ae-objs		:= rtw8821ae.o
-
 obj-m	   += rtw_8821au.o
 rtw_8821au-objs		:= rtw8821au.o
 
@@ -132,9 +113,6 @@ rtw_8821cs-objs		:= rtw8821cs.o
 obj-m	+= rtw_8821cu.o
 rtw_8821cu-objs		:= rtw8821cu.o
 
-obj-m			+= rtw_pci.o
-rtw_pci-objs		:= pci.o
-
 obj-m			+= rtw_sdio.o
 rtw_sdio-objs		:= sdio.o
 
