--- a/Makefile
+++ b/Makefile
@@ -24,10 +24,6 @@ NO_SKIP_SIGN := y
 endif
 
 EXTRA_CFLAGS += -O2
-EXTRA_CFLAGS += -DCONFIG_RTW88_8822BE=1
-EXTRA_CFLAGS += -DCONFIG_RTW88_8821CE=1
-EXTRA_CFLAGS += -DCONFIG_RTW88_8822CE=1
-EXTRA_CFLAGS += -DCONFIG_RTW88_8723DE=1
 #EXTRA_CFLAGS += -DCONFIG_RTW88_DEBUG=1
 #EXTRA_CFLAGS += -DCONFIG_RTW88_DEBUGFS=1
 #EXTRA_CFLAGS += -DCONFIG_RTW88_REGD_USER_REG_HINTS
@@ -54,42 +50,27 @@ rtw_core-objs += main.o \
 obj-m       += rtw_8822b.o
 rtw_8822b-objs                := rtw8822b.o rtw8822b_table.o
 
-obj-m      += rtw_8822be.o
-rtw_8822be-objs               := rtw8822be.o
-
 obj-m      += rtw88_8822bu.o
 rtw88_8822bu-objs		:= rtw8822bu.o
 
 obj-m       += rtw_8822c.o
 rtw_8822c-objs                := rtw8822c.o rtw8822c_table.o
 
-obj-m      += rtw_8822ce.o
-rtw_8822ce-objs               := rtw8822ce.o
-
 obj-m      += rtw88_8822cu.o
 rtw88_8822cu-objs		:= rtw8822cu.o
 
 obj-m       += rtw_8723d.o
 rtw_8723d-objs                := rtw8723d.o rtw8723d_table.o
 
-obj-m      += rtw_8723de.o
-rtw_8723de-objs               := rtw8723de.o
-
 obj-m      += rtw88_8723du.o
 rtw88_8723du-objs		:= rtw8723du.o
 
 obj-m	+= rtw_8821c.o
 rtw_8821c-objs		:= rtw8821c.o rtw8821c_table.o
 
-obj-m	+= rtw_8821ce.o
-rtw_8821ce-objs		:= rtw8821ce.o
-
 obj-m	+= rtw88_8821cu.o
 rtw88_8821cu-objs		:= rtw8821cu.o
 
-obj-m			+= rtw_pci.o
-rtw_pci-objs		:= pci.o
-
 obj-m			+= rtw_usb.o
 rtw_usb-objs		:= usb.o
 
