--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,9 +1,12 @@
 # Project
 project(libftdi)
+set(PACKAGE libftdi)
 set(MAJOR_VERSION 0)
 set(MINOR_VERSION 20)
+set(VERSION ${VERSION_STRING})
 set(VERSION_STRING ${MAJOR_VERSION}.${MINOR_VERSION})
-SET(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}")
+set(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}")
+set(top_srcdir ${CMAKE_SOURCE_DIR})
 
 # CMake
 if("${CMAKE_BUILD_TYPE}" STREQUAL "")
@@ -98,12 +101,6 @@ option(DOCUMENTATION "Generate API docum
 
 find_package(Doxygen)
 if(DOCUMENTATION AND DOXYGEN_FOUND)
-
-   # Set variables
-   set(PACKAGE libftdi)
-   set(VERSION ${VERSION_STRING})
-   set(top_srcdir ${CMAKE_SOURCE_DIR})
-
    # Find doxy config
    message(STATUS "Doxygen found.")
    set(DOXY_DIR "${CMAKE_SOURCE_DIR}/doc")
