From f05ba671931e2b4e38459899f6f63f79f99869fe Mon Sep 17 00:00:00 2001
From: Azat Khuzhin <a3at.mail@gmail.com>
Date: Sat, 3 Aug 2019 14:32:21 +0300
Subject: [PATCH] Add Uninstall.cmake.in into dist archive

Fixes: #863
---
 Makefile.am | 1 +
 1 file changed, 1 insertion(+)

diff --git a/Makefile.am b/Makefile.am
index af75a963..92f9433c 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -127,6 +127,7 @@ CMAKE_FILES = \
 	cmake/LibeventConfig.cmake.in \
 	cmake/LibeventConfigVersion.cmake.in \
 	cmake/VersionViaGit.cmake \
+	cmake/Uninstall.cmake.in \
 	event-config.h.cmake \
 	evconfig-private.h.cmake \
 	CMakeLists.txt
-- 
2.22.0

