--- a/configfile.cc
+++ b/configfile.cc
@@ -24,6 +24,7 @@
 #include <stdlib.h>
 #include <stdio.h>
 #include <errno.h>
+#include <time.h>
 #include "configfile.h"
 #include "util.h"
 
--- a/dumpfile_tuntap.cc
+++ b/dumpfile_tuntap.cc
@@ -20,8 +20,8 @@
 
 #include <errno.h>
 
-#include "dumpfile_tuntap.h"
 #include "ifcontrol.h"
+#include "dumpfile_tuntap.h"
 #include "ipc_remote.h"
 
 #ifndef SYS_CYGWIN
--- a/ifcontrol.h
+++ b/ifcontrol.h
@@ -39,7 +39,6 @@
 #include <net/if_arp.h>
 #include <sys/ioctl.h>
 #include <netinet/in.h>
-#include <netinet/if_ether.h>
 #include <arpa/inet.h>
 #endif
 
