From c8d3576df41025b341eb767168914990598a0caa Mon Sep 17 00:00:00 2001
From: Camelia Groza <camelia.groza@nxp.com>
Date: Tue, 22 Aug 2017 18:35:11 +0300
Subject: [PATCH] net: phy: add 10G fixed-link support

Signed-off-by: Camelia Groza <camelia.groza@nxp.com>
---
 drivers/net/phy/swphy.c | 1 +
 1 file changed, 1 insertion(+)

--- a/drivers/net/phy/swphy.c
+++ b/drivers/net/phy/swphy.c
@@ -71,6 +71,7 @@ static const struct swmii_regs duplex[]
 static int swphy_decode_speed(int speed)
 {
 	switch (speed) {
+	case 10000:
 	case 1000:
 		return SWMII_SPEED_1000;
 	case 100:
