From c579c23d57c0ef7c832cf0844c86c7be42c197d3 Mon Sep 17 00:00:00 2001
From: Radu Alexe <radu.alexe@nxp.com>
Date: Thu, 13 Apr 2017 09:36:32 +0300
Subject: [PATCH] crypto: tcrypt - include rsa test

Signed-off-by: Radu Alexe <radu.alexe@nxp.com>
Signed-off-by: Tudor Ambarus <tudor-dan.ambarus@nxp.com>
---
 crypto/tcrypt.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

--- a/crypto/tcrypt.c
+++ b/crypto/tcrypt.c
@@ -71,9 +71,8 @@ static char *check[] = {
 	"cast6", "arc4", "michael_mic", "deflate", "crc32c", "tea", "xtea",
 	"khazad", "wp512", "wp384", "wp256", "tnepres", "xeta",  "fcrypt",
 	"camellia", "seed", "salsa20", "rmd128", "rmd160", "rmd256", "rmd320",
-	"lzo", "lzo-rle", "cts", "sha3-224", "sha3-256", "sha3-384",
-	"sha3-512", "streebog256", "streebog512",
-	NULL
+	"lzo", "lzo-rle", "cts", "zlib", "sha3-224", "sha3-256", "sha3-384",
+	"sha3-512", "streebog256", "streebog512", "rsa", NULL
 };
 
 static u32 block_sizes[] = { 16, 64, 256, 1024, 1472, 8192, 0 };
@@ -1983,6 +1982,10 @@ static int do_test(const char *alg, u32
 		ret += tcrypt_test("hmac(streebog512)");
 		break;
 
+	case 117:
+		ret += tcrypt_test("rsa");
+		break;
+
 	case 150:
 		ret += tcrypt_test("ansi_cprng");
 		break;
