--- a/server/drivers/port.h
+++ b/server/drivers/port.h
@@ -94,7 +94,7 @@ static inline int port_deny_multiple(uns
 /*  ---------------------------- Linux ------------------------------------ */
 /*  Use ioperm, inb and outb in <sys/io.h> (Linux) */
 /*  And iopl for higher addresses of PCI LPT cards */
-#if defined HAVE_IOPERM
+#if defined(__GLIBC__) || (defined(__x86__) || defined(__x86_64__))
 
 /* Glibc2 and Glibc1 */
 # ifdef HAVE_SYS_IO_H
