--- a/clients/lcdproc/iface.c
+++ b/clients/lcdproc/iface.c
@@ -32,6 +32,7 @@
 #define UNSET_INT -1
 #define UNSET_STR "\01"
 
+IfaceInfo iface[MAX_INTERFACES];
 
 static int iface_count = 0;	/* number of interfaces */
 static char unit_label[10] = "B";	/* default unit label is Bytes */
--- a/clients/lcdproc/iface.h
+++ b/clients/lcdproc/iface.h
@@ -18,7 +18,7 @@
 /** max number of interfaces in multi-interface mode */
 #define MAX_INTERFACES 3
 
-IfaceInfo iface[MAX_INTERFACES];	/* interface info */
+extern IfaceInfo iface[MAX_INTERFACES];	/* interface info */
 
 /** Update screen content */
 int iface_screen(int rep, int display, int *flags_ptr);
