--- a/src/include/linux/ifxos_linux_thread.h
+++ b/src/include/linux/ifxos_linux_thread.h
@@ -206,7 +206,7 @@ typedef struct
 /**
    LINUX User Thread - map the Thread ID.
 */
-typedef int    IFXOS_thread_t;
+typedef pthread_t    IFXOS_thread_t;
 
 /**
    LINUX Kernel Process - map the Process ID.
--- a/src/linux/ifxos_linux_socket_appl.c
+++ b/src/linux/ifxos_linux_socket_appl.c
@@ -363,8 +363,8 @@ IFX_int_t IFXOS_SocketSendTo(
    IFXOS_RETURN_IF_POINTER_NULL(pBuffer, IFX_ERROR);
    IFXOS_RETURN_IF_ARG_LE_ZERO(bufSize_byte, IFX_ERROR);
 
-   ret = (IFX_int_t)sendto((int)socFd, (const char*)pBuffer, 
-		(int)bufSize_byte, 0, pSocAddr, sizeof(IFXOS_sockAddr_t));
+   ret = (IFX_int_t)sendto(socFd, pBuffer, 
+		bufSize_byte, 0, (struct sockaddr *)pSocAddr, sizeof(IFXOS_sockAddr_t));
 
    return ret;
 }
