--- a/src/include/linux/ifxos_linux_common.h
+++ b/src/include/linux/ifxos_linux_common.h
@@ -49,6 +49,7 @@
    IFX Linux adaptation - Includes (Linux Kernel)
    ========================================================================= */
 #include <linux/kernel.h>
+#include <linux/version.h>
 #include <asm/byteorder.h>
 
 /* ============================================================================
