--- a/src/linux/ifxos_linux_socket_drv.c
+++ b/src/linux/ifxos_linux_socket_drv.c
@@ -28,7 +28,7 @@
 #include <linux/net.h>
 #include <linux/fs.h>
 #include <linux/inet.h>
-#include <asm/uaccess.h>
+#include <linux/uaccess.h>
 
 #include "ifx_types.h"
 #include "ifxos_rt_if_check.h"
--- a/src/linux/ifxos_linux_socket_ipv6_drv.c
+++ b/src/linux/ifxos_linux_socket_ipv6_drv.c
@@ -25,7 +25,7 @@
 #include <linux/version.h>
 #include <linux/in.h>
 #include <linux/net.h>
-#include <asm/uaccess.h>
+#include <linux/uaccess.h>
 
 #include "ifx_types.h"
 #include "ifxos_rt_if_check.h"
--- a/src/include/linux/ifxos_linux_interrupt.h
+++ b/src/include/linux/ifxos_linux_interrupt.h
@@ -31,7 +31,7 @@
    IFX Linux adaptation - Includes
    ========================================================================= */
 #include "ifx_types.h"
-#include <asm/irq.h>
+#include <linux/interrupt.h>
 
 /* ============================================================================
    IFX Linux adaptation - supported features
