--- a/dcwapd.linuxjsonstatic/main.cxx
+++ b/dcwapd.linuxjsonstatic/main.cxx
@@ -10,6 +10,7 @@
 #include "dcwlinux/ap_configuration.h"
 #include "dcwlinux/vap_manager.h"
 #include "dcwlinux/json_configuration_provider.h"
+#include "dcwlinux/uci_configuration_provider.h"
 
 #include "dcw/dcwlog.h"
 
@@ -19,7 +20,8 @@ int
 main( void ) {
 
   try {
-    dcwlinux::JsonConfigurationProvider      configProvider("./dcwapdconf.json");
+    //dcwlinux::JsonConfigurationProvider      configProvider("./dcwapdconf.json");
+    dcwlinux::UciConfigurationProvider      configProvider("dcwapd");
 
     dcwposix::ProcessSignalManager           sigman;
     dcwposix::SelectEventReactor             eventReactor;
