--- a/src/utils.c
+++ b/src/utils.c
@@ -258,8 +258,6 @@ usage()
 {
     printf("\n");
     printf("shadowsocks-libev %s with %s\n\n", VERSION, USING_CRYPTO);
-    printf(
-        "  maintained by Max Lv <max.c.lv@gmail.com> and Linus Yang <laokongzi@gmail.com>\n\n");
     printf("  usage:\n\n");
 #ifdef MODULE_LOCAL
     printf("    ss-local\n");
@@ -299,6 +297,25 @@ usage()
         "                                  The default cipher is rc4-md5.\n");
     printf("\n");
     printf(
+        "       -o <obfs>                  Obfs of your remote server: plain,\n");
+    printf(
+        "                                  http_simple, http_post and tls1.2_ticket_auth.\n");
+    printf(
+        "       -g <obfs-param>            Obfs-Param of your remote server.\n");
+    printf(
+        "       -O <protocol>              Protocol of your remote server: orgin,\n");
+    printf(
+        "                                  auth_sha1, auth_sha1_v2, auth_sha1_v4,\n");
+    printf(
+        "                                  auth_aes128_md5, auth_aes128_sha1,\n");
+    printf(
+        "                                  auth_chain_a, auth_chain_b, auth_chain_c,\n");
+    printf(
+        "                                  auth_chain_d, auth_chain_e and auth_chain_f.\n");
+    printf(
+        "       -G <protocol-param>        Protocol-Param of your remote server.\n");
+    printf("\n");
+    printf(
         "       [-a <user>]                Run as another user.\n");
     printf(
         "       [-f <pid_file>]            The file path to store pid.\n");
