From 9918c683fcc2f148328332d58d030ec5750a1473 Mon Sep 17 00:00:00 2001
From: Paul Spooren <mail@aparcar.org>
Date: Sat, 19 Feb 2022 17:20:37 +0100
Subject: [PATCH 1/4] openwrt: move layer db to temp folder

Signed-off-by: Paul Spooren <mail@aparcar.org>
---
 src/database.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/src/database.c
+++ b/src/database.c
@@ -1626,7 +1626,7 @@ const char *apk_db_layer_name(int layer)
 {
 	switch (layer) {
 	case APK_DB_LAYER_ROOT: return "lib/apk/db";
-	case APK_DB_LAYER_UVOL: return "lib/apk/db-uvol";
+	case APK_DB_LAYER_UVOL: return "tmp/run/uvol/.meta/apk";
 	default:
 		assert(!"invalid layer");
 		return 0;
