--- a/src/genhelp_apk.lua
+++ b/src/genhelp_apk.lua
@@ -65,7 +65,7 @@ local function render_options(doc, out,
 end
 
 local function render_footer(doc, out)
-	table.insert(out, ("\nFor more information: man %s %s\n"):format(doc.mansection, doc.manpage))
+	table.insert(out, ("\nFor more information:\n  https://openwrt.org/docs/guide-user/additional-software/apk\n"))
 end
 
 local function render_optgroups(doc, out, groups)
