--- a/portifmgr/nss_portifmgr.c
+++ b/portifmgr/nss_portifmgr.c
@@ -187,16 +187,20 @@ drop:
 }
 
 /*
- * nss_portifmgr_get_stats()
+ * nss_portifmgr_get_stats64()
  *	Netdev get stats function to get port stats
  */
-static struct rtnl_link_stats64 *nss_portifmgr_get_stats(struct net_device *dev, struct rtnl_link_stats64 *stats)
+/*
+ * nss_nlgre_redir_cmn_dev_stats64
+ *	Report packet statistics to linux
+ */
+static void nss_portifmgr_get_stats64(struct net_device *dev,
+		struct rtnl_link_stats64 *stats)
 {
 	struct nss_portifmgr_priv *priv = (struct nss_portifmgr_priv *)netdev_priv(dev);
 	BUG_ON(priv == NULL);
 
 	nss_portid_get_stats(priv->if_num, stats);
-	return stats;
 }
 
 /*
@@ -225,7 +229,7 @@ static const struct net_device_ops nss_p
 	.ndo_start_xmit		= nss_portifmgr_start_xmit,
 	.ndo_set_mac_address	= eth_mac_addr,
 	.ndo_change_mtu		= nss_portifmgr_change_mtu,
-	.ndo_get_stats64	= nss_portifmgr_get_stats,
+	.ndo_get_stats64	= nss_portifmgr_get_stats64,
 };
 
 /*
