--- a/vxlanmgr/nss_vxlanmgr.c
+++ b/vxlanmgr/nss_vxlanmgr.c
@@ -84,7 +84,7 @@ int32_t nss_vxlanmgr_bind_ipsec_by_ip(un
 {
 	int32_t ipsec_if_num;
 	nss_vxlanmgr_get_ipsec_if_num_by_ip_callback_t ipsec_cb;
-	struct nss_ctx_instance *nss_ctx = nss_vxlan_get_ctx();
+	__attribute__((unused)) struct nss_ctx_instance *nss_ctx = nss_vxlan_get_ctx();
 
 	/*
 	 * Check if the VxLAN interface is applied over an IPsec interface by querying the IPsec.
