--- a/match/nss_match_cmd.c
+++ b/match/nss_match_cmd.c
@@ -460,10 +460,10 @@ static int nss_match_cmd_procfs_reset_ne
 	char *cmd_buf = nss_match_data;
 	nss_tx_status_t nss_tx_status;
 	struct nss_ctx_instance *nss_ctx = nss_match_get_context();
-	struct nss_ctx_instance *wifi_nss_ctx = nss_wifi_get_context();
+	struct nss_ctx_instance *wifili_nss_ctx = nss_wifili_get_context();
 
-	if (!nss_ctx || !wifi_nss_ctx) {
-		pr_warn("%px: NSS Context not found. wifi_nss_ctx: %px. Reset nexthop failed", nss_ctx, wifi_nss_ctx);
+	if (!nss_ctx || !wifili_nss_ctx) {
+		pr_warn("%px: NSS Context not found. wifili_nss_ctx: %px. Reset nexthop failed", nss_ctx, wifili_nss_ctx);
 		return -ENOMEM;
 	}
 
@@ -495,9 +495,9 @@ static int nss_match_cmd_procfs_reset_ne
 	 * nss_phys_if_reset_nexthop: Used for physical interfaces.
 	 * nss_if_reset_nexthop: used for VAP interfaces.
 	 */
-	type = nss_dynamic_interface_get_type(wifi_nss_ctx, if_num);
+	type = nss_dynamic_interface_get_type(wifili_nss_ctx, if_num);
 	if (type == NSS_DYNAMIC_INTERFACE_TYPE_VAP) {
-		nss_tx_status = nss_if_reset_nexthop(wifi_nss_ctx, if_num);
+		nss_tx_status = nss_if_reset_nexthop(wifili_nss_ctx, if_num);
 	} else if (if_num < NSS_MAX_PHYSICAL_INTERFACES) {
 		nss_tx_status = nss_phys_if_reset_nexthop(nss_ctx, if_num);
 	} else {
@@ -528,7 +528,7 @@ static int nss_match_cmd_procfs_set_if_n
 	uint32_t nh_if_num;
 	int table_id;
 	struct nss_ctx_instance *nss_ctx = nss_match_get_context();
-	struct nss_ctx_instance *wifi_nss_ctx = nss_wifi_get_context();
+	struct nss_ctx_instance *wifili_nss_ctx = nss_wifili_get_context();
 	char *dev_name, *nexthop_msg;
 	char *cmd_buf = NULL;
 	size_t count = *lenp;
@@ -539,8 +539,8 @@ static int nss_match_cmd_procfs_set_if_n
 		return ret;
 	}
 
-	if (!nss_ctx || !wifi_nss_ctx) {
-		pr_warn("%px: NSS Context not found. wifi_nss_ctx: %px. Set nexthop failed", nss_ctx, wifi_nss_ctx);
+	if (!nss_ctx || !wifili_nss_ctx) {
+		pr_warn("%px: NSS Context not found. wifili_nss_ctx: %px. Set nexthop failed", nss_ctx, wifili_nss_ctx);
 		return -ENOMEM;
 	}
 
@@ -607,9 +607,9 @@ static int nss_match_cmd_procfs_set_if_n
 	 * nss_phys_if_set_nexthop: Used for physical interfaces.
 	 * nss_if_set_nexthop: used for VAP interfaces.
 	 */
-	type = nss_dynamic_interface_get_type(wifi_nss_ctx, if_num);
+	type = nss_dynamic_interface_get_type(wifili_nss_ctx, if_num);
 	if (type == NSS_DYNAMIC_INTERFACE_TYPE_VAP) {
-		nss_tx_status = nss_if_set_nexthop(wifi_nss_ctx, if_num, nh_if_num);
+		nss_tx_status = nss_if_set_nexthop(wifili_nss_ctx, if_num, nh_if_num);
 	} else if (if_num < NSS_MAX_PHYSICAL_INTERFACES) {
 		nss_tx_status = nss_phys_if_set_nexthop(nss_ctx, if_num, nh_if_num);
 	} else {
--- a/gre/nss_connmgr_gre.c
+++ b/gre/nss_connmgr_gre.c
@@ -1672,7 +1672,7 @@ int nss_connmgr_gre_set_wifi_next_hop(st
 		return GRE_ERR_NEXT_NODE_UNREG_IN_AE;
 	}
 
-	ctx = nss_wifi_get_context();
+	ctx = nss_wifili_get_context();
 	status = nss_wifi_vdev_set_next_hop(ctx, ifnumber, NSS_GRE_INTERFACE);
 	if (status != NSS_TX_SUCCESS) {
 		nss_connmgr_gre_info("%px: wifi drv api failed to set next hop\n", wifi_vdev);
--- a/mirror/nss_mirror_ctl.c
+++ b/mirror/nss_mirror_ctl.c
@@ -480,7 +480,7 @@ static int nss_mirror_ctl_parse_enable_i
 	type = nss_dynamic_interface_get_type(nss_ctx, if_num);
 
 	if (type == NSS_DYNAMIC_INTERFACE_TYPE_VAP) {
-		status = nss_wifi_vdev_set_next_hop(nss_wifi_get_context(), if_num, mirror_if_num);
+		status = nss_wifi_vdev_set_next_hop(nss_wifili_get_context(), if_num, mirror_if_num);
 	} else if (if_num < NSS_MAX_PHYSICAL_INTERFACES) {
 		status = nss_phys_if_set_nexthop(nss_ctx, if_num, mirror_if_num);
 	} else {
--- a/netlink/nss_nlgre_redir_cmn.c
+++ b/netlink/nss_nlgre_redir_cmn.c
@@ -905,7 +905,7 @@ int nss_nlgre_redir_cmn_set_next_hop(uin
 	}
 
 	nss_nl_info("%px: next hop interface number is %d\n", nss_ctx, next_dev_ifnum);
-	ctx = nss_wifi_get_context();
+	ctx = nss_wifili_get_context();
 
 	ret = nss_wifi_vdev_set_next_hop(ctx, ifnumber, next_dev_ifnum);
 	if (ret != NSS_TX_SUCCESS) {
