--- a/dtls/v2.0/nss_dtlsmgr_ctx_dev.c
+++ b/dtls/v2.0/nss_dtlsmgr_ctx_dev.c
@@ -532,7 +532,8 @@ void nss_dtlsmgr_ctx_dev_setup(struct ne
 #else
 	dev->priv_destructor = nss_dtlsmgr_ctx_dev_free;
 #endif
-	memcpy(dev->dev_addr, "\xaa\xbb\xcc\xdd\xee\xff", dev->addr_len);
-	memset(dev->broadcast, 0xff, dev->addr_len);
+	const uint8_t mac_addr[ETH_ALEN] = { 0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff };
+	eth_hw_addr_set(dev, mac_addr);
+	eth_broadcast_addr(dev->broadcast);
 	memcpy(dev->perm_addr, dev->dev_addr, dev->addr_len);
 }
