From d44264c8735f79da3253520024841311c555ca31 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Rafa=C5=82=20Mi=C5=82ecki?= <rafal@milecki.pl>
Date: Fri, 14 Apr 2017 22:25:12 +0200
Subject: [PATCH] thermal: broadcom: fix compilation of Northstar driver

---
 drivers/thermal/broadcom/ns-thermal.c | 1 +
 1 file changed, 1 insertion(+)

--- a/drivers/thermal/broadcom/ns-thermal.c
+++ b/drivers/thermal/broadcom/ns-thermal.c
@@ -6,6 +6,7 @@
  * published by the Free Software Foundation.
  */
 
+#include <asm/io.h>
 #include <linux/module.h>
 #include <linux/of_address.h>
 #include <linux/platform_device.h>
