--- a/lib/vsprintf.c
+++ b/lib/vsprintf.c
@@ -618,8 +618,10 @@ char *symbol_string(char *buf, char *end
 		    struct printf_spec spec, const char *fmt)
 {
 	unsigned long value;
-#ifdef CONFIG_KALLSYMS
 	char sym[KSYM_SYMBOL_LEN];
+#ifndef CONFIG_KALLSYMS
+	struct module *mod;
+	int len;
 #endif
 
 	if (fmt[1] == 'R')
@@ -633,15 +635,15 @@ char *symbol_string(char *buf, char *end
 		sprint_symbol(sym, value);
 	else
 		sprint_symbol_no_offset(sym, value);
-
-	return string(buf, end, sym, spec);
 #else
-	spec.field_width = 2 * sizeof(void *);
-	spec.flags |= SPECIAL | SMALL | ZEROPAD;
-	spec.base = 16;
+	len = snprintf(sym, sizeof(sym), "0x%lx", value);
 
-	return number(buf, end, value, spec);
+	mod = __module_address(value);
+	if (mod)
+		snprintf(sym + len, sizeof(sym) - len, " [%s@%p+0x%x]",
+			 mod->name, mod->module_core, mod->core_size);
 #endif
+	return string(buf, end, sym, spec);
 }
 
 static noinline_for_stack
