--- a/net/wireless/Kconfig
+++ b/net/wireless/Kconfig
@@ -201,7 +201,7 @@ config CFG80211_WEXT_EXPORT
 	  wext compatibility symbols to be exported.
 
 config LIB80211
-	tristate
+	tristate "LIB80211"
 	default n
 	help
 	  This options enables a library of common routines used
@@ -210,13 +210,16 @@ config LIB80211
 	  Drivers should select this themselves if needed.
 
 config LIB80211_CRYPT_WEP
-	tristate
+	tristate "LIB80211_CRYPT_WEP"
+	select LIB80211
 
 config LIB80211_CRYPT_CCMP
-	tristate
+	tristate "LIB80211_CRYPT_CCMP"
+	select LIB80211
 
 config LIB80211_CRYPT_TKIP
-	tristate
+	tristate "LIB80211_CRYPT_TKIP"
+	select LIB80211
 
 config LIB80211_DEBUG
 	bool "lib80211 debugging messages"
