--- a/drivers/misc/Kconfig
+++ b/drivers/misc/Kconfig
@@ -150,6 +150,18 @@ config SGI_IOC4
 	  If you have an SGI Altix with an IOC4-based card say Y.
 	  Otherwise say N.
 
+config OWL_LOADER
+	tristate "Owl loader for initializing Atheros PCI(e) Wifi chips"
+	depends on PCI
+	---help---
+	This kernel module helps to initialize certain Qualcomm
+	Atheros' PCI(e) Wifi chips, which have the init data
+	(which contains the PCI device ID for example) stored
+	together with the calibration data in the file system.
+
+	This is necessary for devices like the Cisco Meraki Z1, say M.
+	Otherwise say N.
+
 config TIFM_CORE
 	tristate "TI Flash Media interface support"
 	depends on PCI
--- a/drivers/misc/Makefile
+++ b/drivers/misc/Makefile
@@ -15,6 +15,7 @@ obj-$(CONFIG_BMP085_SPI)	+= bmp085-spi.o
 obj-$(CONFIG_DUMMY_IRQ)		+= dummy-irq.o
 obj-$(CONFIG_ICS932S401)	+= ics932s401.o
 obj-$(CONFIG_LKDTM)		+= lkdtm.o
+obj-$(CONFIG_OWL_LOADER)	+= owl-loader.o
 obj-$(CONFIG_TIFM_CORE)       	+= tifm_core.o
 obj-$(CONFIG_TIFM_7XX1)       	+= tifm_7xx1.o
 obj-$(CONFIG_PHANTOM)		+= phantom.o
