From: Kees Cook <kees@debian.org>
Subject: put stack markings in i386 assembly to avoid executable stack
Bug-Debian: http://bugs.debian.org/528280
X-Debian-version: 3.0-2

--- a/crc_i386.S
+++ b/crc_i386.S
@@ -302,3 +302,5 @@ _crc32:                         /* ulg c
 #endif /* i386 || _i386 || _I386 || __i386 */
 
 #endif /* !USE_ZLIB && !CRC_TABLE_ONLY */
+.section .note.GNU-stack, "", @progbits
+.previous
--- a/match.S
+++ b/match.S
@@ -405,3 +405,5 @@ L__return:
 #endif /* i386 || _I386 || _i386 || __i386  */
 
 #endif /* !USE_ZLIB */
+.section .note.GNU-stack, "", @progbits
+.previous
