--- /dev/null
+++ b/t4240rdb/a008098.rcw
@@ -0,0 +1,31 @@
+/*
+ * Work-around for erratum A-008098
+ *
+ * Description:
+ * The PCI Express controller may fail to discard a bad packet when there
+ * are excessive, correctable errors resulting in corrupted memory.
+ *
+ * Impact:
+ * While running in an environment with many correctable errors the PCI
+ * Express controller can corrupt memory.
+ *
+ * T4240RDB have only 0x100 bytes for PBI in eeprom. Unused pcie must be deleted.
+ */
+
+.pbi
+write 0x240000, 0x87000048
+write 0x240004, 0x80802040
+write 0x240000, 0x8700004c
+write 0x240004, 0x02c02000
+write 0x240000, 0x87000050
+write 0x240004, 0x00002000
+
+write 0x260000, 0x87000048
+write 0x260004, 0x80802040
+write 0x260000, 0x8700004c
+write 0x260004, 0x02c02000
+write 0x260000, 0x87000050
+write 0x260004, 0x00002000
+
+flush
+.end
--- a/t4240rdb/SSFFPPH_27_55_1_9/rcw_27_55_1_9_1666MHz.rcw
+++ b/t4240rdb/SSFFPPH_27_55_1_9/rcw_27_55_1_9_1666MHz.rcw
@@ -60,4 +60,4 @@ write 0x124010, 0x00000101
 write 0x124130, 0x0000000c
 .end
 
-#include <../t4240qds/a008098.rcw>
+#include <a008098.rcw>
