From 05c745c001c8c82bbba8a6d953ad77ad25c92c5f Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Thu, 7 Nov 2019 14:11:08 +0000
Subject: [PATCH] fixup! clk-raspberrypi: Also support v3d clock

---
 drivers/clk/bcm/clk-raspberrypi.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

--- a/drivers/clk/bcm/clk-raspberrypi.c
+++ b/drivers/clk/bcm/clk-raspberrypi.c
@@ -351,21 +351,21 @@ static const struct clk_ops raspberrypi_
 	.is_prepared = raspberrypi_fw_pll_is_on,
 	.recalc_rate = raspberrypi_fw_pll_get_rate,
 	.set_rate = raspberrypi_fw_pll_set_rate,
-	.determine_rate = raspberrypi_pll_determine_rate,
+	//.determine_rate = raspberrypi_pll_determine_rate,
 };
 
 static const struct clk_ops raspberrypi_firmware_pll_divider_clk_ops = {
 	.is_prepared = raspberrypi_fw_pll_div_is_on,
 	.recalc_rate = raspberrypi_fw_pll_div_get_rate,
 	.set_rate = raspberrypi_fw_pll_div_set_rate,
-	.determine_rate = raspberrypi_pll_div_determine_rate,
+	//.determine_rate = raspberrypi_pll_div_determine_rate,
 };
 
 static const struct clk_ops raspberrypi_firmware_clk_ops = {
 	.is_prepared = raspberrypi_fw_clock_is_on,
 	.recalc_rate = raspberrypi_fw_clock_get_rate,
 	.set_rate = raspberrypi_fw_clock_set_rate,
-	.determine_rate = raspberrypi_clock_determine_rate,
+	//.determine_rate = raspberrypi_clock_determine_rate,
 };
 
 
