--- a/drivers/char/random.c
+++ b/drivers/char/random.c
@@ -275,11 +275,11 @@
 /*
  * Configuration information
  */
-#define INPUT_POOL_SHIFT	12
+#define INPUT_POOL_SHIFT	13
 #define INPUT_POOL_WORDS	(1 << (INPUT_POOL_SHIFT-5))
-#define OUTPUT_POOL_SHIFT	10
+#define OUTPUT_POOL_SHIFT	11
 #define OUTPUT_POOL_WORDS	(1 << (OUTPUT_POOL_SHIFT-5))
-#define SEC_XFER_SIZE		512
+#define SEC_XFER_SIZE		1024
 #define EXTRACT_SIZE		10
 
 #define DEBUG_RANDOM_BOOT 0
