--- a/arch/arm/mach-ixp4xx/wg302v1-setup.c
+++ b/arch/arm/mach-ixp4xx/wg302v1-setup.c
@@ -117,6 +117,35 @@ static struct platform_device *wg302v1_d
 	&wg302v1_eth[0],
 };
 
+static char wg302v1_mem_fixup[] __initdata = " mem=32M";
+
+static void __init wg302v1_fixup(struct tag *tags, char **cmdline,
+				 struct meminfo *mi)
+{
+	struct tag *t = tags;
+	char *p = *cmdline;
+	size_t fixlen, cmdlen;
+
+	/* Find the end of the tags table, taking note of any cmdline tag. */
+	for (; t->hdr.size; t = tag_next(t)) {
+		if (t->hdr.tag == ATAG_CMDLINE) {
+			p = t->u.cmdline.cmdline;
+		}
+	}
+
+	fixlen = strlen(wg302v1_mem_fixup);
+	cmdlen = strlen(p);
+	if (fixlen + cmdlen >= COMMAND_LINE_SIZE)
+		return;
+
+	/* append the fixup to the cmdline */
+	memmove(p + cmdlen, wg302v1_mem_fixup, fixlen + 1);
+
+	/* Adjust the size of the atag if there was one */
+	if (t->hdr.size)
+		t->hdr.size += fixlen;
+}
+
 static void __init wg302v1_init(void)
 {
 	ixp4xx_sys_init();
