--- a/drivers/net/ethernet/xscale/ixp4xx_eth.c
+++ b/drivers/net/ethernet/xscale/ixp4xx_eth.c
@@ -1474,6 +1474,10 @@ static int eth_init_one(struct platform_
 		goto err_free_mem;
 	}
 
+	/* mask with MAC supported features */
+	port->phydev->supported &= PHY_BASIC_FEATURES;
+	port->phydev->advertising = port->phydev->supported;
+
 	port->phydev->irq = PHY_POLL;
 
 	if ((err = register_netdev(dev)))
