--- a/arch/arm/mach-ixp4xx/gtwx5715-setup.c
+++ b/arch/arm/mach-ixp4xx/gtwx5715-setup.c
@@ -27,6 +27,7 @@
 #include <linux/serial.h>
 #include <linux/tty.h>
 #include <linux/serial_8250.h>
+#include <linux/spi/spi_gpio_old.h>
 #include <asm/types.h>
 #include <asm/setup.h>
 #include <asm/memory.h>
@@ -146,9 +147,41 @@ static struct platform_device gtwx5715_f
 	.resource	= &gtwx5715_flash_resource,
 };
 
+static int gtwx5715_spi_boardinfo_setup(struct spi_board_info *bi,
+		struct spi_master *master, void *data)
+{
+
+	strlcpy(bi->modalias, "spi-ks8995", sizeof(bi->modalias));
+
+	bi->max_speed_hz = 5000000 /* Hz */;
+	bi->bus_num = master->bus_num;
+	bi->mode = SPI_MODE_0;
+
+	return 0;
+}
+
+static struct spi_gpio_platform_data gtwx5715_spi_bus_data = {
+	.pin_cs			= GTWX5715_KSSPI_SELECT,
+	.pin_clk		= GTWX5715_KSSPI_CLOCK,
+	.pin_miso		= GTWX5715_KSSPI_RXD,
+	.pin_mosi		= GTWX5715_KSSPI_TXD,
+	.cs_activelow		= 1,
+	.no_spi_delay		= 1,
+	.boardinfo_setup	= gtwx5715_spi_boardinfo_setup,
+};
+
+static struct platform_device gtwx5715_spi_bus = {
+	.name		= "spi-gpio",
+	.id		= 0,
+	.dev		= {
+		.platform_data = &gtwx5715_spi_bus_data,
+	},
+};
+
 static struct platform_device *gtwx5715_devices[] __initdata = {
 	&gtwx5715_uart_device,
 	&gtwx5715_flash,
+	&gtwx5715_spi_bus,
 };
 
 static void __init gtwx5715_init(void)
