--- a/src/lib/ffmpeg/meson.build
+++ b/src/lib/ffmpeg/meson.build
@@ -6,11 +6,7 @@ conf.set('HAVE_LIBAVUTIL', libavutil_dep
 enable_ffmpeg = libavformat_dep.found() and libavcodec_dep.found() and libavutil_dep.found()
 conf.set('ENABLE_FFMPEG', enable_ffmpeg)
 
-if enable_ffmpeg
-  libavfilter_dep = dependency('libavfilter', required: false)
-else
   libavfilter_dep = dependency('', required: false)
-endif
 conf.set('HAVE_LIBAVFILTER', libavfilter_dep.found())
 
 if not enable_ffmpeg
