--- a/drivers/net/mdio/Kconfig
+++ b/drivers/net/mdio/Kconfig
@@ -27,6 +27,13 @@ config OF_MDIO
 	help
 	  OpenFirmware MDIO bus (Ethernet PHY) accessors
 
+config AR40XX_PHY
+   tristate "Driver for Qualcomm Atheros IPQ40XX switches"
+   depends on HAS_IOMEM && OF && OF_MDIO
+   select SWCONFIG
+   help
+      This is the driver for Qualcomm Atheros IPQ40XX ESS switches.
+
 if MDIO_BUS
 
 config MDIO_DEVRES
--- a/drivers/net/mdio/Makefile
+++ b/drivers/net/mdio/Makefile
@@ -21,6 +21,8 @@ obj-$(CONFIG_MDIO_SUN4I)		+= mdio-sun4i.
 obj-$(CONFIG_MDIO_THUNDER)		+= mdio-thunder.o
 obj-$(CONFIG_MDIO_XGENE)		+= mdio-xgene.o
 
+obj-$(CONFIG_AR40XX_PHY)		+= ar40xx.o
+
 obj-$(CONFIG_MDIO_BUS_MUX)		+= mdio-mux.o
 obj-$(CONFIG_MDIO_BUS_MUX_BCM_IPROC)	+= mdio-mux-bcm-iproc.o
 obj-$(CONFIG_MDIO_BUS_MUX_GPIO)		+= mdio-mux-gpio.o
