--- a/examples/ldns-signzone.c
+++ b/examples/ldns-signzone.c
@@ -535,6 +535,7 @@ init_openssl_engine ( const char * const
 static void
 shutdown_openssl ( ENGINE * const e )
 {
+#if OPENSSL_VERSION_NUMBER < 0x10100000 || defined(HAVE_LIBRESSL)
 	if ( e != NULL ) {
 		ENGINE_free ( e );
 		ENGINE_cleanup ();
@@ -544,6 +545,7 @@ shutdown_openssl ( ENGINE * const e )
 	EVP_cleanup ();
 	CRYPTO_cleanup_all_ex_data ();
 	ERR_free_strings ();
+#endif
 }
 #endif
 
