--- a/Makefile.in
+++ b/Makefile.in
@@ -84,12 +84,7 @@ showkey:	showkey.o $(LIBNEWT)
 
 _snack.$(SOEXT):   snack.c $(LIBNEWTSH)
 	@[ -n "$(PYTHONVERS)" ] && for ver in $(PYTHONVERS); do \
-		pyconfig=$$ver-config; \
-		if ! $$pyconfig --cflags > /dev/null 2>&1 && \
-				python-config --cflags > /dev/null 2>&1; then \
-			echo $$pyconfig not found, using python-config; \
-			pyconfig=python-config; \
-		fi; \
+		pyconfig=$(PYTHON_CONFIG_PATH)/$$ver-config; \
 		mkdir -p $$ver; \
 		PCFLAGS=`$$pyconfig --cflags`; \
 		PIFLAGS=`$$pyconfig --includes`; \
@@ -109,7 +104,7 @@ whiptcl.$(SOEXT): $(WHIPTCLOBJS) $(LIBNE
 	$(CC) -shared $(SHCFLAGS) $(LDFLAGS) -o whiptcl.$(SOEXT) $(WHIPTCLOBJS) -L. -lnewt  $(LIBTCL) -lpopt $(LIBS)
 
 $(LIBNEWT): $(LIBOBJS)
-	ar rv $@ $^
+	$(AR) rv $@ $^
 
 newt.o $(SHAREDDIR)/newt.o: newt.c Makefile
 
