--- a/lib/util/Makefile.in
+++ b/lib/util/Makefile.in
@@ -223,10 +223,10 @@ libsudo_util.la: $(LTOBJS) @LT_LDDEP@
 	esac
 
 siglist.c: mksiglist
-	./mksiglist > $@
+	mksiglist > $@
 
 signame.c: mksigname
-	./mksigname > $@
+	mksigname > $@
 
 mksiglist: $(srcdir)/mksiglist.c $(srcdir)/mksiglist.h $(incdir)/sudo_compat.h $(top_builddir)/config.h
 	$(CC) $(CPPFLAGS) $(CFLAGS) $(srcdir)/mksiglist.c -o $@
