--- a/include/nasmlib.h
+++ b/include/nasmlib.h
@@ -188,11 +188,9 @@ int64_t readnum(char *str, bool *error);
 int64_t readstrnum(char *str, int length, bool *warn);
 
 /*
- * seg_init: Initialise the segment-number allocator.
  * seg_alloc: allocate a hitherto unused segment number.
  */
-void pure_func seg_init(void);
-int32_t pure_func seg_alloc(void);
+int32_t seg_alloc(void);
 
 /*
  * many output formats will be able to make use of this: a standard
