diff --git a/utils.c b/utils.c
index 9fe148e..196a04e 100644
--- a/utils.c
+++ b/utils.c
@@ -62,7 +62,9 @@ int red_recv_udp_pkt(int fd, char *buf, size_t buflen, struct sockaddr_in *inadd
             if (
                 cmsg->cmsg_level == SOL_IP &&
                 cmsg->cmsg_type == IP_ORIGDSTADDR &&
-                cmsg->cmsg_len >= CMSG_LEN(sizeof(*toaddr))
+                (cmsg->cmsg_len == CMSG_LEN(sizeof(struct sockaddr_in))
+                 || cmsg->cmsg_len == CMSG_LEN(sizeof(struct sockaddr_in6))) &&
+                cmsg->cmsg_len <= CMSG_LEN(sizeof(*toaddr))
             ) {
                 struct sockaddr_in* cmsgaddr = (struct sockaddr_in*)CMSG_DATA(cmsg);
                 memcpy(toaddr, cmsgaddr, sizeof(*toaddr));
