--- a/kcm/cache.c
+++ b/kcm/cache.c
@@ -102,7 +102,7 @@
     for (p = ccache_head; p != NULL; p = p->next) {
 	if ((p->flags & KCM_FLAGS_VALID) == 0)
 	    continue;
-	if (memcmp(p->uuid, uuid, sizeof(uuid)) == 0) {
+	if (memcmp(p->uuid, uuid, sizeof(*uuid)) == 0) {
 	    ret = 0;
 	    break;
 	}
--- a/lib/hcrypto/engine.c
+++ b/lib/hcrypto/engine.c
@@ -88,7 +88,7 @@
     if(engine->destroy)
 	(*engine->destroy)(engine);
 
-    memset(engine, 0, sizeof(engine));
+    memset(engine, 0, sizeof(*engine));
     engine->references = -1;
 
 
--- a/lib/hcrypto/md2.c
+++ b/lib/hcrypto/md2.c
@@ -130,5 +130,5 @@
     MD2_Update(m, pad, 16);
 
     memcpy(res, m->state, MD2_DIGEST_LENGTH);
-    memset(m, 0, sizeof(m));
+    memset(m, 0, sizeof(*m));
 }
--- a/lib/hcrypto/test_rand.c
+++ b/lib/hcrypto/test_rand.c
@@ -164,7 +164,7 @@
 
 	for (bit = 0; bit < 8; bit++) {
 
-	    res = ((double)abs(len - bits[bit] * 2)) / (double)len;
+	    res = abs(len - bits[bit] * 2) / (double)len;
 	    if (res > 0.005)
 		errx(1, "head%d vs tail%d > 0.5%%%% %lf == %d vs %d",
 		     bit, bit, res, len, bits[bit]);
