--- a/configure.ac
+++ b/configure.ac
@@ -246,7 +246,13 @@ PKG_CHECK_MODULES([check], [check >= 0.9
 
 # Third-party libraries
 lldp_CHECK_LIBEVENT
-lldp_CHECK_LIBCAP
+
+#libcap is not a core package so disable this capability
+#lldp_CHECK_LIBCAP
+libcap_LIBS=
+libcap_CFLAGS=
+AC_SUBST([libcap_LIBS])
+AC_SUBST([libcap_CFLAGS])
 
 # Compatibility with pkg.m4 < 0.27
 m4_ifdef([PKG_INSTALLDIR], [PKG_INSTALLDIR],
