From: Felix Fietkau <nbd@nbd.name>
Date: Sun, 13 Sep 2020 08:27:24 +0200
Subject: [PATCH] net: ethernet: mtk_eth_soc: set PPE flow hash as skb hash
 if present

This improves GRO performance

Signed-off-by: Felix Fietkau <nbd@nbd.name>
---

--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -1318,6 +1318,10 @@ static int mtk_poll_rx(struct napi_struc
 		skb->protocol = eth_type_trans(skb, netdev);
 		bytes += pktlen;
 
+		hash = trxd.rxd4 & GENMASK(13, 0);
+		if (hash != GENMASK(13, 0))
+			skb_set_hash(skb, hash, PKT_HASH_TYPE_L4);
+
 		if (netdev->features & NETIF_F_HW_VLAN_CTAG_RX &&
 		    (trxd.rxd2 & RX_DMA_VTAG))
 			__vlan_hwaccel_put_tag(skb, htons(ETH_P_8021Q),
