--- a/arch/arm64/boot/dts/mediatek/mt7622-rfb1.dts
+++ b/arch/arm64/boot/dts/mediatek/mt7622-rfb1.dts
@@ -589,17 +589,17 @@
 
 			factory: partition@1c0000 {
 				label = "Factory";
-				reg = <0x1c0000 0x0040000>;
+				reg = <0x1c0000 0x0100000>;
 			};
 
 			partition@200000 {
 				label = "firmware";
-				reg = <0x200000 0x2000000>;
+				reg = <0x2c0000 0x2000000>;
 			};
 
 			partition@2200000 {
 				label = "User_data";
-				reg = <0x2200000 0x4000000>;
+				reg = <0x22c0000 0x4000000>;
 			};
 		};
 	};
