--- a/drivers/mtd/ubi/build.c
+++ b/drivers/mtd/ubi/build.c
@@ -900,9 +900,8 @@ int ubi_attach_mtd_dev(struct mtd_info *
 	 * will die soon and you will lose all your data.
 	 */
 	if (mtd->type == MTD_MLCNANDFLASH) {
-		pr_err("ubi: refuse attaching mtd%d - MLC NAND is not supported\n",
+		pr_warn("ubi: mtd%d has been identified as a MLC NAND!! - MLC is not supported by UBI and UBIFS and _WILL_ cause corruption in the near future!\n",
 			mtd->index);
-		return -EINVAL;
 	}
 
 	if (ubi_num == UBI_DEV_NUM_AUTO) {
