From bb1e203d150d2f9f6723e889b8b6f444fad537e1 Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.com>
Date: Fri, 1 Dec 2023 15:59:45 +0000
Subject: [PATCH] dtc update

---
 scripts/dtc/checks.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/scripts/dtc/checks.c
+++ b/scripts/dtc/checks.c
@@ -724,7 +724,7 @@ static void check_alias_paths(struct che
 			continue;
 		}
 		if (strspn(prop->name, LOWERCASE DIGITS "-") != strlen(prop->name))
-			FAIL(c, dti, node, "aliases property name must include only lowercase and '-'");
+		    FAIL(c, dti, node, "aliases property name (%s) must include only lowercase and '-'", prop->name);
 	}
 }
 WARNING(alias_paths, check_alias_paths, NULL);
