diff --git a/pci.c b/pci.c
index ae07f5a..72d542f 100644
--- a/pci.c
+++ b/pci.c
@@ -1495,7 +1495,11 @@ static struct acxpci_device_info acxpci_info_tbl[] __devinitdata = {
 #endif
 
 #ifdef CONFIG_PCI
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 8, 0)
 static DEFINE_PCI_DEVICE_TABLE(acxpci_id_tbl) = {
+#else
+static const struct pci_device_id acxpci_id_tbl[] = {
+#endif
 	{ PCI_VDEVICE(TI, PCI_DEVICE_ID_TI_TNETW1100A),
 	  .driver_data = CHIPTYPE_ACX100,
 	},
