diff -Naur a/gdb/nat/aarch64-sve-linux-ptrace.h b/gdb/nat/aarch64-sve-linux-ptrace.h
--- a/gdb/nat/aarch64-sve-linux-ptrace.h	2018-09-05 15:27:13.000000000 +0800
+++ b/gdb/nat/aarch64-sve-linux-ptrace.h	2020-05-17 11:18:55.994338014 +0800
@@ -20,15 +20,11 @@
 #ifndef AARCH64_SVE_LINUX_PTRACE_H
 #define AARCH64_SVE_LINUX_PTRACE_H
 
-#include <asm/sigcontext.h>
+#include <signal.h>
 #include <sys/utsname.h>
 #include <sys/ptrace.h>
 #include <asm/ptrace.h>
 
-#ifndef SVE_SIG_ZREGS_SIZE
-#include "aarch64-sve-linux-sigcontext.h"
-#endif
-
 /* Indicates whether a SVE ptrace header is followed by SVE registers or a
    fpsimd structure.  */
 
