--- a/drivers/mtd/parsers/Kconfig
+++ b/drivers/mtd/parsers/Kconfig
@@ -204,3 +204,9 @@ config MTD_ROUTERBOOT_PARTS
 	 flash, some of which are fixed and some of which are located at
 	 variable offsets. This parser handles both cases via properly
 	 formatted DTS.
+
+config MTD_SERCOMM_PARTS
+	tristate "SERCOMM partitioning information support"
+	depends on MTD_OF_PARTS
+	help
+	  This provides partition table parser for SERCOMM partition map
--- a/drivers/mtd/parsers/Makefile
+++ b/drivers/mtd/parsers/Makefile
@@ -14,3 +14,4 @@ obj-$(CONFIG_MTD_PARSER_TRX)		+= parser_
 obj-$(CONFIG_MTD_SHARPSL_PARTS)		+= sharpslpart.o
 obj-$(CONFIG_MTD_REDBOOT_PARTS)		+= redboot.o
 obj-$(CONFIG_MTD_ROUTERBOOT_PARTS)		+= routerbootpart.o
+obj-$(CONFIG_MTD_SERCOMM_PARTS)		+= scpart.o
