--- a/src/bin/wave.c
+++ b/src/bin/wave.c
@@ -168,7 +168,7 @@ unsigned char wave_open(const char *fnam
 }
 
 #ifdef SHINE_BIG_ENDIAN
-#if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 2))
+#if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 8))
 #define bswap_16(x) __builtin_bswap16(x)
 #else
 #define bswap_16(x) ((((x) >> 8) & 0xff) | (((x) & 0xff) << 8))
