--- a/gcc/gcc.c
+++ b/gcc/gcc.c
@@ -9198,8 +9198,10 @@ getenv_spec_function (int argc, const ch
     value = varname;
 
   if (!value)
-    fatal_error (input_location,
-		 "environment variable %qs not defined", varname);
+    {
+      warning (input_location, "environment variable %qs not defined", varname);
+      value = "";
+    }
 
   /* We have to escape every character of the environment variable so
      they are not interpreted as active spec characters.  A
