From: Giuseppe Lippolis <giu.lippolis@gmail.com>
Subject: Add the linux,spidev compatible in spidev Several device in ramips have this binding in the dts

Signed-off-by: Giuseppe Lippolis <giu.lippolis@gmail.com>
---
 drivers/spi/spidev.c | 1 +
 1 file changed, 1 insertion(+)

--- a/drivers/spi/spidev.c
+++ b/drivers/spi/spidev.c
@@ -717,6 +717,7 @@ static const struct of_device_id spidev_
 	{ .compatible = "menlo,m53cpld", .data = &spidev_of_check },
 	{ .compatible = "cisco,spi-petra", .data = &spidev_of_check },
 	{ .compatible = "micron,spi-authenta", .data = &spidev_of_check },
+	{ .compatible = "siliconlabs,si3210" },
 	{},
 };
 MODULE_DEVICE_TABLE(of, spidev_dt_ids);
