From 31514742de07d595ac23c2b0abf0e092f8b26140 Mon Sep 17 00:00:00 2001
From: Yousong Zhou <zhouyousong@yunionyun.com>
Date: Tue, 21 Aug 2018 13:02:21 +0000
Subject: [PATCH] build: trim build

Signed-off-by: Yousong Zhou <zhouyousong@yunionyun.com>
---
 Makefile.am | 2 --
 1 file changed, 2 deletions(-)

--- a/Makefile.am
+++ b/Makefile.am
@@ -475,12 +475,10 @@ dist-docs:
 	VERSION=$(VERSION) MAKE='$(MAKE)' $(srcdir)/build-aux/dist-docs $(srcdir) $(docs)
 .PHONY: dist-docs
 
-include Documentation/automake.mk
 include m4/automake.mk
 include lib/automake.mk
 include ofproto/automake.mk
 include utilities/automake.mk
-include tests/automake.mk
 include include/automake.mk
 include third-party/automake.mk
 include debian/automake.mk
