From e678a6319a0b122e9b22bcf06a17444a9b63b599 Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.com>
Date: Thu, 29 Oct 2020 14:10:56 +0000
Subject: [PATCH 0261/1085] phy: broadcom: Add bcm54213pe configuration

Signed-off-by: Phil Elwell <phil@raspberrypi.com>
---
 drivers/net/phy/broadcom.c | 8 ++++++++
 1 file changed, 8 insertions(+)

--- a/drivers/net/phy/broadcom.c
+++ b/drivers/net/phy/broadcom.c
@@ -101,6 +101,11 @@ static int bcm54210e_config_init(struct
 	return 0;
 }
 
+static int bcm54213pe_config_init(struct phy_device *phydev)
+{
+	return bcm54210e_config_init(phydev);
+}
+
 static int bcm54612e_config_init(struct phy_device *phydev)
 {
 	int reg;
@@ -391,6 +396,9 @@ static int bcm54xx_config_init(struct ph
 	case PHY_ID_BCM54612E:
 		err = bcm54612e_config_init(phydev);
 		break;
+	case PHY_ID_BCM54213PE:
+		err = bcm54213pe_config_init(phydev);
+		break;
 	case PHY_ID_BCM54616S:
 		err = bcm54616s_config_init(phydev);
 		break;
