From 28dacfcbc4d10579988f2c64e6c7055899680dd5 Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.com>
Date: Tue, 7 Mar 2023 12:12:36 +0000
Subject: [PATCH 0418/1085] hwmon: (ds1621) Add DT compatible strings

Signed-off-by: Phil Elwell <phil@raspberrypi.com>
---
 drivers/hwmon/ds1621.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

--- a/drivers/hwmon/ds1621.c
+++ b/drivers/hwmon/ds1621.c
@@ -378,6 +378,16 @@ static const struct i2c_device_id ds1621
 };
 MODULE_DEVICE_TABLE(i2c, ds1621_id);
 
+static const struct of_device_id ds1621_of_ids[] = {
+	{ .compatible = "dallas,ds1621", },
+	{ .compatible = "dallas,ds1625", },
+	{ .compatible = "dallas,ds1631", },
+	{ .compatible = "dallas,ds1721", },
+	{ .compatible = "dallas,ds1731", },
+	{ }
+};
+MODULE_DEVICE_TABLE(of, ds1621_of_ids);
+
 /* This is the driver that will be inserted */
 static struct i2c_driver ds1621_driver = {
 	.class		= I2C_CLASS_HWMON,
