From c23bd4a7b5c897a4a299c16ff54495207c335864 Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.com>
Date: Tue, 7 Mar 2023 21:47:58 +0000
Subject: [PATCH 0420/1085] hwmon: (sht3x) Add DT compatible string

Signed-off-by: Phil Elwell <phil@raspberrypi.com>
---
 drivers/hwmon/sht3x.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

--- a/drivers/hwmon/sht3x.c
+++ b/drivers/hwmon/sht3x.c
@@ -911,8 +911,18 @@ static int sht3x_probe(struct i2c_client
 	return PTR_ERR_OR_ZERO(hwmon_dev);
 }
 
+static const struct of_device_id sht3x_of_ids[] = {
+	{ .compatible = "sensirion,sht3x" },
+	{ .compatible = "sensirion,sts3x" },
+	{}
+};
+MODULE_DEVICE_TABLE(of, sht3x_of_ids);
+
 static struct i2c_driver sht3x_i2c_driver = {
-	.driver.name = "sht3x",
+	.driver = {
+		.name = "sht3x",
+		.of_match_table = sht3x_of_ids,
+	},
 	.probe       = sht3x_probe,
 	.id_table    = sht3x_ids,
 };
