From 6bbf82441c1c75191af97f92daad5eff8302788f Mon Sep 17 00:00:00 2001
From: Lee Jackson <info@arducam.com>
Date: Thu, 21 Apr 2022 14:18:44 +0800
Subject: [PATCH 0367/1085] media: dt-bindings: media: i2c: Add Arducam
 Pivariety Series CMOS sensor binding

Add YAML device tree binding for Arducam Pivariety CMOS image sensor, and
the relevant MAINTAINERS entries.

Signed-off-by: Lee Jackson <info@arducam.com>
---
 .../bindings/media/i2c/arducam-pivariety.yaml | 112 ++++++++++++++++++
 MAINTAINERS                                   |   8 ++
 2 files changed, 120 insertions(+)
 create mode 100644 Documentation/devicetree/bindings/media/i2c/arducam-pivariety.yaml

--- /dev/null
+++ b/Documentation/devicetree/bindings/media/i2c/arducam-pivariety.yaml
@@ -0,0 +1,112 @@
+# SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/media/i2c/arducam-pivariety.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: Arducam Pivariety Series CMOS Digital Image Sensor
+
+maintainers:
+  - Lee Jackson <info@arducam.com>
+
+description: |-
+  Arducam Pivariety series cameras make compatibility layers for various CMOS
+  sensors and provide a unified command interface. It is programmable through
+  I2C interface. The I2C address is fixed to 0x0C. Image data is sent through
+  MIPI CSI-2, which is configured as either 1, 2 or 4 data lanes.
+
+properties:
+  compatible:
+    const: arducam,arducam-pivariety
+
+  reg:
+    description: I2C device address
+    maxItems: 1
+
+  clocks:
+    maxItems: 1
+
+  VDIG-supply:
+    description:
+      Digital I/O voltage supply, 1.05 volts
+
+  VANA-supply:
+    description:
+      Analog voltage supply, 2.8 volts
+
+  VDDL-supply:
+    description:
+      Digital core voltage supply, 1.8 volts
+
+  reset-gpios:
+    description: |-
+      Reference to the GPIO connected to the xclr pin, if any.
+      Must be released (set high) after all supplies and INCK are applied.
+
+  # See ../video-interfaces.txt for more details
+  port:
+    type: object
+    properties:
+      endpoint:
+        type: object
+        properties:
+          data-lanes:
+            description: |-
+              The sensor supports either two-lane, or four-lane operation.
+              For two-lane operation the property must be set to <1 2>.
+            items:
+              - const: 1
+              - const: 2
+
+          clock-noncontinuous:
+            type: boolean
+            description: |-
+              MIPI CSI-2 clock is non-continuous if this property is present,
+              otherwise it's continuous.
+
+          link-frequencies:
+            allOf:
+              - $ref: /schemas/types.yaml#/definitions/uint64-array
+            description:
+              Allowed data bus frequencies.
+
+        required:
+          - link-frequencies
+
+required:
+  - compatible
+  - reg
+  - clocks
+  - VANA-supply
+  - VDIG-supply
+  - VDDL-supply
+  - port
+
+additionalProperties: false
+
+examples:
+  - |
+    i2c0 {
+        #address-cells = <1>;
+        #size-cells = <0>;
+
+        arducam_pivariety: sensor@0c {
+            compatible = "arducam,arducam-pivariety";
+            reg = <0x0c>;
+            clocks = <&arducam_pivariety_clk>;
+            VANA-supply = <&arducam_pivariety_vana>;   /* 2.8v */
+            VDIG-supply = <&arducam_pivariety_vdig>;   /* 1.05v */
+            VDDL-supply = <&arducam_pivariety_vddl>;   /* 1.8v */
+
+            port {
+                arducam_pivariety_0: endpoint {
+                    remote-endpoint = <&csi1_ep>;
+                    data-lanes = <1 2>;
+                    clock-noncontinuous;
+                    link-frequencies = /bits/ 64 <493500000>;
+                };
+            };
+        };
+    };
+
+...
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -1563,6 +1563,14 @@ S:	Maintained
 F:	drivers/net/arcnet/
 F:	include/uapi/linux/if_arcnet.h
 
+ARDUCAM PIVARIETY SENSOR DRIVER
+M:	Arducam Kernel Maintenance <info@arducam.com>
+L:	linux-media@vger.kernel.org
+S:	Maintained
+T:	git git://linuxtv.org/media_tree.git
+F:	Documentation/devicetree/bindings/media/i2c/arducam-pivariety.yaml
+F:	drivers/media/i2c/arducam-pivariety.c
+
 ARM AND ARM64 SoC SUB-ARCHITECTURES (COMMON PARTS)
 M:	Arnd Bergmann <arnd@arndb.de>
 M:	Olof Johansson <olof@lixom.net>
