From ab2e08485d3018ad5625f4fcac39b8064fa59f36 Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.com>
Date: Fri, 15 Mar 2024 15:09:08 +0000
Subject: [PATCH 0968/1085] fixup! overlays: i2c-mux: Add 'base' parameter

---
 drivers/i2c/i2c-mux.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/i2c/i2c-mux.c
+++ b/drivers/i2c/i2c-mux.c
@@ -359,7 +359,7 @@ int i2c_mux_add_adapter(struct i2c_mux_c
 		u32 reg;
 
 		of_property_read_u32(dev_node, "base-nr", &base_nr);
-		if (!force_nr)
+		if (!force_nr && base_nr)
 			force_nr = base_nr + chan_id;
 
 		if (muxc->arbitrator)
