From 8649c7deae7ed7a33ef6ba630d9df2e23d1ca24f Mon Sep 17 00:00:00 2001
From: Dom Cobley <popcornmix@gmail.com>
Date: Fri, 7 Jul 2023 20:16:06 +0100
Subject: [PATCH 0570/1085] dt-bindings: rtc: new binding for Raspberry Pi RTC
 driver

Add binding for the new RTC driver for Raspberry Pi.
This platform has an RTC managed by firmware, and this RTC
driver provides the simple mailbox interface to access it.

Signed-off-by: Dom Cobley <popcornmix@gmail.com>
---
 .../devicetree/bindings/rtc/rtc-rpi.txt         | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)
 create mode 100644 Documentation/devicetree/bindings/rtc/rtc-rpi.txt

--- /dev/null
+++ b/Documentation/devicetree/bindings/rtc/rtc-rpi.txt
@@ -0,0 +1,17 @@
+* Raspberry Pi RTC
+
+This is a Linux interface to an RTC managed by firmware, hence it's
+virtual from a Linux perspective.
+
+The interface uses the firmware mailbox api to access the RTC registers.
+
+Required properties:
+compatible: should be "raspberrypi,rpi-rtc"
+firmware:   Reference to the RPi firmware device node.
+
+Example:
+
+	rpi_rtc: rpi_rtc {
+		compatible = "raspberrypi,rpi-rtc";
+		firmware = <&firmware>;
+	};
