From e51e9120a35d0b1600e4337e6ecbd020c074af80 Mon Sep 17 00:00:00 2001
From: Maxime Ripard <maxime@cerno.tech>
Date: Mon, 6 Jun 2022 11:02:16 +0200
Subject: [PATCH 0039/1085] arm64: setup: Fix build warning

Signed-off-by: Maxime Ripard <maxime@cerno.tech>
---
 arch/arm64/kernel/setup.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/arch/arm64/kernel/setup.c
+++ b/arch/arm64/kernel/setup.c
@@ -225,9 +225,9 @@ static void __init request_standard_reso
 	size_t res_size;
 
 	kernel_code.start   = __pa_symbol(_stext);
-	kernel_code.end     = __pa_symbol(__init_begin - 1);
+	kernel_code.end     = __pa_symbol(__init_begin) - 1;
 	kernel_data.start   = __pa_symbol(_sdata);
-	kernel_data.end     = __pa_symbol(_end - 1);
+	kernel_data.end     = __pa_symbol(_end) - 1;
 	insert_resource(&iomem_resource, &kernel_code);
 	insert_resource(&iomem_resource, &kernel_data);
 
