From 9c38b74db3b289d276384d63efbc5b09ae3d0d80 Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.com>
Date: Tue, 7 Mar 2023 21:49:30 +0000
Subject: [PATCH 0422/1085] iio: light: veml6070: Add DT compatible string

Signed-off-by: Phil Elwell <phil@raspberrypi.com>
---
 drivers/iio/light/veml6070.c | 7 +++++++
 1 file changed, 7 insertions(+)

--- a/drivers/iio/light/veml6070.c
+++ b/drivers/iio/light/veml6070.c
@@ -194,9 +194,16 @@ static const struct i2c_device_id veml60
 };
 MODULE_DEVICE_TABLE(i2c, veml6070_id);
 
+static const struct of_device_id veml6070_of_id[] = {
+	{ .compatible = "vishay,veml6070" },
+	{ }
+};
+MODULE_DEVICE_TABLE(of, veml6070_of_id);
+
 static struct i2c_driver veml6070_driver = {
 	.driver = {
 		.name   = VEML6070_DRV_NAME,
+		.of_match_table = veml6070_of_id,
 	},
 	.probe = veml6070_probe,
 	.remove  = veml6070_remove,
