From 89bac9679487d1382a883185c3b028359a9fcc8b Mon Sep 17 00:00:00 2001
From: hdoverobinson <hdoverobinson@gmail.com>
Date: Tue, 13 Mar 2018 06:58:39 -0400
Subject: [PATCH 0119/1085] added capture_clear option to pps-gpio via
 dtoverlay (#2433)

---
 drivers/pps/clients/pps-gpio.c | 3 +++
 1 file changed, 3 insertions(+)

--- a/drivers/pps/clients/pps-gpio.c
+++ b/drivers/pps/clients/pps-gpio.c
@@ -113,6 +113,9 @@ static int pps_gpio_setup(struct device
 	data->assert_falling_edge =
 		device_property_read_bool(dev, "assert-falling-edge");
 
+	data->capture_clear =
+		device_property_read_bool(dev, "capture-clear");
+
 	data->echo_pin = devm_gpiod_get_optional(dev, "echo", GPIOD_OUT_LOW);
 	if (IS_ERR(data->echo_pin))
 		return dev_err_probe(dev, PTR_ERR(data->echo_pin),
