From 7c8b067eee519acaa8e9052771e26d4252fa7ad9 Mon Sep 17 00:00:00 2001
From: Dom Cobley <popcornmix@gmail.com>
Date: Tue, 14 Mar 2023 16:54:37 +0000
Subject: [PATCH 0424/1085] ad5398_vcm: Fixup ad5398_probe prototype

---
 drivers/media/i2c/ad5398_vcm.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

--- a/drivers/media/i2c/ad5398_vcm.c
+++ b/drivers/media/i2c/ad5398_vcm.c
@@ -253,8 +253,7 @@ static int ad5398_regulator_notifier(str
 	return NOTIFY_OK;
 }
 
-static int ad5398_probe(struct i2c_client *client,
-			const struct i2c_device_id *devid)
+static int ad5398_probe(struct i2c_client *client)
 {
 	struct ad5398_device *coil;
 	int ret;
