From f1076a9d7a269d72b6707283560d0d38203cb07a Mon Sep 17 00:00:00 2001
From: KubaTaba1uga <luna7337anul@proton.me>
Date: Tue, 4 Mar 2025 09:50:30 +0100
Subject: [PATCH] delete the comment

---
 drivers/dma-buf/heaps/system_heap.c | 4 ----
 1 file changed, 4 deletions(-)

--- a/drivers/dma-buf/heaps/system_heap.c
+++ b/drivers/dma-buf/heaps/system_heap.c
@@ -50,10 +50,6 @@ static gfp_t order_flags[] = {HIGH_ORDER
  * to match with the sizes often found in IOMMUs. Using order 4 pages instead
  * of order 0 pages can significantly improve the performance of many IOMMUs
  * by reducing TLB pressure and time spent updating page tables.
- *
- * Note: `module_max_order` must be set explicitly instead of using
- * `orders[0]` to avoid Clang's "initializer element is not a
- * compile-time constant" error.
  */
 #define MAX_ORDERS_VALUE 8
 static const unsigned int orders[] = {MAX_ORDERS_VALUE, 4, 0};
