From 07b87bf513ffe58ce8dd00eb2edf68b5b9bde518 Mon Sep 17 00:00:00 2001
From: Robert Marko <robimarko@gmail.com>
Date: Mon, 12 May 2025 17:10:38 +0200
Subject: [PATCH] hal: syn: fix missing prototypes

These Synopsys HAL ops were not covered under the previous missing
prototype commit, so cover them now.

Signed-off-by: Robert Marko <robimarko@gmail.com>
---
 hal/dp_ops/syn_gmac_dp/syn_dp_tx.c | 6 +++---
 hal/gmac_ops/syn/gmac/syn_if.c     | 2 +-
 2 files changed, 4 insertions(+), 4 deletions(-)

--- a/hal/dp_ops/syn_gmac_dp/syn_dp_tx.c
+++ b/hal/dp_ops/syn_gmac_dp/syn_dp_tx.c
@@ -117,7 +117,7 @@ static inline struct dma_desc_tx *syn_dp
  * syn_dp_tx_nr_frags()
  *	TX routine for Synopsys GMAC SG packets with nr frags
  */
-int syn_dp_tx_nr_frags(struct syn_dp_info_tx *tx_info, struct sk_buff *skb)
+static int syn_dp_tx_nr_frags(struct syn_dp_info_tx *tx_info, struct sk_buff *skb)
 {
 	dma_addr_t dma_addr;
 	unsigned int length = skb_headlen(skb);
@@ -209,7 +209,7 @@ int syn_dp_tx_nr_frags(struct syn_dp_inf
  * syn_dp_tx_frag_list()
  *	TX routine for Synopsys GMAC SG packets with frag lists
  */
-int syn_dp_tx_frag_list(struct syn_dp_info_tx *tx_info, struct sk_buff *skb)
+static int syn_dp_tx_frag_list(struct syn_dp_info_tx *tx_info, struct sk_buff *skb)
 {
 	dma_addr_t dma_addr;
 	struct sk_buff *iter_skb;
@@ -508,7 +508,7 @@ int syn_dp_tx_complete(struct syn_dp_inf
  * syn_dp_tx_sg()
  *	Tx routine for Synopsys GMAC scatter gather packets
  */
-int syn_dp_tx_sg(struct syn_dp_info_tx *tx_info, struct sk_buff *skb)
+static int syn_dp_tx_sg(struct syn_dp_info_tx *tx_info, struct sk_buff *skb)
 {
 	struct net_device *netdev = tx_info->netdev;
 	bool has_frag_list;
--- a/hal/gmac_ops/syn/gmac/syn_if.c
+++ b/hal/gmac_ops/syn/gmac/syn_if.c
@@ -455,7 +455,7 @@ static int32_t syn_set_max_frame_size(st
 /*
  * syn_get_mib_stats()
  */
-int syn_get_mib_stats(struct nss_gmac_hal_dev *nghd, fal_mib_counter_t *stats)
+static int syn_get_mib_stats(struct nss_gmac_hal_dev *nghd, fal_mib_counter_t *stats)
 {
 	if (fal_mib_counter_get(0, nghd->mac_id, stats) < 0) {
 		return -1;
