From: Chuanhong Guo <gch981213@gmail.com>
Date: Sat, 14 Sep 2024 16:51:36 +0800
Subject: [PATCH 15/20] reset: add support for sf21a6826/sf21h8898

---
 drivers/reset/Kconfig  | 5 +++++
 drivers/reset/Makefile | 1 +
 2 files changed, 6 insertions(+)

--- a/drivers/reset/Kconfig
+++ b/drivers/reset/Kconfig
@@ -250,6 +250,11 @@ config RESET_SF19A2890_PERIPH
 	  This enables reset controller driver for peripheral reset blocks
 	  found on Siflower SF19A2890 SoC.
 
+config RESET_SF21
+	tristate "Siflower SF21A6826/SF21H8898 Reset Controller Driver"
+	help
+	  This enables the reset controller driver for Siflower SF21A6826/SF21H8898.
+
 config RESET_SIMPLE
 	bool "Simple Reset Controller Driver" if COMPILE_TEST || EXPERT
 	default ARCH_ASPEED || ARCH_BCMBCA || ARCH_BITMAIN || ARCH_REALTEK || ARCH_SOPHGO || ARCH_STM32 || (ARCH_INTEL_SOCFPGA && ARM64) || ARCH_SUNXI || ARC
--- a/drivers/reset/Makefile
+++ b/drivers/reset/Makefile
@@ -33,6 +33,7 @@ obj-$(CONFIG_RESET_RASPBERRYPI) += reset
 obj-$(CONFIG_RESET_RZG2L_USBPHY_CTRL) += reset-rzg2l-usbphy-ctrl.o
 obj-$(CONFIG_RESET_SCMI) += reset-scmi.o
 obj-$(CONFIG_RESET_SF19A2890_PERIPH) += reset-sf19a2890-periph.o
+obj-$(CONFIG_RESET_SF21) += reset-sf21.o
 obj-$(CONFIG_RESET_SIMPLE) += reset-simple.o
 obj-$(CONFIG_RESET_SOCFPGA) += reset-socfpga.o
 obj-$(CONFIG_RESET_SUNPLUS) += reset-sunplus.o
