From 3232e537ccaba4417b25d9d70264e4a5533042da Mon Sep 17 00:00:00 2001
From: Nicholas Marriott <nicholas.marriott@gmail.com>
Date: Mon, 18 Mar 2019 13:04:00 +0000
Subject: [PATCH] Fix bas64 declarations, from makepost at firemail dot cc.

---
 fdm.h | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

--- a/fdm.h
+++ b/fdm.h
@@ -20,7 +20,6 @@
 #define FDM_H
 
 #include <sys/param.h>
-#include <sys/cdefs.h>
 #include <sys/stat.h>
 
 #ifdef HAVE_QUEUE_H
@@ -725,8 +724,8 @@ size_t		 strlcat(char *, const char *, s
 
 #ifndef HAVE_B64_NTOP
 /* base64.c */
-int b64_ntop(src, srclength, target, targsize);
-int b64_pton(src, target, targsize);
+int b64_ntop(u_char const *, size_t, char *, size_t);
+int b64_pton(char const *, u_char *, size_t);
 #endif
 
 /* shm.c */
