--- a/src/main.cpp
+++ b/src/main.cpp
@@ -422,12 +422,14 @@ int main(int argc, char** argv) {
                 "No modules found. Perhaps you didn't install ZNC properly?");
             CUtils::PrintError(
                 "Read https://wiki.znc.in/Installation for instructions.");
+#if 0
             if (!CUtils::GetBoolInput(
                     "Do you really want to run ZNC without any modules?",
                     false)) {
                 CZNC::DestroyInstance();
                 return 1;
             }
+#endif
         }
         CUtils::PrintStatus(true, "");
     }
