diff -ruNp trunk.orig/config.h trunk/config.h
--- trunk.orig/config.h	2018-03-25 14:31:53.147349681 +0300
+++ trunk/config.h	2018-03-25 14:46:42.842348711 +0300
@@ -22,7 +22,7 @@
 //#define IPV6SUPPORT 1
 #define MODULE_MONITOR 1
 
-//#define MODULE_CAMD33 1
+#define MODULE_CAMD33 1
 #define MODULE_CAMD35 1
 #define MODULE_CAMD35_TCP 1
 #define MODULE_NEWCAMD 1
@@ -62,10 +62,14 @@
 #define CARDREADER_STINGER 1
 #define CARDREADER_DRECAS 1
 
+#define WITH_PCSC 1
+
 #ifdef WITH_PCSC
 #define CARDREADER_PCSC 1
 #endif
 
+#define WITH_LIBUSB 1
+
 #ifdef WITH_LIBUSB
 #define CARDREADER_SMART 1
 #endif
@@ -83,11 +87,11 @@
 #endif
 
 #ifdef WITH_STAPI
-#define CARDREADER_STAPI 1
+//#define CARDREADER_STAPI 1
 #endif
 
 #ifdef WITH_STAPI5
-#define CARDREADER_STAPI5 1
+//#define CARDREADER_STAPI5 1
 #endif
 
 #ifdef READER_DRE
