--- a/unixos.c
+++ b/unixos.c
@@ -25,6 +25,7 @@
 #include <stdio.h>
 #include <ctype.h>
 #include <string.h>
+#include <stdlib.h>
 #include <errno.h>
 #include <sys/types.h>
 #include <sys/param.h>
@@ -39,8 +40,6 @@
 #endif
 
 extern int errno;
-extern char *malloc();
-extern char *getenv();
 
 int overwrite_files = 0;
 int didchat;
--- a/xmalloc.c
+++ b/xmalloc.c
@@ -24,7 +24,7 @@
  */
 #include <stdio.h>
 #include <string.h>
-extern char *malloc(), *realloc();
+#include <stdlib.h>
 
 char *xmalloc (int size)
 {
