commit bb6242675ad0c7447daef148fffced882e5b4a61
Author: Nicholas Marriott <nicholas.marriott@gmail.com>
Date:   Thu Apr 15 06:45:19 2021 +0100

    Add crosscompiling fallbacks, from Hasso Tepper.

--- a/configure.ac
+++ b/configure.ac
@@ -163,6 +163,7 @@ AC_RUN_IFELSE([AC_LANG_PROGRAM(
 		[return (reallocarray(NULL, 1, 1) == NULL);]
 	)],
 	AC_MSG_RESULT(yes),
+	[AC_LIBOBJ(reallocarray) AC_MSG_RESULT([no])],
 	[AC_LIBOBJ(reallocarray) AC_MSG_RESULT([no])]
 )
 AC_MSG_CHECKING([for working recallocarray])
@@ -171,6 +172,7 @@ AC_RUN_IFELSE([AC_LANG_PROGRAM(
 		[return (recallocarray(NULL, 1, 1, 1) == NULL);]
 	)],
 	AC_MSG_RESULT(yes),
+	[AC_LIBOBJ(recallocarray) AC_MSG_RESULT([no])],
 	[AC_LIBOBJ(recallocarray) AC_MSG_RESULT([no])]
 )
 
