--- a/ibrcommon/data/File.cpp
+++ b/ibrcommon/data/File.cpp
@@ -35,9 +35,7 @@
 #include <cerrno>
 #include <fstream>
 
-#if !defined(HAVE_FEATURES_H) || defined(ANDROID)
 #include <libgen.h>
-#endif
 
 #ifdef __WIN32__
 #include <io.h>
@@ -226,7 +224,7 @@ namespace ibrcommon
 	std::string File::getBasename() const
 	{
 #if !defined(ANDROID) && defined(HAVE_FEATURES_H)
-		return std::string(basename(_path.c_str()));
+		return std::string(basename((char *)_path.c_str()));
 #else
 		char path[_path.length()+1];
 		::memcpy(&path, _path.c_str(), _path.length()+1);
