--- a/src/proxy/app.c
+++ b/src/proxy/app.c
@@ -391,7 +391,7 @@ dnscrypt_proxy_main(int argc, char *argv
     }
     logger_noformat(&proxy_context, LOG_NOTICE, "Starting " PACKAGE_STRING);
     sodium_mlock(&proxy_context, sizeof proxy_context);
-    randombytes_set_implementation(&randombytes_salsa20_implementation);
+    randombytes_set_implementation(&randombytes_internal_implementation);
 
 #ifdef PLUGINS
     if (plugin_support_context_load(app_context.dcps_context) != 0) {
