Index: genext2fs/m4/ac_func_scanf_can_malloc.m4
===================================================================
--- genext2fs.orig/m4/ac_func_scanf_can_malloc.m4	2011-09-03 21:28:49.000000000 +0200
+++ genext2fs/m4/ac_func_scanf_can_malloc.m4	2011-09-03 21:29:41.000000000 +0200
@@ -9,7 +9,7 @@
 # --------------------------------------
 AC_DEFUN([AC_FUNC_SCANF_CAN_MALLOC],
   [ AC_CHECK_HEADERS([stdlib.h])
-    AC_CACHE_CHECK([whether scanf can malloc], [ac_scanf_can_malloc],
+    AC_CACHE_CHECK([whether scanf can malloc], [ac_cv_func_scanf_can_malloc],
     [ AC_RUN_IFELSE(
       [ AC_LANG_PROGRAM(
         [
