--- a/cf/cf/Makefile
+++ b/cf/cf/Makefile
@@ -80,10 +80,10 @@ install:
 install-cf:  install-sendmail-cf install-submit-cf
 
 install-sendmail-cf: $(CF).cf
-	$(INSTALL) -c -o $(CFOWN) -g $(CFGRP) -m $(CFMODE) $(CF).cf ${DESTDIR}$(MAILDIR)/sendmail.cf
+	$(INSTALL) -c -m $(CFMODE) $(CF).cf ${DESTDIR}$(MAILDIR)/sendmail.cf
 
 install-submit-cf: $(SUBMIT).cf
-	$(INSTALL) -c -o $(CFOWN) -g $(CFGRP) -m $(CFMODE) $(SUBMIT).cf ${DESTDIR}$(MAILDIR)/submit.cf
+	$(INSTALL) -c -m $(CFMODE) $(SUBMIT).cf ${DESTDIR}$(MAILDIR)/submit.cf
 
 depend:
 
