From e316b4c7d421afa6ca5f6b6ae86d0a8219a9bd56 Mon Sep 17 00:00:00 2001
From: Ioana Ciornei <ioana.ciornei@nxp.com>
Date: Tue, 24 Oct 2017 16:29:42 +0000
Subject: [PATCH 03/12] scripts: remove unnecessary arrays

Signed-off-by: Ioana Ciornei <ioana.ciornei@nxp.com>
---
 scripts/ls-main | 22 ++--------------------
 1 file changed, 2 insertions(+), 20 deletions(-)

diff --git a/scripts/ls-main b/scripts/ls-main
index 09bef5a..ea0df92 100755
--- a/scripts/ls-main
+++ b/scripts/ls-main
@@ -87,8 +87,6 @@ root_c=
 # Type of endpoint object
 toe=
 
-dpnis=()
-dpmacs=()
 containers=()
 # Full path containers
 fpc=()
@@ -944,16 +942,8 @@ process_listni() {
 				fi
 			fi
 
-			dpnis+=("$dpni $details")
+			echo "${i}/${dpni} ${details}"
 		done
-
-		if [ -n "$dpnis" ]; then
-			for j in "${dpnis[@]}"; do
-				echo "$i/$j"
-			done
-
-			dpnis=()
-		fi
 	done
 }
 
@@ -984,16 +974,8 @@ process_listmac() {
 				fi
 			fi
 
-			dpmacs+=("$dpmac $details")
+			echo "${i}/${dpmac} ${details}"
 		done
-
-		if [ -n "$dpmacs" ]; then
-			for j in "${dpmacs[@]}"; do
-				echo "$i/$j"
-			done
-
-			dpmacs=()
-		fi
 	done
 }
 
-- 
2.14.1

