From 899d6c74286cbc672e1bad1abd7eff15b1b7298d Mon Sep 17 00:00:00 2001
From: Ioana Ciornei <ioana.ciornei@nxp.com>
Date: Tue, 24 Oct 2017 16:29:49 +0000
Subject: [PATCH 05/12] scripts: use proper arithmetic operation syntax

Signed-off-by: Ioana Ciornei <ioana.ciornei@nxp.com>
---
 scripts/ls-main | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/scripts/ls-main b/scripts/ls-main
index 72f6c77..a39df2c 100755
--- a/scripts/ls-main
+++ b/scripts/ls-main
@@ -966,7 +966,7 @@ fi
 
 # Check compatibility with MC version
 mc_major=$($restool --mc-version | cut -f2 -d':' | cut -f1 -d'.' | tr -d ' ')
-if (( $mc_major != 10 )); then
+if [ $mc_major != 10 ]; then
 	echo "Restool wrapper scripts only support the latest major MC version\n"
 	echo "that currently is MC10.x. Use with caution."
 fi
-- 
2.14.1

