From 66ae6e5313da8be5742a502c0d7027689e84a17d Mon Sep 17 00:00:00 2001
From: Ioana Ciornei <ioana.ciornei@nxp.com>
Date: Wed, 25 Oct 2017 11:30:08 +0000
Subject: [PATCH 10/12] scripts: use proper arithmetic operations

Signed-off-by: Ioana Ciornei <ioana.ciornei@nxp.com>
---
 scripts/ls-main | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/scripts/ls-main b/scripts/ls-main
index 7ce73f0..0dcd2fa 100755
--- a/scripts/ls-main
+++ b/scripts/ls-main
@@ -409,7 +409,7 @@ process_addmux() {
 
 	# check the status
 	object_exists $container $dpdmux
-	if (( $object_exists_status == 1 )); then
+	if [ $object_exists_status == 1 ]; then
 
 		if [ "$root_c" == "$container" ]; then
 			evb=$(ls $SYS_DPRC/"$root_c"/"$dpdmux"/net/ | grep -v "p")
@@ -598,7 +598,7 @@ process_addsw() {
 
 	# check the status
 	object_exists $container $dpsw
-	if (( $object_exists_status == 1 )); then
+	if [ $object_exists_status == 1 ]; then
 		echo "Created ETHSW object $dpsw with ${num_ifs} ports"
 
 		if [ $num_ifs -gt $ifcnt ]; then
@@ -850,7 +850,7 @@ process_addni() {
 
 	# check the status
 	object_exists $container $dpni
-	if (( $object_exists_status == 1 )); then
+	if [ $object_exists_status == 1 ]; then
 		if [ "$root_c" == "$container" ]; then
 			ni=$(ls $SYS_DPRC/"$root_c"/"$dpni"/net/)
 		fi
-- 
2.14.1

