From 6af86a65f2dcfd42f572e04207eef45da3573b92 Mon Sep 17 00:00:00 2001
From: Ioana Ciornei <ioana.ciornei@nxp.com>
Date: Wed, 25 Oct 2017 11:30:33 +0000
Subject: [PATCH 11/12] scripts: do not compare strings with regexp expressions

Signed-off-by: Ioana Ciornei <ioana.ciornei@nxp.com>
---
 scripts/ls-main | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/scripts/ls-main b/scripts/ls-main
index 0dcd2fa..526c052 100755
--- a/scripts/ls-main
+++ b/scripts/ls-main
@@ -805,7 +805,8 @@ process_addni() {
 	fi
 
 	# if no --num-queues is specified then set it to number of cores
-	if [[ $dpni_args != *"--num-queues"* ]]; then
+	num_queues_present=$(echo "$dpni_args" | grep -o "\-\-num-queues" || true)
+	if [[ -z "$num_queues_present" ]]; then
 		dpni_args=$dpni_args" --num-queues="$(nproc)
 	fi
 
-- 
2.14.1

