From 0e5f9f0a7f0ef1947984cd82ade9dbc14ba2c80d Mon Sep 17 00:00:00 2001
From: Ioana Ciornei <ioana.ciornei@nxp.com>
Date: Wed, 25 Oct 2017 12:48:04 +0000
Subject: [PATCH 12/12] scripts: replace 'nproc' with cross-platform
 implementation

Signed-off-by: Ioana Ciornei <ioana.ciornei@nxp.com>
---
 scripts/ls-main | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/scripts/ls-main b/scripts/ls-main
index 526c052..7d127f6 100755
--- a/scripts/ls-main
+++ b/scripts/ls-main
@@ -807,7 +807,7 @@ process_addni() {
 	# if no --num-queues is specified then set it to number of cores
 	num_queues_present=$(echo "$dpni_args" | grep -o "\-\-num-queues" || true)
 	if [[ -z "$num_queues_present" ]]; then
-		dpni_args=$dpni_args" --num-queues="$(nproc)
+		dpni_args=$dpni_args" --num-queues="$(grep -c ^processor /proc/cpuinfo)
 	fi
 
 	# Check if --no-link the endpoint have been provided otherwise display the usage
-- 
2.14.1

