--- a/Makefile
+++ b/Makefile
@@ -53,14 +53,13 @@ restool: $(OBJ)
 %.1: %.md
 	pandoc --standalone --to man $^ -o $@
 
-install: restool scripts/ls-main scripts/ls-append-dpl scripts/ls-debug scripts/restool_completion.sh $(MANPAGE)
+install: restool scripts/ls-main scripts/ls-append-dpl scripts/ls-debug scripts/restool_completion.sh
 	install -D -m 755 restool $(DESTDIR)$(bindir)/restool
 	install -D -m 755 scripts/ls-main $(DESTDIR)$(bindir)/ls-main
 	install -D -m 755 scripts/ls-append-dpl $(DESTDIR)$(bindir)/ls-append-dpl
 	install -D -m 755 scripts/ls-debug $(DESTDIR)$(bindir)/ls-debug
 	$(foreach symlink, $(RESTOOL_SCRIPT_SYMLINKS), sh -c "cd $(DESTDIR)$(bindir) && ln -sf ls-main $(symlink)" ;)
 	install -D -m 755 scripts/restool_completion.sh $(DESTDIR)$(bindir_completion)/restool
-	install -m 0644 -D $(MANPAGE) $(call get_manpage_destination,$(MANPAGE))
 
 clean:
 	rm -f $(OBJ) $(MANPAGE) \
