--- a/libevent.c
+++ b/libevent.c
@@ -37,10 +37,12 @@
 # define LIBEVENT_SOCKETS_SUPPORT
 #endif
 
-#ifdef PHP_WIN32
+//#ifdef PHP_WIN32
 /* XXX compiling with 2.x on Windows. Luckily the ext code works thanks to the
 compat exports from the libevent. However it might need to be adapted to the
 never version, so this ifdefs would go away. */
+#if 1
+/* force use of libevent2 on OpenWrt and by-pass compat event.h */
 # include <event2/event.h>
 # include <event2/event_compat.h>
 # include <event2/event_struct.h>
