--- a/autologin/Makefile.in
+++ b/autologin/Makefile.in
@@ -9,7 +9,7 @@ sysconfdir = @sysconfdir@
 
 ### Installation programs and flags
 INSTALL = @INSTALL@
-INSTALL_PROGRAM = @INSTALL_PROGRAM@ -s
+INSTALL_PROGRAM = @INSTALL_PROGRAM@
 LN_S = @LN_S@
 MKDIR = @MKDIR@
 
--- a/conserver/Makefile.in
+++ b/conserver/Makefile.in
@@ -14,7 +14,7 @@ exampledir = $(datadir)/examples/conserv
 
 ### Installation programs and flags
 INSTALL = @INSTALL@
-INSTALL_PROGRAM = @INSTALL_PROGRAM@ -s
+INSTALL_PROGRAM = @INSTALL_PROGRAM@
 LN_S = @LN_S@
 MKDIR = @MKDIR@
 
--- a/console/Makefile.in
+++ b/console/Makefile.in
@@ -10,7 +10,7 @@ mandir = @mandir@
 
 ### Installation programs and flags
 INSTALL = @INSTALL@
-INSTALL_PROGRAM = @INSTALL_PROGRAM@ -s
+INSTALL_PROGRAM = @INSTALL_PROGRAM@
 LN_S = @LN_S@
 MKDIR = @MKDIR@
 
--- a/contrib/chat/Makefile.in
+++ b/contrib/chat/Makefile.in
@@ -10,7 +10,7 @@ mandir = @mandir@
 
 ### Installation programs and flags
 INSTALL = @INSTALL@
-INSTALL_PROGRAM = @INSTALL_PROGRAM@ -s
+INSTALL_PROGRAM = @INSTALL_PROGRAM@
 LN_S = @LN_S@
 MKDIR = @MKDIR@
 
