--- a/Makefile
+++ b/Makefile
@@ -115,9 +115,6 @@ install: $(TARGET) $(TRANSLATIONS)
 	$(INSTALLMAN) $(MAN_RU) $(DESTDIR)/$(MANDIR)/ru/man1
 	$(INSTALLDIR) $(DESTDIR)/$(DOCDIR)
 	$(INSTALLDOC) $(DOCS) $(DESTDIR)/$(DOCDIR)
-ifneq ($(DEBUG),yes)
-	$(STRIP) $(DESTDIR)/$(BINDIR)/$(TARGET)
-endif
 	mkdir -p $(DESTDIR)/$(PREFIX)/share/locale/nl/LC_MESSAGES
 	cp nl.mo $(DESTDIR)/$(PREFIX)/share/locale/nl/LC_MESSAGES/httping.mo
 	mkdir -p $(DESTDIR)/$(PREFIX)/share/locale/ru/LC_MESSAGES
