--- a/utils.c
+++ b/utils.c
@@ -18,12 +18,11 @@
 double get_ts(void)
 {
 	struct timeval ts;
-       	struct timezone tz;
 
-	if (gettimeofday(&ts, &tz) == -1)
+	if (gettimeofday(&ts, NULL) == -1)
 		error_exit(gettext("gettimeofday failed"));
 
-	return (double)ts.tv_sec + ((double)ts.tv_usec)/1000000.0 + (double)(tz.tz_minuteswest * 60);
+	return (double)ts.tv_sec + ((double)ts.tv_usec)/1000000.0;
 }
 
 void split_string(const char *in, const char *split, char ***list, int *list_n)
