--- a/src/opensslfingerprint.cxx
+++ b/src/opensslfingerprint.cxx
@@ -73,9 +73,11 @@ std::string getCertFingerprint(const std::string certfile, const bool Debug = fa
 	// closes file
 	close(fd);
 
+#if (OPENSSL_VERSION_NUMBER < 0x10100000L) || defined(LIBRESSL_VERSION_NUMBER)
 	// initialize OpenSSL
 	SSL_load_error_strings();
 	SSL_library_init();
+#endif
 
 	// creates BIO buffer
 	BIO * bio = BIO_new_mem_buf(buff, len);
