--- a/drivers/db2/dbd_db2.c
+++ b/drivers/db2/dbd_db2.c
@@ -57,8 +57,7 @@ static const dbi_info_t driver_info = {
   "IBM DB2 database support (using DB2 Call Level Interface)",
   "João Henrique F. Freitas <joaohf@users.sourceforge.net>",
   "http://libdbi-drivers.sourceforge.net",
-  "dbd_db2 v" VERSION,
-  __DATE__
+  "dbd_db2 v" VERSION
 };
 
 static const char *custom_functions[] = {NULL}; // TODO
--- a/drivers/firebird/dbd_firebird.c
+++ b/drivers/firebird/dbd_firebird.c
@@ -67,8 +67,7 @@ static const dbi_info_t driver_info = {
 	"Firebird/Interbase database support",
 	"Christian M. Stamgren <cms@cention.se>",
 	"http://libdbi-drivers.sourceforge.net",
-	"dbd_firebird v" VERSION,
-	__DATE__
+	"dbd_firebird v" VERSION
 };
 
 
--- a/drivers/freetds/dbd_freetds.c
+++ b/drivers/freetds/dbd_freetds.c
@@ -63,8 +63,7 @@ static const dbi_info_t driver_info = {
     "MS SQL and Sybase databases support (using libct)",
     "Vadym Kononenko <konan_v@users.sourceforge.net>",
     "http://libdbi.sourceforge.net",
-    "dbd_freetds v" VERSION,
-    __DATE__
+    "dbd_freetds v" VERSION
 };
 
 static const char APP_NAME[] = "libdbi-freetds-driver";
--- a/drivers/ingres/dbd_ingres.c
+++ b/drivers/ingres/dbd_ingres.c
@@ -44,8 +44,7 @@ static const dbi_info_t driver_info = {
 	"Ingres database support",
 	"Toby Thain <qu1j0t3@sourceforge.net>",
 	"http://libdbi-drivers.sourceforge.net",
-	"dbd_ingres v" VERSION,
-	__DATE__
+	"dbd_ingres v" VERSION
 };
 
 static const char *custom_functions[] = {NULL};
--- a/drivers/msql/dbd_msql.c
+++ b/drivers/msql/dbd_msql.c
@@ -57,8 +57,7 @@ static const dbi_info_t driver_info = {
 	"Mini SQL (mSQL) database support",
 	"Christian M. Stamgren <christian@centiongroup.com>",
 	"libdbi-drivers.sourceforge.net", 
-	"dbd_msql v" VERSION,
-	__DATE__
+	"dbd_msql v" VERSION
 };
 
 static const char *custom_functions[] = {NULL}; 
--- a/drivers/mysql/dbd_mysql.c
+++ b/drivers/mysql/dbd_mysql.c
@@ -59,8 +59,7 @@ static const dbi_info_t driver_info = {
 	"MySQL database support (using libmysqlclient)",
 	"Mark M. Tobenkin <mark@brentwoodradio.com>",
 	"http://libdbi-drivers.sourceforge.net",
-	"dbd_mysql v" VERSION,
-	__DATE__
+	"dbd_mysql v" VERSION
 };
 
 static const char *custom_functions[] = MYSQL_CUSTOM_FUNCTIONS;
--- a/drivers/oracle/dbd_oracle.c
+++ b/drivers/oracle/dbd_oracle.c
@@ -54,8 +54,7 @@ static const dbi_info_t driver_info = {
 	"Oracle database support (using Oracle Call Interface)",
 	"Ashish Ranjan <ashishwave@yahoo.com>", 
 	"http://libdbi-drivers.sourceforge.net",
-	"dbd_Oracle v" VERSION,
-	__DATE__
+	"dbd_Oracle v" VERSION
 };
 
 static const char *custom_functions[] = {NULL}; 
--- a/drivers/pgsql/dbd_pgsql.c
+++ b/drivers/pgsql/dbd_pgsql.c
@@ -61,8 +61,7 @@ static const dbi_info_t driver_info = {
 	"PostgreSQL database support (using libpq)",
 	"David A. Parker <david@neongoat.com>",
 	"http://libdbi-drivers.sourceforge.net",
-	"dbd_pgsql v" VERSION,
-	__DATE__
+	"dbd_pgsql v" VERSION
 };
 
 static const char *custom_functions[] = PGSQL_CUSTOM_FUNCTIONS;
--- a/drivers/sqlite/dbd_sqlite.c
+++ b/drivers/sqlite/dbd_sqlite.c
@@ -65,8 +65,7 @@ static const dbi_info_t driver_info = {
   "SQLite database support (using libsqlite)",
   "Markus Hoenicka <mhoenicka@users.sourceforge.net>",
   "http://libdbi-drivers.sourceforge.net",
-  "dbd_sqlite v" VERSION,
-  __DATE__
+  "dbd_sqlite v" VERSION
 };
 
 static const char *custom_functions[] = SQLITE_CUSTOM_FUNCTIONS;
--- a/drivers/sqlite3/dbd_sqlite3.c
+++ b/drivers/sqlite3/dbd_sqlite3.c
@@ -65,8 +65,7 @@ static const dbi_info_t driver_info = {
   "SQLite3 database support (using libsqlite3)",
   "Markus Hoenicka <mhoenicka@users.sourceforge.net>",
   "http://libdbi-drivers.sourceforge.net",
-  "dbd_sqlite3 v" VERSION,
-  __DATE__
+  "dbd_sqlite3 v" VERSION
 };
 
 static const char *custom_functions[] = SQLITE3_CUSTOM_FUNCTIONS;
