--- a/libip6t_FULLCONENAT.c
+++ b/libip6t_FULLCONENAT.c
@@ -214,6 +214,7 @@ static struct xtables_target fullconenat
 	.x6_options	= FULLCONENAT_opts,
 };
 
+#define _init __attribute__((constructor)) _INIT
 void _init(void)
 {
 	xtables_register_target(&fullconenat_tg_reg);
--- a/libipt_FULLCONENAT.c
+++ b/libipt_FULLCONENAT.c
@@ -235,6 +235,7 @@ static struct xtables_target fullconenat
 	.x6_options	= FULLCONENAT_opts,
 };
 
+#define _init __attribute__((constructor)) _INIT
 void _init(void)
 {
 	xtables_register_target(&fullconenat_tg_reg);
