--- a/drivers/net/Kconfig
+++ b/drivers/net/Kconfig
@@ -263,6 +263,10 @@ source "drivers/net/dsa/Kconfig"
 
 source "drivers/net/ethernet/Kconfig"
 
+config ADM5120_ENET
+	tristate "ADM5120 Ethernet switch support"
+	depends on ADM5120
+
 source "drivers/net/fddi/Kconfig"
 
 source "drivers/net/hippi/Kconfig"
--- a/drivers/net/Makefile
+++ b/drivers/net/Makefile
@@ -34,6 +34,7 @@ obj-$(CONFIG_CAN) += can/
 obj-$(CONFIG_ETRAX_ETHERNET) += cris/
 obj-$(CONFIG_NET_DSA) += dsa/
 obj-$(CONFIG_ETHERNET) += ethernet/
+obj-$(CONFIG_ADM5120_ENET) += adm5120sw.o
 obj-$(CONFIG_FDDI) += fddi/
 obj-$(CONFIG_HIPPI) += hippi/
 obj-$(CONFIG_HAMRADIO) += hamradio/
