--- a/drivers/mtd/chips/jedec_probe.c
+++ b/drivers/mtd/chips/jedec_probe.c
@@ -115,6 +115,10 @@
 #define UPD29F064115	0x221C
 
 /* PMC */
+#define PM39LV512	0x001B
+#define PM39LV010	0x001C
+#define PM39LV020	0x003D
+#define PM39LV040	0x003E
 #define PM49FL002	0x006D
 #define PM49FL004	0x006E
 #define PM49FL008	0x006A
@@ -1259,6 +1263,54 @@ static const struct amd_flash_info jedec
 			ERASEINFO(0x02000,2),
 			ERASEINFO(0x04000,1),
 		}
+        }, {
+		.mfr_id		= CFI_MFR_PMC,
+		.dev_id		= PM39LV512,
+		.name		= "PMC Pm39LV512",
+		.devtypes	= CFI_DEVICETYPE_X8,
+		.uaddr		= MTD_UADDR_0x0555_0x02AA,
+		.dev_size	= SIZE_64KiB,
+		.cmd_set	= P_ID_AMD_STD,
+		.nr_regions	= 1,
+		.regions	= {
+			ERASEINFO(0x01000,16),
+		}
+        }, {
+		.mfr_id		= CFI_MFR_PMC,
+		.dev_id		= PM39LV010,
+		.name		= "PMC Pm39LV010",
+		.devtypes	= CFI_DEVICETYPE_X8,
+		.uaddr		= MTD_UADDR_0x0555_0x02AA,
+		.dev_size	= SIZE_128KiB,
+		.cmd_set	= P_ID_AMD_STD,
+		.nr_regions	= 1,
+		.regions	= {
+			ERASEINFO(0x01000,32),
+		}
+        }, {
+		.mfr_id		= CFI_MFR_PMC,
+		.dev_id		= PM39LV020,
+		.name		= "PMC Pm39LV020",
+		.devtypes	= CFI_DEVICETYPE_X8,
+		.uaddr		= MTD_UADDR_0x0555_0x02AA,
+		.dev_size	= SIZE_256KiB,
+		.cmd_set	= P_ID_AMD_STD,
+		.nr_regions	= 1,
+		.regions	= {
+			ERASEINFO(0x01000,64),
+		}
+        }, {
+		.mfr_id		= CFI_MFR_PMC,
+		.dev_id		= PM39LV040,
+		.name		= "PMC Pm39LV040",
+		.devtypes	= CFI_DEVICETYPE_X8,
+		.uaddr		= MTD_UADDR_0x0555_0x02AA,
+		.dev_size	= SIZE_512KiB,
+		.cmd_set	= P_ID_AMD_STD,
+		.nr_regions	= 1,
+		.regions	= {
+			ERASEINFO(0x01000,128),
+		}
 	}, {
 		.mfr_id		= CFI_MFR_PMC,
 		.dev_id		= PM49FL002,
