--- a/drivers/ata/Makefile
+++ b/drivers/ata/Makefile
@@ -98,6 +98,7 @@ obj-$(CONFIG_PATA_PCMCIA)	+= pata_pcmcia
 obj-$(CONFIG_PATA_PALMLD)	+= pata_palmld.o
 obj-$(CONFIG_PATA_PLATFORM)	+= pata_platform.o
 obj-$(CONFIG_PATA_OF_PLATFORM)	+= pata_of_platform.o
+obj-$(CONFIG_PATA_RB153_CF)	+= pata_rb153_cf.o
 obj-$(CONFIG_PATA_RB532)	+= pata_rb532_cf.o
 obj-$(CONFIG_PATA_RZ1000)	+= pata_rz1000.o
 obj-$(CONFIG_PATA_SAMSUNG_CF)	+= pata_samsung_cf.o
--- a/drivers/ata/Kconfig
+++ b/drivers/ata/Kconfig
@@ -955,6 +955,15 @@ config PATA_QDI
 	help
 	  Support for QDI 6500 and 6580 PATA controllers on VESA local bus.
 
+config PATA_RB153_CF
+	tristate "RouterBOARD 153 Compact Flash support"
+	depends on ADM5120_MACH_RB_153
+	help
+	  This option enables support for a Compact Flash connected on
+	  the RouterBOARD 153.
+
+	  If unsure, say N.
+
 config PATA_RB532
 	tristate "RouterBoard 532 PATA CompactFlash support"
 	depends on MIKROTIK_RB532
