--- a/Makefile.in
+++ b/Makefile.in
@@ -80,7 +80,7 @@ SYSCONFIG_FILE = @SYSCONFIG_FILE@
 # Compiler flags
 CFLAGS = @CFLAGS@ -Wall -Wstrict-prototypes -pipe -O2
 DFLAGS = @DFLAGS@
-LFLAGS = @STRIP@
+LDFLAGS += @STRIP@
 DFLAGS += -DSA_DIR=\"$(SA_DIR)\" -DSADC_PATH=\"$(SADC_PATH)\"
 DFLAGS += $(DFSENSORS)
 DFLAGS += $(DFPCP)
@@ -120,7 +120,7 @@ ifeq ($(SYSPARAM),y)
 	DFLAGS += -DHAVE_SYS_PARAM_H
 endif
 NLS = @NLS@
-LFLAGS += @LFINTL@
+LDFLAGS += @LFINTL@
 ifeq ($(NLS),y)
 REQUIRE_NLS = -DUSE_NLS -DPACKAGE=\"$(PACKAGE)\" -DLOCALEDIR=\"$(NLS_DIR)\"
 endif
@@ -198,7 +198,7 @@ NLSPOT= $(NLSPO:.po=.pot)
 	$(CC) -o $@ -c $(CFLAGS) $(DFLAGS) $<
 
 % : %.o
-	$(CC) -o $@ $(CFLAGS) $^ $(LFLAGS)
+	$(CC) -o $@ $(CFLAGS) $^ $(LDFLAGS)
 
 all: sadc sar sadf iostat tapestat mpstat pidstat cifsiostat locales
 
@@ -286,7 +286,7 @@ librdsensors.a: rd_sensors.o
 
 sadc.o: sadc.c sa.h version.h common.h rd_stats.h rd_sensors.h
 
-sadc: LFLAGS += $(LFSENSORS)
+sadc: LDFLAGS += $(LFSENSORS)
 
 sadc: sadc.o act_sadc.o sa_wrap.o sa_common_sadc.o common_sadc.o systest.o librdstats.a librdsensors.a
 
@@ -296,7 +296,7 @@ sar: sar.o act_sar.o format_sar.o sa_com
 
 sadf.o: sadf.c sadf.h version.h sa.h common.h rd_stats.h rd_sensors.h
 
-sadf: LFLAGS += $(LFPCP)
+sadf: LDFLAGS += $(LFPCP)
 
 sadf: sadf.o act_sadf.o format_sadf.o sadf_misc.o pcp_def_metrics.o sa_conv.o rndr_stats.o xml_stats.o json_stats.o svg_stats.o raw_stats.o pcp_stats.o sa_common.o librdstats_light.a libsyscom.a
 
