--- a/lib/err/errPosix.c
+++ b/lib/err/errPosix.c
@@ -61,13 +61,7 @@ ErrErrno2String(Err_Number errorNumber,
                 char *buf,		// OUT: return buffer
 		size_t bufSize)		// IN: size of buffer
 {
-   char *p;
-
-#if defined(__linux__) && !defined(__ANDROID__)
-   p = strerror_r(errorNumber, buf, bufSize);
-#else
-   p = strerror(errorNumber);
-#endif
+   char *p = strerror(errorNumber);
    ASSERT(p != NULL);
    return p;
 }
