From: coolsnowwolf <coolsnowwolf@gmail.com>
Subject: [PATCH] platform: mikrotik: fix bin attribute callback for Linux 6.18

Signed-off-by: coolsnowwolf <coolsnowwolf@gmail.com>
---
--- a/drivers/platform/mikrotik/rb_hardconfig.c
+++ b/drivers/platform/mikrotik/rb_hardconfig.c
@@ -35,6 +35,7 @@
 #include <linux/string.h>
 #include <linux/mtd/mtd.h>
 #include <linux/sysfs.h>
+#include <linux/version.h>
 #include <linux/lzo.h>
 
 #include "rb_hardconfig.h"
@@ -340,9 +341,15 @@
 	return out - buf;
 }
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 18, 0)
 static ssize_t hc_wlan_data_bin_read(struct file *filp, struct kobject *kobj,
 				     struct bin_attribute *attr, char *buf,
 				     loff_t off, size_t count);
+#else
+static ssize_t hc_wlan_data_bin_read(struct file *filp, struct kobject *kobj,
+				     const struct bin_attribute *attr, char *buf,
+				     loff_t off, size_t count);
+#endif
 
 static struct hc_wlan_attr {
 	const u16 erd_tag_id;
@@ -615,9 +622,15 @@
  * at boot time to load wlan caldata), this makes it possible to save memory for
  * the system.
  */
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 18, 0)
 static ssize_t hc_wlan_data_bin_read(struct file *filp, struct kobject *kobj,
 				     struct bin_attribute *attr, char *buf,
 				     loff_t off, size_t count)
+#else
+static ssize_t hc_wlan_data_bin_read(struct file *filp, struct kobject *kobj,
+				     const struct bin_attribute *attr, char *buf,
+				     loff_t off, size_t count)
+#endif
 {
 	struct hc_wlan_attr *hc_wattr;
 	size_t outlen;

