--- a/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_compat.h
+++ b/src/SDIO/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_compat.h
@@ -47,8 +47,8 @@
 #else
 #define HIGH_KERNEL_VERSION KERNEL_VERSION(6, 0, 0)
 #define HIGH_KERNEL_VERSION2 KERNEL_VERSION(6, 1, 0)
-#define HIGH_KERNEL_VERSION3 KERNEL_VERSION(6, 3, 0)
-#define HIGH_KERNEL_VERSION4 KERNEL_VERSION(6, 3, 0)
+#define HIGH_KERNEL_VERSION3 KERNEL_VERSION(6, 9, 0)
+#define HIGH_KERNEL_VERSION4 KERNEL_VERSION(6, 9, 0)
 #endif
 
 
