--- a/nss_qdisc/nss_qdisc.c
+++ b/nss_qdisc/nss_qdisc.c
@@ -2708,9 +2708,11 @@ static int nss_qdisc_if_event_cb(struct
 	case NETDEV_BR_JOIN:
 		nss_qdisc_info("Reveived NETDEV_BR_JOIN on interface %s\n",
 				dev->name);
+		goto fall_through;
 	case NETDEV_BR_LEAVE:
 		nss_qdisc_info("Reveived NETDEV_BR_LEAVE on interface %s\n",
 				dev->name);
+fall_through:
 		br = nss_qdisc_get_dev_master(dev);
 		if_num = nss_cmn_get_interface_number(nss_qdisc_ctx, dev);
 
