--- a/binutils/Makefile
+++ b/binutils/Makefile
@@ -107,7 +107,7 @@ docs:	manpages
 indep: docs
 	$(Q)$(MAKE) -C po all
 
-all:	arch indep
+all:	arch
 
 .PHONY: coverage
 coverage:
@@ -147,7 +147,7 @@ tests: $(BINTOOLS) $(SBINTOOLS) $(TESTS)
 	echo "no tests atm"
 
 .PHONY: install
-install: install-indep install-arch
+install: install-arch
 
 .PHONY: install-arch
 install-arch: arch
--- a/parser/Makefile
+++ b/parser/Makefile
@@ -195,7 +195,7 @@ extra_docs: pdf
 indep: docs
 	$(Q)$(MAKE) -C po all
 
-all:	arch indep
+all:	arch
 
 .PHONY: coverage
 coverage:
@@ -396,7 +396,6 @@ endif
 
 .PHONY: install
 install:
-	$(MAKE) install-indep
 	$(MAKE) install-arch
 
 .PHONY: install-arch
--- a/utils/Makefile
+++ b/utils/Makefile
@@ -31,9 +31,7 @@ MANPAGES = ${TOOLS:=.8} logprof.conf.5
 
 PYFLAKES ?= pyflakes3
 
-all: docs
-	$(MAKE) -C po all
-	$(MAKE) -C vim all
+all: clean
 
 .PHONY: docs
 docs: ${MANPAGES} ${HTMLMANPAGES}
@@ -49,15 +47,12 @@ po/${NAME}.pot: ${TOOLS} ${PYMODULES}
 	$(MAKE) -C po ${NAME}.pot NAME=${NAME} SOURCES="${TOOLS} ${PYMODULES}"
 
 .PHONY: install
-install: ${MANPAGES} ${HTMLMANPAGES}
+install:
 	install -d ${CONFDIR}
 	install -m 644 logprof.conf severity.db notify.conf ${CONFDIR}
 	install -d ${BINDIR}
 	# aa-easyprof is installed by python-tools-setup.py
 	install -m 755 $(filter-out aa-easyprof, ${TOOLS}) ${BINDIR}
-	$(MAKE) -C po install DESTDIR=${DESTDIR} NAME=${NAME}
-	$(MAKE) install_manpages DESTDIR=${DESTDIR}
-	$(MAKE) -C vim install DESTDIR=${DESTDIR}
 	${PYTHON} ${PYSETUP} install --prefix=${PYPREFIX} --root=${DESTDIR} --version=${VERSION}
 
 .PHONY: clean
