--- a/os-daq-modules/daq_ipfw.c
+++ b/os-daq-modules/daq_ipfw.c
@@ -23,10 +23,10 @@
 #include <stdlib.h>
 #include <string.h>
 #include <stdio.h>
+#include <unistd.h>
 
 #include <sys/types.h>
 #include <sys/time.h>
-#include <sys/unistd.h>
 
 #include <netinet/in.h>
 #include <sys/socket.h>
--- a/os-daq-modules/daq_ipq.c
+++ b/os-daq-modules/daq_ipq.c
@@ -24,10 +24,10 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <unistd.h>
 
 #include <sys/types.h>
 #include <sys/time.h>
-#include <sys/unistd.h>
 
 #include <netinet/ip.h>
 
--- a/os-daq-modules/daq_nfq.c
+++ b/os-daq-modules/daq_nfq.c
@@ -24,10 +24,10 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <unistd.h>
 
 #include <sys/types.h>
 #include <sys/time.h>
-#include <sys/unistd.h>
 
 #include <netinet/ip.h>
 
