--- a/key.dns_resolver.c
+++ b/key.dns_resolver.c
@@ -717,12 +717,12 @@ int main(int argc, char *argv[])
 	keyend = buf + ktlen + 1;
 
 	/* the actual key description follows the last semicolon */
-	keyend = rindex(keyend, ';');
+	keyend = strrchr(keyend, ';');
 	if (!keyend)
 		error("Invalid key description: %s", buf);
 	keyend++;
 
-	name = index(keyend, ':');
+	name = strchr(keyend, ':');
 	if (!name)
 		dns_query_a_or_aaaa(keyend, callout_info);
 
