--- a/Makefile.in	2020-07-16 20:18:18.754621366 +0800
+++ b/Makefile.in	2020-07-16 20:18:12.370623476 +0800
@@ -251,10 +251,10 @@ uninstall-pc:
 maybe-uninstall-pc: uninstall-pc
 
 install-static: installdirs $(STATIC_LIBS) install-headers install-doc ${install_examples} install-pc
-	-$(MV) $(DESTDIR)$(libdir)/libreadline.a $(DESTDIR)$(libdir)/libreadline.old
+	[ ! -f $(DESTDIR)$(libdir)/libreadline.a ] || $(MV) $(DESTDIR)$(libdir)/libreadline.a $(DESTDIR)$(libdir)/libreadline.old
 	$(INSTALL_DATA) libreadline.a $(DESTDIR)$(libdir)/libreadline.a
 	-test -n "$(RANLIB)" && $(RANLIB) $(DESTDIR)$(libdir)/libreadline.a
-	-$(MV) $(DESTDIR)$(libdir)/libhistory.a $(DESTDIR)$(libdir)/libhistory.old
+	[ ! -f $(DESTDIR)$(libdir)/libhistory.a ] || $(MV) $(DESTDIR)$(libdir)/libhistory.a $(DESTDIR)$(libdir)/libhistory.old
 	$(INSTALL_DATA) libhistory.a $(DESTDIR)$(libdir)/libhistory.a
 	-test -n "$(RANLIB)" && $(RANLIB) $(DESTDIR)$(libdir)/libhistory.a
 
