--- a/Makefile.am
+++ b/Makefile.am
@@ -9,7 +9,7 @@ EXTRA_DIST = MAINTAINERS TESTERS install
 EXTRA_DIST += HACKING.md SECURITY.md
 
 # Note: @subdirs@ lists all the directories from AC_CONFIG_SUBDIRS()
-SUBDIRS = @subdirs@ libgphoto2 camlibs tests examples po packaging doc gphoto-m4
+SUBDIRS = @subdirs@ libgphoto2 camlibs packaging gphoto-m4
 
 EXTRA_DIST    += libgphoto2.pc.in
 pkgconfig_DATA = libgphoto2.pc
--- a/Makefile.in
+++ b/Makefile.in
@@ -503,7 +503,7 @@ EXTRA_DIST = MAINTAINERS TESTERS install
 	README.packaging
 
 # Note: @subdirs@ lists all the directories from AC_CONFIG_SUBDIRS()
-SUBDIRS = @subdirs@ libgphoto2 camlibs tests examples po packaging doc gphoto-m4
+SUBDIRS = @subdirs@ libgphoto2 camlibs packaging gphoto-m4
 pkgconfig_DATA = libgphoto2.pc
 noinst_DATA = libgphoto2-uninstalled.pc
 doc_DATA = AUTHORS COPYING NEWS ABOUT-NLS ChangeLog README.md \
--- a/configure.ac
+++ b/configure.ac
@@ -706,20 +706,11 @@ gphoto-m4/Makefile
 libgphoto2/Makefile
 libgphoto2.pc
 libgphoto2-uninstalled.pc
-examples/Makefile
-tests/Makefile
-tests/ddb/Makefile
-tests/ddb/check-ddb.sh
 packaging/Makefile
 packaging/linux-hotplug/Makefile
 packaging/generic/Makefile
 packaging/rpm/Makefile
 packaging/rpm/package.spec
-po/Makefile.in
-doc/Makefile
-doc/Doxyfile
-doc/Doxyfile-internals
-doc/api/Makefile
 ],[
 dnl This relies on this code being called for each of the above files
 dnl with ac_file set to the filename.
--- a/libgphoto2_port/Makefile.am
+++ b/libgphoto2_port/Makefile.am
@@ -25,7 +25,7 @@ udevscript_PROGRAMS =
 bin_SCRIPTS = gphoto2-port-config
 
 # The . stands for the current dir, i.e. the iolibs to build.
-SUBDIRS = po libgphoto2_port test . doc gphoto-m4
+SUBDIRS = libgphoto2_port . gphoto-m4
 
 
 ########################################################################
--- a/libgphoto2_port/Makefile.in
+++ b/libgphoto2_port/Makefile.in
@@ -581,7 +581,7 @@ EXTRA_LTLIBRARIES = disk.la ptpip.la ser
 bin_SCRIPTS = gphoto2-port-config
 
 # The . stands for the current dir, i.e. the iolibs to build.
-SUBDIRS = po libgphoto2_port test . doc gphoto-m4
+SUBDIRS = libgphoto2_port . gphoto-m4
 
 ########################################################################
 # All iolibs are defined as EXTRA_LTLIBRARIES. This requires that
--- a/libgphoto2_port/configure.ac
+++ b/libgphoto2_port/configure.ac
@@ -536,13 +536,10 @@ AC_SUBST([AM_LDFLAGS])
 # ---------------------------------------------------------------------------
 AC_CONFIG_FILES([
 Makefile
-po/Makefile.in
 libgphoto2_port/Makefile
 libgphoto2_port.pc
 libgphoto2_port-uninstalled.pc
 gphoto2-port-config
-test/Makefile
-doc/Makefile
 gphoto-m4/Makefile
 ])
 AC_OUTPUT
