--- a/arch/arm/boot/dts/xilinx/zynq-antminer-s9.dts
+++ b/arch/arm/boot/dts/xilinx/zynq-antminer-s9.dts
@@ -14,6 +14,11 @@
 		ethernet0 = &gem0;
 		serial0 = &uart1;
 		mmc0 = &sdhci0;
+
+		led-boot = &led_green;
+		led-failsafe = &led_green;
+		led-running = &led_green;
+		led-upgrade = &led_green;
 	};
 
 	memory@0 {
@@ -40,7 +45,7 @@
 			default-state = "off";
 		};
 
-		led_green {
+		led_green: led_green {
 			label = "s9:green:usr";
 			gpios = <&gpio0 38 0x0>;
 			default-state = "off";
