--- a/util/Makefile.in
+++ b/util/Makefile.in
@@ -723,7 +723,7 @@ mount_util.c: $(top_srcdir)/lib/mount_ut
 
 install-exec-hook:
 	-chmod u+s $(DESTDIR)$(bindir)/fusermount
-	@if test ! -e $(DESTDIR)/dev/fuse; then \
+	-@if test ! -e $(DESTDIR)/dev/fuse; then \
 		$(MKDIR_P) $(DESTDIR)/dev; \
 		echo "mknod $(DESTDIR)/dev/fuse -m 0666 c 10 229 || true"; \
 		mknod $(DESTDIR)/dev/fuse -m 0666 c 10 229 || true; \
