--- a/clish/shell/shell_execute.c
+++ b/clish/shell/shell_execute.c
@@ -19,12 +19,14 @@
 #include <signal.h>
 #include <fcntl.h>
 
+#if defined(__UCLIBC__) && !defined(__UCLIBC_HAS_OBSOLETE_BSD_SIGNAL__)
 /* Empty signal handler to ignore signal but don't use SIG_IGN. */
 static void sigignore(int signo)
 {
 	signo = signo; /* Happy compiler */
 	return;
 }
+#endif
 
 /*-------------------------------------------------------- */
 static int clish_shell_lock(const char *lock_path)
