--- a/drivers/leds/Kconfig
+++ b/drivers/leds/Kconfig
@@ -568,6 +568,13 @@ config LEDS_SEAD3
 
 comment "LED driver for blink(1) USB RGB LED is under Special HID drivers (HID_THINGM)"
 
+config LEDS_NU801
+	tristate "LED driver for NU801 RGB LED"
+	depends on LEDS_CLASS && (ATH79_MACH_MR18 || ATH79_MACH_Z1)
+	help
+	  This option enables support for NU801 RGB LED driver chips
+	  accessed via GPIO.
+
 config LEDS_BLINKM
 	tristate "LED support for the BlinkM I2C RGB LED"
 	depends on LEDS_CLASS
--- a/drivers/leds/Makefile
+++ b/drivers/leds/Makefile
@@ -55,6 +55,7 @@ obj-$(CONFIG_LEDS_ADP5520)		+= leds-adp5
 obj-$(CONFIG_LEDS_DELL_NETBOOKS)	+= dell-led.o
 obj-$(CONFIG_LEDS_MC13783)		+= leds-mc13783.o
 obj-$(CONFIG_LEDS_RB750)		+= leds-rb750.o
+obj-$(CONFIG_LEDS_NU801)		+= leds-nu801.o
 obj-$(CONFIG_LEDS_NS2)			+= leds-ns2.o
 obj-$(CONFIG_LEDS_NETXBIG)		+= leds-netxbig.o
 obj-$(CONFIG_LEDS_ASIC3)		+= leds-asic3.o
