--- a/Makefile.am
+++ b/Makefile.am
@@ -4,7 +4,7 @@
 
 ## Process this file with automake to produce Makefile.in
 
-SUBDIRS = m4 po intl src emacs doc check
+SUBDIRS = m4 src emacs doc check
 
 EXTRA_DIST = m4/ChangeLog config.rpath README-WIN
 
--- a/Makefile.in
+++ b/Makefile.in
@@ -117,7 +117,7 @@ am__CONFIG_DISTCLEAN_FILES = config.stat
  configure.lineno config.status.lineno
 mkinstalldirs = $(SHELL) $(top_srcdir)/mkinstalldirs
 CONFIG_HEADER = config.h
-CONFIG_CLEAN_FILES = intl/Makefile
+CONFIG_CLEAN_FILES =
 CONFIG_CLEAN_VPATH_FILES =
 AM_V_P = $(am__v_P_@AM_V@)
 am__v_P_ = $(am__v_P_@AM_DEFAULT_V@)
@@ -177,7 +177,7 @@ CTAGS = ctags
 CSCOPE = cscope
 DIST_SUBDIRS = $(SUBDIRS)
 am__DIST_COMMON = $(srcdir)/Makefile.in $(srcdir)/config.h.in \
-	$(top_srcdir)/intl/Makefile.in ABOUT-NLS AUTHORS COPYING \
+	ABOUT-NLS AUTHORS COPYING \
 	ChangeLog INSTALL NEWS README compile config.guess \
 	config.rpath config.sub install-sh ltmain.sh missing \
 	mkinstalldirs
@@ -403,7 +403,7 @@ target_alias = @target_alias@
 top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
-SUBDIRS = m4 po intl src emacs doc check
+SUBDIRS = m4 src emacs doc check
 EXTRA_DIST = m4/ChangeLog config.rpath README-WIN
 ACLOCAL_AMFLAGS = -I m4
 all: config.h
--- a/configure.ac
+++ b/configure.ac
@@ -149,17 +149,6 @@ AC_MSG_RESULT($UINT32_TYPE)
 AC_DEFINE_UNQUOTED(UINT32_TYPE,$UINT32_TYPE,unsigned 32 bit integer type)
 
 dnl ----------------------------------------------------------------------
-dnl Internationalization
-
-GETTEXT_PACKAGE=ccrypt
-AC_SUBST(GETTEXT_PACKAGE)
-AC_DEFINE_UNQUOTED(GETTEXT_PACKAGE, "$GETTEXT_PACKAGE", [Package name for gettext])
-
-AM_GNU_GETTEXT
-AM_GNU_GETTEXT_VERSION([0.19.8])
-dnl IT_PO_SUBDIR(po)
-
-dnl ----------------------------------------------------------------------
 dnl Libtool (needed by intl/)
 
 LT_INIT
@@ -180,9 +169,7 @@ AC_SUBST(TAR)
 dnl ----------------------------------------------------------------------
 AC_CONFIG_FILES([doc/ccrypt.1
 	   doc/ccguess.1
-	   po/Makefile.in 
 	   m4/Makefile 
-	   intl/Makefile 
 	   Makefile 
 	   src/Makefile 
 	   emacs/Makefile
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -56,4 +56,4 @@ uninstall-local:
 # internationalization stuff
 localedir = $(datadir)/locale
 AM_CPPFLAGS = @EXTRA_INCLUDES@ -I../intl -I$(top_srcdir)/intl -DLOCALEDIR=\"$(localedir)\"
-LIBS = @LIBINTL@ @LIBS@
+LIBS = @LIBS@
--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -278,7 +278,7 @@ LIBMULTITHREAD = @LIBMULTITHREAD@
 LIBOBJS = @LIBOBJS@
 LIBPTH = @LIBPTH@
 LIBPTH_PREFIX = @LIBPTH_PREFIX@
-LIBS = @LIBINTL@ @LIBS@
+LIBS = @LIBS@
 LIBTHREAD = @LIBTHREAD@
 LIBTOOL = @LIBTOOL@
 LIPO = @LIPO@
