--- a/tacacs-F4.0.4.28/configure
+++ b/tacacs-F4.0.4.28/configure
@@ -3244,11 +3244,9 @@ case "${host_os}" in
     ;;
     *linux* )
 	# XXX: not sure if /usr/local is necessary.
-	# XXX: linux libwrap needs -lnsl. configure should check for
-	#      existence of libnsl instead of hard-coding
 	CPPFLAGS="$CFLAGS -I/usr/local/include"; export CPPFLAGS
 	LDFLAGS="$LDFLAGS -L/usr/local/lib -L/lib"; export LDFLAGS
-	LIBS="-lnsl -lcrypt $LIBS"; export LIBS
+	LIBS="-lcrypt $LIBS"; export LIBS
 	$as_echo "#define LINUX 1" >>confdefs.h
 
 
