From 2e2b3aeda9af9c029bf347c974911fe96cd79c43 Mon Sep 17 00:00:00 2001
From: Evgeniy Didin <didin@synopsys.com>
Date: Mon, 23 Mar 2020 15:57:18 +0300
Subject: [PATCH] arch/arc: Add compiler option for gcc8.4

Signed-off-by: Evgeniy Didin <didin@synopsys.com>
---
 arch/arc/Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/arch/arc/Makefile
+++ b/arch/arc/Makefile
@@ -11,7 +11,7 @@ endif
 
 cflags-y	+= -fno-common -pipe -fno-builtin -mmedium-calls -D__linux__
 cflags-$(CONFIG_ISA_ARCOMPACT)	+= -mA7
-cflags-$(CONFIG_ISA_ARCV2)	+= -mcpu=hs38
+cflags-$(CONFIG_ISA_ARCV2)	+= -mcpu=hs38 -mmpy-option=2
 
 ifdef CONFIG_ARC_CURR_IN_REG
 # For a global register defintion, make sure it gets passed to every file
