--- a/SConstruct
+++ b/SConstruct
@@ -1068,6 +1068,10 @@ else:
                 ncurseslibs += pkg_config('tinfo', rpath_hack=True)
         # It's not yet known whether rpath_hack is appropriate for
         # ncurses5-config.
+        elif WhereIs('ncurses6-config'):
+            ncurseslibs = ['!ncurses6-config --libs --cflags']
+        elif WhereIs('ncursesw6-config'):
+            ncurseslibs = ['!ncursesw6-config --libs --cflags']
         elif WhereIs('ncurses5-config'):
             ncurseslibs = ['!ncurses5-config --libs --cflags']
         elif WhereIs('ncursesw5-config'):
