From 0db3fb3119eda8c2360454c2a01f84602a879c38 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C3=81lvaro=20Fern=C3=A1ndez=20Rojas?= <noltari@gmail.com>
Date: Tue, 16 Jan 2024 15:32:12 +0100
Subject: [PATCH] raspinfo: adapt to OpenWrt
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>
---
 raspinfo/raspinfo | 206 +++-------------------------------------------
 1 file changed, 11 insertions(+), 195 deletions(-)

--- a/raspinfo/raspinfo
+++ b/raspinfo/raspinfo
@@ -8,173 +8,13 @@
 
 alias filter="sed -e 's/\([0-9a-fA-F]\{1,4\}:\)\{7,7\}[0-9a-fA-F]\{1,4\}/y.y.y.y.y.y.y.y/g' | sed -e 's/[0-9a-fA-F]\{1,4\}:\(:[0-9a-fA-F]\{1,4\}\)\{1,4\}/y::y.y.y.y/g' | sed -e 's/\([0-9]\{1,3\}\.\)\{3,3\}[0-9]\{1,3\}/x.x.x.x/g' | sed -e 's/\([0-9a-fA-F]\{2,2\}\:\)\{5,5\}[0-9a-fA-F]\{2,2\}/m.m.m.m/g'"
 
-display_info_drm() {
-   # kmsprint is more useful, but not always installed
-   if command -v kmsprint > /dev/null; then
-      kmsprint
-      echo
-      kmsprint -m
-      echo
-      kmsprint -p
-   else
-      echo "kmsprint is not installed. Install with: sudo apt install kms++-utils"
-   fi
-
-   echo
-
-   # dump the /sys/class entries for the displays
-   for card in /sys/class/drm/card[0-9]-*;
-   do
-      echo $card
-
-      if ! command -v kmsprint > /dev/null; then
-         if [ -f $card/modes ];
-         then
-            cat $card/modes
-         else
-            echo "No modes found"
-         fi
-         echo
-      fi
-
-      if [ -f $card/edid ];
-      then
-         base64 $card/edid
-      else
-         echo "No EDID found"
-      fi
-      echo
-   done
-
-   # dump the /sys/kernel/debug entries for the displays
-   cardfound=0
-   for card in `seq 0 9`;
-   do
-      if sudo test -f "/sys/kernel/debug/dri/${card}/state";
-      then
-         for hdmi in 0 1;
-         do
-            if sudo test -f "/sys/kernel/debug/dri/${card}/hdmi${hdmi}_regs";
-            then
-               echo "HDMI${hdmi}: $(sudo cat /sys/kernel/debug/dri/$card/hdmi${hdmi}_regs | grep HOTPLUG)"
-            fi
-         done
-         echo
-         # dump the state if we don't have kmsprint
-         if ! command -v kmsprint > /dev/null; then
-            echo "/sys/kernel/debug/dri/$card/state:"
-            sudo cat "/sys/kernel/debug/dri/$card/state"
-            echo
-         fi
-         cardfound=1
-      fi
-   done
-   if [ "$cardfound" = "0" ];
-   then
-      echo "kms state not found"
-      echo
-   fi
-}
-
-display_info_legacy() {
-   # Legacy mode
-   for card in `seq 0 9`;
-   do
-      F="/dev/fb${card}"
-      if test -e $F;
-      then
-         echo Framebuffer: $F
-         fbset -s -fb $F
-      fi
-   done
-
-   if command -v tvservice > /dev/null; then
-      disps=`tvservice -l | awk '/Display Number/{print substr($3,1,1)}'`
-
-      tmp=$(mktemp)
-
-      for display in $disps
-      do
-         echo
-         echo "Display: " $display
-
-         tvservice -v $display -s
-         tvservice -v $display -n
-         tvservice -v $display -m CEA
-         tvservice -v $display -m DMT
-
-         echo
-         tvservice -v $display -d $tmp > /dev/null
-         base64 $tmp
-         echo
-      done
-
-      rm $tmp
-   else
-      echo "tvservice not installed"
-   fi
-
-   if pgrep Xorg > /dev/null &&
-      command -v xrandr > /dev/null &&
-      DISPLAY=${DISPLAY:-:0} xrandr --listmonitors >/dev/null 2>&1; then
-      DISPLAY=${DISPLAY:-:0} xrandr --verbose
-      echo
-   fi
-}
-
-display_info() {
-   # Check if we are running a KMS/DRM system
-
-   if [ -f /proc/device-tree/soc/firmwarekms@*/status ] &&
-      grep -q okay /proc/device-tree/soc/firmwarekms@*/status; then
-      KMS=fkms
-   elif [ -f /proc/device-tree/*/v3d@*/status ] &&
-      grep -q okay /proc/device-tree/*/v3d@*/status; then
-      KMS=kms
-   else
-      KMS=legacy
-   fi
-
-   if pgrep labwc > /dev/null; then
-      WIN=labwc
-   elif pgrep wayfire > /dev/null; then
-      WIN=wayfire
-   elif pgrep Xorg > /dev/null; then
-      WIN=Xorg
-   else
-      WIN=console
-   fi
-   echo "Using $WIN with $KMS"
-
-   if [ -d "/dev/dri" ];
-   then
-      display_info_drm
-   else
-      display_info_legacy
-   fi
-}
-
-audio_info() {
-   aplay -l
-   echo
-   aplay -L
-   echo
-   systemctl --user status pipewire.socket pipewire.service pulseaudio.service pulseaudio.socket
-   echo
-   if command -v pactl > /dev/null; then
-      pactl info
-   else
-      echo pactl not installed
-   fi
-}
-
 network_info() {
    cat /proc/net/wireless 2>/dev/null || echo "/proc/net/wireless not available"
    echo
    if command -v rfkill >/dev/null; then
       rfkill list
    else
-      echo "rfkill not installed"
+      echo "rfkill not found"
    fi
    echo
 
@@ -188,7 +28,7 @@ network_info() {
             if [ -d "$iface/phy80211" ]; then
                ethtool -i $iface_name
             else
-               sudo ethtool $iface_name
+               ethtool $iface_name
                echo
                ethtool --show-eee $iface_name
                echo
@@ -198,7 +38,7 @@ network_info() {
          fi
       done
    else
-      echo "ethtool not installed"
+      echo "ethtool not found"
    fi
 }
 
@@ -222,8 +62,6 @@ echo
 cat /etc/os-release | head -4
 echo
 
-cat /etc/rpi-issue
-echo
 uname -a
 
 cat /proc/cpuinfo | tail -3
@@ -260,18 +98,6 @@ echo
 cat /proc/swaps
 
 echo
-echo "Package version information"
-echo "---------------------------"
-echo
-
-apt-cache policy raspberrypi-ui-mods | head -2
-apt-cache policy raspberrypi-sys-mods | head -2
-apt-cache policy openbox | head -2
-apt-cache policy lxpanel | head -2
-apt-cache policy pcmanfm | head -2
-apt-cache policy rpd-plym-splash | head -2
-
-echo
 echo "Networking Information"
 echo "----------------------"
 echo
@@ -283,21 +109,11 @@ echo "USB Information"
 echo "---------------"
 echo
 
-lsusb -t
-
-echo
-echo "Display Information"
-echo "-------------------"
-echo
-
-display_info
-
-echo
-echo "Audio Information"
-echo "-------------------"
-echo
-
-audio_info
+if command -v lsusb > /dev/null; then
+   lsusb -t 2>&1
+else
+   echo "usbutils not found"
+fi
 
 echo
 echo "Firmware Configuration"
@@ -322,7 +138,7 @@ echo "-----------------"
 echo
 
 if command -v pinctrl > /dev/null; then
-   sudo pinctrl 2>&1
+   pinctrl 2>&1
 elif command -v raspi-gpio > /dev/null; then
    raspi-gpio get 2>&1
 else
@@ -335,9 +151,9 @@ echo "------------------"
 echo
 
 if command -v vcdbg > /dev/null; then
-   sudo vcdbg log msg 2>&1
+   vcdbg log msg 2>&1
 elif command -v vclog > /dev/null; then
-   sudo vclog --msg 2>&1
+   vclog --msg 2>&1
 else
    echo "vcdbg not found"
 fi
@@ -347,16 +163,16 @@ echo "dmesg log"
 echo "---------"
 echo
 
-sudo dmesg | filter
+dmesg | filter
 
-if  grep -q "^Revision\s*:\s*[ 123][0-9a-fA-F][0-9a-fA-F][0-9a-fA-F]1[13457][0-9a-fA-F]$" /proc/cpuinfo
+if  grep -q "^Revision\s*:\s*[ 123][0-9a-fA-F][0-9a-fA-F][0-9a-fA-F]1[0-9][0-9a-fA-F]$" /proc/cpuinfo
 then
    echo
    echo "EEPROM"
    echo "------"
    echo
    if command -v rpi-eeprom-update > /dev/null; then
-      sudo rpi-eeprom-update
+      rpi-eeprom-update
    else
       echo "rpi-eeprom-update not found"
    fi
