From 9f550ca53801b2b9c6c1c7a4d02ad525c704b145 Mon Sep 17 00:00:00 2001
From: Robert Marko <robert.marko@sartura.hr>
Date: Thu, 14 May 2020 13:40:53 +0200
Subject: [PATCH] Remove distro check

Signed-off-by: Robert Marko <robert.marko@sartura.hr>
---
 configure.ac | 28 ----------------------------
 1 file changed, 28 deletions(-)

--- a/configure.ac
+++ b/configure.ac
@@ -78,34 +78,6 @@ esac
 LT_INIT
 AC_SUBST([LIBTOOL_DEPS])
 
-# Detect the distribution. This is used for the default configuration and
-# for some distro-specific build options.
-AC_MSG_CHECKING([host distribution])
-AC_ARG_WITH(distro, AS_HELP_STRING([--with-distro=DISTRO], [Specify the Linux distribution to target: One of redhat, redhatenterpriseserver, oracle, centos, fedora, suse, gentoo, debian, arch, slackware, plamo, paldo, openmandriva, pardus, sparclinux, altlinux.]))
-if type lsb_release >/dev/null 2>&1 && test "z$with_distro" = "z"; then
-	with_distro=`lsb_release -is`
-fi
-if test "z$with_distro" = "z"; then
-	AC_CHECK_FILE(/etc/redhat-release,with_distro="redhat")
-	AC_CHECK_FILE(/etc/oracle-release,with_distro="oracle")
-	AC_CHECK_FILE(/etc/sparclinux-release,with_distro="sparclinux")
-	AC_CHECK_FILE(/etc/centos-release,with_distro="centos")
-	AC_CHECK_FILE(/etc/fedora-release,with_distro="fedora")
-	AC_CHECK_FILE(/etc/SuSE-release,with_distro="suse")
-	AC_CHECK_FILE(/etc/gentoo-release,with_distro="gentoo")
-	AC_CHECK_FILE(/etc/debian_version,with_distro="debian")
-	AC_CHECK_FILE(/etc/arch-release,with_distro="arch")
-	AC_CHECK_FILE(/etc/slackware-version,with_distro="slackware")
-	AC_CHECK_FILE(/etc/plamo-version,with_distro="plamo")
-	AC_CHECK_FILE(/etc/frugalware-release,with_distro="frugalware")
-	AC_CHECK_FILE(/etc/mandrakelinux-release, with_distro="openmandriva")
-	AC_CHECK_FILE(/etc/mandriva-release,with_distro="openmandriva")
-	AC_CHECK_FILE(/etc/pardus-release,with_distro="pardus")
-	AC_CHECK_FILE(/etc/altlinux-release,with_distro="altlinux")
-	AC_CHECK_FILE(/etc/pld-release,with_distro="pld")
-fi
-with_distro=`echo ${with_distro} | tr '[[:upper:]]' '[[:lower:]]'`
-
 if test "z$with_distro" = "zforsparc"; then
 	with_distro="sparclinux"
 fi
