From 4233be14a34d754a70b8b6f6fa42d21f35c6e030 Mon Sep 17 00:00:00 2001
From: Alan Modra <amodra@gmail.com>
Date: Sat, 10 Sep 2022 07:30:57 +0930
Subject: [PATCH 043/160] Re: PR29466, APP/NO_APP with linefile

It looks like I copied the SIZE init across from
binutils/testsuite/config/default.exp without some necessary editing.

	PR 29466
	* testsuite/config/default.exp (SIZE): Adjust relative path.

(cherry picked from commit 1180f540d5f2f7751b5309bdd6c38d69fcf699e7)
---
 gas/testsuite/config/default.exp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/gas/testsuite/config/default.exp
+++ b/gas/testsuite/config/default.exp
@@ -53,7 +53,7 @@ if ![info exists NMFLAGS] then {
 }
 
 if ![info exists SIZE] then {
-    set SIZE [findfile $base_dir/size]
+    set SIZE [findfile $base_dir/../../binutils/size]
 }
 
 if ![info exists SIZEFLAGS] then {
