--- a/configure
+++ b/configure
@@ -24873,17 +24873,12 @@ echo "$as_me:$LINENO: result: and out ag
 echo "${ECHO_T}and out again" >&6
 
 echo $ECHO_N "ordering CD from http://people.ee.ethz.ch/~oetiker/wish $ac_c" 1>&6
-sleep 1
 echo $ECHO_N ".$ac_c" 1>&6
-sleep 2
 echo $ECHO_N ".$ac_c" 1>&6
-sleep 1
 echo $ECHO_N ".$ac_c" 1>&6
-sleep 3
 echo $ECHO_N ".$ac_c" 1>&6
 echo $ECHO_N ".$ac_c" 1>&6
 echo $ECHO_N ".$ac_c" 1>&6
-sleep 2
 echo "$as_me:$LINENO: result:  just kidding ;-)" >&5
 echo "${ECHO_T} just kidding ;-)" >&6
 echo
