--- a/setup.py
+++ b/setup.py
@@ -125,7 +125,7 @@ setup(name='zope.interface',
           "Framework :: Zope :: 3",
           "Topic :: Software Development :: Libraries :: Python Modules",
       ],
-      packages=find_packages('src'),
+      packages=find_packages('src', exclude=["*.tests", "*.tests.*"]),
       package_dir={'': 'src'},
       namespace_packages=["zope"],
       cmdclass={
@@ -133,6 +133,7 @@ setup(name='zope.interface',
       },
       test_suite='zope.interface.tests',
       include_package_data=True,
+      exclude_package_data={'':['tests/*']},
       zip_safe=False,
       tests_require=tests_require,
       install_requires=['setuptools'],
