--- a/m4/boost.m4
+++ b/m4/boost.m4
@@ -671,12 +671,15 @@ LDFLAGS=$boost_filesystem_save_LDFLAGS
 BOOST_DEFUN([Context],
 [boost_context_save_LIBS=$LIBS
  boost_context_save_LDFLAGS=$LDFLAGS
-if test $boost_major_version -ge 157; then
-  BOOST_THREAD([$1], [$2])
-  m4_pattern_allow([^BOOST_THREAD_(LIBS|LDFLAGS)$])dnl
-  LIBS="$LIBS $BOOST_THREAD_LIBS"
-  LDFLAGS="$LDFLAGS $BOOST_THREAD_LDFLAGS"
-fi
+# This check is flawed and doesn't actually verify if BOOST::THREAD
+# is required on the system, producing only false positives on the
+# platforms supported by OpenWRT.
+#if test $boost_major_version -ge 157; then
+#  BOOST_THREAD([$1], [$2])
+#  m4_pattern_allow([^BOOST_THREAD_(LIBS|LDFLAGS)$])dnl
+#  LIBS="$LIBS $BOOST_THREAD_LIBS"
+#  LDFLAGS="$LDFLAGS $BOOST_THREAD_LDFLAGS"
+#fi
 
 if test $boost_major_version -ge 169; then
 
