diff -uprN a/src/mips/getcontext.S b/src/mips/getcontext.S
--- a/src/mips/getcontext.S	2012-10-06 12:54:38.000000000 +0800
+++ b/src/mips/getcontext.S	2016-06-08 13:35:25.033051679 +0800
@@ -24,12 +24,12 @@ OF CONTRACT, TORT OR OTHERWISE, ARISING
 WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
 
 #include "offsets.h"
-#include <endian.h>
 
 	.text
+	.set nomips16
 
 #if _MIPS_SIM == _ABIO32
-# if __BYTE_ORDER == __BIG_ENDIAN
+# if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
 #  define OFFSET 4
 # else
 #  define OFFSET 0

