--- a/ipt2socks.c
+++ b/ipt2socks.c
@@ -28,6 +28,36 @@
 
 #define IPT2SOCKS_VERSION "ipt2socks v1.1.2 <https://github.com/zfl9/ipt2socks>"
 
+#ifdef __MUSL__
+#include <unistd.h>
+#include <sys/syscall.h>
+
+ssize_t splice(int __fdin, __off64_t *__offin, int __fdout,
+        __off64_t *__offout, size_t __len, unsigned int __flags) {
+#ifdef __NR_splice
+	return syscall(__NR_splice, __fdin, __offin, __fdout, __offout, __len, __flags);
+#else
+	(void)__fdin;
+	(void)__offin;
+	(void)__fdout;
+	(void)__offout;
+	(void)__len;
+	(void)__flags;
+	errno = ENOSYS;
+	return -1;
+#endif
+}
+
+#undef SPLICE_F_MOVE
+#undef SPLICE_F_NONBLOCK
+#undef SPLICE_F_MORE
+
+#define SPLICE_F_MOVE 1
+#define SPLICE_F_NONBLOCK 2
+#define SPLICE_F_MORE 4
+
+#endif /* __MUSL__ */
+
 enum {
     OPT_ENABLE_TCP         = 0x01 << 0, // enable tcp proxy
     OPT_ENABLE_UDP         = 0x01 << 1, // enable udp proxy

--- a/Makefile
+++ b/Makefile
@@ -1,5 +1,5 @@
 CC = gcc
-CFLAGS = -std=c99 -Wall -Wextra -O2 -pthread
+CFLAGS = -std=c99 -Wall -Wextra -O2 -pthread -D __MUSL__
 LIBS = -lm
 SRCS = ipt2socks.c logutils.c lrucache.c netutils.c protocol.c
 OBJS = $(SRCS:.c=.o)