--- a/drivers/pci/dwc/Kconfig
+++ b/drivers/pci/dwc/Kconfig
@@ -6,7 +6,6 @@ config PCIE_DW
 config PCIE_DW_HOST
         bool
 	depends on PCI
-	depends on PCI_MSI_IRQ_DOMAIN
         select PCIE_DW
 
 config PCIE_DW_EP
@@ -74,7 +73,6 @@ config PCI_IMX6
 	bool "Freescale i.MX6 PCIe controller"
 	depends on PCI
 	depends on SOC_IMX6Q
-	depends on PCI_MSI_IRQ_DOMAIN
 	select PCIEPORTBUS
 	select PCIE_DW_HOST
 
