--- a/raddb/certs/ca.cnf
+++ b/raddb/certs/ca.cnf
@@ -14,7 +14,7 @@ private_key		= $dir/ca.key
 RANDFILE		= $dir/.rand
 name_opt		= ca_default
 cert_opt		= ca_default
-default_days		= 60
+default_days		= 365
 default_crl_days	= 30
 default_md		= sha256
 preserve		= no
--- a/raddb/certs/client.cnf
+++ b/raddb/certs/client.cnf
@@ -14,7 +14,7 @@ private_key		= $dir/ca.key
 RANDFILE		= $dir/.rand
 name_opt		= ca_default
 cert_opt		= ca_default
-default_days		= 60
+default_days		= 365
 default_crl_days	= 30
 default_md		= sha256
 preserve		= no
--- a/raddb/certs/server.cnf
+++ b/raddb/certs/server.cnf
@@ -14,7 +14,7 @@ private_key		= $dir/server.key
 RANDFILE		= $dir/.rand
 name_opt		= ca_default
 cert_opt		= ca_default
-default_days		= 60
+default_days		= 365
 default_crl_days	= 30
 default_md		= sha256
 preserve		= no
