--- a/src/rtl8372n.c
+++ b/src/rtl8372n.c
@@ -1126,6 +1126,7 @@
 
 
 	switch (proto) {
+#ifdef CONFIG_NET_DSA_TAG_MXL_862XX_8021Q
 	case DSA_TAG_PROTO_MXL862_8021Q:
 		ret = rtl8372n_teardown_tag_rtl(ds);
 		if (ret)
@@ -1134,6 +1135,7 @@
 		if (ret)
 			return ret;
 		break;
+#endif
 	case DSA_TAG_PROTO_RTL8_4:
 		ret = rtl8372n_teardown_tag_8021q(ds);
 		if (ret)
@@ -1893,9 +1895,11 @@
 
 	rtnl_lock();
 	switch (priv->tag_proto) {
+#ifdef CONFIG_NET_DSA_TAG_MXL_862XX_8021Q
 	case DSA_TAG_PROTO_MXL862_8021Q:
 		ret = rtl8372n_set_tag_8021q(ds);
 		break;
+#endif
 	case DSA_TAG_PROTO_RTL8_4:
 		ret = rtl8372n_set_tag_rtl(ds);
 		break;
