--- a/lib/Automake/Config.in
+++ b/lib/Automake/Config.in
@@ -32,7 +32,7 @@ our $PACKAGE = '@PACKAGE@';
 our $PACKAGE_BUGREPORT = '@PACKAGE_BUGREPORT@';
 our $VERSION = '@VERSION@';
 our $RELEASE_YEAR = '@RELEASE_YEAR@';
-our $libdir = '@datadir@/@PACKAGE@-@APIVERSION@';
+our $libdir = $ENV{'STAGING_DIR'} ? $ENV{'STAGING_DIR'} . '/../host/share/@PACKAGE@-@APIVERSION@' : '@datadir@/@PACKAGE@-@APIVERSION@';
 
 our $perl_threads = 0;
 # We need at least this version for CLONE support.
--- a/bin/aclocal.in
+++ b/bin/aclocal.in
@@ -1,10 +1,12 @@
-#!@PERL@ -w
+#!@PERL@
 # -*- perl -*-
 # @configure_input@
 
 eval 'case $# in 0) exec @PERL@ -S "$0";; *) exec @PERL@ -S "$0" "$@";; esac'
     if 0;
 
+$^W = 1;
+
 # aclocal - create aclocal.m4 by scanning configure.ac
 
 # Copyright (C) 1996-2017 Free Software Foundation, Inc.
@@ -27,7 +29,7 @@ eval 'case $# in 0) exec @PERL@ -S "$0";
 
 BEGIN
 {
-  @Aclocal::perl_libdirs = ('@datadir@/@PACKAGE@-@APIVERSION@')
+  @Aclocal::perl_libdirs = ($ENV{'STAGING_DIR'} ? $ENV{'STAGING_DIR'} . '/../host/share/@PACKAGE@-@APIVERSION@' : '@datadir@/@PACKAGE@-@APIVERSION@')
     unless @Aclocal::perl_libdirs;
   unshift @INC, @Aclocal::perl_libdirs;
 }
@@ -69,8 +71,8 @@ $perl_threads = 0;
 # ACLOCAL_PATH environment variable, and reset with the '--system-acdir'
 # option.
 my @user_includes = ();
-my @automake_includes = ("@datadir@/aclocal-$APIVERSION");
-my @system_includes = ('@datadir@/aclocal');
+my @automake_includes = ($ENV{'STAGING_DIR'} ? $ENV{'STAGING_DIR'} . "/../host/share/aclocal-$APIVERSION" : "@datadir@/aclocal-$APIVERSION");
+my @system_includes = ($ENV{'STAGING_DIR'} ? $ENV{'STAGING_DIR'} . '/../host/share/aclocal' : '@datadir@/aclocal');
 
 # Whether we should copy M4 file in $user_includes[0].
 my $install = 0;
--- a/bin/automake.in
+++ b/bin/automake.in
@@ -1,10 +1,12 @@
-#!@PERL@ -w
+#!@PERL@
 # -*- perl -*-
 # @configure_input@
 
 eval 'case $# in 0) exec @PERL@ -S "$0";; *) exec @PERL@ -S "$0" "$@";; esac'
     if 0;
 
+$^W = 1;
+
 # automake - create Makefile.in from Makefile.am
 # Copyright (C) 1994-2017 Free Software Foundation, Inc.
 
@@ -31,7 +33,7 @@ use strict;
 
 BEGIN
 {
-  @Automake::perl_libdirs = ('@datadir@/@PACKAGE@-@APIVERSION@')
+  @Automake::perl_libdirs = ($ENV{'STAGING_DIR'} ? $ENV{'STAGING_DIR'} . '/../host/share/@PACKAGE@-@APIVERSION@' : '@datadir@/@PACKAGE@-@APIVERSION@')
     unless @Automake::perl_libdirs;
   unshift @INC, @Automake::perl_libdirs;
 
--- a/t/wrap/aclocal.in
+++ b/t/wrap/aclocal.in
@@ -1,6 +1,8 @@
-#!@PERL@ -w
+#!@PERL@
 # @configure_input@
 
+$^W = 1;
+
 # Copyright (C) 2012-2017 Free Software Foundation, Inc.
 
 # This program is free software; you can redistribute it and/or modify
--- a/t/wrap/automake.in
+++ b/t/wrap/automake.in
@@ -1,6 +1,8 @@
-#!@PERL@ -w
+#!@PERL@
 # @configure_input@
 
+$^W = 1;
+
 # Copyright (C) 2012-2017 Free Software Foundation, Inc.
 
 # This program is free software; you can redistribute it and/or modify
