--- a/bin/automake.in
+++ b/bin/automake.in
@@ -4513,7 +4513,7 @@ sub handle_gettext ()
 	    && grep ($_ eq 'intl', @subdirs));
     }
 
-  require_file ($ac_gettext_location, GNU, 'ABOUT-NLS');
+  require_file ($ac_gettext_location, GNITS, 'ABOUT-NLS');
 }
 
 # Emit makefile footer.
@@ -5641,7 +5641,7 @@ sub check_gnu_standards ()
           # otherwise require non-.md.
           my $required
            = (! -f $file && -f "$file.md") ? "$file.md" : $file;
-          require_file ("$am_file.am", GNU, $required);
+          require_file ("$am_file.am", GNITS, $required);
         }
 
       # Accept one of these three licenses; default to COPYING.
@@ -5655,7 +5655,7 @@ sub check_gnu_standards ()
 	      last;
 	    }
 	}
-      require_file ("$am_file.am", GNU, 'COPYING')
+      require_file ("$am_file.am", GNITS, 'COPYING')
 	unless $license;
     }
 
