diff -ruN automake-1.15/m4/silent.m4 automake-1.15.mod/m4/silent.m4
--- automake-1.15/m4/silent.m4	2014-12-30 22:53:05.000000000 +0100
+++ automake-1.15.mod/m4/silent.m4	2015-03-11 12:00:26.280586399 +0100
@@ -41,7 +41,8 @@
 else
   am_cv_make_support_nested_variables=no
 fi])
-if test $am_cv_make_support_nested_variables = yes; then
+#if test $am_cv_make_support_nested_variables = yes; then
+if false; then
   dnl Using '$V' instead of '$(V)' breaks IRIX make.
   AM_V='$(V)'
   AM_DEFAULT_V='$(AM_DEFAULT_VERBOSITY)'
