--- a/lib/am/clean.am
+++ b/lib/am/clean.am
@@ -39,7 +39,7 @@ distclean-generic:
 ## If you change distclean here, you probably also want to change
 ## maintainer-clean below.
 distclean:
-	-rm -f %MAKEFILE%
+	-touch %MAKEFILE%
 
 maintainer-clean-am: maintainer-clean-generic distclean-am
 maintainer-clean-generic:
@@ -51,7 +51,7 @@ maintainer-clean-generic:
 
 ## See comment for distclean.
 maintainer-clean:
-	-rm -f %MAKEFILE%
+	-touch %MAKEFILE%
 
 .PHONY: clean mostlyclean distclean maintainer-clean \
 clean-generic mostlyclean-generic distclean-generic maintainer-clean-generic
--- a/lib/am/libtool.am
+++ b/lib/am/libtool.am
@@ -25,4 +25,4 @@ clean-libtool:
 
 ?TOPDIR_P?distclean-am: distclean-libtool
 ?TOPDIR_P?distclean-libtool:
-?TOPDIR_P?	-rm -f libtool config.lt
+?TOPDIR_P?	-touch libtool config.lt
