From: coolsnowwolf <coolsnowwolf@gmail.com>
Subject: [PATCH] libnftnl: support the Broadcom masquerade fullcone attribute

Signed-off-by: coolsnowwolf <coolsnowwolf@gmail.com>

--- a/include/libnftnl/expr.h
+++ b/include/libnftnl/expr.h
@@ -269,6 +269,7 @@
 	NFTNL_EXPR_MASQ_FLAGS		= NFTNL_EXPR_BASE,
 	NFTNL_EXPR_MASQ_REG_PROTO_MIN,
 	NFTNL_EXPR_MASQ_REG_PROTO_MAX,
+	NFTNL_EXPR_MASQ_REG_FULLCONE,
 	__NFTNL_EXPR_MASQ_MAX
 };
 
--- a/include/linux/netfilter/nf_tables.h
+++ b/include/linux/netfilter/nf_tables.h
@@ -1505,6 +1505,7 @@
 	NFTA_MASQ_FLAGS,
 	NFTA_MASQ_REG_PROTO_MIN,
 	NFTA_MASQ_REG_PROTO_MAX,
+	NFTA_MASQ_REG_FULLCONE,
 	__NFTA_MASQ_MAX
 };
 #define NFTA_MASQ_MAX		(__NFTA_MASQ_MAX - 1)
--- a/src/expr/masq.c
+++ b/src/expr/masq.c
@@ -20,6 +20,7 @@
 	uint32_t		flags;
 	enum nft_registers	sreg_proto_min;
 	enum nft_registers	sreg_proto_max;
+	uint8_t			fullcone;
 };
 
 static int
@@ -38,6 +39,11 @@
 	case NFTNL_EXPR_MASQ_REG_PROTO_MAX:
 		memcpy(&masq->sreg_proto_max, data, data_len);
 		break;
+	case NFTNL_EXPR_MASQ_REG_FULLCONE:
+		if (data_len != sizeof(masq->fullcone))
+			return -1;
+		memcpy(&masq->fullcone, data, data_len);
+		break;
 	}
 	return 0;
 }
@@ -58,6 +64,9 @@
 	case NFTNL_EXPR_MASQ_REG_PROTO_MAX:
 		*data_len = sizeof(masq->sreg_proto_max);
 		return &masq->sreg_proto_max;
+	case NFTNL_EXPR_MASQ_REG_FULLCONE:
+		*data_len = sizeof(masq->fullcone);
+		return &masq->fullcone;
 	}
 	return NULL;
 }
@@ -77,6 +86,10 @@
 		if (mnl_attr_validate(attr, MNL_TYPE_U32) < 0)
 			abi_breakage();
 		break;
+	case NFTA_MASQ_REG_FULLCONE:
+		if (mnl_attr_validate(attr, MNL_TYPE_U8) < 0)
+			abi_breakage();
+		break;
 	}
 
 	tb[type] = attr;
@@ -96,6 +109,8 @@
 	if (e->flags & (1 << NFTNL_EXPR_MASQ_REG_PROTO_MAX))
 		mnl_attr_put_u32(nlh, NFTA_MASQ_REG_PROTO_MAX,
 				 htobe32(masq->sreg_proto_max));
+	if (e->flags & (1 << NFTNL_EXPR_MASQ_REG_FULLCONE))
+		mnl_attr_put_u8(nlh, NFTA_MASQ_REG_FULLCONE, masq->fullcone);
 }
 
 static int
@@ -121,6 +136,10 @@
 			be32toh(mnl_attr_get_u32(tb[NFTA_MASQ_REG_PROTO_MAX]));
 		e->flags |= (1 << NFTNL_EXPR_MASQ_REG_PROTO_MAX);
 	}
+	if (tb[NFTA_MASQ_REG_FULLCONE]) {
+		masq->fullcone = mnl_attr_get_u8(tb[NFTA_MASQ_REG_FULLCONE]);
+		e->flags |= (1 << NFTNL_EXPR_MASQ_REG_FULLCONE);
+	}
 
 	return 0;
 }
@@ -145,6 +164,10 @@
 		ret = snprintf(buf + offset, remain, "flags 0x%x ", masq->flags);
 		SNPRINTF_BUFFER_SIZE(ret, remain, offset);
 	}
+	if (e->flags & (1 << NFTNL_EXPR_MASQ_REG_FULLCONE)) {
+		ret = snprintf(buf + offset, remain, "fullcone %u ", masq->fullcone);
+		SNPRINTF_BUFFER_SIZE(ret, remain, offset);
+	}
 
 	return offset;
 }
@@ -153,6 +176,7 @@
 	[NFTNL_EXPR_MASQ_FLAGS]         = { .maxlen = sizeof(uint32_t) },
 	[NFTNL_EXPR_MASQ_REG_PROTO_MIN] = { .maxlen = sizeof(uint32_t) },
 	[NFTNL_EXPR_MASQ_REG_PROTO_MAX] = { .maxlen = sizeof(uint32_t) },
+	[NFTNL_EXPR_MASQ_REG_FULLCONE]  = { .maxlen = sizeof(uint8_t) },
 };
 
 struct expr_ops expr_ops_masq = {
