diff -urN gengetopt-2.22.6/Makefile.am gengetopt-2.22.6.new/Makefile.am
--- gengetopt-2.22.6/Makefile.am	2012-11-02 06:26:54.000000000 -0700
+++ gengetopt-2.22.6.new/Makefile.am	2015-05-21 17:43:15.183083542 -0700
@@ -18,6 +18,6 @@
 ACLOCAL_AMFLAGS = -I m4 -I gl/m4
 
 EXTRA_DIST = configure TODO LICENSE gl/m4/gnulib-cache.m4
-SUBDIRS = gl src doc tests
+SUBDIRS = gl src
 
-gengetoptdoc_DATA = ChangeLog COPYING NEWS THANKS INSTALL README LICENSE
\ No newline at end of file
+gengetoptdoc_DATA = ChangeLog COPYING NEWS THANKS INSTALL README LICENSE
