--- a/Makefile.in
+++ b/Makefile.in
@@ -451,7 +451,7 @@ target_alias = @target_alias@
 top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
-SUBDIRS = include lib sanei backend frontend tools doc po testsuite
+SUBDIRS = include lib sanei backend frontend
 DIST_SUBDIRS = include lib sanei backend frontend tools doc po japi testsuite
 dist_doc_DATA = AUTHORS COPYING LICENSE NEWS PROBLEMS PROJECTS README \
 	README.aix README.beos README.darwin README.djpeg \
