--- a/config
+++ b/config
@@ -24,6 +24,14 @@ ifeq ($(KVER),$(filter 5.4%,$(KVER)))
 	OS_VER=5_4
 endif
 
+ifeq ($(KVER),$(filter 5.10%,$(KVER)))
+OS_VER=5_10
+endif
+
+ifeq ($(KVER),$(filter 5.15%,$(KVER)))
+OS_VER=5_15
+endif
+
 ifeq ($(KVER), 3.4.0)
 	OS_VER=3_4
 endif
@@ -132,7 +140,7 @@ ifeq ($(ARCH), arm)
 endif
 
 ifeq ($(ARCH), arm64)
-	ifeq ($(KVER),$(filter 4.1% 4.4% 4.9% 5.4%,$(KVER)))
+	ifeq ($(KVER),$(filter 4.1% 4.4% 4.9% 5.4% 5.10% 5.15%,$(KVER)))
 		CPU_CFLAG=  -DMODULE -Os -pipe -march=armv8-a -mcpu=cortex-a53+crypto -fno-caller-saves -fno-strict-aliasing -Werror -fno-common -Wno-format-security -Wno-pointer-sign -Wno-unused-but-set-variable -Wno-error=unused-result -mcmodel=large
 	endif
 endif
--- a/make/linux_opt.mk
+++ b/make/linux_opt.mk
@@ -422,7 +422,7 @@ ifeq (KSLIB, $(MODULE_TYPE))
       KASAN_SHADOW_SCALE_SHIFT := 3
   endif
 
-  ifeq (5_4, $(OS_VER))
+  ifeq ($(OS_VER),$(filter 5_4 5_10 5_15, $(OS_VER)))
       ifeq ($(ARCH), arm64)
           KASAN_OPTION += -DKASAN_SHADOW_SCALE_SHIFT=$(KASAN_SHADOW_SCALE_SHIFT)
        endif
@@ -453,7 +453,7 @@ ifeq (KSLIB, $(MODULE_TYPE))
 
   endif
 
-  ifeq ($(OS_VER),$(filter 4_4 5_4, $(OS_VER)))
+  ifeq ($(OS_VER),$(filter 4_4 5_4 5_10 5_15, $(OS_VER)))
                 MODULE_CFLAG += -DKVER34
                 MODULE_CFLAG += -DKVER32
             MODULE_CFLAG += -DLNX26_22
