--- a/src/init/ssdk_plat.c
+++ b/src/init/ssdk_plat.c
@@ -580,7 +580,7 @@ static int miibus_get(a_uint32_t dev_id)
 		if (!mdio_node)
 			mdio_node = of_find_compatible_node(NULL, NULL, "qcom,qca-mdio");
 	} else
-		mdio_node = of_find_compatible_node(NULL, NULL, "virtual,mdio-gpio");
+		mdio_node = of_find_compatible_node(NULL, NULL, "qcom,ipq8064-mdio");
 
 	if (!mdio_node) {
 		SSDK_ERROR("No MDIO node found in DTS!\n");
