--- a/src/init/ssdk_clk.c
+++ b/src/init/ssdk_clk.c
@@ -1282,10 +1282,8 @@ ssdk_mp_reset_init(void)
 
 	for (i = 0; i < MP_BCR_RST_MAX; i++) {
 		rst = of_reset_control_get(rst_node, mp_rst_ids[i]);
-		if (IS_ERR(rst)) {
-			SSDK_ERROR("%s not exist!\n", mp_rst_ids[i]);
-			return;
-		}
+		if (IS_ERR(rst))
+			continue;
 		ssdk_gcc_reset(rst, SSDK_RESET_ASSERT);
 		msleep(200);
 		ssdk_gcc_reset(rst, SSDK_RESET_DEASSERT);
