--- a/src/hsl/phy/hsl_phy.c
+++ a/src/hsl/phy/hsl_phy.c
@@ -762,7 +762,7 @@
 	*phydev = miibus->phy_map[phy_addr];
 	if(*phydev == NULL)
 	{
-		SSDK_ERROR("phy_addr %d phydev is NULL\n", phy_addr);
+		// SSDK_ERROR("phy_addr %d phydev is NULL\n", phy_addr);
 		return SW_NOT_INITIALIZED;
 	}
 	pdev_addr = (*phydev)->addr;
@@ -772,7 +772,7 @@
 	*phydev = mdiobus_get_phy(miibus, phy_addr);
 	if(*phydev == NULL)
 	{
-		SSDK_ERROR("phy_addr %d phydev is NULL\n", phy_addr);
+		// SSDK_ERROR("phy_addr %d phydev is NULL\n", phy_addr);
 		return SW_NOT_INITIALIZED;
 	}
 	pdev_addr = (*phydev)->mdio.addr;
--- a/src/adpt/hppe/adpt_hppe_portctrl.c
+++ b/src/adpt/hppe/adpt_hppe_portctrl.c
@@ -2277,9 +2277,9 @@
 		{
 			if(*mode0 != PORT_WRAPPER_MAX && *mode0 != PORT_WRAPPER_PSGMII)
 			{
-				SSDK_ERROR("when the port_interface_mode of port %d is %d, "
+				/* SSDK_ERROR("when the port_interface_mode of port %d is %d, "
 					"mode0:%d cannot be supported\n",
-					port_id, phy_info->port_mode[port_id], *mode0);
+					port_id, phy_info->port_mode[port_id], *mode0); */
 				return SW_NOT_SUPPORTED;
 			}
 			*mode0 = PORT_WRAPPER_PSGMII;
@@ -2296,9 +2296,9 @@
 			if((*mode0 != PORT_WRAPPER_MAX && *mode0 != PORT_WRAPPER_QSGMII) ||
 				port_id == SSDK_PHYSICAL_PORT5)
 			{
-				SSDK_ERROR("when the port_interface_mode of port %d is %d, "
+				/* SSDK_ERROR("when the port_interface_mode of port %d is %d, "
 					"mode0:%d cannot be supported\n",
-					port_id, phy_info->port_mode[port_id], *mode0);
+					port_id, phy_info->port_mode[port_id], *mode0); */
 				return SW_NOT_SUPPORTED;
 			}
 			*mode0 = PORT_WRAPPER_QSGMII;
@@ -2310,9 +2310,9 @@
 				if((*mode0 != PORT_WRAPPER_MAX && *mode0 != PORT_WRAPPER_UQXGMII) ||
 					port_id == SSDK_PHYSICAL_PORT5)
 				{
-					SSDK_ERROR("when the port_interface_mode of port %d is %d, "
+					/* SSDK_ERROR("when the port_interface_mode of port %d is %d, "
 						"mode0:%d cannot be supported\n",
-						port_id, phy_info->port_mode[port_id], *mode0);
+						port_id, phy_info->port_mode[port_id], *mode0); */
 					return SW_NOT_SUPPORTED;
 				}
 				*mode0 = PORT_WRAPPER_UQXGMII;
@@ -2326,10 +2326,10 @@
 			{
 				if(port_id != SSDK_PHYSICAL_PORT5)
 				{
-					SSDK_ERROR("when the port_interface_mode of port %d is %d, "
+					/* SSDK_ERROR("when the port_interface_mode of port %d is %d, "
 						"mode0:%d cannot be supported\n",
 						port_id, phy_info->port_mode[port_id],
-						*mode0);
+						*mode0); */
 					return SW_NOT_SUPPORTED;
 				}
 				else
