From: Piotr Dymacz <pepe2k@gmail.com>
Subject: kernel/mtd: add support for EON EN25Q128

Signed-off-by: Piotr Dymacz <pepe2k@gmail.com>
---
 drivers/mtd/spi-nor/spi-nor.c | 1 +
 1 file changed, 1 insertion(+)

--- a/drivers/mtd/spi-nor/eon.c
+++ b/drivers/mtd/spi-nor/eon.c
@@ -18,6 +18,11 @@ static const struct flash_info eon_nor_p
 		.name = "en25p64",
 		.size = SZ_8M,
 	}, {
+		.id = SNOR_ID(0x1c, 0x30, 0x18),
+		.name = "en25q128",
+		.size = SZ_16M,
+		.no_sfdp_flags = SECT_4K,
+	}, {
 		.id = SNOR_ID(0x1c, 0x30, 0x14),
 		.name = "en25q80a",
 		.size = SZ_1M,
