From d68b4aa22e8c625685bfad642dd7337948dc0ad1 Mon Sep 17 00:00:00 2001
From: Koen Vandeputte <koen.vandeputte@ncentric.com>
Date: Mon, 6 Jan 2020 13:07:56 +0100
Subject: [PATCH] mtd: spi-nor: add support for Gigadevice GD25D05

Signed-off-by: Koen Vandeputte <koen.vandeputte@ncentric.com>
---
 drivers/mtd/spi-nor/spi-nor.c | 5 +++++
 1 file changed, 5 insertions(+)

--- a/drivers/mtd/spi-nor/gigadevice.c
+++ b/drivers/mtd/spi-nor/gigadevice.c
@@ -35,6 +35,12 @@ static const struct spi_nor_fixups gd25q
 
 static const struct flash_info gigadevice_nor_parts[] = {
 	{
+		.id = SNOR_ID(0xc8, 0x40, 0x10),
+		.name = "gd25q05",
+		.size = SZ_64K,
+		.flags = SPI_NOR_HAS_LOCK | SPI_NOR_HAS_TB,
+		.no_sfdp_flags = SECT_4K | SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ,
+	}, {
 		.id = SNOR_ID(0xc8, 0x40, 0x15),
 		.name = "gd25q16",
 		.size = SZ_2M,
