--- a/packet_plugin/rjv3/packet_plugin_rjv3_priv.c
+++ b/packet_plugin/rjv3/packet_plugin_rjv3_priv.c
@@ -132,7 +132,8 @@ static void rjv3_set_hdd_serial(uint8_t* serial_buf, char* fake_serial) {
         memmove(serial_buf, fake_serial, strnlen(fake_serial, MAX_PROP_LEN));
         return;
     }
-#ifdef __linux__
+    PR_WARN("请使用 --fake-serial 选项手动指定硬盘序列号");
+#if 0
     FILE* _fp = fopen("/etc/mtab", "r");
     char _line_buf[MAX_LINE_LEN] = {0};
     char* _line_buf_dev, *_line_buf_mountpoint;
