--- a/arch/x86/kernel/reboot.c
+++ b/arch/x86/kernel/reboot.c
@@ -448,6 +448,16 @@ static const struct dmi_system_id reboot
 		},
 	},
 
+	/* PC Engines */
+	{	/* Handle problems with rebooting on PC Engines apu2 */
+		.callback = set_pci_reboot,
+		.ident = "PC Engines apu2",
+		.matches = {
+			DMI_MATCH(DMI_BOARD_VENDOR, "PC Engines"),
+			DMI_MATCH(DMI_BOARD_NAME, "apu2"),
+		},
+	},
+
 	/* Sony */
 	{	/* Handle problems with rebooting on Sony VGN-Z540N */
 		.callback = set_bios_reboot,
