--- a/glob.h
+++ b/glob.h
@@ -9,6 +9,12 @@
 
 #include <linux/ctype.h>
 
+#include <linux/lockdep.h>
+
+#ifndef lockdep_assert_not_held
+#define lockdep_assert_not_held(l)		do { (void)(l); } while (0)
+#endif
+
 #include "unicode.h"
 #include "vfs_cache.h"
 
--- a/vfs.c
+++ b/vfs.c
@@ -1705,7 +1705,7 @@
 
 	err = -ENOTEMPTY;
 	if (dst_dent != trap_dent && !d_really_is_positive(dst_dent)) {
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 12, 0)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0)
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 3, 0)
 		struct renamedata rd = {
 			.old_mnt_idmap	= src_idmap,
@@ -1715,7 +1715,7 @@
 			.new_dir	= d_inode(dst_dent_parent),
 			.new_dentry	= dst_dent,
 		};
-#else
+#elif LINUX_VERSION_CODE >= KERNEL_VERSION(5, 12, 0)
 		struct renamedata rd = {
 			.old_mnt_userns	= src_user_ns,
 			.old_dir	= d_inode(src_dent_parent),
@@ -1724,6 +1724,15 @@
 			.new_dir	= d_inode(dst_dent_parent),
 			.new_dentry	= dst_dent,
 		};
+#else
+		struct renamedata rd = {
+			.old_dir		= d_inode(src_dent_parent),
+			.old_dentry		= src_dent,
+			.new_dir		= d_inode(dst_dent_parent),
+			.new_dentry		= dst_dent,
+			.delegated_inode	= NULL,
+			.flags			= 0,
+		};
 #endif
 		err = vfs_rename(&rd);
 #else
