--- a/smb1pdu.c
+++ b/smb1pdu.c
@@ -887,7 +887,11 @@
 	}
 
 	ksmbd_update_fstate(&work->sess->file_table, fp, FP_INITED);
+	#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 4, 0)
+	rc = ksmbd_vfs_rename(work, &fp->filp->f_path, newname, 0);
+	#else
 	rc = ksmbd_vfs_fp_rename(work, fp, newname);
+	#endif
 	if (rc) {
 		rsp->hdr.Status.CifsError = STATUS_NO_MEMORY;
 		path_put(&path);
@@ -1712,7 +1716,11 @@
 				goto out;
 			}
 			cmd = F_SETLKW;
+			#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 9, 0)
+			flock->c.flc_type = F_RDLCK;
+			#else
 			flock->fl_type = F_RDLCK;
+			#endif
 		} else {
 			pr_err("received exclusive request\n");
 			if (!(filp->f_mode & FMODE_WRITE)) {
@@ -1722,8 +1730,13 @@
 				goto out;
 			}
 			cmd = F_SETLKW;
+			#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 9, 0)
+			flock->c.flc_type = F_WRLCK;
+			flock->c.flc_flags |= FL_SLEEP;
+			#else
 			flock->fl_type = F_WRLCK;
 			flock->fl_flags |= FL_SLEEP;
+			#endif
 		}
 
 		if (req->LockType & LOCKING_ANDX_LARGE_FILES) {
@@ -1785,8 +1798,13 @@
 		hash_for_each(conn_list, bkt, conn, hlist) {
 			spin_lock(&conn->llist_lock);
 			list_for_each_entry_safe(cmp_lock, tmp2, &conn->lock_list, clist) {
+				#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 9, 0)
+				if (file_inode(cmp_lock->fl->c.flc_file) !=
+				    file_inode(smb_lock->fl->c.flc_file))
+				#else
 				if (file_inode(cmp_lock->fl->fl_file) !=
 					file_inode(smb_lock->fl->fl_file))
+				#endif
 					continue;
 
 				if (smb_lock->zero_len &&
@@ -1915,7 +1933,11 @@
 		if (!flock)
 			goto out;
 
+		#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 9, 0)
+		flock->c.flc_type = F_UNLCK;
+		#else
 		flock->fl_type = F_UNLCK;
+		#endif
 		cmd = 0;
 
 		if (req->LockType & LOCKING_ANDX_LARGE_FILES) {
@@ -1951,8 +1973,13 @@
 		hash_for_each(conn_list, bkt, conn, hlist) {
 			spin_lock(&conn->llist_lock);
 			list_for_each_entry(cmp_lock, &conn->lock_list, clist) {
+				#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 9, 0)
+				if (file_inode(cmp_lock->fl->c.flc_file) !=
+				    file_inode(flock->c.flc_file))
+				#else
 				if (file_inode(cmp_lock->fl->fl_file) !=
 					file_inode(flock->fl_file))
+				#endif
 					continue;
 
 				if ((cmp_lock->start == offset &&
@@ -2021,7 +2048,11 @@
 		struct file_lock *rlock = NULL;
 
 		rlock = smb_flock_init(filp);
+		#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 9, 0)
+		rlock->c.flc_type = F_UNLCK;
+		#else
 		rlock->fl_type = F_UNLCK;
+		#endif
 		rlock->fl_start = smb_lock->start;
 		rlock->fl_end = smb_lock->end;
 
@@ -4962,7 +4993,7 @@
 	struct ksmbd_share_config *share = work->tcon->share_conf;
 	struct open_psx_req *psx_req;
 	struct open_psx_rsp *psx_rsp;
-	struct path path, parent_path;
+	struct path path;
 	struct kstat stat;
 	__u16 data_offset, rsp_info_level, file_info = 0;
 	__u32 oplock_flags, posix_open_flags, may_flags;
@@ -7337,7 +7368,11 @@
 	}
 
 	ksmbd_debug(SMB, "new name(%s)\n", newname);
+	#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 4, 0)
+	rc = ksmbd_vfs_rename(work, &fp->filp->f_path, newname, 0);
+	#else
 	rc = ksmbd_vfs_fp_rename(work, fp, newname);
+	#endif
 	if (rc) {
 		rsp->hdr.Status.CifsError = STATUS_UNEXPECTED_IO_ERROR;
 		goto out;
