From 25aa88e0632bfd01aa9301e6468141412216f859 Mon Sep 17 00:00:00 2001
From: Anthony Ruhier <aruhier@mailbox.org>
Date: Fri, 14 May 2021 21:12:17 +0000
Subject: [PATCH] fs/ntfs3: Fix unsupported flags by clang (#146)

Signed-off-by: Alexandre Frade <kernel@xanmod.org>
---
 fs/ntfs3/Makefile | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/fs/ntfs3/Makefile b/fs/ntfs3/Makefile
index b9aacc061781..5adc54ebac5a 100644
--- a/fs/ntfs3/Makefile
+++ b/fs/ntfs3/Makefile
@@ -3,6 +3,13 @@
 # Makefile for the ntfs3 filesystem support.
 #
 
+# to check robot warnings
+ccflags-y += -Wint-to-pointer-cast
+condflags := \
+	$(call cc-option, -Wunused-but-set-variable) \
+	$(call cc-option, -Wold-style-declaration)
+ccflags-y += $(condflags)
+
 obj-$(CONFIG_NTFS3_FS) += ntfs3.o
 
 ntfs3-y :=	attrib.o \
@@ -28,4 +35,4 @@ ntfs3-$(CONFIG_NTFS3_LZX_XPRESS) += $(addprefix lib/,\
 		decompress_common.o \
 		lzx_decompress.o \
 		xpress_decompress.o \
-		)
\ No newline at end of file
+		)
-- 
2.30.2

