--- a/CMakeLists.txt	2017-08-10 04:19:45.000000000 -0700
+++ b/CMakeLists.txt	2022-11-28 17:21:03.453548350 -0800
@@ -50,6 +50,9 @@
 
 project(lzop VERSION 1.04 LANGUAGES C)
 
+# configuration options
+option(ENABLE_DOCS     "Install documentation."    ON)
+
 # install directories
 if(NOT CMAKE_INSTALL_PREFIX)
     message(FATAL_ERROR "ERROR: CMAKE_INSTALL_PREFIX is not defined.")
@@ -186,9 +189,11 @@
 
 install(TARGETS lzop DESTINATION "${CMAKE_INSTALL_FULL_BINDIR}")
 
+if(ENABLE_DOCS)
 set(f AUTHORS COPYING NEWS README THANKS doc/lzop.html doc/lzop.man doc/lzop.pod doc/lzop.ps doc/lzop.tex doc/lzop.txt)
 install(FILES ${f} DESTINATION "${CMAKE_INSTALL_FULL_DOCDIR}")
 install(FILES doc/lzop.1 DESTINATION "${CMAKE_INSTALL_FULL_MANDIR}/man1")
+endif() # ENABLE_DOCS
 
 endif() # CMAKE_INSTALL_FULL_LIBDIR
 
