--- a/tools/bpf/bpftool/btf_dumper.c
+++ b/tools/bpf/bpftool/btf_dumper.c
@@ -5,6 +5,7 @@
 #include <stdio.h> /* for (FILE *) used by json_writer */
 #include <string.h>
 #include <unistd.h>
+#include <linux/compiler.h>
 #include <asm/byteorder.h>
 #include <linux/bitops.h>
 #include <linux/btf.h>
--- a/tools/bpf/bpftool/map_perf_ring.c
+++ b/tools/bpf/bpftool/map_perf_ring.c
@@ -16,6 +16,7 @@
 #include <time.h>
 #include <unistd.h>
 #include <linux/bpf.h>
+#include <linux/compiler.h>
 #include <linux/perf_event.h>
 #include <sys/ioctl.h>
 #include <sys/mman.h>
