--- a/sycls.c
+++ b/sycls.c
@@ -1329,6 +1329,7 @@ unsigned int Sleep(unsigned int seconds)
    return retval;
 }
 
+#if 0
 /* obsolete by POSIX.1-2001 */
 void Usleep(unsigned long usec) {
    Debug1("usleep(%lu)", usec);
@@ -1336,6 +1337,7 @@ void Usleep(unsigned long usec) {
    Debug("usleep() ->");
    return;
 }
+#endif
 
 #if HAVE_NANOSLEEP
 unsigned int Nanosleep(const struct timespec *req, struct timespec *rem) {
