--- a/src/xupnpd.lua
+++ b/src/xupnpd.lua
@@ -5,10 +5,10 @@ cfg.ui_auth_file='auth.txt'
 cfg.ui_session_file='/tmp/xupnpd.session'
 
 -- multicast interface for SSDP exchange, 'eth0', 'br0', 'br-lan' for example
-cfg.ssdp_interface='lo'
+cfg.ssdp_interface='br-lan'
 
 -- 'cfg.ssdp_loop' enables multicast loop (if player and server in one host)
-cfg.ssdp_loop=1
+cfg.ssdp_loop=0
 
 -- SSDP announcement interval
 cfg.ssdp_notify_interval=15
