--- a/nginx-dav-ext-module/config	2018-12-17 08:45:12.000000000 +0000
+++ b/nginx-dav-ext-module/config	2019-10-11 13:49:41.446919258 +0000
@@ -8,9 +8,8 @@ ngx_module_name=ngx_http_dav_ext_module
 # building nginx with the xslt module, in which case libxslt will
 # be linked anyway.  In other cases libxslt is just redundant.
 # If that's a big deal, libxml2 can be linked directly:
-# ngx_module_libs=-lxml2
+ngx_module_libs=-lxml2
+ngx_module_incs=$(pkg-config --cflags-only-I libxml-2.0 | sed 's/^-I//')
 
-ngx_module_libs=LIBXSLT
-
 ngx_module_srcs="$ngx_addon_dir/ngx_http_dav_ext_module.c"
 
