From ba648449b01373768ecbf16b1610fcc071c4099c Mon Sep 17 00:00:00 2001
From: Samuel Holland <samuel@sholland.org>
Date: Tue, 31 Oct 2023 02:30:52 -0500
Subject: [PATCH 56/68] fixup! sunxi: Globally enable SUPPORT_SPL

---
 arch/arm/mach-sunxi/Kconfig | 1 -
 1 file changed, 1 deletion(-)

--- a/arch/arm/mach-sunxi/Kconfig
+++ b/arch/arm/mach-sunxi/Kconfig
@@ -354,7 +354,6 @@ config MACH_SUN8I_R528
 	select SUNXI_GEN_NCAT2
 	select SUNXI_NEW_PINCTRL
 	select MMC_SUNXI_HAS_NEW_MODE
-	select SUPPORT_SPL
 	select DRAM_SUN20I_D1
 
 config MACH_SUN8I_V3S
