From 9e824a4cf12362c38a995cf4751d15b2e281a5e1 Mon Sep 17 00:00:00 2001
From: Samuel Holland <samuel@sholland.org>
Date: Tue, 31 Oct 2023 03:40:19 -0500
Subject: [PATCH 63/68] regulator: Don't fail when already enabled

Signed-off-by: Samuel Holland <samuel@sholland.org>
---
 drivers/power/regulator/regulator_common.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/drivers/power/regulator/regulator_common.c
+++ b/drivers/power/regulator/regulator_common.c
@@ -77,7 +77,7 @@ int regulator_common_set_enable(const st
 	/* If previously enabled, increase count */
 	if (enable && plat->enable_count > 0) {
 		plat->enable_count++;
-		return -EALREADY;
+		return 0;
 	}
 
 	if (!enable) {
@@ -87,7 +87,7 @@ int regulator_common_set_enable(const st
 			return -EBUSY;
 		} else if (!plat->enable_count) {
 			/* If already disabled, do nothing */
-			return -EALREADY;
+			return 0;
 		}
 	}
 
