From 62419798c137c9eec2d2c1011e417d1520aecffb Mon Sep 17 00:00:00 2001
From: Zoltan HERPAI <wigyori@uid0.hu>
Date: Tue, 6 Jun 2023 19:46:30 +0000
Subject: [PATCH 90/90] drivers: phy: fix typo

Signed-off-by: Zoltan HERPAI <wigyori@uid0.hu>
---
 drivers/phy/allwinner/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/phy/allwinner/Kconfig
+++ b/drivers/phy/allwinner/Kconfig
@@ -3,7 +3,7 @@
 #
 config PHY_SUN4I_USB
 	bool "Allwinner Sun4I USB PHY driver"
-	depends on depends on BOARD_SUNXI
+	depends on BOARD_SUNXI
 	default y
 	select DM_REGULATOR
 	select PHY
