--- a/src/formats.c
+++ b/src/formats.c
@@ -413,7 +413,7 @@ static void UNUSED rewind_pipe(FILE * fp
 #if defined _FSTDIO || defined _NEWLIB_VERSION || defined __APPLE__
   fp->_p -= PIPE_AUTO_DETECT_SIZE;
   fp->_r += PIPE_AUTO_DETECT_SIZE;
-#elif defined __GLIBC__
+#elif defined __GLIBC__ && ! defined __UCLIBC__
   fp->_IO_read_ptr = fp->_IO_read_base;
 #elif defined _MSC_VER || defined _WIN32 || defined _WIN64 || \
       defined _ISO_STDIO_ISO_H || defined __sgi
@@ -422,7 +422,6 @@ static void UNUSED rewind_pipe(FILE * fp
   /* To fix this #error, either simply remove the #error line and live without
    * file-type detection with pipes, or add support for your compiler in the
    * lines above.  Test with cat monkey.wav | ./sox --info - */
-  #error FIX NEEDED HERE
   #define NO_REWIND_PIPE
   (void)fp;
 #endif
