--- a/Makefile.am
+++ b/Makefile.am
@@ -3,7 +3,7 @@
 AUTOMAKE_OPTIONS = foreign dist-zip
 ACLOCAL_AMFLAGS = -I m4
 
-SUBDIRS = src conf doc web admin win32 examples
+SUBDIRS = src conf web admin
 
 EXTRA_DIST = HACKING m4/acx_pthread.m4 m4/ogg.m4 \
     m4/theora.m4 m4/vorbis.m4 m4/speex.m4 \
@@ -11,7 +11,7 @@ EXTRA_DIST = HACKING m4/acx_pthread.m4 m
     m4/xiph_types.m4 m4/xiph_xml2.m4
 
 docdir = $(datadir)/doc/$(PACKAGE)
-doc_DATA = README AUTHORS COPYING NEWS TODO ChangeLog
+doc_DATA = README AUTHORS COPYING
 
 debug:
 	$(MAKE) all CFLAGS="@DEBUG@"
