--- /dev/null
+++ b/Makefile.am
@@ -0,0 +1,2 @@
+ACLOCAL_AMFLAGS = -I m4
+SUBDIRS = .
--- /dev/null
+++ b/libcharset/Makefile.am
@@ -0,0 +1 @@
+ACLOCAL_AMFLAGS = -I m4
--- a/configure.ac
+++ b/configure.ac
@@ -54,7 +54,7 @@ AC_ISC_POSIX
 
 dnl           checks for installer options
 
-AC_RELOCATABLE
+gl_RELOCATABLE
 
 AC_ARG_ENABLE([extra-encodings],
   [AC_HELP_STRING([--enable-extra-encodings],
--- a/libcharset/configure.ac
+++ b/libcharset/configure.ac
@@ -41,7 +41,7 @@ AC_CANONICAL_HOST
 
 dnl           checks for installer options
 
-AC_RELOCATABLE_LIBRARY
+gl_RELOCATABLE_LIBRARY
 
 dnl           checks for programs
 
