--- a/m4/canonicalize.m4
+++ b/m4/canonicalize.m4
@@ -11,8 +11,6 @@ AC_DEFUN([gl_CANONICALIZE],
   AC_CHECK_FUNCS(canonicalize_file_name)
   if test $ac_cv_func_canonicalize_file_name = no; then
     AC_LIBOBJ(canonicalize)
-    AC_DEFINE([realpath], [rpl_realpath],
-      [Define to a replacement function name for realpath().])
     gl_PREREQ_CANONICALIZE
   fi
 ])
--- a/configure
+++ b/configure
@@ -35532,9 +35532,6 @@ done
   if test $ac_cv_func_canonicalize_file_name = no; then
     SRCLIBOBJS="$SRCLIBOBJS canonicalize.$ac_objext"
 
-cat >>confdefs.h <<\_ACEOF
-#define realpath rpl_realpath
-_ACEOF
 
 
 
