diff --git a/cmd/version.c b/cmd/version.c
index b2fffe9..bcbbeb1 100644
--- a/cmd/version.c
+++ b/cmd/version.c
@@ -18,7 +18,7 @@ static int do_version(cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[])
 {
	char buf[DISPLAY_OPTIONS_BANNER_LENGTH];

-	printf(display_options_get_banner(false, buf, sizeof(buf)));
+	printf("%s",display_options_get_banner(false, buf, sizeof(buf)));
 #ifdef CC_VERSION_STRING
	puts(CC_VERSION_STRING "\n");
 #endif
