From 8f9b9d829ea0e67760b2e67c9339f6c417084fdc Mon Sep 17 00:00:00 2001
From: Pan Jiafei <Jiafei.Pan@nxp.com>
Date: Thu, 28 Jan 2016 12:10:24 +0800
Subject: [PATCH 05/70] armv8: aarch32: change FS file name in ITS

Signed-off-by: Pan Jiafei <Jiafei.Pan@nxp.com>
---
 kernel-ls1043a-rdb-aarch32.its |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/kernel-ls1043a-rdb-aarch32.its
+++ b/kernel-ls1043a-rdb-aarch32.its
@@ -33,7 +33,7 @@
 		};
 		ramdisk@1 {
 			description = "LS1 Ramdisk";
-                        data = /incbin/("./fsl-image-core-ls1021atwr-wifi.rootfs.ext2.gz");
+                        data = /incbin/("./fsl-image-core-ls1043ardb-32b.ext2.gz");
 			type = "ramdisk";
 			arch = "arm";
 			os = "linux";
