From 6e4322d938f01f58156b0a5929daa648ecad4754 Mon Sep 17 00:00:00 2001
From: Madalin Bucur <madalin.bucur@freescale.com>
Date: Thu, 14 Apr 2016 12:03:52 +0300
Subject: [PATCH 67/70] net: phy: add SGMII 2500 PHY

Signed-off-by: Madalin Bucur <madalin.bucur@freescale.com>
---
 include/linux/phy.h |    1 +
 1 file changed, 1 insertion(+)

--- a/include/linux/phy.h
+++ b/include/linux/phy.h
@@ -77,6 +77,7 @@ typedef enum {
 	PHY_INTERFACE_MODE_XGMII,
 	PHY_INTERFACE_MODE_MOCA,
 	PHY_INTERFACE_MODE_QSGMII,
+	PHY_INTERFACE_MODE_SGMII_2500,
 	PHY_INTERFACE_MODE_MAX,
 } phy_interface_t;
 
