From 53ff2bc222c446dfb6ffa9b708a23e7b8b82abb3 Mon Sep 17 00:00:00 2001
From: Pan Jiafei <Jiafei.Pan@nxp.com>
Date: Fri, 3 Jun 2016 06:45:57 +0000
Subject: [PATCH 62/70] armv8: aarch32: enable pci_domains for armv8-32bit

Signed-off-by: Pan Jiafei <Jiafei.Pan@nxp.com>
---
 arch/arm/mach-imx/Kconfig |    1 +
 1 file changed, 1 insertion(+)

--- a/arch/arm/mach-imx/Kconfig
+++ b/arch/arm/mach-imx/Kconfig
@@ -617,6 +617,7 @@ config ARCH_LAYERSCAPE
 	select ARM_GIC
 	select HAVE_ARM_ARCH_TIMER
 	select PCI_LAYERSCAPE if PCI
+	select PCI_DOMAINS if PCI
 	select LS1_MSI if PCI_MSI
 
 	help
