--- a/Makefile.am
+++ b/Makefile.am
@@ -1,6 +1,6 @@
 ## Process this file with automake to produce Makefile.in
 
-SUBDIRS = lib bc dc doc
+SUBDIRS = lib bc dc
 
 MAINTAINERCLEANFILES =  aclocal.m4 config.h.in configure Makefile.in \
 			stamp-h $(distdir).tar.gz h/number.h depcomp missing \
--- a/Makefile.in
+++ b/Makefile.in
@@ -288,7 +288,7 @@ target_alias = @target_alias@
 top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
-SUBDIRS = lib bc dc doc
+SUBDIRS = lib bc dc
 MAINTAINERCLEANFILES = aclocal.m4 config.h.in configure Makefile.in \
 			stamp-h $(distdir).tar.gz h/number.h depcomp missing \
 			bc/libmath.h
