--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -64,6 +64,32 @@ static struct board_info __initdata boar
 		.use_fullspeed		= 0,
 		.port_no		= 0,
 	},
+
+	.has_enetsw = 1,
+	.enetsw = {
+		.used_ports = {
+			[0] = {
+				.used = 1,
+				.phy_id = 1,
+				.name = "Port 1",
+			},
+			[1] = {
+				.used = 1,
+				.phy_id = 2,
+				.name = "Port 2",
+			},
+			[2] = {
+				.used = 1,
+				.phy_id = 3,
+				.name = "Port 3",
+			},
+			[3] = {
+				.used = 1,
+				.phy_id = 4,
+				.name = "Port 4",
+			},
+		},
+	},
 };
 #endif /* CONFIG_BCM63XX_CPU_6328 */
 
