--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -226,6 +226,33 @@ static struct board_info __initdata boar
 };
 
 
+static struct board_info __initdata board_ct536_ct5621 = {
+	.name				= "CT536_CT5621",
+	.expected_cpu_id		= 0x6348,
+
+	.has_enet0			= 0,
+	.has_enet1			= 1,
+	.has_pci			= 1,
+	.use_fallback_sprom		= 1,
+
+	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
+		.force_speed_100	= 1,
+		.force_duplex_full	= 1,
+	},
+
+	.has_ohci0 = 1,
+	.has_pccard = 1,
+	.has_ehci0 = 1,
+
+	.fallback_sprom = {
+		.type 				= SPROM_BCM4318,
+		.pci_bus			= 0,
+		.pci_dev			= 1,
+	},
+};
+
 static struct board_info __initdata board_96348gw = {
 	.name				= "96348GW",
 	.expected_cpu_id		= 0x6348,
@@ -747,6 +774,7 @@ static const struct board_info __initcon
 	&board_96348sv,
 	&board_V2500V_BB,
 	&board_V2110,
+	&board_ct536_ct5621,
 #endif
 
 #ifdef CONFIG_BCM63XX_CPU_6358
@@ -788,6 +816,8 @@ static struct of_device_id const bcm963x
 	{ .compatible = "brcm,bcm96348gw-a", .data = &board_96348gw_a, },
 	{ .compatible = "bt,voyager-2110", .data = &board_V2110, },
 	{ .compatible = "bt,voyager-2500v-bb", .data = &board_V2500V_BB, },
+	{ .compatible = "comtrend,ct-536plus", .data = &board_ct536_ct5621, },
+	{ .compatible = "comtrend,ct-5621", .data = &board_ct536_ct5621, },
 	{ .compatible = "d-link,dsl-2640b-b", .data = &board_96348_D4PW, },
 	{ .compatible = "davolink,dv-201amr", .data = &board_DV201AMR, },
 	{ .compatible = "dynalink,rta1025w", .data = &board_rta1025w_16, },
