From: "mexit@o2.pl" <mexit@o2.pl>
Date: Sun, 24 Nov 2013 21:33:38 +0000
Subject: [PATCH 4/5] bcm63xx: add support for Asmax AR 1004g router

Support for Asmax AR 1004g router

Signed-off-by: Adrian Feliks <mexit@o2.pl>
---
--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -647,6 +647,21 @@ static struct board_info __initdata boar
 	.has_ehci0 = 1,
 };
 
+static struct board_info __initdata board_96348gw_10_AR1004G = {
+	.name				= "AR1004G",
+	.expected_cpu_id		= 0x6348,
+
+	.has_enet1			= 1,
+	.has_pci			= 1,
+	
+	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
+		.force_speed_100	= 1,
+		.force_duplex_full	= 1,
+	},
+};
+
 
 /* BT Voyager 2110 */
 static struct board_info __initdata board_V2110 = {
@@ -1649,6 +1664,7 @@ static const struct board_info __initcon
 	&board_96348A_122,
 	&board_CPVA502plus,
 	&board_96348W3,
+	&board_96348gw_10_AR1004G,
 #endif
 
 #ifdef CONFIG_BCM63XX_CPU_6358
@@ -1713,6 +1729,7 @@ static struct of_device_id const bcm963x
 	{ .compatible = "dynalink,rta770w", .data = &board_rta770w, },
 #endif
 #ifdef CONFIG_BCM63XX_CPU_6348
+	{ .compatible = "asmax,ar-1004g", .data = &board_96348gw_10_AR1004G, },
 	{ .compatible = "belkin,f5d7633", .data = &board_96348gw_10, },
 	{ .compatible = "brcm,bcm96348r", .data = &board_96348r, },
 	{ .compatible = "brcm,bcm96348gw-10", .data = &board_96348gw_10, },
