--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -1710,6 +1710,30 @@ static struct board_info __initdata boar
 	.has_ehci0 = 1,
 };
 
+static struct board_info __initdata board_DGND3700v1_3800B = {
+	.name				= "DGND3700v1_3800B",
+	.expected_cpu_id		= 0x6368,
+
+	.has_pci			= 1,
+	.has_ohci0			= 1,
+	.has_ehci0			= 1,
+	.num_usbh_ports			= 2,
+
+	.has_enetsw			= 1,
+	.enetsw = {
+		.used_ports = {
+			[5] = {
+				.used  = 1,
+				.phy_id  = 0xff,
+				.bypass_link = 1,
+				.force_speed = 1000,
+				.force_duplex_full = 1,
+				.name  = "RGMII",
+			},
+		},
+	},
+};
+
 static struct sprom_fixup __initdata vr3025u_fixups[] = {
 	{ .offset = 97, .value = 0xfeb3 },
 	{ .offset = 98, .value = 0x1618 },
@@ -2141,6 +2165,7 @@ static const struct board_info __initcon
 #ifdef CONFIG_BCM63XX_CPU_6368
 	&board_96368mvwg,
 	&board_96368mvngr,
+	&board_DGND3700v1_3800B,
 	&board_P870HW51A_V2,
 	&board_VR3025u,
 	&board_VR3025un,
@@ -2243,6 +2268,7 @@ static struct of_device_id const bcm963x
 	{ .compatible = "comtrend,vr-3025u", .data = &board_VR3025u, },
 	{ .compatible = "comtrend,vr-3025un", .data = &board_VR3025un, },
 	{ .compatible = "comtrend,wap-5813n", .data = &board_WAP5813n, },
+	{ .compatible = "netgear,dgnd3700-v1", .data = &board_DGND3700v1_3800B, },
 	{ .compatible = "zyxel,p870hw-51a-v2", .data = &board_P870HW51A_V2, },
 #endif
 #ifdef CONFIG_BCM63XX_CPU_63268
