--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -12,6 +12,7 @@
 #include <linux/init.h>
 #include <linux/kernel.h>
 #include <linux/string.h>
+#include <linux/pci_ids.h>
 #include <asm/addrspace.h>
 #include <bcm63xx_board.h>
 #include <bcm63xx_cpu.h>
@@ -1603,6 +1604,94 @@ static struct board_info __initdata boar
 	},
 };
 
+static struct board_info __initdata board_HW556_A = {
+	.name = "HW556_A",
+	.expected_cpu_id = 0x6358,
+
+	.has_pci = 1,
+	.has_ohci0 = 1,
+	.has_ehci0 = 1,
+	.num_usbh_ports = 2,
+
+	.has_caldata = 1,
+	.caldata = {
+		{
+			.vendor = PCI_VENDOR_ID_ATHEROS,
+			.caldata_offset = 0xf7e000,
+			.slot = 1,
+			.endian_check = 1,
+			.led_pin = 2,
+			.led_active_high = 1,
+		},
+	},
+
+	.has_enet1 = 1,
+	.enet1 = {
+		.has_phy = 1,
+		.phy_id = 0,
+		.force_speed_100 = 1,
+		.force_duplex_full = 1,
+	},
+};
+
+static struct board_info __initdata board_HW556_B = {
+	.name = "HW556_B",
+	.expected_cpu_id = 0x6358,
+
+	.has_pci = 1,
+	.has_ohci0 = 1,
+	.has_ehci0 = 1,
+	.num_usbh_ports = 2,
+
+	.has_caldata = 1,
+	.caldata = {
+		{
+			.vendor = PCI_VENDOR_ID_ATHEROS,
+			.caldata_offset = 0xefe000,
+			.slot = 1,
+			.endian_check = 1,
+			.led_pin = 2,
+			.led_active_high = 1,
+		},
+	},
+
+	.has_enet1 = 1,
+	.enet1 = {
+		.has_phy = 1,
+		.phy_id = 0,
+		.force_speed_100 = 1,
+		.force_duplex_full = 1,
+	},
+};
+
+static struct board_info __initdata board_HW556_C = {
+	.name = "HW556_C",
+	.expected_cpu_id = 0x6358,
+
+	.has_pci = 1,
+	.has_ohci0 = 1,
+	.has_ehci0 = 1,
+	.num_usbh_ports = 2,
+
+	.has_caldata = 1,
+	.caldata = {
+		{
+			.vendor = PCI_VENDOR_ID_RALINK,
+			.caldata_offset = 0xeffe00,
+			.slot = 1,
+			.eeprom = "rt2x00.eeprom",
+		},
+	},
+
+	.has_enet1 = 1,
+	.enet1 = {
+		.has_phy = 1,
+		.phy_id = 0,
+		.force_speed_100 = 1,
+		.force_duplex_full = 1,
+	},
+};
+
  /* T-Home Speedport W 303V Typ B */
 static struct board_info __initdata board_spw303v = {
 	.name = "96358-502V",
@@ -1714,6 +1803,9 @@ static const struct board_info __initcon
 	&board_nb4_fxc_r1,
 	&board_ct6373_1,
 	&board_HW553,
+	&board_HW556_A,
+	&board_HW556_B,
+	&board_HW556_C,
 	&board_spw303v,
 	&board_DVAG3810BN,
 #endif /* CONFIG_BCM63XX_CPU_6358 */
@@ -1792,6 +1884,9 @@ static struct of_device_id const bcm963x
 	{ .compatible = "d-link,dsl-2650u", .data = &board_96358vw2, },
 	{ .compatible = "d-link,dva-g3810bn-tl", .data = &board_DVAG3810BN, },
 	{ .compatible = "huawei,echolife-hg553", .data = &board_HW553, },
+	{ .compatible = "huawei,echolife-hg556a-a", .data = &board_HW556_A, },
+	{ .compatible = "huawei,echolife-hg556a-b", .data = &board_HW556_B, },
+	{ .compatible = "huawei,echolife-hg556a-c", .data = &board_HW556_C, },
 	{ .compatible = "pirelli,a226g", .data = &board_DWVS0, },
 	{ .compatible = "pirelli,a226m", .data = &board_DWVS0, },
 	{ .compatible = "pirelli,a226m-fwb", .data = &board_DWVS0, },
