--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -1579,6 +1579,31 @@ static struct board_info __initdata boar
 	},
 };
 
+static struct board_info __initdata board_homehub2a = {
+	.name = "HOMEHUB2A",
+	.expected_cpu_id = 0x6358,
+
+	.has_pci = 1,
+	.has_ohci0 = 1,
+	.has_ehci0 = 1,
+	.num_usbh_ports = 2,
+
+	.has_enet1 = 1,
+	.enet1 = {
+		.has_phy = 1,
+		.phy_id = 0,
+		.force_speed_100 = 1,
+		.force_duplex_full = 1,
+	},
+
+	.use_fallback_sprom = 1,
+	.fallback_sprom = {
+		.type = SPROM_BCM4322,
+		.pci_bus = 0,
+		.pci_dev = 1,
+	},
+};
+
 static struct board_info __initdata board_HW520 = {
 	.name = "HW6358GW_B",
 	.expected_cpu_id = 0x6358,
@@ -1832,6 +1857,7 @@ static const struct board_info __initcon
 	&board_nb4_ser_r0,
 	&board_nb4_fxc_r1,
 	&board_ct6373_1,
+	&board_homehub2a,
 	&board_HW520,
 	&board_HW553,
 	&board_HW556_A,
@@ -1927,6 +1953,7 @@ static struct of_device_id const bcm963x
 	{ .compatible = "sfr,neufbox-4-foxconn-r1", .data = &board_nb4_fxc_r1, },
 	{ .compatible = "t-com,speedport-w-303v", .data = &board_spw303v, },
 	{ .compatible = "telsey,cpva642", .data = &board_CPVA642, },
+	{ .compatible = "bt,home-hub-2-a", .data = &board_homehub2a, },
 #endif /* CONFIG_BCM63XX_CPU_6358 */
 #ifdef CONFIG_BCM63XX_CPU_6362
 #endif /* CONFIG_BCM63XX_CPU_6362 */
