--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -1865,6 +1865,37 @@ static struct board_info __initdata boar
 		},
 	},
 };
+
+static struct board_info __initdata board_hg253s_v2 = {
+	.name = "hg253s",
+	.expected_cpu_id = 0x6362,
+
+	.has_ohci0 = 1,
+	.has_ehci0 = 1,
+	.num_usbh_ports = 1,
+
+	.has_enetsw = 1,
+	.enetsw = {
+		.used_ports = {
+			[4] = {
+				.used = 1,
+				.phy_id = 0xff,
+				.bypass_link = 1,
+				.force_speed = 1000,
+				.force_duplex_full = 1,
+				.timing_sel = 1,
+				.name = "RGMII",
+			},
+
+			[5] = {
+				.used = 1,
+				.phy_id = 24,
+				.timing_sel = 1,
+				.name = "WAN",
+			},
+		},
+	},
+};
 #endif /* CONFIG_BCM63XX_CPU_6362 */
 
 /*
@@ -1950,6 +1981,7 @@ static const struct board_info __initcon
 	&board_nb6,
 	&board_dgnd3700v2,
 	&board_fast2504n,
+	&board_hg253s_v2,
 #endif /* CONFIG_BCM63XX_CPU_6362 */
 };
 
@@ -2041,6 +2073,7 @@ static struct of_device_id const bcm963x
 	{ .compatible = "bt,home-hub-2-a", .data = &board_homehub2a, },
 #endif /* CONFIG_BCM63XX_CPU_6358 */
 #ifdef CONFIG_BCM63XX_CPU_6362
+	{ .compatible = "huawei,hg253s-v2", .data = &board_hg253s_v2, },
 	{ .compatible = "netgear,dgnd3700-v2", .data = &board_dgnd3700v2, },
 	{ .compatible = "sagem,fast-2504n", .data = &board_fast2504n, },
 	{ .compatible = "sfr,neufbox-6-sercomm-r0", .data = &board_nb6, },
