--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -2580,6 +2580,48 @@ static struct board_info __initdata boar
 		},
 	},
 };
+
+static struct board_info __initdata board_vw6339gu = {
+	.name = "VW6339GU",
+	.expected_cpu_id = 0x63268,
+
+	.has_ohci0 = 1,
+	.has_ehci0 = 1,
+	.num_usbh_ports = 1,
+
+	.has_enetsw = 1,
+	.enetsw = {
+		.used_ports = {
+			[0] = {
+				.used = 1,
+				.phy_id = 1,
+				.name = "LAN2",
+			},
+			[1] = {
+				.used = 1,
+				.phy_id = 2,
+				.name = "LAN3",
+			},
+			[2] = {
+				.used = 1,
+				.phy_id = 3,
+				.name = "LAN4",
+			},
+			[3] = {
+				.used = 1,
+				.phy_id = 4,
+				.name = "LAN1",
+			},
+			[4] = {
+				.used = 1,
+				.phy_id = 7,
+				.name = "WAN",
+				.mii_override = 1,
+				.timing_sel = 1,
+			},
+		},
+	},
+};
 #endif /* CONFIG_BCM63XX_CPU_63268 */
 
 /*
@@ -2685,6 +2727,7 @@ static const struct board_info __initcon
 #endif /* CONFIG_BCM63XX_CPU_6368 */
 #ifdef CONFIG_BCM63XX_CPU_63268
 	&board_963269bhr,
+	&board_vw6339gu,
 #endif /* CONFIG_BCM63XX_CPU_63268 */
 };
 
@@ -2799,6 +2842,7 @@ static struct of_device_id const bcm963x
 #endif /* CONFIG_BCM63XX_CPU_6368 */
 #ifdef CONFIG_BCM63XX_CPU_63268
 	{ .compatible = "brcm,bcm963269bhr", .data = &board_963269bhr, },
+	{ .compatible = "inteno,vg50", .data = &board_vw6339gu, },
 #endif /* CONFIG_BCM63XX_CPU_63268 */
 #endif /* CONFIG_OF */
 	{ },
