--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -2777,6 +2777,36 @@ static struct board_info __initdata boar
 		},
 	},
 };
+
+static struct board_info __initdata board_H500s = {
+	.name = "BXK00C-1.6",
+	.expected_cpu_id = 0x63268,
+
+	.has_ohci0 = 1,
+	.has_ehci0 = 1,
+	.num_usbh_ports = 1,
+
+	.has_enetsw = 1,
+	.enetsw = {
+		.used_ports = {
+			[3] = {
+				.used = 1,
+				.phy_id = 12,
+				.name = "WAN",
+			},
+			[4] = {
+				.used = 1,
+				.phy_id = 0,
+				.bypass_link = 1,
+				.force_speed = 1000,
+				.force_duplex_full = 1,
+				.mii_override = 1,
+				.timing_sel = 1,
+				.name = "RGMII",
+			},
+		},
+	},
+};
 #endif /* CONFIG_BCM63XX_CPU_63268 */
 
 /*
@@ -2887,6 +2917,7 @@ static const struct board_info __initcon
 	&board_VR3032u,
 	&board_vw6339gu,
 	&board_BSKYB_63168,
+	&board_H500s,
 #endif /* CONFIG_BCM63XX_CPU_63268 */
 };
 
@@ -3005,6 +3036,8 @@ static struct of_device_id const bcm963x
 	{ .compatible = "comtrend,vg-8050", .data = &board_VG8050, },
 	{ .compatible = "comtrend,vr-3032u", .data = &board_VR3032u, },
 	{ .compatible = "inteno,vg50", .data = &board_vw6339gu, },
+	{ .compatible = "sercomm,h500-s-lowi", .data = &board_H500s, },
+	{ .compatible = "sercomm,h500-s-vfes", .data = &board_H500s, },
 	{ .compatible = "sky,sr102", .data = &board_BSKYB_63168, },
 #endif /* CONFIG_BCM63XX_CPU_63268 */
 #endif /* CONFIG_OF */
