--- a/Makefile
+++ b/Makefile
@@ -584,7 +584,10 @@ UBOOTINCLUDE    := \
 		-I$(srctree)/arch/$(ARCH)/include \
 		-include $(srctree)/include/linux/kconfig.h
 
-NOSTDINC_FLAGS += -nostdinc -isystem $(shell $(CC) -print-file-name=include)
+ifneq ($(shell uname),FreeBSD)
+  NOSTDINC_FLAGS += -nostdinc -isystem $(shell $(CC) -print-file-name=include)
+endif
+
 CHECKFLAGS     += $(NOSTDINC_FLAGS)
 
 # FIX ME
