--- a/configure.py
+++ b/configure.py
@@ -198,6 +198,8 @@ parser = OptionParser()
 profilers = ['gmon', 'pprof']
 parser.add_option('--bootstrap', action='store_true',
                   help='bootstrap a ninja binary from nothing')
+parser.add_option('--no-rebuild', action='store_true',
+                  help='let user execute ninja after build.ninja generation')
 parser.add_option('--verbose', action='store_true',
                   help='enable verbose build')
 parser.add_option('--platform',
@@ -756,7 +758,11 @@ n.build('all', 'phony', all_targets)
 n.close()
 print('wrote %s.' % BUILD_FILENAME)
 
-if options.bootstrap:
+if options.bootstrap and options.no_rebuild:
+    print('bootstrap complete. execute ninja in this directory...')
+    print(os.getcwd())
+
+elif options.bootstrap:
     print('bootstrap complete.  rebuilding...')
 
     rebuild_args = []
