--- a/nss_stats.c
+++ b/nss_stats.c
@@ -407,9 +407,9 @@ void nss_stats_create_dentry(char *name,
 /*
  * gmac_stats_ops
  */
-#ifdef NSS_DATA_PLANE_GENERIC_SUPPORT
-NSS_STATS_DECLARE_FILE_OPERATIONS(gmac);
-#endif
+// #ifdef NSS_DATA_PLANE_GENERIC_SUPPORT
+// NSS_STATS_DECLARE_FILE_OPERATIONS(gmac);
+// #endif
 
 /*
  * wt_stats_ops
@@ -464,9 +464,9 @@ void nss_stats_init(void)
 	/*
 	 * gmac_stats
 	 */
-#ifdef NSS_DATA_PLANE_GENERIC_SUPPORT
-	nss_stats_create_dentry("gmac", &nss_gmac_stats_ops);
-#endif
+// #ifdef NSS_DATA_PLANE_GENERIC_SUPPORT
+// 	nss_stats_create_dentry("gmac", &nss_gmac_stats_ops);
+// #endif
 
 	/*
 	 * Per-project stats
--- a/Makefile
+++ b/Makefile
@@ -26,7 +26,6 @@ qca-nss-drv-objs := \
 			nss_n2h_stats.o \
 			nss_n2h_strings.o \
 			nss_pm.o \
-			nss_profiler.o \
 			nss_project.o \
 			nss_rps.o \
 			nss_stats.o \
@@ -40,7 +39,6 @@ qca-nss-drv-objs += nss_hal/nss_hal.o
 
 ifneq "$(NSS_DRV_POINT_OFFLOAD)" "y"
 qca-nss-drv-objs += \
-			nss_gmac_stats.o \
 			nss_if.o \
 			nss_if_log.o \
 			nss_phys_if.o \
@@ -381,7 +379,9 @@ endif
 
 ifeq ($(SoC),$(filter $(SoC),ipq806x))
 qca-nss-drv-objs += nss_data_plane/nss_data_plane_gmac.o \
-		    nss_hal/ipq806x/nss_hal_pvt.o
+		    nss_hal/ipq806x/nss_hal_pvt.o \
+		    nss_gmac_stats.o \
+		    nss_profiler.o
 
 ifneq "$(NSS_DRV_C2C_ENABLE)" "n"
 ccflags-y += -DNSS_DRV_C2C_ENABLE
