--- a/Makefile
+++ b/Makefile
@@ -350,10 +350,17 @@ endif
 
 ifneq "$(NSS_DRV_WIFIOFFLOAD_ENABLE)" "n"
 ccflags-y += -DNSS_DRV_WIFIOFFLOAD_ENABLE
+ifneq "$(NSS_DRV_WIFI_LEGACY_ENABLE)" "n"
+ccflags-y += -DNSS_DRV_WIFI_LEGACY_ENABLE
 qca-nss-drv-objs += \
 			nss_wifi.o \
 			nss_wifi_log.o \
-			nss_wifi_stats.o \
+			nss_wifi_stats.o
+ifeq ($(SoC),$(filter $(SoC),ipq806x))
+ccflags-y += -DNSS_HAL_IPQ806x_SUPPORT
+endif
+endif
+qca-nss-drv-objs += \
 			nss_wifi_vdev.o \
 			nss_wifili.o \
 			nss_wifili_log.o \
--- a/nss_hal/nss_hal.c
+++ b/nss_hal/nss_hal.c
@@ -587,7 +587,9 @@ int nss_hal_probe(struct platform_device
 	if (npd->wifioffload_enabled == NSS_FEATURE_ENABLED) {
 		nss_top->wifi_handler_id = nss_dev->id;
 		nss_top->dynamic_interface_table[NSS_DYNAMIC_INTERFACE_TYPE_VAP] = nss_dev->id;
+#if defined(NSS_HAL_IPQ806x_SUPPORT)
 		nss_wifi_register_handler();
+#endif
 		nss_wifili_register_handler();
 #ifdef NSS_DRV_WIFI_EXT_VDEV_ENABLE
 		nss_wifi_ext_vdev_register_handler();
