--- a/nss_stats.c
+++ b/nss_stats.c
@@ -395,8 +395,9 @@ size_t nss_stats_print(char *node, char
  */
 void nss_stats_create_dentry(char *name, const struct file_operations *ops)
 {
-	if (!debugfs_create_file(name, 0400, nss_top_main.stats_dentry, &nss_top_main, ops)) {
-		nss_warning("Failed to create debug entry for subsystem %s\n", name);
+	if (!debugfs_lookup(name, nss_top_main.stats_dentry))
+	  if (!debugfs_create_file(name, 0400, nss_top_main.stats_dentry, &nss_top_main, ops)) {
+	    nss_warning("Failed to create debug entry for subsystem %s\n", name);
 	}
 }
 
