--- a/exports/nss_wifili_if.h
+++ b/exports/nss_wifili_if.h
@@ -751,6 +751,15 @@ struct nss_wifili_stats_notification {
 	struct nss_wifili_stats stats;	/**< Wifili statistics. */
 };
 
+/**
+ * nss_wifili_get_context
+ *	Gets the Wi-Fi Li context used in NSS GRE transmit.
+ *
+ * @return
+ * Pointer to the NSS core context.
+ */
+extern struct nss_ctx_instance *nss_wifili_get_context(void);
+
 #ifdef __KERNEL__ /* only kernel will use. */
 
 /**
