--- a/nss_wifi_mesh_strings.c
+++ b/nss_wifi_mesh_strings.c
@@ -80,8 +80,8 @@ struct nss_stats_info nss_wifi_mesh_stri
 	{"not_found",			NSS_STATS_TYPE_SPECIAL},
 	{"delete_success",		NSS_STATS_TYPE_SPECIAL},
 	{"update_success",		NSS_STATS_TYPE_SPECIAL},
-	{"mesh_path_expired",		NSS_STATS_TYPE_SPECIAL},
-	{"mesh_path_refresh_needed",	NSS_STATS_TYPE_SPECIAL},
+	{"path_expired",		NSS_STATS_TYPE_SPECIAL},
+	{"path_refresh_needed",	NSS_STATS_TYPE_SPECIAL},
 	{"add_requests",		NSS_STATS_TYPE_SPECIAL},
 	{"del_requests",		NSS_STATS_TYPE_SPECIAL},
 	{"update_requests",		NSS_STATS_TYPE_SPECIAL},
@@ -91,8 +91,8 @@ struct nss_stats_info nss_wifi_mesh_stri
 	{"metric_updations",		NSS_STATS_TYPE_SPECIAL},
 	{"block_mesh_fwd_updations",	NSS_STATS_TYPE_SPECIAL},
 	{"delete_failures",		NSS_STATS_TYPE_SPECIAL},
-	{"mesh_path_dummy_add_failures",NSS_STATS_TYPE_SPECIAL},
-	{"mesh_path_dummy_add_success",	NSS_STATS_TYPE_SPECIAL}
+	{"path_dummy_add_failures",	NSS_STATS_TYPE_SPECIAL},
+	{"path_dummy_add_success",	NSS_STATS_TYPE_SPECIAL}
 
 };
 
@@ -126,10 +126,10 @@ struct nss_stats_info nss_wifi_mesh_stri
 	{"update_requests",				NSS_STATS_TYPE_SPECIAL},
 	{"mda_updations",				NSS_STATS_TYPE_SPECIAL},
 	{"flag_updations",				NSS_STATS_TYPE_SPECIAL},
-	{"mesh_proxy_path_dummy_lookup_success",	NSS_STATS_TYPE_SPECIAL},
-	{"mesh_proxy_path_dummy_lookup_failures",	NSS_STATS_TYPE_SPECIAL},
-	{"mesh_proxy_path_dummy_add_failures",		NSS_STATS_TYPE_SPECIAL},
-	{"mesh_proxy_path_dummy_add_success",		NSS_STATS_TYPE_SPECIAL}
+	{"proxy_path_dummy_lookup_success",	NSS_STATS_TYPE_SPECIAL},
+	{"proxy_path_dummy_lookup_failures",	NSS_STATS_TYPE_SPECIAL},
+	{"proxy_path_dummy_add_failures",		NSS_STATS_TYPE_SPECIAL},
+	{"proxy_path_dummy_add_success",		NSS_STATS_TYPE_SPECIAL}
 };
 
 /*
