--- a/nss_core.c
+++ b/nss_core.c
@@ -2700,7 +2700,7 @@ static inline bool nss_core_skb_can_reus
 
 	if (unlikely(irqs_disabled()))
 		return false;
-#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 6, 0))
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 14, 0))
 	if (unlikely(skb_shinfo(nbuf)->tx_flags & SKBTX_DEV_ZEROCOPY))
 		return false;
 #else
--- a/nss_coredump.c
+++ b/nss_coredump.c
@@ -26,7 +26,7 @@
 #include "nss_log.h"
 #include <linux/kernel.h>
 
-#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 6, 0))
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 14, 0))
 #include <linux/notifier.h>	/* for panic_notifier_list */
 #else
 #include <linux/panic_notifier.h>	/* for panic_notifier_list */
--- a/nss_core.h
+++ b/nss_core.h
@@ -125,7 +125,7 @@
 #define h2n_hlos_index_to_dma(_if_map_addr, _index) (_if_map_addr) + h2n_hlos_index_offset + (sizeof(uint32_t) * (_index))
 #define n2h_hlos_index_to_dma(_if_map_addr, _index) (_if_map_addr) + n2h_hlos_index_offset + (sizeof(uint32_t) * (_index))
 
-#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 6, 0))
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 14, 0))
 #define nss_ioremap ioremap_nocache
 #else
 #define nss_ioremap ioremap
