--- a/drivers/mtd/nand/raw/qcom_nandc.c
+++ b/drivers/mtd/nand/raw/qcom_nandc.c
@@ -1867,7 +1867,7 @@ static int qcom_param_page_type_exec(str
 
 	reg_base = NAND_READ_LOCATION_0;
 
-	if (nandc->props->qpic_v2)
+	if (nandc->props->qpic_version2)
 		reg_base = NAND_READ_LOCATION_LAST_CW_0;
 
 	ret = qcom_parse_instructions(chip, subop, &q_op);
@@ -1920,7 +1920,7 @@ static int qcom_param_page_type_exec(str
 	op_id = q_op.data_instr_idx;
 	len = nand_subop_get_data_len(subop, op_id);
 
-	if (nandc->props->qpic_v2)
+	if (nandc->props->qpic_version2)
 		nandc_set_read_loc_last(chip, reg_base, 0, len, 1);
 	else
 		nandc_set_read_loc_first(chip, reg_base, 0, len, 1);
