--- a/arch/arm/boot/dts/marvell/armada-385-linksys.dtsi
+++ b/arch/arm/boot/dts/marvell/armada-385-linksys.dtsi
@@ -116,7 +116,7 @@
 };
 
 &eth2 {
-	status = "okay";
+	status = "disabled";
 	phy-mode = "sgmii";
 	buffer-manager = <&bm>;
 	bm,pool-long = <2>;
@@ -200,10 +200,10 @@
 				label = "wan";
 			};
 
-			port@5 {
-				reg = <5>;
+			port@6 {
+				reg = <6>;
 				phy-mode = "sgmii";
-				ethernet = <&eth2>;
+				ethernet = <&eth0>;
 
 				fixed-link {
 					speed = <1000>;
