--- a/GNUmakefile
+++ b/GNUmakefile
@@ -357,6 +357,8 @@ socket_accept4.o socket_accept6.o socket
 socket_local6.o socket_recv4.o socket_recv6.o socket_remote4.o \
 socket_remote6.o: havesl.h
 
+socket_remote4.o: havescope.h
+
 dns_nd6.o fmt_xlong.o scan_xlong.o fmt_ip6_flat.o $(TEXTCODE_OBJS): haveinline.h
 
 iob_send.o scan_ip6if.o: havealloca.h
@@ -386,6 +388,6 @@ update:
 	dl -n http://www.w3.org/TR/html5/entities.json
 
 entities.h: entities.json ent
-	./ent
+	libowfat-ent
 
 scan_html.o: entities.h
