--- a/arch/arm64/boot/dts/qcom/ipq8074-ac-cpu.dtsi
+++ b/arch/arm64/boot/dts/qcom/ipq8074-ac-cpu.dtsi
@@ -32,16 +32,37 @@
 		compatible = "operating-points-v2";
 		opp-shared;
 
-		opp-1017600000 {
-			opp-hz = /bits/ 64 <1017600000>;
-			opp-microvolt = <1>;
-			clock-latency-ns = <200000>;
-		};
-		opp-1382400000 {
-			opp-hz = /bits/ 64 <1382400000>;
-			opp-microvolt = <2>;
-			clock-latency-ns = <200000>;
-		};
+                opp-1017600000 {
+                        opp-hz = /bits/ 64 <1017600000>;
+                        opp-microvolt = <1>;
+                        clock-latency-ns = <100000>;
+                };
+                opp-1382400000 {
+                        opp-hz = /bits/ 64 <1382400000>;
+                        opp-microvolt = <2>;
+                        clock-latency-ns = <100000>;
+                };
+                opp-1651100000 {
+                        opp-hz = /bits/ 64 <1651100000>;
+                        opp-microvolt = <3>;
+                        clock-latency-ns = <100000>;
+                };
+                opp-1843100000 {
+                        opp-hz = /bits/ 64 <1843100000>;
+                        opp-microvolt = <4>;
+                        clock-latency-ns = <100000>;
+                };
+                opp-1910000000 {
+                        opp-hz = /bits/ 64 <1910000000>;
+                        opp-microvolt = <5>;
+                        clock-latency-ns = <100000>;
+                };
+                opp-2208000000 {
+                        opp-hz = /bits/ 64 <2208000000>;
+                        opp-microvolt = <6>;
+                        clock-latency-ns = <100000>;
+                };
+
 	};
 };
 
