From 1608172019c68c5a1d567d2639843d10f19b60aa Mon Sep 17 00:00:00 2001
From: Eric Anholt <eric@anholt.net>
Date: Thu, 2 Jun 2016 12:29:45 -0700
Subject: [PATCH 129/454] BCM270X: Add the DSI panel to the defconfig.

Signed-off-by: Eric Anholt <eric@anholt.net>
---
 arch/arm64/configs/bcmrpi3_defconfig | 2 ++
 1 file changed, 2 insertions(+)

--- a/arch/arm64/configs/bcmrpi3_defconfig
+++ b/arch/arm64/configs/bcmrpi3_defconfig
@@ -813,6 +813,8 @@ CONFIG_VIDEO_OV7640=m
 CONFIG_VIDEO_MT9V011=m
 CONFIG_DRM=m
 CONFIG_DRM_LOAD_EDID_FIRMWARE=y
+CONFIG_DRM_PANEL_SIMPLE=m
+CONFIG_DRM_PANEL_RASPBERRYPI_TOUCHSCREEN=m
 CONFIG_DRM_UDL=m
 CONFIG_DRM_VC4=m
 CONFIG_FB=y
