From 47008c3c01cc8785371cb30afc3e1ab056195d3e Mon Sep 17 00:00:00 2001
From: eccgecko <36884529+eccgecko@users.noreply.github.com>
Date: Sun, 20 May 2018 17:00:39 +0200
Subject: [PATCH 305/454] config: Add LZ4 compression support to arm64 kernel
 (#2559)

---
 arch/arm64/configs/bcmrpi3_defconfig | 1 +
 1 file changed, 1 insertion(+)

--- a/arch/arm64/configs/bcmrpi3_defconfig
+++ b/arch/arm64/configs/bcmrpi3_defconfig
@@ -1270,6 +1270,7 @@ CONFIG_CRYPTO_TGR192=m
 CONFIG_CRYPTO_WP512=m
 CONFIG_CRYPTO_CAST5=m
 CONFIG_CRYPTO_DES=y
+CONFIG_CRYPTO_LZ4=m
 CONFIG_CRYPTO_USER_API_SKCIPHER=m
 CONFIG_ARM64_CRYPTO=y
 CONFIG_CRC_ITU_T=y
