Index: u-boot-2017.09/cmd/Kconfig
===================================================================
--- u-boot-2017.09.orig/cmd/Kconfig
+++ u-boot-2017.09/cmd/Kconfig
@@ -589,6 +589,14 @@ config CMD_ARMFLASH
 	help
 	  ARM Ltd reference designs flash partition access
 
+config CMD_ADC
+	bool "adc - Access Analog to Digital Converters info and data"
+	select ADC
+	depends on DM_REGULATOR
+	help
+	  Shows ADC device info and permit printing one-shot analog converted
+	  data from a named Analog to Digital Converter.
+
 config CMD_CLK
 	bool "clk - Show clock frequencies"
 	help
Index: u-boot-2017.09/cmd/Makefile
===================================================================
--- u-boot-2017.09.orig/cmd/Makefile
+++ u-boot-2017.09/cmd/Makefile
@@ -13,6 +13,7 @@ obj-y += help.o
 obj-y += version.o
 
 # command
+obj-$(CONFIG_CMD_ADC) += adc.o
 obj-$(CONFIG_CMD_AES) += aes.o
 obj-$(CONFIG_CMD_AMBAPP) += ambapp.o
 obj-$(CONFIG_CMD_ANDROID_AB_SELECT) += android_ab_select.o android_cmds.o
