From 40361189b449992e14ff0bd0be18c92167ff8d00 Mon Sep 17 00:00:00 2001
From: jjm2473 <1129525450@qq.com>
Date: Tue, 6 Jun 2023 12:27:39 +0800
Subject: [PATCH] fix efi detection

---
 disk/part_efi.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/disk/part_efi.c b/disk/part_efi.c
index d4d03de..b803d17 100644
--- a/disk/part_efi.c
+++ b/disk/part_efi.c
@@ -351,7 +351,7 @@ int part_get_info_efi(struct blk_desc *dev_desc, int part,
 	return 0;
 }
 
-#ifdef CONFIG_RKIMG_BOOTLOADER
+#if 0
 #if defined(CONFIG_SPL_KERNEL_BOOT) || !defined(CONFIG_SPL_BUILD)
 static void gpt_entry_modify(struct blk_desc *dev_desc,
 			     gpt_entry *gpt_pte,
@@ -452,7 +452,7 @@ static int part_test_efi(struct blk_desc *dev_desc)
 		|| (is_pmbr_valid(legacymbr) != 1)) {
 		return -1;
 	}
-#ifdef CONFIG_RKIMG_BOOTLOADER
+#if 0
 #if defined(CONFIG_SPL_KERNEL_BOOT) || !defined(CONFIG_SPL_BUILD)
 	gpt_entry *h_gpt_pte = NULL;
 	gpt_header *h_gpt_head = NULL;
@@ -1084,7 +1084,7 @@ static int is_pmbr_valid(legacy_mbr * mbr)
 {
 	int i = 0;
 
-#ifdef CONFIG_ARCH_ROCKCHIP
+#if 0
 	/*
 	 * In sd-update card, we use RKPARM partition in bootloader to load
 	 * firmware, and use MS-DOS partition in recovery to update system.
-- 
2.31.0

