From 600dbd1d1ca42e82aa11540954725a7c19952bef Mon Sep 17 00:00:00 2001
From: jjm2473 <1129525450@qq.com>
Date: Mon, 5 Jun 2023 16:58:02 +0800
Subject: [PATCH] supports legacy image

---
 include/configs/rockchip-common.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/configs/rockchip-common.h b/include/configs/rockchip-common.h
index c35b7db..b695a7d 100644
--- a/include/configs/rockchip-common.h
+++ b/include/configs/rockchip-common.h
@@ -179,6 +179,8 @@
 	"run distro_bootcmd;"
 #endif
 
+#define CONFIG_IMAGE_FORMAT_LEGACY /* enable also legacy image format */
+
 #endif /* CONFIG_SPL_BUILD */
 
 #define CONFIG_DISPLAY_BOARDINFO_LATE
-- 
2.31.0

