From ca4343eb308b019c6bc6930361b56306cd3ba55d Mon Sep 17 00:00:00 2001
From: jjm2473 <1129525450@qq.com>
Date: Mon, 5 Jun 2023 18:48:45 +0800
Subject: [PATCH] fix boorargs not in effect

---
 arch/arm/mach-rockchip/board.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/arch/arm/mach-rockchip/board.c b/arch/arm/mach-rockchip/board.c
index 9ffb017..1db4711 100644
--- a/arch/arm/mach-rockchip/board.c
+++ b/arch/arm/mach-rockchip/board.c
@@ -1115,6 +1115,7 @@ void board_quiesce_devices(void *images)
 #endif
 }
 
+#ifdef CONFIG_CMD_BOOT_ANDROID
 char *board_fdt_chosen_bootargs(void *fdt)
 {
 	/* bootargs_ext is used when dtbo is applied. */
@@ -1240,6 +1241,7 @@ char *board_fdt_chosen_bootargs(void *fdt)
 
 	return (char *)bootargs;
 }
+#endif
 
 int ft_verify_fdt(void *fdt)
 {
-- 
2.31.0

