Index: u-boot-2017.09/include/configs/rockchip-common.h
===================================================================
--- u-boot-2017.09.orig/include/configs/rockchip-common.h
+++ u-boot-2017.09/include/configs/rockchip-common.h
@@ -106,9 +106,7 @@
 	BOOT_TARGET_MMC(func) \
 	BOOT_TARGET_MTD(func) \
 	BOOT_TARGET_RKNAND(func) \
-	BOOT_TARGET_USB(func) \
-	BOOT_TARGET_PXE(func) \
-	BOOT_TARGET_DHCP(func)
+	BOOT_TARGET_USB(func)
 
 #ifdef CONFIG_ARM64
 #define ROOT_UUID "B921B045-1DF0-41C3-AF44-4C6F280D3FAE;\0"
