From f9dced95ce10893799be34534f605b34d15f3c39 Mon Sep 17 00:00:00 2001
From: jjm2473 <1129525450@qq.com>
Date: Mon, 5 Jun 2023 16:43:41 +0800
Subject: [PATCH] skip android boot when cmd disabled

---
 include/configs/rockchip-common.h | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/include/configs/rockchip-common.h b/include/configs/rockchip-common.h
index 002dcd6..c35b7db 100644
--- a/include/configs/rockchip-common.h
+++ b/include/configs/rockchip-common.h
@@ -168,12 +168,15 @@
 #elif defined(CONFIG_FIT_SIGNATURE)
 #define RKIMG_BOOTCOMMAND			\
 	"boot_fit;"
-#else
+#elif defined(CONFIG_CMD_BOOT_ANDROID)
 #define RKIMG_BOOTCOMMAND			\
 	"boot_android ${devtype} ${devnum};"	\
 	"boot_fit;"				\
 	"bootrkp;"				\
 	"run distro_bootcmd;"
+#else
+#define RKIMG_BOOTCOMMAND			\
+	"run distro_bootcmd;"
 #endif
 
 #endif /* CONFIG_SPL_BUILD */
-- 
2.31.0

