--- a/Makefile.am
+++ b/Makefile.am
@@ -8,7 +8,7 @@ else
 EXAMPLES_DIR =
 endif
 
-SUBDIRS = lib include doc tests m4 $(EXAMPLES_DIR)
+SUBDIRS = lib include m4 $(EXAMPLES_DIR)
 
 
 # we include the whole debian/ dir in EXTRA_DIST because there's a problem
--- a/Makefile.in
+++ b/Makefile.in
@@ -164,7 +164,7 @@ AUTOMAKE_OPTIONS = foreign 1.6 dist-zip
 @THEORA_ENABLE_EXAMPLES_TRUE@EXAMPLES_DIR = examples
 @THEORA_ENABLE_EXAMPLES_FALSE@EXAMPLES_DIR = 
 
-SUBDIRS = lib include doc tests m4 $(EXAMPLES_DIR)
+SUBDIRS = lib include m4 $(EXAMPLES_DIR)
 
 # we include the whole debian/ dir in EXTRA_DIST because there's a problem
 # with autotools and HFS+ MacOSX file systems that caused debian/Makefile.am
@@ -200,7 +200,7 @@ DIST_COMMON = README AUTHORS COPYING Mak
 	missing mkinstalldirs theora-uninstalled.pc.in theora.pc.in \
 	theoradec-uninstalled.pc.in theoradec.pc.in \
 	theoraenc-uninstalled.pc.in theoraenc.pc.in
-DIST_SUBDIRS = lib include doc tests m4 examples
+DIST_SUBDIRS = lib include m4 examples
 all: config.h
 	$(MAKE) $(AM_MAKEFLAGS) all-recursive
 
