--- a/Makefile
+++ b/Makefile
@@ -62,8 +62,7 @@ HDR_SUBDIRS = include libxfs
 LIBFROG_SUBDIR = libfrog
 DLIB_SUBDIRS = libxlog libxcmd libhandle
 LIB_SUBDIRS = libxfs $(DLIB_SUBDIRS)
-TOOL_SUBDIRS = copy db estimate fsck fsr growfs io logprint mkfs quota \
-		mdrestore repair rtcp m4 man doc debian spaceman
+TOOL_SUBDIRS = db fsck growfs io mkfs repair
 
 ifeq ("$(ENABLE_SCRUB)","yes")
 TOOL_SUBDIRS += scrub
