--- a/arch/mips/bcm63xx/Kconfig
+++ b/arch/mips/bcm63xx/Kconfig
@@ -25,6 +25,8 @@ config BCM63XX_CPU_6318
 	select HAVE_PCI
 	select BCM63XX_OHCI
 	select BCM63XX_EHCI
+	select PINCTRL
+	select PINCTRL_BCM6318
 
 config BCM63XX_CPU_6328
 	bool "support 6328 CPU"
@@ -32,6 +34,8 @@ config BCM63XX_CPU_6328
 	select HAVE_PCI
 	select BCM63XX_OHCI
 	select BCM63XX_EHCI
+	select PINCTRL
+	select PINCTRL_BCM6328
 
 config BCM63XX_CPU_6338
 	bool "support 6338 CPU"
@@ -47,6 +51,8 @@ config BCM63XX_CPU_6348
 	select SYS_HAS_CPU_BMIPS32_3300
 	select HAVE_PCI
 	select BCM63XX_OHCI
+	select PINCTRL
+	select PINCTRL_BCM6348
 
 config BCM63XX_CPU_6358
 	bool "support 6358 CPU"
@@ -54,6 +60,8 @@ config BCM63XX_CPU_6358
 	select HAVE_PCI
 	select BCM63XX_OHCI
 	select BCM63XX_EHCI
+	select PINCTRL
+	select PINCTRL_BCM6358
 
 config BCM63XX_CPU_6362
 	bool "support 6362 CPU"
@@ -61,6 +69,8 @@ config BCM63XX_CPU_6362
 	select HAVE_PCI
 	select BCM63XX_OHCI
 	select BCM63XX_EHCI
+	select PINCTRL
+	select PINCTRL_BCM6362
 
 config BCM63XX_CPU_6368
 	bool "support 6368 CPU"
@@ -68,6 +78,8 @@ config BCM63XX_CPU_6368
 	select HAVE_PCI
 	select BCM63XX_OHCI
 	select BCM63XX_EHCI
+	select PINCTRL
+	select PINCTRL_BCM6368
 
 config BCM63XX_CPU_63268
 	bool "support 63268 CPU"
@@ -75,6 +87,8 @@ config BCM63XX_CPU_63268
 	select HAVE_PCI
 	select BCM63XX_OHCI
 	select BCM63XX_EHCI
+	select PINCTRL
+	select PINCTRL_BCM63268
 endmenu
 
 source "arch/mips/bcm63xx/boards/Kconfig"
