--- a/contrib/i2pd.conf
+++ b/contrib/i2pd.conf
@@ -8,12 +8,12 @@
 
 ## Tunnels config file
 ## Default: ~/.i2pd/tunnels.conf or /var/lib/i2pd/tunnels.conf
-# tunconf = /var/lib/i2pd/tunnels.conf
+tunconf = /etc/i2pd/tunnels.conf
 
 ## Tunnels config files path
 ## Use that path to store separated tunnels in different config files.
 ## Default: ~/.i2pd/tunnels.d or /var/lib/i2pd/tunnels.d
-# tunnelsdir = /var/lib/i2pd/tunnels.d
+tunnelsdir = /etc/i2pd/tunnels.d
 
 ## Where to write pidfile (default: i2pd.pid, not used in Windows)
 # pidfile = /run/i2pd.pid
@@ -26,12 +26,12 @@
 ##  * stdout - print log entries to stdout
 ##  * file - log entries to a file
 ##  * syslog - use syslog, see man 3 syslog
-# log = file
+log = syslog
 ## Path to logfile (default - autodetect)
 # logfile = /var/log/i2pd/i2pd.log
 ## Log messages above this level (debug, info, *warn, error, none)
 ## If you set it to none, logging will be disabled
-# loglevel = warn
+loglevel = error
 ## Write full CLF-formatted date and time to log (default: write only time)
 # logclftime = true
 
@@ -84,7 +84,7 @@ ipv6 = false
 
 ## Router will not accept transit tunnels, disabling transit traffic completely
 ## (default = false)
-# notransit = true
+notransit = true
 
 ## Router will be floodfill
 ## Note: that mode uses much more network connections and CPU!
@@ -95,8 +95,10 @@ ipv6 = false
 ## Uncomment and set to 'false' to disable Web Console
 # enabled = true
 ## Address and port service will listen on
-address = 127.0.0.1
+address = 192.168.1.1
+# address = ::
 port = 7070
+# strictheaders = false
 ## Path to web console, default "/"
 # webroot = /
 ## Uncomment following lines to enable Web Console authentication
@@ -108,7 +110,7 @@ port = 7070
 ## Uncomment and set to 'false' to disable HTTP Proxy
 # enabled = true
 ## Address and port service will listen on
-address = 127.0.0.1
+address = 192.168.1.1
 port = 4444
 ## Optional keys file for proxy local destination
 # keys = http-proxy-keys.dat
@@ -122,7 +124,7 @@ port = 4444
 ## Uncomment and set to 'false' to disable SOCKS Proxy
 # enabled = true
 ## Address and port service will listen on
-address = 127.0.0.1
+address = 192.168.1.1
 port = 4447
 ## Optional keys file for proxy local destination
 # keys = socks-proxy-keys.dat
@@ -237,9 +239,9 @@ verify = true
 
 [persist]
 ## Save peer profiles on disk (default: true)
-# profiles = true
+profiles = false
 ## Save full addresses on disk (default: true)
-# addressbook = true
+addressbook = false
 
 [cpuext]
 ## Use CPU AES-NI instructions set when work with cryptography when available (default: true)
