--- a/arch/mips/pci/Makefile
+++ b/arch/mips/pci/Makefile
@@ -58,6 +58,7 @@ obj-$(CONFIG_MIKROTIK_RB532)	+= pci-rc32
 obj-$(CONFIG_CAVIUM_OCTEON_SOC) += pci-octeon.o pcie-octeon.o
 obj-$(CONFIG_CPU_XLR)		+= pci-xlr.o
 obj-$(CONFIG_CPU_XLP)		+= pci-xlp.o
+obj-$(CONFIG_ADM8668)		+= pci-adm8668.o
 
 ifdef CONFIG_PCI_MSI
 obj-$(CONFIG_CAVIUM_OCTEON_SOC) += msi-octeon.o
--- a/include/linux/pci_ids.h
+++ b/include/linux/pci_ids.h
@@ -1820,6 +1820,9 @@
 
 #define PCI_VENDOR_ID_CB		0x1307	/* Measurement Computing */
 
+#define PCI_VENDOR_ADMTEK		0x1317
+#define PCI_DEVICE_ID_ADM8668		0x8688
+
 #define PCI_VENDOR_ID_SIIG		0x131f
 #define PCI_SUBVENDOR_ID_SIIG		0x131f
 #define PCI_DEVICE_ID_SIIG_1S_10x_550	0x1000
