--- a/scan/sane/OrbliteScan/LinuxCommon.h
+++ b/scan/sane/OrbliteScan/LinuxCommon.h
@@ -18,10 +18,8 @@ typedef u_int32_t			UInt32;
 typedef int32_t				SInt32;
 //typedef unsigned long			UInt32;
 //typedef signed long			SInt32;
-typedef __S64_TYPE			SInt64;
-typedef __U64_TYPE			UInt64;
-typedef __S64_TYPE			int64_t;
-typedef __U64_TYPE			uint64_t;
+typedef int64_t 			SInt64;
+typedef uint64_t			UInt64;
 
 //typedef unsigned long	ULONG;
 //typedef void*			LPVOID;
