--- a/scripts/fakeroot.in
+++ b/scripts/fakeroot.in
@@ -30,13 +30,20 @@ fatal ()
 }
 
 # strip /bin/fakeroot to find install prefix
-FAKEROOT_PREFIX=@prefix@
-FAKEROOT_BINDIR=@bindir@
+if [ -n "$STAGING_DIR_HOST" ]; then
+    USEABSLIBPATH=1
+    FAKEROOT_LIB=${STAGING_DIR_HOST}/lib/lib@fakeroot_transformed@@DLSUFFIX@
+    FAKED=${STAGING_DIR_HOST}/bin/faked
+    PATHS=
+else
+    FAKEROOT_PREFIX=@prefix@
+    FAKEROOT_BINDIR=@bindir@
 
-USEABSLIBPATH=@LDPRELOADABS@
-FAKEROOT_LIB=lib@fakeroot_transformed@@DLSUFFIX@
-PATHS=@libdir@:${FAKEROOT_PREFIX}/lib64/libfakeroot:${FAKEROOT_PREFIX}/lib32/libfakeroot
-FAKED=${FAKEROOT_BINDIR}/@faked_transformed@
+    USEABSLIBPATH=@LDPRELOADABS@
+    FAKEROOT_LIB=lib@fakeroot_transformed@@DLSUFFIX@
+    PATHS=@libdir@:${FAKEROOT_PREFIX}/lib64/libfakeroot:${FAKEROOT_PREFIX}/lib32/libfakeroot
+    FAKED=${FAKEROOT_BINDIR}/@faked_transformed@
+fi
 
 FAKED_MODE="unknown-is-root"
 export FAKED_MODE
