--- a/macros/db3-check.m4
+++ b/macros/db3-check.m4
@@ -148,9 +148,9 @@ if test "x$bdb_required" = "xyes"; then
                         dnl -- LD_LIBRARY_PATH on many platforms. This will be fairly
                         dnl -- portable hopefully. Reference:
                         dnl -- http://lists.gnu.org/archive/html/autoconf/2009-03/msg00040.html
-                        eval export $shlibpath_var=$bdblibdir
+#                        eval export $shlibpath_var=$bdblibdir
                         NETATALK_BDB_TRY_LINK
-                        eval export $shlibpath_var=$saved_shlibpath_var
+#                        eval export $shlibpath_var=$saved_shlibpath_var
 
                         if test x"${atalk_cv_bdb_version}" = x"yes"; then
                             BDB_CFLAGS="-I${bdbdir}/include${subdir}"
@@ -177,9 +177,9 @@ if test "x$bdb_required" = "xyes"; then
                            CPPFLAGS="-I${bdbdir}/include${subdir} $CPPFLAGS"
                            LDFLAGS="-L$bdblibdir $LDFLAGS"
 
-                           eval export $shlibpath_var=$bdblibdir
+#                           eval export $shlibpath_var=$bdblibdir
                            NETATALK_BDB_TRY_LINK
-                           eval export $shlibpath_var=$saved_shlibpath_var
+#                           eval export $shlibpath_var=$saved_shlibpath_var
 
                            if test x"${atalk_cv_bdb_version}" = x"yes"; then
                               BDB_CFLAGS="-I${bdbdir}/include${subdir}"
