From 32df6e155ccfc83216321925273c3e75e631ebe6 Mon Sep 17 00:00:00 2001
From: Andrew Bauer <zonexpertconsulting@outlook.com>
Date: Wed, 22 Jan 2020 09:59:47 -0600
Subject: [PATCH] fix ftbs multiple def of invalid_dircache_entries

---
 etc/afpd/directory.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/etc/afpd/directory.h
+++ b/etc/afpd/directory.h
@@ -91,7 +91,7 @@ struct maccess {
 #define	AR_UWRITE	(1<<2)
 #define	AR_UOWN		(1<<7)
 
-q_t *invalid_dircache_entries;
+extern q_t *invalid_dircache_entries;
 
 typedef int (*dir_loop)(struct dirent *, char *, void *);
 
