--- a/gost_pmeth.c
+++ b/gost_pmeth.c
@@ -608,7 +608,7 @@ static int pkey_gost_mac_ctrl_str(EVP_PK
     if (strcmp(type, hexkey_ctrl_string) == 0) {
         long keylen;
         int ret;
-        unsigned char *keybuf = string_to_hex(value, &keylen);
+        unsigned char *keybuf = OPENSSL_hexstr2buf(value, &keylen);
         if (!keybuf || keylen != 32) {
             GOSTerr(GOST_F_PKEY_GOST_MAC_CTRL_STR,
                     GOST_R_INVALID_MAC_KEY_LENGTH);
