--- a/lib/berkeley-db-1.xx/PORT/bsd.4.4/include/db.h
+++ b/lib/berkeley-db-1.xx/PORT/bsd.4.4/include/db.h
@@ -37,7 +37,7 @@
 #define	_DB_H_
 
 #include <sys/types.h>
-#include <sys/cdefs.h>
+#include "cdefs.h"
 
 #include <limits.h>
 
--- a/lib/berkeley-db-1.xx/PORT/clib/memmove.c
+++ b/lib/berkeley-db-1.xx/PORT/clib/memmove.c
@@ -38,7 +38,7 @@
 static char sccsid[] = "@(#)bcopy.c	8.1 (Berkeley) 6/4/93";
 #endif /* LIBC_SCCS and not lint */
 
-#include <sys/cdefs.h>
+#include "cdefs.h"
 #include <string.h>
 
 /*
--- a/lib/berkeley-db-1.xx/PORT/clib/snprintf.c
+++ b/lib/berkeley-db-1.xx/PORT/clib/snprintf.c
@@ -1,5 +1,5 @@
 #include <sys/types.h>
-#include <sys/cdefs.h>
+#include "cdefs.h"
 
 #include <compat.h>
 
--- a/lib/berkeley-db-1.xx/PORT/hpux.9.01/local/hp_siglist.c
+++ b/lib/berkeley-db-1.xx/PORT/hpux.9.01/local/hp_siglist.c
@@ -2,7 +2,7 @@
  * Derived from:
  * static char sccsid[] = "@(#)siglist.c	8.1 (Berkeley) 6/4/93";
  */
-#include <sys/cdefs.h>
+#include "cdefs.h"
 
 #include <signal.h>
 
--- a/lib/berkeley-db-1.xx/PORT/ultrix.4.2/include/db.h
+++ b/lib/berkeley-db-1.xx/PORT/ultrix.4.2/include/db.h
@@ -37,7 +37,7 @@
 #define	_DB_H_
 
 #include <sys/types.h>
-#include <sys/cdefs.h>
+#include "cdefs.h"
 
 #include <limits.h>
 
--- a/lib/berkeley-db-1.xx/include/db.h
+++ b/lib/berkeley-db-1.xx/include/db.h
@@ -37,7 +37,7 @@
 #define	_DB_H_
 
 #include <sys/types.h>
-#include <sys/cdefs.h>
+#include "cdefs.h"
 
 #include <limits.h>
 
