--- a/drivers/net/phy/Kconfig
+++ b/drivers/net/phy/Kconfig
@@ -431,6 +431,8 @@ config QSEMI_PHY
 
 source "drivers/net/phy/realtek/Kconfig"
 
+source "drivers/net/phy/rtl8261n/Kconfig"
+
 config RENESAS_PHY
 	tristate "Renesas PHYs"
 	help
--- a/drivers/net/phy/Makefile
+++ b/drivers/net/phy/Makefile
@@ -110,6 +110,7 @@ obj-$(CONFIG_NXP_TJA11XX_PHY)	+= nxp-tja
 obj-y				+= qcom/
 obj-$(CONFIG_QSEMI_PHY)		+= qsemi.o
 obj-$(CONFIG_REALTEK_PHY)	+= realtek/
+obj-y				+= rtl8261n/
 obj-$(CONFIG_RENESAS_PHY)	+= uPD60620.o
 obj-$(CONFIG_ROCKCHIP_PHY)	+= rockchip.o
 obj-$(CONFIG_SMSC_PHY)		+= smsc.o
