--- a/configure.in
+++ b/configure.in
@@ -29,6 +29,8 @@ AC_C_VOLATILE
 #AC_FUNC_STRTOD
 #AC_CHECK_FUNCS([ftime gethostbyname gettimeofday localtime_r memset select socket strchr strerror strstr strtoull])
 
+AC_SEARCH_LIBS([clock_gettime],[rt])
+
 #
 # save the configure arguments
 #
--- a/src/dsl_cpe_linux.h
+++ b/src/dsl_cpe_linux.h
@@ -45,7 +45,8 @@
 #include <arpa/inet.h>
 #include <sys/socket.h>          /* socket */
 #include <sys/sem.h>             /* semget */
-#include <semaphore.h>           /* sem_t */ 
+#include <semaphore.h>           /* sem_t */
+#include <limits.h>
 
 #ifdef DSL_DEBUG_TOOL_INTERFACE
 #include <sys/socket.h>
