--- a/frontends/nss/ecm_nss_common.c
+++ b/frontends/nss/ecm_nss_common.c
@@ -67,6 +67,7 @@ bool ecm_nss_ipv6_is_conn_limit_reached(

 	return false;
 }
+EXPORT_SYMBOL(ecm_nss_ipv6_is_conn_limit_reached);
 #endif

 /*
@@ -116,3 +117,4 @@ bool ecm_nss_ipv4_is_conn_limit_reached(

 	return false;
 }
+EXPORT_SYMBOL(ecm_nss_ipv4_is_conn_limit_reached);
--- a/frontends/nss/ecm_nss_non_ported_ipv4.c
+++ b/frontends/nss/ecm_nss_non_ported_ipv4.c
@@ -1831,6 +1831,7 @@ struct ecm_front_end_connection_instance

 	return feci;
 }
+EXPORT_SYMBOL(ecm_nss_non_ported_ipv4_connection_instance_alloc);

 /*
  * ecm_nss_non_ported_ipv4_debugfs_init()
--- a/frontends/nss/ecm_nss_non_ported_ipv4.h
+++ b/frontends/nss/ecm_nss_non_ported_ipv4.h
@@ -19,7 +19,7 @@

 extern bool ecm_nss_non_ported_ipv4_debugfs_init(struct dentry *dentry);

-extern struct ecm_front_end_connection_instance *ecm_nss_non_ported_ipv4_connection_instance_alloc(
+struct ecm_front_end_connection_instance *ecm_nss_non_ported_ipv4_connection_instance_alloc(
 								uint32_t accel_flags,
 								int protocol,
 								struct ecm_db_connection_instance **nci);
--- a/frontends/nss/ecm_nss_non_ported_ipv6.c
+++ b/frontends/nss/ecm_nss_non_ported_ipv6.c
@@ -1657,6 +1657,7 @@ struct ecm_front_end_connection_instance

 	return feci;
 }
+EXPORT_SYMBOL(ecm_nss_non_ported_ipv6_connection_instance_alloc);

 /*
  * ecm_nss_non_ported_ipv6_debugfs_init()
--- a/frontends/nss/ecm_nss_non_ported_ipv6.h
+++ b/frontends/nss/ecm_nss_non_ported_ipv6.h
@@ -19,7 +19,7 @@

 extern bool ecm_nss_non_ported_ipv6_debugfs_init(struct dentry *dentry);

-extern struct ecm_front_end_connection_instance *ecm_nss_non_ported_ipv6_connection_instance_alloc(
+struct ecm_front_end_connection_instance *ecm_nss_non_ported_ipv6_connection_instance_alloc(
 								uint32_t accel_flags,
 								int protocol,
 								struct ecm_db_connection_instance **nci);
--- a/frontends/nss/ecm_nss_ported_ipv4.c
+++ b/frontends/nss/ecm_nss_ported_ipv4.c
@@ -1906,6 +1906,7 @@ struct ecm_front_end_connection_instance

 	return feci;
 }
+EXPORT_SYMBOL(ecm_nss_ported_ipv4_connection_instance_alloc);

 /*
  * ecm_nss_ported_ipv4_debugfs_init()
--- a/frontends/nss/ecm_nss_ported_ipv4.h
+++ b/frontends/nss/ecm_nss_ported_ipv4.h
@@ -19,7 +19,7 @@

 extern bool ecm_nss_ported_ipv4_debugfs_init(struct dentry *dentry);

-extern struct ecm_front_end_connection_instance *ecm_nss_ported_ipv4_connection_instance_alloc(
+struct ecm_front_end_connection_instance *ecm_nss_ported_ipv4_connection_instance_alloc(
 								uint32_t accel_flags,
 								int protocol,
 								struct ecm_db_connection_instance **nci);
--- a/frontends/nss/ecm_nss_ported_ipv6.c
+++ b/frontends/nss/ecm_nss_ported_ipv6.c
@@ -1812,6 +1812,7 @@ struct ecm_front_end_connection_instance

 	return feci;
 }
+EXPORT_SYMBOL(ecm_nss_ported_ipv6_connection_instance_alloc);

 /*
  * ecm_nss_ported_ipv6_debugfs_init()
--- a/frontends/nss/ecm_nss_ported_ipv6.h
+++ b/frontends/nss/ecm_nss_ported_ipv6.h
@@ -19,7 +19,7 @@

 extern bool ecm_nss_ported_ipv6_debugfs_init(struct dentry *dentry);

-extern struct ecm_front_end_connection_instance *ecm_nss_ported_ipv6_connection_instance_alloc(
+struct ecm_front_end_connection_instance *ecm_nss_ported_ipv6_connection_instance_alloc(
 								uint32_t accel_flags,
 								int protocol,
 								struct ecm_db_connection_instance **nci);
