--- a/frontends/include/ecm_front_end_common.h
+++ b/frontends/include/ecm_front_end_common.h
@@ -339,15 +339,19 @@ static inline bool ecm_front_end_ppppoe_
 	fe_type = ecm_front_end_type_get();
 	switch (fe_type) {
 #ifdef ECM_FRONT_END_NSS_ENABLE
+#ifdef ECM_INTERFACE_PPPOE_ENABLE
 	case ECM_FRONT_END_TYPE_NSS:
 		ret = (nss_pppoe_get_br_accel_mode() == NSS_PPPOE_BR_ACCEL_MODE_DIS);
 		break;
 #endif
+#endif
 #ifdef ECM_FRONT_END_SFE_ENABLE
+#ifdef ECM_INTERFACE_PPPOE_ENABLE
 	case ECM_FRONT_END_TYPE_SFE:
 		ret = (sfe_pppoe_get_br_accel_mode() == SFE_PPPOE_BR_ACCEL_MODE_DISABLED);
 		break;
 #endif
+#endif
 	default:
 		DEBUG_TRACE("front end type: %d\n", fe_type);
 		break;
@@ -369,15 +373,19 @@ static inline bool ecm_front_end_ppppoe_
 	fe_type = ecm_front_end_type_get();
 	switch (fe_type) {
 #ifdef ECM_FRONT_END_NSS_ENABLE
+#ifdef ECM_INTERFACE_PPPOE_ENABLE
 	case ECM_FRONT_END_TYPE_NSS:
 		ret = (nss_pppoe_get_br_accel_mode() == NSS_PPPOE_BR_ACCEL_MODE_EN_3T);
 		break;
 #endif
+#endif
 #ifdef ECM_FRONT_END_SFE_ENABLE
+#ifdef ECM_INTERFACE_PPPOE_ENABLE
 	case ECM_FRONT_END_TYPE_SFE:
 		ret = (sfe_pppoe_get_br_accel_mode() == SFE_PPPOE_BR_ACCEL_MODE_EN_3T);
 		break;
 #endif
+#endif
 	default:
 		DEBUG_WARN("front end type: %d is not supported\n", fe_type);
 		break;
