--- a/frontends/nss/ecm_nss_non_ported_ipv4.c
+++ b/frontends/nss/ecm_nss_non_ported_ipv4.c
@@ -855,7 +855,7 @@ static void ecm_nss_non_ported_ipv4_conn
 #endif
 			break;
 		case ECM_DB_IFACE_TYPE_RAWIP:
-#ifdef ECM_INTERFACE_RAWIP_ENABLE
+#if defined(NSS_FIRMWARE_VERSION_12_5) && defined(ECM_INTERFACE_RAWIP_ENABLE)
 			nircm->valid_flags |= NSS_IPV4_RULE_CREATE_RAWIP_VALID;
 #else
 			rule_invalid = true;
@@ -1072,7 +1072,7 @@ static void ecm_nss_non_ported_ipv4_conn
 #endif
 			break;
 		case ECM_DB_IFACE_TYPE_RAWIP:
-#ifdef ECM_INTERFACE_RAWIP_ENABLE
+#if defined(NSS_FIRMWARE_VERSION_12_5) && defined(ECM_INTERFACE_RAWIP_ENABLE)
 			nircm->valid_flags |= NSS_IPV4_RULE_CREATE_RAWIP_VALID;
 #else
 			rule_invalid = true;
