--- a/ecm_interface.c
+++ b/ecm_interface.c
@@ -4057,11 +4057,7 @@ static uint32_t ecm_interface_multicast_
 				 * For MLO bond netdevice, destination for multicast is bond netdevice itself
 				 * Therefore, slave lookup is not needed.
 				 */
-#ifdef ECM_FRONT_END_SFE_ENABLE
-				if (ecm_front_end_is_lag_master(dest_dev) && !bond_is_mlo_device(dest_dev)) {
-#else
 				if (ecm_front_end_is_lag_master(dest_dev)) {
-#endif
 					/*
 					 * Link aggregation
 					 * Figure out which slave device of the link aggregation will be used to reach the destination.
