--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -224,7 +224,6 @@ install-exec-hook: $(bin_PROGRAMS)
 
 # Extra rule for linux-2.6 kernel object
 if KERNEL_2_6
-drv_vmmc_CFLAGS += -fno-common
 drv_vmmc_OBJS = "$(subst .c,.o, $(drv_vmmc_SOURCES) $(nodist_drv_vmmc_SOURCES))"
 
 drv_vmmc.ko: $(drv_vmmc_SOURCES) $(EXTRA_DIST)
@@ -239,7 +238,7 @@ drv_vmmc.ko: $(drv_vmmc_SOURCES) $(EXTRA
 	@echo "# drv_vmmc: Generated to build Linux 2.6.x kernel object" > $(PWD)/Kbuild
 	@echo "obj-m := $(subst .ko,.o,$@)"			>> $(PWD)/Kbuild
 	@echo "$(subst .ko,,$@)-y := $(drv_vmmc_OBJS)"	>> $(PWD)/Kbuild
-	@echo "EXTRA_CFLAGS := -DHAVE_CONFIG_H  $(CFLAGS) $(drv_vmmc_CFLAGS) $(INCLUDES)"	>> $(PWD)/Kbuild
+	@echo "EXTRA_CFLAGS := -DHAVE_CONFIG_H $(drv_vmmc_CFLAGS) $(INCLUDES)"	>> $(PWD)/Kbuild
 	$(MAKE) ARCH=@KERNEL_ARCH@ -C @KERNEL_BUILD_PATH@ O=@KERNEL_BUILD_PATH@ M=$(PWD) modules
 
 clean-generic:
