--- a/tn7atm.c
+++ b/tn7atm.c
@@ -633,7 +633,7 @@ static int turbodsl_check_priority_type(
  *  Description: tnetd73xx SAR interrupt.
  *
  *~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
-static void tn7atm_sar_irq (int irq, void *voiddev, struct pt_regs *regs)
+static irqreturn_t tn7atm_sar_irq (int irq, void *voiddev)
 {
   struct atm_dev *atmdev;
   Tn7AtmPrivate *priv;
@@ -660,6 +660,7 @@ static void tn7atm_sar_irq (int irq, voi
 #ifdef TIATM_INST_SUPP
   psp_trace_par (ATM_DRV_SAR_ISR_EXIT, retval);
 #endif
+  return IRQ_HANDLED;
 }
 
 /*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
@@ -669,7 +670,7 @@ static void tn7atm_sar_irq (int irq, voi
  *  Description: tnetd73xx DSL interrupt.
  *
  *~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
-static void tn7atm_dsl_irq (int irq, void *voiddev, struct pt_regs *regs)
+static irqreturn_t tn7atm_dsl_irq (int irq, void *voiddev)
 {
   struct atm_dev *atmdev;
   Tn7AtmPrivate *priv;
@@ -691,6 +692,8 @@ static void tn7atm_dsl_irq (int irq, voi
 #ifdef TIATM_INST_SUPP
   psp_trace_par (ATM_DRV_DSL_ISR_EXIT, retval);
 #endif
+
+  return IRQ_HANDLED;
 }
 
 /*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
