--- a/tn7dsl.c
+++ b/tn7dsl.c
@@ -492,7 +492,6 @@ static void avsar_release(struct device
 }
 
 static struct device avsar = {
-	.bus_id    = "vlynq",
 	.release   = avsar_release,
 };
 
@@ -501,6 +500,8 @@ int shim_osLoadFWImage(unsigned char *pt
 	const struct firmware *fw_entry;
 	size_t size;
 
+	dev_set_name(&avsar, "avsar");
+
 	printk("requesting firmware image \"ar0700xx.bin\"\n");
 	if(device_register(&avsar) < 0) {
 		printk(KERN_ERR
