--- a/tn7atm.c
+++ b/tn7atm.c
@@ -78,8 +78,8 @@
 #include <linux/ctype.h>
 
 #include <asm/io.h>
-#include <asm/ar7/ar7.h>
-#include <asm/ar7/prom.h>
+#include <asm/mach-ar7/ar7.h>
+#include <asm/mach-ar7/prom.h>
 
 #include "dsl_hal_api.h"
 #ifdef AR7_EFM
--- a/tn7dsl.c
+++ b/tn7dsl.c
@@ -119,8 +119,8 @@
 #include <linux/firmware.h>
 
 #include <asm/io.h>
-#include <asm/ar7/ar7.h>
-#include <asm/ar7/prom.h>
+#include <asm/mach-ar7/ar7.h>
+#include <asm/mach-ar7/prom.h>
 
 /* Modules specific header files */
 #ifdef AR7_EFM
--- a/tn7sar.c
+++ b/tn7sar.c
@@ -55,8 +55,8 @@
 #include <linux/ctype.h>
 
 #include <asm/io.h>
-#include <asm/ar7/ar7.h>
-#include <asm/ar7/prom.h>
+#include <asm/mach-ar7/ar7.h>
+#include <asm/mach-ar7/prom.h>
 
 #define _CPHAL_AAL5
 #define _CPHAL_SAR
