--- a/iptables/xtables-multi.c
+++ b/iptables/xtables-multi.c
@@ -26,8 +26,10 @@ static const struct subcommand multi_sub
 	{"iptables-restore",    iptables_restore_main},
 	{"restore4",            iptables_restore_main},
 #endif
+#ifdef ENABLE_XML
 	{"iptables-xml",        iptables_xml_main},
 	{"xml",                 iptables_xml_main},
+#endif
 #ifdef ENABLE_IPV6
 	{"ip6tables",           ip6tables_main},
 	{"main6",               ip6tables_main},
