--- a/libxtables/xtables.c
+++ b/libxtables/xtables.c
@@ -887,12 +887,6 @@ static void xtables_check_options(const
 
 void xtables_register_match(struct xtables_match *me)
 {
-	if (me->next) {
-		fprintf(stderr, "%s: match \"%s\" already registered\n",
-			xt_params->program_name, me->name);
-		exit(1);
-	}
-
 	if (me->version == NULL) {
 		fprintf(stderr, "%s: match %s<%u> is missing a version\n",
 		        xt_params->program_name, me->name, me->revision);
@@ -1080,12 +1074,6 @@ void xtables_register_matches(struct xta
 
 void xtables_register_target(struct xtables_target *me)
 {
-	if (me->next) {
-		fprintf(stderr, "%s: target \"%s\" already registered\n",
-			xt_params->program_name, me->name);
-		exit(1);
-	}
-
 	if (me->version == NULL) {
 		fprintf(stderr, "%s: target %s<%u> is missing a version\n",
 		        xt_params->program_name, me->name, me->revision);
