--- a/libxtables/xtables.c
+++ b/libxtables/xtables.c
@@ -476,7 +476,7 @@ char *xtables_strdup(const char *s)
 	return dup;
 }
 
-static char *get_modprobe(void)
+__attribute__((unused)) static char *get_modprobe(void)
 {
 	int procfile;
 	char *ret;
@@ -511,6 +511,7 @@ static char *get_modprobe(void)
 
 int xtables_insmod(const char *modname, const char *modprobe, bool quiet)
 {
+#if 0
 	char *buf = NULL;
 	char *argv[4];
 	int status;
@@ -545,6 +546,7 @@ int xtables_insmod(const char *modname,
 	free(buf);
 	if (WIFEXITED(status) && WEXITSTATUS(status) == 0)
 		return 0;
+#endif
 	return -1;
 }
 
