--- a/libxtables/xtables.c
+++ b/libxtables/xtables.c
@@ -360,6 +360,7 @@ static char *get_modprobe(void)
 
 int xtables_insmod(const char *modname, const char *modprobe, bool quiet)
 {
+#if 0
 	char *buf = NULL;
 	char *argv[4];
 	int status;
@@ -394,6 +395,7 @@ int xtables_insmod(const char *modname,
 	free(buf);
 	if (WIFEXITED(status) && WEXITSTATUS(status) == 0)
 		return 0;
+#endif
 	return -1;
 }
 
