--- a/extensions/libxt_conntrack.c
+++ b/extensions/libxt_conntrack.c
@@ -1397,6 +1397,7 @@ static int conntrack3_mt6_xlate(struct x
 }
 
 static struct xtables_match conntrack_mt_reg[] = {
+#ifndef NO_LEGACY
 	{
 		.version       = XTABLES_VERSION,
 		.name          = "conntrack",
@@ -1472,6 +1473,7 @@ static struct xtables_match conntrack_mt
 		.alias	       = conntrack_print_name_alias,
 		.x6_options    = conntrack2_mt_opts,
 	},
+#endif
 	{
 		.version       = XTABLES_VERSION,
 		.name          = "conntrack",
@@ -1504,6 +1506,7 @@ static struct xtables_match conntrack_mt
 		.x6_options    = conntrack3_mt_opts,
 		.xlate	       = conntrack3_mt6_xlate,
 	},
+#ifndef NO_LEGACY
 	{
 		.family        = NFPROTO_UNSPEC,
 		.name          = "state",
@@ -1534,6 +1537,8 @@ static struct xtables_match conntrack_mt
 		.x6_parse      = state_ct23_parse,
 		.x6_options    = state_opts,
 	},
+#endif
+#ifndef NO_LEGACY
 	{
 		.family        = NFPROTO_UNSPEC,
 		.name          = "state",
@@ -1563,6 +1568,7 @@ static struct xtables_match conntrack_mt
 		.x6_parse      = state_parse,
 		.x6_options    = state_opts,
 	},
+#endif
 };
 
 void _init(void)
--- a/extensions/libxt_CT.c
+++ b/extensions/libxt_CT.c
@@ -349,6 +349,7 @@ static void notrack_ct2_tg_init(struct x
 }
 
 static struct xtables_target ct_target_reg[] = {
+#ifndef NO_LEGACY
 	{
 		.family		= NFPROTO_UNSPEC,
 		.name		= "CT",
@@ -374,6 +375,7 @@ static struct xtables_target ct_target_r
 		.x6_parse	= ct_parse_v1,
 		.x6_options	= ct_opts_v1,
 	},
+#endif
 	{
 		.family		= NFPROTO_UNSPEC,
 		.name		= "CT",
@@ -388,6 +390,7 @@ static struct xtables_target ct_target_r
 		.x6_parse	= ct_parse_v1,
 		.x6_options	= ct_opts_v1,
 	},
+#ifndef NO_LEGACY
 	{
 		.family        = NFPROTO_UNSPEC,
 		.name          = "NOTRACK",
@@ -425,6 +428,7 @@ static struct xtables_target ct_target_r
 		.revision      = 0,
 		.version       = XTABLES_VERSION,
 	},
+#endif
 };
 
 void _init(void)
--- a/extensions/libxt_multiport.c
+++ b/extensions/libxt_multiport.c
@@ -571,6 +571,7 @@ static int multiport_xlate6_v1(struct xt
 }
 
 static struct xtables_match multiport_mt_reg[] = {
+#ifndef NO_LEGACY
 	{
 		.family        = NFPROTO_IPV4,
 		.name          = "multiport",
@@ -601,6 +602,7 @@ static struct xtables_match multiport_mt
 		.x6_options    = multiport_opts,
 		.xlate         = multiport_xlate6,
 	},
+#endif
 	{
 		.family        = NFPROTO_IPV4,
 		.name          = "multiport",
