--- a/extensions/libxt_conntrack.c
+++ b/extensions/libxt_conntrack.c
@@ -1385,6 +1385,7 @@ static int conntrack3_mt6_xlate(struct x
 }
 
 static struct xtables_match conntrack_mt_reg[] = {
+#ifndef NO_LEGACY
 	{
 		.version       = XTABLES_VERSION,
 		.name          = "conntrack",
@@ -1460,6 +1461,7 @@ static struct xtables_match conntrack_mt
 		.alias	       = conntrack_print_name_alias,
 		.x6_options    = conntrack2_mt_opts,
 	},
+#endif
 	{
 		.version       = XTABLES_VERSION,
 		.name          = "conntrack",
@@ -1492,6 +1494,7 @@ static struct xtables_match conntrack_mt
 		.x6_options    = conntrack3_mt_opts,
 		.xlate	       = conntrack3_mt6_xlate,
 	},
+#ifndef NO_LEGACY
 	{
 		.family        = NFPROTO_UNSPEC,
 		.name          = "state",
@@ -1522,6 +1525,8 @@ static struct xtables_match conntrack_mt
 		.x6_parse      = state_ct23_parse,
 		.x6_options    = state_opts,
 	},
+#endif
+#ifndef NO_LEGACY
 	{
 		.family        = NFPROTO_UNSPEC,
 		.name          = "state",
@@ -1551,6 +1556,7 @@ static struct xtables_match conntrack_mt
 		.x6_parse      = state_parse,
 		.x6_options    = state_opts,
 	},
+#endif
 };
 
 void _init(void)
--- a/extensions/libxt_CT.c
+++ b/extensions/libxt_CT.c
@@ -363,6 +363,7 @@ static int xlate_ct1_tg(struct xt_xlate
 }
 
 static struct xtables_target ct_target_reg[] = {
+#ifndef NO_LEGACY
 	{
 		.family		= NFPROTO_UNSPEC,
 		.name		= "CT",
@@ -388,6 +389,7 @@ static struct xtables_target ct_target_r
 		.x6_parse	= ct_parse_v1,
 		.x6_options	= ct_opts_v1,
 	},
+#endif
 	{
 		.family		= NFPROTO_UNSPEC,
 		.name		= "CT",
@@ -403,6 +405,7 @@ static struct xtables_target ct_target_r
 		.x6_options	= ct_opts_v1,
 		.xlate		= xlate_ct1_tg,
 	},
+#ifndef NO_LEGACY
 	{
 		.family        = NFPROTO_UNSPEC,
 		.name          = "NOTRACK",
@@ -441,6 +444,7 @@ static struct xtables_target ct_target_r
 		.revision      = 0,
 		.version       = XTABLES_VERSION,
 	},
+#endif
 };
 
 void _init(void)
--- a/extensions/libxt_multiport.c
+++ b/extensions/libxt_multiport.c
@@ -591,6 +591,7 @@ static int multiport_xlate6_v1(struct xt
 }
 
 static struct xtables_match multiport_mt_reg[] = {
+#ifndef NO_LEGACY
 	{
 		.family        = NFPROTO_IPV4,
 		.name          = "multiport",
@@ -621,6 +622,7 @@ static struct xtables_match multiport_mt
 		.x6_options    = multiport_opts,
 		.xlate         = multiport_xlate6,
 	},
+#endif
 	{
 		.family        = NFPROTO_IPV4,
 		.name          = "multiport",
