From: coolsnowwolf <coolsnowwolf@gmail.com>
Subject: [PATCH] iptables: add BCM fullcone NAT mode

Signed-off-by: coolsnowwolf <coolsnowwolf@gmail.com>
---
--- a/extensions/libxt_NAT.c
+++ b/extensions/libxt_NAT.c
@@ -39,6 +39,7 @@
 	O_TO_SRC,
 	O_TO_PORTS,
 	O_RANDOM,
+	O_MODE,
 	O_RANDOM_FULLY,
 	O_PERSISTENT,
 };
@@ -61,7 +62,9 @@
 " --random\n"
 "				Randomize source port.\n"
 " --random-fully\n"
-"				Fully randomize source port.\n");
+"				Fully randomize source port.\n"
+" --mode <fullcone|symmetric>\n"
+"				NAT mode.\n");
 }
 
 static void DNAT_help(void)
@@ -104,6 +107,7 @@
 	{.name = "to-ports", .id = O_TO_PORTS, .type = XTTYPE_STRING},
 	{.name = "random", .id = O_RANDOM, .type = XTTYPE_NONE},
 	{.name = "random-fully", .id = O_RANDOM_FULLY, .type = XTTYPE_NONE},
+	{.name = "mode", .id = O_MODE, .type = XTTYPE_STRING},
 	XTOPT_TABLEEND,
 };
 
@@ -280,6 +284,18 @@
 	case O_RANDOM_FULLY:
 		range->flags |= NF_NAT_RANGE_PROTO_RANDOM_FULLY;
 		break;
+	case O_MODE:
+		if (family != AF_INET)
+			xtables_error(PARAMETER_PROBLEM,
+				       "NAT mode is only supported for IPv4");
+		if (strcasecmp(cb->arg, "fullcone") == 0)
+			range->min_addr.ip = 1;
+		else if (strcasecmp(cb->arg, "symmetric") == 0)
+			range->min_addr.ip = 0;
+		else
+			xtables_error(PARAMETER_PROBLEM,
+				       "Unknown mode %s", cb->arg);
+		break;
 	}
 }
 
@@ -306,6 +322,9 @@
 	case O_RANDOM_FULLY:
 		mr->range->flags |= range.flags;
 		break;
+	case O_MODE:
+		mr->range->min_ip = range.min_addr.ip;
+		break;
 	}
 }
 
@@ -489,8 +508,34 @@
 PSX_GEN(SNAT6, RANGE2_INIT_FROM_RANGE, \
 	AF_INET6, "to:", "--to-source ", false, "snat")
 
-PSX_GEN(MASQUERADE, RANGE2_INIT_FROM_IPV4_MRC, \
-	AF_INET, "masq ports: ", "--to-ports ", true, "masquerade")
+static void MASQUERADE_print(const void *ip,
+			     const struct xt_entry_target *target, int numeric)
+{
+	struct nf_nat_range2 range = RANGE2_INIT_FROM_IPV4_MRC(target->data);
+
+	__NAT_print(&range, AF_INET, "masq ports: ", "", true);
+	if (range.min_addr.ip == 1)
+		printf(" mode: fullcone");
+}
+
+static void MASQUERADE_save(const void *ip,
+			    const struct xt_entry_target *target)
+{
+	struct nf_nat_range2 range = RANGE2_INIT_FROM_IPV4_MRC(target->data);
+
+	__NAT_print(&range, AF_INET, "--to-ports ", "--", true);
+	if (range.min_addr.ip == 1)
+		printf(" --mode fullcone");
+}
+
+static int MASQUERADE_xlate(struct xt_xlate *xl,
+			    const struct xt_xlate_tg_params *params)
+{
+	struct nf_nat_range2 range =
+		RANGE2_INIT_FROM_IPV4_MRC(params->target->data);
+
+	return __NAT_xlate(xl, &range, AF_INET, "masquerade");
+}
 
 PSX_GEN(MASQUERADE6, RANGE2_INIT_FROM_RANGE, \
 	AF_INET6, "masq ports: ", "--to-ports ", true, "masquerade")
