--- a/runtime/doc/Makefile
+++ b/runtime/doc/Makefile
@@ -323,10 +323,6 @@ all: tags vim.man evim.man vimdiff.man v
 # Use Vim to generate the tags file.  Can only be used when Vim has been
 # compiled and installed.  Supports multiple languages.
 vimtags: $(DOCS)
-	@if command -v $(VIMEXE); then \
-	    $(VIMEXE) --clean -eX -u doctags.vim >/dev/null && \
-	      echo "help tags updated"; \
-	  else echo "vim executable $(VIMEXE) not found; help tags not updated"; fi
 
 # Use "doctags" to generate the tags file.  Only works for English!
 tags: doctags $(DOCS)
