From 12d5ea5ca2d9a47a1cab06caf2b36967667a3daf Mon Sep 17 00:00:00 2001
From: Leo <thinkabit.ukim@gmail.com>
Date: Sun, 24 Nov 2019 20:58:20 +0100
Subject: [PATCH] Add missing include of limits.h for PATH_MAX

---
 evtest.c | 1 +
 1 file changed, 1 insertion(+)

--- a/evtest.c
+++ b/evtest.c
@@ -59,6 +59,7 @@
 #include <sys/time.h>
 #include <sys/types.h>
 #include <unistd.h>
+#include <limits.h> /* PATH_MAX */
 
 #define BITS_PER_LONG (sizeof(long) * 8)
 #define NBITS(x) ((((x)-1)/BITS_PER_LONG)+1)
