--- a/minissdp.c	2017-04-21 19:24:23.000000000 +0800
+++ b/minissdp.c	2021-06-22 07:01:24.000000000 +0800
@@ -865,8 +865,8 @@
 	lan_addr = get_lan_for_peer(sender);
 	if(lan_addr == NULL)
 	{
-		syslog(LOG_WARNING, "SSDP packet sender %s not from a LAN, ignoring",
-		       sender_str);
+		/* syslog(LOG_WARNING, "SSDP packet sender %s not from a LAN, ignoring",
+		       sender_str);  */
 		return;
 	}
 
