--- a/CMakeLists.txt	2020-09-07 14:23:49.000000000 +0800
+++ b/CMakeLists.txt	2020-09-07 14:23:24.000000000 +0800
@@ -11,7 +11,12 @@
 endif(WIN32)
 
 add_executable(ssocks  src/ssocks.c  ${AUX})
+INSTALL(TARGETS ssocks RUNTIME DESTINATION bin)
 add_executable(ssocksd src/ssocksd.c ${AUX})
+INSTALL(TARGETS ssocksd RUNTIME DESTINATION bin)
 add_executable(nsocks  src/nsocks.c  ${AUX})
+INSTALL(TARGETS nsocks RUNTIME DESTINATION bin)
 add_executable(rcsocks src/rcsocks.c ${AUX})
+INSTALL(TARGETS rcsocks RUNTIME DESTINATION bin)
 add_executable(rssocks src/rssocks.c ${AUX})
+INSTALL(TARGETS rssocks RUNTIME DESTINATION bin)
