From 08452889f1bd46f4b4f34b915a3095523de1758a Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Tue, 21 May 2019 15:17:33 +0100
Subject: [PATCH 499/725] .gitignore: Add *.dtbo explicitly

Signed-off-by: popcornmix <popcornmix@gmail.com>
---
 .gitignore | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/.gitignore
+++ b/.gitignore
@@ -15,7 +15,8 @@
 *.bin
 *.bz2
 *.c.[012]*.*
-*.dtb*
+*.dtb
+*.dtbo
 *.dtb.S
 *.dwo
 *.elf
