From f5d74e55edf6504f6f857b23b06dca7b4d3d3306 Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.org>
Date: Mon, 10 Jun 2019 17:22:44 +0100
Subject: [PATCH 651/773] bcm2711 dts: Disable the v3d node by default

Signed-off-by: Phil Elwell <phil@raspberrypi.org>
---
 arch/arm/boot/dts/bcm2711.dtsi | 4 ++++
 1 file changed, 4 insertions(+)

--- a/arch/arm/boot/dts/bcm2711.dtsi
+++ b/arch/arm/boot/dts/bcm2711.dtsi
@@ -13,6 +13,10 @@
 	};
 };
 
+&v3d {
+	status = "disabled";
+};
+
 &dma {
 	brcm,dma-channel-mask = <0x7ef5>;
 };
