From d188a0a3472f80003d2558639ebeb41e190a4b9b Mon Sep 17 00:00:00 2001
From: Klaus Schulz <klsschlz@gmail.com>
Date: Thu, 16 May 2019 13:35:32 +0200
Subject: [PATCH 501/703] sound: pcm512x-codec: Adding 352.8kHz samplerate
 support

---
 sound/soc/codecs/pcm512x.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/sound/soc/codecs/pcm512x.c
+++ b/sound/soc/codecs/pcm512x.c
@@ -542,7 +542,7 @@ static unsigned long pcm512x_ncp_target(
 
 static const u32 pcm512x_dai_rates[] = {
 	8000, 11025, 16000, 22050, 32000, 44100, 48000, 64000,
-	88200, 96000, 176400, 192000, 384000,
+	88200, 96000, 176400, 192000, 352800, 384000,
 };
 
 static const struct snd_pcm_hw_constraint_list constraints_slave = {
