From e7bf2e5b1b86a11a0ccf59222cb31b06213d9276 Mon Sep 17 00:00:00 2001
From: Giedrius <giedrius@blokas.io>
Date: Fri, 12 Jul 2019 17:45:55 +0300
Subject: [PATCH 707/773] Pisound: Remove spinlock usage around spi_sync

---
 sound/soc/bcm/pisound.c | 5 -----
 1 file changed, 5 deletions(-)

--- a/sound/soc/bcm/pisound.c
+++ b/sound/soc/bcm/pisound.c
@@ -286,9 +286,6 @@ static irqreturn_t data_available_interr
 	return IRQ_HANDLED;
 }
 
-static DEFINE_SPINLOCK(spilock);
-static unsigned long spilockflags;
-
 static uint16_t spi_transfer16(uint16_t val)
 {
 	uint8_t txbuf[2];
@@ -333,9 +330,7 @@ static void spi_transfer(const uint8_t *
 	transfer.delay_usecs = 10;
 	spi_message_add_tail(&transfer, &msg);
 
-	spin_lock_irqsave(&spilock, spilockflags);
 	err = spi_sync(pisnd_spi_device, &msg);
-	spin_unlock_irqrestore(&spilock, spilockflags);
 
 	if (err < 0) {
 		printe("spi_sync error %d\n", err);
