From 4e11b14186707dfb88be344f227737eaf8adc608 Mon Sep 17 00:00:00 2001
From: Ezekiel Bethel <zek@9net.org>
Date: Wed, 12 Dec 2018 19:11:13 +0000
Subject: [PATCH 297/725] bcm2835_smi: re-add dereference to fix DMA transfers

---
 drivers/misc/bcm2835_smi.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/misc/bcm2835_smi.c
+++ b/drivers/misc/bcm2835_smi.c
@@ -879,7 +879,7 @@ static int bcm2835_smi_probe(struct plat
 		goto err;
 	}
 	addr = of_get_address(node, 0, NULL, NULL);
-	inst->smi_regs_busaddr = be32_to_cpu(addr);
+	inst->smi_regs_busaddr = be32_to_cpu(*addr);
 
 	err = bcm2835_smi_dma_setup(inst);
 	if (err)
