From f1b841cc75bcd8c655c9cd6c3dbed439a1f4d24a Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.org>
Date: Tue, 8 Nov 2016 21:35:38 +0000
Subject: [PATCH 013/725] spi-bcm2835: Remove unused code

---
 drivers/spi/spi-bcm2835.c | 61 ---------------------------------------
 1 file changed, 61 deletions(-)

--- a/drivers/spi/spi-bcm2835.c
+++ b/drivers/spi/spi-bcm2835.c
@@ -678,17 +678,8 @@ static void bcm2835_spi_set_cs(struct sp
 	bcm2835_wr(bs, BCM2835_SPI_CS, cs);
 }
 
-static int chip_match_name(struct gpio_chip *chip, void *data)
-{
-	return !strcmp(chip->label, data);
-}
-
 static int bcm2835_spi_setup(struct spi_device *spi)
 {
-	int err;
-	struct gpio_chip *chip;
-	struct device_node *pins;
-	u32 pingroup_index;
 	/*
 	 * sanity checking the native-chipselects
 	 */
@@ -706,58 +697,6 @@ static int bcm2835_spi_setup(struct spi_
 		return -EINVAL;
 	}
 
-#if 0
-	/* now translate native cs to GPIO */
-	/* first look for chip select pins in the devices pin groups */
-	for (pingroup_index = 0;
-	     (pins = of_parse_phandle(spi->master->dev.of_node,
-				     "pinctrl-0",
-				      pingroup_index)) != 0;
-	     pingroup_index++) {
-		u32 pin;
-		u32 pin_index;
-		for (pin_index = 0;
-		     of_property_read_u32_index(pins,
-						"brcm,pins",
-						pin_index,
-						&pin) == 0;
-		     pin_index++) {
-			if (((spi->chip_select == 0) &&
-			     ((pin == 8) || (pin == 36) || (pin == 46))) ||
-			    ((spi->chip_select == 1) &&
-			     ((pin == 7) || (pin == 35)))) {
-				spi->cs_gpio = pin;
-				break;
-			}
-		}
-		of_node_put(pins);
-	}
-	/* if that fails, assume GPIOs 7-11 are used */
-	if (!gpio_is_valid(spi->cs_gpio) ) {
-		/* get the gpio chip for the base */
-		chip = gpiochip_find("pinctrl-bcm2835", chip_match_name);
-		if (!chip)
-			return 0;
-
-		/* and calculate the real CS */
-		spi->cs_gpio = chip->base + 8 - spi->chip_select;
-	}
-
-	/* and set up the "mode" and level */
-	dev_info(&spi->dev, "setting up native-CS%i as GPIO %i\n",
-		 spi->chip_select, spi->cs_gpio);
-
-	/* set up GPIO as output and pull to the correct level */
-	err = gpio_direction_output(spi->cs_gpio,
-				    (spi->mode & SPI_CS_HIGH) ? 0 : 1);
-	if (err) {
-		dev_err(&spi->dev,
-			"could not set CS%i gpio %i as output: %i",
-			spi->chip_select, spi->cs_gpio, err);
-		return err;
-	}
-#endif
-
 	return 0;
 }
 
