--- a/Makefile
+++ b/Makefile
@@ -157,11 +157,7 @@ wipe.o	:	wipe.c random.h misc.h version.
 		$(CC) $(CCO) $(CCOC) wipe.c -o wipe.o
 
 version.h: always
-		if which git >/dev/null 2>&1 ; then \
-			git rev-list --max-count=1 HEAD | sed -e 's/^/#define WIPE_GIT "/' -e 's/$$/"/' >version.h ; \
-	  else \
-			echo '#define WIPE_GIT "(unknown, compiled without git)"' >version.h ; \
-	  fi
+		echo '#define WIPE_GIT "(unknown, compiled without git)"' >version.h ; \
 
 random.o	:	random.c misc.h md5.h
 		$(CC) $(CCO) $(CCOC) random.c -o random.o
--- a/wipe.c
+++ b/wipe.c
@@ -24,7 +24,7 @@
 
 /*** defines */
 
-#define WIPE_VERSION "0.23"
+#define WIPE_VERSION "0.24"
 #define WIPE_DATE "2016-11-03"
 
 /* exit codes */
