--- a/src/mklibs
+++ b/src/mklibs
@@ -616,7 +616,7 @@ while 1:
             # compile in only used symbols
             cmd = []
             cmd.extend(gcc_options)
-            cmd.append("-nostdlib -nostartfiles -shared -Wl,-soname=" + soname)
+            cmd.append("-nostdlib -nostartfiles -shared -Wl,--gc-sections -Wl,-soname=" + soname)
             cmd.extend(["-u%s" % a.linker_name() for a in symbols])
             cmd.extend(["-o", dest_path + "/" + so_file_name + "-so"])
             cmd.extend(extra_pre_obj)
