--- a/src/mklibs-readelf/main.cpp
+++ b/src/mklibs-readelf/main.cpp
@@ -84,7 +84,7 @@ static void process_symbols_provided (co
       continue;
     if (shndx == SHN_UNDEF || shndx == SHN_ABS)
       continue;
-    if (type != STT_NOTYPE && type != STT_OBJECT && type != STT_FUNC && type != STT_GNU_IFUNC && type != STT_COMMON && type != STT_TLS)
+    if (type != STT_NOTYPE && type != STT_OBJECT && type != STT_FUNC && type != STT_COMMON && type != STT_TLS)
       continue;
     if (!name.size())
       continue;
@@ -115,7 +115,7 @@ static void process_symbols_undefined (c
       continue;
     if (shndx != SHN_UNDEF)
       continue;
-    if (type != STT_NOTYPE && type != STT_OBJECT && type != STT_FUNC && type != STT_GNU_IFUNC && type != STT_COMMON && type != STT_TLS)
+    if (type != STT_NOTYPE && type != STT_OBJECT && type != STT_FUNC && type != STT_COMMON && type != STT_TLS)
       continue;
     if (!name.size())
       continue;
