--- a/src/lex_config.l
+++ b/src/lex_config.l
@@ -1,6 +1,6 @@
 /* Special state for handling include files */
 %x src
-%option noinput nounput
+%option noinput nounput noyywrap
 
 %{
 /*
@@ -77,10 +77,6 @@ module		/* skip */ ;
 
 %%
 
-#ifndef yywrap
-int yywrap() { return 1; }
-#endif
-
 /*======================================================================
 
     Stuff to parse basic data types
