From 6a07ec52fd6e8f595550f0b2eef63b2d5ae9d100 Mon Sep 17 00:00:00 2001
From: jjm2473 <1129525450@qq.com>
Date: Thu, 9 Feb 2023 14:07:13 +0800
Subject: [PATCH] drivers/net/stmmac: dwmac-rk set mac delay on phy attached

---
 .../net/ethernet/stmicro/stmmac/dwmac-rk.c    | 40 +++++++++++++++++++
 1 file changed, 40 insertions(+)

--- a/drivers/net/ethernet/stmicro/stmmac/dwmac-rk.c
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac-rk.c
@@ -1207,6 +1207,43 @@ static int phy_power_on(struct rk_priv_d
 	return 0;
 }
 
+static void rk_fix_phy(void *priv, struct phy_device *phy)
+{
+	int ret;
+	int phy_id;
+	int tx_delay;
+	int rx_delay;
+	struct device_node *child;
+	struct device_node *delay_cfg;
+	struct rk_priv_data *bsp_priv = priv;
+	struct device *dev = &bsp_priv->pdev->dev;
+
+	delay_cfg = of_get_child_by_name(dev->of_node, "delay_cfg");
+	if (!delay_cfg) {
+		return;
+	}
+	for_each_available_child_of_node(delay_cfg, child) {
+		ret = of_property_read_u32(child, "phy_id", &phy_id);
+		if (ret)
+			continue;
+		if (phy_id != phy->phy_id)
+			continue;
+		ret = of_property_read_u32(child, "tx_delay", &tx_delay);
+		if (ret)
+			break;
+		ret = of_property_read_u32(child, "rx_delay", &rx_delay);
+		if (ret)
+			break;
+
+		dev_info(dev, "Set MAC tx_delay: 0x%x rx_delay: 0x%x for PHY 0x%08x\n", tx_delay, rx_delay, phy->phy_id);
+		bsp_priv->ops->set_to_rgmii(bsp_priv, tx_delay,
+					    rx_delay);
+		bsp_priv->tx_delay = tx_delay;
+		bsp_priv->rx_delay = rx_delay;
+		break;
+	}
+}
+
 static struct rk_priv_data *rk_gmac_setup(struct platform_device *pdev,
 					  struct plat_stmmacenet_data *plat,
 					  const struct rk_gmac_ops *ops)
@@ -1266,6 +1303,9 @@ static struct rk_priv_data *rk_gmac_setu
 		bsp_priv->rx_delay = value;
 	}
 
+	if (bsp_priv->phy_iface == PHY_INTERFACE_MODE_RGMII && of_get_child_by_name(dev->of_node, "delay_cfg"))
+		plat->fix_phy = rk_fix_phy;
+
 	bsp_priv->grf = syscon_regmap_lookup_by_phandle(dev->of_node,
 							"rockchip,grf");
 
