From 5adedeeb7618f75967674c529379d542cdf49c19 Mon Sep 17 00:00:00 2001
From: jjm2473 <1129525450@qq.com>
Date: Thu, 11 May 2023 19:05:38 +0800
Subject: [PATCH 35/42] drivers: add rockchip video

---
 drivers/video/Kconfig  | 4 ++++
 drivers/video/Makefile | 1 +
 2 files changed, 5 insertions(+)

--- a/drivers/video/Kconfig
+++ b/drivers/video/Kconfig
@@ -25,6 +25,10 @@ endmenu
 
 source "drivers/video/backlight/Kconfig"
 
+menu "Rockchip Misc Video driver"
+source "drivers/video/rockchip/Kconfig"
+endmenu
+
 config VGASTATE
        tristate
        default n
--- a/drivers/video/Makefile
+++ b/drivers/video/Makefile
@@ -6,6 +6,7 @@ obj-$(CONFIG_VT)		  += console/
 obj-$(CONFIG_FB_STI)		  += console/
 obj-$(CONFIG_LOGO)		  += logo/
 obj-y				  += backlight/
+obj-y				  += rockchip/
 
 obj-y				  += fbdev/
 
