From a72a0085aeac23ac8aa76c90c476e20626046ea7 Mon Sep 17 00:00:00 2001
From: Phil Howard <github@gadgetoid.com>
Date: Tue, 12 May 2026 18:36:13 +0100
Subject: [PATCH] dtoverlays: hyperpixel2r: Document PWM backlight support.

Signed-off-by: Phil Howard <github@gadgetoid.com>
---
 arch/arm/boot/dts/overlays/README | 1 +
 1 file changed, 1 insertion(+)

--- a/arch/arm/boot/dts/overlays/README
+++ b/arch/arm/boot/dts/overlays/README
@@ -5722,6 +5722,7 @@ Params: disable-touch           Disables
         touchscreen-inverted-y  Inverts Y direction of touch controller
         touchscreen-swapped-x-y Swaps X & Y axes of touch controller
         rotate                  Display rotation {0,90,180,270} (default 0)
+        backlight-pwm           Use PWM instead of GPIO for backlight
 
 
 Name:   vc4-kms-dpi-hyperpixel4
