From bb81a9b70e0a34350dd1b1d94e61e8368539344d Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.com>
Date: Wed, 22 Jul 2026 10:28:29 +0100
Subject: [PATCH] fixup! misc: rp1-pio: Reverse christmas tree declarations

Signed-off-by: Phil Elwell <phil@raspberrypi.com>
---
 drivers/misc/rp1-pio.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/drivers/misc/rp1-pio.c
+++ b/drivers/misc/rp1-pio.c
@@ -614,7 +614,8 @@ static int rp1_pio_sm_config_xfer_intern
 	struct dma_slave_config config = {};
 	struct device *dev = &pdev->dev;
 	struct dma_slave_caps dma_caps;
-	struct dma_info *dma = NULL; 
+	struct dma_info *dma = NULL;
+	bool reconfigure = false;
 	phys_addr_t fifo_addr;
 	uint32_t dma_mask;
 	char chan_name[4];
