--- a/gdb/configure
+++ b/gdb/configure
@@ -16818,6 +16818,7 @@ else
   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 #include <stdlib.h>
+#include <string.h>
   #include "bfd.h"
   #include "elf-bfd.h"
 
--- a/gdb/acinclude.m4
+++ b/gdb/acinclude.m4
@@ -362,6 +362,7 @@ AC_DEFUN([GDB_AC_CHECK_BFD], [
   AC_CACHE_CHECK([$1], [$2],
   [AC_TRY_LINK(
   [#include <stdlib.h>
+  #include <string.h>
   #include "bfd.h"
   #include "$4"
   ],
