--- a/gdb/gdbserver/configure
+++ b/gdb/gdbserver/configure
@@ -2508,7 +2508,7 @@ $as_echo "$as_me: error: \`$ac_var' was
       ac_cache_corrupted=: ;;
     ,);;
     *)
-      if test "x$ac_old_val" != "x$ac_new_val"; then
+      if test "`echo x$ac_old_val`" != "`echo x$ac_new_val`"; then
 	# differences in whitespace do not lead to failure.
 	ac_old_val_w=`echo x $ac_old_val`
 	ac_new_val_w=`echo x $ac_new_val`
