--- a/arch/mips/rb532/prom.c
+++ b/arch/mips/rb532/prom.c
@@ -67,6 +67,7 @@ static inline unsigned long tag2ul(char
 	return simple_strtoul(num, 0, 10);
 }
 
+extern char __image_cmdline[];
 void __init prom_setup_cmdline(void)
 {
 	static char cmd_line[COMMAND_LINE_SIZE] __initdata;
@@ -108,6 +109,9 @@ void __init prom_setup_cmdline(void)
 	}
 	*(cp++) = ' ';
 
+	strcpy(cp,(__image_cmdline));
+	cp += strlen(__image_cmdline);
+
 	i = strlen(arcs_cmdline);
 	if (i > 0) {
 		*(cp++) = ' ';
