--- a/arch/mips/rb532/devices.c
+++ b/arch/mips/rb532/devices.c
@@ -199,11 +199,11 @@ static struct platform_device nand_slot0
 
 static struct mtd_partition rb532_partition_info[] = {
 	{
-		.name = "Routerboard NAND boot",
+		.name = "kernel",
 		.offset = 0,
-		.size = 4 * 1024 * 1024,
+		.size = 8 * 1024 * 1024,
 	}, {
-		.name = "rootfs",
+		.name = "ubi",
 		.offset = MTDPART_OFS_NXTBLK,
 		.size = MTDPART_SIZ_FULL,
 	}
