--- a/drivers/gpu/drm/i915/gt/uc/intel_uc.c
+++ b/drivers/gpu/drm/i915/gt/uc/intel_uc.c
@@ -26,10 +26,14 @@ static void uc_expand_default_options(st
 		return;
 
 	/* Don't enable GuC/HuC on pre-Gen12 */
-	if (GRAPHICS_VER(i915) < 12) {
+	if (GRAPHICS_VER(i915) < 9) {
 		i915->params.enable_guc = 0;
 		return;
 	}
+	if(GRAPHICS_VER(i915) < 11){
+		i915->params.enable_guc = 2;
+		return;
+	}
 
 	/* Don't enable GuC/HuC on older Gen12 platforms */
 	if (IS_TIGERLAKE(i915) || IS_ROCKETLAKE(i915)) {
