--- a/CPP/7zip/Common/FileStreams.h
+++ b/CPP/7zip/Common/FileStreams.h
@@ -72,6 +72,7 @@ class COutFileStream:
   public IOutStream,
   public CMyUnknownImp
 {
+public:
   #ifdef USE_WIN_FILE
   NWindows::NFile::NIO::COutFile File;
   #else
--- a/CPP/Common/MyWindows.h
+++ b/CPP/Common/MyWindows.h
@@ -101,8 +101,11 @@ typedef LONG SCODE;
 
 #ifdef __cplusplus
 
+#ifndef INITGUID
+#define INITGUID
 DEFINE_GUID(IID_IUnknown,
 0x00000000, 0x0000, 0x0000, 0xC0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x46);
+#endif
 struct IUnknown
 {
   STDMETHOD(QueryInterface) (REFIID iid, void **outObject) PURE;
