--- a/CPP/7zip/Compress/LZMA_Alone/LzmaAlone.cpp
+++ b/CPP/7zip/Compress/LZMA_Alone/LzmaAlone.cpp
@@ -101,6 +101,8 @@ static const int kNumSwitches = sizeof(k
 
 static void PrintHelp()
 {
+  fprintf(stderr, "\nLZMA " MY_VERSION_COPYRIGHT_DATE "\n");
+
   fprintf(stderr, "\nUsage:  LZMA <e|d> inputFile outputFile [<switches>...]\n"
              "  e: encode file\n"
              "  d: decode file\n"
@@ -168,8 +170,6 @@ int main2(int n, const char *args[])
   g_IsNT = IsItWindowsNT();
   #endif
 
-  fprintf(stderr, "\nLZMA " MY_VERSION_COPYRIGHT_DATE "\n");
-
   if (n == 1)
   {
     PrintHelp();
