--- a/gcc/gcc.c
+++ b/gcc/gcc.c
@@ -8828,8 +8828,10 @@ getenv_spec_function (int argc, const ch
 
   value = getenv (argv[0]);
   if (!value)
-    fatal_error (input_location,
-		 "environment variable %qs not defined", argv[0]);
+    {
+      warning (input_location, "environment variable %qs not defined", argv[0]);
+      value = "";
+    }
 
   /* We have to escape every character of the environment variable so
      they are not interpreted as active spec characters.  A
