--- a/setup.cfg
+++ b/setup.cfg
@@ -98,7 +98,6 @@ console_scripts =
 	conch = twisted.conch.scripts.conch:run
 	mailmail = twisted.mail.scripts.mailmail:run
 	pyhtmlizer = twisted.scripts.htmlizer:run
-	tkconch = twisted.conch.scripts.tkconch:run
 	trial = twisted.scripts.trial:run
 	twist = twisted.application.twist._twist:Twist.main
 	twistd = twisted.scripts.twistd:run
--- a/src/twisted/python/twisted-completion.zsh
+++ b/src/twisted/python/twisted-completion.zsh
@@ -1,4 +1,4 @@
-#compdef twist twistd trial conch cftp ckeygen pyhtmlizer tkconch
+#compdef twist twistd trial conch cftp ckeygen pyhtmlizer
 #
 # This is the ZSH completion file for Twisted commands. It calls the current
 # command-line with the special "--_shell-completion" option which is handled
