diff --git a/libfstools/rootdisk.c b/libfstools/rootdisk.c
index dd00c1b..f3b87fc 100644
--- a/libfstools/rootdisk.c
+++ b/libfstools/rootdisk.c
@@ -270,7 +270,7 @@ static int rootdisk_volume_init(struct volume *v)
 		if (rootdisk_use_f2fs(p))
 			snprintf(str, sizeof(str), "mkfs.f2fs -q -l rootfs_data %s", v->blk);
 		else
-			snprintf(str, sizeof(str), "mkfs.ext4 -q -L rootfs_data %s", v->blk);
+			snprintf(str, sizeof(str), "mkfs.ext4 -q -E lazy_itable_init=0,lazy_journal_init=0 -L rootfs_data %s", v->blk);
 		ret = system(str);
 		break;
 	default:
