--- a/block.c
+++ b/block.c
@@ -530,7 +530,7 @@
 		printf("\toption\tuuid\t'%s'\n", pr->uuid);
 	else
 		printf("\toption\tdevice\t'%s'\n", pr->dev);
-	printf("\toption\tenabled\t'0'\n\n");
+	printf("\toption\tenabled\t'1'\n\n");
 
 	return 0;
 }
@@ -1454,7 +1454,7 @@
 	cache_load(0);
 	printf("config 'global'\n");
 	printf("\toption\tanon_swap\t'0'\n");
-	printf("\toption\tanon_mount\t'0'\n");
+	printf("\toption\tanon_mount\t'1'\n");
 	printf("\toption\tauto_swap\t'1'\n");
 	printf("\toption\tauto_mount\t'1'\n");
 	printf("\toption\tdelay_root\t'5'\n");
