Description: Fix FTBFS on armhf by correctly detecting the architecture
Origin: vendor
Bug-Debian: http://bugs.debian.org/730844
Author: Alessandro Ghedini <ghedo@debian.org>
Last-Update: 2013-11-30

--- a/configure.ac
+++ b/configure.ac
@@ -275,7 +275,7 @@ case "${host_cpu}" in
         ARCH_MAX="s390x"
         ;;
 
-     armv7*)
+     arm*)
 	AC_MSG_RESULT([ok (${host_cpu})])
 	ARCH_MAX="arm"
 	;;
