--- a/Makefile
+++ b/Makefile
@@ -32,6 +32,7 @@ SCRIPTPATH = /etc/comgt/
 SCRIPTSRC = ./scripts/
 BIN     = $(CPROG) 
 MANP	= comgt.1 sigmon.1
+CC	= cc
 
 CFLAGS  = -c
 LDFLAGS =
@@ -70,10 +71,5 @@ clean:
 	-rm *~
 	-rm $(SCRIPTSRC)*~
 
-
-comgt: comgt.o
-	cc comgt.o $(LDFLAGS) -o comgt
-
-comgt.o: comgt.c comgt.h
-	cc comgt.c $(CFLAGS) 
-
+comgt:  comgt.c comgt.h
+	$(CC) $(CFLAGS) -o comgt $< $(LDFLAGS)
