--- a/comgt.c
+++ b/comgt.c
@@ -1281,7 +1281,7 @@ void opengt(void) {
   parity=stbuf.c_cflag & (PARENB | PARODD);
   stbuf.c_iflag &= ~(IGNCR | ICRNL | IUCLC | INPCK | IXON | IXANY | IGNPAR );
   stbuf.c_oflag &= ~(OPOST | OLCUC | OCRNL | ONLCR | ONLRET);
-  stbuf.c_lflag &= ~(ICANON | XCASE | ECHO | ECHOE | ECHONL);
+  stbuf.c_lflag &= ~(ICANON | ECHO | ECHOE | ECHONL);
   stbuf.c_lflag &= ~(ECHO | ECHOE);
   stbuf.c_cc[VMIN] = 1;
   stbuf.c_cc[VTIME] = 0;
@@ -1336,7 +1336,7 @@ void opendevice(void) {
   parity=stbuf.c_cflag & (PARENB | PARODD);
   stbuf.c_iflag &= ~(IGNCR | ICRNL | IUCLC | INPCK | IXON | IXANY | IGNPAR );
   stbuf.c_oflag &= ~(OPOST | OLCUC | OCRNL | ONLCR | ONLRET);
-  stbuf.c_lflag &= ~(ICANON | XCASE | ECHO | ECHOE | ECHONL);
+  stbuf.c_lflag &= ~(ICANON | ECHO | ECHOE | ECHONL);
   stbuf.c_lflag &= ~(ECHO | ECHOE);
   stbuf.c_cc[VMIN] = 1;
   stbuf.c_cc[VTIME] = 0;
