--- a/src/Cedar/Server.c
+++ b/src/Cedar/Server.c
@@ -4,7 +4,7 @@
 
 // Server.c
 // VPN Server module
-
+#include <syslog.h>
 #include "CedarPch.h"
 
 static SERVER *server = NULL;
@@ -6492,6 +6492,10 @@ UINT SiWriteConfigurationFile(SERVER *s)
 	{
 		return 0;
 	}
+	
+	openlog("softethervpn-server", LOG_CONS | LOG_PID | LOG_NDELAY, LOG_DAEMON);
+	syslog(LOG_NOTICE, "Writing vpn_server.config (AutoSaveConfigSpan set to < %d > seconds)", (s->AutoSaveConfigSpan / 1000));
+	closelog();
 
 	Lock(s->SaveCfgLock);
 	{
