--- a/src/nettest_omni.c
+++ b/src/nettest_omni.c
@@ -454,16 +454,6 @@ static  int remote_cpu_method;
 static int client_port_min = 5000;
 static int client_port_max = 65535;
 
- /* different options for the sockets				*/
-
-int
-  loc_nodelay,		/* don't/do use NODELAY	locally		*/
-  rem_nodelay,		/* don't/do use NODELAY remotely	*/
-  loc_sndavoid,		/* avoid send copies locally		*/
-  loc_rcvavoid,		/* avoid recv copies locally		*/
-  rem_sndavoid,		/* avoid send copies remotely		*/
-  rem_rcvavoid; 	/* avoid recv_copies remotely		*/
-
 extern int
   loc_tcpcork,
   rem_tcpcork,
