--- a/Makefile.am
+++ b/Makefile.am
@@ -20,6 +20,6 @@
 # LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 # OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 # WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
-SUBDIRS = lib tests examples doc
+SUBDIRS = lib tests
 
 ACLOCAL_AMFLAGS = -I m4
--- a/configure.ac
+++ b/configure.ac
@@ -123,9 +123,6 @@ AC_CONFIG_FILES([
   lib/includes/Makefile
   lib/includes/wslay/wslayver.h
   tests/Makefile
-  examples/Makefile
-  doc/Makefile
-  doc/sphinx/conf.py
 ])
 AC_OUTPUT
 
