+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_mdio.c
@@ -498,6 +498,11 @@ int stmmac_mdio_register(struct net_devi
 	if (priv->plat->has_xgmac)
 		stmmac_xgmac2_mdio_read(new_bus, 0, MII_ADDR_C45);
 
+		stmmac_mdio_write(new_bus,0,31,2627);
+		stmmac_mdio_write(new_bus,0,25,0x1801);
+		stmmac_mdio_write(new_bus,0,31,0);
+		stmmac_mdio_write(new_bus,0,0,0x8000);
+
 	if (priv->plat->phy_node || mdio_node)
 		goto bus_register_done;
 
--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_mdio.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_mdio.c
@@ -498,6 +498,16 @@ int stmmac_mdio_register(struct net_devi
 	if (priv->plat->has_xgmac)
 		stmmac_xgmac2_mdio_read(new_bus, 0, MII_ADDR_C45);
 
+		stmmac_mdio_write(new_bus,0,31,2627);
+		stmmac_mdio_write(new_bus,0,25,0x1801);
+		stmmac_mdio_write(new_bus,0,31,0);
+		stmmac_mdio_write(new_bus,0,0,0x8000);
+
+		stmmac_mdio_write(new_bus,0,31,2627);
+		stmmac_mdio_write(new_bus,0,25,0x1801);
+		stmmac_mdio_write(new_bus,0,31,0);
+		stmmac_mdio_write(new_bus,0,0,0x8000);
+
 	if (priv->plat->phy_node || mdio_node)
 		goto bus_register_done;
 
--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
@@ -2852,6 +2852,8 @@ static int stmmac_init_dma_engine(struct
 	if (priv->extend_desc && (priv->mode == STMMAC_RING_MODE))
 		atds = 1;
 
+	msleep(1500);
+
 	ret = stmmac_reset(priv, priv->ioaddr);
 	if (ret) {
 		dev_err(priv->device, "Failed to reset the dma\n");
