--- a/conf_options.c
+++ b/conf_options.c
@@ -519,7 +519,7 @@ void config_parse_file_and_cmdline(int a
 			conf_filename = optarg;
 			break;
 		case 'v':
-			printf("%s (build date: %s %s)\n", VERSION, __DATE__, __TIME__);
+			printf("%s\n", VERSION);
 			exit(0);
 		case 'h':
 		case '?':
--- a/display-help.c
+++ b/display-help.c
@@ -36,7 +36,7 @@ void update_help_win(WINDOW *win)
 	print_centered(win, 2, COLS, "HORST - Horsts OLSR Radio Scanning Tool (or)");
 	print_centered(win, 3, COLS, "HORST - Highly Optimized Radio Scanning Tool");
 
-	print_centered(win, 5, COLS, "Version " VERSION " (build date " __DATE__ " " __TIME__ ")");
+	print_centered(win, 5, COLS, "Version " VERSION);
 	print_centered(win, 6, COLS, "(C) 2005-2016 Bruno Randolf, Licensed under the GPLv2");
 
 	mvwprintw(win, 8, 2, "Known IEEE802.11 Packet Types:");
