--- a/core/Makefile
+++ b/core/Makefile
@@ -3,7 +3,7 @@ DESTDIR ?=
 LIBDIR ?= /usr/lib
 
 #IRSSI_INCLUDE:=/home/phh/irssi-0.8.16-rc1.phh/
-IRSSI_INCLUDE:=/usr/include/irssi/
+IRSSI_INCLUDE?=/usr/include/irssi/
 IRSSI_LIB?=$(DESTDIR)/$(LIBDIR)/irssi
 IRSSI_CFLAGS+=-I$(IRSSI_INCLUDE)/src/
 IRSSI_CFLAGS+=-I$(IRSSI_INCLUDE)/src/core/
@@ -28,7 +28,7 @@ else
     LDFLAGS += -lquasselc
 endif
 
-CFLAGS=-std=gnu11 -Wall -Wextra -Werror -g -O2 $(IRSSI_CFLAGS) $(QUASSELC_FLAGS) -Wmissing-prototypes -Wmissing-declarations
+CFLAGS+=-std=gnu11 -Wall -Wextra -g $(IRSSI_CFLAGS) $(QUASSELC_FLAGS) -Wmissing-prototypes -Wmissing-declarations
 
 CFLAGS += $(SSL_CFLAGS)
 LDFLAGS+= $(SSL_LDLAGS)
