From 74be1b50f96b3ada0a4fc18f0283659739edbadf Mon Sep 17 00:00:00 2001
From: Yangbo Lu <yangbo.lu@nxp.com>
Date: Fri, 22 Sep 2017 11:00:12 +0800
Subject: [PATCH 1/3] armv8: ls1043a: add LEDE boot support in environment

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>
---
 include/configs/ls1043a_common.h | 20 +++++++++++++++++++-
 1 file changed, 19 insertions(+), 1 deletion(-)

diff --git a/include/configs/ls1043a_common.h b/include/configs/ls1043a_common.h
index 0b3fb4e..5c01111 100644
--- a/include/configs/ls1043a_common.h
+++ b/include/configs/ls1043a_common.h
@@ -336,7 +336,19 @@
 	"sd_bootcmd=echo Trying load from SD ..;"   \
 		"mmcinfo; mmc read $load_addr "	    \
 		"$kernel_addr_sd $kernel_size_sd ;" \
-		" bootm $load_addr#$board\0"
+		" bootm $load_addr#$board\0" \
+	"lede_setenv=setenv loadaddr 82000000 && " \
+		"setenv fdtaddr 8f000000 && " \
+		"setenv bootargs root=/dev/mtdblock8 " \
+		"rootfstype=squashfs,jffs2 noinitrd " \
+		"earlycon=uart8250,mmio,0x21c0500 console=ttyS0,115200 " \
+		"mtdparts=60000000.nor:1M(rcw),2M(u-boot),1M(u-boot-env)," \
+		"5M(reserved-1),256k(fman),5888k(reserved-2),1M(dtb)," \
+		"16M(kernel),32M(rootfs),64M(otherbank)\0" \
+	"lede_run=cp.b 60f00000 $fdtaddr 100000 && " \
+		"cp.b 61000000 $loadaddr 1000000 && " \
+		"bootm $loadaddr - $fdtaddr\0" \
+	"lede_boot=run lede_setenv;run lede_run\0"
 
 
 #undef CONFIG_BOOTCOMMAND
@@ -351,6 +363,12 @@
 			   "env exists secureboot && esbc_halt;"
 #endif
 
+#undef CONFIG_BOOTCOMMAND
+#define CONFIG_BOOTCOMMAND "run lede_boot"
+
+#undef CONFIG_BOOTDELAY
+#define CONFIG_BOOTDELAY		3
+
 #define CONFIG_BOOTARGS			"console=ttyS0,115200 root=/dev/ram0 " \
 					"earlycon=uart8250,mmio,0x21c0500 "    \
 					MTDPARTS_DEFAULT
-- 
2.7.4

