From 8d6efca7eb2de012b98493206c87f31ee9b9e967 Mon Sep 17 00:00:00 2001
From: Yangbo Lu <yangbo.lu@nxp.com>
Date: Thu, 28 Sep 2017 11:25:58 +0800
Subject: [PATCH 3/3] armv8: ls1012ardb: add LEDE boot support in environment

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>
---
 include/configs/ls1012ardb.h | 21 ++++++++++++++++++++-
 1 file changed, 20 insertions(+), 1 deletion(-)

diff --git a/include/configs/ls1012ardb.h b/include/configs/ls1012ardb.h
index 5fe3218..242cdf0 100644
--- a/include/configs/ls1012ardb.h
+++ b/include/configs/ls1012ardb.h
@@ -127,13 +127,32 @@
 		"$kernel_addr $kernel_size; env exists secureboot "	\
 		"&& sf read $kernelheader_addr_r $kernelheader_addr "	\
 		"$kernelheader_size && esbc_validate ${kernelheader_addr_r}; " \
-		"bootm $load_addr#$board\0"
+		"bootm $load_addr#$board\0" \
+	"lede_setenv=setenv loadaddr 82000000 && " \
+		"setenv fdtaddr 8f000000 && " \
+		"setenv bootargs root=/dev/mtdblock8 " \
+		"rootfstype=ext4 noinitrd " \
+		"earlycon=uart8250,mmio,0x21c0500 console=ttyS0,115200 " \
+		"mtdparts=1550000.quadspi:1M(rcw),2M(u-boot),1M(u-boot-env)," \
+		"5M(reserved-1),256k(fman),5888k(reserved-2),1M(dtb)," \
+		"16M(kernel),30M(ext4rfs),2M(user)\0" \
+	"lede_run=pfe stop && sf probe 0:0 && " \
+		"sf read $fdtaddr f00000 100000 && " \
+		"sf read $loadaddr 1000000 1000000 && " \
+		"bootm $loadaddr - $fdtaddr\0" \
+	"lede_boot=run lede_setenv;run lede_run\0"
 
 #undef CONFIG_BOOTCOMMAND
 #if defined(CONFIG_QSPI_BOOT) || defined(CONFIG_SD_BOOT_QSPI)
 #define CONFIG_BOOTCOMMAND "run distro_bootcmd; run qspi_bootcmd; " \
 			   "env exists secureboot && esbc_halt;"
 #endif
+#undef CONFIG_BOOTCOMMAND
+#define CONFIG_BOOTCOMMAND		"run lede_boot"
+
+#undef CONFIG_BOOTDELAY
+#define CONFIG_BOOTDELAY		3
+
 
 #include <asm/fsl_secure_boot.h>
 
-- 
2.7.4

