From a5e454a3bb33a456655f08592606e4df34e40110 Mon Sep 17 00:00:00 2001
From: Yangbo Lu <yangbo.lu@nxp.com>
Date: Tue, 24 Oct 2017 17:48:08 +0800
Subject: [PATCH] armv8: ls1012afrdm: add LEDE boot support in environment

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>
---
 include/configs/ls1012afrdm.h | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/include/configs/ls1012afrdm.h b/include/configs/ls1012afrdm.h
index a3f8824d27..36dff19bdd 100644
--- a/include/configs/ls1012afrdm.h
+++ b/include/configs/ls1012afrdm.h
@@ -73,12 +73,31 @@
 	"qspi_bootcmd=echo Trying load from qspi..;"	\
 		"sf probe && sf read $load_addr "	\
 		"$kernel_addr $kernel_size && bootm $load_addr#$board\0" \
+	"lede_setenv=setenv loadaddr 82000000 && " \
+		"setenv fdtaddr 8f000000 && " \
+		"setenv bootargs ubi.mtd=8 root=ubi0:rootfs rw " \
+		"rootfstype=ubifs noinitrd " \
+		"earlycon=uart8250,mmio,0x21c0500 console=ttyS0,115200 " \
+		"mtdparts=1550000.quadspi:1M(rcw),2M(u-boot),1M(u-boot-env)," \
+		"5M(reserved-1),256k(fman),5888k(reserved-2),1M(dtb)," \
+		"16M(kernel),32M(ubifs)\0" \
+	"lede_run=pfe stop && sf probe 0:0 && " \
+		"sf read $fdtaddr f00000 100000 && " \
+		"sf read $loadaddr 1000000 1000000 && " \
+		"bootm $loadaddr - $fdtaddr\0" \
+	"lede_boot=run lede_setenv;run lede_run\0"
 
 #undef CONFIG_BOOTCOMMAND
 #if defined(CONFIG_QSPI_BOOT) || defined(CONFIG_SD_BOOT_QSPI)
 #define CONFIG_BOOTCOMMAND "run distro_bootcmd;run qspi_bootcmd"
 #endif
 
+#undef CONFIG_BOOTCOMMAND
+#define CONFIG_BOOTCOMMAND		"run lede_boot"
+
+#undef CONFIG_BOOTDELAY
+#define CONFIG_BOOTDELAY		3
+
 /*
 * USB
 */
-- 
2.14.1

