--- a/arch/mips/boot/dts/img/pistachio_marduk.dts
+++ b/arch/mips/boot/dts/img/pistachio_marduk.dts
@@ -19,6 +19,11 @@
 		ethernet0 = &enet;
 		spi0 = &spfi0;
 		spi1 = &spfi1;
+
+		led-boot = &led_heartbeat;
+		led-failsafe = &led_heartbeat;
+		led-running = &led_heartbeat;
+		led-upgrade = &led_heartbeat;
 	};
 
 	chosen {
@@ -89,11 +94,10 @@
 	led-controller {
 		compatible = "pwm-leds";
 
-		led-1 {
+		led_heartbeat: heartbeat {
 			label = "marduk:red:heartbeat";
 			pwms = <&pwm 3 300000>;
 			max-brightness = <255>;
-			linux,default-trigger = "heartbeat";
 		};
 	};
 
