--- a/updates/ids_tls_update.c
+++ b/updates/ids_tls_update.c
@@ -68,11 +68,13 @@ setup_context(const char *hostname, int
     int rc;
     unsigned long ssl_err = 0;
 
+#if OPENSSL_API_COMPAT < 0x10100000L
     SSL_load_error_strings();
     SSL_library_init();
     OpenSSL_add_all_algorithms();
     ERR_load_BIO_strings();
     ERR_load_crypto_strings();
+#endif
 
 #ifdef HAVE_TLS_METHOD
     method = TLS_method();
