--- a/configure.ac
+++ b/configure.ac
@@ -23,7 +23,7 @@ AC_HEADER_SYS_WAIT
 AC_CHECK_HEADERS([arpa/inet.h errno.h malloc.h \
 		stddef.h stdlib.h string.h sys/socket.h \
 		sys/time.h syslog.h unistd.h time.h \
-		assert.h sys/poll.h signal.h pwd.h grp.h stdarg.h\
+		assert.h poll.h signal.h pwd.h grp.h stdarg.h\
 		limits.h sys/wait.h sched.h sys/ioctl.h sys/uio.h])
 AC_HEADER_TIME
 
--- a/src/apinger.c
+++ b/src/apinger.c
@@ -33,8 +33,8 @@
 #ifdef HAVE_SYS_WAIT_H
 # include <sys/wait.h>
 #endif
-#ifdef HAVE_SYS_POLL_H
-# include <sys/poll.h>
+#ifdef HAVE_POLL_H
+# include <poll.h>
 #endif
 #ifdef HAVE_ARPA_INET_H
 # include <arpa/inet.h>
