--- a/spamc/libspamc.c
+++ b/spamc/libspamc.c
@@ -1213,9 +1213,11 @@ int message_filter(struct transport *tp,
 
     if (flags & SPAMC_USE_SSL) {
 #ifdef SPAMC_SSL
+#if 0
 	SSLeay_add_ssl_algorithms();
-	meth = SSLv23_client_method();
 	SSL_load_error_strings();
+#endif
+	meth = SSLv23_client_method();
 	ctx = SSL_CTX_new(meth);
 #else
 	UNUSED_VARIABLE(ssl);
@@ -1600,9 +1602,11 @@ int message_tell(struct transport *tp, c
 
     if (flags & SPAMC_USE_SSL) {
 #ifdef SPAMC_SSL
+#if 0
 	SSLeay_add_ssl_algorithms();
-	meth = SSLv23_client_method();
 	SSL_load_error_strings();
+#endif
+	meth = SSLv23_client_method();
 	ctx = SSL_CTX_new(meth);
 #else
 	UNUSED_VARIABLE(ssl);
