--- a/Makefile.in
+++ b/Makefile.in
@@ -393,7 +393,7 @@ target_alias = @target_alias@
 top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
-SUBDIRS = m4m po libexif test doc binary-dist contrib
+SUBDIRS = m4m po libexif test binary-dist contrib
 EXTRA_DIST = @PACKAGE_TARNAME@.spec README-Win32.txt libexif.pc.in \
 	libexif-uninstalled.pc.in SECURITY.md
 pkgconfigdir = $(libdir)/pkgconfig
