--- a/misc/terminfo.src
+++ b/misc/terminfo.src
@@ -4984,6 +4984,7 @@ xterm-xfree86|xterm terminal emulator (X
 
 xterm+nofkeys|building block for xterm fkey-variants,
 	npc,
+	kbs=\177,
 	kcbt=\E[Z, kent=\EOM, nel=\EE, use=ecma+index,
 	use=ansi+rep, use=ecma+strikeout, use=vt420+lrmm,
 	use=xterm+sm+1006, use=xterm+tmux, use=ecma+italics,
@@ -6689,6 +6690,7 @@ mlterm-256color|mlterm 3.0 with xterm 25
 rxvt-basic|rxvt terminal base (X Window System),
 	OTbs, am, bce, eo, mir, msgr, xenl, xon, XT,
 	cols#80, it#8, lines#24,
+	kbs=\177,
 	acsc=``aaffggjjkkllmmnnooppqqrrssttuuvvwwxxyyzz{{||}}~~,
 	bel=^G, blink=\E[5m, bold=\E[1m, clear=\E[H\E[2J, cr=\r,
 	csr=\E[%i%p1%d;%p2%dr, cub=\E[%p1%dD, cub1=^H,
@@ -6698,7 +6700,7 @@ rxvt-basic|rxvt terminal base (X Window
 	enacs=\E(B\E)0, flash=\E[?5h$<100/>\E[?5l, home=\E[H,
 	ht=^I, hts=\EH, ich=\E[%p1%d@, il=\E[%p1%dL, il1=\E[L,
 	ind=\n, is1=\E[?47l\E=\E[?1l,
-	is2=\E[r\E[m\E[2J\E[H\E[?7h\E[?1;3;4;6l\E[4l, kbs=^H,
+	is2=\E[r\E[m\E[2J\E[H\E[?7h\E[?1;3;4;6l\E[4l,
 	kcbt=\E[Z, kmous=\E[M, rc=\E8, rev=\E[7m, ri=\EM, rmacs=^O,
 	rmir=\E[4l, rmkx=\E>, rmso=\E[27m, rmul=\E[24m,
 	rs1=\E>\E[1;3;4;5;6l\E[?7h\E[m\E[r\E[2J\E[H,
@@ -8347,6 +8349,7 @@ dumb-emacs-ansi|Emacs dumb terminal with
 screen-base|VT 100/ANSI X3.64 virtual terminal (base),
 	OTbs, OTpt, am, km, mir, msgr, xenl, G0,
 	colors#8, cols#80, it#8, lines#24, ncv@, pairs#64, U8#1,
+	kbs=\177,
 	acsc=++\,\,--..00``aaffgghhiijjkkllmmnnooppqqrrssttuuvvwwxxy
 	     yzz{{||}}~~,
 	bel=^G, blink=\E[5m, bold=\E[1m, cbt=\E[Z, civis=\E[?25l,
@@ -8358,7 +8361,7 @@ screen-base|VT 100/ANSI X3.64 virtual te
 	dl=\E[%p1%dM, dl1=\E[M, ed=\E[J, el=\E[K, el1=\E[1K,
 	enacs=\E(B\E)0, flash=\Eg, home=\E[H, hpa=\E[%i%p1%dG,
 	ht=^I, hts=\EH, ich=\E[%p1%d@, il=\E[%p1%dL, il1=\E[L,
-	ind=\n, indn=\E[%p1%dS, is2=\E)0, kbs=^H, kcbt=\E[Z,
+	ind=\n, indn=\E[%p1%dS, is2=\E)0, kcbt=\E[Z,
 	kcub1=\EOD, kcud1=\EOB, kcuf1=\EOC, kcuu1=\EOA, kf1=\EOP,
 	kf10=\E[21~, kf11=\E[23~, kf12=\E[24~, kf2=\EOQ, kf3=\EOR,
 	kf4=\EOS, kf5=\E[15~, kf6=\E[17~, kf7=\E[18~, kf8=\E[19~,
@@ -8500,6 +8503,7 @@ screen.xterm-r6|screen customized for X1
 # on Solaris because Sun's curses implementation gets confused.
 screen.teraterm|disable ncv in teraterm,
 	ncv#127,
+	kbs=^H,
 	acsc=+\020\,\021-\030.^Y0\333`\004a\261f\370g\361h\260i
 	     \316j\331k\277l\332m\300n\305o~p\304q\304r\304s_t\303u
 	     \264v\301w\302x\263y\363z\362{\343|\330}\234~\376,
