--- a/aclocal.m4
+++ b/aclocal.m4
@@ -454,7 +454,7 @@ AC_REQUIRE([CF_PROG_AR])
 
 AC_CACHE_CHECK(for options to update archives, cf_cv_ar_flags,[
 	cf_cv_ar_flags=unknown
-	for cf_ar_flags in -curvU -curv curv -crv crv -cqv cqv -rv rv
+	for cf_ar_flags in -curvD -curv curv -crv crv -cqv cqv -rv rv
 	do
 
 		# check if $ARFLAGS already contains this choice
--- a/configure
+++ b/configure
@@ -4751,7 +4751,7 @@ if test "${cf_cv_ar_flags+set}" = set; t
 else
 
 	cf_cv_ar_flags=unknown
-	for cf_ar_flags in -curvU -curv curv -crv crv -cqv cqv -rv rv
+	for cf_ar_flags in -curvD -curv curv -crv crv -cqv cqv -rv rv
 	do
 
 		# check if $ARFLAGS already contains this choice
