--- a/aclocal.m4
+++ b/aclocal.m4
@@ -505,7 +505,7 @@ AC_CACHE_CHECK(for options to update arc
 		;;
 	(*)
 		cf_cv_ar_flags=unknown
-		for cf_ar_flags in -curvU -curv curv -crv crv -cqv cqv -rv rv
+		for cf_ar_flags in -curvD -curv curv -crv crv -cqv cqv -rv rv
 		do
 
 			# check if $ARFLAGS already contains this choice
--- a/configure
+++ b/configure
@@ -5072,7 +5072,7 @@ else
 		;;
 	(*)
 		cf_cv_ar_flags=unknown
-		for cf_ar_flags in -curvU -curv curv -crv crv -cqv cqv -rv rv
+		for cf_ar_flags in -curvD -curv curv -crv crv -cqv cqv -rv rv
 		do
 
 			# check if $ARFLAGS already contains this choice
