--- a/drivers/thermal/Kconfig
+++ b/drivers/thermal/Kconfig
@@ -522,4 +522,11 @@ config LOONGSON2_THERMAL
 	  is higher than the high temperature threshold or lower than the low
 	  temperature threshold, the interrupt will occur.
 
+config REALTEK_THERMAL
+	tristate "Realtek RTL838x and RTL930x thermal sensor support"
+	depends on RTL838X || RTL839X || RTL930X || COMPILE_TEST
+	depends on THERMAL_OF
+	help
+	  Support thermal sensor in Realtek RTL838x, RTL839x and RTL930x SoCs
+
 endif
--- a/drivers/thermal/Makefile
+++ b/drivers/thermal/Makefile
@@ -64,3 +64,4 @@ obj-$(CONFIG_AMLOGIC_THERMAL)     += aml
 obj-$(CONFIG_SPRD_THERMAL)	+= sprd_thermal.o
 obj-$(CONFIG_KHADAS_MCU_FAN_THERMAL)	+= khadas_mcu_fan.o
 obj-$(CONFIG_LOONGSON2_THERMAL)	+= loongson2_thermal.o
+obj-$(CONFIG_REALTEK_THERMAL)	+= realtek-thermal.o
