--- a/configure.ac
+++ b/configure.ac
@@ -74,6 +74,21 @@ AM_SILENT_RULES
 AC_CANONICAL_HOST
 AB_INIT
 
+case "${host}" in
+    x86_64-openwrt-linux-gnu|i?86-openwrt-linux-gnu)
+        host=$(echo $host | sed 's/openwrt/pc/g')
+        ;;
+    arm-openwrt-linux-gnu|armeb-openwrt-linux-gnu)
+        host=arm-unknown-linux-gnueabi
+        ;;
+    mips64-openwrt-linux-gnu)
+        host=mips64el-unknown-linux-gnuabi64
+        ;;
+    *)
+        host=$(echo $host | sed 's/openwrt/unknown/g')
+        ;;
+esac
+
 # Checks for programs.
 AC_PROG_CC
 AM_PROG_CC_C_O
