--- a/ipvsadm-restore
+++ b/ipvsadm-restore
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/bin/sh
 #      ipvsadm-restore - Restore IPVS rules
 #
 #      A very simple wrapper to restore IPVS rules
@@ -11,6 +11,8 @@
 #      This file:
 #
 #      ChangeLog
+#      M. Mozzarelli       :        Amended to use /bin/sh for compatibility
+#                          :        with embedded systems using busybox
 #      Horms               :        Clear IPVS rules before adding from STDIN
 #      Horms               :        Filter out "^#"
 #
--- a/ipvsadm-save
+++ b/ipvsadm-save
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/bin/sh
 #      ipvsadm-save - Save IPVS rules
 #
 #      A very simple wrapper to save IPVS rules
@@ -12,6 +12,8 @@
 #
 #      ChangeLog
 #
+#    M. Mozzarelli          :  Amended to use /bin/sh for compatibility
+#                           :  with embedded systems using busybox
 #	 Wensong Zhang          :  Added the "-n" option and the help()
 #
 
