--- a/src/PCIE/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_main.c
+++ b/src/PCIE/driver_fw/driver/aic8800/aic8800_fdrv/rwnx_main.c
@@ -535,8 +535,8 @@ int wifi_fail = 0;
 module_param(testmode, int, 0660);
 module_param(adap_test, int, 0660);
 module_param(wifi_fail, int, 0660);
-char aic_fw_path[200]="/lib/firmware/aic8800D80";
-char aic_fw_path_8800d80x2[200]="/lib/firmware/aic8800D80X2";
+char aic_fw_path[200]=CONFIG_AIC_FW_PATH;
+char aic_fw_path_8800d80x2[200]=CONFIG_AIC_FW_PATH;
 u8 chip_sub_id = 0;
 u8 chip_mcu_id = 0;
 u8 chip_id = 0;
--- a/src/USB/driver_fw/drivers/aic8800/aic8800_fdrv/aicwf_compat_8800d80.c
+++ b/src/USB/driver_fw/drivers/aic8800/aic8800_fdrv/aicwf_compat_8800d80.c
@@ -39,7 +39,7 @@ int	rwnx_plat_userconfig_load_8800d80(st
     u32 *dst=NULL;
     char *filename = FW_USERCONFIG_NAME_8800D80;
 
-#ifndef ANDROID_PLATFORM
+#if 0
             sprintf(aic_fw_path, "%s/%s", aic_fw_path, "aic8800D80");
 #endif
 
--- a/src/USB/driver_fw/drivers/aic8800/aic8800_fdrv/aicwf_compat_8800d80x2.c
+++ b/src/USB/driver_fw/drivers/aic8800/aic8800_fdrv/aicwf_compat_8800d80x2.c
@@ -40,7 +40,7 @@ int	rwnx_plat_userconfig_load_8800d80x2(
     u32 *dst=NULL;
     char *filename = FW_USERCONFIG_NAME_8800D80X2;
 
-#ifndef ANDROID_PLATFORM
+#if 0
     sprintf(aic_fw_path, "%s/%s", aic_fw_path, "aic8800D80X2");
 #endif
 
--- a/src/USB/driver_fw/drivers/aic8800/aic8800_fdrv/rwnx_platform.c
+++ b/src/USB/driver_fw/drivers/aic8800/aic8800_fdrv/rwnx_platform.c
@@ -1668,7 +1668,7 @@ static int rwnx_plat_patch_load(struct r
 
     if(rwnx_hw->usbdev->chipid == PRODUCT_ID_AIC8800DC ||
         rwnx_hw->usbdev->chipid == PRODUCT_ID_AIC8800DW){
-#ifndef ANDROID_PLATFORM
+#if 0
         sprintf(aic_fw_path, "%s/%s", aic_fw_path, "aic8800DC");
 #endif
         AICWFDBG(LOGINFO, "testmode=%d\n", testmode);
--- a/src/USB/driver_fw/drivers/aic8800/aic_load_fw/Makefile
+++ b/src/USB/driver_fw/drivers/aic8800/aic_load_fw/Makefile
@@ -29,6 +29,7 @@ CONFIG_AIC_LOADFW_SUPPORT = m
 MODULE_NAME = aic_load_fw
 # CONFIG_AIC_FW_PATH = "/vendor/etc/firmware"
 # export CONFIG_AIC_FW_PATH
+ccflags-y += -DCONFIG_AIC_FW_PATH=\"$(CONFIG_AIC_FW_PATH)\"
 
 ifeq ($(CONFIG_USB_SUPPORT), y)
 ccflags-y += -DAICWF_USB_SUPPORT
--- a/src/USB/driver_fw/drivers/aic8800/aic_load_fw/aicbluetooth.c
+++ b/src/USB/driver_fw/drivers/aic8800/aic_load_fw/aicbluetooth.c
@@ -144,7 +144,7 @@ enum aicbsp_cpmode_type {
 
 #define FW_PATH_MAX 200
 #if defined(CONFIG_PLATFORM_UBUNTU)
-static const char* aic_default_fw_path = "/lib/firmware";
+static const char* aic_default_fw_path = CONFIG_AIC_FW_PATH;
 #else
 static const char* aic_default_fw_path = "/vendor/etc/firmware";
 #endif
@@ -321,7 +321,7 @@ static int aic_load_firmware(u32 ** fw_b
 		printk("%s: use customer define fw_path\n", __func__);
 		len = snprintf(path, FW_PATH_MAX, "%s/%s", aic_fw_path, name);
     } else {
-    #if defined(CONFIG_PLATFORM_UBUNTU)
+    #if 0
         if (usb_dev->chipid == PRODUCT_ID_AIC8800) {
             len = snprintf(path, FW_PATH_MAX, "%s/%s/%s",aic_default_fw_path, "aic8800", name);
         } else if (usb_dev->chipid == PRODUCT_ID_AIC8800D80) {
--- a/src/USB/driver_fw/drivers/aic_btusb/Makefile
+++ b/src/USB/driver_fw/drivers/aic_btusb/Makefile
@@ -4,6 +4,7 @@ CONFIG_AIC8800_BTUSB_SUPPORT = m
 CONFIG_SUPPORT_VENDOR_APCF = n
 # Need to set fw path in BOARD_KERNEL_CMDLINE
 CONFIG_USE_FW_REQUEST = n
+ccflags-y += -DCONFIG_AIC_FW_PATH=\"$(CONFIG_AIC_FW_PATH)\"
 
 ifeq ($(CONFIG_SUPPORT_VENDOR_APCF), y)
 obj-$(CONFIG_AIC8800_BTUSB_SUPPORT) := $(MODULE_NAME).o aic_btusb_external_featrue.o
--- a/src/USB/driver_fw/drivers/aic_btusb/aic_btusb.c
+++ b/src/USB/driver_fw/drivers/aic_btusb/aic_btusb.c
@@ -2844,7 +2844,7 @@ struct aicbsp_info_t aicbsp_info = {
 
 char aic_fw_path[FW_PATH_MAX];
 #if (CONFIG_BLUEDROID == 0) || defined (CONFIG_PLATFORM_UBUNTU)
-static const char* aic_default_fw_path = "/lib/firmware";
+static const char* aic_default_fw_path = CONFIG_AIC_FW_PATH;
 #else
 static const char* aic_default_fw_path = "/vendor/etc/firmware";
 #endif
