--- a/src/PCIE/driver_fw/driver/aic8800d80x2/aic8800_fdrv/Makefile
+++ b/src/PCIE/driver_fw/driver/aic8800d80x2/aic8800_fdrv/Makefile
@@ -11,7 +11,7 @@ RWNX_VERS_NUM := 6.4.3.0
 CONFIG_COUNTRY_CODE = "00"
 
 
-MODULE_NAME = aic8800D80_fdrv
+MODULE_NAME = aic8800D80x2_fdrv
 CONFIG_AIC8800_WLAN_SUPPORT = m
 
 # Support of bootrom start
--- a/src/USB/driver_fw/drivers/aic8800/aic8800_fdrv/Makefile
+++ b/src/USB/driver_fw/drivers/aic8800/aic8800_fdrv/Makefile
@@ -5,7 +5,7 @@ EXTRA_CFLAGS += -Wno-implicit-fallthroug
 RWNX_VERS_NUM := 6.4.3.0
 
 CONFIG_AIC8800_WLAN_SUPPORT = m
-MODULE_NAME = aic8800_fdrv
+MODULE_NAME = aic8800DC_fdrv
 CONFIG_COUNTRY_CODE = "00"
 
 # Support of bootrom start
