--- a/src/USB/driver_fw/drivers/aic8800/aic8800_fdrv/aicwf_usb.c
+++ b/src/USB/driver_fw/drivers/aic8800/aic8800_fdrv/aicwf_usb.c
@@ -1252,7 +1252,7 @@ int usb_bustx_thread(void *data)
 
 #ifdef CONFIG_THREAD_INFO_IN_TASK
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 16, 0))
-	AICWFDBG(LOGINFO, "%s the cpu is:%d\n", __func__, current->thread_info.cpu);
+	AICWFDBG(LOGINFO, "%s the cpu is:%d\n", __func__, smp_processor_id());
 #else
     AICWFDBG(LOGINFO, "%s the cpu is:%d\n", __func__, current->cpu);
 #endif
@@ -1261,7 +1261,7 @@ int usb_bustx_thread(void *data)
 #ifdef CONFIG_THREAD_INFO_IN_TASK
     AICWFDBG(LOGINFO, "%s set_cpu_ret is:%d\n", __func__, set_cpu_ret);
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 16, 0))
-	AICWFDBG(LOGINFO, "%s change cpu to:%d\n", __func__, current->thread_info.cpu);
+	AICWFDBG(LOGINFO, "%s change cpu to:%d\n", __func__, smp_processor_id());
 #else
     AICWFDBG(LOGINFO, "%s change cpu to:%d\n", __func__, current->cpu);
 #endif
@@ -1319,7 +1319,7 @@ int usb_busrx_thread(void *data)
     
 #ifdef CONFIG_THREAD_INFO_IN_TASK
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 16, 0))
-	AICWFDBG(LOGINFO, "%s the cpu is:%d\n", __func__, current->thread_info.cpu);
+	AICWFDBG(LOGINFO, "%s the cpu is:%d\n", __func__, smp_processor_id());
 #else
     AICWFDBG(LOGINFO, "%s the cpu is:%d\n", __func__, current->cpu);
 #endif
@@ -1328,7 +1328,7 @@ int usb_busrx_thread(void *data)
 #ifdef CONFIG_THREAD_INFO_IN_TASK
     AICWFDBG(LOGINFO, "%s set_cpu_ret is:%d\n", __func__, set_cpu_ret);
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 16, 0))
-	AICWFDBG(LOGINFO, "%s change cpu to:%d\n", __func__, current->thread_info.cpu);
+	AICWFDBG(LOGINFO, "%s change cpu to:%d\n", __func__, smp_processor_id());
 #else
     AICWFDBG(LOGINFO, "%s change cpu to:%d\n", __func__, current->cpu);
 #endif
