--- a/src/security/SecurityCertificateManager.cpp
+++ b/src/security/SecurityCertificateManager.cpp
@@ -23,6 +23,7 @@
 #include "Configuration.h"
 
 #include <cstdlib>
+#include <cstring>
 
 #include <ibrcommon/Logger.h>
 #include <ibrcommon/ssl/TLSStream.h>
--- a/src/security/SecurityManager.cpp
+++ b/src/security/SecurityManager.cpp
@@ -28,6 +28,7 @@
 #include <ibrdtn/security/PayloadConfidentialBlock.h>
 #include <ibrdtn/security/ExtensionSecurityBlock.h>
 #include <ibrcommon/Logger.h>
+#include <cstring>
 
 #ifdef __DEVELOPMENT_ASSERTIONS__
 #include <cassert>
