--- a/src/security/SecurityKeyManager.cpp
+++ b/src/security/SecurityKeyManager.cpp
@@ -29,6 +29,7 @@
 #include <fstream>
 #include <fcntl.h>
 
+#include <openssl/bn.h>
 #include <openssl/pem.h>
 #include <openssl/rsa.h>
 #include <openssl/err.h>
--- a/src/security/exchange/DHProtocol.cpp
+++ b/src/security/exchange/DHProtocol.cpp
@@ -28,6 +28,7 @@
 #include <ibrcommon/ssl/HMacStream.h>
 #include <ibrcommon/Logger.h>
 
+#include <openssl/dh.h>
 #include <openssl/rand.h>
 #include <openssl/pem.h>
 #include "openssl_compat.h"
