--- a/ifxmips_ptm_adsl.c
+++ b/ifxmips_ptm_adsl.c
@@ -183,7 +183,7 @@ static int proc_read_version(char *, cha
 static int proc_read_wanmib(char *, char **, off_t, int, int *, void *);
 static int proc_write_wanmib(struct file *, const char *, unsigned long, void *);
 #endif
-#if defined(ENABLE_FW_PROC) && ENABLE_FW_PROC
+#if defined(ENABLE_FW_PROC) && ENABLE_FW_PROC && defined(ENABLE_DBG_PROC) && ENABLE_DBG_PROC
   static int proc_read_genconf(char *, char **, off_t, int, int *, void *);
 #endif
 #if defined(ENABLE_DBG_PROC) && ENABLE_DBG_PROC
@@ -194,8 +194,8 @@ static int proc_write_wanmib(struct file
 /*
  *  Proc Help Functions
  */
-static INLINE int stricmp(const char *, const char *);
 #if defined(ENABLE_DBG_PROC) && ENABLE_DBG_PROC
+  static INLINE int stricmp(const char *, const char *);
   static INLINE int strincmp(const char *, const char *, int);
 #endif
 static INLINE int ifx_ptm_version(char *);
@@ -1206,8 +1206,6 @@ static int proc_write_dbg(struct file *f
     return count;
 }
 
-#endif  //  defined(ENABLE_DBG_PROC) && ENABLE_DBG_PROC
-
 static INLINE int stricmp(const char *p1, const char *p2)
 {
     int c1, c2;
@@ -1225,7 +1223,6 @@ static INLINE int stricmp(const char *p1
     return *p1 - *p2;
 }
 
-#if defined(ENABLE_DBG_PROC) && ENABLE_DBG_PROC
 static INLINE int strincmp(const char *p1, const char *p2, int n)
 {
     int c1 = 0, c2;
