From f3cc8c83993486b9d35557f2bd85038d5bb96bc5 Mon Sep 17 00:00:00 2001
From: Yousong Zhou <yszhou4tech@gmail.com>
Date: Wed, 29 Jul 2020 20:09:56 +0800
Subject: [PATCH] build: skip check and use of libunbound

Signed-off-by: Yousong Zhou <yszhou4tech@gmail.com>
---
 configure.ac | 1 -
 1 file changed, 1 deletion(-)

--- a/configure.ac
+++ b/configure.ac
@@ -130,7 +130,6 @@ OVS_CHECK_LINUX_HOST
 OVS_LIBTOOL_VERSIONS
 OVS_CHECK_CXX
 AX_FUNC_POSIX_MEMALIGN
-OVN_CHECK_UNBOUND
 
 OVS_CHECK_INCLUDE_NEXT([stdio.h string.h])
 AC_CONFIG_FILES([lib/libovn.sym])
