From cea7c74ecc4a0843f29ce1882d7876fff199a721 Mon Sep 17 00:00:00 2001
From: Yousong Zhou <yszhou4tech@gmail.com>
Date: Mon, 23 Mar 2020 14:18:30 +0800
Subject: [PATCH] build: skip tests and docs

Signed-off-by: Yousong Zhou <yszhou4tech@gmail.com>
---
 Makefile.am | 2 --
 1 file changed, 2 deletions(-)

--- a/Makefile.am
+++ b/Makefile.am
@@ -483,11 +483,9 @@ dist-docs:
 
 
 include automake.mk
-include Documentation/automake.mk
 include m4/automake.mk
 include lib/automake.mk
 include utilities/automake.mk
-include tests/automake.mk
 include include/automake.mk
 include debian/automake.mk
 include lib/ovsdb_automake.mk
