From ae90737100d901723af9890e69798bd1dac9a86d Mon Sep 17 00:00:00 2001
From: Daniel Mierswa <impulze@impulze.org>
Date: Sun, 31 Jul 2011 23:53:21 +0000
Subject: [PATCH 2/2] udevd: add -lrt for message queue symbols

---
 Makefile.am |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index 3d8ce92..7d8a4b6 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -164,7 +164,7 @@ udev_udevd_SOURCES = \
 	udev/udevd.c \
 	udev/sd-daemon.h \
 	udev/sd-daemon.c
-udev_udevd_LDADD = libudev/libudev-private.la
+udev_udevd_LDADD = libudev/libudev-private.la -lrt
 
 udev_udevadm_SOURCES = \
 	$(udev_common_sources) \
-- 
1.7.6

