--- a/drivers/net/wireless/realtek/rtw89/pci.c
+++ b/drivers/net/wireless/realtek/rtw89/pci.c

/drivers/net/wireless/realtek/rtw89/pci.c
@@ -3295,8 +3295,11 @@
 {
 	unsigned long flags = 0;
 	int ret;
-
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 10, 0))
 	flags |= PCI_IRQ_LEGACY | PCI_IRQ_MSI;
+#else
+	flags |= PCI_IRQ_INTX | PCI_IRQ_MSI;
+#endif
 	ret = pci_alloc_irq_vectors(pdev, 1, 1, flags);
 	if (ret < 0) {
 		rtw89_err(rtwdev, "failed to alloc irq vectors, ret %d\n", ret);
