--- a/config
+++ b/config
@@ -1,16 +1,16 @@
 # Installation directories
 
 # Default prefix
-PREFIX = /usr/local
+PREFIX = /usr
 
 # Lua version
 LUA_SYS_VER ?= 5.2
 
 # System's libraries directory (where binary libraries are installed)
-LUA_LIBDIR= $(PREFIX)/lib/lua/$(LUA_SYS_VER)
+LUA_LIBDIR= $(PREFIX)/lib/lua
 
 # System's lua directory (where Lua libraries are installed)
-LUA_DIR= $(PREFIX)/share/lua/$(LUA_SYS_VER)
+LUA_DIR= $(PREFIX)/share/lua
 
 # Lua includes directory
 LUA_INC= $(PREFIX)/include
