--- a/super.c
+++ b/super.c
@@ -292,14 +292,14 @@ static const struct fs_parameter_spec exfat_param_specs[] = {
 #endif
 	fsparam_flag("discard",			Opt_discard),
 	fsparam_s32("time_offset",		Opt_time_offset),
-	__fsparam(NULL, "utf8",			Opt_utf8, fs_param_deprecated,
-		  NULL),
-	__fsparam(NULL, "debug",		Opt_debug, fs_param_deprecated,
-		  NULL),
+	__fsparam(NULL, "utf8",			Opt_utf8, fs_param_deprecated
+		  ),
+	__fsparam(NULL, "debug",		Opt_debug, fs_param_deprecated
+		  ),
 	__fsparam(fs_param_is_u32, "namecase",	Opt_namecase,
-		  fs_param_deprecated, NULL),
+		  fs_param_deprecated),
 	__fsparam(fs_param_is_u32, "codepage",	Opt_codepage,
-		  fs_param_deprecated, NULL),
+		  fs_param_deprecated),
 	{}
 };
 
