From d789256e237343f17097eed7ef29be554b5a2265 Mon Sep 17 00:00:00 2001
From: ailick <277498654@qq.com>
Date: Sat, 31 Jul 2021 00:49:23 +0800
Subject: [PATCH] Fix compilation issues

---
 eip93-prng.c | 1 +
 eip93-ring.c | 1 +
 2 files changed, 2 insertions(+)

diff --git a/eip93-prng.c b/eip93-prng.c
index 9c02522..87cd4dc 100644
--- a/eip93-prng.c
+++ b/eip93-prng.c
@@ -5,6 +5,7 @@
  * Richard van Schagen <vschagen@cs.com>
  */
 
+#include <linux/dma-mapping.h>
 #include "eip93-common.h"
 #include "eip93-core.h"
 #include "eip93-regs.h"
diff --git a/eip93-ring.c b/eip93-ring.c
index fff5c0b..b553bba 100644
--- a/eip93-ring.c
+++ b/eip93-ring.c
@@ -5,6 +5,7 @@
  * Richard van Schagen <vschagen@cs.com>
  */
 
+#include <linux/device.h>
 #include "eip93-common.h"
 #include "eip93-core.h"
 
-- 
2.30.2

