--- a/configure.ac
+++ b/configure.ac
@@ -125,15 +125,7 @@ fi
 AC_SUBST(GOBJECT_CFLAGS)
 AC_SUBST(GOBJECT_LIBS)
 
-# Have gdkpixbuf?
-PKG_CHECK_MODULES(GDKPIXBUF, gdk-pixbuf-2.0, HAVE_GDKPIXBUF=yes, HAVE_GDKPIXBUF=no)
-if test x"$HAVE_GDKPIXBUF" = "xyes"; then
-	AC_DEFINE(HAVE_GDKPIXBUF, 1, [Define if gdk-pixbuf support is enabled])
-else
-	AC_MSG_WARN([Gdk-pixbuf library not present; now-playing artwork might be affected])
-fi
-
-AM_CONDITIONAL(USE_GDKPIXBUF, test x"$HAVE_GDKPIXBUF" = "xyes")
+HAVE_GDKPIXBUF=no
 
 AC_SUBST(GDKPIXBUF_CFLAGS)
 AC_SUBST(GDKPIXBUF_LIBS)
