--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -409,6 +409,7 @@ static int mtk_mdio_init(struct mtk_eth
 
 	snprintf(eth->mii_bus->id, MII_BUS_ID_SIZE, "%s", mii_np->name);
 	ret = of_mdiobus_register(eth->mii_bus, mii_np);
+printk("%s:%s[%d]%d %p\n", __FILE__, __func__, __LINE__, ret, eth->mii_bus);
 
 err_put_node:
 	of_node_put(mii_np);
@@ -1472,7 +1473,10 @@ static void mtk_hwlro_rx_uninit(struct m
 	for (i = 0; i < 10; i++) {
 		val = mtk_r32(eth, MTK_PDMA_LRO_CTRL_DW0);
 		if (val & MTK_LRO_RING_RELINQUISH_DONE) {
-			msleep(20);
+			if (in_atomic())
+				mdelay(20);
+			else
+				msleep(20);
 			continue;
 		}
 		break;
@@ -1869,7 +1873,10 @@ static void mtk_stop_dma(struct mtk_eth
 	for (i = 0; i < 10; i++) {
 		val = mtk_r32(eth, glo_cfg);
 		if (val & (MTK_TX_DMA_BUSY | MTK_RX_DMA_BUSY)) {
-			msleep(20);
+			if (in_atomic())
+				mdelay(20);
+			else
+				msleep(20);
 			continue;
 		}
 		break;
@@ -1907,7 +1914,10 @@ static void ethsys_reset(struct mtk_eth
 			   reset_bits,
 			   reset_bits);
 
-	usleep_range(1000, 1100);
+	if (in_atomic())
+		udelay(1000);
+	else
+		usleep_range(1000, 1100);
 	regmap_update_bits(eth->ethsys, ETHSYS_RSTCTRL,
 			   reset_bits,
 			   ~reset_bits);
