--- a/ipsecmgr/v1.0/nss_ipsecmgr.c
+++ b/ipsecmgr/v1.0/nss_ipsecmgr.c
@@ -377,7 +377,7 @@ free:
  * nss_ipsecmgr_tunnel_stats()
  * 	get tunnel statistics
  */
-static struct rtnl_link_stats64 *nss_ipsecmgr_tunnel_stats64(struct net_device *dev, struct rtnl_link_stats64 *stats)
+void nss_ipsecmgr_tunnel_stats64(struct net_device *dev, struct rtnl_link_stats64 *stats)
 {
 	struct nss_ipsecmgr_priv *priv = netdev_priv(dev);
 
@@ -389,8 +389,6 @@ static struct rtnl_link_stats64 *nss_ips
 	read_lock_bh(&ipsecmgr_ctx->lock);
 	memcpy(stats, &priv->stats, sizeof(struct rtnl_link_stats64));
 	read_unlock_bh(&ipsecmgr_ctx->lock);
-
-	return stats;
 }
 
 /*
@@ -442,7 +440,7 @@ static void nss_ipsecmgr_tunnel_setup(st
 	dev->header_ops = NULL;
 	dev->netdev_ops = &nss_ipsecmgr_tunnel_ops;
 
-	dev->destructor = nss_ipsecmgr_tunnel_free;
+	dev->priv_destructor = nss_ipsecmgr_tunnel_free;
 
 	/*
 	 * get the MAC address from the ethernet device
