--- a/gen_oui.sh
+++ b/gen_oui.sh
@@ -3,7 +3,7 @@
 VER="1.2"
 
 # Location of tmp file
-TMPFILE="/tmp/oui.tmp"
+TMPFILE="./oui.tmp"
 
 # File to write
 OUIFILE="oui.txt"
@@ -22,10 +22,7 @@ exit 1
 
 get_oui ()
 {
-echo -n "Downloading OUI file from IEEE..."
-wget --quiet -O $TMPFILE http://standards.ieee.org/develop/regauth/oui/oui.txt || \
-	ErrorMsg ERR "Unable to contact IEEE server!"
-
+[ -f "$TMPFILE" ] || ErrorMsg ERR "Unable to find $TMPFILE"
 echo "OK"
 }
 
@@ -44,7 +41,6 @@ echo "OK"
 clean_all ()
 {
 echo -n "Removing files..."
-rm -f $TMPFILE
 rm -f $OUIFILE
 echo "OK"
 }
