--- a/src/server.c
+++ b/src/server.c
@@ -2013,11 +2013,8 @@ static void dispatch_method(lo_server s,
 
                     tmp = (char*) malloc(strlen(it->path + len) + 1);
                     strcpy(tmp, it->path + len);
-#if defined(WIN32) || defined(_MSC_VER)
                     sec = strchr(tmp, '/');
-#else
-                    sec = index(tmp, '/');
-#endif
+
                     if (sec)
                         *sec = '\0';
                     slend = sl;
