--- a/src/tools/oscsendfile.c
+++ b/src/tools/oscsendfile.c
@@ -354,7 +354,7 @@ int send_file(lo_address target, double
             lo_timetag_now(&tt_now);
             double wait_time = timetag_diff(*tt_last, tt_now);
             if (wait_time > 0.) {
-                usleep(wait_time * 1000000);
+                sleep(wait_time);
             }
             if (b) {
                 ret = lo_send_bundle(target, b);
@@ -376,7 +376,7 @@ int send_file(lo_address target, double
         lo_timetag_now(&tt_now);
         double wait_time = timetag_diff(*tt_last, tt_now);
         if (wait_time > 0.) {
-            usleep(wait_time * 1000000);
+            sleep(wait_time);
         }
         lo_send_bundle(target, b);
     }
