diff -ur a/Makefile b/Makefile
--- a/Makefile	2002-09-20 18:23:55.000000000 +0000
+++ b/Makefile	2007-05-17 13:59:55.000000000 +0000
@@ -16,8 +16,8 @@
 CCOPT=-D_GNU_SOURCE -O2 -Wstrict-prototypes -Wall -g
 CFLAGS=$(CCOPT) $(GLIBCFIX) $(DEFINES) 
 
-IPV4_TARGETS=tracepath ping clockdiff rdisc arping tftpd rarpd
-IPV6_TARGETS=tracepath6 traceroute6 ping6
+IPV4_TARGETS=tracepath ping clockdiff arping tftpd
+IPV6_TARGETS=tracepath6 traceroute6 ping6
 TARGETS=$(IPV4_TARGETS) $(IPV6_TARGETS)
 
 LASTTAG:=`git describe HEAD | sed -e 's/-.*//'`
