--- a/Makefile
+++ b/Makefile
@@ -30,9 +30,13 @@ all: $(TARGETS)
 
 
 tftpd: tftpd.o tftpsubs.o
-arping: arping.o -lsysfs
+arping: arping.o 
+	$(CC) $(CFLAGS) -o $@ arping.o -lsysfs
 ping: ping.o ping_common.o
-ping6: ping6.o ping_common.o -lresolv -lcrypto
+ping6: ping6.o ping_common.o
+ifneq ($(CONFIG_USE_UCLIBC),y)
+	$(CC) $(CFLAGS) -o $@ ping6.o ping_common.o -lresolv -lcrypto
+endif
 ping.o ping6.o ping_common.o: ping_common.h
 tftpd.o tftpsubs.o: tftp.h
 
