--- a/arch/arm64/boot/dts/mediatek/mt7622-bananapi-bpi-r64.dts
+++ b/arch/arm64/boot/dts/mediatek/mt7622-bananapi-bpi-r64.dts
@@ -20,6 +20,7 @@
 
 	aliases {
 		serial0 = &uart0;
+		ethernet0 = &gmac0;
 	};
 
 	chosen {
@@ -164,22 +165,22 @@
 
 				port@1 {
 					reg = <1>;
-					label = "lan0";
+					label = "lan1";
 				};
 
 				port@2 {
 					reg = <2>;
-					label = "lan1";
+					label = "lan2";
 				};
 
 				port@3 {
 					reg = <3>;
-					label = "lan2";
+					label = "lan3";
 				};
 
 				port@4 {
 					reg = <4>;
-					label = "lan3";
+					label = "lan4";
 				};
 
 				port@6 {
