--- a/drivers/mtd/nand/spi/core.c
+++ b/drivers/mtd/nand/spi/core.c
@@ -724,7 +724,7 @@ static int spinand_mtd_write(struct mtd_
 static bool spinand_isbad(struct nand_device *nand, const struct nand_pos *pos)
 {
 	struct spinand_device *spinand = nand_to_spinand(nand);
-	u8 marker[2] = { };
+	u8 marker[1] = { };
 	struct nand_page_io_req req = {
 		.pos = *pos,
 		.ooblen = sizeof(marker),
@@ -735,7 +735,7 @@ static bool spinand_isbad(struct nand_de
 
 	spinand_select_target(spinand, pos->target);
 	spinand_read_page(spinand, &req);
-	if (marker[0] != 0xff || marker[1] != 0xff)
+	if (marker[0] != 0xff)
 		return true;
 
 	return false;
