--- a/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts
+++ b/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts
@@ -17,6 +17,10 @@
 		serial2 = &uart2;
 		mmc0 = &mmc0;
 		mmc1 = &mmc1;
+		led-boot = &led_system_green;
+		led-failsafe = &led_system_blue;
+		led-running = &led_system_green;
+		led-upgrade = &led_system_blue;
 	};
 
 	chosen {
@@ -112,13 +116,13 @@
 		pinctrl-names = "default";
 		pinctrl-0 = <&led_pins_a>;
 
-		blue {
+		led_system_blue: blue {
 			label = "bpi-r2:pio:blue";
 			gpios = <&pio 240 GPIO_ACTIVE_LOW>;
 			default-state = "off";
 		};
 
-		green {
+		led_system_green: green {
 			label = "bpi-r2:pio:green";
 			gpios = <&pio 241 GPIO_ACTIVE_LOW>;
 			default-state = "off";
