---
 drivers/leds/Kconfig  |   10 ++++++++++
 drivers/leds/Makefile |    1 +
 2 files changed, 11 insertions(+)

--- a/drivers/leds/Kconfig
+++ b/drivers/leds/Kconfig
@@ -874,6 +874,16 @@ source "drivers/leds/flash/Kconfig"
 comment "RGB LED drivers"
 source "drivers/leds/rgb/Kconfig"
 
+config LEDS_SMARTRG_LED
+	tristate "LED support for Adtran SmartRG"
+	depends on LEDS_CLASS && I2C && OF
+	help
+	  This option enables support for the Adtran SmartRG platform
+	  system LED driver.
+
+	  To compile this driver as a module, choose M here: the module
+	  will be called leds-smartrg-system.
+
 comment "LED Triggers"
 source "drivers/leds/trigger/Kconfig"
 
--- a/drivers/leds/Makefile
+++ b/drivers/leds/Makefile
@@ -76,6 +76,7 @@ obj-$(CONFIG_LEDS_PWM)			+= leds-pwm.o
 obj-$(CONFIG_LEDS_REGULATOR)		+= leds-regulator.o
 obj-$(CONFIG_LEDS_S3C24XX)		+= leds-s3c24xx.o
 obj-$(CONFIG_LEDS_SC27XX_BLTC)		+= leds-sc27xx-bltc.o
+obj-$(CONFIG_LEDS_SMARTRG_LED)		+= leds-smartrg-system.o
 obj-$(CONFIG_LEDS_SUNFIRE)		+= leds-sunfire.o
 obj-$(CONFIG_LEDS_SYSCON)		+= leds-syscon.o
 obj-$(CONFIG_LEDS_TCA6507)		+= leds-tca6507.o
