Index: EasyRSA-3.0.4/easyrsa
===================================================================
--- EasyRSA-3.0.4.orig/easyrsa
+++ EasyRSA-3.0.4/easyrsa
@@ -1033,7 +1033,7 @@ vars_setup() {
 	vars=
 
 	# set up program path
-	prog_vars="${0%/*}/vars"
+	prog_vars="/etc/easy-rsa/vars"
 	# set up PKI path
 	pki_vars="${EASYRSA_PKI:-$PWD/pki}/vars"
 
@@ -1060,7 +1060,7 @@ Note: using Easy-RSA configuration from:
 	fi
 	
 	# Set defaults, preferring existing env-vars if present
-	set_var EASYRSA		"${0%/*}"
+	set_var EASYRSA		"/etc/easy-rsa"
 	set_var EASYRSA_OPENSSL	openssl
 	set_var EASYRSA_PKI	"$PWD/pki"
 	set_var EASYRSA_DN	cn_only
