diff --git a/lib/getline.h b/lib/getline.h
index 991184c..e41865c 100644
--- a/lib/getline.h
+++ b/lib/getline.h
@@ -27,6 +27,7 @@ Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */
 #  endif
 # endif
 
+# ifdef _MUSL_
 # if __GLIBC__ < 2
 int
 getline PARAMS ((char **_lineptr, size_t *_n, FILE *_stream));
@@ -34,5 +35,6 @@ getline PARAMS ((char **_lineptr, size_t *_n, FILE *_stream));
 int
 getdelim PARAMS ((char **_lineptr, size_t *_n, int _delimiter, FILE *_stream));
 # endif
+# endif
 
 #endif /* not GETLINE_H_ */
