--- a/src/r8125_n.c
+++ b/src/r8125_n.c
@@ -15345,6 +15345,7 @@ rtl8125_init_board(struct pci_dev *pdev,
         void __iomem *ioaddr;
         struct net_device *dev;
         struct rtl8125_private *tp;
+        const char *devname = of_get_property(pdev->dev.of_node, "label", NULL);
         int rc = -ENOMEM, i, pm_cap;
 
         assert(ioaddr_out != NULL);
@@ -15359,6 +15360,9 @@ rtl8125_init_board(struct pci_dev *pdev,
                 goto err_out;
         }
 
+        if (devname)
+                strscpy(dev->name, devname, IFNAMSIZ);
+
         SET_MODULE_OWNER(dev);
         SET_NETDEV_DEV(dev, &pdev->dev);
         tp = netdev_priv(dev);
