--- a/source3/wscript
+++ b/source3/wscript
@@ -68,6 +68,12 @@ def set_options(opt):
                    help=("Directory under which libcephfs is installed"),
                    action="store", dest='libcephfs_dir', default=None)
 
+    opt.add_option('--program-suffix',
+                   help=("Directory under which libcephfs is installed"))
+
+    opt.add_option('--disable-nls',
+                   help=("Directory under which libcephfs is installed"))
+
     opt.SAMBA3_ADD_OPTION('glusterfs', with_name="enable", without_name="disable", default=True)
 
     opt.add_option('--enable-vxfs',
