--- a/source3/rpc_server/rpc_service_setup.c
+++ b/source3/rpc_server/rpc_service_setup.c
@@ -560,6 +560,7 @@ bool dcesrv_ep_setup(struct tevent_conte
 	}
 #endif
 
+#ifdef EXTRA_SERVICES
 	ok = rpc_setup_svcctl(ev_ctx, msg_ctx);
 	if (!ok) {
 		goto done;
@@ -574,6 +575,7 @@ bool dcesrv_ep_setup(struct tevent_conte
 	if (!ok) {
 		goto done;
 	}
+#endif	
 
 	ok = rpc_setup_initshutdown(ev_ctx, msg_ctx);
 	if (!ok) {
--- a/source3/smbd/server_exit.c
+++ b/source3/smbd/server_exit.c
@@ -195,9 +195,11 @@ static void exit_server_common(enum serv
 #endif
 		rpc_netdfs_shutdown();
 		rpc_initshutdown_shutdown();
+#ifdef EXTRA_SERVICES
 		rpc_eventlog_shutdown();
 		rpc_ntsvcs_shutdown();
 		rpc_svcctl_shutdown();
+#endif
 #ifdef PRINTER_SUPPORT
 		rpc_spoolss_shutdown();
 #endif
--- a/source3/rpcclient/rpcclient.c
+++ b/source3/rpcclient/rpcclient.c
@@ -631,9 +631,11 @@ static struct cmd_set *rpcclient_command
 	shutdown_commands,
  	test_commands,
 	wkssvc_commands,
+#ifdef EXTRA_SERVICES
 	ntsvcs_commands,
 	drsuapi_commands,
 	eventlog_commands,
+#endif
 	winreg_commands,
 	fss_commands,
 	witness_commands,
