--- a/source3/rpc_server/rpc_service_setup.c
+++ b/source3/rpc_server/rpc_service_setup.c
@@ -396,10 +396,12 @@ static bool eventlog_init_cb(void *ptr)
 		talloc_get_type_abort(ptr, struct messaging_context);
 	bool ok;
 
+#ifdef WINREG_SUPPORT
 	ok = eventlog_init_winreg(msg_ctx);
 	if (!ok) {
 		return false;
 	}
+#endif
 
 	return true;
 }
@@ -506,10 +508,12 @@ bool dcesrv_ep_setup(struct tevent_conte
 		return false;
 	}
 
+#ifdef WINREG_SUPPORT
 	ok = rpc_setup_winreg(ev_ctx, msg_ctx);
 	if (!ok) {
 		goto done;
 	}
+#endif
 
 	ok = rpc_setup_srvsvc(ev_ctx, msg_ctx);
 	if (!ok) {
--- a/source3/smbd/server_exit.c
+++ b/source3/smbd/server_exit.c
@@ -205,7 +205,9 @@ static void exit_server_common(enum serv
 #endif
 
 		rpc_srvsvc_shutdown();
+#ifdef WINREG_SUPPORT
 		rpc_winreg_shutdown();
+#endif
 
 		rpc_netlogon_shutdown();
 		rpc_samr_shutdown();
--- a/source3/rpc_server/svcctl/srv_svcctl_nt.c
+++ b/source3/rpc_server/svcctl/srv_svcctl_nt.c
@@ -96,9 +96,11 @@ bool init_service_op_table( void )
 	svcctl_ops[i].ops  = &netlogon_svc_ops;
 	i++;
 
+#ifdef WINREG_SUPPORT
 	svcctl_ops[i].name = talloc_strdup( svcctl_ops, "RemoteRegistry" );
 	svcctl_ops[i].ops  = &winreg_svc_ops;
 	i++;
+#endif
 
 	svcctl_ops[i].name = talloc_strdup( svcctl_ops, "WINS" );
 	svcctl_ops[i].ops  = &wins_svc_ops;
--- a/source3/services/svc_winreg_glue.c
+++ b/source3/services/svc_winreg_glue.c
@@ -88,6 +88,10 @@ struct security_descriptor *svcctl_get_s
 	NTSTATUS status;
 	WERROR result = WERR_OK;
 
+#ifndef WINREG_SUPPORT
+	return NULL;
+#endif
+
 	key = talloc_asprintf(mem_ctx,
 			      "%s\\%s\\Security",
 			      TOP_LEVEL_SERVICES_KEY, name);
@@ -161,6 +165,10 @@ bool svcctl_set_secdesc(struct messaging
 	NTSTATUS status;
 	WERROR result = WERR_OK;
 
+#ifndef WINREG_SUPPORT
+	return false;
+#endif
+
 	tmp_ctx = talloc_stackframe();
 	if (tmp_ctx == NULL) {
 		return false;
@@ -272,6 +280,10 @@ const char *svcctl_get_string_value(TALL
 	NTSTATUS status;
 	WERROR result = WERR_OK;
 
+#ifndef WINREG_SUPPORT
+	return NULL;
+#endif
+
 	tmp_ctx = talloc_stackframe();
 	if (tmp_ctx == NULL) {
 		return NULL;
--- a/source3/rpcclient/rpcclient.c
+++ b/source3/rpcclient/rpcclient.c
@@ -636,7 +636,9 @@ static struct cmd_set *rpcclient_command
 	drsuapi_commands,
 	eventlog_commands,
 #endif
+#ifdef WINREG_SUPPORT
 	winreg_commands,
+#endif
 	fss_commands,
 	witness_commands,
 	clusapi_commands,
