--- a/hostapd/Makefile
+++ b/hostapd/Makefile
@@ -323,6 +323,7 @@ ifdef CONFIG_FILS
 CFLAGS += -DCONFIG_FILS
 OBJS += ../src/ap/fils_hlp.o
 NEED_SHA384=y
+NEED_HMAC_SHA384_KDF=y
 NEED_AES_SIV=y
 ifdef CONFIG_FILS_SK_PFS
 CFLAGS += -DCONFIG_FILS_SK_PFS
--- a/wpa_supplicant/Makefile
+++ b/wpa_supplicant/Makefile
@@ -312,6 +312,7 @@ endif
 ifdef CONFIG_FILS
 CFLAGS += -DCONFIG_FILS
 NEED_SHA384=y
+NEED_HMAC_SHA384_KDF=y
 NEED_AES_SIV=y
 ifdef CONFIG_FILS_SK_PFS
 CFLAGS += -DCONFIG_FILS_SK_PFS
