--- a/wpa_supplicant/wpa_supplicant.c
+++ b/wpa_supplicant/wpa_supplicant.c
@@ -259,9 +259,10 @@ void wpa_supplicant_cancel_auth_timeout(
  */
 void wpa_supplicant_initiate_eapol(struct wpa_supplicant *wpa_s)
 {
+	struct wpa_ssid *ssid = wpa_s->current_ssid;
+
 #ifdef IEEE8021X_EAPOL
 	struct eapol_config eapol_conf;
-	struct wpa_ssid *ssid = wpa_s->current_ssid;
 
 #ifdef CONFIG_IBSS_RSN
 	if (ssid->mode == WPAS_MODE_IBSS &&
