From: Felix Fietkau <nbd@openwrt.org>
Date: Sat, 23 Oct 2010 23:39:54 +0000
Subject: [PATCH] nl80211_del_beacon_bss.patch

--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -6253,8 +6253,7 @@ static void nl80211_teardown_ap(struct i
 		nl80211_mgmt_unsubscribe(bss, "AP teardown");
 
 	nl80211_put_wiphy_data_ap(bss);
-	if (bss->flink)
-		bss->flink->beacon_set = 0;
+	wpa_driver_nl80211_del_beacon_all(bss);
 }
 
 
