--- a/hostapd/hostapd_cli.c
+++ b/hostapd/hostapd_cli.c
@@ -447,7 +447,6 @@ static int hostapd_cli_cmd_sa_query(stru
 #endif /* CONFIG_IEEE80211W */
 
 
-#ifdef CONFIG_WPS
 static int hostapd_cli_cmd_wps_pin(struct wpa_ctrl *ctrl, int argc,
 				   char *argv[])
 {
@@ -673,7 +672,6 @@ static int hostapd_cli_cmd_wps_config(st
 			 ssid_hex, argv[1]);
 	return wpa_ctrl_command(ctrl, buf);
 }
-#endif /* CONFIG_WPS */
 
 
 static int hostapd_cli_cmd_disassoc_imminent(struct wpa_ctrl *ctrl, int argc,
@@ -1367,7 +1365,6 @@ static const struct hostapd_cli_cmd host
 	{ "sa_query", hostapd_cli_cmd_sa_query, NULL,
 	  "<addr> = send SA Query to a station" },
 #endif /* CONFIG_IEEE80211W */
-#ifdef CONFIG_WPS
 	{ "wps_pin", hostapd_cli_cmd_wps_pin, NULL,
 	  "<uuid> <pin> [timeout] [addr] = add WPS Enrollee PIN" },
 	{ "wps_check_pin", hostapd_cli_cmd_wps_check_pin, NULL,
@@ -1392,7 +1389,6 @@ static const struct hostapd_cli_cmd host
 	  "<SSID> <auth> <encr> <key> = configure AP" },
 	{ "wps_get_status", hostapd_cli_cmd_wps_get_status, NULL,
 	  "= show current WPS status" },
-#endif /* CONFIG_WPS */
 	{ "disassoc_imminent", hostapd_cli_cmd_disassoc_imminent, NULL, NULL },
 	{ "ess_disassoc", hostapd_cli_cmd_ess_disassoc, NULL, NULL },
 	{ "bss_tm_req", hostapd_cli_cmd_bss_tm_req, NULL, NULL },
