--- a/wpa_supplicant/wpa_supplicant.c
+++ b/wpa_supplicant/wpa_supplicant.c
@@ -2154,11 +2154,13 @@ void ibss_mesh_setup_freq(struct wpa_sup
 	for (j = 0; j < wpa_s->last_scan_res_used; j++) {
 		struct wpa_bss *bss = wpa_s->last_scan_res[j];
 
-		if (ssid->mode != WPAS_MODE_IBSS)
+		/* Don't adjust control freq in case of fixed_freq */
+		if (ssid->fixed_freq) {
+			obss_scan = 0;
 			break;
+		}
 
-		/* Don't adjust control freq in case of fixed_freq */
-		if (ssid->fixed_freq)
+		if (ssid->mode != WPAS_MODE_IBSS)
 			break;
 
 		if (!bss_is_ibss(bss))
