--- a/setup.py
+++ b/setup.py
@@ -4,6 +4,7 @@
 Setup script for the Parsley distribution.
 """
 
+import io
 from distutils.core import setup
 setup(
     name="Parsley",
@@ -13,7 +14,7 @@ setup(
     author="Allen Short",
     author_email="washort42@gmail.com",
     license="MIT License",
-    long_description=open("README").read(),
+    long_description=io.open("README", encoding="utf-8").read(),
     packages=["ometa", "terml", "ometa._generated", "terml._generated",
               "ometa.test", "terml.test"],
     py_modules=["parsley"]
