--- a/setup.py
+++ b/setup.py
@@ -15,7 +15,6 @@ setup(
     author_email="washort42@gmail.com",
     license="MIT License",
     long_description=io.open("README", encoding="utf-8").read(),
-    packages=["ometa", "terml", "ometa._generated", "terml._generated",
-              "ometa.test", "terml.test"],
+    packages=["ometa", "terml", "ometa._generated", "terml._generated"],
     py_modules=["parsley"]
 )
