From: Sven Eckelmann <sven@narfation.org>
Date: Fri, 27 Mar 2026 00:17:23 +0100
Subject: Fix endianness when building on big endian systems

Signed-off-by: Sven Eckelmann <sven@narfation.org>
Forwarded: https://github.com/balika011/realtek_phy_firmware/pull/2

--- a/phy_patch.c
+++ b/phy_patch.c
@@ -3,6 +3,7 @@
 #include <linux/errno.h>
 #include <string.h>
 #include <stdlib.h>
+#include <endian.h>
 
 #include "phy_patch.h"
 
@@ -12,11 +13,11 @@ int phy_patch_op(rt_phy_patch_db_t *pPhy
 
     op.patch_op = patch_op;
     op.portmask = portmask;
-    op.pagemmd  = pagemmd;
-    op.addr     = addr;
+    op.pagemmd  = htole16(pagemmd);
+    op.addr     = htole16(addr);
     op.msb      = msb;
     op.lsb      = lsb;
-    op.data     = data;
+    op.data     = htole16(data);
 
     return pPhy_patchDb->fPatch_op(phydev, &op);
 }
@@ -41,11 +42,11 @@ static int _phy_patch_process(struct phy
 	{
 		patch.patch_op = pPatch[i].patch_op;
 		patch.portmask = pPatch[i].portmask;
-		patch.pagemmd = pPatch[i].pagemmd;
-		patch.addr = pPatch[i].addr;
+		patch.pagemmd = htole16(pPatch[i].pagemmd);
+		patch.addr = htole16(pPatch[i].addr);
 		patch.msb = pPatch[i].msb;
 		patch.lsb = pPatch[i].lsb;
-		patch.data = pPatch[i].data;
+		patch.data = htole16(pPatch[i].data);
 
 		ret = pPatchDb->fPatch_op(phydev, &patch);
 		if (ret < 0)
