From 001a8dcb56ced58c518aaa10a4f0ba5e878705b6 Mon Sep 17 00:00:00 2001
From: Georgi Djakov <georgi.djakov@linaro.org>
Date: Tue, 17 May 2016 16:15:43 +0300
Subject: [PATCH 56/69] cpufreq-dt: Add missing rcu locks

Signed-off-by: Georgi Djakov <georgi.djakov@linaro.org>
---
 drivers/cpufreq/cpufreq-dt.c | 2 ++
 1 file changed, 2 insertions(+)

--- a/drivers/cpufreq/cpufreq-dt.c
+++ b/drivers/cpufreq/cpufreq-dt.c
@@ -146,8 +146,10 @@ static int opp_notifier(struct notifier_
 			ret = PTR_ERR(cpu_reg);
 			goto out;
 		}
+		rcu_read_lock();
 		volt = dev_pm_opp_get_voltage(opp);
 		freq = dev_pm_opp_get_freq(opp);
+		rcu_read_unlock();
 
 		mutex_lock(&priv->lock);
 		if (freq == priv->opp_freq) {
