--- a/drivers/thermal/qcom/tsens-common.c
+++ b/drivers/thermal/qcom/tsens-common.c
@@ -128,6 +128,7 @@ int __init init_common(struct tsens_devi
 {
 	void __iomem *base;
 	struct resource *res;
+	resource_size_t size;
 	struct platform_device *op = of_find_device_by_node(tmdev->dev->of_node);
 
 	if (!op)
@@ -142,7 +143,8 @@ int __init init_common(struct tsens_devi
 	}
 
 	res = platform_get_resource(op, IORESOURCE_MEM, 0);
-	base = devm_ioremap_resource(&op->dev, res);
+	size = resource_size(res);
+	base = devm_ioremap(&op->dev, res->start, size);
 	if (IS_ERR(base))
 		return PTR_ERR(base);
 
