--- a/drivers/net/ethernet/broadcom/bcm63xx_enet.c
+++ b/drivers/net/ethernet/broadcom/bcm63xx_enet.c
@@ -2784,12 +2784,6 @@ static int bcm_enetsw_probe(struct platf
 	if (ret)
 		goto out;
 
-	if (!request_mem_region(res_mem->start, resource_size(res_mem),
-				"bcm63xx_enetsw")) {
-		ret = -EBUSY;
-		goto out;
-	}
-
 	priv->base = ioremap(res_mem->start, resource_size(res_mem));
 	if (priv->base == NULL) {
 		ret = -ENOMEM;
