--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -681,6 +681,26 @@ static struct board_info __initdata boar
 	},
 };
 
+static struct board_info __initdata board_CPVA642 = {
+	.name                           = "CPVA642",
+	.expected_cpu_id                = 0x6358,
+
+	.has_uart0			= 1,
+	.has_enet1                      = 1,
+	.has_pci                        = 1,
+
+	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
+		.force_speed_100        = 1,
+		.force_duplex_full      = 1,
+ 	},
+
+	.has_ohci0 = 1,
+	.has_ehci0 = 1,
+};
+
+
 static struct board_info __initdata board_AGPFS0 = {
 	.name				= "AGPF-S0",
 	.expected_cpu_id		= 0x6358,
@@ -821,6 +841,7 @@ static const struct board_info __initcon
 	&board_96358vw,
 	&board_96358vw2,
 	&board_AGPFS0,
+	&board_CPVA642,
 	&board_DWVS0,
 	&board_nb4_ser_r0,
 	&board_nb4_fxc_r1,
@@ -868,6 +889,7 @@ static struct of_device_id const bcm963x
 	{ .compatible = "pirelli,agpf-s0", .data = &board_AGPFS0, },
 	{ .compatible = "sfr,nb4-ser-r0", .data = &board_nb4_ser_r0, },
 	{ .compatible = "sfr,nb4-fxc-r1", .data = &board_nb4_fxc_r1, },
+	{ .compatible = "telsey,cpva642", .data = &board_CPVA642, },
 #endif
 #ifdef CONFIG_BCM63XX_CPU_6368
 #endif
