--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -998,6 +998,21 @@ static struct board_info __initdata boar
 		.pci_dev			= 1,
 	},
 };
+
+ /* T-Home Speedport W 303V Typ B */
+static struct board_info __initdata board_spw303v = {
+	.name			= "96358-502V",
+	.expected_cpu_id	= 0x6358,
+
+	.has_uart0		= 1,
+	.has_enet0		= 1,
+	.has_pci		= 1,
+
+	.enet0 = {
+		.has_phy 		= 1,
+		.use_internal_phy 	= 1,
+	},
+};
 #endif /* CONFIG_BCM63XX_CPU_6358 */
 
 /*
@@ -1045,6 +1060,7 @@ static const struct board_info __initcon
 	&board_nb4_ser_r0,
 	&board_nb4_fxc_r1,
 	&board_HW553,
+	&board_spw303v,
 #endif
 };
 
@@ -1096,6 +1112,7 @@ static struct of_device_id const bcm963x
 	{ .compatible = "pirelli,agpf-s0", .data = &board_AGPFS0, },
 	{ .compatible = "sfr,nb4-ser-r0", .data = &board_nb4_ser_r0, },
 	{ .compatible = "sfr,nb4-fxc-r1", .data = &board_nb4_fxc_r1, },
+	{ .compatible = "t-com,spw303v", .data = &board_spw303v, },
 	{ .compatible = "telsey,cpva642", .data = &board_CPVA642, },
 #endif
 #ifdef CONFIG_BCM63XX_CPU_6368
