--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -600,6 +600,24 @@ static struct board_info __initdata boar
 	.has_ohci0 = 1,
 };
 
+/* NetGear DG834G v4 */
+static struct board_info __initdata board_96348W3 = {
+ 	.name	 			= "96348W3",
+	.expected_cpu_id		= 0x6348,
+
+	.has_enet1			= 1,
+	.has_pci			= 1,
+
+	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
+		.force_speed_100	= 1,
+		.force_duplex_full	= 1,
+	},
+
+	.has_ohci0	 		= 1,
+};
+
 static struct board_info __initdata board_96348_D4PW = {
 	.name				= "D-4P-W",
 	.expected_cpu_id		= 0x6348,
@@ -1076,6 +1094,7 @@ static const struct board_info __initcon
 	&board_ct536_ct5621,
 	&board_96348A_122,
 	&board_CPVA502plus,
+	&board_96348W3,
 #endif
 
 #ifdef CONFIG_BCM63XX_CPU_6358
@@ -1131,6 +1150,7 @@ static struct of_device_id const bcm963x
 	{ .compatible = "davolink,dv-201amr", .data = &board_DV201AMR, },
 	{ .compatible = "dynalink,rta1025w", .data = &board_rta1025w_16, },
 	{ .compatible = "netgear,dg834gtpn", .data = &board_96348gw_10, },
+ 	{ .compatible = "netgear,dg834g-v4", .data = &board_96348W3, },
 	{ .compatible = "sagem,f@st2404", .data = &board_FAST2404, },
 	{ .compatible = "t-com,spw500v", .data = &board_spw500v, },
 	{ .compatible = "tecom,gw6000", .data = &board_gw6000, },
