--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -426,6 +426,25 @@ static struct board_info __initdata boar
 		.force_duplex_full	= 1,
 	},
 };
+
+// Actually this board is the very same as the rta770bw,
+// where the additional 'b' within the name just
+// just indicates 'Annex B'. The ADSL Modem itself is able
+// to handle both Annex A as well as Annex B -
+// the loaded firmware makes the only difference
+static struct board_info __initdata board_rta770w = {
+	.name				= "RTA770W",
+	.expected_cpu_id		= 0x6345,
+
+	.has_enet0			= 1,
+
+	.enet0 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
+		.force_speed_100	= 1,
+		.force_duplex_full	= 1,
+	},
+};
 #endif /* CONFIG_BCM63XX_CPU_6345 */
 
 /*
@@ -1443,6 +1462,7 @@ static const struct board_info __initcon
 #ifdef CONFIG_BCM63XX_CPU_6345
 	&board_96345gw2,
 	&board_rta770bw,
+	&board_rta770w,
 #endif
 #ifdef CONFIG_BCM63XX_CPU_6348
 	&board_96348r,
@@ -1517,6 +1537,7 @@ static struct of_device_id const bcm963x
 #ifdef CONFIG_BCM63XX_CPU_6345
 	{ .compatible = "brcm,bcm96345gw2", .data = &board_96345gw2, },
 	{ .compatible = "dynalink,rta770bw", .data = &board_rta770bw, },
+	{ .compatible = "dynalink,rta770w", .data = &board_rta770w, },
 #endif
 #ifdef CONFIG_BCM63XX_CPU_6348
 	{ .compatible = "belkin,f5d7633", .data = &board_96348gw_10, },
