--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -50,6 +50,55 @@ static struct board_info __initdata boar
 #endif /* CONFIG_BCM63XX_CPU_3368 */
 
 /*
+ * known 6318 boards
+ */
+#ifdef CONFIG_BCM63XX_CPU_6318
+static struct board_info __initdata board_96318ref = {
+	.name				= "96318REF",
+	.expected_cpu_id		= 0x6318,
+
+	.has_pci			= 1,
+
+	.has_usbd			= 1,
+
+	.usbd = {
+		.use_fullspeed		= 0,
+		.port_no		= 0,
+	},
+
+	.has_enetsw			= 1,
+
+	.has_ehci0			= 1,
+	.num_usbh_ports			= 1,
+
+	.enetsw = {
+		.used_ports = {
+			[0] = {
+				.used	= 1,
+				.phy_id = 1,
+				.name	= "Port 1",
+			},
+			[1] = {
+				.used	= 1,
+				.phy_id	= 2,
+				.name	= "Port 2",
+			},
+			[2] = {
+				.used	= 1,
+				.phy_id	= 3,
+				.name	= "Port 3",
+			},
+			[3] = {
+				.used	= 1,
+				.phy_id	= 4,
+				.name	= "Port 4",
+			},
+		},
+	},
+};
+#endif /* CONFIG_BCM63XX_CPU_6318 */
+
+/*
  * known 6328 boards
  */
 #ifdef CONFIG_BCM63XX_CPU_6328
@@ -1517,6 +1566,9 @@ static const struct board_info __initcon
 #ifdef CONFIG_BCM63XX_CPU_3368
 	&board_cvg834g,
 #endif
+#ifdef CONFIG_BCM63XX_CPU_6318
+	&board_96318ref,
+#endif
 #ifdef CONFIG_BCM63XX_CPU_6328
 	&board_96328avng,
 	&board_AR5381u,
@@ -1594,6 +1646,9 @@ static struct of_device_id const bcm963x
 #ifdef CONFIG_BCM63XX_CPU_3368
 	{ .compatible = "netgear,cvg834g", .data = &board_cvg834g, },
 #endif
+#ifdef CONFIG_BCM63XX_CPU_6318
+	{ .compatible = "brcm,bcm96318ref", .data = &board_96318ref, },
+#endif
 #ifdef CONFIG_BCM63XX_CPU_6328
 	{ .compatible = "adb,a4001n1", .data = &board_A4001N1, },
 	{ .compatible = "brcm,bcm963281TAN", .data = &board_963281TAN, },
