--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -1622,6 +1622,64 @@ static struct board_info __initdata boar
  * known 63268/63269 boards
  */
 #ifdef CONFIG_BCM63XX_CPU_63268
+static struct board_info __initdata board_963268bu_p300 = {
+	.name				= "963268BU_P300",
+	.expected_cpu_id		= 0x63268,
+
+	.has_ehci0			= 1,
+	.has_ohci0			= 1,
+	.num_usbh_ports			= 1,
+
+	.has_usbd			= 1,
+
+	.usbd = {
+		.use_fullspeed		= 0,
+		.port_no		= 0,
+	},
+
+	.has_enetsw			= 1,
+
+	.enetsw = {
+		.used_ports = {
+			[0] = {
+				.used	= 1,
+				.phy_id	= 17,
+				.name	= "FE1",
+			},
+
+			[3] = {
+				.used	= 1,
+				.phy_id	= 4,
+				.name	= "GbE2",
+			},
+
+			[4] = {
+				.used	= 1,
+				.phy_id	= 0,
+				.name	= "GbE3",
+			},
+
+			[5] = {
+				.used	= 1,
+				.phy_id	= 1,
+				.name	= "GbE1",
+			},
+
+			[6] = {
+				.used	= 1,
+				.phy_id	= 24,
+				.name	= "GbE4",
+			},
+
+			[7] = {
+				.used	= 1,
+				.phy_id	= 25,
+				.name	= "GbE5",
+			},
+		},
+	},
+};
+
 static struct board_info __initdata board_963269bhr = {
 	.name				= "963269BHR",
 	.expected_cpu_id		= 0x63268,
@@ -1790,6 +1848,7 @@ static const struct board_info __initcon
 	&board_96368mvngr,
 #endif
 #ifdef CONFIG_BCM63XX_CPU_63268
+	&board_963268bu_p300,
 	&board_963269bhr,
 	&board_vw6339gu,
 #endif
@@ -1881,6 +1940,7 @@ static struct of_device_id const bcm963x
 	{ .compatible = "brcm,bcm96368mvwg", .data = &board_96368mvwg, },
 #endif
 #ifdef CONFIG_BCM63XX_CPU_63268
+	{ .compatible = "brcm,bcm963268bu_p300", .data = &board_963268bu_p300, },
 	{ .compatible = "brcm,bcm963269bhr", .data = &board_963269bhr, },
 	{ .compatible = "inteno,vg50", .data = &board_vw6339gu, },
 #endif
