--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -2116,6 +2116,29 @@ static struct board_info __initdata boar
 	},
 };
 
+static struct board_info __initdata board_R1000H = {
+	.name				= "R1000H",
+	.expected_cpu_id		= 0x6368,
+
+	.has_pci			= 1,
+	.has_ohci0			= 1,
+	.has_ehci0			= 1,
+
+	.has_enetsw			= 1,
+	.enetsw = {
+		.used_ports = {
+			[5] = {
+				.used  = 1,
+				.phy_id  = 0xff,
+				.bypass_link = 1,
+				.force_speed = 1000,
+				.force_duplex_full = 1,
+				.name  = "RGMII",
+			},
+		},
+	},
+};
+
 static struct board_info __initdata board_VR3025u = {
 	.name					= "96368M-1541N",
 	.expected_cpu_id		= 0x6368,
@@ -2586,6 +2609,7 @@ static const struct board_info __initcon
 	&board_HG622,
 	&board_HG655b,
 	&board_P870HW51A_V2,
+	&board_R1000H,
 	&board_VH4032N,
 	&board_VR3025u,
 	&board_VR3025un,
@@ -2687,6 +2711,7 @@ static struct of_device_id const bcm963x
 	{ .compatible = "sfr,nb6-ser-r0", .data = &board_nb6, },
 #endif
 #ifdef CONFIG_BCM63XX_CPU_6368
+	{ .compatible = "actiontec,r1000h", .data = &board_R1000H, },
 	{ .compatible = "adb,av4202n", .data = &board_AV4202N, },
 	{ .compatible = "brcm,bcm96368mvngr", .data = &board_96368mvngr, },
 	{ .compatible = "brcm,bcm96368mvwg", .data = &board_96368mvwg, },
