--- a/drivers/net/ethernet/broadcom/bcm63xx_enet.c
+++ b/drivers/net/ethernet/broadcom/bcm63xx_enet.c
@@ -2270,6 +2270,10 @@ static int bcm_enetsw_open(struct net_de
 
 		rgmii_ctrl = enetsw_readb(priv, ENETSW_RGMII_CTRL_REG(i));
 		rgmii_ctrl |= ENETSW_RGMII_CTRL_GMII_CLK_EN;
+		if (BCMCPU_IS_63268()) {
+			rgmii_ctrl |= ENETSW_RGMII_CTRL_TIMING_SEL_EN;
+			rgmii_ctrl |= ENETSW_RGMII_CTRL_MII_OVERRIDE_EN;
+		}
 		enetsw_writeb(priv, rgmii_ctrl, ENETSW_RGMII_CTRL_REG(i));
 	}
 
