--- a/linux/mon.c
+++ b/linux/mon.c
@@ -44,7 +44,6 @@
 #include <signal.h>
 #include <sys/ioctl.h>
 #include <sys/socket.h>
-#include <sys/sysctl.h>
 #include <sys/queue.h>
 #include <netinet/in.h>
 #include <net/if.h>
--- a/linux/meshd.c
+++ b/linux/meshd.c
@@ -44,7 +44,6 @@
 #include <signal.h>
 #include <sys/ioctl.h>
 #include <sys/socket.h>
-#include <sys/sysctl.h>
 #include <sys/queue.h>
 #include <netinet/in.h>
 #include <net/if.h>
