--- a/arch/powerpc/platforms/44x/Kconfig
+++ b/arch/powerpc/platforms/44x/Kconfig
@@ -143,6 +143,17 @@ config CANYONLANDS
 	help
 	  This option enables support for the AMCC PPC460EX evaluation board.
 
+config APOLLO3G
+	bool "Apollo3G"
+	depends on 44x
+	default n
+	select PPC44x_SIMPLE
+	select APM821xx
+	select IBM_EMAC_RGMII
+	select 460EX
+	help
+	  This option enables support for the AMCC Apollo 3G board.
+
 config GLACIER
 	bool "Glacier"
 	depends on 44x
--- a/arch/powerpc/platforms/44x/ppc44x_simple.c
+++ b/arch/powerpc/platforms/44x/ppc44x_simple.c
@@ -50,6 +50,7 @@ machine_device_initcall(ppc44x_simple, p
  * board.c file for it rather than adding it to this list.
  */
 static char *board[] __initdata = {
+	"amcc,apollo3g",
 	"amcc,arches",
 	"amcc,bamboo",
 	"apm,bluestone",
