--- a/arch/powerpc/platforms/44x/Kconfig
+++ b/arch/powerpc/platforms/44x/Kconfig
@@ -30,6 +30,19 @@ config BLUESTONE
 	help
 	  This option enables support for the APM APM821xx Evaluation board.
 
+config BUCKMINSTER
+	bool "Buckminster"
+	depends on 44x
+	default n
+	select APM821xx
+	select PCI_MSI
+	select PPC4xx_MSI
+	select PPC4xx_PCI_EXPRESS
+	select IBM_EMAC_RGMII
+	select 460EX
+	help
+	  This option enables support for the Cisco Meraki MX60/MX60W (Buckminster) Security Appliance
+
 config EBONY
 	bool "Ebony"
 	depends on 44x
--- a/arch/powerpc/platforms/44x/ppc44x_simple.c
+++ b/arch/powerpc/platforms/44x/ppc44x_simple.c
@@ -63,6 +63,7 @@ static char *board[] __initdata = {
 	"amcc,sequoia",
 	"amcc,taishan",
 	"amcc,yosemite",
+	"meraki,buckminster",
 	"meraki,ikarem",
 	"mosaixtech,icon"
 };
