From 3ccdf969a87be79df3daa9fe2d42f68f90ab7774 Mon Sep 17 00:00:00 2001
From: W_Y_CPP <383152993@qq.com>
Date: Mon, 17 Jan 2022 21:34:38 +0900
Subject: [PATCH] add MODULE_SUPPORTED_DEVICE macro

---
 include/linux/module.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/include/linux/module.h b/include/linux/module.h
index c9f1200b2..a4190c861 100644
--- a/include/linux/module.h
+++ b/include/linux/module.h
@@ -32,6 +32,9 @@
 #include <linux/percpu.h>
 #include <asm/module.h>
 
+/* Not Yet Implemented */
+#define MODULE_SUPPORTED_DEVICE(name)
+
 #define MODULE_NAME_LEN MAX_PARAM_PREFIX_LEN
 
 struct modversion_info {
-- 
2.17.1

