From 453ebd5d6b535388972fcea747025ced3afca5cc Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Thu, 10 Aug 2017 14:47:06 +0200
Subject: [PATCH 22/57] nand: make bootrom work with upstream driver

Signed-off-by: John Crispin <john@phrozen.org>
---
 drivers/mtd/nand/mtk_nand.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

--- a/drivers/mtd/nand/mtk_nand.c
+++ b/drivers/mtd/nand/mtk_nand.c
@@ -1073,8 +1073,8 @@ static int mtk_nfc_ooblayout_free(struct
 	if (section >= eccsteps)
 		return -ERANGE;
 
-	oob_region->length = fdm->reg_size - fdm->ecc_size;
-	oob_region->offset = section * fdm->reg_size + fdm->ecc_size;
+	oob_region->length = fdm->reg_size - 1;
+	oob_region->offset = section * fdm->reg_size + 1;
 
 	return 0;
 }
@@ -1114,7 +1114,7 @@ static void mtk_nfc_set_fdm(struct mtk_n
 		fdm->reg_size = NFI_FDM_MAX_SIZE;
 
 	/* bad block mark storage */
-	fdm->ecc_size = 1;
+	fdm->ecc_size = fdm->reg_size;
 }
 
 static void mtk_nfc_set_bad_mark_ctl(struct mtk_nfc_bad_mark_ctl *bm_ctl,
