--- a/mt7921/dma.c
+++ b/mt7921/dma.c
@@ -283,10 +283,15 @@ int mt7921_dma_init(struct mt7921_dev *d
 	if (ret < 0)
 		return ret;
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 1, 0)
 	netif_tx_napi_add(&dev->mt76.tx_napi_dev, &dev->mt76.tx_napi,
 			  mt7921_poll_tx, NAPI_POLL_WEIGHT);
 	napi_enable(&dev->mt76.tx_napi);
-
+#else
+	netif_napi_add_tx_weight(&dev->mt76.tx_napi_dev, &dev->mt76.tx_napi,
+			  mt7921_poll_tx, NAPI_POLL_WEIGHT);
+	napi_enable(&dev->mt76.tx_napi);
+#endif
 	return mt7921_dma_enable(dev);
 }
 
--- a/dma.c
+++ b/dma.c
@@ -910,7 +910,11 @@ mt76_dma_init(struct mt76_dev *dev,
 	dev->napi_dev.threaded = 1;
 
 	mt76_for_each_q_rx(dev, i) {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6,1,0)
 		netif_napi_add(&dev->napi_dev, &dev->napi[i], poll, 64);
+#else
+		netif_napi_add_weight(&dev->napi_dev, &dev->napi[i], poll, 64);
+#endif
 		mt76_dma_rx_fill(dev, &dev->q_rx[i]);
 		napi_enable(&dev->napi[i]);
 	}
--- a/mt76.h
+++ b/mt76.h
@@ -37,6 +37,9 @@
 #define MT_WED_Q_TX(_n)		__MT_WED_Q(MT76_WED_Q_TX, _n)
 #define MT_WED_Q_RX(_n)		__MT_WED_Q(MT76_WED_Q_RX, _n)
 #define MT_WED_Q_TXFREE		__MT_WED_Q(MT76_WED_Q_TXFREE, 0)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,1,0)
+#define netif_tx_napi_add netif_napi_add_tx_weight
+#endif
 
 struct mt76_dev;
 struct mt76_phy;
