--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -510,11 +510,11 @@ if(WIN32)
 else(WIN32)
     set(DBUS_SESSION_BUS_LISTEN_ADDRESS "unix:tmpdir=${DBUS_SESSION_SOCKET_DIR}" CACHE STRING "session bus default listening address")
     set(DBUS_SESSION_BUS_CONNECT_ADDRESS "autolaunch:" CACHE STRING "session bus fallback address for clients")
-    set(sysconfdir "")
+    set(sysconfdir "/usr/share")
     set(configdir ${sysconfdir}/dbus-1 )
     set(DBUS_SYSTEM_CONFIG_FILE  ${configdir}/system.conf)
     set(DBUS_SESSION_CONFIG_FILE ${configdir}/session.conf)
-    set(DBUS_USER "messagebus")
+    set(DBUS_USER "root")
     set(DBUS_TEST_USER "nobody")
   # For best security, assume that all non-Windows platforms can do
   # credentials-passing.
