--- a/drivers/bcma/driver_chipcommon_sflash.c
+++ b/drivers/bcma/driver_chipcommon_sflash.c
@@ -93,7 +93,7 @@ static void bcma_sflash_cmd(struct bcma_
 	bcma_err(cc->core->bus, "SFLASH control command failed (timeout)!\n");
 }
 
-const struct bcma_sflash_tbl_e *bcma_sflash_shrink_flash(u32 id)
+static const struct bcma_sflash_tbl_e *bcma_sflash_shrink_flash(u32 id)
 {
 	enum bcm47xx_board board = bcm47xx_board_get();
 	const struct bcma_sflash_tbl_e *e;
--- a/drivers/firmware/broadcom/cfe_env.c
+++ b/drivers/firmware/broadcom/cfe_env.c
@@ -21,6 +21,8 @@
 static char _nvdata[NVRAM_SIZE];
 static char _valuestr[256];
 
+char *cfe_env_get(unsigned char *nv_buf, const char *name);
+
 /*
  * TLV types.  These codes are used in the "type-length-value"
  * encoding of the items stored in the NVRAM device (flash or EEPROM)
