--- a/include/uapi/linux/pkt_cls.h
+++ b/include/uapi/linux/pkt_cls.h
@@ -139,6 +139,7 @@ enum tca_id {
 	TCA_ID_MPLS,
 	TCA_ID_CT,
 	TCA_ID_GATE,
+	TCA_ID_MIRRED_NSS, /* QCA NSS Qdisc IGS Support */
 	/* other actions go here */
 	__TCA_ID_MAX = 255
 };
@@ -797,4 +798,14 @@ enum {
 	TCF_EM_OPND_LT
 };
 
+/* QCA NSS Qdisc Support - Start */
+#define _TC_MAKE32(x)		((x))
+#define _TC_MAKEMASK1(n)	(_TC_MAKE32(1) << _TC_MAKE32(n))
+
+#define TC_NCLS                 _TC_MAKEMASK1(8)
+#define TC_NCLS_NSS		_TC_MAKEMASK1(12)
+#define SET_TC_NCLS_NSS(v)	( TC_NCLS_NSS | ((v) & ~TC_NCLS_NSS))
+#define CLR_TC_NCLS_NSS(v)	( (v) & ~TC_NCLS_NSS)
+/* QCA NSS Qdisc Support - End */
+
 #endif
