# HG changeset patch
# User Marc Glisse <marc.glisse@inria.fr>
# Date 1743513946 -7200
#      Tue Apr 01 15:25:46 2025 +0200
# Node ID d66d66d82dbbe4f561920d28c1e1cbe6818452c7
# Parent  1a2ad0e32507e842486c8ac9d5cdc772fd0c335e
Add parameter names to function prototype

 2025-02-01  Marc Glisse <marc.glisse@inria.fr>
 
 	* longlong.h (loongarch64 umul_ppmm): __int128__ -> __int128.
--- a/acinclude.m4
+++ b/acinclude.m4
@@ -609,7 +609,7 @@ GMP_PROG_CC_WORKS_PART([$1], [long long
 
 #if defined (__GNUC__) && ! defined (__cplusplus)
 typedef unsigned long long t1;typedef t1*t2;
-void g(int,t1 const*,t1,t2,t1 const*,int){}
+void g(int a,t1 const*b,t1 c,t2 d,t1 const*e,int f){}
 void h(){}
 static __inline__ t1 e(t2 rp,t2 up,int n,t1 v0)
 {t1 c,x,r;int i;if(v0){c=1;for(i=1;i<n;i++){x=up[i];r=x+1;rp[i]=r;}}return c;}
