From 9c2354cb1d868bb7caad0133f29358c36f562aac Mon Sep 17 00:00:00 2001
From: Junhao Xie <bigfoot@radxa.com>
Date: Wed, 15 Oct 2025 12:58:41 +0800
Subject: [PATCH 035/268] kbuild: Force all dtbs generated symbols to use
 overlay

Signed-off-by: Junhao Xie <bigfoot@radxa.com>
---
 scripts/Makefile.dtbs | 3 +++
 1 file changed, 3 insertions(+)

--- a/scripts/Makefile.dtbs
+++ b/scripts/Makefile.dtbs
@@ -91,6 +91,9 @@ $(call multi_depend, $(multi-dtb-y), .dt
 DTC ?= $(objtree)/scripts/dtc/dtc
 DTC_FLAGS += -Wno-unique_unit_address
 
+# Force generation of symbols even if there are no overlays
+DTC_FLAGS += -@
+
 # Disable noisy checks by default
 ifeq ($(findstring 1,$(KBUILD_EXTRA_WARN)),)
 DTC_FLAGS += -Wno-unit_address_vs_reg \
