From 6041c5f0150f49ab503d5e3c681c2752c7c13020 Mon Sep 17 00:00:00 2001
From: SongJun Li <lisongjun@radxa.com>
Date: Thu, 29 Jan 2026 09:28:17 +0800
Subject: [PATCH 054/268] media: i2c: imx214: add reset GPIO handling in
 power-on sequence

Signed-off-by: SongJun Li <lisongjun@radxa.com>
---
 drivers/media/i2c/imx214.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

--- a/drivers/media/i2c/imx214.c
+++ b/drivers/media/i2c/imx214.c
@@ -271,6 +271,7 @@ struct imx214 {
 
 	struct regulator_bulk_data	supplies[IMX214_NUM_SUPPLIES];
 
+	struct gpio_desc *reset_gpio;
 	struct gpio_desc *enable_gpio;
 };
 
@@ -548,6 +549,16 @@ static int __maybe_unused imx214_power_o
 		return ret;
 	}
 
+	if (!IS_ERR(imx214->reset_gpio))
+		gpiod_set_value_cansleep(imx214->reset_gpio, 1);
+
+	usleep_range(500, 1000);
+
+	if (!IS_ERR(imx214->reset_gpio))
+		gpiod_set_value_cansleep(imx214->reset_gpio, 0);
+
+	usleep_range(500, 1000);
+
 	gpiod_set_value_cansleep(imx214->enable_gpio, 1);
 	usleep_range(12000, 15000);
 
@@ -1408,6 +1419,11 @@ static int imx214_probe(struct i2c_clien
 	if (ret < 0)
 		return dev_err_probe(dev, ret, "failed to get regulators\n");
 
+	imx214->reset_gpio = devm_gpiod_get(dev, "reset", GPIOD_OUT_LOW);
+	if (IS_ERR(imx214->reset_gpio))
+		return dev_err_probe(dev, PTR_ERR(imx214->reset_gpio),
+				     "failed to get reset gpio\n");
+
 	imx214->enable_gpio = devm_gpiod_get(dev, "enable", GPIOD_OUT_LOW);
 	if (IS_ERR(imx214->enable_gpio))
 		return dev_err_probe(dev, PTR_ERR(imx214->enable_gpio),
