--- a/drivers/interconnect/qcom/sc8280xp.c
+++ b/drivers/interconnect/qcom/sc8280xp.c
@@ -21,6 +21,11 @@ static struct qcom_icc_node qhm_qspi = {
 	.id = SC8280XP_MASTER_QSPI_0,
 	.channels = 1,
 	.buswidth = 4,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x5000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_A1NOC_SNOC },
 };
@@ -30,6 +35,11 @@ static struct qcom_icc_node qhm_qup1 = {
 	.id = SC8280XP_MASTER_QUP_1,
 	.channels = 1,
 	.buswidth = 4,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x6000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_A1NOC_SNOC },
 };
@@ -39,6 +49,11 @@ static struct qcom_icc_node qhm_qup2 = {
 	.id = SC8280XP_MASTER_QUP_2,
 	.channels = 1,
 	.buswidth = 4,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xc000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_A1NOC_SNOC },
 };
@@ -57,6 +72,12 @@ static struct qcom_icc_node qxm_ipa = {
 	.id = SC8280XP_MASTER_IPA,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x8000 },
+		.prio = 2,
+		.urg_fwd = 1,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_A1NOC_SNOC },
 };
@@ -66,6 +87,11 @@ static struct qcom_icc_node xm_emac_1 =
 	.id = SC8280XP_MASTER_EMAC_1,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xd000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_A1NOC_SNOC },
 };
@@ -75,6 +101,11 @@ static struct qcom_icc_node xm_sdc4 = {
 	.id = SC8280XP_MASTER_SDCC_4,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x7000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_A1NOC_SNOC },
 };
@@ -84,6 +115,11 @@ static struct qcom_icc_node xm_ufs_mem =
 	.id = SC8280XP_MASTER_UFS_MEM,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x9000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_A1NOC_SNOC },
 };
@@ -93,6 +129,11 @@ static struct qcom_icc_node xm_usb3_0 =
 	.id = SC8280XP_MASTER_USB3_0,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xb000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_USB_NOC_SNOC },
 };
@@ -102,6 +143,11 @@ static struct qcom_icc_node xm_usb3_1 =
 	.id = SC8280XP_MASTER_USB3_1,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xa000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_USB_NOC_SNOC },
 };
@@ -111,6 +157,11 @@ static struct qcom_icc_node xm_usb3_mp =
 	.id = SC8280XP_MASTER_USB3_MP,
 	.channels = 1,
 	.buswidth = 16,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xe000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_USB_NOC_SNOC },
 };
@@ -120,6 +171,11 @@ static struct qcom_icc_node xm_usb4_host
 	.id = SC8280XP_MASTER_USB4_0,
 	.channels = 1,
 	.buswidth = 16,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xf000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_USB_NOC_SNOC },
 };
@@ -129,6 +185,11 @@ static struct qcom_icc_node xm_usb4_host
 	.id = SC8280XP_MASTER_USB4_1,
 	.channels = 1,
 	.buswidth = 16,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x10000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_USB_NOC_SNOC },
 };
@@ -138,6 +199,11 @@ static struct qcom_icc_node qhm_qdss_bam
 	.id = SC8280XP_MASTER_QDSS_BAM,
 	.channels = 1,
 	.buswidth = 4,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xa000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_A2NOC_SNOC },
 };
@@ -147,6 +213,11 @@ static struct qcom_icc_node qhm_qup0 = {
 	.id = SC8280XP_MASTER_QUP_0,
 	.channels = 1,
 	.buswidth = 4,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xb000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_A2NOC_SNOC },
 };
@@ -165,6 +236,12 @@ static struct qcom_icc_node qxm_crypto =
 	.id = SC8280XP_MASTER_CRYPTO,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xe000 },
+		.prio = 2,
+		.urg_fwd = 1,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_A2NOC_SNOC },
 };
@@ -192,6 +269,11 @@ static struct qcom_icc_node xm_emac_0 =
 	.id = SC8280XP_MASTER_EMAC,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x11000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_A2NOC_SNOC },
 };
@@ -201,6 +283,11 @@ static struct qcom_icc_node xm_pcie3_0 =
 	.id = SC8280XP_MASTER_PCIE_0,
 	.channels = 1,
 	.buswidth = 16,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x12000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_ANOC_PCIE_GEM_NOC },
 };
@@ -210,6 +297,11 @@ static struct qcom_icc_node xm_pcie3_1 =
 	.id = SC8280XP_MASTER_PCIE_1,
 	.channels = 1,
 	.buswidth = 16,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xf000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_ANOC_PCIE_GEM_NOC },
 };
@@ -219,6 +311,11 @@ static struct qcom_icc_node xm_pcie3_2a
 	.id = SC8280XP_MASTER_PCIE_2A,
 	.channels = 1,
 	.buswidth = 16,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x13000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_ANOC_PCIE_GEM_NOC },
 };
@@ -228,6 +325,11 @@ static struct qcom_icc_node xm_pcie3_2b
 	.id = SC8280XP_MASTER_PCIE_2B,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x14000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_ANOC_PCIE_GEM_NOC },
 };
@@ -237,6 +339,11 @@ static struct qcom_icc_node xm_pcie3_3a
 	.id = SC8280XP_MASTER_PCIE_3A,
 	.channels = 1,
 	.buswidth = 16,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x15000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_ANOC_PCIE_GEM_NOC },
 };
@@ -246,6 +353,11 @@ static struct qcom_icc_node xm_pcie3_3b
 	.id = SC8280XP_MASTER_PCIE_3B,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x16000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_ANOC_PCIE_GEM_NOC },
 };
@@ -255,6 +367,11 @@ static struct qcom_icc_node xm_pcie3_4 =
 	.id = SC8280XP_MASTER_PCIE_4,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x10000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_ANOC_PCIE_GEM_NOC },
 };
@@ -264,6 +381,11 @@ static struct qcom_icc_node xm_qdss_etr
 	.id = SC8280XP_MASTER_QDSS_ETR,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x7000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_A2NOC_SNOC },
 };
@@ -273,6 +395,11 @@ static struct qcom_icc_node xm_sdc2 = {
 	.id = SC8280XP_MASTER_SDCC_2,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x8000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_A2NOC_SNOC },
 };
@@ -282,6 +409,11 @@ static struct qcom_icc_node xm_ufs_card
 	.id = SC8280XP_MASTER_UFS_CARD,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x9000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_A2NOC_SNOC },
 };
@@ -430,6 +562,11 @@ static struct qcom_icc_node alm_gpu_tcu
 	.id = SC8280XP_MASTER_GPU_TCU,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xa0000 },
+		.prio = 1,
+	},
 	.num_links = 2,
 	.links = { SC8280XP_SLAVE_GEM_NOC_CNOC,
 		   SC8280XP_SLAVE_LLCC
@@ -441,6 +578,11 @@ static struct qcom_icc_node alm_pcie_tcu
 	.id = SC8280XP_MASTER_PCIE_TCU,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xa1000 },
+		.prio = 3,
+	},
 	.num_links = 2,
 	.links = { SC8280XP_SLAVE_GEM_NOC_CNOC,
 		   SC8280XP_SLAVE_LLCC
@@ -452,6 +594,11 @@ static struct qcom_icc_node alm_sys_tcu
 	.id = SC8280XP_MASTER_SYS_TCU,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xa2000 },
+		.prio = 6,
+	},
 	.num_links = 2,
 	.links = { SC8280XP_SLAVE_GEM_NOC_CNOC,
 		   SC8280XP_SLAVE_LLCC
@@ -475,6 +622,11 @@ static struct qcom_icc_node qnm_cmpnoc0
 	.id = SC8280XP_MASTER_COMPUTE_NOC,
 	.channels = 2,
 	.buswidth = 32,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 2,
+		.port_offsets = { 0x1b000, 0x66000 },
+		.urg_fwd = 1,
+	},
 	.num_links = 2,
 	.links = { SC8280XP_SLAVE_GEM_NOC_CNOC,
 		   SC8280XP_SLAVE_LLCC
@@ -486,6 +638,11 @@ static struct qcom_icc_node qnm_cmpnoc1
 	.id = SC8280XP_MASTER_COMPUTE_NOC_1,
 	.channels = 2,
 	.buswidth = 32,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 2,
+		.port_offsets = { 0x24000, 0x6f000 },
+		.urg_fwd = 1,
+	},
 	.num_links = 2,
 	.links = { SC8280XP_SLAVE_GEM_NOC_CNOC,
 		   SC8280XP_SLAVE_LLCC
@@ -509,6 +666,11 @@ static struct qcom_icc_node qnm_gpu = {
 	.id = SC8280XP_MASTER_GFX3D,
 	.channels = 4,
 	.buswidth = 32,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 4,
+		.port_offsets = { 0x1c000, 0x1d000, 0x67000, 0x68000 },
+		.urg_fwd = 1,
+	},
 	.num_links = 2,
 	.links = { SC8280XP_SLAVE_GEM_NOC_CNOC,
 		   SC8280XP_SLAVE_LLCC
@@ -520,6 +682,11 @@ static struct qcom_icc_node qnm_mnoc_hf
 	.id = SC8280XP_MASTER_MNOC_HF_MEM_NOC,
 	.channels = 2,
 	.buswidth = 32,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 2,
+		.port_offsets = { 0x1e000, 0x69000 },
+		.urg_fwd = 1,
+	},
 	.num_links = 2,
 	.links = { SC8280XP_SLAVE_LLCC,
 		   SC8280XP_SLAVE_GEM_NOC_PCIE_CNOC
@@ -531,6 +698,11 @@ static struct qcom_icc_node qnm_mnoc_sf
 	.id = SC8280XP_MASTER_MNOC_SF_MEM_NOC,
 	.channels = 2,
 	.buswidth = 32,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 2,
+		.port_offsets = { 0x1f000, 0x6a000 },
+		.urg_fwd = 1,
+	},
 	.num_links = 2,
 	.links = { SC8280XP_SLAVE_GEM_NOC_CNOC,
 		   SC8280XP_SLAVE_LLCC
@@ -542,6 +714,12 @@ static struct qcom_icc_node qnm_pcie = {
 	.id = SC8280XP_MASTER_ANOC_PCIE_GEM_NOC,
 	.channels = 1,
 	.buswidth = 32,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xa3000 },
+		.prio = 2,
+		.urg_fwd = 1,
+	},
 	.num_links = 2,
 	.links = { SC8280XP_SLAVE_GEM_NOC_CNOC,
 		   SC8280XP_SLAVE_LLCC
@@ -553,6 +731,11 @@ static struct qcom_icc_node qnm_snoc_gc
 	.id = SC8280XP_MASTER_SNOC_GC_MEM_NOC,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xa4000 },
+		.urg_fwd = 1,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_LLCC },
 };
@@ -562,6 +745,11 @@ static struct qcom_icc_node qnm_snoc_sf
 	.id = SC8280XP_MASTER_SNOC_SF_MEM_NOC,
 	.channels = 1,
 	.buswidth = 16,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xa5000 },
+		.urg_fwd = 1,
+	},
 	.num_links = 3,
 	.links = { SC8280XP_SLAVE_GEM_NOC_CNOC,
 		   SC8280XP_SLAVE_LLCC,
@@ -610,6 +798,11 @@ static struct qcom_icc_node qnm_camnoc_h
 	.id = SC8280XP_MASTER_CAMNOC_HF,
 	.channels = 2,
 	.buswidth = 32,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 2,
+		.port_offsets = { 0x10000, 0x10180 },
+		.urg_fwd = 1,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_MNOC_HF_MEM_NOC },
 };
@@ -619,6 +812,11 @@ static struct qcom_icc_node qnm_mdp0_0 =
 	.id = SC8280XP_MASTER_MDP0,
 	.channels = 1,
 	.buswidth = 32,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x16000 },
+		.urg_fwd = 1,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_MNOC_HF_MEM_NOC },
 };
@@ -628,6 +826,11 @@ static struct qcom_icc_node qnm_mdp0_1 =
 	.id = SC8280XP_MASTER_MDP1,
 	.channels = 1,
 	.buswidth = 32,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x16080 },
+		.urg_fwd = 1,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_MNOC_HF_MEM_NOC },
 };
@@ -637,6 +840,11 @@ static struct qcom_icc_node qnm_mdp1_0 =
 	.id = SC8280XP_MASTER_MDP_CORE1_0,
 	.channels = 1,
 	.buswidth = 32,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xf000 },
+		.urg_fwd = 1,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_MNOC_HF_MEM_NOC },
 };
@@ -646,6 +854,11 @@ static struct qcom_icc_node qnm_mdp1_1 =
 	.id = SC8280XP_MASTER_MDP_CORE1_1,
 	.channels = 1,
 	.buswidth = 32,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xf800 },
+		.urg_fwd = 1,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_MNOC_HF_MEM_NOC },
 };
@@ -664,6 +877,11 @@ static struct qcom_icc_node qnm_rot_0 =
 	.id = SC8280XP_MASTER_ROTATOR,
 	.channels = 1,
 	.buswidth = 32,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x17000 },
+		.urg_fwd = 1,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_MNOC_SF_MEM_NOC },
 };
@@ -673,6 +891,11 @@ static struct qcom_icc_node qnm_rot_1 =
 	.id = SC8280XP_MASTER_ROTATOR_1,
 	.channels = 1,
 	.buswidth = 32,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xe000 },
+		.urg_fwd = 1,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_MNOC_SF_MEM_NOC },
 };
@@ -682,6 +905,11 @@ static struct qcom_icc_node qnm_video0 =
 	.id = SC8280XP_MASTER_VIDEO_P0,
 	.channels = 1,
 	.buswidth = 32,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x14000 },
+		.urg_fwd = 1,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_MNOC_SF_MEM_NOC },
 };
@@ -691,6 +919,11 @@ static struct qcom_icc_node qnm_video1 =
 	.id = SC8280XP_MASTER_VIDEO_P1,
 	.channels = 1,
 	.buswidth = 32,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x14080 },
+		.urg_fwd = 1,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_MNOC_SF_MEM_NOC },
 };
@@ -700,6 +933,11 @@ static struct qcom_icc_node qnm_video_cv
 	.id = SC8280XP_MASTER_VIDEO_PROC,
 	.channels = 1,
 	.buswidth = 32,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x15000 },
+		.urg_fwd = 1,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_MNOC_SF_MEM_NOC },
 };
@@ -709,6 +947,11 @@ static struct qcom_icc_node qxm_camnoc_i
 	.id = SC8280XP_MASTER_CAMNOC_ICP,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x11000 },
+		.urg_fwd = 1,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_MNOC_SF_MEM_NOC },
 };
@@ -718,6 +961,11 @@ static struct qcom_icc_node qxm_camnoc_s
 	.id = SC8280XP_MASTER_CAMNOC_SF,
 	.channels = 1,
 	.buswidth = 32,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x12000 },
+		.urg_fwd = 1,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_MNOC_SF_MEM_NOC },
 };
@@ -794,6 +1042,11 @@ static struct qcom_icc_node qnm_lpass_no
 	.id = SC8280XP_MASTER_LPASS_ANOC,
 	.channels = 1,
 	.buswidth = 16,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0xe000 },
+		.urg_fwd = 1,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_SNOC_GEM_NOC_SF },
 };
@@ -812,6 +1065,12 @@ static struct qcom_icc_node qxm_pimem =
 	.id = SC8280XP_MASTER_PIMEM,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x10000 },
+		.prio = 2,
+		.urg_fwd = 1,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_SNOC_GEM_NOC_GC },
 };
@@ -821,6 +1080,11 @@ static struct qcom_icc_node xm_gic = {
 	.id = SC8280XP_MASTER_GIC,
 	.channels = 1,
 	.buswidth = 8,
+	.qosbox = &(const struct qcom_icc_qosbox) {
+		.num_ports = 1,
+		.port_offsets = { 0x12000 },
+		.prio = 2,
+	},
 	.num_links = 1,
 	.links = { SC8280XP_SLAVE_SNOC_GEM_NOC_GC },
 };
@@ -2014,11 +2278,21 @@ static struct qcom_icc_node * const aggr
 	[SLAVE_SERVICE_A1NOC] = &srvc_aggre1_noc,
 };
 
+static const struct regmap_config sc8280xp_aggre1_noc_regmap_config = {
+	.reg_bits = 32,
+	.reg_stride = 4,
+	.val_bits = 32,
+	.max_register = 0x3af80,
+	.fast_io = true,
+};
+
 static const struct qcom_icc_desc sc8280xp_aggre1_noc = {
+	.config = &sc8280xp_aggre1_noc_regmap_config,
 	.nodes = aggre1_noc_nodes,
 	.num_nodes = ARRAY_SIZE(aggre1_noc_nodes),
 	.bcms = aggre1_noc_bcms,
 	.num_bcms = ARRAY_SIZE(aggre1_noc_bcms),
+	.qos_requires_clocks = true,
 };
 
 static struct qcom_icc_bcm * const aggre2_noc_bcms[] = {
@@ -2050,11 +2324,21 @@ static struct qcom_icc_node * const aggr
 	[SLAVE_SERVICE_A2NOC] = &srvc_aggre2_noc,
 };
 
+static const struct regmap_config sc8280xp_aggre2_noc_regmap_config = {
+	.reg_bits = 32,
+	.reg_stride = 4,
+	.val_bits = 32,
+	.max_register = 0x3af80,
+	.fast_io = true,
+};
+
 static const struct qcom_icc_desc sc8280xp_aggre2_noc = {
+	.config = &sc8280xp_aggre2_noc_regmap_config,
 	.nodes = aggre2_noc_nodes,
 	.num_nodes = ARRAY_SIZE(aggre2_noc_nodes),
 	.bcms = aggre2_noc_bcms,
 	.num_bcms = ARRAY_SIZE(aggre2_noc_bcms),
+	.qos_requires_clocks = true,
 };
 
 static struct qcom_icc_bcm * const clk_virt_bcms[] = {
@@ -2176,7 +2460,16 @@ static struct qcom_icc_node * const conf
 	[SLAVE_TCU] = &xs_sys_tcu_cfg,
 };
 
+static const struct regmap_config sc8280xp_config_noc_regmap_config = {
+	.reg_bits = 32,
+	.reg_stride = 4,
+	.val_bits = 32,
+	.max_register = 0x2c000,
+	.fast_io = true,
+};
+
 static const struct qcom_icc_desc sc8280xp_config_noc = {
+	.config = &sc8280xp_config_noc_regmap_config,
 	.nodes = config_noc_nodes,
 	.num_nodes = ARRAY_SIZE(config_noc_nodes),
 	.bcms = config_noc_bcms,
@@ -2192,7 +2485,16 @@ static struct qcom_icc_node * const dc_n
 	[SLAVE_GEM_NOC_CFG] = &qns_gemnoc,
 };
 
+static const struct regmap_config sc8280xp_dc_noc_regmap_config = {
+	.reg_bits = 32,
+	.reg_stride = 4,
+	.val_bits = 32,
+	.max_register = 0x5080,
+	.fast_io = true,
+};
+
 static const struct qcom_icc_desc sc8280xp_dc_noc = {
+	.config = &sc8280xp_dc_noc_regmap_config,
 	.nodes = dc_noc_nodes,
 	.num_nodes = ARRAY_SIZE(dc_noc_nodes),
 	.bcms = dc_noc_bcms,
@@ -2226,7 +2528,16 @@ static struct qcom_icc_node * const gem_
 	[SLAVE_SERVICE_GEM_NOC] = &srvc_sys_gemnoc,
 };
 
+static const struct regmap_config sc8280xp_gem_noc_regmap_config = {
+	.reg_bits = 32,
+	.reg_stride = 4,
+	.val_bits = 32,
+	.max_register = 0xb8400,
+	.fast_io = true,
+};
+
 static const struct qcom_icc_desc sc8280xp_gem_noc = {
+	.config = &sc8280xp_gem_noc_regmap_config,
 	.nodes = gem_noc_nodes,
 	.num_nodes = ARRAY_SIZE(gem_noc_nodes),
 	.bcms = gem_noc_bcms,
@@ -2249,7 +2560,16 @@ static struct qcom_icc_node * const lpas
 	[SLAVE_SERVICE_LPASS_AG_NOC] = &srvc_niu_lpass_agnoc,
 };
 
+static const struct regmap_config sc8280xp_lpass_ag_noc_regmap_config = {
+	.reg_bits = 32,
+	.reg_stride = 4,
+	.val_bits = 32,
+	.max_register = 0xf080,
+	.fast_io = true,
+};
+
 static const struct qcom_icc_desc sc8280xp_lpass_ag_noc = {
+	.config = &sc8280xp_lpass_ag_noc_regmap_config,
 	.nodes = lpass_ag_noc_nodes,
 	.num_nodes = ARRAY_SIZE(lpass_ag_noc_nodes),
 	.bcms = lpass_ag_noc_bcms,
@@ -2297,7 +2617,16 @@ static struct qcom_icc_node * const mmss
 	[SLAVE_SERVICE_MNOC] = &srvc_mnoc,
 };
 
+static const struct regmap_config sc8280xp_mmss_noc_regmap_config = {
+	.reg_bits = 32,
+	.reg_stride = 4,
+	.val_bits = 32,
+	.max_register = 0x1fa80,
+	.fast_io = true,
+};
+
 static const struct qcom_icc_desc sc8280xp_mmss_noc = {
+	.config = &sc8280xp_mmss_noc_regmap_config,
 	.nodes = mmss_noc_nodes,
 	.num_nodes = ARRAY_SIZE(mmss_noc_nodes),
 	.bcms = mmss_noc_bcms,
@@ -2317,7 +2646,16 @@ static struct qcom_icc_node * const nspa
 	[SLAVE_SERVICE_NSP_NOC] = &service_nsp_noc,
 };
 
+static const struct regmap_config sc8280xp_nspa_noc_regmap_config = {
+	.reg_bits = 32,
+	.reg_stride = 4,
+	.val_bits = 32,
+	.max_register = 0x10000,
+	.fast_io = true,
+};
+
 static const struct qcom_icc_desc sc8280xp_nspa_noc = {
+	.config = &sc8280xp_nspa_noc_regmap_config,
 	.nodes = nspa_noc_nodes,
 	.num_nodes = ARRAY_SIZE(nspa_noc_nodes),
 	.bcms = nspa_noc_bcms,
@@ -2337,7 +2675,16 @@ static struct qcom_icc_node * const nspb
 	[SLAVE_SERVICE_NSPB_NOC] = &service_nspb_noc,
 };
 
+static const struct regmap_config sc8280xp_nspb_noc_regmap_config = {
+	.reg_bits = 32,
+	.reg_stride = 4,
+	.val_bits = 32,
+	.max_register = 0x10000,
+	.fast_io = true,
+};
+
 static const struct qcom_icc_desc sc8280xp_nspb_noc = {
+	.config = &sc8280xp_nspb_noc_regmap_config,
 	.nodes = nspb_noc_nodes,
 	.num_nodes = ARRAY_SIZE(nspb_noc_nodes),
 	.bcms = nspb_noc_bcms,
@@ -2366,7 +2713,16 @@ static struct qcom_icc_node * const syst
 	[SLAVE_SERVICE_SNOC] = &srvc_snoc,
 };
 
+static const struct regmap_config sc8280xp_system_noc_regmap_config = {
+	.reg_bits = 32,
+	.reg_stride = 4,
+	.val_bits = 32,
+	.max_register = 0x1a400,
+	.fast_io = true,
+};
+
 static const struct qcom_icc_desc sc8280xp_system_noc_main = {
+	.config = &sc8280xp_system_noc_regmap_config,
 	.nodes = system_noc_main_nodes,
 	.num_nodes = ARRAY_SIZE(system_noc_main_nodes),
 	.bcms = system_noc_main_bcms,
