From 4988c98fd3079603a4edb6179b497b3742a05ce4 Mon Sep 17 00:00:00 2001
From: Jiali Chen <chenjiali@radxa.com>
Date: Wed, 9 Jul 2025 03:33:32 +0000
Subject: [PATCH 156/268] drm/edid: add device-tree quirk flags supports

Signed-off-by: ZHANG Yuntian <yt@radxa.com>
Signed-off-by: Jiali Chen <chenjiali@radxa.com>
---
 drivers/gpu/drm/drm_edid.c | 34 +++++++++++++++++++++++++++++++++-
 1 file changed, 33 insertions(+), 1 deletion(-)

--- a/drivers/gpu/drm/drm_edid.c
+++ b/drivers/gpu/drm/drm_edid.c
@@ -40,6 +40,8 @@
 #include <linux/seq_buf.h>
 #include <linux/slab.h>
 #include <linux/vga_switcheroo.h>
+#include <linux/of.h>
+#include <linux/of_device.h>
 
 #include <drm/drm_drv.h>
 #include <drm/drm_edid.h>
@@ -96,6 +98,8 @@ enum drm_edid_internal_quirk {
 	EDID_QUIRK_NON_DESKTOP,
 	/* Cap the DSC target bitrate to 15bpp */
 	EDID_QUIRK_CAP_DSC_15BPP,
+	/* Prefer Full HD */
+	CONNECTOR_QUIRK_PREFER_FHD
 };
 
 #define MICROSOFT_IEEE_OUI	0xca125c
@@ -2962,6 +2966,26 @@ static u32 edid_get_quirks(const struct
 	return 0;
 }
 
+static u32 connector_get_quirks(struct drm_connector *connector)
+{
+	struct device_node *np = connector->dev->dev->of_node;
+	u32 quirks = 0;
+
+	if (!np)
+		return 0;
+
+	if (of_property_read_bool(np, "quirk-prefer-fhd"))
+		quirks |= BIT(CONNECTOR_QUIRK_PREFER_FHD);
+
+	if (of_property_read_bool(np, "quirk-prefer-large-60"))
+		quirks |= BIT(EDID_QUIRK_PREFER_LARGE_60);
+
+	if (of_property_read_bool(np, "quirk-prefer-large-75"))
+		quirks |= BIT(EDID_QUIRK_PREFER_LARGE_75);
+
+	return quirks;
+}
+
 static bool drm_edid_has_internal_quirk(struct drm_connector *connector,
 					enum drm_edid_internal_quirk quirk)
 {
@@ -3004,6 +3028,14 @@ static void edid_fixup_preferred(struct
 		if (cur_mode == preferred_mode)
 			continue;
 
+		if (drm_edid_has_internal_quirk(connector, CONNECTOR_QUIRK_PREFER_FHD) &&
+			MODE_SIZE(cur_mode) > 1920 * 1080)
+			continue;
+
+		if (drm_edid_has_internal_quirk(connector, CONNECTOR_QUIRK_PREFER_FHD) &&
+			MODE_SIZE(preferred_mode) > 1920 * 1080)
+			preferred_mode = cur_mode;
+
 		/* Largest mode is preferred */
 		if (MODE_SIZE(cur_mode) > MODE_SIZE(preferred_mode))
 			preferred_mode = cur_mode;
@@ -6682,7 +6714,7 @@ static void update_display_info(struct d
 
 	edid = drm_edid->edid;
 
-	info->quirks = edid_get_quirks(drm_edid);
+	info->quirks = edid_get_quirks(drm_edid) | connector_get_quirks(connector);
 
 	info->width_mm = edid->width_cm * 10;
 	info->height_mm = edid->height_cm * 10;
