diff -Naur /tmp/qwrt-ufs-ah8/a/drivers/ufs/host/ufs-qcom.c /tmp/qwrt-ufs-ah8/b/drivers/ufs/host/ufs-qcom.c
--- a/drivers/ufs/host/ufs-qcom.c	2026-07-01 20:52:36.035046328 +0800
+++ b/drivers/ufs/host/ufs-qcom.c	2026-07-01 20:52:47.425142834 +0800
@@ -683,6 +683,13 @@
 	return 0;
 }
 
+static void ufs_qcom_link_startup_post_change(struct ufs_hba *hba)
+{
+	if (ufshcd_is_auto_hibern8_supported(hba))
+		ufshcd_rmwl(hba, UFS_HW_CLK_CTRL_EN, UFS_HW_CLK_CTRL_EN,
+			    UFS_AH8_CFG);
+}
+
 static int ufs_qcom_link_startup_notify(struct ufs_hba *hba,
 					enum ufs_notify_change_status status)
 {
@@ -709,6 +716,9 @@
 		err = ufshcd_disable_host_tx_lcc(hba);
 
 		break;
+	case POST_CHANGE:
+		ufs_qcom_link_startup_post_change(hba);
+		break;
 	default:
 		break;
 	}
diff -Naur /tmp/qwrt-ufs-ah8/a/drivers/ufs/host/ufs-qcom.h /tmp/qwrt-ufs-ah8/b/drivers/ufs/host/ufs-qcom.h
--- a/drivers/ufs/host/ufs-qcom.h	2026-07-01 20:52:36.036072828 +0800
+++ b/drivers/ufs/host/ufs-qcom.h	2026-07-01 20:52:56.106323411 +0800
@@ -228,6 +228,17 @@
  */
 #define NUM_TX_R1W1 13
 
+/* bit definitions for UFS_AH8_CFG register */
+#define CC_UFS_SYS_CLK_REQ_EN		BIT(2)
+#define CC_UFS_ICE_CORE_CLK_REQ_EN	BIT(3)
+#define CC_UFS_UNIPRO_CORE_CLK_REQ_EN	BIT(4)
+#define CC_UFS_AUXCLK_REQ_EN		BIT(5)
+
+#define UFS_HW_CLK_CTRL_EN	(CC_UFS_SYS_CLK_REQ_EN |\
+				CC_UFS_ICE_CORE_CLK_REQ_EN |\
+				CC_UFS_UNIPRO_CORE_CLK_REQ_EN |\
+				CC_UFS_AUXCLK_REQ_EN)
+
 static inline void
 ufs_qcom_get_controller_revision(struct ufs_hba *hba,
 				 u8 *major, u16 *minor, u16 *step)
