--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -24,8 +24,6 @@ SET (LIBGD_SRC_FILES
 	gd_io_dp.c
 	gd_io_file.c
 	gd_io_ss.c
-	gd_io_stream.cxx
-	gd_io_stream.h
 	gd_jpeg.c
 	gd_matrix.c
 	gd_nnquant.c
@@ -61,8 +59,6 @@ SET (LIBGD_SRC_FILES
 	gdhelpers.c
 	gdhelpers.h
 	gdkanji.c
-	gdpp.cxx
-	gdpp.h
 	gdtables.c
 	gdxpm.c
 	jisx0208.h
@@ -203,7 +199,6 @@ install(FILES
 	gdfonts.h
 	gdfontt.h
 	gdfx.h
-	gdpp.h
 	DESTINATION include)
 
 CONFIGURE_FILE(../config/gdlib.pc.cmake gdlib.pc @ONLY)
