--- a/drivers/net/ethernet/atheros/Kconfig
+++ b/drivers/net/ethernet/atheros/Kconfig
@@ -17,13 +17,7 @@ config NET_VENDOR_ATHEROS
 
 if NET_VENDOR_ATHEROS
 
-config AG71XX
-	tristate "Atheros AR7XXX/AR9XXX built-in ethernet mac support"
-	depends on ATH79
-	select PHYLINK
-	help
-	  If you wish to compile a kernel for AR7XXX/91XXX and enable
-	  ethernet support, then you should always answer Y to this.
+source "drivers/net/ethernet/atheros/ag71xx/Kconfig"
 
 config ATL2
 	tristate "Atheros L2 Fast Ethernet support"
--- a/drivers/net/ethernet/atheros/Makefile
+++ b/drivers/net/ethernet/atheros/Makefile
@@ -3,7 +3,7 @@
 # Makefile for the Atheros network device drivers.
 #
 
-obj-$(CONFIG_AG71XX) += ag71xx.o
+obj-$(CONFIG_AG71XX) += ag71xx/
 obj-$(CONFIG_ATL1) += atlx/
 obj-$(CONFIG_ATL2) += atlx/
 obj-$(CONFIG_ATL1E) += atl1e/
