--- a/src/qgen/Makefile.am
+++ b/src/qgen/Makefile.am
@@ -2,7 +2,7 @@ noinst_PROGRAMS = qgen
 
 qgen_SOURCES = common.c common.h file.c file.h first.c ql_y.y ql_l.l qgen.c \
 		qgen.h second.c third.c
-qgen_LDADD = -lfl
+qgen_LDADD = 
 
 COMPILE = @CC_FOR_BUILD@ @CFLAGS_FOR_BUILD@
 LINK = @CC_FOR_BUILD@ @CFLAGS_FOR_BUILD@ -o $@
--- a/src/qgen/Makefile.in
+++ b/src/qgen/Makefile.in
@@ -204,7 +204,7 @@ top_srcdir = @top_srcdir@
 qgen_SOURCES = common.c common.h file.c file.h first.c ql_y.y ql_l.l qgen.c \
 		qgen.h second.c third.c
 
-qgen_LDADD = -lfl
+qgen_LDADD =
 COMPILE = @CC_FOR_BUILD@ @CFLAGS_FOR_BUILD@
 LINK = @CC_FOR_BUILD@ @CFLAGS_FOR_BUILD@ -o $@
 
--- a/src/sigd/Makefile.am
+++ b/src/sigd/Makefile.am
@@ -8,7 +8,7 @@ atmsigd_XTRAS = mess.o $(top_builddir)/s
 			$(top_builddir)/src/q2931/qd.dump.o \
 			$(top_builddir)/src/lib/libatm.la \
 			$(top_builddir)/src/saal/libsaal.a
-atmsigd_LDADD = $(atmsigd_XTRAS) -lfl
+atmsigd_LDADD = $(atmsigd_XTRAS) 
 atmsigd_DEPENDENCIES = mess.c $(atmsigd_XTRAS)
 
 CLEANFILES = mess.c
--- a/src/sigd/Makefile.in
+++ b/src/sigd/Makefile.in
@@ -245,7 +245,7 @@ atmsigd_XTRAS = mess.o $(top_builddir)/s
 			$(top_builddir)/src/lib/libatm.la \
 			$(top_builddir)/src/saal/libsaal.a
 
-atmsigd_LDADD = $(atmsigd_XTRAS) -lfl
+atmsigd_LDADD = $(atmsigd_XTRAS) 
 atmsigd_DEPENDENCIES = mess.c $(atmsigd_XTRAS)
 CLEANFILES = mess.c
 sysconf_DATA = atmsigd.conf
--- a/src/switch/debug/debug.c
+++ b/src/switch/debug/debug.c
@@ -20,6 +20,11 @@
 
 #define PRV(call) ((FAB *) (call)->fab)
 
+int yywrap(void)
+{
+        return 1;
+}
+
 
 typedef struct _fab {
     CALL *next; /* relay.c may not keep track of calls, but WE are */
--- a/src/switch/debug/Makefile.am
+++ b/src/switch/debug/Makefile.am
@@ -5,7 +5,7 @@ INCLUDES = -I$(srcdir)/../../q2931
 sw_debug_SOURCES = debug.c
 sw_debug_XTRAS = $(top_builddir)/src/switch/libsw.a \
 			$(top_builddir)/src/lib/libatm.la
-sw_debug_LDADD = $(sw_debug_XTRAS) -lfl
+sw_debug_LDADD = $(sw_debug_XTRAS) 
 			
 sw_debug_DEPENDENCIES = $(sw_debug_XTRAS)
 
--- a/src/switch/debug/Makefile.in
+++ b/src/switch/debug/Makefile.in
@@ -200,7 +200,8 @@ sw_debug_SOURCES = debug.c
 sw_debug_XTRAS = $(top_builddir)/src/switch/libsw.a \
 			$(top_builddir)/src/lib/libatm.la
 
-sw_debug_LDADD = $(sw_debug_XTRAS) -lfl
+sw_debug_LDADD = $(sw_debug_XTRAS)
+
 sw_debug_DEPENDENCIES = $(sw_debug_XTRAS)
 EXTRA_DIST = demo README
 all: all-am
--- a/src/switch/tcp/Makefile.am
+++ b/src/switch/tcp/Makefile.am
@@ -5,7 +5,7 @@ INCLUDES = -I$(srcdir)/../../q2931
 sw_tcp_SOURCES = tcpsw.c
 sw_tcp_XTRAS = $(top_builddir)/src/switch/libsw.a \
 		$(top_builddir)/src/lib/libatm.la
-sw_tcp_LDADD = $(sw_tcp_XTRAS) -lfl
+sw_tcp_LDADD = $(sw_tcp_XTRAS) 
 sw_tcp_DEPENDENCIES = $(sw_tcp_XTRAS)
 
 EXTRA_DIST = mkfiles README
--- a/src/switch/tcp/Makefile.in
+++ b/src/switch/tcp/Makefile.in
@@ -200,7 +200,7 @@ sw_tcp_SOURCES = tcpsw.c
 sw_tcp_XTRAS = $(top_builddir)/src/switch/libsw.a \
 		$(top_builddir)/src/lib/libatm.la
 
-sw_tcp_LDADD = $(sw_tcp_XTRAS) -lfl
+sw_tcp_LDADD = $(sw_tcp_XTRAS) 
 sw_tcp_DEPENDENCIES = $(sw_tcp_XTRAS)
 EXTRA_DIST = mkfiles README
 all: all-am
--- a/src/switch/tcp/tcpsw.c
+++ b/src/switch/tcp/tcpsw.c
@@ -35,6 +35,10 @@
 #define MAX_PACKET (ATM_MAX_AAL5_PDU+sizeof(struct atmtcp_hdr))
 #define BUFFER_SIZE (MAX_PACKET*2)
 
+int yywrap(void)
+{
+        return 1;
+}
 
 typedef struct _table {
     struct _link *out;	/* output port */
--- a/src/test/Makefile.am
+++ b/src/test/Makefile.am
@@ -20,7 +20,7 @@ br_SOURCES = br.c
 bw_SOURCES = bw.c
 isp_SOURCES = isp.c isp.h ispl_y.y ispl_l.l
 isp_XTRAS = $(LDADD)
-isp_LDADD = $(isp_XTRAS) -lfl
+isp_LDADD = $(isp_XTRAS)
 isp_DEPENDENCIES = $(isp_XTRAS)
 window_SOURCES = window.c
 
--- a/src/test/Makefile.in
+++ b/src/test/Makefile.in
@@ -283,7 +283,7 @@ br_SOURCES = br.c
 bw_SOURCES = bw.c
 isp_SOURCES = isp.c isp.h ispl_y.y ispl_l.l
 isp_XTRAS = $(LDADD)
-isp_LDADD = $(isp_XTRAS) -lfl
+isp_LDADD = $(isp_XTRAS)
 isp_DEPENDENCIES = $(isp_XTRAS)
 window_SOURCES = window.c
 CLEANFILES = errnos.inc
--- a/src/test/ispl_l.l
+++ b/src/test/ispl_l.l
@@ -18,6 +18,11 @@
 #include "ispl_y.h"
 
 
+int yywrap(void)
+{
+       return 1;
+}
+
 static int lineno = 1;
 
 %}
--- a/src/qgen/ql_l.l
+++ b/src/qgen/ql_l.l
@@ -11,6 +11,11 @@
 #include "ql_y.h"
 
 
+int yywrap(void)
+{
+        return 1;
+}
+
 typedef struct _tree {
     struct _tree *left,*right;
     const char str[0];
--- a/src/sigd/cfg_l.l
+++ b/src/sigd/cfg_l.l
@@ -16,6 +16,10 @@
 
 #include "cfg_y.h"
 
+int yywrap(void)
+{
+        return 1;
+}
 
 static int lineno = 1;
 static int token; /* f@#%ing flex doesn't grok return after BEGIN */
