--- a/src/extra/Makefile.am
+++ b/src/extra/Makefile.am
@@ -7,6 +7,8 @@ EXTRA_DIST = linux-atm.spec.in \
 BUILT_SOURCES = pca200e.bin pca200e_ecd.bin2 sba200e_ecd.bin2
 CLEANFILES = pca200e.bin pca200e_ecd.bin2 sba200e_ecd.bin2
 
+OBJCOPY = objcopy
+
 install-exec-hook:
 	$(MKDIR_P) $(DESTDIR)/lib/firmware
 	$(INSTALL_DATA) $(srcdir)/pca200e.bin $(DESTDIR)/lib/firmware
@@ -14,7 +16,7 @@ install-exec-hook:
 	$(INSTALL_DATA) $(srcdir)/sba200e_ecd.bin2 $(DESTDIR)/lib/firmware
 
 %.bin %.bin2: %.data
-	objcopy -Iihex $< -Obinary $@.gz
+	$(OBJCOPY) -Iihex $< -Obinary $@.gz
 	gzip -n -df $@.gz
 
 
--- a/src/extra/Makefile.in
+++ b/src/extra/Makefile.in
@@ -187,6 +187,8 @@ CLEANFILES = pca200e.bin pca200e_ecd.bin
 all: $(BUILT_SOURCES)
 	$(MAKE) $(AM_MAKEFLAGS) all-am
 
+OBJCOPY = objcopy
+
 .SUFFIXES:
 $(srcdir)/Makefile.in:  $(srcdir)/Makefile.am  $(am__configure_deps)
 	@for dep in $?; do \
@@ -385,7 +387,7 @@ install-exec-hook:
 	$(INSTALL_DATA) $(srcdir)/sba200e_ecd.bin2 $(DESTDIR)/lib/firmware
 
 %.bin %.bin2: %.data
-	objcopy -Iihex $< -Obinary $@.gz
+	$(OBJCOPY) -Iihex $< -Obinary $@.gz
 	gzip -n -df $@.gz
 
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
