--- a/src/maint/saaldump.c	2020-03-29 22:58:01.089711789 +0200
+++ b/src/maint/saaldump.c	2020-03-29 22:59:17.564639387 +0200
@@ -6,6 +6,7 @@
 #include <config.h>
 #endif

+#include <linux/sockios.h>
 #include <stdlib.h>
 #include <stdarg.h>
 #include <stdio.h>
--- a/src/maint/atmdump.c	2020-03-29 22:58:18.573694469 +0200
+++ b/src/maint/atmdump.c	2020-03-29 22:58:49.956729365 +0200
@@ -6,6 +6,7 @@
 #include <config.h>
 #endif

+#include <linux/sockios.h>
 #include <stdlib.h>
 #include <stdio.h>
 #include <unistd.h>

