--- a/arch/arm/boot/dts/kirkwood-iconnect.dts
+++ b/arch/arm/boot/dts/kirkwood-iconnect.dts
@@ -16,8 +16,6 @@
 	chosen {
 		bootargs = "console=ttyS0,115200n8 earlyprintk";
 		stdout-path = &uart0;
-		linux,initrd-start = <0x4500040>;
-		linux,initrd-end   = <0x4800000>;
 	};
 
 	ocp@f1000000 {
@@ -146,28 +144,23 @@
 	status = "okay";
 
 	partition@0 {
-		label = "uboot";
-		reg = <0x0000000 0xc0000>;
+		label = "u-boot";
+		reg = <0x0000000 0xe0000>;
 	};
 
-	partition@a0000 {
-		label = "env";
-		reg = <0xa0000 0x20000>;
+	partition@e0000 {
+		label = "u-boot environment";
+		reg = <0xe0000 0x100000>;
 	};
 
 	partition@100000 {
-		label = "zImage";
-		reg = <0x100000 0x300000>;
+		label = "second stage u-boot";
+		reg = <0x100000 0x200000>;
 	};
 
-	partition@540000 {
-		label = "initrd";
-		reg = <0x540000 0x300000>;
-	};
-
-	partition@980000 {
-		label = "boot";
-		reg = <0x980000 0x1f400000>;
+	partition@200000 {
+		label = "ubi";
+		reg = <0x200000 0x1fe00000>;
 	};
 };
 
