--- a/arch/arm/boot/dts/kirkwood-pogoplug-series-4.dts
+++ b/arch/arm/boot/dts/kirkwood-pogoplug-series-4.dts
@@ -24,6 +24,7 @@
 	};
 
 	chosen {
+		bootargs = "console=ttyS0,115200";
 		stdout-path = "uart0:115200n8";
 	};
 
@@ -37,8 +38,8 @@
 		eject {
 			debounce-interval = <50>;
 			wakeup-source;
-			linux,code = <KEY_EJECTCD>;
-			label = "Eject Button";
+			linux,code = <KEY_RESTART>;
+			label = "Reset";
 			gpios = <&gpio0 29 GPIO_ACTIVE_LOW>;
 		};
 	};
@@ -137,29 +138,19 @@
 		#size-cells = <1>;
 
 		partition@0 {
-			label = "u-boot";
-			reg = <0x00000000 0x200000>;
+			label = "uboot";
+			reg = <0x00000000 0x1c0000>;
 			read-only;
 		};
 
-		partition@200000 {
-			label = "uImage";
-			reg = <0x00200000 0x300000>;
-		};
-
-		partition@500000 {
-			label = "uImage2";
-			reg = <0x00500000 0x300000>;
-		};
-
-		partition@800000 {
-			label = "failsafe";
-			reg = <0x00800000 0x800000>;
+		partition@1c0000 {
+			label = "uboot_env";
+			reg = <0x001c0000 0x40000>;
 		};
 
-		partition@1000000 {
-			label = "root";
-			reg = <0x01000000 0x7000000>;
+		partition@200000 {
+			label = "ubi";
+			reg = <0x00200000 0x7e00000>;
 		};
 	};
 };
