Index: refpolicy-2.20200229/Makefile
===================================================================
--- refpolicy-2.20200229.orig/Makefile
+++ refpolicy-2.20200229/Makefile
@@ -648,6 +648,6 @@ ifneq ($(generated_fc),)
 endif
 endif
 
-.PHONY: install-src install-appconfig install-headers generate xml conf html bare tags
+.PHONY: install-src install-appconfig install-headers generate conf bare tags
 .SUFFIXES:
 .SUFFIXES: .c
