--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
@@ -7112,6 +7112,7 @@ int stmmac_dvr_probe(struct device *devi
 {
 	struct net_device *ndev = NULL;
 	struct stmmac_priv *priv;
+	const char *devname = of_get_property(device->of_node, "label", NULL);
 	u32 rxq;
 	int i, ret = 0;
 
@@ -7120,6 +7121,9 @@ int stmmac_dvr_probe(struct device *devi
 	if (!ndev)
 		return -ENOMEM;
 
+	if (devname)
+		strlcpy(ndev->name, devname, IFNAMSIZ);
+
 	SET_NETDEV_DEV(ndev, device);
 
 	priv = netdev_priv(ndev);
