Index: u-boot-2017-01-29-6946ebba/fritz/fritzcreator.sh
===================================================================
--- u-boot-2017-01-29-6946ebba.orig/fritz/fritzcreator.sh
+++ u-boot-2017-01-29-6946ebba/fritz/fritzcreator.sh
@@ -1,4 +1,4 @@
-#!/bin/sh -xe
+#!/bin/bash -xe
 
 # A helpful
 #
@@ -11,6 +11,7 @@ UBOOT_FRITZ4040="uboot-fritz4040.bin"
 FRITZ_DTS="fritz/fritz4040.dts"
 FRITZ_DTB="fritz4040.dtb"
 UBOOT_LOADADDR=0x841FFFF8
+DTC="${DTC:-dtc}"
 
 rm -f "$UBOOT_FRITZ4040"
 
@@ -44,11 +45,11 @@ printf "\x00\x00\x00\xea\xf8\x01\x33\xc4
 cat "$UBOOT_BIN" >> "$UBOOT_FRITZ4040"
 
 # Pad file to 1M
-dd if="$UBOOT_FRITZ4040" of="$UBOOT_FRITZ4040.new" bs=1M count=1 conv=sync
+dd if="$UBOOT_FRITZ4040" of="$UBOOT_FRITZ4040.new" bs=1024k count=1 conv=sync
 mv "$UBOOT_FRITZ4040.new" "$UBOOT_FRITZ4040"
 
 # Compile DTS
-dtc "$FRITZ_DTS" -o "$FRITZ_DTB" --space 49152
+$DTC "$FRITZ_DTS" -o "$FRITZ_DTB" --space 49152
 
 # Append the compiled DTBs
 cat "$FRITZ_DTB" "$FRITZ_DTB" "$FRITZ_DTB" "$FRITZ_DTB" >> $UBOOT_FRITZ4040
