--- a/arch/mips/alchemy/board-mtx1.c
+++ b/arch/mips/alchemy/board-mtx1.c
@@ -98,6 +98,9 @@ void __init board_setup(void)
 	alchemy_gpio_direction_output(211, 1);	/* green on */
 	alchemy_gpio_direction_output(212, 0);	/* red off */
 
+	/* Set watchdog pin low */
+	alchemy_gpio_direction_output(215, 0);
+
 	pm_power_off = mtx1_power_off;
 	_machine_halt = mtx1_power_off;
 	_machine_restart = mtx1_reset;
