--- a/arch/mips/alchemy/devboards/db1300.c
+++ b/arch/mips/alchemy/devboards/db1300.c
@@ -712,6 +712,7 @@ static struct platform_device db1300_lcd
 
 /**********************************************************************/
 
+#ifdef CONFIG_TOUCHSCREEN_WM97XX
 static void db1300_wm97xx_irqen(struct wm97xx *wm, int enable)
 {
 	if (enable)
@@ -744,6 +745,15 @@ static int db1300_wm97xx_probe(struct pl
 	return wm97xx_register_mach_ops(wm, &db1300_wm97xx_ops);
 }
 
+#else
+
+static int db1300_wm97xx_probe(struct platform_device *pdev)
+{
+	return -1;
+}
+
+#endif
+
 static struct platform_driver db1300_wm97xx_driver = {
 	.driver.name	= "wm97xx-touch",
 	.driver.owner	= THIS_MODULE,
