--- a/drivers/net/wireless/ath/ath11k/pci.c
--- b/drivers/net/wireless/ath/ath11k/pci.c
@@ -427,7 +427,11 @@
 	pci_read_config_dword(pci_dev, pci_dev->msi_cap + PCI_MSI_ADDRESS_LO,
 			      &ab->pci.msi.addr_lo);
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6,1,0)
 	if (msi_desc->msi_attrib.is_64) {
+#else
+	if (msi_desc->pci.msi_attrib.is_64) {
+#endif
 		pci_read_config_dword(pci_dev, pci_dev->msi_cap + PCI_MSI_ADDRESS_HI,
 				      &ab->pci.msi.addr_hi);
 	} else {
--- a/drivers/net/wireless/ath/ath11k/pcic.c
+++ b/drivers/net/wireless/ath/ath11k/pcic.c
@@ -548,8 +548,13 @@
 		irq_grp->ab = ab;
 		irq_grp->grp_id = i;
 		init_dummy_netdev(&irq_grp->napi_ndev);
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6,1,0)
 		netif_napi_add(&irq_grp->napi_ndev, &irq_grp->napi,
 			       ath11k_pcic_ext_grp_napi_poll, NAPI_POLL_WEIGHT);
+#else
+		netif_napi_add_weight(&irq_grp->napi_ndev, &irq_grp->napi,
+			       ath11k_pcic_ext_grp_napi_poll, NAPI_POLL_WEIGHT);
+#endif
 
 		if (ab->hw_params.ring_mask->tx[i] ||
 		    ab->hw_params.ring_mask->rx[i] ||
--- a/net/qrtr/qrtr.c
+++ b/net/qrtr/qrtr.c
@@ -1035,8 +1035,12 @@
 		return -EADDRNOTAVAIL;
 	}
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6,1,0)
 	skb = skb_recv_datagram(sk, flags & ~MSG_DONTWAIT,
 				flags & MSG_DONTWAIT, &rc);
+#else
+	skb = skb_recv_datagram(sk, flags, &rc);
+#endif
 	if (!skb) {
 		release_sock(sk);
 		return rc;
