--- a/arch/arm/mach-ixp4xx/ixp4xx_npe.c
+++ b/arch/arm/mach-ixp4xx/ixp4xx_npe.c
@@ -586,6 +586,8 @@ int npe_load_firmware(struct npe *npe, c
 	npe_reset(npe);
 #endif
 
+	print_npe(KERN_INFO, npe, "firmware's license can be found in /usr/share/doc/LICENSE.IPL\n");
+
 	print_npe(KERN_INFO, npe, "firmware functionality 0x%X, "
 		  "revision 0x%X:%X\n", (image->id >> 16) & 0xFF,
 		  (image->id >> 8) & 0xFF, image->id & 0xFF);
