--- a/net/core/skbuff.c
+++ b/net/core/skbuff.c
@@ -215,6 +215,9 @@ struct sk_buff *__alloc_skb(unsigned int
 
 	if (sk_memalloc_socks() && (flags & SKB_ALLOC_RX))
 		gfp_mask |= __GFP_MEMALLOC;
+#ifdef CONFIG_ARCH_IXP4XX
+	gfp_mask |= GFP_DMA;
+#endif
 
 	/* Get the HEAD */
 	skb = kmem_cache_alloc_node(cache, gfp_mask & ~__GFP_DMA, node);
@@ -1222,6 +1225,10 @@ int pskb_expand_head(struct sk_buff *skb
 	if (skb_shared(skb))
 		BUG();
 
+#ifdef CONFIG_ARCH_IXP4XX
+	gfp_mask |= GFP_DMA;
+#endif
+
 	size = SKB_DATA_ALIGN(size);
 
 	if (skb_pfmemalloc(skb))
