--- a/meson.build
+++ b/meson.build
@@ -390,12 +390,11 @@ if dl_dep.found()
 endif
 
 have_iconv = false
-if cc.has_function('iconv_open')
+iconv_dep = cc.find_library('iconv', required : false)
+have_iconv = iconv_dep.found()
+if not have_iconv and cc.has_function('iconv_open')
   iconv_dep = dependency('', required : false)
   have_iconv = true
-else
-  iconv_dep = cc.find_library('iconv', required : false)
-  have_iconv = iconv_dep.found()
 endif
 if have_iconv
   cdata.set('HAVE_ICONV', 1)
