--- a/makeconfig
+++ b/makeconfig
@@ -414,47 +414,6 @@ int main(void)
 fi
 
 cat >$tmp2.c <<\!
-#include <gssapi/gssapi.h>
-
-int main(void)
-{
-	gss_import_name(0, 0, GSS_C_NT_HOSTBASED_SERVICE, 0);
-	gss_init_sec_context(0,0,0,0,0,0,0,0,0,0,0,0,0);
-	return 0;
-}
-!
-
-<$tmp2.c link_check gssapi 'for GSSAPI in libgss' \
-		'#define USE_GSSAPI' '-lgss' ||
-	<$tmp2.c link_check gssapi 'for GSSAPI in libgssapi_krb5' \
-			'#define USE_GSSAPI' '-lgssapi_krb5' ||
-		link_check gssapi 'for GSSAPI in libgssapi_krb5, old-style' \
-				'#define USE_GSSAPI
-#define GSSAPI_OLD_STYLE' '-lgssapi_krb5' <<\! || \
-			link_check gssapi 'for GSSAPI in libgssapi' \
-				'#define USE_GSSAPI
-#define	GSSAPI_REG_INCLUDE' '-lgssapi' <<\%
-#include <gssapi/gssapi.h>
-#include <gssapi/gssapi_generic.h>
-
-int main(void)
-{
-	gss_import_name(0, 0, gss_nt_service_name, 0);
-	gss_init_sec_context(0,0,0,0,0,0,0,0,0,0,0,0,0);
-	return 0;
-}
-!
-#include <gssapi.h>
-
-int main(void)
-{
-	gss_import_name(0, 0, GSS_C_NT_HOSTBASED_SERVICE, 0);
-	gss_init_sec_context(0,0,0,0,0,0,0,0,0,0,0,0,0);
-	return 0;
-}
-%
-
-cat >$tmp2.c <<\!
 #include "config.h"
 #ifdef HAVE_NL_LANGINFO
 #include <langinfo.h>
