--- a/arch/arm64/boot/dts/mediatek/mt7622.dtsi
+++ b/arch/arm64/boot/dts/mediatek/mt7622.dtsi
@@ -111,7 +111,7 @@
 	};
 
 	psci {
-		compatible  = "arm,psci-0.2";
+		compatible  = "arm,psci-1.0";
 		method      = "smc";
 	};
 
@@ -127,6 +127,13 @@
 		#size-cells = <2>;
 		ranges;
 
+		/* 64 KiB reserved for ramoops/pstore */
+		ramoops@0x42ff0000 {
+			compatible = "ramoops";
+			reg = <0 0x42ff0000 0 0x10000>;
+			record-size = <0x1000>;
+		};
+
 		/* 192 KiB reserved for ARM Trusted Firmware (BL31) */
 		secmon_reserved: secmon@43000000 {
 			reg = <0 0x43000000 0 0x30000>;
