--- a/arch/arm/boot/dts/kirkwood-ib62x0.dts
+++ b/arch/arm/boot/dts/kirkwood-ib62x0.dts
@@ -117,13 +117,13 @@
 	};
 
 	partition@100000 {
-		label = "uImage";
-		reg = <0x0100000 0x600000>;
+		label = "second stage u-boot";
+		reg = <0x100000 0x200000>;
 	};
 
-	partition@700000 {
-		label = "root";
-		reg = <0x0700000 0xf900000>;
+	partition@200000 {
+		label = "ubi";
+		reg = <0x200000 0xfe00000>;
 	};
 
 };
