--- a/arch/arm/boot/dts/kirkwood-dockstar.dts
+++ b/arch/arm/boot/dts/kirkwood-dockstar.dts
@@ -77,18 +77,22 @@
 
 	partition@0 {
 		label = "u-boot";
-		reg = <0x0000000 0x100000>;
-		read-only;
+		reg = <0x0000000 0xe0000>;
+	};
+
+	partition@e0000 {
+		label = "u-boot environment";
+		reg = <0xe0000 0x100000>;
 	};
 
 	partition@100000 {
-		label = "uImage";
-		reg = <0x0100000 0x400000>;
+		label = "second stage u-boot";
+		reg = <0x100000 0x200000>;
 	};
 
-	partition@500000 {
-		label = "data";
-		reg = <0x0500000 0xfb00000>;
+	partition@200000 {
+		label = "ubi";
+		reg = <0x200000 0xfe00000>;
 	};
 };
 
