--- a/src/pm/drv_dsl_cpe_api_pm_vrx.c
+++ b/src/pm/drv_dsl_cpe_api_pm_vrx.c
@@ -1435,9 +1435,16 @@ DSL_Error_t DSL_DRV_PM_DEV_ReTxCountersG
       /* ignore zero value*/
       if (nEftrMin)
       {
-         /* Fw Format:  kBit/s */
-         /* API format: bit/s */
-         pCounters->nEftrMin = nEftrMin*1000;
+         if (nDirection == DSL_NEAR_END)
+         {
+            /* Fw Format:  kBit/s */
+            /* API format: bit/s */
+            pCounters->nEftrMin = nEftrMin*1000;
+         }
+         else
+         {
+            pCounters->nEftrMin = nEftrMin;
+         }
       }
    }
    else
