--- a/configure.ac
+++ b/configure.ac
@@ -46,7 +46,7 @@ AC_GCC_TRADITIONAL
 AC_SYS_LARGEFILE
 
 AC_ISC_POSIX
-AM_C_PROTOTYPES
+AC_C_PROTOTYPES
 
 AC_CHECK_HEADERS(float.h fnmatch.h glob.h langinfo.h libintl.h mcheck.h unistd.h)
 
--- a/Makefile.am
+++ b/Makefile.am
@@ -35,9 +35,10 @@ tdict_LDADD = $(usrlib_LTLIBRARIES)
 noinst_SCRIPTS = testit.sh
 
 TESTS_ENVIRONMENT = \
+TESTS_DIR="$(top_srcdir)" \
 test1="$(top_builddir)/test1"
 
-TESTS = $(top_srcdir)/testit.sh
+TESTS = testit.sh
 
 include_HEADERS = popt.h
 
