--- a/iwinfo_cli.c
+++ b/iwinfo_cli.c
@@ -343,16 +343,36 @@
 	return buf;
 }
 
+static const uint16_t cli_ht_chan_width[] = {
+	20,
+	2040,
+};
+
+static const uint16_t cli_vht_chan_width[] = {
+	40,
+	80,
+	160,
+	8080,
+};
+
+static const uint16_t cli_eht_chan_width[] = {
+	20,
+	40,
+	80,
+	160,
+	320,
+};
+
 static const char* format_chan_width(bool vht, uint8_t width)
 {
-	if (!vht && width < ARRAY_SIZE(ht_chan_width))
-		switch (ht_chan_width[width]) {
+	if (!vht && width < ARRAY_SIZE(cli_ht_chan_width))
+		switch (cli_ht_chan_width[width]) {
 			case 20: return "20 MHz";
 			case 2040: return "40 MHz or higher";
 		}
 
-	if (vht && width < ARRAY_SIZE(vht_chan_width))
-		switch (vht_chan_width[width]) {
+	if (vht && width < ARRAY_SIZE(cli_vht_chan_width))
+		switch (cli_vht_chan_width[width]) {
 			case 40: return "20 or 40 MHz";
 			case 80: return "80 MHz";
 			case 8080: return "80+80 MHz";
@@ -364,8 +384,8 @@
 
 static const char* format_6ghz_chan_width(uint8_t width)
 {
-	if (width < ARRAY_SIZE(eht_chan_width))
-		switch (eht_chan_width[width]) {
+	if (width < ARRAY_SIZE(cli_eht_chan_width))
+		switch (cli_eht_chan_width[width]) {
 			case 20: return "20 MHz";
 			case 40: return "40 MHz";
 			case 80: return "80 MHz";
