--- a/hcxhashtool.c
+++ b/hcxhashtool.c
@@ -104,9 +104,6 @@ static void closelists()
 {
 if(hashlist != NULL) free(hashlist);
 if(ouilist != NULL) free(ouilist);
-EVP_cleanup();
-CRYPTO_cleanup_all_ex_data();
-ERR_free_strings();
 return;
 }
 /*===========================================================================*/
@@ -125,8 +122,6 @@ eapolwrittencount = 0;
 essidwrittencount = 0;
 hccapxwrittencount = 0;
 hccapwrittencount = 0;
-ERR_load_crypto_strings();
-OpenSSL_add_all_algorithms();
 if((hashlist = (hashlist_t*)calloc(hashlistcount, HASHLIST_SIZE)) == NULL) return false;
 if((ouilist = (ouilist_t*)calloc(ouilistcount, OUILIST_SIZE)) == NULL) return false;
 return true;
--- a/hcxpcapngtool.c
+++ b/hcxpcapngtool.c
@@ -358,9 +358,6 @@ if(eapmschapv2msglist != NULL) free(eapm
 if(eapmschapv2hashlist != NULL) free(eapmschapv2hashlist);
 if(tacacsplist != NULL) free(tacacsplist);
 
-EVP_cleanup();
-CRYPTO_cleanup_all_ex_data();
-ERR_free_strings();
 return;
 }
 /*===========================================================================*/
@@ -369,8 +366,6 @@ static bool initlists()
 static unsigned long opensslversion;
 static const char nastring[] = { "N/A" };
 
-ERR_load_crypto_strings();
-OpenSSL_add_all_algorithms();
 opensslversion = OpenSSL_version_num();
 opensslversionmajor = (opensslversion & 0x10000000L) >> 28;
 opensslversionminor = (opensslversion & 0x01100000L) >> 20;
--- a/hcxpmktool.c
+++ b/hcxpmktool.c
@@ -920,8 +920,6 @@ while((auswahl = getopt_long(argc, argv,
 		}
 	}
 
-ERR_load_crypto_strings();
-OpenSSL_add_all_algorithms();
 printf("\n");
 if((essidstring != NULL) && (pskstring != NULL) && (pmkstring == NULL) && (hashlinestring == NULL))
 	{
@@ -953,9 +951,6 @@ else if((essidstring != NULL) && (pskstr
 	}
 
 printf("\n");
-EVP_cleanup();
-CRYPTO_cleanup_all_ex_data();
-ERR_free_strings();
 return EXIT_SUCCESS;
 }
 /*===========================================================================*/
--- a/hcxpsktool.c
+++ b/hcxpsktool.c
@@ -61,8 +61,6 @@ essidglen = 32;
 t = time(NULL);
 tm = localtime(&t);
 thisyear = tm->tm_year +1900;
-ERR_load_crypto_strings();
-OpenSSL_add_all_algorithms();
 return;
 }
 /*===========================================================================*/
@@ -2770,10 +2768,6 @@ if(pskname != NULL)
 	fclose(fhpsk);
 	}
 
-EVP_cleanup();
-CRYPTO_cleanup_all_ex_data();
-ERR_free_strings();
-
 return EXIT_SUCCESS;
 }
 /*===========================================================================*/
