--- /dev/null
+++ b/include/net/ppp_defs.h
@@ -0,0 +1,10 @@
+#ifndef _NET_PPP_DEFS_H
+#define _NET_PPP_DEFS_H 1
+
+#define __need_time_t
+#include <time.h>
+
+#include <asm/types.h>
+#include <linux/ppp_defs.h>
+
+#endif /* net/ppp_defs.h */
--- a/src/libsstp-log/sstp-log-syslog.c
+++ b/src/libsstp-log/sstp-log-syslog.c
@@ -32,6 +32,7 @@
 #include <sys/uio.h>
 #include <sys/un.h>
 #include <sys/socket.h>
+#include <sys/types.h>
 #include <unistd.h>
 
 #include <sstp-common.h>
--- a/src/libsstp-log/sstp-log-std.c
+++ b/src/libsstp-log/sstp-log-std.c
@@ -25,6 +25,7 @@
 #include <stdio.h>
 #include <stdint.h>
 #include <string.h>
+#include <sys/types.h>
 #include <sys/uio.h>
 #include <unistd.h>
 
