--- a/src/sstp-client.c
+++ b/src/sstp-client.c
@@ -542,6 +542,7 @@ static status_t sstp_init_ssl(sstp_clien
     int retval = SSTP_FAIL;
     int status = 0;
 
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
     /* Initialize the OpenSSL library */
     status = SSL_library_init();
     if (status != 1)
@@ -555,6 +556,9 @@ static status_t sstp_init_ssl(sstp_clien
 
     /* Create a new crypto context */
     client->ssl_ctx = SSL_CTX_new(SSLv23_client_method());
+#else
+    client->ssl_ctx = SSL_CTX_new(TLS_client_method());
+#endif
     if (client->ssl_ctx == NULL)
     {
         log_err("Could not get SSL crypto context");
