--- a/htpdate.c
+++ b/htpdate.c
@@ -359,7 +359,7 @@ static int htpdate_adjtimex( double drif
 
 	/* Read current kernel frequency */
 	tmx.modes = 0;
-	ntp_adjtime(&tmx);
+	adjtimex(&tmx);
 
 	/* Calculate new frequency */
 	freq = (long)(65536e6 * drift);
@@ -377,7 +377,7 @@ static int htpdate_adjtimex( double drif
 		printlog( 1, "seteuid()" );
 		exit(1);
 	} else {
-		return( ntp_adjtime(&tmx) );
+		return( adjtimex(&tmx) );
 	}
 
 }
