--- a/net/wireless/reg.c
+++ b/net/wireless/reg.c
@@ -2860,6 +2860,8 @@ void regulatory_hint_country_ie(struct w
 	enum environment_cap env = ENVIRON_ANY;
 	struct regulatory_request *request = NULL, *lr;
 
+	return;
+
 	/* IE len must be evenly divisible by 2 */
 	if (country_ie_len & 0x01)
 		return;
@@ -3066,6 +3068,7 @@ static void restore_regulatory_settings(
 
 void regulatory_hint_disconnect(void)
 {
+	return;
 	pr_debug("All devices are disconnected, going to restore regulatory settings\n");
 	restore_regulatory_settings(false);
 }
