--- a/drivers/net/wireless/ath/ath10k/pci.c
+++ b/drivers/net/wireless/ath/ath10k/pci.c
@@ -128,7 +128,7 @@ static struct ce_attr host_ce_config_wla
 		.flags = CE_ATTR_FLAGS,
 		.src_nentries = 0,
 		.src_sz_max = 2048,
-		.dest_nentries = 512,
+		.dest_nentries = 128,
 		.recv_cb = ath10k_pci_htt_htc_rx_cb,
 	},
 
@@ -137,7 +137,7 @@ static struct ce_attr host_ce_config_wla
 		.flags = CE_ATTR_FLAGS,
 		.src_nentries = 0,
 		.src_sz_max = 2048,
-		.dest_nentries = 128,
+		.dest_nentries = 64,
 		.recv_cb = ath10k_pci_htc_rx_cb,
 	},
 
@@ -164,7 +164,7 @@ static struct ce_attr host_ce_config_wla
 		.flags = CE_ATTR_FLAGS,
 		.src_nentries = 0,
 		.src_sz_max = 512,
-		.dest_nentries = 512,
+		.dest_nentries = 128,
 		.recv_cb = ath10k_pci_htt_rx_cb,
 	},
 
@@ -189,7 +189,7 @@ static struct ce_attr host_ce_config_wla
 		.flags = CE_ATTR_FLAGS,
 		.src_nentries = 0,
 		.src_sz_max = 2048,
-		.dest_nentries = 128,
+		.dest_nentries = 96,
 		.recv_cb = ath10k_pci_pktlog_rx_cb,
 	},
 
