--- a/src/drv_tapi_linux.c
+++ b/src/drv_tapi_linux.c
@@ -556,7 +556,7 @@ static ssize_t ifx_tapi_write (struct fi
    IFX_uint8_t         *pData;
    IFX_size_t           buf_size;
 #endif /* TAPI_PACKET */
-   IFX_ssize_t          size = 0;
+   ssize_t          size = 0;
 
 #ifdef TAPI_PACKET
    if (pTapiDev->bInitialized == IFX_FALSE)
--- a/src/drv_tapi_osmap.h
+++ b/src/drv_tapi_osmap.h
@@ -17,39 +17,6 @@
 */
 
 #include "ifx_types.h"     /* ifx type definitions */
-
-#ifndef HAVE_IFX_ULONG_T
-   #warning please update your ifx_types.h, using local definition of IFX_ulong_t
-   /* unsigned long type - valid for 32bit systems only */
-   typedef unsigned long               IFX_ulong_t;
-   #define HAVE_IFX_ULONG_T
-#endif /* HAVE_IFX_ULONG_T */
-
-#ifndef HAVE_IFX_LONG_T
-   #warning please update your ifx_types.h, using local definition of IFX_long_t
-   /* long type - valid for 32bit systems only */
-   typedef long                        IFX_long_t;
-   #define HAVE_IFX_LONG_T
-#endif /* HAVE_IFX_LONG_T */
-
-#ifndef HAVE_IFX_INTPTR_T
-   #warning please update your ifx_types.h, using local definition of IFX_intptr_t
-   typedef IFX_long_t                  IFX_intptr_t;
-   #define HAVE_IFX_INTPTR_T
-#endif /* HAVE_IFX_INTPTR_T */
-
-#ifndef HAVE_IFX_SIZE_T
-   #warning please update your ifx_types.h, using local definition of IFX_size_t
-   typedef IFX_ulong_t                 IFX_size_t;
-   #define HAVE_IFX_SIZE_T
-#endif /* HAVE_IFX_SIZE_T */
-
-#ifndef HAVE_IFX_SSIZE_T
-   #warning please update your ifx_types.h, using local definition of IFX_ssize_t
-   typedef IFX_long_t                  IFX_ssize_t;
-   #define HAVE_IFX_SSIZE_T
-#endif /* HAVE_IFX_SSIZE_T */
-
 #include "ifxos_interrupt.h"
 #include "ifxos_memory_alloc.h"
 #include "ifxos_copy_user_space.h"
--- a/include/drv_tapi_ll_interface.h
+++ b/include/drv_tapi_ll_interface.h
@@ -40,13 +40,6 @@
 #include "ifxos_select.h"
 #endif /* TAPI_PACKET */
 
-#ifndef HAVE_IFX_ULONG_T
-   #warning please update your ifx_types.h, using local definition of IFX_ulong_t
-   /* unsigned long type - valid for 32bit systems only */
-   typedef unsigned long               IFX_ulong_t;
-   #define HAVE_IFX_ULONG_T
-#endif /* HAVE_IFX_ULONG_T */
-
 /* ============================= */
 /* Local Macros  Definitions    */
 /* ============================= */
--- a/src/lib/lib_bufferpool/lib_bufferpool.c
+++ b/src/lib/lib_bufferpool/lib_bufferpool.c
@@ -85,24 +85,6 @@
 #include <stdlib.h>
 #endif /*VXWORKS*/
 
-
-/* ============================= */
-/* Extra type definitions        */
-/* ============================= */
-#ifndef HAVE_IFX_ULONG_T
-   #warning please update your ifx_types.h, using local definition of IFX_ulong_t
-   /* unsigned long type - valid for 32bit systems only */
-   typedef unsigned long               IFX_ulong_t;
-   #define HAVE_IFX_ULONG_T
-#endif /* HAVE_IFX_ULONG_T */
-
-#ifndef HAVE_IFX_UINTPTR_T
-   #warning please update your ifx_types.h, using local definition of IFX_uintptr_t
-   typedef IFX_ulong_t                 IFX_uintptr_t;
-   #define HAVE_IFX_UINTPTR_T
-#endif /* HAVE_IFX_UINTPTR_T */
-
-
 /* ============================= */
 /* Local Macros & Definitions    */
 /* ============================= */
