Index: drv_tapi-3.13.0/src/drv_tapi_linux.c
===================================================================
--- drv_tapi-3.13.0.orig/src/drv_tapi_linux.c	2013-09-05 22:28:16.868419283 +0200
+++ drv_tapi-3.13.0/src/drv_tapi_linux.c	2013-09-05 22:32:37.396425814 +0200
@@ -93,6 +93,8 @@
 #include "drv_tapi_announcements.h"
 #endif /* TAPI_ANNOUNCEMENTS */
 
+#undef CONFIG_PROC_FS
+
 #define TAPI_IOCTL_STACKSIZE                 4000 /* allow some overhead 4 k */
 
 /* ================================== */
