From a427929b0267a85ef6924aa3449077420beb23e0 Mon Sep 17 00:00:00 2001
From: Tianling Shen <i@cnsztl.eu.org>
Date: Sun, 24 Oct 2021 17:11:10 +0800
Subject: [PATCH] fix: read version from `Version` variable

Signed-off-by: Tianling Shen <i@cnsztl.eu.org>
---
 main.go | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/main.go
+++ b/main.go
@@ -22,7 +22,7 @@ func init() {
 	cache.Init()
 }
 
-var Version = "v1.0.3"
+var Version = "v1.0.12"
 
 type Task struct {
 	Id string `json:"id"`
@@ -49,7 +49,7 @@ func main() {
 	flag.Parse()
 
 	if *versin {
-		fmt.Println("v1.0.9")
+		fmt.Println(Version)
 		return
 	}
 	if len(*refreshToken) == 0 || len(os.Args) < 3 || os.Args[1] != "-rt" {
