From 5c294df1715d673f94f3b0a6e1ea3a426ca35e6e Mon Sep 17 00:00:00 2001
From: Georgi Djakov <georgi.djakov@linaro.org>
Date: Thu, 28 Apr 2016 16:20:12 +0300
Subject: [PATCH 48/69] PM / OPP: HACK: Allow to set regulator without opp_list

Signed-off-by: Georgi Djakov <georgi.djakov@linaro.org>
---
 drivers/base/power/opp/core.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/drivers/base/power/opp/core.c
+++ b/drivers/base/power/opp/core.c
@@ -1280,11 +1280,13 @@ struct opp_table *dev_pm_opp_set_regulat
 	if (!opp_table)
 		return ERR_PTR(-ENOMEM);
 
+#if 0
 	/* This should be called before OPPs are initialized */
 	if (WARN_ON(!list_empty(&opp_table->opp_list))) {
 		ret = -EBUSY;
 		goto err;
 	}
+#endif
 
 	/* Already have regulators set */
 	if (opp_table->regulators) {
