From 626cf45df06382116f3ff41e82dddfa9100ca24d Mon Sep 17 00:00:00 2001
From: Pavel Kubelun <be.dissent@gmail.com>
Date: Fri, 22 Dec 2017 23:15:12 +0300
Subject: [PATCH] firmware: scm: add ipq806x compatible string

---
 drivers/firmware/qcom_scm.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/firmware/qcom_scm.c b/drivers/firmware/qcom_scm.c
index bb16510..5f9be0d 100644
--- a/drivers/firmware/qcom_scm.c
+++ b/drivers/firmware/qcom_scm.c
@@ -413,6 +413,9 @@ static const struct of_device_id qcom_scm_dt_match[] = {
 	{ .compatible = "qcom,scm-apq8064",
 	  /* FIXME: This should have .data = (void *) SCM_HAS_CORE_CLK */
 	},
+	{ .compatible = "qcom,scm-ipq806x",
+	  .data = NULL, /* no clocks */
+	},
 	{ .compatible = "qcom,scm-msm8660",
 	  .data = (void *) SCM_HAS_CORE_CLK,
 	},
--
Working Copy 3.0.6

