--- a/arch/arm64/boot/dts/mediatek/mt7622.dtsi
+++ b/arch/arm64/boot/dts/mediatek/mt7622.dtsi
@@ -848,6 +848,12 @@
 			#address-cells = <0>;
 			#interrupt-cells = <1>;
 		};
+
+		slot0: pcie@0,0 {
+			reg = <0x0000 0 0 0 0>;
+			#address-cells = <3>;
+			#size-cells = <2>;
+		};
 	};
 
 	pcie1: pcie@1a145000 {
@@ -886,6 +892,12 @@
 			#address-cells = <0>;
 			#interrupt-cells = <1>;
 		};
+
+		slot1: pcie@1,0 {
+			reg = <0x0800 0 0 0 0>;
+			#address-cells = <3>;
+			#size-cells = <2>;
+		};
 	};
 
 	sata: sata@1a200000 {
