--- a/Modules/CTest.cmake
+++ b/Modules/CTest.cmake
@@ -47,7 +47,7 @@ the :variable:`CTEST_USE_LAUNCHERS` vari
 in the ``CTestConfig.cmake`` file.
 #]=======================================================================]
 
-option(BUILD_TESTING "Build the testing tree." ON)
+option(BUILD_TESTING "Build the testing tree." OFF)
 
 # function to turn generator name into a version string
 # like vs9 or vs10
--- a/Modules/Dart.cmake
+++ b/Modules/Dart.cmake
@@ -33,7 +33,7 @@ whether testing support should be enable
 #
 #
 
-option(BUILD_TESTING "Build the testing tree." ON)
+option(BUILD_TESTING "Build the testing tree." OFF)
 
 if(BUILD_TESTING)
   find_package(Dart QUIET)
--- a/Tests/Contracts/VTK/Dashboard.cmake.in
+++ b/Tests/Contracts/VTK/Dashboard.cmake.in
@@ -25,7 +25,7 @@ ctest_empty_binary_directory(${CTEST_BIN
 
 file(WRITE "${CTEST_BINARY_DIRECTORY}/CMakeCache.txt" "
   VTK_BUILD_EXAMPLES:BOOL=ON
-  VTK_BUILD_TESTING:STRING=WANT
+  VTK_BUILD_TESTING:STRING=OFF
   VTK_WRAP_PYTHON:BOOL=ON
   ExternalData_OBJECT_STORES:FILEPATH=@base_dir@/ExternalData
 ")
