--- a/bootstrap
+++ b/bootstrap
@@ -1423,7 +1423,10 @@ int main(){ printf("1%c", (char)0x0a); r
 ' > "test.c"
 cmake_original_make_flags="${cmake_make_flags}"
 if test "x${cmake_parallel_make}" != "x"; then
-  cmake_make_flags="${cmake_make_flags} -j ${cmake_parallel_make}"
+  case "$cmake_parallel_make" in
+    [0-9]*) cmake_parallel_make="-j ${cmake_parallel_make}";;
+  esac
+  cmake_make_flags="${cmake_make_flags} ${cmake_parallel_make}"
 fi
 for a in ${cmake_make_processors}; do
   if test -z "${cmake_make_processor}" && cmake_try_make "${a}" "${cmake_make_flags}" >> ../cmake_bootstrap.log 2>&1; then
