--- a/Utilities/cmlibarchive/CMakeLists.txt
+++ b/Utilities/cmlibarchive/CMakeLists.txt
@@ -656,7 +656,7 @@ IF(ENABLE_ZSTD)
     SET(ZSTD_FIND_QUIETLY TRUE)
   ENDIF (ZSTD_INCLUDE_DIR)
 
-  IF(0) # CMake does not let pkg-config override its search paths.
+  IF(1) # CMake does not let pkg-config override its search paths.
   IF(UNIX)
     FIND_PACKAGE(PkgConfig QUIET)
     PKG_SEARCH_MODULE(PC_ZSTD libzstd)
