--- a/Source/CMakeLists.txt
+++ b/Source/CMakeLists.txt
@@ -858,7 +858,7 @@ if(CMake_USE_XCOFF_PARSER)
 endif()
 
 # Xcode only works on Apple
-if(APPLE)
+if(APPLE_DISABLED)
   target_sources(
     CMakeLib
     PRIVATE
--- a/Source/cmake.cxx
+++ b/Source/cmake.cxx
@@ -134,7 +134,7 @@
 #  include "cmGlobalGhsMultiGenerator.h"
 #endif
 
-#if defined(__APPLE__)
+#if defined(__APPLE_DISABLED__)
 #  if !defined(CMAKE_BOOTSTRAP)
 #    include "cmGlobalXCodeGenerator.h"
 
