--- a/Source/CMakeLists.txt
+++ b/Source/CMakeLists.txt
@@ -903,7 +903,7 @@ if(CMake_USE_XCOFF_PARSER)
 endif()
 
 # Xcode only works on Apple
-if(APPLE)
+if(APPLE_DISABLED)
   target_sources(
     CMakeLib
     PRIVATE
--- a/Source/cmake.cxx
+++ b/Source/cmake.cxx
@@ -143,7 +143,7 @@
 #  endif
 #endif
 
-#if defined(__APPLE__)
+#if defined(__APPLE_DISABLED__)
 #  if !defined(CMAKE_BOOTSTRAP)
 #    include "cmGlobalXCodeGenerator.h"
 
