--- a/drivers/gpio/Kconfig
+++ b/drivers/gpio/Kconfig
@@ -441,6 +441,18 @@ config GPIO_REG
 	  A 32-bit single register GPIO fixed in/out implementation.  This
 	  can be used to represent any register as a set of GPIO signals.
 
+config GPIO_RTL8231
+	tristate "RTL8231 GPIO"
+	depends on GPIO_RTL838X
+	help
+	  Say yes here to support Realtek RTL8231 GPIO expansion chips.
+
+config GPIO_RTL838X
+	tristate "RTL838X GPIO"
+	depends on RTL838X
+	help
+	  Say yes here to support RTL838X GPIO devices.
+
 config GPIO_SAMA5D2_PIOBU
 	tristate "SAMA5D2 PIOBU GPIO support"
 	depends on MFD_SYSCON
--- a/drivers/gpio/Makefile
+++ b/drivers/gpio/Makefile
@@ -117,6 +117,8 @@ obj-$(CONFIG_GPIO_RC5T583)		+= gpio-rc5t
 obj-$(CONFIG_GPIO_RCAR)			+= gpio-rcar.o
 obj-$(CONFIG_GPIO_RDC321X)		+= gpio-rdc321x.o
 obj-$(CONFIG_GPIO_REG)			+= gpio-reg.o
+obj-$(CONFIG_GPIO_RTL8231)		+= gpio-rtl8231.o
+obj-$(CONFIG_GPIO_RTL838X)		+= gpio-rtl838x.o
 obj-$(CONFIG_ARCH_SA1100)		+= gpio-sa1100.o
 obj-$(CONFIG_GPIO_SAMA5D2_PIOBU)	+= gpio-sama5d2-piobu.o
 obj-$(CONFIG_GPIO_SCH311X)		+= gpio-sch311x.o
