--- a/include/linux/phy.h
+++ b/include/linux/phy.h
@@ -644,6 +644,10 @@ struct phy_driver {
 			    struct ethtool_tunable *tuna,
 			    const void *data);
 	int (*set_loopback)(struct phy_device *dev, bool enable);
+	int (*get_port)(struct phy_device *dev);
+	int (*set_port)(struct phy_device *dev, int port);
+	int (*get_eee)(struct phy_device *dev, struct ethtool_eee *e);
+	int (*set_eee)(struct phy_device *dev, struct ethtool_eee *e);
 };
 #define to_phy_driver(d) container_of(to_mdio_common_driver(d),		\
 				      struct phy_driver, mdiodrv)
