--- a/drivers/net/phy/Kconfig
+++ b/drivers/net/phy/Kconfig
@@ -540,6 +540,12 @@ config REALTEK_PHY
 	---help---
 	  Supports the Realtek 821x PHY.
 
+config REALTEK_SOC_PHY
+	tristate "Realtek SoC PHYs"
+	depends on RTL838X
+	---help---
+	  Supports the PHYs found in combination with Realtek Switch SoCs
+
 config RENESAS_PHY
 	tristate "Driver for Renesas PHYs"
 	---help---
--- a/drivers/net/phy/Makefile
+++ b/drivers/net/phy/Makefile
@@ -102,6 +102,7 @@ obj-$(CONFIG_NATIONAL_PHY)	+= national.o
 obj-$(CONFIG_NXP_TJA11XX_PHY)	+= nxp-tja11xx.o
 obj-$(CONFIG_QSEMI_PHY)		+= qsemi.o
 obj-$(CONFIG_REALTEK_PHY)	+= realtek.o
+obj-$(CONFIG_REALTEK_SOC_PHY)   += rtl83xx-phy.o
 obj-$(CONFIG_RENESAS_PHY)	+= uPD60620.o
 obj-$(CONFIG_ROCKCHIP_PHY)	+= rockchip.o
 obj-$(CONFIG_SMSC_PHY)		+= smsc.o
