commit 3790bd05ed86b1692719df22857ef76c1293b01c
Author: Harald Geyer <harald@ccbib.org>
Date:   Sun Feb 18 19:24:48 2018 +0100

    Make modem processes terminate when main process dies

--- a/src/smsd.c
+++ b/src/smsd.c
@@ -34,6 +34,7 @@ Either version 2 of the License, or (at
 #ifndef DISABLE_INOTIFY
 #include <sys/inotify.h>
 #endif
+#include <sys/prctl.h>
 
 #include "extras.h"
 #include "locking.h"
@@ -8100,6 +8101,7 @@ int main(int argc,char** argv)
 
       if (pid == 0)
       {
+        prctl(PR_SET_PDEATHSIG, SIGTERM, 0, 0, 0);
         process_id = i;
         strcpy(process_title, DEVICE.name);
         apply_process_name(argc, argv, process_title);
