--- a/lib/vtls/mbedtls.c
+++ b/lib/vtls/mbedtls.c
@@ -811,7 +811,7 @@ static void Curl_mbedtls_session_free(vo
 
 static size_t Curl_mbedtls_version(char *buffer, size_t size)
 {
-  unsigned int version = mbedtls_version_get_number();
+  unsigned int version = MBEDTLS_VERSION_NUMBER;
   return msnprintf(buffer, size, "mbedTLS/%u.%u.%u", version>>24,
                    (version>>16)&0xff, (version>>8)&0xff);
 }
