--- a/drivers/mtd/Kconfig
+++ b/drivers/mtd/Kconfig
@@ -178,6 +178,12 @@ menu "Partition parsers"
 source "drivers/mtd/parsers/Kconfig"
 endmenu
 
+config MTD_CYBERTAN_PARTS
+	tristate "Cybertan partitioning support"
+	depends on ATH79
+	---help---
+	   Cybertan partitioning support
+
 config MTD_MYLOADER_PARTS
 	tristate "MyLoader partition parsing"
 	depends on ADM5120 || ATH25 || ATH79
--- a/drivers/mtd/Makefile
+++ b/drivers/mtd/Makefile
@@ -17,6 +17,7 @@ obj-$(CONFIG_MTD_BCM63XX_PARTS)	+= bcm63
 obj-$(CONFIG_MTD_BCM47XX_PARTS)	+= bcm47xxpart.o
 obj-$(CONFIG_MTD_MYLOADER_PARTS) += myloader.o
 obj-y				+= parsers/
+obj-$(CONFIG_MTD_CYBERTAN_PARTS) += cybertan_part.o
 
 # 'Users' - code which presents functionality to userspace.
 obj-$(CONFIG_MTD_BLKDEVS)	+= mtd_blkdevs.o
