--- a/arch/mips/ath79/machtypes.h
+++ b/arch/mips/ath79/machtypes.h
@@ -154,6 +154,7 @@
 	ATH79_MACH_LAN_TURTLE,			/* Hak5 LAN Turtle */
 	ATH79_MACH_LIMA,			/* 8devices Lima */
 	ATH79_MACH_MC_MAC1200R,			/* MERCURY MAC1200R */
+	ATH79_MACH_MC_MW4530R,		/* MerCury MW4530R */
 	ATH79_MACH_MR12,			/* Cisco Meraki MR12 */
 	ATH79_MACH_MR16,			/* Cisco Meraki MR16 */
 	ATH79_MACH_MR1750,			/* OpenMesh MR1750 */
===================================================================
--- a/arch/mips/ath79/Makefile
+++ b/arch/mips/ath79/Makefile
@@ -154,6 +154,7 @@
 obj-$(CONFIG_ATH79_MACH_LAN_TURTLE)		+= mach-lan-turtle.o
 obj-$(CONFIG_ATH79_MACH_LIMA)			+= mach-lima.o
 obj-$(CONFIG_ATH79_MACH_MC_MAC1200R)		+= mach-mc-mac1200r.o
+obj-$(CONFIG_ATH79_MACH_MC_MW4530R)		+= mach-mc-mw4530r.o
 obj-$(CONFIG_ATH79_MACH_MR12)			+= mach-mr12.o
 obj-$(CONFIG_ATH79_MACH_MR16)			+= mach-mr16.o
 obj-$(CONFIG_ATH79_MACH_MR1750)			+= mach-mr1750.o
===================================================================
--- a/arch/mips/ath79/Kconfig.openwrt
+++ b/arch/mips/ath79/Kconfig.openwrt
@@ -1126,6 +1126,17 @@
 	select ATH79_DEV_M25P80
 	select ATH79_DEV_WMAC
 
+config ATH79_MACH_MC_MW4530R
+	bool "MerCury MW4530R board support"
+	select SOC_AR934X
+	select ATH79_DEV_AP9X_PCI if PCI
+	select ATH79_DEV_ETH
+	select ATH79_DEV_GPIO_BUTTONS
+	select ATH79_DEV_LEDS_GPIO
+	select ATH79_DEV_M25P80
+	select ATH79_DEV_USB
+	select ATH79_DEV_WMAC
+
 config ATH79_MACH_RB4XX
 	bool "MikroTik RouterBOARD 4xx series support"
 	select SOC_AR71XX
