commit 31d5b6b580a356c5149dbc4f53118b7f83c60822
Author: Philip Prindeville <philipp@redfish-solutions.com>
Date:   Sun Jul 30 18:54:46 2017 -0600
Subject: Add required RSA header for OpenSSL
    
Signed-off-by: Philip Prindeville <philipp@redfish-solutions.com>

--- a/src/iperf_auth.c	2017-06-26 11:42:56.000000000 -0600
+++ b/src/iperf_auth.c	2017-07-30 17:42:43.070447875 -0600
@@ -40,6 +40,7 @@
 #include <openssl/pem.h>
 #include <openssl/sha.h>
 #include <openssl/buffer.h>
+#include <openssl/rsa.h>
 
 void sha256(const char *string, char outputBuffer[65])
 {
