--- a/drivers/net/wireless/realtek/rtw88/usb.c
+++ b/drivers/net/wireless/realtek/rtw88/usb.c
@@ -965,7 +965,11 @@ static int rtw_usb_init_rx(struct rtw_de
 	struct sk_buff *rx_skb;
 	int i;
 
+#if LINUX_VERSION_IS_GEQ(6,9,0)
 	rtwusb->rxwq = alloc_workqueue("rtw88_usb: rx wq", WQ_BH, 0);
+#else
+	rtwusb->rxwq = create_singlethread_workqueue("rtw88_usb: rx wq");
+#endif
 	if (!rtwusb->rxwq) {
 		rtw_err(rtwdev, "failed to create RX work queue\n");
 		return -ENOMEM;
