--- a/drivers/mtd/spi-nor/esmt.c
+++ b/drivers/mtd/spi-nor/esmt.c
@@ -10,6 +10,8 @@
 
 static const struct flash_info esmt_parts[] = {
 	/* ESMT */
+	{ "f25l16pa-2s", INFO(0x8c2115, 0, 64 * 1024, 32,
+			   SECT_4K | SPI_NOR_HAS_LOCK) },
 	{ "f25l32pa", INFO(0x8c2016, 0, 64 * 1024, 64,
 			   SECT_4K | SPI_NOR_HAS_LOCK | SPI_NOR_SWP_IS_VOLATILE) },
 	{ "f25l32qa", INFO(0x8c4116, 0, 64 * 1024, 64,
