From 9a0e73c5f6cbcf825eced89d26273a24f7266522 Mon Sep 17 00:00:00 2001
From: "Schanzenbach, Martin" <martin.schanzenbach@aisec.fraunhofer.de>
Date: Sun, 30 Sep 2018 20:38:37 +0200
Subject: [PATCH] shared library

---
 Makefile.in | 13 ++++++-------
 1 file changed, 6 insertions(+), 7 deletions(-)

--- a/Makefile.in
+++ b/Makefile.in
@@ -18,13 +18,12 @@ LDFLAGS = @LDFLAGS@ \
 
 DISTNAME = @PACKAGE_TARNAME@-@PACKAGE_VERSION@
 
-all: libgabe.a TAGS
+all: libgabe.so TAGS
 
 # compilation and library making
 
-libgabe.a: core.o misc.o
-	rm -f $@
-	ar rc $@ $^
+libgabe.so: core.o misc.o
+	$(CC) -shared -o $@ $^
 
 # test: test.o libgabe.a
 # 	$(CC) -o $@ $(LDFLAGS) $^
@@ -44,14 +43,14 @@ dist: AUTHORS COPYING INSTALL NEWS READM
 	tar zc $(DISTNAME) > $(DISTNAME).tar.gz
 	rm -rf $(DISTNAME)
 
-install: libgabe.a gabe.h
+install: libgabe.so gabe.h
 	$(top_srcdir)/mkinstalldirs -m 755 $(DESTDIR)$(libdir)
 	$(top_srcdir)/mkinstalldirs -m 755 $(DESTDIR)$(includedir)
-	$(top_srcdir)/install-sh -m 755 libgabe.a $(DESTDIR)$(libdir)
+	$(top_srcdir)/install-sh -m 755 libgabe.so $(DESTDIR)$(libdir)
 	$(top_srcdir)/install-sh -m 644 gabe.h $(DESTDIR)$(includedir)
 
 uninstall:
-	/bin/rm -f $(DESTDIR)$(libdir)/libgabe.a
+	/bin/rm -f $(DESTDIR)$(libdir)/libgabe.so
 	/bin/rm -f $(DESTDIR)$(includedir)/gabe.h
 
 # development and meta stuff
