--- a/arch/arm/boot/dts/marvell/armada-370.dtsi
+++ b/arch/arm/boot/dts/marvell/armada-370.dtsi
@@ -254,7 +254,7 @@
 				clocks = <&gateclk 23>;
 				clock-names = "cesa0";
 				marvell,crypto-srams = <&crypto_sram>;
-				marvell,crypto-sram-size = <0x7e0>;
+				marvell,crypto-sram-size = <0x800>;
 			};
 		};
 
@@ -275,12 +275,17 @@
 			 * cpuidle workaround.
 			 */
 			idle-sram@0 {
+				status = "disabled";
 				reg = <0x0 0x20>;
 			};
 		};
 	};
 };
 
+&coherencyfab {
+	broken-idle;
+};
+
 /*
  * Default UART pinctrl setting without RTS/CTS, can be overwritten on
  * board level if a different configuration is used.
