--- a/arch/arm/boot/dts/armada-388-helios4.dts
+++ b/arch/arm/boot/dts/armada-388-helios4.dts
@@ -24,6 +24,10 @@
 	aliases {
 		/* So that mvebu u-boot can update the MAC addresses */
 		ethernet1 = &eth0;
+		led-boot = &led_status;
+		led-failsafe = &led_status;
+		led-running = &led_status;
+		led-upgrade = &led_status;
 	};
 
 	chosen {
@@ -73,10 +77,9 @@
 		pinctrl-names = "default";
 		pinctrl-0 = <&helios_system_led_pins>;
 
-		status-led {
+		led_status: status-led {
 			label = "helios4:green:status";
 			gpios = <&gpio0 24 GPIO_ACTIVE_LOW>;
-			linux,default-trigger = "heartbeat";
 			default-state = "on";
 		};
 
