--- a/drivers/crypto/rockchip/rk3288_crypto.c
+++ b/drivers/crypto/rockchip/rk3288_crypto.c
@@ -24,18 +24,6 @@ static struct rockchip_ip rocklist = {
 	.lock = __SPIN_LOCK_UNLOCKED(rocklist.lock),
 };
 
-struct rk_crypto_info *get_rk_crypto(void)
-{
-	struct rk_crypto_info *first;
-
-	spin_lock(&rocklist.lock);
-	first = list_first_entry_or_null(&rocklist.dev_list,
-					 struct rk_crypto_info, list);
-	list_rotate_left(&rocklist.dev_list);
-	spin_unlock(&rocklist.lock);
-	return first;
-}
-
 static const struct rk_variant rk3288_variant = {
 	.num_clks = 4,
 	.rkclks = {
