--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -524,9 +524,9 @@ remove_flags(-Wno-pedantic -pedantic-err
 if (LIBCXX_HAS_MUSL_LIBC OR LIBCXX_TARGETING_CLANG_CL)
   # musl's pthread implementations uses volatile types in their structs which is
   # not a constexpr in C++11 but is in C++14, so we use C++14 with musl.
-  set(LIBCXX_STANDARD_VER c++14 CACHE STRING "internal option to change build dialect")
+  set(LIBCXX_STANDARD_VER c++2a CACHE STRING "internal option to change build dialect")
 else()
-  set(LIBCXX_STANDARD_VER c++11 CACHE STRING "internal option to change build dialect")
+  set(LIBCXX_STANDARD_VER c++2a CACHE STRING "internal option to change build dialect")
 endif()
 add_compile_flags_if_supported(-std=${LIBCXX_STANDARD_VER})
 add_compile_flags_if_supported("/std:${LIBCXX_STANDARD_VER}")
