do not use host path to fix cross compiling

--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -160,7 +160,7 @@
 	(echo '/* GNU ld script'; \
 	 echo ' * The MD5 functions are provided by the libmd library. */'; \
 	 cat format.ld; \
-	 echo "GROUP($(runtimelibdir)/$$soname AS_NEEDED($(MD5_LIBS)))"; \
+	 echo "GROUP($$soname AS_NEEDED($(MD5_LIBS)))"; \
 	)>$(DESTDIR)$(libdir)/libbsd.so
 else
 	if [ "$(libdir)" != "$(runtimelibdir)" ]; then \
