--- a/arch/arm64/boot/dts/mediatek/mt7986a-bananapi-bpi-r3-emmc.dtso
+++ b/arch/arm64/boot/dts/mediatek/mt7986a-bananapi-bpi-r3-emmc.dtso
@@ -21,5 +21,24 @@
 	non-removable;
 	no-sd;
 	no-sdio;
+	#address-cells = <1>;
+	#size-cells = <0>;
 	status = "okay";
+
+	card@0 {
+		compatible = "mmc-card";
+		reg = <0>;
+
+		partitions {
+			compatible = "gpt-partitions";
+
+			emmc_rootdisk: block-partition-production {
+				partname = "production";
+			};
+		};
+	};
+};
+
+&{/chosen} {
+	rootdisk-emmc = <&emmc_rootdisk>;
 };
--- a/arch/arm64/boot/dts/mediatek/mt7986a-bananapi-bpi-r3-nand.dtso
+++ b/arch/arm64/boot/dts/mediatek/mt7986a-bananapi-bpi-r3-nand.dtso
@@ -29,25 +29,24 @@
 
 			partition@0 {
 				label = "bl2";
-				reg = <0x0 0x100000>;
-				read-only;
+				reg = <0x0 0x200000>;
 			};
 
-			partition@100000 {
-				label = "reserved";
-				reg = <0x100000 0x280000>;
-			};
-
-			partition@380000 {
-				label = "fip";
-				reg = <0x380000 0x200000>;
-				read-only;
-			};
-
-			partition@580000 {
+			partition@200000 {
 				label = "ubi";
-				reg = <0x580000 0x7a80000>;
+				reg = <0x200000 0x7e00000>;
+				compatible = "linux,ubi";
+
+				volumes {
+					nand_rootdisk: ubi-volume-fit {
+						volname = "fit";
+					};
+				};
 			};
 		};
 	};
 };
+
+&{/chosen} {
+	rootdisk-spim-nand = <&nand_rootdisk>;
+};
--- a/arch/arm64/boot/dts/mediatek/mt7986a-bananapi-bpi-r3-nor.dtso
+++ b/arch/arm64/boot/dts/mediatek/mt7986a-bananapi-bpi-r3-nor.dtso
@@ -52,7 +52,7 @@
 				reg = <0x180000 0xa80000>;
 			};
 
-			partition@c00000 {
+			nor_rootdisk: partition@c00000 {
 				label = "fit";
 				reg = <0xc00000 0x1400000>;
 				compatible = "denx,fit";
@@ -60,3 +60,7 @@
 		};
 	};
 };
+
+&{/chosen} {
+	rootdisk-nor = <&nor_rootdisk>;
+};
--- a/arch/arm64/boot/dts/mediatek/mt7986a-bananapi-bpi-r3-sd.dtso
+++ b/arch/arm64/boot/dts/mediatek/mt7986a-bananapi-bpi-r3-sd.dtso
@@ -15,5 +15,24 @@
 	bus-width = <4>;
 	max-frequency = <52000000>;
 	cap-sd-highspeed;
+	#address-cells = <1>;
+	#size-cells = <0>;
 	status = "okay";
+
+	card@0 {
+		compatible = "mmc-card";
+		reg = <0>;
+
+		partitions {
+			compatible = "gpt-partitions";
+
+			sd_rootdisk: block-partition-production {
+				partname = "production";
+			};
+		};
+	};
+};
+
+&{/chosen} {
+	rootdisk-sd = <&sd_rootdisk>;
 };
