From: coolsnowwolf <coolsnowwolf@gmail.com>
Subject: [PATCH] net: phy: add Realtek RTL8261N support

Adapted from ImmortalWrt 8bc0f53b13b249792c6a689ff6173996ae2e16fd.

Signed-off-by: coolsnowwolf <coolsnowwolf@gmail.com>

--- a/drivers/net/phy/Kconfig
+++ b/drivers/net/phy/Kconfig
@@ -439,6 +439,8 @@
 	help
 	  Currently supports the qs6612
 
+source "drivers/net/phy/rtl8261n/Kconfig"
+
 source "drivers/net/phy/realtek/Kconfig"
 
 config RENESAS_PHY
--- a/drivers/net/phy/Makefile
+++ b/drivers/net/phy/Makefile
@@ -120,3 +120,5 @@
 obj-$(CONFIG_TERANETICS_PHY)	+= teranetics.o
 obj-$(CONFIG_VITESSE_PHY)	+= vitesse.o
 obj-$(CONFIG_XILINX_GMII2RGMII) += xilinx_gmii2rgmii.o
+
+obj-y				+= rtl8261n/
