--- a/drivers/cpufreq/mediatek-cpufreq.c
+++ b/drivers/cpufreq/mediatek-cpufreq.c
@@ -741,6 +741,16 @@ static struct platform_driver mtk_cpufre
 	.probe		= mtk_cpufreq_probe,
 };
 
+struct mtk_cpufreq_corr_data mt7988_volt_corr_data[] = {
+	{
+		.freq = 1800000000,
+		.vbase = 850000,
+		.vscale = 10000,
+		.vmax = 1120000,
+	},
+	{ } /* sentinel */
+};
+
 static const struct mtk_cpufreq_platform_data mt2701_platform_data = {
 	.min_volt_shift = 100000,
 	.max_volt_shift = 200000,
@@ -769,10 +779,12 @@ static const struct mtk_cpufreq_platform
 static const struct mtk_cpufreq_platform_data mt7988_platform_data = {
 	.min_volt_shift = 100000,
 	.max_volt_shift = 200000,
-	.proc_max_volt = 900000,
+	.proc_max_volt = 1120000,
 	.sram_min_volt = 0,
 	.sram_max_volt = 1150000,
 	.ccifreq_supported = true,
+	.nvmem_volt_corr = 1,
+	.corr_data = mt7988_volt_corr_data,
 };
 
 static const struct mtk_cpufreq_platform_data mt8183_platform_data = {
