--- a/drivers/cpufreq/mediatek-cpufreq.c
+++ b/drivers/cpufreq/mediatek-cpufreq.c
@@ -35,6 +35,8 @@ struct mtk_cpufreq_platform_data {
 	bool ccifreq_supported;
 	/* whether voltage correction via nvmem is supported */
 	bool nvmem_volt_corr;
+	/* Flag indicating whether the processor voltage is fixed */
+	bool proc_fixed_volt;
 };
 
 /*
@@ -176,6 +178,9 @@ static int mtk_cpufreq_set_voltage(struc
 	const struct mtk_cpufreq_platform_data *soc_data = info->soc_data;
 	int ret;
 
+	if (soc_data->proc_fixed_volt)
+		return 0;
+
 	if (info->need_voltage_tracking)
 		ret = mtk_cpufreq_voltage_tracking(info, vproc);
 	else
