--- a/drivers/cpufreq/mediatek-cpufreq.c
+++ b/drivers/cpufreq/mediatek-cpufreq.c
@@ -781,6 +781,12 @@ static const struct mtk_cpufreq_platform
 	.ccifreq_supported = false,
 };
 
+static const struct mtk_cpufreq_platform_data mt7987_platform_data = {
+	.proc_max_volt = 1023000,
+	.ccifreq_supported = false,
+	.proc_fixed_volt = true,
+};
+
 static const struct mtk_cpufreq_platform_data mt7988_platform_data = {
 	.min_volt_shift = 100000,
 	.max_volt_shift = 200000,
@@ -825,6 +831,7 @@ static const struct of_device_id mtk_cpu
 	{ .compatible = "mediatek,mt2712", .data = &mt2701_platform_data },
 	{ .compatible = "mediatek,mt7622", .data = &mt7622_platform_data },
 	{ .compatible = "mediatek,mt7623", .data = &mt7623_platform_data },
+	{ .compatible = "mediatek,mt7987", .data = &mt7987_platform_data },
 	{ .compatible = "mediatek,mt7988a", .data = &mt7988_platform_data },
 	{ .compatible = "mediatek,mt7988d", .data = &mt7988_platform_data },
 	{ .compatible = "mediatek,mt8167", .data = &mt8516_platform_data },
