--- a/arch/arm64/boot/dts/mediatek/mt7622-bananapi-bpi-r64.dts
+++ b/arch/arm64/boot/dts/mediatek/mt7622-bananapi-bpi-r64.dts
@@ -32,6 +32,9 @@
 	chosen {
 		stdout-path = "serial0:115200n8";
 		bootargs = "earlycon=uart8250,mmio32,0x11002000 console=ttyS0,115200n1 swiotlb=512";
+		rootdisk-emmc = <&emmc_rootfs>;
+		rootdisk-sd = <&sd_rootfs>;
+		rootdisk-snfi = <&ubi_rootfs>;
 	};
 
 	cpus {
@@ -245,6 +248,28 @@
 	assigned-clocks = <&topckgen CLK_TOP_MSDC30_0_SEL>;
 	assigned-clock-parents = <&topckgen CLK_TOP_UNIV48M>;
 	non-removable;
+	#address-cells = <1>;
+	#size-cells = <0>;
+
+	card@0 {
+		compatible = "mmc-card";
+		reg = <0>;
+
+		block {
+			compatible = "block-device";
+			partitions {
+				block-partition-env {
+					partname = "ubootenv";
+					nvmem-layout {
+						compatible = "u-boot,env";
+					};
+				};
+				emmc_rootfs: block-partition-production {
+					partname = "production";
+				};
+			};
+		};
+	};
 };
 
 &mmc1 {
@@ -260,6 +285,28 @@
 	vqmmc-supply = <&reg_3p3v>;
 	assigned-clocks = <&topckgen CLK_TOP_MSDC30_1_SEL>;
 	assigned-clock-parents = <&topckgen CLK_TOP_UNIV48M>;
+	#address-cells = <1>;
+	#size-cells = <0>;
+
+	card@0 {
+		compatible = "mmc-card";
+		reg = <0>;
+
+		block {
+			compatible = "block-device";
+			partitions {
+				block-partition-env {
+					partname = "ubootenv";
+					nvmem-layout {
+						compatible = "u-boot,env";
+					};
+				};
+				sd_rootfs: block-partition-production {
+					partname = "production";
+				};
+			};
+		};
+	};
 };
 
 &nandc {
@@ -293,15 +340,30 @@
 				read-only;
 			};
 
-			partition@80000 {
-				label = "fip";
-				reg = <0x80000 0x200000>;
-				read-only;
-			};
-
-			ubi: partition@280000 {
+			ubi: partition@80000 {
 				label = "ubi";
-				reg = <0x280000 0x7d80000>;
+				reg = <0x80000 0x7f80000>;
+				compatible = "linux,ubi";
+
+				volumes {
+					ubi-volume-ubootenv {
+						volname = "ubootenv";
+						nvmem-layout {
+							compatible = "u-boot,env-redundant-bool";
+						};
+					};
+
+					ubi-volume-ubootenv2 {
+						volname = "ubootenv2";
+						nvmem-layout {
+							compatible = "u-boot,env-redundant-bool";
+						};
+					};
+
+					ubi_rootfs: ubi-volume-fit {
+						volname = "fit";
+					};
+				};
 			};
 		};
 	};
