--- a/lib/traceevent/plugins/plugin_mac80211.c
+++ b/lib/traceevent/plugins/plugin_mac80211.c
@@ -165,12 +165,15 @@ static int drv_config(struct trace_seq *
 		{ 2, "IDLE" },
 		{ 3, "QOS"},
 	);
-	tep_print_num_field(s, " chan:%d/", event, "center_freq", record, 1);
-	print_enum(s, event, "channel_type", data,
-		{ 0, "noht" },
-		{ 1, "ht20" },
-		{ 2, "ht40-" },
-		{ 3, "ht40+" });
+	tep_print_num_field(s, " chan:%d@", event, "control_freq", record, 1);
+	print_enum(s, event, "chan_width", data,
+		{ 0, "20_noht" },
+		{ 1, "20" },
+		{ 2, "40" },
+		{ 3, "80" },
+		{ 4, "80p80" },
+		{ 5, "160" });
+
 	trace_seq_putc(s, ' ');
 	SF("power_level");
 
