diff -Nurp emailrelay-1.9/src/glib/gstr.cpp emailrelay-1.9-new/src/glib/gstr.cpp
--- emailrelay-1.9/src/glib/gstr.cpp	2013-12-07 23:55:43.000000000 +0100
+++ emailrelay-1.9-new/src/glib/gstr.cpp	2016-05-21 03:50:36.178681099 +0200
@@ -546,7 +546,7 @@ void G::Str::readLineFrom( std::istream
 
 	if( pre_erase )
 		line.erase() ;
-
+/*
 	// this is a special speed optimisation for a two-character terminator with a one-character initial string ;-)
 	if( eol.length() == 2U && eol[0] != eol[1] && line.length() == 1U )
 	{
@@ -577,9 +577,9 @@ void G::Str::readLineFrom( std::istream
 		}
 	}
 	else
-	{
+	{ */
 		readLineFromImp( stream , eol , line ) ;
-	}
+//	}
 }
 
 void G::Str::readLineFromImp( std::istream & stream , const std::string & eol , std::string & line )
diff -Nurp emailrelay-1.9/src/gsmtp/gserverprotocol.cpp emailrelay-1.9-new/src/gsmtp/gserverprotocol.cpp
--- emailrelay-1.9/src/gsmtp/gserverprotocol.cpp	2013-12-07 23:55:46.000000000 +0100
+++ emailrelay-1.9-new/src/gsmtp/gserverprotocol.cpp	2016-05-21 17:52:18.682367416 +0200
@@ -608,7 +608,6 @@ std::string GSmtp::ServerProtocol::comma
 std::string GSmtp::ServerProtocol::commandLine( const std::string & line_in ) const
 {
 	std::string line( line_in ) ;
-	G::Str::trimLeft( line , " \t" ) ;
 	return line ;
 }
 
