--- a/src/gssl/gssl_openssl.cpp	2016-08-12 16:14:04.122314793 +0200
+++ b/src/gssl/gssl_openssl.cpp	2016-08-12 16:17:11.969303185 +0200
@@ -290,12 +290,7 @@ namespace
 
 GSsl::Context::Context( const std::string & pem_file , unsigned int flags )
 {
-	if( (flags&3U) == 2U )
-		m_ssl_ctx = SSL_CTX_new(SSLv23_method()) ;
-	else if( (flags&3U) == 3U )
-		m_ssl_ctx = SSL_CTX_new(SSLv3_method()) ;
-	else
-		m_ssl_ctx = SSL_CTX_new(TLSv1_method()) ;
+	m_ssl_ctx = SSL_CTX_new(SSLv23_method()) ;
 
 	if( m_ssl_ctx == NULL )
 		throw Error( "SSL_CTX_new" , ERR_get_error() ) ;
