--- a/source3/librpc/rpc/rpc_common.c
+++ b/source3/librpc/rpc/rpc_common.c
@@ -120,9 +120,11 @@ static bool initialize_interfaces(void)
 	if (!smb_register_ndr_interface(&ndr_table_netdfs)) {
 		return false;
 	}
+#ifdef DEVELOPER
 	if (!smb_register_ndr_interface(&ndr_table_rpcecho)) {
 		return false;
 	}
+#endif
 	if (!smb_register_ndr_interface(&ndr_table_initshutdown)) {
 		return false;
 	}
--- a/source3/rpcclient/rpcclient.c
+++ b/source3/rpcclient/rpcclient.c
@@ -631,7 +631,9 @@ static struct cmd_set *rpcclient_command
 	netlogon_commands,
 	srvsvc_commands,
 	dfs_commands,
+#ifdef DEVELOPER
 	echo_commands,
+#endif
 	epmapper_commands,
 	shutdown_commands,
  	test_commands,
