--- a/source3/rpc_server/rpc_service_setup.c
+++ b/source3/rpc_server/rpc_service_setup.c
@@ -508,6 +508,7 @@
 	}
 #endif
 
+#ifdef EXTRA_SERVICES
 	ok = rpc_setup_svcctl(ev_ctx, msg_ctx);
 	if (!ok) {
 		goto done;
@@ -522,6 +523,7 @@
 	if (!ok) {
 		goto done;
 	}
+#endif	
 
 	ok = rpc_setup_initshutdown(ev_ctx, msg_ctx);
 	if (!ok) {
--- a/source3/smbd/server_exit.c
+++ b/source3/smbd/server_exit.c
@@ -183,9 +183,11 @@ static void exit_server_common(enum serv
 #endif
 		rpc_netdfs_shutdown();
 		rpc_initshutdown_shutdown();
+#ifdef EXTRA_SERVICES
 		rpc_eventlog_shutdown();
 		rpc_ntsvcs_shutdown();
 		rpc_svcctl_shutdown();
+#endif
 #ifdef PRINTER_SUPPORT
 		rpc_spoolss_shutdown();
 #endif
--- a/source3/librpc/rpc/rpc_common.c
+++ b/source3/librpc/rpc/rpc_common.c
@@ -130,6 +130,7 @@ static bool initialize_interfaces(void)
 	if (!smb_register_ndr_interface(&ndr_table_initshutdown)) {
 		return false;
 	}
+#ifdef EXTRA_SERVICES
 	if (!smb_register_ndr_interface(&ndr_table_svcctl)) {
 		return false;
 	}
@@ -139,6 +140,7 @@ static bool initialize_interfaces(void)
 	if (!smb_register_ndr_interface(&ndr_table_ntsvcs)) {
 		return false;
 	}
+#endif
 	if (!smb_register_ndr_interface(&ndr_table_epmapper)) {
 		return false;
 	}
--- a/source3/rpcclient/rpcclient.c
+++ b/source3/rpcclient/rpcclient.c
@@ -640,9 +640,11 @@ static struct cmd_set *rpcclient_command
 	shutdown_commands,
  	test_commands,
 	wkssvc_commands,
+#ifdef EXTRA_SERVICES
 	ntsvcs_commands,
 	drsuapi_commands,
 	eventlog_commands,
+#endif
 	winreg_commands,
 	fss_commands,
 	NULL
