--- a/source3/librpc/rpc/rpc_common.c	2013-02-04 14:33:53.119988894 +0100
+++ b/source3/librpc/rpc/rpc_common.c	2013-02-04 14:35:04.167465045 +0100
@@ -146,12 +146,14 @@
 	if (!smb_register_ndr_interface(&ndr_table_epmapper)) {
 		return false;
 	}
+#ifdef DSREP_SERVICES
 	if (!smb_register_ndr_interface(&ndr_table_drsuapi)) {
 		return false;
 	}
 	if (!smb_register_ndr_interface(&ndr_table_FileServerVssAgent)) {
 		return false;
 	}
+#endif
 	return true;
 }
 
