--- a/lib/util/debug.h
+++ b/lib/util/debug.h
@@ -45,7 +45,7 @@ bool dbghdr( int level, const char *loca
  * Redefine DEBUGLEVEL because so we don't have to change every source file
  * that *unnecessarily* references it.
  */
-#define DEBUGLEVEL DEBUGLEVEL_CLASS[DBGC_ALL]
+#define DEBUGLEVEL 0
 
 /*
  * Define all new debug classes here. A class is represented by an entry in
--- a/source3/nmbd/asyncdns.c
+++ b/source3/nmbd/asyncdns.c
@@ -85,7 +85,7 @@ static void asyncdns_process(void)
 	struct query_record r;
 	unstring qname;
 
-	DEBUGLEVEL = -1;
+	DEBUGLEVEL_CLASS[DBGC_ALL] = -1;
 
 	while (1) {
 		NTSTATUS status;
