--- a/source3/auth/auth.c	2013-02-11 22:09:08.516106878 +0100
+++ b/source3/auth/auth.c	2013-02-11 22:09:55.158449459 +0100
@@ -449,7 +449,7 @@
 		case ROLE_DOMAIN_MEMBER:
 			DEBUG(5,("Making default auth method list for server role = 'domain member'\n"));
 			auth_method_list = str_list_make_v3(
-				talloc_tos(), "guest sam winbind:ntdomain",
+				talloc_tos(), "guest sam",
 				NULL);
 			break;
 		case ROLE_DOMAIN_BDC:
@@ -457,7 +457,7 @@
 			DEBUG(5,("Making default auth method list for DC\n"));
 			auth_method_list = str_list_make_v3(
 				talloc_tos(),
-				"guest sam winbind:trustdomain",
+				"guest sam",
 				NULL);
 			break;
 		case ROLE_STANDALONE:
