--- a/e2fsck/e2fsck.h
+++ b/e2fsck/e2fsck.h
@@ -74,7 +74,7 @@
  * Exit codes used by fsck-type programs
  */
 #define FSCK_OK          0	/* No errors */
-#define FSCK_NONDESTRUCT 1	/* File system errors corrected */
+#define FSCK_NONDESTRUCT 0	/* File system errors corrected */
 #define FSCK_REBOOT      2	/* System should be rebooted */
 #define FSCK_UNCORRECTED 4	/* File system errors left uncorrected */
 #define FSCK_ERROR       8	/* Operational error */
