diff -urN a/open-vm-tools/configure.ac b/open-vm-tools/configure.ac
--- a/open-vm-tools/configure.ac
+++ b/open-vm-tools/configure.ac
@@ -949,7 +949,7 @@
       AC_VMW_CHECK_LIB([dnet],
                        [DNET],
                        [],
-                       [dnet-config],
+                       [],
                        [],
                        [dnet.h],
                        [intf_open],
@@ -959,7 +959,7 @@
 
    if test $have_dnet = "no"; then
 		AC_MSG_ERROR(
-		   [dnet-config was not found on your PATH. Please configure without dnet (using --without-dnet) or install dnet - http://libdnet.sourceforge.net])
+		   [dnet was not found. Please configure without dnet (using --without-dnet) or install dnet - http://libdnet.sourceforge.net])
    fi
 fi
 
