--- a/drivers/watchdog/Kconfig
+++ b/drivers/watchdog/Kconfig
@@ -997,10 +997,10 @@ config RTD119X_WATCHDOG
 
 config REALTEK_OTTO_WDT
 	tristate "Realtek Otto MIPS watchdog support"
-	depends on MACH_REALTEK_RTL || COMPILE_TEST
+	depends on RTL83XX
 	depends on COMMON_CLK
 	select WATCHDOG_CORE
-	default MACH_REALTEK_RTL
+	default RTL83XX
 	help
 	  Say Y here to include support for the watchdog timer on Realtek
 	  RTL838x, RTL839x, RTL930x SoCs. This watchdog has pretimeout
