--- a/src/config/torrc.sample.in
+++ b/src/config/torrc.sample.in
@@ -39,7 +39,7 @@
 ## Send every possible message to @LOCALSTATEDIR@/log/tor/debug.log
 #Log debug file @LOCALSTATEDIR@/log/tor/debug.log
 ## Use the system log instead of Tor's logfiles
-#Log notice syslog
+Log notice syslog
 ## To send all messages to stderr:
 #Log debug stderr
 
@@ -50,7 +50,7 @@
 
 ## The directory for keeping all the keys/etc. By default, we store
 ## things in $HOME/.tor on Unix, and in Application Data\tor on Windows.
-#DataDirectory @LOCALSTATEDIR@/lib/tor
+DataDirectory @LOCALSTATEDIR@/lib/tor
 
 ## The port on which Tor will listen for local connections from Tor
 ## controller applications, as documented in control-spec.txt.
@@ -251,3 +251,4 @@
 ## The %include option can be used recursively.
 #%include /etc/torrc.d/*.conf
 
+User tor
