--- a/bind/Makefile.in
+++ b/bind/Makefile.in
@@ -56,7 +56,12 @@ bind1:
 		echo Configuring BIND libraries for DHCP. ;             \
 		rm -rf ${cleandirs} ${cleanfiles} ;                     \
 		(cd ${bindsrcdir} &&                                    \
-                 ./configure ${bindconfig} > ${binddir}/configure.log); \
+		 export CC=${CROSS_CC} &&                               \
+                ./configure ${bindconfig} --disable-atomic             \
+		 --disable-kqueue --disable-epoll --disable-kqueue      \
+                --disable-epoll --disable-devpoll --without-openssl    \
+                --without-libxml2 --disable-threads --without-gssapi   \
+                --without-randomdev > ${binddir}/configure.log);       \
 	fi
 
 atf:
