--- a/bind/Makefile.in
+++ b/bind/Makefile.in
@@ -57,6 +57,7 @@ bind1:
 		rm -rf ${cleandirs} ${cleanfiles} ;                     \
 		(cd ${bindsrcdir} &&                                    \
 		 export CC=${CROSS_CC} &&                               \
+		 ac_cv_func_catgets=no					\
                 ./configure ${bindconfig} --disable-atomic             \
 		 --disable-kqueue --disable-epoll --disable-kqueue      \
                 --disable-epoll --disable-devpoll --without-openssl    \
