From 0866cb3f7cfe4b8bae1edc8d0dbf18c85e9ca74f Mon Sep 17 00:00:00 2001
From: Mathieu Parent <math.parent@gmail.com>
Date: Tue, 27 Oct 2009 18:29:45 +0100
Subject: [PATCH 4/7] add header files

---
 pico/Makefile.am |   36 ++++++++++++++++++++++++++++++++++++
 1 files changed, 36 insertions(+), 0 deletions(-)

--- a/pico/Makefile.am
+++ b/pico/Makefile.am
@@ -34,6 +34,42 @@ libttspico_la_SOURCES = \
 	lib/picotrns.c \
 	lib/picowa.c
 
+libttspico_ladir = $(includedir)
+libttspico_la_HEADERS = \
+    lib/picoacph.h \
+    lib/picoapid.h \
+    lib/picoapi.h \
+    lib/picobase.h \
+    lib/picocep.h \
+    lib/picoctrl.h \
+    lib/picodata.h \
+    lib/picodbg.h \
+    lib/picodefs.h \
+    lib/picodsp.h \
+    lib/picoextapi.h \
+    lib/picofftsg.h \
+    lib/picokdbg.h \
+    lib/picokdt.h \
+    lib/picokfst.h \
+    lib/picoklex.h \
+    lib/picoknow.h \
+    lib/picokpdf.h \
+    lib/picokpr.h \
+    lib/picoktab.h \
+    lib/picoos.h \
+    lib/picopal.h \
+    lib/picopam.h \
+    lib/picopltf.h \
+    lib/picopr.h \
+    lib/picorsrc.h \
+    lib/picosa.h \
+    lib/picosig2.h \
+    lib/picosig.h \
+    lib/picospho.h \
+    lib/picotok.h \
+    lib/picotrns.h \
+    lib/picowa.h
+
 bin_PROGRAMS = pico2wave
 pico2wave_SOURCES = \
 	bin/pico2wave.c
