--- a/pico/Makefile.am
+++ b/pico/Makefile.am
@@ -70,6 +70,8 @@ libttspico_la_HEADERS = \
     lib/picotrns.h \
     lib/picowa.h
 
+libttspico_la_LIBADD = -lm
+
 picolangdir = $(datadir)/pico/lang
 picolang_DATA = \
     lang/de-DE*.bin \
@@ -83,6 +85,6 @@ bin_PROGRAMS = pico2wave
 pico2wave_SOURCES = \
 	bin/pico2wave.c
 pico2wave_LDADD = \
-	libttspico.la -lm -lpopt
+	libttspico.la -lpopt
 pico2wave_CFLAGS = -Wall -Dpicolangdir=\"$(picolangdir)\" -I lib
 
