--- a/quickassist/build_system/build_files/env_files/environment.mk
+++ b/quickassist/build_system/build_files/env_files/environment.mk
@@ -59,11 +59,7 @@ DIRECT_PATH=$(ICP_ROOT)/quickassist/look
 KERNEL_PATH=$(ICP_ROOT)/quickassist/lookaside/access_layer/src/qat_kernel/
 CMN_MEM_PATH=$(ICP_ROOT)/quickassist/utilities/libusdm_drv
 ICP_DEFENSES_ENABLED ?= y
-ifeq ($(shell cat /proc/kallsyms | grep -w "__stack_chk_fail" | wc -l), 1)
-KERNEL_DEFENSES_STACK_PROTECTION = y
-else
-KERNEL_DEFENSES_STACK_PROTECTION = n
-endif
+KERNEL_DEFENSES_STACK_PROTECTION ?= n
 
 #----------------------------------------------------------------------
 # ADF paths
