Index: quickassist-c3xxx-1.7/quickassist/utilities/osal/src/linux/user_space/OsalServices.c
===================================================================
--- quickassist-c3xxx-1.7.orig/quickassist/utilities/osal/src/linux/user_space/OsalServices.c
+++ quickassist-c3xxx-1.7/quickassist/utilities/osal/src/linux/user_space/OsalServices.c
@@ -42,7 +42,7 @@
 #define __STDC_WANT_LIB_EXT1__ 1
 #include <stdarg.h>
 #include <stdio.h>
-#ifdef __linux__
+#if defined(__linux__) && defined(__GLIBC__)
 #include <bits/time.h>
 #include <string.h>
 #endif
