Index: quickassist-c3xxx-1.7/quickassist/utilities/libusdm_drv/linux/user_space/qae_page_table.h
===================================================================
--- quickassist-c3xxx-1.7.orig/quickassist/utilities/libusdm_drv/linux/user_space/qae_page_table.h
+++ quickassist-c3xxx-1.7/quickassist/utilities/libusdm_drv/linux/user_space/qae_page_table.h
@@ -82,7 +82,9 @@
 #endif
 
 #ifndef __FreeBSD__ /* FreeBSD, already defined in machine param.h */
+#if defined(__GLIBC__)
 #define PAGE_SIZE (0x1000)
+#endif
 #define PAGE_SHIFT (12)
 #endif
 
