--- a/Makefile
+++ b/Makefile
@@ -13,7 +13,7 @@ endif
 
 DOCBOOK2XMAN=xsltproc --param man.authors.section.enabled 0 $(DOCBOOKXSL)/manpages/docbook.xsl
 
-all: $(BINS) $(MANS)
+all: $(BINS)
 
 clean:
 	rm -f $(BINS) $(MANS) dump.c errnos.h errno.o \
@@ -28,9 +28,6 @@ install:
 	$(INSTALL_BIN) $(BINS) $(DESTDIR)$(PREFIX)/bin
 	install $(PERLSCRIPTS) $(DESTDIR)$(PREFIX)/bin
 
-	mkdir -p $(DESTDIR)$(PREFIX)/share/man/man1
-	install $(MANS) $(DESTDIR)$(PREFIX)/share/man/man1
-
 check: isutf8
 	./is_utf8/test.sh
 
