--- a/conf/post-install
+++ b/conf/post-install
@@ -366,10 +366,10 @@ test -f $config_directory/main.cf && {
 
 # Sanity checks
 
-case $manpage_directory in
- no) echo $0: Error: manpage_directory no longer accepts \"no\" values. 1>&2
-     echo Try again with \"$0 manpage_directory=/pathname ...\". 1>&2; exit 1;;
-esac
+#case $manpage_directory in
+# no) echo $0: Error: manpage_directory no longer accepts \"no\" values. 1>&2
+#     echo Try again with \"$0 manpage_directory=/pathname ...\". 1>&2; exit 1;;
+#esac
 
 case $setgid_group in
  no) echo $0: Error: setgid_group no longer accepts \"no\" values. 1>&2
@@ -377,7 +377,7 @@ case $setgid_group in
 esac
 
 for path in "$daemon_directory" "$command_directory" "$queue_directory" \
-    "$sendmail_path" "$newaliases_path" "$mailq_path" "$manpage_directory" \
+    "$sendmail_path" "$newaliases_path" "$mailq_path" \
     "$meta_directory"
 do
    case "$path" in
@@ -386,7 +386,7 @@ do
    esac
 done
 
-for path in "$html_directory" "$readme_directory" "$shlib_directory"
+for path in "$html_directory" "$readme_directory" "$shlib_directory" "$manpage_directory"
 do
    case "$path" in
    /*) ;;
--- a/postfix-install
+++ b/postfix-install
@@ -563,13 +563,13 @@ case "$setgid_group" in
      exit 1;;
 esac
 
-case "$manpage_directory" in
- no) (echo $0: Error: the manpage_directory parameter no longer accepts 
-     echo \"no\" values.  Try again with \"manpage_directory=/path/name\" 
-     echo on the command line or execute \"make install\" and specify
-     echo manpage_directory interactively.) | ${FMT} 1>&2
-     exit 1;;
-esac
+#case "$manpage_directory" in
+# no) (echo $0: Error: the manpage_directory parameter no longer accepts 
+#     echo \"no\" values.  Try again with \"manpage_directory=/path/name\" 
+#     echo on the command line or execute \"make install\" and specify
+#     echo manpage_directory interactively.) | ${FMT} 1>&2
+#     exit 1;;
+#esac
 
 for path in "$html_directory" "$readme_directory" "$shlib_directory"
 do
@@ -582,7 +582,7 @@ do
 done
 
 for path in "$daemon_directory" "$data_directory" "$command_directory" "$queue_directory" \
-    "$sendmail_path" "$newaliases_path" "$mailq_path" "$manpage_directory" \
+    "$sendmail_path" "$newaliases_path" "$mailq_path" \
     "$meta_directory"
 do
    case "$path" in
@@ -789,8 +789,8 @@ do
 		compare_or_replace $mode "$owner" "$group" html/$file \
 		    $HTML_DIRECTORY/$file || exit 1;;
 	'$manpage_directory')
-	    check_parent $MANPAGE_DIRECTORY/$file || exit 1
-	    compare_or_replace $mode "$owner" "$group" man/$file \
+	    test "$manpage_directory" = "no" || check_parent $MANPAGE_DIRECTORY/$file || exit 1
+	    test "$manpage_directory" = "no" || compare_or_replace $mode "$owner" "$group" man/$file \
 		$MANPAGE_DIRECTORY/$file || exit 1;;
 	'$readme_directory')
 	    test "$readme_directory" = "no" ||
