--- a/src/util/dict_db.c
+++ b/src/util/dict_db.c
@@ -750,8 +750,8 @@ static DICT *dict_db_open(const char *cl
 	msg_fatal("create DB database: %m");
     if (db == 0)
 	msg_panic("db_create null result");
-    if (type == DB_HASH && db->set_h_nelem(db, DICT_DB_NELM) != 0)
-	msg_fatal("set DB hash element count %d: %m", DICT_DB_NELM);
+//    if (type == DB_HASH && db->set_h_nelem(db, DICT_DB_NELM) != 0)
+//	msg_fatal("set DB hash element count %d: %m", DICT_DB_NELM);
     db_base_buf = vstring_alloc(100);
 #if DB_VERSION_MAJOR == 6 || DB_VERSION_MAJOR == 5 || \
 	(DB_VERSION_MAJOR == 4 && DB_VERSION_MINOR > 0)
