--- a/src/util/sys_defs.h
+++ b/src/util/sys_defs.h
@@ -760,9 +760,8 @@ extern int initgroups(const char *, int)
 #define INTERNAL_LOCK	MYFLOCK_STYLE_FLOCK
 #define DEF_MAILBOX_LOCK "fcntl, dotlock"	/* RedHat >= 4.x */
 #define HAS_FSYNC
-#define HAS_DB
 #define NATIVE_DB_TYPE	"hash"
-#define ALIAS_DB_MAP	DEF_DB_TYPE ":/etc/aliases"
+#define ALIAS_DB_MAP	DEF_DB_TYPE ":/etc/postfix/aliases"
 #ifndef NO_NIS
 #define HAS_NIS
 #endif
