--- a/postfix-install
+++ b/postfix-install
@@ -861,23 +861,23 @@ do
     esac
 done
 
-bin/postconf -c $CONFIG_DIRECTORY -e \
-    "daemon_directory = $daemon_directory" \
-    "data_directory = $data_directory" \
-    "command_directory = $command_directory" \
-    "queue_directory = $queue_directory" \
-    "mail_owner = $mail_owner" \
-    "setgid_group = $setgid_group" \
-    "sendmail_path = $sendmail_path" \
-    "mailq_path = $mailq_path" \
-    "newaliases_path = $newaliases_path" \
-    "html_directory = $html_directory" \
-    "manpage_directory = $manpage_directory" \
-    "sample_directory = $sample_directory" \
-    "readme_directory = $readme_directory" \
-    "shlib_directory = $shlib_directory" \
-    "meta_directory = $meta_directory" \
-|| exit 1
+#bin/postconf -c $CONFIG_DIRECTORY -e \
+#    "daemon_directory = $daemon_directory" \
+#    "data_directory = $data_directory" \
+#    "command_directory = $command_directory" \
+#    "queue_directory = $queue_directory" \
+#    "mail_owner = $mail_owner" \
+#    "setgid_group = $setgid_group" \
+#    "sendmail_path = $sendmail_path" \
+#    "mailq_path = $mailq_path" \
+#    "newaliases_path = $newaliases_path" \
+#    "html_directory = $html_directory" \
+#    "manpage_directory = $manpage_directory" \
+#    "sample_directory = $sample_directory" \
+#    "readme_directory = $readme_directory" \
+#    "shlib_directory = $shlib_directory" \
+#    "meta_directory = $meta_directory" \
+#|| exit 1
 
 # If Postfix is being installed locally from source code, do the
 # post-install processing now.
