From a758e0870c6d1e4b0272f6e7f9efa9face5534bb Mon Sep 17 00:00:00 2001
From: John Crispin <blogic@openwrt.org>
Date: Sun, 27 Jul 2014 09:49:07 +0100
Subject: [PATCH 32/53] USB: dwc2: add device_reset()

Signed-off-by: John Crispin <blogic@openwrt.org>
---
 drivers/usb/dwc2/hcd.c |    3 +++
 1 file changed, 3 insertions(+)

--- a/drivers/usb/dwc2/hcd.c
+++ b/drivers/usb/dwc2/hcd.c
@@ -48,6 +48,7 @@
 #include <linux/io.h>
 #include <linux/slab.h>
 #include <linux/usb.h>
+#include <linux/reset.h>
 
 #include <linux/usb/hcd.h>
 #include <linux/usb/ch11.h>
@@ -5104,6 +5105,8 @@ int dwc2_hcd_init(struct dwc2_hsotg *hso
 
 	retval = -ENOMEM;
 
+	device_reset(hsotg->dev);
+
 	hcfg = dwc2_readl(hsotg->regs + HCFG);
 	dev_dbg(hsotg->dev, "hcfg=%08x\n", hcfg);
 
