From a7eb46e0ea4a11e4dfb56ab129bf816d1059a6c5 Mon Sep 17 00:00:00 2001
From: John Crispin <blogic@openwrt.org>
Date: Mon, 7 Dec 2015 17:31:08 +0100
Subject: [PATCH 51/53] serial: add ugly custom baud rate hack

Signed-off-by: John Crispin <blogic@openwrt.org>
---
 drivers/tty/serial/serial_core.c |    3 +++
 1 file changed, 3 insertions(+)

--- a/drivers/tty/serial/serial_core.c
+++ b/drivers/tty/serial/serial_core.c
@@ -428,6 +428,9 @@ uart_get_baud_rate(struct uart_port *por
 		break;
 	}
 
+	if (tty_termios_baud_rate(termios) == 2500000)
+		return 250000;
+
 	for (try = 0; try < 2; try++) {
 		baud = tty_termios_baud_rate(termios);
 
