--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -58,9 +58,9 @@ find_library(JPEG_LIB jpeg)
 
 add_subdirectory(plugins/input_file)
 add_subdirectory(plugins/input_http)
-add_subdirectory(plugins/input_opencv)
-add_subdirectory(plugins/input_raspicam)
-add_subdirectory(plugins/input_ptp2)
+#add_subdirectory(plugins/input_opencv)
+#add_subdirectory(plugins/input_raspicam)
+#add_subdirectory(plugins/input_ptp2)
 add_subdirectory(plugins/input_uvc)
 
 #
@@ -71,7 +71,7 @@ add_subdirectory(plugins/output_file)
 add_subdirectory(plugins/output_http)
 add_subdirectory(plugins/output_rtsp)
 add_subdirectory(plugins/output_udp)
-add_subdirectory(plugins/output_viewer)
+#add_subdirectory(plugins/output_viewer)
 add_subdirectory(plugins/output_zmqserver)
 
 #
