--- a/src/shairplay.c
+++ b/src/shairplay.c
@@ -350,7 +350,7 @@ main(int argc, char *argv[])
 	raop_cbs.audio_destroy = audio_destroy;
 	raop_cbs.audio_set_volume = audio_set_volume;
 
-	raop = raop_init_from_keyfile(10, &raop_cbs, "airport.key", NULL);
+	raop = raop_init_from_keyfile(10, &raop_cbs, "/usr/share/shairplay/airport.key", NULL);
 	if (raop == NULL) {
 		fprintf(stderr, "Could not initialize the RAOP service\n");
 		fprintf(stderr, "Please make sure the airport.key file is in the current directory.\n");
