From 031c3d802e51bbc233b1044f812402a66bfcf237 Mon Sep 17 00:00:00 2001
From: Memphiz <memphis@machzwo.de>
Date: Fri, 21 Dec 2018 20:39:11 +0100
Subject: [PATCH] [configure] - only check for dns_sd.h in case libdl was not
 found

---
 configure.ac | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/configure.ac
+++ b/configure.ac
@@ -19,7 +19,7 @@ LT_LIB_DLLOAD
 
 # Checks for header files.
 AC_HEADER_STDC
-if test yes = "$libltdl_cv_func_dlopen" || test yes = "$libltdl_cv_lib_dl_dlopen"
+if test no = "$libltdl_cv_func_dlopen" && test no = "$libltdl_cv_lib_dl_dlopen"
 then
   AC_CHECK_HEADERS([dns_sd.h], [],
                    [AC_MSG_ERROR([Could not find dns_sd.h header, please install libavahi-compat-libdnssd-dev or equivalent.])])
