From fee63b81ee432d0b3e03ffcf9f54cda99476f267 Mon Sep 17 00:00:00 2001
From: Jan Engelhardt <jengelh@inai.de>
Date: Sun, 24 Jun 2012 14:57:48 +0200
Subject: [PATCH 2/6] build: remove unused lines in Makefile.am

Signed-off-by: Jan Engelhardt <jengelh@inai.de>
---
 Makefile.am |    2 --
 1 file changed, 2 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index b758e56..9a1cbcb 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -1,6 +1,4 @@
-AM_CPPFLAGS = -I$(top_srcdir)/include
 SUBDIRS	= include src utils
-LINKOPTS = -lnfnetlink
 
 ACLOCAL_AMFLAGS = -I m4
 
-- 
1.7.10.4

