--- a/configure.ac
+++ b/configure.ac
@@ -4140,7 +4140,6 @@ AC_CONFIG_FILES([support/wolfssl.pc])
 AC_CONFIG_FILES([rpm/spec])
 
 AX_CREATE_GENERIC_CONFIG
-AX_AM_JOBSERVER([yes])
 
 AC_OUTPUT
 
--- a/Makefile.am
+++ b/Makefile.am
@@ -20,8 +20,6 @@ dist_noinst_SCRIPTS =
 noinst_SCRIPTS =
 check_SCRIPTS =
 
-#includes additional rules from aminclude.am
-@INC_AMINCLUDE@
 DISTCLEANFILES+= aminclude.am
 
 CLEANFILES+= cert.der \
