From e5b5361651940ff5c0c1784dfd0130abec7ab535 Mon Sep 17 00:00:00 2001
From: wevsty <ty@wevs.org>
Date: Mon, 24 Aug 2020 02:27:11 +0800
Subject: [PATCH] rockchip: rk3328: enable hardware rng for NanoPi R2S


Signed-off-by: wevsty <ty@wevs.org>
---

--- a/arch/arm64/boot/dts/rockchip/rk3328-nanopi-r2s.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3328-nanopi-r2s.dts
@@ -411,3 +411,7 @@
 		realtek,led-data = <0x87>;
 	};
 };
+
+&rng {
+	status = "okay";
+};
