From 0fcb001e301b7dec639f948de49f69666fe91669 Mon Sep 17 00:00:00 2001
From: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Date: Fri, 2 Sep 2022 18:31:11 -0500
Subject: [PATCH 1010/1011] Simplify "ipq6018: rproc: Add non secure Q6 bringup
 sequence"

---
 drivers/remoteproc/qcom_q6v5_wcss.c | 25 +++++++++----------------
 1 file changed, 9 insertions(+), 16 deletions(-)

--- a/drivers/remoteproc/qcom_q6v5_wcss.c
+++ b/drivers/remoteproc/qcom_q6v5_wcss.c
@@ -628,18 +628,13 @@ static void q6v5_wcss_halt_axi_port(stru
 	unsigned long timeout;
 	unsigned int val;
 	int ret;
-	const struct wcss_data *desc = device_get_match_data(wcss->dev);
 
-	if (desc != &wcss_ipq6018_res_init) {
-		/* Check if we're already idle */
-		ret = regmap_read(halt_map, offset + AXI_IDLE_REG, &val);
-		if (!ret && val)
-			return;
-	}
+	/* Check if we're already idle */
+	ret = regmap_read(halt_map, offset + AXI_IDLE_REG, &val);
+	if (!ret && val)
+		return;
 
 	/* Assert halt request */
-	regmap_read(halt_map, offset + AXI_HALTREQ_REG, &val);
-	val |= BIT(0);
 	regmap_write(halt_map, offset + AXI_HALTREQ_REG, 1);
 
 	/* Wait for halt */
@@ -652,14 +647,12 @@ static void q6v5_wcss_halt_axi_port(stru
 		msleep(1);
 	}
 
-	if (desc != &wcss_ipq6018_res_init) {
-		ret = regmap_read(halt_map, offset + AXI_IDLE_REG, &val);
-		if (ret || !val)
-			dev_err(wcss->dev, "port failed halt\n");
+	ret = regmap_read(halt_map, offset + AXI_IDLE_REG, &val);
+	if (ret || !val)
+		dev_err(wcss->dev, "port failed halt\n");
 
-		/* Clear halt request (port will remain halted until reset) */
-		regmap_write(halt_map, offset + AXI_HALTREQ_REG, 0);
-	}
+	/* Clear halt request (port will remain halted until reset) */
+	regmap_write(halt_map, offset + AXI_HALTREQ_REG, 0);
 }
 
 static int q6v5_qcs404_wcss_shutdown(struct q6v5_wcss *wcss)
