--- a/sysincludes.h
+++ b/sysincludes.h
@@ -101,14 +101,8 @@ typedef void *caddr_t;
 #if defined __GNUC__ && defined __STDC__
 /* gcc -traditional doesn't have PACKED, UNUSED and NORETURN */
 # define PACKED __attribute__ ((packed))
-# if __GNUC__ == 2 && __GNUC_MINOR__ > 6 || __GNUC__ >= 3
-/* gcc 2.6.3 doesn't have "unused" */		/* mool */
-#  define UNUSED(x) x __attribute__ ((unused));x
-#  define UNUSEDP __attribute__ ((unused))
-# else
-#  define UNUSED(x) x
-#  define UNUSEDP /* */
-# endif
+# define UNUSED(x) x
+# define UNUSEDP /* */
 # define NORETURN __attribute__ ((noreturn))
 #else
 # define UNUSED(x) x
