--- a/drivers/net/phy/Kconfig
+++ b/drivers/net/phy/Kconfig
@@ -135,6 +135,8 @@ config RTL8367B_PHY
 
 endif # RTL8366_SMI
 
+source "drivers/net/phy/mtk/mt753x/Kconfig"
+
 comment "MII PHY device drivers"
 
 config AMD_PHY
--- a/drivers/net/phy/Makefile
+++ b/drivers/net/phy/Makefile
@@ -94,3 +94,4 @@ obj-$(CONFIG_STE10XP)		+= ste10Xp.o
 obj-$(CONFIG_TERANETICS_PHY)	+= teranetics.o
 obj-$(CONFIG_VITESSE_PHY)	+= vitesse.o
 obj-$(CONFIG_XILINX_GMII2RGMII) += xilinx_gmii2rgmii.o
+obj-$(CONFIG_MT753X_GSW)	+= mtk/mt753x/

--- a/drivers/net/ethernet/Kconfig	2022-03-28 15:57:11.000000000 +0800
+++ b/drivers/net/ethernet/Kconfig	2022-04-30 15:35:42.047114136 +0800
@@ -119,6 +119,7 @@
 
 source "drivers/net/ethernet/marvell/Kconfig"
 source "drivers/net/ethernet/mediatek/Kconfig"
+source "drivers/net/ethernet/mediateksdk/Kconfig"
 source "drivers/net/ethernet/mellanox/Kconfig"
 source "drivers/net/ethernet/micrel/Kconfig"
 source "drivers/net/ethernet/microchip/Kconfig"

--- a/drivers/net/ethernet/Makefile	2022-03-28 15:57:11.000000000 +0800
+++ b/drivers/net/ethernet/Makefile	2022-04-30 15:34:53.069392279 +0800
@@ -52,6 +52,7 @@
 obj-$(CONFIG_LANTIQ_XRX200) += lantiq_xrx200.o
 obj-$(CONFIG_NET_VENDOR_MARVELL) += marvell/
 obj-$(CONFIG_NET_VENDOR_MEDIATEK) += mediatek/
+obj-$(CONFIG_NET_VENDOR_MEDIATEK_SDK) += mediateksdk/
 obj-$(CONFIG_NET_VENDOR_MELLANOX) += mellanox/
 obj-$(CONFIG_NET_VENDOR_MICREL) += micrel/
 obj-$(CONFIG_NET_VENDOR_MICROCHIP) += microchip/

