--- a/include/linux/crypto.h	2021-07-19 15:45:03.000000000 +0800
+++ b/include/linux/crypto.h	2021-07-31 00:41:39.404278501 +0800
@@ -101,6 +101,11 @@
 #define CRYPTO_NOLOAD			0x00008000
 
 /*
+ * Transform masks and values (for crt_flags).
+ */
+#define CRYPTO_TFM_RES_BAD_KEY_LEN   	0x00200000
+
+/*
  * The algorithm may allocate memory during request processing, i.e. during
  * encryption, decryption, or hashing.  Users can request an algorithm with this
  * flag unset if they can't handle memory allocation failures.
