--- a/arch/arm/boot/dts/broadcom/bcm958625-meraki-alamo.dtsi
+++ b/arch/arm/boot/dts/broadcom/bcm958625-meraki-alamo.dtsi
@@ -254,11 +254,17 @@
 		port@0 {
 			label = "wan1";
 			reg = <0>;
+
+			nvmem-cells = <&mac_address 1>;
+			nvmem-cell-names = "mac-address";
 		};
 
 		port@1 {
 			label = "wan2";
 			reg = <1>;
+
+			nvmem-cells = <&mac_address 1>;
+			nvmem-cell-names = "mac-address";
 		};
 
 		sgmii0: port@4 {
--- a/arch/arm/boot/dts/broadcom/bcm958625-meraki-kingpin.dtsi
+++ b/arch/arm/boot/dts/broadcom/bcm958625-meraki-kingpin.dtsi
@@ -154,6 +154,9 @@
 		port@4 {
 			label = "wan";
 			reg = <4>;
+
+			nvmem-cells = <&mac_address 1>;
+			nvmem-cell-names = "mac-address";
 		};
 
 		port@8 {
--- a/arch/arm/boot/dts/broadcom/bcm958625-meraki-mx6x-common.dtsi
+++ b/arch/arm/boot/dts/broadcom/bcm958625-meraki-mx6x-common.dtsi
@@ -39,7 +39,7 @@
 
 &amac2 {
 	status = "okay";
-	nvmem-cells = <&mac_address>;
+	nvmem-cells = <&mac_address 0>;
 	nvmem-cell-names = "mac-address";
 };
 
@@ -62,7 +62,9 @@
 			#size-cells = <1>;
 
 			mac_address: mac-address@66 {
+				compatible = "mac-base";
 				reg = <0x66 0x6>;
+				#nvmem-cell-cells = <1>;
 			};
 		};
 	};
