--- ./simulated-driver/sfe_drv.c.orig	2020-06-16 12:49:47.680153371 +0800
+++ ./simulated-driver/sfe_drv.c	2020-06-16 12:50:18.540153371 +0800
@@ -1167,7 +1167,7 @@ int sfe_drv_recv(struct sk_buff *skb)
 	 * If ingress Qdisc configured, and packet not processed by ingress Qdisc yet
 	 * We can not accelerate this packet.
 	 */
-	if (dev->ingress_queue && !(skb->tc_verd & TC_NCLS)) {
+	if (dev->ingress_queue && !(skb->tc_verd_qca_nss & TC_NCLS)) {
 		return 0;
 	}
 #endif
