--- a/lib/stdio.in.h
+++ b/lib/stdio.in.h
@@ -835,14 +835,6 @@ _GL_WARN_ON_USE (getline, "getline is un
 # endif
 #endif
 
-/* It is very rare that the developer ever has full control of stdin,
-   so any use of gets warrants an unconditional warning; besides, C11
-   removed it.  */
-#undef gets
-#if HAVE_RAW_DECL_GETS && !defined __cplusplus
-_GL_WARN_ON_USE (gets, "gets is a security hole - use fgets instead");
-#endif
-
 #if defined _WIN32 && !defined __CYGWIN__
 # undef getw
 # define getw _getw
