--- a/Makefile
+++ b/Makefile
@@ -1,21 +1,7 @@
 OBJ = progress
 CFLAGS ?= -g
 override CFLAGS += -Wall -D_FILE_OFFSET_BITS=64
-override LDFLAGS += -lm
-UNAME := $(shell uname)
-PKG_CONFIG ?= pkg-config
-ifeq ($(UNAME), Linux)
-    ifeq (, $(shell which $(PKG_CONFIG) 2> /dev/null))
-    $(error "pkg-config command not found")
-    endif
-    ifeq (, $(shell $(PKG_CONFIG) ncurses --libs 2> /dev/null))
-    $(error "ncurses package not found")
-    endif
-    override LDFLAGS += $(shell $(PKG_CONFIG) ncurses --libs)
-endif
-ifeq ($(UNAME), Darwin)
-    override LDFLAGS += -lncurses
-endif
+override LDFLAGS += -lm -lncursesw
 PREFIX ?= /usr/local
 BINDIR = $(PREFIX)/bin
 MANDIR = $(PREFIX)/share/man/man1
