--- a/arch/mips/bcm63xx/boards/board_common.c
+++ b/arch/mips/bcm63xx/boards/board_common.c
@@ -98,6 +98,8 @@ void __init board_early_setup(const stru
 		if (BCMCPU_IS_6348())
 			val |= GPIO_MODE_6348_G3_EXT_MII |
 				GPIO_MODE_6348_G0_EXT_MII;
+		else if (BCMCPU_IS_6358())
+			val |= GPIO_MODE_6358_ENET1_MII_CLK_INV;
 	}
 
 	bcm_gpio_writel(val, GPIO_MODE_REG);
--- a/arch/mips/include/asm/mach-bcm63xx/bcm63xx_regs.h
+++ b/arch/mips/include/asm/mach-bcm63xx/bcm63xx_regs.h
@@ -652,6 +652,8 @@
 #define GPIO_MODE_6358_EXTRA_SPI_SS	(1 << 7)
 #define GPIO_MODE_6358_SERIAL_LED	(1 << 10)
 #define GPIO_MODE_6358_UTOPIA		(1 << 12)
+#define GPIO_MODE_6358_ENET1_MII_CLK_INV (1 << 30)
+#define GPIO_MODE_6358_ENET0_MII_CLK_INV (1 << 31)
 
 #define GPIO_MODE_6368_ANALOG_AFE_0	(1 << 0)
 #define GPIO_MODE_6368_ANALOG_AFE_1	(1 << 1)
