--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -77,6 +77,8 @@ static struct board_info __initdata boar
 
 	.has_enet0			= 1,
 	.enet0 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -90,6 +92,8 @@ static struct board_info __initdata boar
 
 	.has_enet0			= 1,
 	.enet0 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -138,6 +142,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -162,6 +168,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -186,6 +194,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -208,6 +218,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -231,6 +243,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -252,6 +266,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -271,6 +287,8 @@ static struct board_info __initdata boar
 		.use_internal_phy	= 1,
 	},
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -298,6 +316,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -323,6 +343,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -348,6 +370,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
@@ -371,6 +395,8 @@ static struct board_info __initdata boar
 	},
 
 	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
 		.force_speed_100	= 1,
 		.force_duplex_full	= 1,
 	},
