--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -46,7 +46,7 @@ static struct board_info __initdata boar
 	.ephy_reset_gpio		= 36,
 	.ephy_reset_gpio_flags		= GPIO_ACTIVE_LOW,
 };
-#endif
+#endif /* CONFIG_BCM63XX_CPU_3368 */
 
 /*
  * known 6328 boards
@@ -65,7 +65,7 @@ static struct board_info __initdata boar
 		.port_no		= 0,
 	},
 };
-#endif
+#endif /* CONFIG_BCM63XX_CPU_6328 */
 
 /*
  * known 6338 boards
@@ -98,7 +98,7 @@ static struct board_info __initdata boar
 		.force_duplex_full	= 1,
 	},
 };
-#endif
+#endif /* CONFIG_BCM63XX_CPU_6338 */
 
 /*
  * known 6345 boards
@@ -108,7 +108,7 @@ static struct board_info __initdata boar
 	.name				= "96345GW2",
 	.expected_cpu_id		= 0x6345,
 };
-#endif
+#endif /* CONFIG_BCM63XX_CPU_6345 */
 
 /*
  * known 6348 boards
@@ -295,7 +295,7 @@ static struct board_info __initdata boar
 
 	.has_ohci0 = 1,
 };
-#endif
+#endif /* CONFIG_BCM63XX_CPU_6348 */
 
 /*
  * known 6358 boards
@@ -403,7 +403,7 @@ static struct board_info __initdata boar
 
 	.has_ohci0			= 1,
 };
-#endif
+#endif /* CONFIG_BCM63XX_CPU_6358 */
 
 /*
  * all boards
