--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -452,6 +452,22 @@ static struct board_info __initdata boar
 	.has_ohci0			= 1,
 };
 
+/* D-Link DSL-274xB revison C2/C3 */
+static struct board_info __initdata board_dsl_274xb_rev_c = {
+	.name				= "AW4139",
+	.expected_cpu_id		= 0x6358,
+
+	.has_enet1			= 1,
+	.has_pci			= 1,
+
+	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
+		.force_speed_100	= 1,
+		.force_duplex_full	= 1,
+	},
+};
+
 static struct board_info __initdata board_nb4_ser_r0 = {
 	.name				= "NB4-SER-r0",
 	.expected_cpu_id		= 0x6358,
@@ -543,6 +559,7 @@ static const struct board_info __initcon
 	&board_AGPFS0,
 	&board_CPVA642,
 	&board_DWVS0,
+	&board_dsl_274xb_rev_c,
 	&board_nb4_ser_r0,
 	&board_nb4_fxc_r1,
 #endif
@@ -582,6 +599,7 @@ static struct of_device_id const bcm963x
 	{ .compatible = "alcatel,rg100a", .data = &board_96358vw2, },
 	{ .compatible = "brcm,bcm96358vw", .data = &board_96358vw, },
 	{ .compatible = "brcm,bcm96358vw2", .data = &board_96358vw2, },
+	{ .compatible = "d-link,dsl-274xb-c2", .data = &board_dsl_274xb_rev_c, },
 	{ .compatible = "d-link,dsl-2650u", .data = &board_96358vw2, },
 	{ .compatible = "pirelli,a226g", .data = &board_DWVS0, },
 	{ .compatible = "pirelli,a226m", .data = &board_DWVS0, },
