--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -216,6 +216,24 @@ static struct board_info __initdata boar
 	.has_ehci0 = 1,
 };
 
+
+/* BT Voyager 2110 */
+static struct board_info __initdata board_V2110 = {
+	.name				= "V2110",
+	.expected_cpu_id		= 0x6348,
+
+	.has_enet1			= 1,
+	.has_pci			= 1,
+
+	.enet1 = {
+		.has_phy		= 1,
+		.phy_id			= 0,
+		.force_speed_100	= 1,
+		.force_duplex_full	= 1,
+	},
+};
+
+
 static struct board_info __initdata board_96348gw = {
 	.name				= "96348GW",
 	.expected_cpu_id		= 0x6348,
@@ -776,6 +794,7 @@ static const struct board_info __initcon
 	&board_spw500v,
 	&board_96348sv,
 	&board_V2500V_BB,
+	&board_V2110,
 #endif
 
 #ifdef CONFIG_BCM63XX_CPU_6358
@@ -815,6 +834,7 @@ static struct of_device_id const bcm963x
 	{ .compatible = "brcm,bcm96348gw-10", .data = &board_96348gw_10, },
 	{ .compatible = "brcm,bcm96348gw-11", .data = &board_96348gw_11, },
 	{ .compatible = "brcm,bcm96348gw-a", .data = &board_96348gw_a, },
+	{ .compatible = "bt,v2110", .data = &board_V2110, },
 	{ .compatible = "bt,v2500v-bb", .data = &board_V2500V_BB, },
 	{ .compatible = "d-link,dsl-2640b-b", .data = &board_96348_D4PW, },
 	{ .compatible = "davolink,dv-201amr", .data = &board_DV201AMR, },
