--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -1672,6 +1672,59 @@ static struct board_info __initdata boar
 	},
 };
 
+static struct sprom_fixup __initdata vr3025un_fixups[] = {
+	{ .offset = 97, .value = 0xfeb3 },
+	{ .offset = 98, .value = 0x1618 },
+	{ .offset = 99, .value = 0xfab0 },
+	{ .offset = 113, .value = 0xfed1 },
+	{ .offset = 114, .value = 0x1609 },
+	{ .offset = 115, .value = 0xfad9 },
+};
+
+static struct board_info __initdata board_VR3025un = {
+	.name					= "96368M-1341N",
+	.expected_cpu_id		= 0x6368,
+
+	.has_pci				= 1,
+	.use_fallback_sprom			= 1,
+	.has_ohci0				= 1,
+	.has_ehci0				= 1,
+
+	.has_enetsw				= 1,
+	.enetsw = {
+		.used_ports = {
+			[0] = {
+				.used		= 1,
+				.phy_id		= 1,
+				.name		= "port1",
+			},
+			[1] = {
+				.used		= 1,
+				.phy_id		= 2,
+				.name		= "port2",
+			},
+			[2] = {
+				.used		= 1,
+				.phy_id		= 3,
+				.name		= "port3",
+			},
+			[3] = {
+				.used		= 1,
+				.phy_id		= 4,
+				.name		= "port4",
+			},
+		},
+	},
+
+	.fallback_sprom = {
+		.type 				= SPROM_BCM43222,
+		.pci_bus			= 0,
+		.pci_dev			= 1,
+		.board_fixups			= vr3025un_fixups,
+		.num_board_fixups		= ARRAY_SIZE(vr3025un_fixups),
+	},
+};
+
 static struct sprom_fixup __initdata wap5813n_fixups[] = {
 	{ .offset = 97, .value = 0xfeed },
 	{ .offset = 98, .value = 0x15d1 },
@@ -1943,6 +1996,7 @@ static const struct board_info __initcon
 	&board_96368mvwg,
 	&board_96368mvngr,
 	&board_VR3025u,
+	&board_VR3025un,
 	&board_WAP5813n,
 #endif
 #ifdef CONFIG_BCM63XX_CPU_63268
@@ -2037,6 +2091,7 @@ static struct of_device_id const bcm963x
 	{ .compatible = "brcm,bcm96368mvngr", .data = &board_96368mvngr, },
 	{ .compatible = "brcm,bcm96368mvwg", .data = &board_96368mvwg, },
 	{ .compatible = "comtrend,vr-3025u", .data = &board_VR3025u, },
+	{ .compatible = "comtrend,vr-3025un", .data = &board_VR3025un, },
 	{ .compatible = "comtrend,wap-5813n", .data = &board_WAP5813n, },
 #endif
 #ifdef CONFIG_BCM63XX_CPU_63268
