--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -1580,6 +1580,48 @@ static struct sprom_fixup __initdata vr3
 	{ .offset = 115, .value = 0xfad9 },
 };
 
+static struct board_info __initdata board_P870HW51A_V2 = {
+	.name				= "P870HW-51a_v2",
+	.expected_cpu_id		= 0x6368,
+
+	.has_pci			= 1,
+	.use_fallback_sprom		= 1,
+	.has_ohci0			= 1,
+	.has_ehci0			= 1,
+
+	.has_enetsw			= 1,
+	.enetsw = {
+		.used_ports = {
+			[0] = {
+				.used	= 1,
+				.phy_id	= 1,
+				.name	= "port1",
+			},
+			[1] = {
+				.used	= 1,
+				.phy_id	= 2,
+				.name	= "port2",
+			},
+			[2] = {
+				.used	= 1,
+				.phy_id	= 3,
+				.name	= "port3",
+			},
+			[3] = {
+				.used	= 1,
+				.phy_id	= 4,
+				.name	= "port4",
+			},
+		},
+	},
+
+	.fallback_sprom = {
+		.type 				= SPROM_BCM4318,
+		.pci_bus			= 0,
+		.pci_dev			= 1,
+	},
+};
+
 static struct board_info __initdata board_VR3025u = {
 	.name					= "96368M-1541N",
 	.expected_cpu_id		= 0x6368,
@@ -1947,6 +1989,7 @@ static const struct board_info __initcon
 #ifdef CONFIG_BCM63XX_CPU_6368
 	&board_96368mvwg,
 	&board_96368mvngr,
+	&board_P870HW51A_V2,
 	&board_VR3025u,
 	&board_VR3025un,
 	&board_WAP5813n,
@@ -2045,6 +2088,7 @@ static struct of_device_id const bcm963x
 	{ .compatible = "comtrend,vr-3025u", .data = &board_VR3025u, },
 	{ .compatible = "comtrend,vr-3025un", .data = &board_VR3025un, },
 	{ .compatible = "comtrend,wap-5813n", .data = &board_WAP5813n, },
+	{ .compatible = "zyxel,p870hw-51a-v2", .data = &board_P870HW51A_V2, },
 #endif
 #ifdef CONFIG_BCM63XX_CPU_63268
 	{ .compatible = "brcm,bcm963268bu_p300", .data = &board_963268bu_p300, },
