--- a/include/configs/mt7986.h
+++ b/include/configs/mt7986.h
@@ -11,6 +11,11 @@
 
 #include <linux/sizes.h>
 
+#define CONFIG_SYS_MAXARGS		32
+#define CONFIG_SYS_BOOTM_LEN		SZ_128M
+#define CONFIG_SYS_CBSIZE		SZ_1K
+#define CONFIG_SYS_PBSIZE		(CONFIG_SYS_CBSIZE +	\
+					sizeof(CONFIG_SYS_PROMPT) + 16)
 #define CONFIG_SYS_NONCACHED_MEMORY	SZ_1M
 #define CONFIG_SYS_MMC_ENV_DEV		0
 
@@ -19,6 +24,11 @@
 
 /* SPL -> Uboot */
 #define CONFIG_SYS_UBOOT_START		CONFIG_SYS_TEXT_BASE
+#define CONFIG_SYS_INIT_SP_ADDR		(CONFIG_SYS_TEXT_BASE + SZ_2M - \
+					 GENERATED_GBL_DATA_SIZE)
+
+/* Flash */
+#define CONFIG_SYS_NAND_MAX_CHIPS	1
 
 /* DRAM */
 #define CONFIG_SYS_SDRAM_BASE		0x40000000
