--- a/arch/arm/dts/mt7622-bananapi-bpi-r64.dts
+++ b/arch/arm/dts/mt7622-bananapi-bpi-r64.dts
@@ -20,7 +20,7 @@
 	};
 
 	aliases {
-		spi0 = &snfi;
+		spi0 = &snand;
 		ethernet0 = &eth;
 	};
 
@@ -208,17 +208,11 @@
 	};
 };
 
-&snfi {
-	pinctrl-names = "default", "snfi";
-	pinctrl-0 = <&snor_pins>;
-	pinctrl-1 = <&snfi_pins>;
+&snand {
+	pinctrl-names = "default";
+	pinctrl-0 = <&snfi_pins>;
+	quad-spi;
 	status = "okay";
-
-	spi-flash@0{
-		compatible = "jedec,spi-nor";
-		reg = <0>;
-		bootph-all;
-	};
 };
 
 &uart0 {
