--- a/arch/arm/dts/mt7623n-bananapi-bpi-r2.dts
+++ b/arch/arm/dts/mt7623n-bananapi-bpi-r2.dts
@@ -51,19 +51,19 @@
 
 		blue {
 			label = "bpi-r2:pio:blue";
-			gpios = <&gpio 241 GPIO_ACTIVE_HIGH>;
+			gpios = <&gpio 240 GPIO_ACTIVE_LOW>;
 			default-state = "off";
 		};
 
 		green {
 			label = "bpi-r2:pio:green";
-			gpios = <&gpio 240 GPIO_ACTIVE_HIGH>;
+			gpios = <&gpio 241 GPIO_ACTIVE_LOW>;
 			default-state = "off";
 		};
 
 		red {
 			label = "bpi-r2:pio:red";
-			gpios = <&gpio 239 GPIO_ACTIVE_HIGH>;
+			gpios = <&gpio 239 GPIO_ACTIVE_LOW>;
 			default-state = "off";
 		};
 	};
