diff --git a/src/http.h b/src/http.h
index 914815a..e312dd3 100644
--- a/src/http.h
+++ b/src/http.h
@@ -29,6 +29,6 @@
 #include <stdio.h>
 #include "protocol.h"
 
-const protocol_t *const http_protocol;
+extern const protocol_t *const http_protocol;
 
 #endif
diff --git a/src/tls.h b/src/tls.h
index 3998913..ddbee11 100644
--- a/src/tls.h
+++ b/src/tls.h
@@ -28,6 +28,6 @@
 
 #include "protocol.h"
 
-const protocol_t *const tls_protocol;
+extern const protocol_t *const tls_protocol;
 
 #endif
