--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.c
@@ -582,18 +582,28 @@
 	return 0;
 }
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 6, 0)
 static int brcmf_common_pd_remove(struct platform_device *pdev)
+#else
+static void brcmf_common_pd_remove(struct platform_device *pdev)
+#endif
 {
 	brcmf_dbg(INFO, "Enter\n");
 
 	if (brcmfmac_pdata->power_off)
 		brcmfmac_pdata->power_off();
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 6, 0)
 	return 0;
+#endif
 }
 
 static struct platform_driver brcmf_pd = {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 6, 0)
 	.remove		= brcmf_common_pd_remove,
+#else
+	.remove_new	= brcmf_common_pd_remove,
+#endif
 	.driver		= {
 		.name	= BRCMFMAC_PDATA_NAME,
 	}
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/core.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/core.c
@@ -1184,7 +1184,9 @@
 
 static const struct file_operations bus_reset_fops = {
 	.open	= simple_open,
+#if LINUX_VERSION_IS_LESS(6,11,0)
 	.llseek	= no_llseek,
+#endif
 	.write	= bus_reset_write,
 };
 
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/pcie.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/pcie.c
@@ -2707,7 +2707,9 @@
 
 
 static struct pci_driver brcmf_pciedrvr = {
+#if LINUX_VERSION_IS_LESS(6,6,0)
 	.node = {},
+#endif
 	.name = KBUILD_MODNAME,
 	.id_table = brcmf_pcie_devid_table,
 	.probe = brcmf_pcie_probe,
