--- a/iwlib.c
+++ b/iwlib.c
@@ -667,6 +667,7 @@ iw_get_basic_config(int			skfd,
 {
   struct iwreq		wrq;
 
+  memset((char *) &wrq, 0, sizeof(struct iwreq));
   memset((char *) info, 0, sizeof(struct wireless_config));
 
   /* Get wireless name */
--- a/Makefile
+++ b/Makefile
@@ -73,8 +73,8 @@ DYNAMIC_LINK= libiw.so
 # Install directories
 INSTALL_DIR= $(PREFIX)/sbin/
 INSTALL_LIB= $(PREFIX)/lib/
-INSTALL_INC= $(PREFIX)/include/
-INSTALL_MAN= $(PREFIX)/man/
+INSTALL_INC= $(PREFIX)/usr/include/
+INSTALL_MAN= $(PREFIX)/usr/share/man/
 
 # Various commands
 RM = rm -f
@@ -102,9 +102,9 @@ ifdef BUILD_WE_ESSENTIAL
 endif
 
 # Other flags
-CFLAGS=-Os -W -Wall -Wstrict-prototypes -Wmissing-prototypes -Wshadow \
+#CFLAGS=-Os -W -Wall -Wstrict-prototypes -Wmissing-prototypes -Wshadow \
 	-Wpointer-arith -Wcast-qual -Winline -I.
-#CFLAGS=-O2 -W -Wall -Wstrict-prototypes -I.
+CFLAGS=-O2 -W -Wall -Wstrict-prototypes -I.
 DEPFLAGS=-MMD
 XCFLAGS=$(CFLAGS) $(DEPFLAGS) $(WARN) $(HEADERS) $(WELIB_FLAG) $(WEDEF_FLAG)
 PICFLAG=-fPIC
