--- a/src/dsl_cpe_init_cfg.c
+++ b/src/dsl_cpe_init_cfg.c
@@ -38,7 +38,7 @@ DSL_InitData_t gInitCfgData =
       DSL_DEV_HS_TONE_GROUP_CLEANED, \
       DSL_DEV_HS_TONE_GROUP_CLEANED, \
       DSL_DEV_HS_TONE_GROUP_CLEANED, \
-      0x1E116000, 0x37, -1),
+      0x1E116000, 0x3f, -1),
    DSL_CPE_SIC_SET(DSL_TC_ATM, DSL_EMF_TC_CLEANED, DSL_EMF_TC_CLEANED, DSL_SYSTEMIF_MII, \
                    DSL_TC_EFM, DSL_EMF_TC_CLEANED, DSL_EMF_TC_CLEANED, DSL_SYSTEMIF_MII),
    DSL_CPE_MAC_CFG_SET(DSL_EFM_SPEED_100, DSL_EFM_DUPLEX_FULL, DSL_EFM_FLOWCTRL_ON, DSL_EFM_AUTONEG_OFF, \
--- a/src/dsl_cpe_control.c
+++ b/src/dsl_cpe_control.c
@@ -6761,7 +6761,7 @@ DSL_int_t dsl_cpe_daemon (
    for (nDevice = 0; nDevice < DSL_CPE_MAX_DSL_ENTITIES; nDevice++)
    {
 #if defined(INCLUDE_DSL_CPE_API_VRX)
-      sprintf (device, "%s/%d", DSL_CPE_DEVICE_NAME, nDevice);
+      sprintf (device, "%s%d", DSL_CPE_DEVICE_NAME, nDevice);
 #else
       sprintf (device, "%s", DSL_CPE_DEVICE_NAME);
 #endif /* defined(INCLUDE_DSL_CPE_API_VRX)*/
