From 505dde6ac50689ec2ba50a1f0087eb8a9ff1f419 Mon Sep 17 00:00:00 2001
From: Toni Uhlig <matzeton@googlemail.com>
Date: Mon, 18 May 2020 20:12:17 +0200
Subject: [PATCH] fix ncurses check during configure

Signed-off-by: Toni Uhlig <matzeton@googlemail.com>
---
 configure | 5 +----
 1 file changed, 1 insertion(+), 4 deletions(-)

--- a/configure
+++ b/configure
@@ -434,11 +434,8 @@ int main(void)
 EOF
 
 	$CC $CFLAGS $LDFLAGS \
-		$($PKG_CONFIG --cflags ncurses 2>> config.log) \
 		-o $TMPDIR/ncursestest $TMPDIR/ncursestest.c \
-		$($PKG_CONFIG --libs ncurses 2>> config.log \
-	          || echo '-lncurses' ) \
-			>> config.log 2>&1
+		-lncurses 2>> config.log
 	if [ ! -x $TMPDIR/ncursestest ] ; then
 		echo "[NO]"
 		MISSING_DEFS=1
