From 0f1653c82a06567ccb151e17f1fe3ff3c3b465f3 Mon Sep 17 00:00:00 2001
From: Toni Uhlig <matzeton@googlemail.com>
Date: Mon, 18 May 2020 20:27:32 +0200
Subject: [PATCH] removed flowtop pkg-config usage for ncurses

Signed-off-by: Toni Uhlig <matzeton@googlemail.com>
---
 flowtop/Makefile | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

--- a/flowtop/Makefile
+++ b/flowtop/Makefile
@@ -1,7 +1,6 @@
 flowtop-libs =	-lurcu \
 		$(shell PKG_CONFIG_PATH=$(PKG_CONFIG_PATH) $(PKG_CONFIG) --libs libnetfilter_conntrack 2> /dev/null ) \
-		$(shell PKG_CONFIG_PATH=$(PKG_CONFIG_PATH) $(PKG_CONFIG) --libs ncurses 2> /dev/null \
-			|| echo '-lncurses') \
+		-lncurses \
 		$(shell PKG_CONFIG_PATH=$(PKG_CONFIG_PATH) $(PKG_CONFIG) --libs tinfo 2> /dev/null ) \
 		-lpthread
 
@@ -30,8 +29,7 @@ flowtop-objs +=	geoip.o \
 		ioops.o
 endif
 
-flowtop-eflags = $(shell $(PKG_CONFIG) --cflags ncurses) \
-                 $(shell $(PKG_CONFIG) --cflags libnetfilter_conntrack)
+flowtop-eflags = $(shell $(PKG_CONFIG) --cflags libnetfilter_conntrack)
 
 flowtop-confs =	tcp.conf \
 		udp.conf \
