--- a/evhtp.c
+++ b/evhtp.c
@@ -2863,7 +2863,7 @@ htp__accept_cb_(struct evconnlistener *
 }     /* htp__accept_cb_ */
 
 #ifndef EVHTP_DISABLE_SSL
-#ifndef EVHTP_DISABLE_EVTHR
+#if !defined(EVHTP_DISABLE_EVTHR) && (OPENSSL_VERSION_NUMBER < 0x10100000L)
 
 #ifndef WIN32
 #define _HTP_tid (unsigned long)pthread_self()
@@ -4575,7 +4575,7 @@ htp__use_threads_(evhtp_t * htp,
     htp->thread_init_cb = init_cb;
     htp->thread_exit_cb = exit_cb;
 
-#ifndef EVHTP_DISABLE_SSL
+#if !defined(EVHTP_DISABLE_EVTHR) && (OPENSSL_VERSION_NUMBER < 0x10100000L)
     evhtp_ssl_use_threads();
 #endif
 
@@ -4723,7 +4723,7 @@ evhtp_set_post_accept_cb(evhtp_t * htp,
 }
 
 #ifndef EVHTP_DISABLE_SSL
-#ifndef EVHTP_DISABLE_EVTHR
+#if !defined(EVHTP_DISABLE_EVTHR) && (OPENSSL_VERSION_NUMBER < 0x10100000L)
 int
 evhtp_ssl_use_threads(void)
 {
