From 3d9af59ddb472c814fb99ae1fab85030560d2fc6 Mon Sep 17 00:00:00 2001
From: W_Y_CPP <383152993@qq.com>
Date: Sun, 24 Apr 2022 05:50:46 -0400
Subject: [PATCH] remove check icp_sal_versions

---
 configure.ac | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/configure.ac b/configure.ac
index d6e43df74..2d981ddcc 100644
--- a/configure.ac
+++ b/configure.ac
@@ -252,9 +252,6 @@ AC_SUBST(enable_qat_auto_engine_init_on_fork)
 #Default library name is qatengine
 AC_SUBST([LIBQATNAME], "qatengine")
 
-AC_CHECK_FILE(${qat_hw_dir_prefix}/include/qat/icp_sal_versions.h,
-	      [with_icp_sal_versions_h=yes],
-	      [with_icp_sal_versions_h=no])
 if test "x$with_icp_sal_versions_h" = "xyes" -a "x$enable_qat_sw" = "x"
 then
   if test `grep "define SAL_INFO2_DRIVER_SW_VERSION_TYPE \"in-tree\"" ${qat_hw_dir_prefix}/include/qat/icp_sal_versions.h | wc -l` = "1"
-- 
2.17.1

