--- a/.tools/configure.ac
+++ b/.tools/configure.ac
@@ -533,7 +533,7 @@ then
   AC_SUBST(QAT_SHARED_LIB_DEPS_DRIVER)
 fi
 
-if test "x$enable_qat_contig_mem" != "xyes" -a "x$enable_multi_thread" != "xyes" -a "$enable_qat16_driver" = "xyes"
+if test "x$enable_qat_contig_mem" != "xyes" -a "x$enable_multi_thread" != "xyes"
 then
   AC_ARG_WITH(QAT_SHARED_LIB_DEPS_QAE_MEM, AS_HELP_STRING(), , QAT_SHARED_LIB_DEPS_QAE_MEM=["-Wl,-rpath,\$(with_usdm_dir) -L\$(with_usdm_dir) -lusdm_drv_s"])
   AC_SUBST(QAT_SHARED_LIB_DEPS_QAE_MEM)
