Index: openssl-qat-1/e_qat.c
===================================================================
--- openssl-qat-1.orig/e_qat.c
+++ openssl-qat-1/e_qat.c
@@ -1327,8 +1327,10 @@ static int bind_qat(ENGINE *e, const cha
      * configuration as environment variables.
      */
 
+#if defined(__GLIBC__)
 #if __GLIBC_PREREQ(2, 17)
     config_section = secure_getenv("QAT_SECTION_NAME");
+#endif
 #else
     config_section = getenv("QAT_SECTION_NAME");
 #endif
