From 677096b562111a87f52345346ee9c1b877fcdcd9 Mon Sep 17 00:00:00 2001
From: tiagofreire-pt <41837236+tiagofreire-pt@users.noreply.github.com>
Date: Sat, 29 Jan 2022 10:01:36 +0000
Subject: [PATCH] Patching lan78xx for SOF_TIMESTAMPING_TX_SOFTWARE
 support

---
 drivers/net/usb/lan78xx.c | 1 +
 1 file changed, 1 insertion(+)

--- a/drivers/net/usb/lan78xx.c
+++ b/drivers/net/usb/lan78xx.c
@@ -1788,6 +1788,7 @@ static const struct ethtool_ops lan78xx_
 	.set_link_ksettings = lan78xx_set_link_ksettings,
 	.get_regs_len	= lan78xx_get_regs_len,
 	.get_regs	= lan78xx_get_regs,
+	.get_ts_info    = ethtool_op_get_ts_info,
 };
 
 static void lan78xx_init_mac_address(struct lan78xx_net *dev)
