From 1a61ef7d787f940b8ff93f054a332b4a47a512c6 Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Thu, 3 Sep 2020 14:02:41 +0100
Subject: [PATCH] dwc_otg: whitelist_table is now productlist_table

---
 drivers/usb/core/otg_productlist.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/usb/core/otg_productlist.h
+++ b/drivers/usb/core/otg_productlist.h
@@ -139,7 +139,7 @@ static int is_targeted(struct usb_device
 		/* NOTE: can't use usb_match_id() since interface caches
 		 * aren't set up yet. this is cut/paste from that code.
 		 */
-		for (id = whitelist_table; id->match_flags; id++) {
+		for (id = productlist_table; id->match_flags; id++) {
 #ifdef DEBUG
 			dev_dbg(&dev->dev,
 				"ID: V:%04x P:%04x DC:%04x SC:%04x PR:%04x \n",
