From 2320a8d1a847d31ffa678ad61970c138dd626113 Mon Sep 17 00:00:00 2001
From: Dom Cobley <popcornmix@gmail.com>
Date: Wed, 17 Mar 2021 12:45:53 +0000
Subject: [PATCH] hack: fixup bcm2835-unicam

---
 drivers/media/platform/bcm2835/bcm2835-unicam.c | 3 +++
 1 file changed, 3 insertions(+)

--- a/drivers/media/platform/bcm2835/bcm2835-unicam.c
+++ b/drivers/media/platform/bcm2835/bcm2835-unicam.c
@@ -72,6 +72,9 @@
 #include <media/v4l2-fwnode.h>
 #include <media/videobuf2-dma-contig.h>
 
+#include <media/v4l2-async.h>
+#define v4l2_async_notifier_add_subdev __v4l2_async_notifier_add_subdev
+
 #include "vc4-regs-unicam.h"
 
 #define UNICAM_MODULE_NAME	"unicam"
