--- a/Makefile
+++ b/Makefile
@@ -72,7 +72,7 @@ endif
 CPPFLAGS += $(DEFS) $(INCLUDES)
 
 ##############################################################################
-all: Makefile.depend whois mkpasswd pos
+all: Makefile.depend whois
 
 ##############################################################################
 %.o: %.c
@@ -121,7 +121,7 @@ afl-run:
 	nice afl-fuzz -i ../afl_in -o ../afl_out -- ./whois
 
 ##############################################################################
-install: install-whois install-mkpasswd install-pos
+install: install-whois
 
 install-whois: whois
 	$(INSTALL) -d $(BASEDIR)$(prefix)/bin/
