--- a/Makefile
+++ b/Makefile
@@ -42,14 +42,6 @@ ifdef HAVE_LIBIDN
 $(error Please fix your build system to stop defining HAVE_LIBIDN!)
 endif
 
-ifeq ($(shell $(PKG_CONFIG) --exists 'libidn2 >= 2.0.3' || echo NO),)
-whois_LDADD += $(shell $(PKG_CONFIG) --libs libidn2)
-DEFS += -DHAVE_LIBIDN2 $(shell $(PKG_CONFIG) --cflags libidn2)
-else ifeq ($(shell $(PKG_CONFIG) --exists 'libidn' || echo NO),)
-whois_LDADD += $(shell $(PKG_CONFIG) --libs libidn)
-DEFS += -DHAVE_LIBIDN $(shell $(PKG_CONFIG) --cflags libidn)
-endif
-
 ifdef HAVE_ICONV
 whois_OBJECTS += simple_recode.o
 DEFS += -DHAVE_ICONV
