--- a/oor/data-plane/tun/tun.c
+++ b/oor/data-plane/tun/tun.c
@@ -53,6 +53,8 @@ int tun_rm_fwd_from_entry(lisp_addr_t *e
 tun_dplane_data_t * tun_dplane_data_new_init(oor_encap_t encap_type, glist_t *allowed_dst_eids);
 void tun_dplane_data_free(tun_dplane_data_t *data);
 
+int tun_receive_fd;
+int tun_ifindex;
 
 data_plane_struct_t dplane_tun = {
         .datap_init = tun_configure_data_plane,
--- a/oor/data-plane/tun/tun.h
+++ b/oor/data-plane/tun/tun.h
@@ -60,9 +60,9 @@
 
 /* Tun MN variables */
 
-int tun_receive_fd;
-int tun_ifindex;
-uint8_t *tun_receive_buf;
+extern int tun_receive_fd;
+extern int tun_ifindex;
+extern uint8_t *tun_receive_buf;
 
 lisp_addr_t * tun_get_default_output_address(int afi);
 int tun_get_default_output_socket(int);
