--- a/lib/fts.in.h
+++ b/lib/fts.in.h
@@ -291,20 +291,20 @@ typedef struct _ftsent {
 __BEGIN_DECLS
 
  _GL_ATTRIBUTE_NODISCARD
-FTSENT  *fts_children (FTS *, int) __THROW;
+FTSENT  *_GL_FUNCDECL_SYS_NAME (fts_children) (FTS *, int) __THROW;
 
 _GL_ATTRIBUTE_NODISCARD
-int      fts_close (FTS *) __THROW;
+int      _GL_FUNCDECL_SYS_NAME (fts_close) (FTS *) __THROW;
 
 _GL_ATTRIBUTE_NODISCARD
-FTS     *fts_open (char * const *, int,
+FTS     *_GL_FUNCDECL_SYS_NAME (fts_open) (char * const *, int,
                    int (*)(const FTSENT **, const FTSENT **))
   _GL_ATTRIBUTE_DEALLOC (fts_close, 1) __THROW;
 
 _GL_ATTRIBUTE_NODISCARD
-FTSENT  *fts_read (FTS *) __THROW;
+FTSENT  *_GL_FUNCDECL_SYS_NAME (fts_read) (FTS *) __THROW;
 
-int      fts_set (FTS *, FTSENT *, int) __THROW;
+int      _GL_FUNCDECL_SYS_NAME (fts_set) (FTS *, FTSENT *, int) __THROW;
 
 #if GNULIB_FTS_DEBUG
 extern bool fts_debug;
