--- a/lib/fts.in.h
+++ b/lib/fts.in.h
@@ -299,7 +299,8 @@ int      _GL_FUNCDECL_SYS_NAME (fts_clos
 _GL_ATTRIBUTE_NODISCARD
 FTS     *_GL_FUNCDECL_SYS_NAME (fts_open) (char * const *, int,
                    int (*)(const FTSENT **, const FTSENT **))
-  _GL_ATTRIBUTE_DEALLOC (fts_close, 1) __THROW;
+  __THROW
+  _GL_ATTRIBUTE_DEALLOC (fts_close, 1);
 
 _GL_ATTRIBUTE_NODISCARD
 FTSENT  *_GL_FUNCDECL_SYS_NAME (fts_read) (FTS *) __THROW;
