--- a/config_f.h
+++ b/config_f.h
@@ -138,11 +138,8 @@
  *		This can be much slower and no memory statistics will be
  *		provided.
  */
-#if defined(__MACHTEN__) || defined(PURIFY) || defined(MALLOC_TRACE) || defined(_OSD_POSIX) || defined(__MVS__) || defined (__CYGWIN__) || defined(__GLIBC__) || defined(__OpenBSD__) || defined(__APPLE__) || defined (__ANDROID__)
+
 # define SYSMALLOC
-#else
-# undef SYSMALLOC
-#endif
 
 /*
  * USE_ACCESS	Use access(2) rather than stat(2) when POSIX is defined.
