--- a/config-scripts/cups-directories.m4
+++ b/config-scripts/cups-directories.m4
@@ -102,7 +102,7 @@ if test "$libdir" = "\${exec_prefix}/lib"; then
 	case "$host_os_name" in
 		linux*)
 			if test -d /usr/lib64 -a ! -d /usr/lib64/fakeroot; then
-				libdir="$exec_prefix/lib64"
+				libdir="$exec_prefix/lib"
 			fi
 			;;
 	esac
--- "a/configure"
+++ "b/configure"
@@ -6188,7 +6188,7 @@ if test "$libdir" = "\${exec_prefix}/lib"; then
 	case "$host_os_name" in
 		linux*)
 			if test -d /usr/lib64 -a ! -d /usr/lib64/fakeroot; then
-				libdir="$exec_prefix/lib64"
+				libdir="$exec_prefix/lib"
 			fi
 			;;
 	esac
--- a/cups-config.in
+++ b/cups-config.in
@@ -48,7 +48,7 @@ else
 	CFLAGS="$CFLAGS -I$includedir"
     fi
 
-    if test $libdir != /usr/lib -a $libdir != /usr/lib32 -a $libdir != /usr/lib64; then
+    if test $libdir != /usr/lib -a $libdir != /usr/lib -a $libdir != /usr/lib; then
 	LDFLAGS="$LDFLAGS -L$libdir"
     fi
 fi
