Index: shellinabox-2015-12-06/configure.ac
===================================================================
--- shellinabox-2015-12-06.orig/configure.ac	2016-04-27 15:24:07.000000000 +0800
+++ shellinabox-2015-12-06/configure.ac	2016-04-27 16:50:10.792530502 +0800
@@ -75,11 +75,6 @@
             [AC_DEFINE(HAVE_ISNAN, 1,
                        Define to 1 if you have support for isnan)])
 
-dnl Even if utmpx.h exists, not all systems have support for updwtmpx()
-AC_TRY_LINK([#include <utmp.h>],
-            [updwtmp(0, 0);],
-            [AC_DEFINE(HAVE_UPDWTMP, 1,
-                       Define to 1 if you have support for updwtmp)])
 AC_TRY_LINK([#include <utmpx.h>],
             [updwtmpx(0, 0);],
             [AC_DEFINE(HAVE_UPDWTMPX, 1,
