diff --git a/shellinabox/launcher.c b/shellinabox/launcher.c
index ba54b36..c802e00 100644
--- a/shellinabox/launcher.c
+++ b/shellinabox/launcher.c
@@ -769,7 +769,7 @@ void closeAllFds(int *exceptFds, int num) {
 }
 
 #if !defined(HAVE_PTSNAME_R)
-static int ptsname_r(int fd, char *buf, size_t buflen) {
+int ptsname_r(int fd, char *buf, size_t buflen) {
   // It is unfortunate that ptsname_r is not universally available.
   // For the time being, this is not a big problem, as ShellInABox is
   // single-threaded (and so is the launcher process). But if this