--- a/shellinabox/launcher.c    2016-08-01 16:15:25.018260726 +0200
+++ b/shellinabox/launcher.c    2016-08-01 16:16:50.271765498 +0200
@@ -81,6 +81,10 @@
 #include <sys/uio.h>
 #endif

+#ifndef TTYDEF_IFLAG
+#include <sys/ttydefaults.h>
+#endif
+
 #ifdef HAVE_UTIL_H
 #include <util.h>
 #endif
@@ -90,9 +94,12 @@
 #endif

 #ifdef HAVE_UTMPX_H
-#include <utmpx.h>
+//#include <utmpx.h>
+#undef HAVE_UTMPX_H
 #endif

+#define HAVE_PTSNAME_R 1
+
 #if defined(HAVE_SECURITY_PAM_APPL_H)
 #include <security/pam_appl.h>
