--- a/tools/include/linux/compiler.h	2020-04-29 22:27:48.001385200 +0800
+++ b/tools/include/linux/compiler.h	2020-04-29 22:29:30.511340600 +0800
@@ -41,6 +41,10 @@
 
 #define ACCESS_ONCE(x) (*(volatile typeof(x) *)&(x))
 
+#ifdef __cplusplus
+#define READ_ONCE(x) ACCESS_ONCE(x)
+#define WRITE_ONCE(x, val) ACCESS_ONCE(x) = val
+#else
 #include <linux/types.h>
 
 /*
@@ -115,6 +119,7 @@ static __always_inline void __write_once
 #define WRITE_ONCE(x, val) \
 	({ union { typeof(x) __val; char __c[1]; } __u = { .__val = (val) }; __write_once_size(&(x), __u.__c, sizeof(x)); __u.__val; })
 
+#endif /* __cplusplus */
 
 #ifndef __fallthrough
 # if defined(__GNUC__) && __GNUC__ >= 7
