--- a/setup.py
+++ b/setup.py
@@ -31,7 +31,7 @@ setup(
         "Programming Language :: Python :: 3.5",
         "Programming Language :: Python :: 3.6",
     ],
-    packages=find_packages("src", exclude=("exampleproj",)),
+    packages=find_packages("src", exclude=("exampleproj", "*.tests", "*.tests.*")),
     package_dir={"": "src"},
     extras_require={"scripts": ["click>=6.0", "twisted>=16.4.0"]},
     license="MIT",
