--- a/tools/env/fw_env.c
+++ b/tools/env/fw_env.c
@@ -246,7 +246,7 @@ int fw_printenv (int argc, char *argv[])
 	int i, n_flag;
 	int rc = 0;
 
-	if (argc >= 2 && strcmp(argv[1], "-a") == 0) {
+	if (0 && argc >= 2 && strcmp(argv[1], "-a") == 0) {
 		if (argc < 3) {
 			fprintf(stderr,
 				"## Error: '-a' option requires AES key\n");
@@ -325,7 +325,7 @@ int fw_printenv (int argc, char *argv[])
 int fw_env_close(void)
 {
 	int ret;
-	if (aes_flag) {
+	if (0 && aes_flag) {
 		ret = env_aes_cbc_crypt(environment.data, 1);
 		if (ret) {
 			fprintf(stderr,
@@ -1223,7 +1223,7 @@ int fw_env_open(void)
 
 	crc0 = crc32 (0, (uint8_t *) environment.data, ENV_SIZE);
 
-	if (aes_flag) {
+	if (0 && aes_flag) {
 		ret = env_aes_cbc_crypt(environment.data, 0);
 		if (ret)
 			return ret;
@@ -1280,7 +1280,7 @@ int fw_env_open(void)
 
 		crc1 = crc32 (0, (uint8_t *) redundant->data, ENV_SIZE);
 
-		if (aes_flag) {
+		if (0 && aes_flag) {
 			ret = env_aes_cbc_crypt(redundant->data, 0);
 			if (ret)
 				return ret;
