--- a/arch/arm/boot/dts/ox820-cloudengines-pogoplug-series-3.dts
+++ b/arch/arm/boot/dts/ox820-cloudengines-pogoplug-series-3.dts
@@ -74,11 +74,27 @@
 		nand-ecc-algo = "hamming";
 
 		partition@0 {
-			label = "boot";
-			reg = <0x00000000 0x00e00000>;
+			label = "stage1";
+			reg = <0x00000000 0x00040000>;
 			read-only;
 		};
 
+		partition@40000 {
+			label = "u-boot";
+			reg = <0x00040000 0x00380000>;
+			read-only;
+		};
+
+		partition@3c0000 {
+			label = "u-boot-env";
+			reg = <0x003c0000 0x00080000>;
+		};
+
+		partition@440000 {
+			label = "kernel";
+			reg = <0x00440000 0x009c0000>;
+		};
+
 		partition@e00000 {
 			label = "ubi";
 			reg = <0x00e00000 0x07200000>;
