--- a/arch/arm/boot/dts/ox820-cloudengines-pogoplug-series-3.dts
+++ b/arch/arm/boot/dts/ox820-cloudengines-pogoplug-series-3.dts
@@ -28,24 +28,28 @@
 		serial0 = &uart0;
 		gpio0 = &gpio0;
 		gpio1 = &gpio1;
+		led-boot = &led_status;
+		led-failsafe = &led_warn;
+		led-running = &led_act;
+		led-upgrade = &led_warn;
 	};
 
 	leds {
 		compatible = "gpio-leds";
 
-		blue {
+		led_status: blue {
 			label = "pogoplug:blue";
 			gpios = <&gpio0 2 0>;
 			default-state = "keep";
 		};
 
-		orange {
+		led_warn: orange {
 			label = "pogoplug:orange";
 			gpios = <&gpio1 16 1>;
 			default-state = "keep";
 		};
 
-		green {
+		led_act: green {
 			label = "pogoplug:green";
 			gpios = <&gpio1 17 1>;
 			default-state = "keep";
