--- a/arch/arm/boot/dts/ox820-cloudengines-pogoplug-series-3.dts
+++ b/arch/arm/boot/dts/ox820-cloudengines-pogoplug-series-3.dts
@@ -12,7 +12,7 @@
 / {
 	model = "Cloud Engines PogoPlug Series 3";
 
-	compatible = "cloudengines,pogoplugv3", "oxsemi,ox820";
+	compatible = "cloudengines,pogoplug-series-3", "cloudengines,pogoplugv3", "oxsemi,ox820";
 
 	chosen {
 		bootargs = "earlyprintk console=ttyS0,115200";
