From 13556aaf94a2b0738a835a29f76be41d775a2d33 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Sun, 3 Jul 2016 15:51:43 +0200
Subject: [PATCH] Use -lftdi1

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 src/Makefile.am | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -1,6 +1,6 @@
 bin_PROGRAMS = \
   relayctl
 
-relayctl_LDFLAGS = -lftdi
+relayctl_LDFLAGS = -lftdi1
 
 relayctl_SOURCES = relayctl.c
