--- a/gst/gstplugin.c
+++ b/gst/gstplugin.c
@@ -790,15 +790,8 @@ _priv_gst_plugin_load_file_for_registry
   }
 #endif
 
-  flags = G_MODULE_BIND_LOCAL;
-  /* libgstpython.so is the gst-python plugin loader. It needs to be loaded with
-   * G_MODULE_BIND_LAZY.
-   *
-   * Ideally there should be a generic way for plugins to specify that they
-   * need to be loaded with _LAZY.
-   * */
-  if (strstr (filename, "libgstpython"))
-    flags |= G_MODULE_BIND_LAZY;
+  // No need to resolve all bindings until referenced
+  flags = G_MODULE_BIND_LOCAL | G_MODULE_BIND_LAZY;
 
   module = g_module_open (filename, flags);
   if (module == NULL) {
