--- a/arch/mips/ar7/memory.c
+++ b/arch/mips/ar7/memory.c
@@ -66,5 +66,11 @@ void __init prom_meminit(void)
 
 void __init prom_free_prom_memory(void)
 {
-	/* Nothing to free */
+	/* adapted from arch/mips/txx9/generic/setup.c */
+	unsigned long saddr = PHYS_OFFSET + PAGE_SIZE;
+	unsigned long eaddr = __pa_symbol(&_text);
+
+	/* free memory between prom-record and kernel _text base */
+	if (saddr < eaddr)
+		free_init_pages("prom memory", saddr, eaddr);
 }
