--- a/drivers/mtd/Kconfig
+++ b/drivers/mtd/Kconfig
@@ -154,6 +154,11 @@ config MTD_OF_PARTS
 	  the partition map from the children of the flash node,
 	  as described in Documentation/devicetree/bindings/mtd/partition.txt.
 
+config MTD_AC49X_PARTS
+	tristate "AudioCodes AC49X partitioning support"
+	---help---
+	  AudioCodes AC49X partitioning support
+
 config MTD_AR7_PARTS
 	tristate "TI AR7 partitioning support"
 	---help---
--- a/drivers/mtd/Makefile
+++ b/drivers/mtd/Makefile
@@ -11,6 +11,7 @@ obj-$(CONFIG_MTD_SPLIT)		+= mtdsplit/
 obj-$(CONFIG_MTD_OF_PARTS)	+= ofpart.o
 obj-$(CONFIG_MTD_REDBOOT_PARTS) += redboot.o
 obj-$(CONFIG_MTD_CMDLINE_PARTS) += cmdlinepart.o
+obj-$(CONFIG_MTD_AC49X_PARTS)	+= ac49xpart.o
 obj-$(CONFIG_MTD_AFS_PARTS)	+= afs.o
 obj-$(CONFIG_MTD_AR7_PARTS)	+= ar7part.o titanpart.o
 obj-$(CONFIG_MTD_BCM63XX_PARTS)	+= bcm63xxpart.o
--- a/arch/mips/ar7/platform.c
+++ b/arch/mips/ar7/platform.c
@@ -201,7 +201,7 @@ static struct resource physmap_flash_res
 	.end	= 0x11ffffff,
 };
 
-static const char *ar7_probe_types[] = { "ar7part", NULL };
+static const char *ar7_probe_types[] = { "ac49xpart", "ar7part", NULL };
 
 static struct physmap_flash_data physmap_flash_data = {
 	.width	= 2,
