--- a/interface/vmcs_host/CMakeLists.txt
+++ b/interface/vmcs_host/CMakeLists.txt
@@ -24,7 +24,7 @@ if(NOT ARM64)
 
 	# ILCS pulls in EGL for the ILCS/EGL surface API support
 	target_link_libraries(vcilcs brcmEGL brcmGLESv2 khrn_client vchiq_arm vcos)
-	set(INSTALL_TARGETS "${INSTALL_TARGETS} vcilcs")
+	set(INSTALL_TARGETS ${INSTALL_TARGETS} vcilcs)
 endif()
 
 # vchostif needs ilcore as well (vmcs_main pulls it in)
