diff --git a/xt_FULLCONENAT.c b/xt_FULLCONENAT.c
index 9e52eba..8658c5f 100644
--- a/xt_FULLCONENAT.c
+++ b/xt_FULLCONENAT.c
@@ -697,9 +697,11 @@ static struct xt_target tg_reg[] __read_mostly = {
 
 static int __init fullconenat_tg_init(void)
 {
+  printk(KERN_INFO "xt_FULLCONENAT: RFC3489 Full Cone NAT module\n"
+	"xt_FULLCONENAT: Copyright (C) 2018 Chion Tang <tech@chionlab.moe>\n");
   wq = create_singlethread_workqueue("xt_FULLCONENAT");
   if (wq == NULL) {
-    printk("xt_FULLCONENAT: warning: failed to create workqueue\n");
+    printk(KERN_WARNING "xt_FULLCONENAT: warning: failed to create workqueue\n");
   }
 
   return xt_register_targets(tg_reg, ARRAY_SIZE(tg_reg));
