From ea9e2477624adaa40e8a553ef876f60ec8d3150c Mon Sep 17 00:00:00 2001
From: W_Y_CPP <383152993@qq.com>
Date: Fri, 18 Feb 2022 00:53:12 -0500
Subject: [PATCH] refresh

---
 xt_FULLCONENAT.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/xt_FULLCONENAT.c b/xt_FULLCONENAT.c
index f96cfd8a3..237666039 100644
--- a/xt_FULLCONENAT.c
+++ b/xt_FULLCONENAT.c
@@ -1258,8 +1258,7 @@ static int fullconenat_tg_check(const struct xt_tgchk_param *par)
 #endif
 
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 15, 0) && !defined(CONFIG_NF_CONNTRACK_CHAIN_EVENTS)
-    if (!READ_ONCE(par->net->ct.nf_conntrack_event_cb)) {
-      nf_conntrack_register_notifier(par->net, &ct_event_notifier);
+    if (!READ_ONCE(par->net->ct.nf_conntrack_event_cb)&&(nf_conntrack_register_notifier(par->net, &ct_event_notifier)==0)) {
 #else
     if (nf_conntrack_register_notifier(par->net, &ct_event_notifier) == 0) {
 #endif
-- 
2.17.1
