--- a/src/display/canvas.h
+++ b/src/display/canvas.h
@@ -48,7 +48,7 @@ class Canvas {
 public:
   typedef std::vector<Attributes> attributes_list;
 
-  Canvas(int x = 0, int y = 0, int width = 0, int height = 0);
+  Canvas(int x = 0, int y = 0, int width = 1, int height = 1);
   ~Canvas() { if (!m_isDaemon) { delwin(m_window); } }
 
   void                refresh()                                               { if (!m_isDaemon) { wnoutrefresh(m_window); } }
