--- a/setup.py
+++ b/setup.py
@@ -2,14 +2,14 @@ from __future__ import unicode_literals
 
 from setuptools import find_packages, setup
 
-import picklefield
+import os
 
 with open('README.rst') as file_:
     long_description = file_.read()
 
 setup(
     name='django-picklefield',
-    version=picklefield.__version__,
+    version=os.getenv('PKG_VERSION'),
     description='Pickled object field for Django',
     long_description=long_description,
     author='Simon Charette',
