--- a/src/libs/zbxcommon/str.c
+++ b/src/libs/zbxcommon/str.c
@@ -54,7 +54,7 @@ static const char	help_message_footer[]
 void	version(void)
 {
 	printf("%s (Zabbix) %s\n", title_message, ZABBIX_VERSION);
-	printf("Revision %s %s, compilation time: %s %s\n\n", ZABBIX_REVISION, ZABBIX_REVDATE, __DATE__, __TIME__);
+	printf("Revision %s %s\n\n", ZABBIX_REVISION, ZABBIX_REVDATE);
 	puts(copyright_message);
 #if defined(HAVE_GNUTLS) || defined(HAVE_OPENSSL)
 	printf("\n");
