--- a/tn7atm.c
+++ b/tn7atm.c
@@ -73,8 +73,8 @@
 #include <linux/ctype.h>
 
 #include <asm/io.h>
-#include <asm/ar7/ar7.h>
-#include <asm/ar7/prom.h>
+#include <asm/mach-ar7/ar7.h>
+#include <asm/mach-ar7/prom.h>
 
 #include "dsl_hal_api.h"
 #include "tn7atm.h"
--- a/tn7dsl.c
+++ b/tn7dsl.c
@@ -111,8 +111,8 @@
 #include <linux/firmware.h>
 
 #include <asm/io.h>
-#include <asm/ar7/ar7.h>
-#include <asm/ar7/prom.h>
+#include <asm/mach-ar7/ar7.h>
+#include <asm/mach-ar7/prom.h>
 
 /* Modules specific header files */
 #include "tn7atm.h"
--- a/tn7sar.c
+++ b/tn7sar.c
@@ -54,8 +54,8 @@
 #include <linux/ctype.h>
 
 #include <asm/io.h>
-#include <asm/ar7/ar7.h>
-#include <asm/ar7/prom.h>
+#include <asm/mach-ar7/ar7.h>
+#include <asm/mach-ar7/prom.h>
 
 #define _CPHAL_AAL5
 #define _CPHAL_SAR
