--- a/tn7atm.c
+++ b/tn7atm.c
@@ -71,10 +71,16 @@
 #include <linux/proc_fs.h>
 #include <linux/string.h>
 #include <linux/ctype.h>
+#include <linux/version.h>
 
 #include <asm/io.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,31)
 #include <asm/ar7/ar7.h>
 #include <asm/ar7/prom.h>
+#else
+#include <asm/mach-ar7/ar7.h>
+#include <asm/mach-ar7/prom.h>
+#endif
 
 #include "dsl_hal_api.h"
 #include "tn7atm.h"
--- a/tn7dsl.c
+++ b/tn7dsl.c
@@ -112,8 +112,13 @@
 #include <linux/version.h>
 
 #include <asm/io.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,31)
 #include <asm/ar7/ar7.h>
 #include <asm/ar7/prom.h>
+#else
+#include <asm/mach-ar7/ar7.h>
+#include <asm/mach-ar7/prom.h>
+#endif
 
 /* Modules specific header files */
 #include "tn7atm.h"
--- a/tn7sar.c
+++ b/tn7sar.c
@@ -52,10 +52,16 @@
 #include <linux/proc_fs.h>
 #include <linux/string.h>
 #include <linux/ctype.h>
+#include <linux/version.h>
 
 #include <asm/io.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,31)
 #include <asm/ar7/ar7.h>
 #include <asm/ar7/prom.h>
+#else
+#include <asm/mach-ar7/ar7.h>
+#include <asm/mach-ar7/prom.h>
+#endif
 
 #define _CPHAL_AAL5
 #define _CPHAL_SAR
