--- a/cp_sar_reg.h
+++ b/cp_sar_reg.h
@@ -214,4 +214,4 @@
 
 /* END OF FILE */
 
-#endif _INC_SAR_REG
+#endif
--- a/tn7api.h
+++ b/tn7api.h
@@ -172,4 +172,4 @@ void tn7sar_get_sar_firmware_version(uns
 extern struct file_operations tn7sar_proc_oam_ping_fops;
 extern struct file_operations tn7sar_proc_pvc_table_fops;
 int tn7sar_tx_flush(void *privContext, int chan, int queue, int skip);
-#endif __SGAPI_H
+#endif
--- a/tn7atm.h
+++ b/tn7atm.h
@@ -276,4 +276,4 @@ typedef struct
 #define PHYS_TO_K1(X)                             (PHYS_ADDR(X)|K1BASE)
 #endif
 
-#endif __TN7ATM_H
+#endif
--- a/dsl_hal_api.h
+++ b/dsl_hal_api.h
@@ -2448,7 +2448,7 @@ unsigned int dslhal_api_getHLINpsds(tids
 *
 ********************************************************************************************/
 
-unsigned int dslhal_api_getHighPrecision();
+unsigned int dslhal_api_getHighPrecision(void);
 
 /********************************************************************************************
 * FUNCTION NAME: void dslhal_api_setHighPrecision
@@ -2459,7 +2459,7 @@ unsigned int dslhal_api_getHighPrecision
 * Return: None
 ********************************************************************************************/
 
-void dslhal_api_setHighPrecision();
+void dslhal_api_setHighPrecision(void);
 //  UR8_MERGE_END   CQ11054*
 
 #ifdef INTERNAL_BUILD
