--- a/tn7dsl.c
+++ b/tn7dsl.c
@@ -346,7 +346,7 @@ static void tn7dsl_chng_modulation(void*
 static unsigned int tn7dsl_set_modulation(void* data, int flag);
 static void tn7dsl_ctrl_fineGain(int value);
 static void tn7dsl_set_fineGainValue(int value);
-static int dslmod_sysctl (ctl_table * ctl, int write, struct file *filp,
+static int dslmod_sysctl (struct ctl_table * ctl, int write, struct file *filp,
                           void *buffer, size_t * lenp);
 static void tn7dsl_register_dslss_led(void);
 void tn7dsl_dslmod_sysctl_register(void);
@@ -3325,7 +3325,7 @@ unsigned int tn7dsl_get_memory(unsigned
 
 
 
-static int dslmod_sysctl(ctl_table *ctl, int write, struct file * filp,
+static int dslmod_sysctl(struct ctl_table *ctl, int write, struct file * filp,
       void *buffer, size_t *lenp)
 {
   char *ptr;
@@ -3451,7 +3451,7 @@ static int dslmod_sysctl(ctl_table *ctl,
 }
 
 
-ctl_table dslmod_table[] = {
+struct ctl_table dslmod_table[] = {
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,33)
   {DEV_DSLMOD, "dslmod", info, DSL_MOD_SIZE, 0644, NULL, NULL, &dslmod_sysctl, &sysctl_string}
 #else
@@ -3469,7 +3469,7 @@ ctl_table dslmod_table[] = {
   };
 
 /* Make sure that /proc/sys/dev is there */
-ctl_table dslmod_root_table[] = {
+struct ctl_table dslmod_root_table[] = {
 #ifdef CONFIG_PROC_FS
 	#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,33)
   {CTL_DEV, "dev", NULL, 0, 0555, dslmod_table}
