--- a/tn7dsl.c
+++ b/tn7dsl.c
@@ -333,7 +333,7 @@ static void tn7dsl_chng_modulation(void*
 static unsigned int tn7dsl_set_modulation(void* data, int flag);
 static void tn7dsl_ctrl_fineGain(int value);
 static void tn7dsl_set_fineGainValue(int value);
-static int dslmod_sysctl (ctl_table * ctl, int write, void *buffer,
+static int dslmod_sysctl (struct ctl_table * ctl, int write, void *buffer,
                           size_t * lenp, loff_t *ppos);
 static void tn7dsl_register_dslss_led(void);
 void tn7dsl_dslmod_sysctl_register(void);
@@ -3308,7 +3308,7 @@ unsigned int tn7dsl_get_memory(unsigned
 
 
 
-static int dslmod_sysctl(ctl_table *ctl, int write, void *buffer,
+static int dslmod_sysctl(struct ctl_table *ctl, int write, void *buffer,
       size_t *lenp, loff_t *ppos)
 {
   char *ptr;
@@ -3423,7 +3423,7 @@ static int dslmod_sysctl(ctl_table *ctl,
 }
 
 
-ctl_table dslmod_table[] = {
+struct ctl_table dslmod_table[] = {
 	{
 		.procname 		= "dslmod",
 		.data			= info,
@@ -3437,7 +3437,7 @@ ctl_table dslmod_table[] = {
   };
 
 /* Make sure that /proc/sys/dev is there */
-ctl_table dslmod_root_table[] = {
+struct ctl_table dslmod_root_table[] = {
 #ifdef CONFIG_PROC_FS
 	{
 		.procname		= "dev",
