--- a/Makefile.am
+++ b/Makefile.am
@@ -18,7 +18,7 @@ AUTOMAKE_OPTIONS = gnu
 # old Automake version.
 ACLOCAL_AMFLAGS = -I m4
 
-SUBDIRS = doc src tests tune
+SUBDIRS = src
 
 nobase_dist_doc_DATA = AUTHORS BUGS COPYING COPYING.LESSER NEWS TODO \
   examples/ReadMe examples/divworst.c examples/rndo-add.c examples/sample.c \
--- a/Makefile.in
+++ b/Makefile.in
@@ -378,7 +378,7 @@ AUTOMAKE_OPTIONS = gnu
 # libtoolize and in case some developer needs to switch back to an
 # old Automake version.
 ACLOCAL_AMFLAGS = -I m4
-SUBDIRS = doc src tests tune
+SUBDIRS = src
 nobase_dist_doc_DATA = AUTHORS BUGS COPYING COPYING.LESSER NEWS TODO \
   examples/ReadMe examples/divworst.c examples/rndo-add.c examples/sample.c \
   examples/version.c
