--- a/Makefile.am
+++ b/Makefile.am
@@ -24,7 +24,7 @@ AUTOMAKE_OPTIONS = gnu
 # old Automake version.
 ACLOCAL_AMFLAGS = -I m4
 
-SUBDIRS = doc src tests tune tools/bench
+SUBDIRS = src
 
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = mpfr.pc
--- a/Makefile.in
+++ b/Makefile.in
@@ -384,7 +384,7 @@ AUTOMAKE_OPTIONS = gnu
 # libtoolize and in case some developer needs to switch back to an
 # old Automake version.
 ACLOCAL_AMFLAGS = -I m4
-SUBDIRS = doc src tests tune tools/bench
+SUBDIRS = src
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = mpfr.pc
 nobase_dist_doc_DATA = AUTHORS BUGS COPYING COPYING.LESSER NEWS TODO \
