--- a/arch/arm/boot/dts/armada-388-clearfog.dts
+++ b/arch/arm/boot/dts/armada-388-clearfog.dts
@@ -430,6 +430,18 @@
 		};
 	};
 
+	mvsw61xx {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		compatible = "marvell,88e6176";
+		status = "okay";
+		reg = <0x4>;
+		is-indirect;
+
+		mii-bus = <&mdio>;
+		cpu-port-0 = <5>;
+	};
+
 	gpio-keys {
 		compatible = "gpio-keys";
 		pinctrl-0 = <&rear_button_pins>;
