Index: qatzip-1/utils/qzip.c
===================================================================
--- qatzip-1.orig/utils/qzip.c
+++ qatzip-1/utils/qzip.c
@@ -776,7 +776,9 @@ int main(int argc, char **argv)
             printf("For help, type: qzip -h\n");
         } else {
             stream_out = stdout;
+#ifdef __GLIBC___
             stdout = freopen(NULL, "w", stdout);
+#endif
             processStream(&g_sess, stdin, stream_out, g_decompress == 0);
         }
     } else {
