--- a/drivers/pci/controller/dwc/pcie-dw-rockchip.c
+++ b/drivers/pci/controller/dwc/pcie-dw-rockchip.c
@@ -521,9 +521,13 @@ static int rockchip_pcie_probe(struct pl
 	if (ret)
 		return ret;
 
-	ret = reset_control_assert(rockchip->rst);
-	if (ret)
-		return ret;
+	if (of_machine_is_compatible("rockchip,rk3528")) {
+		dev_warn(dev, "RockChip refuses to admit it's a bug\n");
+	} else {
+		ret = reset_control_assert(rockchip->rst);
+		if (ret)
+			return ret;
+	}
 
 	/* DON'T MOVE ME: must be enable before PHY init */
 	rockchip->vpcie3v3 = devm_regulator_get_optional(dev, "vpcie3v3");
