--- a/arch/arm64/boot/dts/rockchip/rk3328-nanopi-r2s.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3328-nanopi-r2s.dts
@@ -9,6 +9,7 @@
 #include <dt-bindings/leds/common.h>
 #include <dt-bindings/gpio/gpio.h>
 #include "rk3328.dtsi"
+#include "rk3328-dram-dmc.dtsi"
 
 / {
 	model = "FriendlyElec NanoPi R2S";
@@ -142,6 +143,10 @@
 	cpu-supply = <&vdd_arm>;
 };
 
+&dfi {
+	status = "okay";
+};
+
 &display_subsystem {
 	status = "disabled";
 };
--- a/arch/arm64/boot/dts/rockchip/rk3328-orangepi-r1-plus.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3328-orangepi-r1-plus.dts
@@ -9,6 +9,7 @@
 #include <dt-bindings/gpio/gpio.h>
 #include <dt-bindings/leds/common.h>
 #include "rk3328.dtsi"
+#include "rk3328-dram-dmc.dtsi"
 
 / {
 	model = "Xunlong Orange Pi R1 Plus";
@@ -108,6 +109,10 @@
 	cpu-supply = <&vdd_arm>;
 };
 
+&dfi {
+	status = "okay";
+};
+
 &display_subsystem {
 	status = "disabled";
 };
--- a/arch/arm64/boot/dts/rockchip/rk3328-orangepi-r1-plus-lts.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3328-orangepi-r1-plus-lts.dts
@@ -14,6 +14,13 @@
 	compatible = "xunlong,orangepi-r1-plus-lts", "rockchip,rk3328";
 };
 
+&dmc_opp_table {
+	/delete-node/ opp-798000000;
+	/delete-node/ opp-840000000;
+	/delete-node/ opp-924000000;
+	/delete-node/ opp-1056000000;
+};
+
 &gmac2io {
 	phy-handle = <&yt8531c>;
 	tx_delay = <0x19>;
