--- a/drivers/phy/rockchip/phy-rockchip-naneng-combphy.c	2025-03-06 21:30:53.981108971 +0900
+++ b/drivers/phy/rockchip/phy-rockchip-naneng-combphy.c	2025-03-06 21:30:06.752107940 +0900
@@ -138,6 +138,7 @@
 	struct combphy_reg pipe_xpcs_phy_ready;
 	struct combphy_reg pipe_pcie1l0_sel;
 	struct combphy_reg pipe_pcie1l1_sel;
+	struct combphy_reg pipe_sgmii_mac_sel;
 };
 
 struct rockchip_combphy_cfg {
@@ -290,6 +291,7 @@
 
 static int rockchip_combphy_parse_dt(struct device *dev, struct rockchip_combphy_priv *priv)
 {
+	int mac_id;
 	int i;
 
 	priv->num_clks = devm_clk_bulk_get_all(dev, &priv->clks);
@@ -325,6 +327,11 @@
 
 	priv->ext_refclk = device_property_present(dev, "rockchip,ext-refclk");
 
+	if (!device_property_read_u32(dev, "rockchip,sgmii-mac-sel", &mac_id)) {
+		rockchip_combphy_param_write(priv->pipe_grf, &priv->cfg->grfcfg->pipe_sgmii_mac_sel,
+					     (mac_id > 0) ? true : false);
+	}
+
 	priv->phy_rst = devm_reset_control_get_exclusive(dev, "phy");
 	/* fallback to old behaviour */
 	if (PTR_ERR(priv->phy_rst) == -ENOENT)
@@ -704,6 +711,7 @@
 	/* pipe-grf */
 	.pipe_con0_for_sata	= { 0x0000, 15, 0, 0x00, 0x2220 },
 	.pipe_xpcs_phy_ready	= { 0x0040, 2, 2, 0x00, 0x01 },
+	.pipe_sgmii_mac_sel	= { 0x0040, 1, 1, 0x00, 0x01 },
 };
 
 static const struct rockchip_combphy_cfg rk3568_combphy_cfgs = {
