From 7f2f7250d1f5ad8369fa4424e9091c6fffde408d Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.com>
Date: Mon, 29 Mar 2021 12:05:06 +0100
Subject: [PATCH] overlays: Add README entry for minipitft13

Signed-off-by: Phil Elwell <phil@raspberrypi.com>
---
 arch/arm/boot/dts/overlays/README | 11 +++++++++++
 1 file changed, 11 insertions(+)

--- a/arch/arm/boot/dts/overlays/README
+++ b/arch/arm/boot/dts/overlays/README
@@ -2018,6 +2018,17 @@ Load:   dtoverlay=midi-uart1
 Params: <None>
 
 
+Name:   minipitft13
+Info:   Overlay for AdaFruit Mini Pi 1.3" TFT via SPI using fbtft driver.
+Load:   dtoverlay=minipitft13,<param>=<val>
+Params: speed                   SPI bus speed (default 32000000)
+        rotate                  Display rotation (0, 90, 180 or 270; default 0)
+        width                   Display width (default 240)
+        height                  Display height (default 240)
+        fps                     Delay between frame updates (default 25)
+        debug                   Debug output level (0-7; default 0)
+
+
 Name:   miniuart-bt
 Info:   Switch the onboard Bluetooth function on Pi 3B, 3B+, 3A+, 4B and Zero W
         to use the mini-UART (ttyS0) and restore UART0/ttyAMA0 over GPIOs 14 &
