--- a/Makefile.am
+++ b/Makefile.am
@@ -14,10 +14,7 @@ PACKAGE_VERSION = @PACKAGE_VERSION@
 
 ACLOCAL_AMFLAGS = -I m4
 SUBDIRS = \
-	include \
-	man-po \
-	po \
-	testsuite
+	include
 
 AM_CFLAGS = -Iproc
 LDADD = ./proc/libprocps.la $(CYGWINFLAGS)
--- a/configure.ac
+++ b/configure.ac
@@ -298,8 +298,5 @@ AC_CHECK_FUNCS([__fpending alarm atexit
 
 AC_CONFIG_FILES([Makefile
                  include/Makefile
-                 man-po/Makefile
-                 po/Makefile.in
-                 proc/libprocps.pc
-                 testsuite/Makefile])
+                 proc/libprocps.pc])
 AC_OUTPUT
