From 70cef63e0c0ec15016cf7c52d5183ed864cc50c4 Mon Sep 17 00:00:00 2001
From: Toni Uhlig <matzeton@googlemail.com>
Date: Sun, 20 Sep 2020 13:58:05 +0200
Subject: [PATCH] Disable non IANA TLS extensions.

Signed-off-by: Toni Uhlig <matzeton@googlemail.com>
---
 openssl_clienthello.c | 2 ++
 1 file changed, 2 insertions(+)

--- a/openssl_clienthello.c
+++ b/openssl_clienthello.c
@@ -93,7 +93,9 @@ static struct lookup_table_element_t kno
 	ELEMENT(TLSEXT_TYPE_certificate_authorities),
 #endif
 	ELEMENT(TLSEXT_TYPE_renegotiate),
+#ifndef OPENSSL_NO_NEXTPROTONEG
 	ELEMENT(TLSEXT_TYPE_next_proto_neg),
+#endif
 	{ 0 }
 };
 
