--- a/super.c
+++ b/super.c
@@ -219,7 +219,7 @@ enum Opt {
 	Opt_showmeta,
 	Opt_acl,
 	Opt_noatime,
-	Opt_nls,
+	Opt_iocharset,
 	Opt_prealloc,
 	Opt_no_acs_rules,
 	Opt_err,
@@ -239,7 +239,7 @@ static const match_table_t ntfs_tokens = {
 	{ Opt_acl, "acl" },
 	{ Opt_noatime, "noatime" },
 	{ Opt_showmeta, "showmeta" },
-	{ Opt_nls, "nls=%s" },
+	{ Opt_iocharset, "iocharset=%s" },
 	{ Opt_prealloc, "prealloc" },
 	{ Opt_no_acs_rules, "no_acs_rules" },
 	{ Opt_err, NULL },
@@ -342,7 +342,7 @@ static noinline int ntfs_parse_options(struct super_block *sb, char *options,
 		case Opt_showmeta:
 			opts->showmeta = 1;
 			break;
-		case Opt_nls:
+		case Opt_iocharset:
 			match_strlcpy(nls_name, &args[0], sizeof(nls_name));
 			break;
 		case Opt_prealloc:
@@ -581,9 +581,9 @@ static int ntfs_show_options(struct seq_file *m, struct dentry *root)
 	if (opts->dmask)
 		seq_printf(m, ",dmask=%04o", ~opts->fs_dmask_inv);
 	if (opts->nls)
-		seq_printf(m, ",nls=%s", opts->nls->charset);
+		seq_printf(m, ",iocharset=%s", opts->nls->charset);
 	else
-		seq_puts(m, ",nls=utf8");
+		seq_puts(m, ",iocharset=utf8");
 	if (opts->sys_immutable)
 		seq_puts(m, ",sys_immutable");
 	if (opts->discard)
