--- a/arch/arm/mach-kirkwood/Kconfig
+++ b/arch/arm/mach-kirkwood/Kconfig
@@ -74,4 +74,7 @@ source "board/Seagate/goflexhome/Kconfig
 source "board/Seagate/nas220/Kconfig"
 source "board/zyxel/nsa310s/Kconfig"
 
+config SECOND_STAGE
+	bool "OpenWrt second stage hack"
+
 endif
--- a/include/configs/dockstar.h
+++ b/include/configs/dockstar.h
@@ -98,4 +98,6 @@
 #define CONFIG_CMD_MTDPARTS
 #define CONFIG_LZO
 
+#include "openwrt-kirkwood-common.h"
+
 #endif /* _CONFIG_DOCKSTAR_H */
--- a/include/configs/ib62x0.h
+++ b/include/configs/ib62x0.h
@@ -108,4 +108,6 @@
 #define CONFIG_RTC_MV
 #endif /* CONFIG_CMD_DATE */
 
+#include "openwrt-kirkwood-common.h"
+
 #endif /* _CONFIG_IB62x0_H */
--- a/include/configs/iconnect.h
+++ b/include/configs/iconnect.h
@@ -103,4 +103,6 @@
 #define CONFIG_MTD_PARTITIONS
 #define CONFIG_CMD_MTDPARTS
 
+#include "openwrt-kirkwood-common.h"
+
 #endif /* _CONFIG_ICONNECT_H */
--- /dev/null
+++ b/include/configs/openwrt-kirkwood-common.h
@@ -0,0 +1,40 @@
+/*
+ * Copyright (C) 2013 Luka Perkov <luka@openwrt.org>
+ *
+ * SPDX-License-Identifier:	GPL-2.0+
+ */
+
+#ifndef __OPENWRT_KIRKWOOD_COMMON_H
+#define __OPENWRT_KIRKWOOD_COMMON_H
+
+/* Commands */
+#define CONFIG_CMD_BOOTZ
+
+/* Ethernet */
+#if defined(CONFIG_CMD_NET)
+#define CONFIG_SERVERIP		192.168.1.2
+#define CONFIG_IPADDR		192.168.1.1
+#endif
+
+/* second stage loader */
+#if defined(CONFIG_SECOND_STAGE)
+#undef CONFIG_ENV_IS_IN_NAND
+#undef CONFIG_ENV_SECT_SIZE
+#define CONFIG_ENV_IS_NOWHERE
+#endif
+
+/* Flattened uImage Tree */
+#define CONFIG_FIT			1
+#define CONFIG_FIT_VERBOSE		1
+
+/* Various */
+#define CONFIG_BZIP2
+#define CONFIG_LZMA
+#define CONFIG_LZO
+
+/* Unnecessary */
+#undef CONFIG_BOOTM_NETBSD
+#undef CONFIG_BOOTM_PLAN9
+#undef CONFIG_BOOTM_RTEMS
+
+#endif /* __OPENWRT_KIRKWOOD_COMMON_H */
--- a/include/configs/pogo_e02.h
+++ b/include/configs/pogo_e02.h
@@ -99,4 +99,6 @@
 #define CONFIG_CMD_MTDPARTS
 #define CONFIG_LZO
 
+#include "openwrt-kirkwood-common.h"
+
 #endif /* _CONFIG_POGO_E02_H */
--- a/include/configs/sheevaplug.h
+++ b/include/configs/sheevaplug.h
@@ -109,4 +109,6 @@
 #define CONFIG_SYS_ATA_IDE1_OFFSET	MV_SATA_PORT1_OFFSET
 #endif /* CONFIG_CMD_IDE */
 
+#include "openwrt-kirkwood-common.h"
+
 #endif /* _CONFIG_SHEEVAPLUG_H */
--- a/include/configs/goflexhome.h
+++ b/include/configs/goflexhome.h
@@ -122,4 +122,6 @@
 #define CONFIG_RTC_MV
 #endif /* CONFIG_CMD_DATE */
 
+#include "openwrt-kirkwood-common.h"
+
 #endif /* _CONFIG_GOFLEXHOME_H */
