--- a/hack/make.sh
+++ b/hack/make.sh
@@ -95,7 +95,7 @@ fi
 # with a newer libdevmapper than the one it was built with.
 if
 	command -v gcc &> /dev/null \
-		&& ! (echo -e '#include <libdevmapper.h>\nint main() { dm_task_deferred_remove(NULL); }' | gcc -xc - -o /dev/null $(pkg-config --libs devmapper) &> /dev/null) \
+		&& ! (echo -e '#include <libdevmapper.h>\nint main() { dm_task_deferred_remove(NULL); }' | gcc -xc - -o /dev/null $(pkg-config --libs libdevmapper) &> /dev/null) \
 		;
 then
 	add_buildtag libdm dlsym_deferred_remove
