diff --git a/drivers/soc/qcom/socinfo.c b/drivers/soc/qcom/socinfo.c
index e0620416e574..2a7ceef338fb 100644
--- a/drivers/soc/qcom/socinfo.c
+++ b/drivers/soc/qcom/socinfo.c
@@ -227,7 +227,11 @@ static const struct soc_id soc_id[] = {
 	{ 341, "SDA845" },
 	{ 356, "SM8250" },
 	{ 402, "IPQ6018" },
+	{ 403, "IPQ6028" },
+	{ 421, "IPQ6000" },
+	{ 422, "IPQ6010" },
 	{ 425, "SC7180" },
+	{ 453, "IPQ6005" },
 };
 
 static const char *socinfo_machine(struct device *dev, unsigned int id)
