 diff --git a/mm/memblock.c b/mm/memblock.c
index b5ea98bfd..b0fecc325 100644
--- a/mm/memblock.c
+++ b/mm/memblock.c
@@ -578,6 +578,7 @@ static int __init_memblock memblock_add_
 	/* special case for empty array */
 	if (type->regions[0].size == 0) {
 		WARN_ON(type->cnt != 1 || type->total_size);
+		type->start_base = base;
 		type->regions[0].base = base;
 		type->regions[0].size = size;
 		type->regions[0].flags = flags;
