--- a/configure.ac
+++ b/configure.ac
@@ -533,11 +533,7 @@ if test "x$ac_system" = "xLinux"; then
     [have_cpuid_h="no (cpuid.h not found)"]
   )
 
-  AC_CHECK_HEADERS([sys/capability.h],
-    [have_capability="yes"],
-    [have_capability="no (<sys/capability.h> not found)"]
-  )
-
+  have_capability="no"
   if test "x$have_capability" = "xyes"; then
     AC_CHECK_LIB([cap], [cap_get_proc],
       [have_capability="yes"],
