--- a/src/types.db
+++ b/src/types.db
@@ -359,6 +359,17 @@ vs_memory               value:GAUGE:0:92
 vs_processes            value:GAUGE:0:65535
 vs_threads              value:GAUGE:0:65535
 
+qdisc_bytes             value:DERIVE:0:U
+qdisc_drops             value:DERIVE:0:U
+qdisc_backlog           value:GAUGE:0:U
+
+qdisct_bytes            value:DERIVE:0:U
+qdisct_thres            value:GAUGE:0:U
+qdisct_drops            drops:DERIVE:0:U, ecn:DERIVE:0:U, ack:DERIVE:0:U
+qdisct_backlog          value:GAUGE:0:U
+qdisct_latencyus	tg:GAUGE:0:U, pk:GAUGE:0:U, av:GAUGE:0:U, sp:GAUGE:0:U
+qdisct_flows		sp:GAUGE:0:U, bu:GAUGE:0:U, un:GAUGE:0:U
+
 #
 # Legacy types
 # (required for the v5 upgrade target)
