diff -Nuar a/arch/arm/boot/dts/bcm5301x.dtsi b/arch/arm/boot/dts/bcm5301x.dtsi
--- a/arch/arm/boot/dts/bcm5301x.dtsi	2021-12-29 16:55:29.441523015 +0800
+++ b/arch/arm/boot/dts/bcm5301x.dtsi	2021-12-29 17:03:44.640694605 +0800
@@ -38,8 +38,6 @@
 			reg = <0x0400 0x100>;
 			interrupts = <GIC_SPI 85 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&iprocslow>;
-			pinctrl-names = "default";
-			pinctrl-0 = <&pinmux_uart1>;
 			status = "disabled";
 		};
 	};
