Index: eudev-3.2.9/src/udev/udevadm.c
===================================================================
--- eudev-3.2.9.orig/src/udev/udevadm.c
+++ eudev-3.2.9/src/udev/udevadm.c
@@ -45,13 +45,7 @@ static const struct udevadm_cmd udevadm_
 
 static const struct udevadm_cmd *udevadm_cmds[] = {
         &udevadm_info,
-        &udevadm_trigger,
-        &udevadm_settle,
-        &udevadm_control,
         &udevadm_monitor,
-        &udevadm_hwdb,
-        &udevadm_test,
-        &udevadm_test_builtin,
         &udevadm_version,
         &udevadm_help,
 };
