--- a/arch/arm/mach-gemini/board-nas4220b.c
+++ b/arch/arm/mach-gemini/board-nas4220b.c
@@ -94,6 +94,7 @@ static void __init ib4220b_init(void)
 	platform_device_register(&ib4220b_led_device);
 	platform_device_register(&ib4220b_key_device);
 	platform_register_rtc();
+	platform_register_watchdog();
 }
 
 MACHINE_START(NAS4220B, "Raidsonic NAS IB-4220-B")
--- a/arch/arm/mach-gemini/board-wbd111.c
+++ b/arch/arm/mach-gemini/board-wbd111.c
@@ -121,6 +121,7 @@ static void __init wbd111_init(void)
 	platform_device_register(&wbd111_leds_device);
 	platform_device_register(&wbd111_keys_device);
 	platform_register_rtc();
+	platform_register_watchdog();
 }
 
 MACHINE_START(WBD111, "Wiliboard WBD-111")
--- a/arch/arm/mach-gemini/board-wbd222.c
+++ b/arch/arm/mach-gemini/board-wbd222.c
@@ -121,6 +121,7 @@ static void __init wbd222_init(void)
 	platform_device_register(&wbd222_leds_device);
 	platform_device_register(&wbd222_keys_device);
 	platform_register_rtc();
+	platform_register_watchdog();
 }
 
 MACHINE_START(WBD222, "Wiliboard WBD-222")
--- a/arch/arm/mach-gemini/board-rut1xx.c
+++ b/arch/arm/mach-gemini/board-rut1xx.c
@@ -80,6 +80,7 @@ static void __init rut1xx_init(void)
 	platform_device_register(&rut1xx_leds);
 	platform_device_register(&rut1xx_keys_device);
 	platform_register_rtc();
+	platform_register_watchdog();
 }
 
 MACHINE_START(RUT100, "Teltonika RUT100")
