From 8f68331e14dff9a101f2d0e1d6bec84a031f27ee Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Thu, 9 Mar 2017 11:03:18 +0100
Subject: [PATCH 69/69] arm: boot: add dts files

Signed-off-by: John Crispin <john@phrozen.org>
---
 arch/arm/boot/dts/Makefile | 8 ++++++++
 1 file changed, 8 insertions(+)

--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -785,11 +785,29 @@ dtb-$(CONFIG_ARCH_QCOM) += \
 	qcom-apq8074-dragonboard.dtb \
 	qcom-apq8084-ifc6540.dtb \
 	qcom-apq8084-mtp.dtb \
+	qcom-ipq4018-a42.dtb \
+	qcom-ipq4018-ea6350v3.dtb \
+	qcom-ipq4018-eap1300.dtb \
+	qcom-ipq4018-ex6100v2.dtb \
+	qcom-ipq4018-ex6150v2.dtb \
+	qcom-ipq4018-fritz4040.dtb \
+	qcom-ipq4018-jalapeno.dtb \
+	qcom-ipq4018-nbg6617.dtb \
+	qcom-ipq4018-rt-ac58u.dtb \
+	qcom-ipq4018-wre6606.dtb \
 	qcom-ipq4019-ap.dk01.1-c1.dtb \
 	qcom-ipq4019-ap.dk04.1-c1.dtb \
 	qcom-ipq4019-ap.dk04.1-c3.dtb \
 	qcom-ipq4019-ap.dk07.1-c1.dtb \
 	qcom-ipq4019-ap.dk07.1-c2.dtb \
+	qcom-ipq4019-a62.dtb \
+	qcom-ipq4019-map-ac2200.dtb \
+	qcom-ipq4019-r619ac.dtb \
+	qcom-ipq4019-hugo-ac1200.dtb \
+	qcom-ipq4019-rt-acrh17.dtb \
+	qcom-ipq4028-wpj428.dtb \
+	qcom-ipq4029-gl-b1300.dtb \
+	qcom-ipq4029-mr33.dtb \
 	qcom-ipq8064-ap148.dtb \
 	qcom-msm8660-surf.dtb \
 	qcom-msm8960-cdp.dtb \
