--- a/setup.py
+++ b/setup.py
@@ -2,6 +2,7 @@
 Setup file for automat
 """
 
+import os
 from setuptools import setup, find_packages
 
 try:
@@ -14,7 +15,7 @@ except(IOError, ImportError):
 
 setup(
     name='Automat',
-    use_scm_version=True,
+    version=os.getenv('PKG_VERSION'),
     url='https://github.com/glyph/Automat',
     description="""
     Self-service finite-state machines for the programmer on the go.
@@ -22,10 +23,6 @@ setup(
     long_description=long_description,
     packages=find_packages(exclude=[]),
     package_dir={'automat': 'automat'},
-    setup_requires=[
-        'setuptools-scm',
-        'm2r',
-    ],
     install_requires=[
         "attrs>=19.2.0",
         "six",
