--- a/setup.py
+++ b/setup.py
@@ -27,15 +27,6 @@ setup(
         "attrs>=19.2.0",
         "six",
     ],
-    extras_require={
-        "visualize": ["graphviz>0.5.1",
-                      "Twisted>=16.1.1"],
-    },
-    entry_points={
-        "console_scripts": [
-            "automat-visualize = automat._visualize:tool"
-        ],
-    },
     author='Glyph',
     author_email='glyph@twistedmatrix.com',
     include_package_data=True,
