--- a/setup.py
+++ b/setup.py
@@ -21,7 +21,7 @@ setup(
     Self-service finite-state machines for the programmer on the go.
     """.strip(),
     long_description=long_description,
-    packages=find_packages(exclude=[]),
+    packages=find_packages(exclude=["*._test", "*._test.*"]),
     package_dir={'automat': 'automat'},
     install_requires=[
         "attrs>=19.2.0",
@@ -30,6 +30,7 @@ setup(
     author='Glyph',
     author_email='glyph@twistedmatrix.com',
     include_package_data=True,
+    exclude_package_data={'':['_test/*']},
     license="MIT",
     keywords='fsm finite state machine automata',
     classifiers=[
