diff -Naur a/include/common.h b/include/common.h
--- a/include/common.h	2019-08-03 10:48:42.000000000 -0700
+++ b/include/common.h	2020-03-01 19:26:14.000000000 -0800
@@ -2,6 +2,7 @@
 #define __COMMON_H__
 
 #include <stdint.h>
+#include <endian.h>
 
 #ifdef DEBUG
 #include <stdio.h>
@@ -10,7 +11,7 @@
 #define DPRINTF(...)
 #endif
 
-#ifdef __BIG_ENDIAN__
+#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
 
 static inline uint16_t BE16(uint16_t x)
 {
