--- /dev/null
+++ b/arch/arm/dts/rk3568-fastrhino-r66s-u-boot.dtsi
@@ -0,0 +1,25 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * (C) Copyright 2021 Rockchip Electronics Co., Ltd
+ */
+
+#include "rk3568-u-boot.dtsi"
+
+/ {
+	chosen {
+		stdout-path = &uart2;
+		u-boot,spl-boot-order = "same-as-spl", &sdhci, &sdmmc0;
+	};
+};
+
+&sdmmc0 {
+	bus-width = <4>;
+	u-boot,dm-spl;
+	u-boot,spl-fifo-mode;
+};
+
+&uart2 {
+	clock-frequency = <24000000>;
+	u-boot,dm-spl;
+	status = "okay";
+};
--- /dev/null
+++ b/arch/arm/dts/rk3568-fastrhino-r66s.dts
@@ -0,0 +1,9 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+
+/dts-v1/;
+#include "rk3568-evb.dts"
+
+/ {
+	model = "FastRhino R66S";
+	compatible = "fastrhino,r66s", "rockchip,rk3568";
+};
--- /dev/null
+++ b/board/fastrhino/fastrhino-r66s-rk3568/Kconfig
@@ -0,0 +1,15 @@
+if TARGET_FASTRHINO_R66S_RK3568
+
+config SYS_BOARD
+	default "fastrhino-r66s-rk3568"
+
+config SYS_VENDOR
+	default "fastrhino"
+
+config SYS_CONFIG_NAME
+	default "fastrhino-r66s-rk3568"
+
+config BOARD_SPECIFIC_OPTIONS # dummy
+	def_bool y
+
+endif
--- /dev/null
+++ b/board/fastrhino/fastrhino-r66s-rk3568/Makefile
@@ -0,0 +1,4 @@
+# SPDX-License-Identifier:     GPL-2.0+
+#
+
+obj-y	+= fastrhino-r66s-rk3568.o
--- /dev/null
+++ b/board/fastrhino/fastrhino-r66s-rk3568/fastrhino-r66s-rk3568.c
@@ -0,0 +1,4 @@
+ // SPDX-License-Identifier: GPL-2.0+
+/*
+ *
+ */
--- /dev/null
+++ b/include/configs/fastrhino-r66s-rk3568.h
@@ -0,0 +1,17 @@
+/* SPDX-License-Identifier: GPL-2.0+ */
+
+#ifndef __FASTRHINO_R66S_RK3568_H
+#define __FASTRHINO_R66S_RK3568_H
+
+#include <configs/rk3568_common.h>
+
+#define CONFIG_SUPPORT_EMMC_RPMB
+
+#define ROCKCHIP_DEVICE_SETTINGS \
+			"stdout=serial,vidconsole\0" \
+			"stderr=serial,vidconsole\0"
+
+#define CONFIG_USB_OHCI_NEW
+#define CONFIG_SYS_USB_OHCI_MAX_ROOT_PORTS     2
+
+#endif
