From d585719ad3f737ca5e5186fe55be964340ecfcc7 Mon Sep 17 00:00:00 2001
From: Marty Jones <mj8263788@gmail.com>
Date: Mon, 4 Oct 2021 16:46:32 -0400
Subject: [PATCH] tools:disable TOOLS_LIBCRYPTO

Signed-off-by: Marty Jones <mj8263788@gmail.com>
---
 tools/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/tools/Kconfig
+++ b/tools/Kconfig
@@ -16,7 +16,7 @@ config TOOLS_CRC32
 
 config TOOLS_LIBCRYPTO
 	bool "Use OpenSSL's libcrypto library for host tools"
-	default y
+	default n
 	help
 	  Cryptographic signature, verification, and encryption of images is
 	  provided by host tools using OpenSSL's libcrypto. Select 'n' here if
