--- a/tools/binman/control.py
+++ b/tools/binman/control.py
@@ -9,8 +9,9 @@ from collections import OrderedDict
 import glob
 try:
     import importlib.resources as importlib_resources
-except ImportError:  # pragma: no cover
-    # for Python 3.6
+    # for Python 3.6, 3.7 and 3.8
+    importlib_resources.files
+except (ImportError, AttributeError):
     import importlib_resources
 import os
 import re
