From d0689253c9e2eb78fc5296adb109aba4d35a13fd Mon Sep 17 00:00:00 2001
From: Chris Leech <cleech@redhat.com>
Date: Mon, 19 Nov 2012 17:09:24 -0800
Subject: [PATCH] remove the offload boot supported ifdef

---
 usr/iface.c | 7 +------
 1 file changed, 1 insertion(+), 6 deletions(-)

--- a/usr/iface.c
+++ b/usr/iface.c
@@ -998,6 +998,7 @@ int iface_setup_from_boot_context(struct
 {
 	struct iscsi_transport *t = NULL;
 	uint32_t hostno;
+	int rc;
 
 	if (strlen(context->initiatorname))
 		strlcpy(iface->iname, context->initiatorname,
@@ -1011,10 +1012,7 @@ int iface_setup_from_boot_context(struct
 			return 0;
 		}
 	} else if (strlen(context->iface)) {
-/* this ifdef is only temp until distros and firmwares are updated */
-#ifdef OFFLOAD_BOOT_SUPPORTED
 		char transport_name[ISCSI_TRANSPORT_NAME_MAXLEN];
-		int rc;
 
 		memset(transport_name, 0, ISCSI_TRANSPORT_NAME_MAXLEN);
 		/* make sure offload driver is loaded */
@@ -1040,9 +1038,6 @@ int iface_setup_from_boot_context(struct
 		}
 
 		strlcpy(iface->netdev, context->iface, sizeof(iface->netdev));
-#else
-		return 0;
-#endif
 	} else
 		return 0;
 
