--- a/src/icons/icons.qrc
+++ b/src/icons/icons.qrc
@@ -250,38 +250,16 @@
         <file>flags/zm.svg</file>
         <file>flags/zw.svg</file>
         <file>L.gif</file>
-        <file>loading.png</file>
         <file>qbt-theme/application-exit.svg</file>
-        <file>qbt-theme/application-rss+xml.svg</file>
-        <file>qbt-theme/application-x-mswinurl.svg</file>
         <file>qbt-theme/checked.svg</file>
         <file>qbt-theme/configure.svg</file>
-        <file>qbt-theme/dialog-cancel.svg</file>
-        <file>qbt-theme/dialog-information.svg</file>
-        <file>qbt-theme/dialog-warning.svg</file>
         <file>qbt-theme/document-edit-verify.svg</file>
         <file>qbt-theme/document-edit.svg</file>
-        <file>qbt-theme/document-encrypt.svg</file>
-        <file>qbt-theme/document-import.svg</file>
-        <file>qbt-theme/document-new.svg</file>
-        <file>qbt-theme/document-properties.svg</file>
-        <file>qbt-theme/document-save.svg</file>
-        <file>qbt-theme/download.svg</file>
-        <file>qbt-theme/edit-clear-history.svg</file>
         <file>qbt-theme/edit-clear.svg</file>
         <file>qbt-theme/edit-copy.svg</file>
-        <file>qbt-theme/edit-cut.svg</file>
         <file>qbt-theme/edit-delete.svg</file>
-        <file>qbt-theme/edit-find-user.svg</file>
         <file>qbt-theme/edit-find.svg</file>
-        <file>qbt-theme/edit-paste.svg</file>
         <file>qbt-theme/edit-rename.svg</file>
-        <file>qbt-theme/folder-documents.svg</file>
-        <file>qbt-theme/folder-download.svg</file>
-        <file>qbt-theme/folder-new.svg</file>
-        <file>qbt-theme/folder-remote.svg</file>
-        <file>qbt-theme/gear.svg</file>
-        <file>qbt-theme/gear32.svg</file>
         <file>qbt-theme/go-bottom.svg</file>
         <file>qbt-theme/go-down.svg</file>
         <file>qbt-theme/go-top.svg</file>
@@ -295,43 +273,15 @@
         <file>qbt-theme/kt-set-max-upload-speed.png</file>
         <file>qbt-theme/list-add.svg</file>
         <file>qbt-theme/list-remove.svg</file>
-        <file>qbt-theme/mail-folder-inbox.svg</file>
-        <file>qbt-theme/mail-mark-read.svg</file>
         <file>qbt-theme/media-playback-pause.svg</file>
         <file>qbt-theme/media-playback-start.svg</file>
         <file>qbt-theme/media-seek-forward.svg</file>
-        <file>qbt-theme/network-server.svg</file>
-        <file>qbt-theme/network-wired.svg</file>
-        <file>qbt-theme/object-locked.svg</file>
-        <file>qbt-theme/office-chart-line.svg</file>
-        <file>qbt-theme/preferences-desktop.svg</file>
-        <file>qbt-theme/preferences-other.svg</file>
-        <file>qbt-theme/preferences-system-network.svg</file>
-        <file>qbt-theme/preferences-web-browser-cookies.svg</file>
-        <file>qbt-theme/rss-config.png</file>
         <file>qbt-theme/security-high.svg</file>
         <file>qbt-theme/security-low.svg</file>
-        <file>qbt-theme/services.svg</file>
-        <file>qbt-theme/speedometer.svg</file>
         <file>qbt-theme/system-log-out.svg</file>
-        <file>qbt-theme/tab-close.svg</file>
-        <file>qbt-theme/task-attention.svg</file>
-        <file>qbt-theme/task-complete.png</file>
-        <file>qbt-theme/task-ongoing.png</file>
-        <file>qbt-theme/task-reject.png</file>
-        <file>qbt-theme/text-plain.svg</file>
-        <file>qbt-theme/tools-report-bug.svg</file>
-        <file>qbt-theme/unavailable.svg</file>
-        <file>qbt-theme/user-group-delete.svg</file>
-        <file>qbt-theme/user-group-new.svg</file>
-        <file>qbt-theme/view-calendar-journal.svg</file>
         <file>qbt-theme/view-categories.svg</file>
-        <file>qbt-theme/view-filter.svg</file>
-        <file>qbt-theme/view-preview.svg</file>
-        <file>qbt-theme/view-refresh.svg</file>
         <file>qbt-theme/view-statistics.svg</file>
         <file>qbt-theme/wallet-open.svg</file>
-        <file>qbt-theme/webui.svg</file>
         <file>skin/arrow-right.gif</file>
         <file>skin/bg-dropdown.gif</file>
         <file>skin/bg-handle-horizontal.gif</file>
@@ -342,7 +292,6 @@
         <file>skin/completed.svg</file>
         <file>skin/connected.svg</file>
         <file>skin/disconnected.svg</file>
-        <file>skin/dock-tabs.gif</file>
         <file>skin/download.svg</file>
         <file>skin/downloading.svg</file>
         <file>skin/error.svg</file>
@@ -353,14 +302,11 @@
         <file>skin/handle-icon-horizontal.gif</file>
         <file>skin/handle-icon.gif</file>
         <file>skin/knob.gif</file>
-        <file>skin/logo-blank.gif</file>
         <file>skin/logo.gif</file>
         <file>skin/logo2.gif</file>
         <file>skin/mascot.png</file>
         <file>skin/paused.svg</file>
         <file>skin/qbittorrent-tray.svg</file>
-        <file>skin/qbittorrent-tray-dark.svg</file>
-        <file>skin/qbittorrent-tray-light.svg</file>
         <file>skin/qbittorrent32.png</file>
         <file>skin/queued.svg</file>
         <file>skin/ratio.svg</file>
@@ -370,17 +316,12 @@
         <file>skin/spacer.gif</file>
         <file>skin/spinner-placeholder.gif</file>
         <file>skin/spinner.gif</file>
-        <file>skin/splash.png</file>
         <file>skin/stalledDL.svg</file>
         <file>skin/stalledUP.svg</file>
         <file>skin/tabs.gif</file>
         <file>skin/toolbox-divider.gif</file>
-        <file>skin/toolbox-divider2.gif</file>
         <file>skin/uploading.svg</file>
         <file>slow.png</file>
         <file>slow_off.png</file>
-        <file>sphere.png</file>
-        <file>sphere2.png</file>
-        <file>url.png</file>
     </qresource>
 </RCC>
