--- a/src/ccache.cpp
+++ b/src/ccache.cpp
@@ -1633,6 +1633,7 @@ calculate_result_and_manifest_key(Contex
                              "CPLUS_INCLUDE_PATH",
                              "OBJC_INCLUDE_PATH",
                              "OBJCPLUS_INCLUDE_PATH", // clang
+                             "GCC_HONOUR_COPTS",
                              nullptr};
     for (const char** p = envvars; *p; ++p) {
       const char* v = getenv(*p);
