--- a/source3/lib/messages.c
+++ b/source3/lib/messages.c
@@ -507,7 +507,7 @@ static NTSTATUS messaging_init_internal(
 		return NT_STATUS_ACCESS_DENIED;
 	}
 
-	priv_path = private_path("msg.sock");
+	priv_path = lock_path(talloc_tos(), "msg.sock");
 	if (priv_path == NULL) {
 		return NT_STATUS_NO_MEMORY;
 	}
@@ -670,7 +670,7 @@ NTSTATUS messaging_reinit(struct messagi
 		msg_ctx->per_process_talloc_ctx,
 		msg_ctx->event_ctx,
 		&msg_ctx->id.unique_id,
-		private_path("msg.sock"),
+		lock_path(talloc_tos(), "msg.sock"),
 		lck_path,
 		messaging_recv_cb,
 		msg_ctx,
