--- a/lib/util/wscript_configure
+++ b/lib/util/wscript_configure
@@ -11,13 +11,13 @@ conf.CHECK_FUNCS_IN('backtrace backtrace
 conf.CHECK_HEADERS('execinfo.h')
 
 conf.SET_TARGET_TYPE('LIBUNWIND', 'EMPTY')
-if conf.check_cfg(package='libunwind-generic',
-                  args='--cflags --libs',
-                  msg='Checking for libunwind',
-                  uselib_store='LIBUNWIND',
-                  mandatory=False):
-    if conf.CHECK_HEADERS('libunwind.h'):
-        conf.SET_TARGET_TYPE('LIBUNWIND', 'SYSLIB')
+# if conf.check_cfg(package='libunwind-generic',
+                  # args='--cflags --libs',
+                  # msg='Checking for libunwind',
+                  # uselib_store='LIBUNWIND',
+                  # mandatory=False):
+    # if conf.CHECK_HEADERS('libunwind.h'):
+        # conf.SET_TARGET_TYPE('LIBUNWIND', 'SYSLIB')
 
 conf.CHECK_STRUCTURE_MEMBER('struct statvfs', 'f_frsize', define='HAVE_FRSIZE', headers='sys/statvfs.h')
 
