commit e72b7bd12fdef06c3494b919376bfe886aa8bb4d
Author: Jens Muecke <jens@nons.de>
Date:   Thu Jan 26 00:35:43 2012 +0100

    100-no_aio

--- a/configure.in
+++ b/configure.in
@@ -92,8 +92,8 @@ AC_PACKAGE_GLOBALS(xfsprogs)
 AC_PACKAGE_UTILITIES(xfsprogs)
 AC_MULTILIB($enable_lib64)
 
-AC_PACKAGE_NEED_AIO_H
-AC_PACKAGE_NEED_LIO_LISTIO
+librt="-lrt"
+AC_SUBST(librt)
 
 AC_PACKAGE_NEED_UUID_H
 AC_PACKAGE_NEED_UUIDCOMPARE
