commit 7b1d0a98e779170232c0a81b4749ab934ec67a7e
Author: Jens Muecke <jens@nons.de>
Date:   Thu Jan 26 00:36:42 2012 +0100

    110-uclibc_no_ustat

--- a/libxfs/linux.c
+++ b/libxfs/linux.c
@@ -21,7 +21,6 @@
 #include <mntent.h>
 #include <sys/stat.h>
 #undef ustat
-#include <sys/ustat.h>
 #include <sys/mount.h>
 #include <sys/ioctl.h>
 #include <sys/sysinfo.h>
@@ -49,6 +48,7 @@ static int max_block_alignment;
 int
 platform_check_ismounted(char *name, char *block, struct stat64 *s, int verbose)
 {
+#if 0
 	/* Pad ust; pre-2.6.28 linux copies out too much in 32bit compat mode */
 	struct ustat	ust[2];
 	struct stat64	st;
@@ -68,6 +68,7 @@ platform_check_ismounted(char *name, cha
 				progname, name);
 		return 1;
 	}
+#endif
 	return 0;
 }
 
