commit 10d6058b24f18cb31889154f830b191849f45106
Author: Jens Muecke <jens@nons.de>
Date:   Thu Jan 26 00:38:27 2012 +0100

    130-uclibc_no_xattr

--- a/fsr/xfs_fsr.c
+++ b/fsr/xfs_fsr.c
@@ -35,7 +35,6 @@
 #include <sys/wait.h>
 #include <sys/vfs.h>
 #include <sys/statvfs.h>
-#include <sys/xattr.h>
 
 
 #ifndef XFS_XFLAG_NODEFRAG
@@ -990,6 +989,7 @@ fsr_setup_attr_fork(
 	int		tfd,
 	xfs_bstat_t	*bstatp)
 {
+#if 0
 	struct stat64	tstatbuf;
 	int		i;
 	int		last_forkoff = 0;
@@ -1108,6 +1108,7 @@ fsr_setup_attr_fork(
 out:
 	if (dflag)
 		fsrprintf(_("set temp attr\n"));
+#endif
 	return 0;
 }
 
