--- a/pdns/Makefile.am
+++ b/pdns/Makefile.am
@@ -71,7 +71,6 @@ CLEANFILES = \
 	backends/gsql/gsqlbackend.gcno \
 	backends/gsql/gsqlbackend.gcov \
 	dnsmessage.pb.cc dnsmessage.pb.h \
-	pdns.conf-dist \
 	apidocfiles.h \
 	api-swagger.yaml \
 	api-swagger.json
@@ -99,7 +98,6 @@ apidocfiles.h: api-swagger.yaml api-swag
 	./incfiles $^ > $@
 
 noinst_SCRIPTS = pdns.init
-sysconf_DATA = pdns.conf-dist
 
 sbin_PROGRAMS = pdns_server
 bin_PROGRAMS = \
@@ -139,7 +137,7 @@ if IXFRDIST
 bin_PROGRAMS += \
 	ixfrdist
 
-sysconf_DATA += \
+sysconf_DATA = \
 	ixfrdist.example.yml
 endif
 
@@ -1278,9 +1276,6 @@ dnspcap2protobuf_LDADD = \
 endif
 endif
 
-pdns.conf-dist: pdns_server
-	$(AM_V_GEN)./pdns_server --config=default 2>/dev/null > $@
-
 testrunner_SOURCES = \
 	arguments.cc \
 	auth-caches.cc auth-caches.hh \
