diff -Naur a/m4/libupnp.m4 b/m4/libupnp.m4
--- a/m4/libupnp.m4	2016-10-06 19:01:54.000000000 +0800
+++ b/m4/libupnp.m4	2016-10-11 07:36:39.198318574 +0800
@@ -52,7 +52,6 @@
 		[export PKG_CONFIG_PATH=$withval/lib/pkgconfig])
 
 dnl	Check for libupnp >= MIN_LIBUPNP_VERSION
-	AS_IF([test $cross_compiling = no], [
 		AC_MSG_CHECKING([for libupnp version >= MIN_LIBUPNP_VERSION])
 		AS_IF([test -n "$PKG_CONFIG"], [
 			AS_IF([$PKG_CONFIG libupnp --exists], [
@@ -79,11 +78,6 @@
 		])
 		AC_MSG_RESULT([$result$resultstr])
 		libupnp_error="libupnp >= MIN_LIBUPNP_VERSION not found$resultstr"
-	], [
-dnl Currently cross-compilation with libupnp is not supported.
-		result=no
-		libupnp_error="cross compiling"
-	])
 
 dnl	Execute the right action.
 	AS_IF([test ${result:-no} = yes], [$2], [$3])
