--- a/src/libs/common/MuleDebug.cpp	2016-10-06 18:49:44.000000000 +0800
+++ b/src/libs/common/MuleDebug.cpp	2021-06-17 02:16:40.000000000 +0800
@@ -268,14 +268,14 @@
 		return;
 	}
 
-	bfd_vma vma = bfd_get_section_vma(abfd, section);
+	bfd_vma vma = bfd_section_vma(section);
 
 	unsigned long address = (unsigned long)_address;
 	if (address < vma) {
 		return;
 	}
 
-	bfd_size_type size = bfd_section_size(abfd, section);
+	bfd_size_type size = bfd_section_size(section);
 	if (address > (vma + size)) {
 		return;
 	}
