--- a/bin/Makefile
+++ b/bin/Makefile
@@ -13,7 +13,7 @@ install:
 	$(INSTALL) -m 755 $(WRAPPER) $(PREFIX)$(UCLIBCXX_RUNTIME_BINDIR)
 
 $(WRAPPER): Makefile
-	echo '#!/bin/sh' > $@
+	echo '#!/usr/bin/env bash' > $@
 	echo '' >> $@
 	echo 'WRAPPER_INCLUDEDIR="$${WRAPPER_INCLUDEDIR:=-I$(UCLIBCXX_RUNTIME_INCLUDEDIR)}"' >> $@
 	echo 'WRAPPER_LIBDIR="$${WRAPPER_LIBDIR:=-L$(UCLIBCXX_RUNTIME_LIBDIR)}"' >> $@
