--- a/include/ostream
+++ b/include/ostream
@@ -294,7 +294,7 @@ namespace std {
 #endif
 #endif
 
-	template <class charT,class traits = char_traits<charT> >
+	template <class charT,class traits>
 		class _UCXXEXPORT basic_ostream<charT,traits>::sentry
 	{
 		bool ok;
--- a/include/istream
+++ b/include/istream
@@ -340,7 +340,7 @@ namespace std{
 
 	};
 
-	template <class charT,class traits = char_traits<charT> > class _UCXXEXPORT basic_istream<charT,traits>::sentry {
+	template <class charT,class traits> class _UCXXEXPORT basic_istream<charT,traits>::sentry {
 		bool ok;
 	public:
 		explicit _UCXXEXPORT sentry(basic_istream<charT,traits>& os, bool noskipws = false){
