--- a/src/del_op.cpp
+++ b/src/del_op.cpp
@@ -24,3 +24,7 @@
 _UCXXEXPORT void operator delete(void* ptr) throw(){
 	free(ptr);
 }
+
+_UCXXEXPORT void operator delete(void* ptr, size_t size) throw(){
+	free(ptr);
+}
--- a/src/del_opv.cpp
+++ b/src/del_opv.cpp
@@ -24,3 +24,7 @@
 _UCXXEXPORT void operator delete[](void * ptr) throw(){
 	free(ptr);
 }
+
+_UCXXEXPORT void operator delete[](void * ptr, size_t size) throw(){
+	free(ptr);
+}
