From: Felix Fietkau <nbd@nbd.name>
Subject: net: phy: disable soft-reset for generic PHY devices to avoid accidentally clearing preinitialized state

Signed-off-by: Felix Fietkau <nbd@nbd.name>
---
 drivers/net/phy/phy_device.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

--- a/drivers/net/phy/phy_device.c
+++ b/drivers/net/phy/phy_device.c
@@ -1524,11 +1524,6 @@ int genphy_config_init(struct phy_device
 	return 0;
 }
 
-static int gen10g_soft_reset(struct phy_device *phydev)
-{
-	/* Do nothing for now */
-	return 0;
-}
 EXPORT_SYMBOL(genphy_config_init);
 
 static int gen10g_config_init(struct phy_device *phydev)
@@ -1811,7 +1806,7 @@ static struct phy_driver genphy_driver[]
 	.phy_id         = 0xffffffff,
 	.phy_id_mask    = 0xffffffff,
 	.name           = "Generic 10G PHY",
-	.soft_reset	= gen10g_soft_reset,
+	.soft_reset	= genphy_no_soft_reset,
 	.config_init    = gen10g_config_init,
 	.features       = 0,
 	.config_aneg    = gen10g_config_aneg,
