--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -372,13 +372,13 @@ install-exec-local:
 	rm -f $(DESTDIR)/$(bindir)/`echo lsb | sed -e '$(transform)'`
 	ln $(DESTDIR)/$(bindir)/`echo lsz |sed -e '$(transform)'` \
 		$(DESTDIR)/$(bindir)/`echo lsb |sed -e '$(transform)'` 
-	rm -f $(bindir)/`echo lsx | sed -e '$(transform)'`
+	rm -f $(DESTDIR)/$(bindir)/`echo lsx | sed -e '$(transform)'`
 	ln $(DESTDIR)/$(bindir)/`echo lsz |sed -e '$(transform)'` \
 		$(DESTDIR)/$(bindir)/`echo lsx |sed -e '$(transform)'` 
-	rm -f $(bindir)/`echo lrb | sed -e '$(transform)'`
+	rm -f $(DESTDIR)/$(bindir)/`echo lrb | sed -e '$(transform)'`
 	ln $(DESTDIR)/$(bindir)/`echo lrz |sed -e '$(transform)'` \
 		$(DESTDIR)/$(bindir)/`echo lrb |sed -e '$(transform)'` 
-	rm -f $(bindir)/`echo lrx | sed -e '$(transform)'`
+	rm -f $(DESTDIR)/$(bindir)/`echo lrx | sed -e '$(transform)'`
 	ln $(DESTDIR)/$(bindir)/`echo lrz |sed -e '$(transform)'` \
 		$(DESTDIR)/$(bindir)/`echo lrx |sed -e '$(transform)'` 
 #	rm -f $(bindir)/`echo lrzszbug | sed -e '$(transform)'`
