--- a/libtoolize.in
+++ b/libtoolize.in
@@ -1066,7 +1066,7 @@ func_install_pkgaux_files ()
           func_ltmain_update "$file" "$pkgauxdir" "$aux_dir" pkgaux_header
           ;;
         *)
-          test subproject = "$ltdl_mode" || continue
+          $opt_install || test subproject = "$ltdl_mode" || continue
           func_copy "$file" "$pkgauxdir" "$aux_dir" pkgaux_header
           ;;
       esac
