diff --git a/processor/processor.go b/processor/processor.go
index a35f8df..a7e3015 100644
--- a/processor/processor.go
+++ b/processor/processor.go
@@ -393,7 +393,7 @@
 				//data["url"] = uri.Scheme + "://" + uri.Host + uri.EscapedPath()
 				//data["url"] = uri.String()
 				if *config.EndPoint {
-					data["url"] = "http://music.163.com/unblockmusic/" + uri.String()
+					data["url"] = "https://music.163.com/unblockmusic/" + uri.String()
 				} else {
 					data["url"] = uri.String()
 				}
