--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -966,6 +966,32 @@ static struct board_info __initdata boar
 };
 #endif /* CONFIG_BCM63XX_CPU_6358 */
 
+#ifdef CONFIG_BCM63XX_CPU_6362
+static struct board_info __initdata board_nb6 = {
+	.name				= "NB6",
+	.expected_cpu_id		= 0x6362,
+
+	.has_ohci0			= 1,
+	.has_ehci0			= 1,
+	.num_usbh_ports			= 2,
+
+	.has_enetsw			= 1,
+
+	.enetsw = {
+		.used_ports = {
+			[4] = {
+				.used	= 1,
+				.phy_id	= 0xff,
+				.bypass_link = 1,
+				.force_speed = 1000,
+				.force_duplex_full = 1,
+				.name	= "RGMII",
+			},
+		},
+	},
+};
+#endif /* CONFIG_BCM63XX_CPU_6362 */
+
 /*
  * known 6368 boards
  */
@@ -1115,6 +1141,10 @@ static const struct board_info __initcon
 	&board_DVAG3810BN,
 #endif
 
+#ifdef CONFIG_BCM63XX_CPU_6362
+	&board_nb6,
+#endif
+
 #ifdef CONFIG_BCM63XX_CPU_6368
 	&board_96368mvwg,
 	&board_96368mvngr,
@@ -1183,6 +1213,9 @@ static struct of_device_id const bcm963x
 	{ .compatible = "t-com,spw303v", .data = &board_spw303v, },
 	{ .compatible = "telsey,cpva642", .data = &board_CPVA642, },
 #endif
+#ifdef CONFIG_BCM63XX_CPU_6362
+	{ .compatible = "sfr,nb6-ser-r0", .data = &board_nb6, },
+#endif
 #ifdef CONFIG_BCM63XX_CPU_6368
 	{ .compatible = "brcm,bcm96368mvngr", .data = &board_96368mvngr, },
 	{ .compatible = "brcm,bcm96368mvwg", .data = &board_96368mvwg, },
