From: Max Staudt <openwrt.max@enpas.org>
Date: Wed, 15 Jan 2014 18:51:13 +0000
Subject: [PATCH] brcm63xx: F@ST2504n board support (Linux-3.10.26)

Signed-off-by: Max Staudt <openwrt.max@enpas.org>
---
--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -1332,6 +1332,41 @@ static struct board_info __initdata boar
 		},
 	},
 };
+
+static struct board_info __initdata board_fast2504n = {
+	.name				= "F@ST2504n",
+	.expected_cpu_id		= 0x6362,
+
+	.has_enetsw			= 1,
+
+	.enetsw = {
+		.used_ports = {
+			[0] = {
+				.used	= 1,
+				.phy_id	= 1,
+				.name	= "Port 1",
+			},
+
+			[1] = {
+				.used	= 1,
+				.phy_id	= 2,
+				.name	= "Port 2",
+			},
+
+			[2] = {
+				.used	= 1,
+				.phy_id	= 3,
+				.name	= "Port 3",
+			},
+
+			[3] = {
+				.used	= 1,
+				.phy_id	= 4,
+				.name	= "Port 4",
+			},
+		},
+	},
+};
 #endif /* CONFIG_BCM63XX_CPU_6362 */
 
 /*
@@ -1495,6 +1530,7 @@ static const struct board_info __initcon
 
 #ifdef CONFIG_BCM63XX_CPU_6362
 	&board_nb6,
+	&board_fast2504n,
 #endif
 
 #ifdef CONFIG_BCM63XX_CPU_6368
@@ -1576,6 +1612,7 @@ static struct of_device_id const bcm963x
 	{ .compatible = "telsey,cpva642", .data = &board_CPVA642, },
 #endif
 #ifdef CONFIG_BCM63XX_CPU_6362
+	{ .compatible = "sagem,f@st2504n", .data = &board_fast2504n, },
 	{ .compatible = "sfr,nb6-ser-r0", .data = &board_nb6, },
 #endif
 #ifdef CONFIG_BCM63XX_CPU_6368
