From 07ce9e9bc4dcd5ac4728e587901112eef95bbe7b Mon Sep 17 00:00:00 2001
From: Stefan Koch <stefan.koch10@gmail.com>
Date: Mon, 13 Mar 2017 23:42:00 +0100
Subject: [PATCH] lantiq: vpe nosmp

Signed-off-by: Stefan Koch <stefan.koch10@gmail.com>
---
 arch/mips/kernel/vpe-mt.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

--- a/arch/mips/kernel/vpe-mt.c
+++ b/arch/mips/kernel/vpe-mt.c
@@ -130,7 +130,10 @@ int vpe_run(struct vpe *v)
 	 * kernels need to turn it on, even if that wasn't the pre-dvpe() state.
 	 */
 #ifdef CONFIG_SMP
-	evpe(vpeflags);
+	if (!setup_max_cpus) /* nosmp is set */
+		evpe(EVPE_ENABLE);
+	else
+		evpe(vpeflags);
 #else
 	evpe(EVPE_ENABLE);
 #endif
