--- a/init/Kconfig
+++ b/init/Kconfig
@@ -2388,7 +2388,7 @@ config PADATA
 	bool
 
 config ASN1
-	tristate
+	tristate "ASN1"
 	help
 	  Build a simple ASN.1 grammar compiler that produces a bytecode output
 	  that can be interpreted by the ASN.1 stream decoder and used to
--- a/lib/Kconfig
+++ b/lib/Kconfig
@@ -614,7 +614,7 @@ config LIBFDT
 	bool
 
 config OID_REGISTRY
-	tristate
+	tristate "OID"
 	help
 	  Enable fast lookup object identifier registry.
 
