--- a/driver/siutils.c
+++ b/driver/siutils.c
@@ -1859,7 +1859,7 @@ BCMINITFN(si_devpath)(si_t *sih, char *p
 	case PCI_BUS:
 		ASSERT((SI_INFO(sih))->osh != NULL);
 		slen = snprintf(path, (size_t)size, "pci/%u/%u/",
-		                OSL_PCI_BUS((SI_INFO(sih))->osh),
+		                OSL_PCI_BUS((SI_INFO(sih))->osh) + 1,
 		                OSL_PCI_SLOT((SI_INFO(sih))->osh));
 		break;
 	case PCMCIA_BUS:
