--- a/driver/wl_linux.c
+++ b/driver/wl_linux.c
@@ -762,7 +762,7 @@ wl_attach(uint16 vendor, uint16 device,
 		dev->name, device);
 
 #ifdef BCMDBG
-	printf(" (Compiled in " SRCBASE " at " __TIME__ " on " __DATE__ ")");
+	printf(" (Compiled in " SRCBASE ")");
 #endif /* BCMDBG */
 	printf("\n");
 
@@ -2298,8 +2298,7 @@ wl_sendup(wl_info_t *wl, wl_if_t *wlif,
 void
 wl_dump_ver(wl_info_t *wl, struct bcmstrbuf *b)
 {
-	bcm_bprintf(b, "wl%d: %s %s version %s\n", wl->pub->unit,
-		__DATE__, __TIME__, EPI_VERSION_STR);
+	bcm_bprintf(b, "wl%d: version %s\n", wl->pub->unit, EPI_VERSION_STR);
 }
 
 #ifdef BCMDBG
