--- a/libs/src/misc/Makefile.in
+++ b/libs/src/misc/Makefile.in
@@ -62,7 +62,7 @@ libopt: clean
 
 
 $(LIBNAME): $(OBJECTS)
-	ar r $@ $?
+	$(AR) r $@ $?
 	$(RANLIB) $@
 
 install: $(LIBNAME)
--- a/libs/src/portable/Makefile.in
+++ b/libs/src/portable/Makefile.in
@@ -57,7 +57,7 @@ libopt: clean
 
 
 $(LIBNAME): $(OBJECTS)
-	ar r $@ $?
+	$(AR) r $@ $?
 	$(RANLIB) $@
 
 install: $(LIBNAME)
--- a/libs/src/pset/Makefile.in
+++ b/libs/src/pset/Makefile.in
@@ -53,7 +53,7 @@ libopt: clean
 	mv $(LIBNAME) $(LIBDIR)/optimized
 
 $(LIBNAME): $(OBJECTS)
-	ar r $@ $?
+	$(AR) r $@ $?
 	$(RANLIB) $@
 
 install: $(LIBNAME)
--- a/libs/src/sio/Makefile.in
+++ b/libs/src/sio/Makefile.in
@@ -52,7 +52,7 @@ libopt: clean
 	$(INSTALL) $(FMODE) $(LIBNAME) $(LIBDIR)/optimized
 
 $(LIBNAME): $(OBJECTS)
-	ar r $@ $?
+	$(AR) r $@ $?
 	$(RANLIB) $@
 
 install: $(LIBNAME)
--- a/libs/src/str/Makefile.in
+++ b/libs/src/str/Makefile.in
@@ -63,7 +63,7 @@ libopt: clean
 	$(INSTALL) $(FMODE) $(LIBNAME) $(LIBDIR)-O
 
 $(LIBNAME): $(OBJECTS)
-	ar r $@ $?
+	$(AR) r $@ $?
 	$(RANLIB) $@
 
 LINT_IGNORE=possible pointer alignment|RCSid unused
--- a/libs/src/xlog/Makefile.in
+++ b/libs/src/xlog/Makefile.in
@@ -58,7 +58,7 @@ libopt: clean
 	$(INSTALL) $(FMODE) $(LIBNAME) $(LIBDIR)/optimized
 
 $(LIBNAME): $(OBJECTS)
-	ar r $@ $?
+	$(AR) r $@ $?
 	$(RANLIB) $@
 
 install: $(LIBNAME)
