From: coolsnowwolf <coolsnowwolf@gmail.com>
Subject: [PATCH] netfilter: fullcone: retain unconfirmed NAT mappings

Signed-off-by: coolsnowwolf <coolsnowwolf@gmail.com>

--- a/src/nf_nat_fullcone.c
+++ b/src/nf_nat_fullcone.c
@@ -683,6 +683,11 @@
 	list_for_each_safe(iter, tmp, &mapping->original_tuple_list) {
 		original_tuple_item = list_entry(iter, struct nat_mapping_original_tuple, node);
 
+#ifdef NF_NAT_MAPPING_ACTIVE
+		if (nf_nat_mapping_active(net, zone, &original_tuple_item->tuple))
+			continue;
+#endif
+
 		tuple_hash = nf_conntrack_find_get(net, zone, &original_tuple_item->tuple);
 
 		if (tuple_hash == NULL) {
@@ -737,6 +742,11 @@
 	list_for_each_safe(iter, tmp, &mapping->original_tuple_list) {
 		original_tuple_item = list_entry(iter, struct nat_mapping_original_tuple, node);
 
+#ifdef NF_NAT_MAPPING_ACTIVE
+		if (nf_nat_mapping_active(net, zone, &original_tuple_item->tuple))
+			continue;
+#endif
+
 		tuple_hash = nf_conntrack_find_get(net, zone, &original_tuple_item->tuple);
 
 		if (tuple_hash == NULL) {
