--- a/ath10k/htt_rx.c
+++ b/ath10k/htt_rx.c
@@ -2514,7 +2514,7 @@ bool ath10k_htt_t2h_msg_handler(struct a
 		u32 freq = __le32_to_cpu(resp->chan_change.freq);
 
 		ar->tgt_oper_chan =
-			__ieee80211_get_channel(ar->hw->wiphy, freq);
+			ieee80211_get_channel(ar->hw->wiphy, freq);
 		ath10k_dbg(ar, ATH10K_DBG_HTT,
 			   "htt chan change freq %u phymode %s\n",
 			   freq, ath10k_wmi_phymode_str(phymode));
