From: Shiji Yang <yangshiji66@outlook.com>
Date: Thu, 26 Mar 2026 22:44:28 +0800
Subject: [PATCH] ath10k: correct driver version number

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
---
 ath10k-7.2/pci.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/ath10k-7.2/pci.c
+++ b/ath10k-7.2/pci.c
@@ -3872,7 +3872,7 @@ static int __ath10k_pci_probe(struct pci
 	int (*pci_hard_reset)(struct ath10k *ar);
 	u32 (*targ_cpu_to_ce_addr)(struct ath10k *ar, u32 addr);
 
-	printk(KERN_INFO "ath10k 6.15 driver, optimized for CT firmware, probing pci device: 0x%x.\n",
+	printk(KERN_INFO "ath10k 7.2 driver, optimized for CT firmware, probing pci device: 0x%x.\n",
 	       pci_dev->device);
 
 	switch (pci_dev->device) {
