--- a/ath10k-7.2/wmi.h
+++ b/ath10k-7.2/wmi.h
@@ -6310,7 +6310,7 @@ struct qca9880_set_ctl_table_cmd {
 	__le32 ctl_len; /* in bytes.  This may be ignored in firmware,
 			 * make sure ctl_info data is sizeof(qca9880_power_ctl) */
 	/** ctl array (len adjusted to  number of words) */
-	__le32 *ctl_info; /* data would be the qca9880_power_ctl table above */
+	__le32 ctl_info[]; /* data would be the qca9880_power_ctl table above */
 };
 
 /* Used by:   WMI_PDEV_SET_MIMOGAIN_TABLE_CMDID */
