--- a/ath10k-5.10/htt_rx.c
+++ b/ath10k-5.10/htt_rx.c
@@ -3026,7 +3026,7 @@
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
+				//dev_warn_once(ar->dev, "htt tx ct: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -3120,7 +3120,7 @@
 
 			/* workaround for possibly firmware bug */
 			if (unlikely(tx_done.tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-				dev_warn_once(ar->dev, "htt tx: fixing invalid VHT TX rate code 0xff\n");
+				//dev_warn_once(ar->dev, "htt tx: fixing invalid VHT TX rate code 0xff\n");
 				tx_done.tx_rate_code = 0;
 			}
 
@@ -3970,11 +3970,11 @@
 	if (txrate.flags == WMI_RATE_PREAMBLE_VHT && txrate.mcs > 9) {
 		static bool done_once = 0;
 		if (!done_once) {
-			ath10k_warn(ar, "Invalid VHT mcs %hhd peer stats",  txrate.mcs);
+			//ath10k_warn(ar, "Invalid VHT mcs %hhd peer stats",  txrate.mcs);
 			done_once = true;
 		}
 		else {
-			ath10k_dbg(ar, ATH10K_DBG_HTT, "Invalid VHT mcs %hhd peer stats",  txrate.mcs);
+			//ath10k_dbg(ar, ATH10K_DBG_HTT, "Invalid VHT mcs %hhd peer stats",  txrate.mcs);
 		}
 		return;
 	}
--- a/ath10k-5.10/wmi.c
+++ b/ath10k-5.10/wmi.c
@@ -6329,7 +6329,7 @@
 
 	/* workaround for possibly firmware bug */
 	if (unlikely(ev->tx_rate_code == ATH10K_CT_TX_BEACON_INVALID_RATE_CODE)) {
-		dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
+		//dev_warn_once(ar->dev, "wmi: fixing invalid VHT TX rate code 0xff\n");
 		ev->tx_rate_code = 0;
 	}
 
