Index: drv_dsl_cpe_api-3.24.4.4/src/common/drv_dsl_cpe_os_linux.c
===================================================================
--- drv_dsl_cpe_api-3.24.4.4.orig/src/common/drv_dsl_cpe_os_linux.c	2012-12-07 21:22:58.020256076 +0100
+++ drv_dsl_cpe_api-3.24.4.4/src/common/drv_dsl_cpe_os_linux.c	2012-12-07 21:31:13.156268489 +0100
@@ -12,6 +12,7 @@
 
 #define DSL_INTERN
 #include <linux/device.h>
+#include <linux/platform_device.h>
 
 #include "drv_dsl_cpe_api.h"
 #include "drv_dsl_cpe_api_ioctl.h"
@@ -1063,7 +1064,7 @@
 #endif
 
 /* Entry point of driver */
-int __init DSL_ModuleInit(void)
+static int __devinit ltq_adsl_probe(struct platform_device *pdev)
 {
    struct class *dsl_class;
    DSL_int_t i;
@@ -1117,7 +1118,7 @@
    return 0;
 }
 
-void __exit DSL_ModuleCleanup(void)
+static int __devexit ltq_adsl_remove(struct platform_device *pdev)
 {
    printk("Module will be unloaded"DSL_DRV_CRLF);
 
@@ -1132,7 +1133,7 @@
                (DSL_uint8_t**)&g_BndFpgaBase);
 #endif /* defined(INCLUDE_DSL_CPE_API_VINAX) && defined(INCLUDE_DSL_BONDING)*/
 
-   return;
+   return 0;
 }
 
 #ifndef _lint
@@ -1148,8 +1149,30 @@
 MODULE_PARM_DESC(debug_level, "set to get more (1) or fewer (4) debug outputs");
 #endif /* #ifndef DSL_DEBUG_DISABLE*/
 
-module_init(DSL_ModuleInit);
-module_exit(DSL_ModuleCleanup);
+static const struct of_device_id ltq_adsl_match[] = {
+#ifdef CONFIG_DANUBE
+	{ .compatible = "lantiq,adsl-danube"},
+#elif defined CONFIG_AMAZON_SE
+	{ .compatible = "lantiq,adsl-ase"},
+#elif defined CONFIG_AR9
+	{ .compatible = "lantiq,adsl-arx100"},
+#endif
+	{},
+};
+MODULE_DEVICE_TABLE(of, ltq_adsl_match);
+
+static struct platform_driver ltq_adsl_driver = {
+	.probe = ltq_adsl_probe,
+	.remove = __devexit_p(ltq_adsl_remove),
+	.driver = {
+		.name = "adsl",
+		.owner = THIS_MODULE,
+		.of_match_table = ltq_adsl_match,
+	},
+};
+
+module_platform_driver(ltq_adsl_driver);
+
 #endif /* #ifndef _lint*/
 
 //EXPORT_SYMBOL(DSL_ModuleInit);
