--- a/src/common/drv_dsl_cpe_os_linux.c
+++ b/src/common/drv_dsl_cpe_os_linux.c
@@ -417,7 +417,11 @@ int DSL_DRV_ErrorToOS(DSL_Error_t nError
 DSL_void_t* DSL_DRV_VMalloc(
    DSL_DRV_size_t    nSize)
 {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5,9,0)
    return __vmalloc((unsigned long)nSize, GFP_KERNEL, PAGE_KERNEL);
+#else
+   return __vmalloc((unsigned long)nSize, GFP_KERNEL);
+#endif
    /*   return vmalloc(nSize);*/
 }
 
