--- a/src/common/drv_dsl_cpe_os_linux.c
+++ b/src/common/drv_dsl_cpe_os_linux.c
@@ -549,7 +549,7 @@ static int DSL_DRV_KernelThreadStartup(v
    retVal = pThrCntrl->pThrFct(&pThrCntrl->thrParams);
    pThrCntrl->thrParams.bRunning = 0;
 
-   complete_and_exit(&pThrCntrl->thrCompletion, (long)retVal);
+   kthread_complete_and_exit(&pThrCntrl->thrCompletion, (long)retVal);
 
    DSL_DEBUG( DSL_DBG_MSG,
       (DSL_NULL, "EXIT - Kernel Thread Startup <%s>" DSL_DRV_CRLF,
