--- a/src/pm/drv_dsl_cpe_api_pm_danube.c
+++ b/src/pm/drv_dsl_cpe_api_pm_danube.c
@@ -182,7 +182,7 @@ DSL_Error_t DSL_DRV_PM_DEV_ChannelCounte
    }
    else
    {
-      if (DSL_DRV_DANUBE_ActLatencyGet(pContext, nChannel, nDirection, &nLPath) >= DSL_SUCCESS)
+      if (DSL_DRV_DANUBE_ActLatencyGet(pContext, nChannel, (DSL_AccessDir_t)nDirection, &nLPath) >= DSL_SUCCESS)
       {
          if (nLPath == DSL_LATENCY_IP_LP0) nIdx = 0;
          else nIdx = 1;
@@ -345,7 +345,7 @@ DSL_Error_t DSL_DRV_PM_DEV_ChannelCounte
 
    if ((!bAdsl1) && (DSL_CHANNELS_PER_LINE > 1))
    {
-      if (DSL_DRV_DANUBE_ActLatencyGet(pContext, nChannel, nDirection, &nLPath) >= DSL_SUCCESS)
+      if (DSL_DRV_DANUBE_ActLatencyGet(pContext, nChannel, (DSL_AccessDir_t)nDirection, &nLPath) >= DSL_SUCCESS)
       {
          if (nLPath == DSL_LATENCY_IP_LP0) nIdx = 0;
          else nIdx = 1;
