--- a/arch/powerpc/boot/dts/kilauea.dts
+++ b/arch/powerpc/boot/dts/kilauea.dts
@@ -102,8 +102,8 @@
 			compatible = "ibm,sdram-405ex", "ibm,sdram-4xx-ddr2";
 			dcr-reg = <0x010 0x002>;
 			interrupt-parent = <&UIC2>;
-			interrupts = <0x5 0x4	/* ECC DED Error */ 
-				      0x6 0x4>;	/* ECC SEC Error */ 
+			interrupts = <0x5 0x4	/* ECC DED Error */
+				      0x6 0x4>;	/* ECC SEC Error */
 		};
 
 		CRYPTO: crypto@ef700000 {
@@ -157,30 +157,30 @@
 					reg = <0x00000000 0x00000000 0x04000000>;
 					#address-cells = <1>;
 					#size-cells = <1>;
-					partition@0 {
+					partition0@0 {
 						label = "kernel";
 						reg = <0x00000000 0x001e0000>;
 					};
-					partition@1e0000 {
+					partition1@1e0000 {
 						label = "dtb";
 						reg = <0x001e0000 0x00020000>;
 					};
-					partition@200000 {
-						label = "root";
-						reg = <0x00200000 0x00200000>;
-					};
-					partition@400000 {
-						label = "user";
-						reg = <0x00400000 0x03b60000>;
+					partition2@200000 {
+						label = "rootfs";
+						reg = <0x00200000 0x03d60000>;
 					};
-					partition@3f60000 {
+					partition3@3f60000 {
 						label = "env";
 						reg = <0x03f60000 0x00040000>;
 					};
-					partition@3fa0000 {
+					partition4@3fa0000 {
 						label = "u-boot";
 						reg = <0x03fa0000 0x00060000>;
 					};
+					partition5@0 {
+						label = "firmware";
+						reg = <0x00000000 0x03f60000>;
+					};
 				};
 
 				ndfc@1,0 {
