--- a/controller/seafile-controller.c
+++ b/controller/seafile-controller.c
@@ -516,7 +516,7 @@ init_pidfile_path (SeafileController *ct
         pid_dir = g_build_filename (topdir, "pids", NULL);
     }
     if (!g_file_test(pid_dir, G_FILE_TEST_EXISTS)) {
-        if (g_mkdir(pid_dir, 0777) < 0) {
+        if (g_mkdir(pid_dir, 0755) < 0) {
             seaf_warning("failed to create pid dir %s: %s", pid_dir, strerror(errno));
             controller_exit(1);
         }
