--- a/scripts/setup-seafile-mysql.py
+++ b/scripts/setup-seafile-mysql.py
@@ -316,9 +316,6 @@ class EnvManager(object):
             os.path.join(self.install_path, 'runtime'),
         ]
 
-        for path in paths:
-            error_if_not_exists(path)
-
         if os.path.exists(ccnet_config.ccnet_dir):
             Utils.error('Ccnet config dir \"%s\" already exists.' % ccnet_config.ccnet_dir)
 
--- a/scripts/setup-seafile.sh
+++ b/scripts/setup-seafile.sh
@@ -415,7 +415,6 @@ if [[ $# -ge 1 && "$1" == "auto" ]]; the
     need_pause=0
 fi
 
-check_sanity;
 if [[ "${need_pause}" == "1" ]]; then
     welcome;
 fi
