--- a/scripts/setup-seafile-mysql.py
+++ b/scripts/setup-seafile-mysql.py
@@ -1543,7 +1543,6 @@ def main():
     seahub_config.do_syncdb()
     seahub_config.prepare_avatar_dir()
     # db_config.create_seahub_admin()
-    user_manuals_handler.copy_user_manuals()
     create_seafile_server_symlink()
 
     set_file_perm()
--- a/scripts/setup-seafile.sh
+++ b/scripts/setup-seafile.sh
@@ -709,7 +709,6 @@ chmod 0700 "$default_conf_dir"
 # -------------------------------------------
 # copy user manuals to library template
 # -------------------------------------------
-copy_user_manuals;
 
 # -------------------------------------------
 # final message
