--- a/scripts/setup-seafile-mysql.py
+++ b/scripts/setup-seafile-mysql.py
@@ -1543,7 +1543,6 @@ def main():
     seahub_config.do_syncdb()
     seahub_config.prepare_avatar_dir()
     # db_config.create_seahub_admin()
-    create_seafile_server_symlink()
 
     set_file_perm()
 
--- a/scripts/setup-seafile.sh
+++ b/scripts/setup-seafile.sh
@@ -689,17 +689,6 @@ fi
 # /data/haiwen/
 #            -- seafile-server-2.0.4
 #            -- seafile-server-latest # symlink to 2.0.4
-seafile_server_symlink=${TOPDIR}/seafile-server-latest
-echo
-echo -n "creating seafile-server-latest symbolic link ... "
-if ! ln -s $(basename ${INSTALLPATH}) ${seafile_server_symlink}; then
-    echo
-    echo
-    echo "Failed to create symbolic link ${seafile_server_symlink}"
-    err_and_quit;
-fi
-echo "done"
-echo
 
 chmod 0600 "$dest_settings_py"
 chmod 0700 "$default_ccnet_conf_dir"
--- a/scripts/upgrade/minor-upgrade.sh
+++ b/scripts/upgrade/minor-upgrade.sh
@@ -162,7 +162,6 @@ make_media_custom_symlink;
 
 move_old_elasticsearch_config_to_latest;
 
-update_latest_symlink;
 
 
 echo "DONE"
