--- a/scripts/setup-seafile.sh
+++ b/scripts/setup-seafile.sh
@@ -418,9 +418,9 @@ fi
 if [[ "${need_pause}" == "1" ]]; then
     welcome;
 fi
-sleep .5
+sleep 1
 check_system_dependency;
-sleep .5
+sleep 1
 
 check_existing_ccnet;
 check_existing_seafile;
@@ -438,7 +438,7 @@ if [[ "$fileserver_port" == "" ]]; then
 fi
 
 
-sleep .5
+sleep 1
 
 printf "\nThis is your config information:\n\n"
 
@@ -475,7 +475,7 @@ fi
 echo
 
 
-sleep 0.5
+sleep 1
 
 # -------------------------------------------
 # Create seafile conf
