--- a/scripts/setup-seafile-mysql.py
+++ b/scripts/setup-seafile-mysql.py
@@ -1556,6 +1556,9 @@ def report_success():
 Your seafile server configuration has been finished successfully.
 -----------------------------------------------------------------
 
+You will need to create an admin account before starting seafile server.
+
+create admin account:   create-seafile-admin
 run seafile server:     service seafile-server { start | stop | restart }
 
 -----------------------------------------------------------------
--- a/scripts/setup-seafile.sh
+++ b/scripts/setup-seafile.sh
@@ -710,6 +710,9 @@ echo "----------------------------------
 echo "Your seafile server configuration has been completed successfully." 
 echo "-----------------------------------------------------------------"
 echo 
+echo "You will need to create an admin account before starting seafile server."
+echo
+echo "create admin account:   create-seafile-admin"
 echo "run seafile server:     service seafile-server { start | stop | restart }"
 echo
 echo "-----------------------------------------------------------------"
