--- a/lib/repo.vala
+++ b/lib/repo.vala
@@ -70,7 +70,7 @@ public class Repo : Object {
         get { return _relay_id; }
         set { _relay_id = value; }
     }
-    public int  last_sync_time { get; set; }
+    public int64  last_sync_time { get; set; }
     public bool auto_sync { get; set; }
     public bool worktree_invalid { get; set; }
 
@@ -164,7 +164,7 @@ public class DeletedEntry : Object {
     public string obj_name { get; set; }
     public string basedir { get; set; }
     public int mode { get; set; }
-    public int delete_time { get; set; }
+    public int64 delete_time { get; set; }
     public int64 file_size { get; set; }
     public string scan_stat { get; set; }
 }
