--- a/src/parse_ether.c
+++ b/src/parse_ether.c
@@ -21,10 +21,18 @@
 #include "parse_ether.h"
 
 /* tcpdump header (ether.h) defines ETHER_HDRLEN) */
-#ifndef ETHER_HDRLEN 
+#ifndef ETHER_HDRLEN
 #define ETHER_HDRLEN 14
 #endif
 
+/* uClibc-ng compatibility */
+#ifndef IPPROTO_BEETPH
+#define IPPROTO_BEETPH	94
+#endif
+
+#ifndef IPPROTO_MPLS
+#define IPPROTO_MPLS	137
+#endif
 
 const char *ip_protcol_str(int p)
 {
