--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -7,6 +7,7 @@ dtb-$(CONFIG_ARCH_BCM2835) += \
 	bcm2708-rpi-0-w.dtb \
 	bcm2709-rpi-2-b.dtb \
 	bcm2710-rpi-3-b.dtb \
+	bcm2710-rpi-3-b-plus.dtb \
 	bcm2710-rpi-cm3.dtb
 
 dtb-$(CONFIG_ARCH_ALPINE) += \
--- /dev/null
+++ b/arch/arm/boot/dts/bcm2710-rpi-3-b-plus.dts
@@ -0,0 +1,220 @@
+/dts-v1/;
+
+#include "bcm2710.dtsi"
+#include "bcm283x-rpi-lan7515.dtsi"
+
+/ {
+      compatible = "raspberrypi,3-model-b-plus", "brcm,bcm2837", "brcm,bcm2836";
+      model = "Raspberry Pi 3 Model B+";
+};
+
+&gpio {
+      sdhost_pins: sdhost_pins {
+              brcm,pins = <48 49 50 51 52 53>;
+              brcm,function = <4>; /* alt0 */
+      };
+
+      spi0_pins: spi0_pins {
+              brcm,pins = <9 10 11>;
+              brcm,function = <4>; /* alt0 */
+      };
+
+      spi0_cs_pins: spi0_cs_pins {
+              brcm,pins = <8 7>;
+              brcm,function = <1>; /* output */
+      };
+
+      i2c0_pins: i2c0 {
+              brcm,pins = <0 1>;
+              brcm,function = <4>;
+      };
+
+      i2c1_pins: i2c1 {
+              brcm,pins = <2 3>;
+              brcm,function = <4>;
+      };
+
+      i2s_pins: i2s {
+              brcm,pins = <18 19 20 21>;
+              brcm,function = <4>; /* alt0 */
+      };
+
+      sdio_pins: sdio_pins {
+              brcm,pins =     <34 35 36 37 38 39>;
+              brcm,function = <7>; // alt3 = SD1
+              brcm,pull =     <0 2 2 2 2 2>;
+      };
+
+      bt_pins: bt_pins {
+              brcm,pins = <43>;
+              brcm,function = <4>; /* alt0:GPCLK2 */
+              brcm,pull = <0>;
+      };
+
+      uart0_pins: uart0_pins {
+              brcm,pins = <32 33>;
+              brcm,function = <7>; /* alt3=UART0 */
+              brcm,pull = <0 2>;
+      };
+
+      uart1_pins: uart1_pins {
+              brcm,pins;
+              brcm,function;
+              brcm,pull;
+      };
+
+      audio_pins: audio_pins {
+              brcm,pins = <40 41>;
+              brcm,function = <4>;
+      };
+};
+
+&sdhost {
+      pinctrl-names = "default";
+      pinctrl-0 = <&sdhost_pins>;
+      bus-width = <4>;
+      status = "okay";
+};
+
+&mmc {
+      pinctrl-names = "default";
+      pinctrl-0 = <&sdio_pins>;
+      non-removable;
+      bus-width = <4>;
+      status = "okay";
+      brcm,overclock-50 = <0>;
+};
+
+&soc {
+      expgpio: expgpio {
+              compatible = "brcm,bcm2835-expgpio";
+              gpio-controller;
+              #gpio-cells = <2>;
+              firmware = <&firmware>;
+              status = "okay";
+      };
+};
+
+&fb {
+      status = "okay";
+};
+
+&uart0 {
+      pinctrl-names = "default";
+      pinctrl-0 = <&uart0_pins &bt_pins>;
+      status = "okay";
+};
+
+&uart1 {
+      pinctrl-names = "default";
+      pinctrl-0 = <&uart1_pins>;
+      status = "okay";
+};
+
+&spi0 {
+      pinctrl-names = "default";
+      pinctrl-0 = <&spi0_pins &spi0_cs_pins>;
+      cs-gpios = <&gpio 8 1>, <&gpio 7 1>;
+
+      spidev0: spidev@0{
+              compatible = "spidev";
+              reg = <0>;      /* CE0 */
+              #address-cells = <1>;
+              #size-cells = <0>;
+              spi-max-frequency = <125000000>;
+      };
+
+      spidev1: spidev@1{
+              compatible = "spidev";
+              reg = <1>;      /* CE1 */
+              #address-cells = <1>;
+              #size-cells = <0>;
+              spi-max-frequency = <125000000>;
+      };
+};
+
+&i2c0 {
+      pinctrl-names = "default";
+      pinctrl-0 = <&i2c0_pins>;
+      clock-frequency = <100000>;
+};
+
+&i2c1 {
+      pinctrl-names = "default";
+      pinctrl-0 = <&i2c1_pins>;
+      clock-frequency = <100000>;
+};
+
+&i2c2 {
+      clock-frequency = <100000>;
+};
+
+&i2s {
+      pinctrl-names = "default";
+      pinctrl-0 = <&i2s_pins>;
+};
+
+&leds {
+      act_led: act {
+              label = "led0";
+              linux,default-trigger = "mmc0";
+              gpios = <&gpio 29 0>;
+      };
+
+      pwr_led: pwr {
+              label = "led1";
+              linux,default-trigger = "default-on";
+              gpios = <&expgpio 2 GPIO_ACTIVE_LOW>;
+      };
+};
+
+&hdmi {
+      hpd-gpios = <&expgpio 4 GPIO_ACTIVE_LOW>;
+};
+
+&audio {
+      pinctrl-names = "default";
+      pinctrl-0 = <&audio_pins>;
+};
+
+/ {
+      chosen {
+              bootargs = "8250.nr_uarts=1";
+      };
+};
+
+/ {
+      __overrides__ {
+              uart0 = <&uart0>,"status";
+              uart1 = <&uart1>,"status";
+              i2s = <&i2s>,"status";
+              spi = <&spi0>,"status";
+              i2c0 = <&i2c0>,"status";
+              i2c1 = <&i2c1>,"status";
+              i2c2_iknowwhatimdoing = <&i2c2>,"status";
+              i2c0_baudrate = <&i2c0>,"clock-frequency:0";
+              i2c1_baudrate = <&i2c1>,"clock-frequency:0";
+              i2c2_baudrate = <&i2c2>,"clock-frequency:0";
+
+              act_led_gpio = <&act_led>,"gpios:4";
+              act_led_activelow = <&act_led>,"gpios:8";
+              act_led_trigger = <&act_led>,"linux,default-trigger";
+
+              pwr_led_gpio = <&pwr_led>,"gpios:4";
+              pwr_led_activelow = <&pwr_led>,"gpios:8";
+              pwr_led_trigger = <&pwr_led>,"linux,default-trigger";
+
+              audio = <&audio>,"status";
+              watchdog = <&watchdog>,"status";
+              random = <&random>,"status";
+              sd_overclock = <&sdhost>,"brcm,overclock-50:0";
+              sd_force_pio = <&sdhost>,"brcm,force-pio?";
+              sd_pio_limit = <&sdhost>,"brcm,pio-limit:0";
+              sd_debug     = <&sdhost>,"brcm,debug";
+      };
+};
+
+&i2c_dsi {
+      gpios = <&gpio 44 0
+               &gpio 45 0>;
+};
--- /dev/null
+++ b/arch/arm/boot/dts/bcm283x-rpi-lan7515.dtsi
@@ -0,0 +1,36 @@
+/ {
+      aliases {
+              ethernet = &ethernet;
+      };
+};
+
+&usb {
+      usb1@1 {
+              compatible = "usb424,2514";
+              reg = <1>;
+              #address-cells = <1>;
+              #size-cells = <0>;
+
+              usb1_1@1 {
+                      compatible = "usb424,2514";
+                      reg = <1>;
+                      #address-cells = <1>;
+                      #size-cells = <0>;
+
+                      ethernet: usbether@1 {
+                              compatible = "usb424,7800";
+                              reg = <1>;
+                              microchip,eee-enabled;
+                              microchip,tx-lpi-timer = <600>; /* non-aggressive*/
+                      };
+              };
+      };
+};
+
+
+/ {
+      __overrides__ {
+              eee = <&ethernet>,"microchip,eee-enabled?";
+              tx_lpi_timer = <&ethernet>,"microchip,tx-lpi-timer:0";
+      };
+};
--- a/arch/arm64/boot/dts/broadcom/Makefile
+++ b/arch/arm64/boot/dts/broadcom/Makefile
@@ -8,6 +8,7 @@ dtb-$(CONFIG_ARCH_BCM2835) += bcm2837-rp
 dtb-$(CONFIG_ARCH_BCM_IPROC) += ns2-svk.dtb
 dtb-$(CONFIG_ARCH_VULCAN) += vulcan-eval.dtb
 dtb-$(CONFIG_ARCH_BCM2835) += bcm2710-rpi-3-b.dtb
+dtb-$(CONFIG_ARCH_BCM2835) += bcm2710-rpi-3-b-plus.dtb
 
 dts-dirs += ../overlays
 
--- /dev/null
+++ b/arch/arm64/boot/dts/broadcom/bcm2710-rpi-3-b-plus.dts
@@ -0,0 +1,3 @@
+#define RPI364
+
+#include "../../../../arm/boot/dts/bcm2710-rpi-3-b-plus.dts"
--- a/arch/arm64/configs/bcmrpi3_defconfig
+++ b/arch/arm64/configs/bcmrpi3_defconfig
@@ -482,6 +482,7 @@ CONFIG_USB_NET_SR9700=m
 CONFIG_USB_NET_SR9800=m
 CONFIG_USB_NET_SMSC75XX=m
 CONFIG_USB_NET_SMSC95XX=y
+CONFIG_USB_LAN78XX=y
 CONFIG_USB_NET_GL620A=m
 CONFIG_USB_NET_NET1080=m
 CONFIG_USB_NET_PLUSB=m
