From ed753b7389b8c65af0b4266c001ef0343fb1c30e Mon Sep 17 00:00:00 2001
From: Phil Elwell <pelwell@users.noreply.github.com>
Date: Tue, 3 Jan 2017 21:27:46 +0000
Subject: [PATCH] Document the si7020 option

---
 arch/arm/boot/dts/overlays/README | 3 +++
 1 file changed, 3 insertions(+)

--- a/arch/arm/boot/dts/overlays/README
+++ b/arch/arm/boot/dts/overlays/README
@@ -549,6 +549,9 @@ Params: bmp085                  Select t
 
         bmp280                  Select the Bosch sensortronic BMP280
 
+        si7020                  Select the Silicon Labs Si7013/20/21 humidity/
+                                temperature sensor
+
 
 Name:   i2c0-bcm2708
 Info:   Enable the i2c_bcm2708 driver for the i2c0 bus. Not all pin combinations
