From 711f7413d72af6328a3958eecd400bd8911af211 Mon Sep 17 00:00:00 2001
From: popcornmix <popcornmix@gmail.com>
Date: Mon, 25 Jul 2016 17:06:50 +0100
Subject: [PATCH] iqaudio-dac: Compile fix - untested

---
 sound/soc/bcm/iqaudio-dac.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

--- a/sound/soc/bcm/iqaudio-dac.c
+++ b/sound/soc/bcm/iqaudio-dac.c
@@ -76,7 +76,11 @@ static void snd_rpi_iqaudio_gpio_unmute(
 static int snd_rpi_iqaudio_set_bias_level(struct snd_soc_card *card,
 	struct snd_soc_dapm_context *dapm, enum snd_soc_bias_level level)
 {
-	struct snd_soc_dai *codec_dai = card->rtd[0].codec_dai;
+	struct snd_soc_pcm_runtime *rtd;
+	struct snd_soc_dai *codec_dai;
+
+	rtd = snd_soc_get_pcm_runtime(card, card->dai_link[0].name);
+	codec_dai = rtd->codec_dai;
 
 	if (dapm->dev != codec_dai->dev)
 		return 0;
