--- a/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-phicomm-n1.dts
+++ b/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-phicomm-n1.dts
@@ -10,6 +10,17 @@
 / {
 	compatible = "phicomm,n1", "amlogic,s905d", "amlogic,meson-gxl";
 	model = "Phicomm N1";
+	
+	aliases {
+		mmc0 = &sd_emmc_b;
+		mmc1 = &sd_emmc_c;
+		mmc2 = &sd_emmc_a;
+
+		led-boot = &led_work;
+		led-failsafe = &led_work;
+		led-running = &led_work;
+		led-upgrade = &led_work;
+	};
 
 	cvbs-connector {
 		status = "disabled";
@@ -18,7 +29,7 @@
 	leds {
 		compatible = "gpio-leds";
 
-		led {
+		led_work: led {
 			label = "n1:white:status";
 			gpios = <&gpio_ao GPIOAO_9 GPIO_ACTIVE_HIGH>;
 			default-state = "on";
@@ -33,3 +44,18 @@
 &usb {
 	dr_mode = "host";
 };
+
+&sd_emmc_a {
+	max-frequency = <80000000>;
+	sd-uhs-ddr50;
+};
+
+&sd_emmc_b {
+	status = "disabled";
+};
+
+&ethmac {
+	snps,aal;
+	snps,txpbl = <0x8>;
+	snps,rxpbl = <0x8>;
+};
