--- a/drv_T6963.c
+++ b/drv_T6963.c
@@ -114,7 +114,9 @@ static void drv_T6_status1(void)
     /* wait for STA0=1 and STA1=1 */
     n = 0;
     do {
+#if 0
 	rep_nop();
+#endif
 	if (++n > 1000) {
 	    debug("hang in status1");
 	    bug = 1;
@@ -150,7 +152,9 @@ static void drv_T6_status2(void)
     /* wait for STA3=1 */
     n = 0;
     do {
+#if 0
 	rep_nop();
+#endif
 	if (++n > 1000) {
 	    debug("hang in status2");
 	    bug = 1;
