--- a/drv_GLCD2USB.c
+++ b/drv_GLCD2USB.c
@@ -48,6 +48,7 @@
 #include <fcntl.h>
 #include <sys/time.h>
 #include <usb.h>
+#include <endian.h>
 
 #include "debug.h"
 #include "cfg.h"
@@ -487,6 +488,9 @@ static int drv_GLCD2USB_start(const char
 	return -1;
     }
 
+    buffer.display_info.width = le16toh(buffer.display_info.width);
+    buffer.display_info.height = le16toh(buffer.display_info.height);
+
     info("%s: display name = %s", Name, buffer.display_info.name);
     info("%s: display resolution = %d * %d", Name, buffer.display_info.width, buffer.display_info.height);
     info("%s: display flags: %x", Name, buffer.display_info.flags);
