--- a/Makefile.am
+++ b/Makefile.am
@@ -458,14 +458,13 @@ testvector_FILES = \
     TestVectors/lea.txt               TestVectors/sosemanuk.txt \
     TestVectors/mars.txt              TestVectors/speck.txt \
     TestVectors/nr.txt                TestVectors/tea.txt \
-    TestVectors/ocb.txt               TestVectors/threefish.txt \
-    TestVectors/panama.txt            TestVectors/ttmac.txt \
-    TestVectors/poly1305aes.txt       TestVectors/vmac.txt \
-    TestVectors/poly1305_tls.txt      TestVectors/wake.txt \
-    TestVectors/rabbit.txt            TestVectors/whrlpool.txt \
-    TestVectors/Readme.txt            TestVectors/xchacha.txt \
-    TestVectors/rsa_oaep.txt          TestVectors/xts.txt \
-    TestVectors/rsa_pkcs1_1_5.txt
+    TestVectors/panama.txt            TestVectors/threefish.txt \
+    TestVectors/poly1305aes.txt       TestVectors/ttmac.txt \
+    TestVectors/poly1305_tls.txt      TestVectors/vmac.txt \
+    TestVectors/rabbit.txt            TestVectors/wake.txt \
+    TestVectors/Readme.txt            TestVectors/whrlpool.txt \
+    TestVectors/rsa_oaep.txt          TestVectors/xchacha.txt \
+    TestVectors/rsa_pkcs1_1_5.txt     TestVectors/xts.txt
 
 # Create with 'ls TestPrograms/test_*.cxx'. These files do not get installed.
 # They exist so the feature tests can be run from the configure directory.
