--- /dev/null
+++ b/arch/arm/boot/dts/bcm4709-arris-sbr-ac3200p.dts
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/*
+ * DTS for ARRIS SBR-AC3200P
+ *
+ * Copyright (C) 2019 lostlonger <lostlonger.g@gmail.com>
+ * Copyright (C) 2019 paldier <paldier@hotmail.com>
+ */
+
+/dts-v1/;
+
+#include "bcm4709.dtsi"
+#include "bcm5301x-nand-cs0-bch4.dtsi"
+
+/ {
+	compatible = "arris,sbr-ac3200p", "brcm,bcm4709", "brcm,bcm4708";
+	model = "ARRIS SBR-AC3200P";
+
+	chosen {
+		bootargs = "console=ttyS0,115200";
+	};
+
+	memory {
+		reg = <0x00000000 0x08000000
+		       0x88000000 0x08000000>;
+	};
+
+	gpio-keys {
+		compatible = "gpio-keys";
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		restart {
+			label = "Reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&chipcommon 19 GPIO_ACTIVE_LOW>;
+		};
+		
+		wps {
+			label = "WPS";
+			linux,code = <KEY_WPS_BUTTON>;
+			gpios = <&chipcommon 20 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&usb2 {
+	vcc-gpio = <&chipcommon 18 GPIO_ACTIVE_HIGH>;
+};
+
+&usb3 {
+	vcc-gpio = <&chipcommon 21 GPIO_ACTIVE_HIGH>;
+};
+
+&usb3_phy {
+	status = "okay";
+};
+
+&nandcs {
+	partitions {
+		compatible = "fixed-partitions";
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		partition@0 {
+			label = "boot";
+			reg = <0x0000000 0x0080000>;
+			read-only;
+		};
+
+		partition@80000 {
+			label = "nvram";
+			reg = <0x0080000 0x0180000>;
+		};
+
+		partition@200000 {
+			label = "firmware";
+			reg = <0x0200000 0x6600000>;
+			compatible = "brcm,trx";
+		};
+
+		partition@6800000 {
+			label = "env";
+			reg = <0x6800000 0x0200000>;
+			read-only;
+		};
+
+		partition@6a00000 {
+			label = "data";
+			reg = <0x6a00000 0x1600000>;
+		};
+	};
+};
+