--- a/configure.ac
+++ b/configure.ac
@@ -580,10 +580,10 @@ AC_TRY_COMPILE([
 
 usable_regex=
 AC_MSG_CHECKING(for usuable C++11 regex)
-AC_TRY_RUN([
+AC_TRY_COMPILE([
 #include <regex>
-#include <iostream>
-int main() {
+#include <iostream>],
+[int main() {
   const std::regex regex(".*");
   const std::string string = "This should match!";
   const auto result = std::regex_search(string, regex);
