--- a/configure
+++ b/configure
@@ -4635,6 +4635,7 @@ $as_echo_n "checking build system compil
 # remove anything that might look like compiler output to our "||" expression
 rm -f conftest* a.out b.out a.exe a_out.exe
 cat >conftest.c <<EOF
+#include <stdlib.h>
 int
 main ()
 {
@@ -4667,6 +4668,7 @@ $as_echo_n "checking build system compil
 # remove anything that might look like compiler output to our "||" expression
 rm -f conftest* a.out b.out a.exe a_out.exe
 cat >conftest.c <<EOF
+#include <stdlib.h>
 int
 main ()
 {
@@ -4703,6 +4705,7 @@ $as_echo_n "checking build system compil
 # remove anything that might look like compiler output to our "||" expression
 rm -f conftest* a.out b.out a.exe a_out.exe
 cat >conftest.c <<EOF
+#include <stdlib.h>
 int
 main ()
 {
@@ -4753,6 +4756,7 @@ else
   gmp_cv_prog_exeext_for_build="$EXEEXT"
 else
   cat >conftest.c <<EOF
+#include <stdlib.h>
 int
 main ()
 {
