--- a/configure
+++ b/configure
@@ -4661,6 +4661,7 @@ $as_echo_n "checking build system compil
 # remove anything that might look like compiler output to our "||" expression
 rm -f conftest* a.out b.out a.exe a_out.exe
 cat >conftest.c <<EOF
+#include <stdlib.h>
 int
 main ()
 {
@@ -4693,6 +4694,7 @@ $as_echo_n "checking build system compil
 # remove anything that might look like compiler output to our "||" expression
 rm -f conftest* a.out b.out a.exe a_out.exe
 cat >conftest.c <<EOF
+#include <stdlib.h>
 int
 main ()
 {
@@ -4729,6 +4731,7 @@ $as_echo_n "checking build system compil
 # remove anything that might look like compiler output to our "||" expression
 rm -f conftest* a.out b.out a.exe a_out.exe
 cat >conftest.c <<EOF
+#include <stdlib.h>
 int
 main ()
 {
@@ -4779,6 +4782,7 @@ else
   gmp_cv_prog_exeext_for_build="$EXEEXT"
 else
   cat >conftest.c <<EOF
+#include <stdlib.h>
 int
 main ()
 {
