--- a/configure
+++ b/configure
@@ -5689,6 +5689,7 @@ printf %s "checking build system compile
 # remove anything that might look like compiler output to our "||" expression
 rm -f conftest* a.out b.out a.exe a_out.exe
 cat >conftest.c <<EOF
+#include <stdlib.h>
 int
 main ()
 {
@@ -5721,6 +5722,7 @@ printf %s "checking build system compile
 # remove anything that might look like compiler output to our "||" expression
 rm -f conftest* a.out b.out a.exe a_out.exe
 cat >conftest.c <<EOF
+#include <stdlib.h>
 int
 main ()
 {
@@ -5757,6 +5759,7 @@ printf %s "checking build system compile
 # remove anything that might look like compiler output to our "||" expression
 rm -f conftest* a.out b.out a.exe a_out.exe
 cat >conftest.c <<EOF
+#include <stdlib.h>
 int
 main ()
 {
@@ -5808,6 +5811,7 @@ else $as_nop
   gmp_cv_prog_exeext_for_build="$EXEEXT"
 else
   cat >conftest.c <<EOF
+#include <stdlib.h>
 int
 main ()
 {
