--- a/configure
+++ b/configure
@@ -4719,6 +4719,7 @@ $as_echo_n "checking build system compil
 # remove anything that might look like compiler output to our "||" expression
 rm -f conftest* a.out b.out a.exe a_out.exe
 cat >conftest.c <<EOF
+#include <stdlib.h>
 int
 main ()
 {
@@ -4751,6 +4752,7 @@ $as_echo_n "checking build system compil
 # remove anything that might look like compiler output to our "||" expression
 rm -f conftest* a.out b.out a.exe a_out.exe
 cat >conftest.c <<EOF
+#include <stdlib.h>
 int
 main ()
 {
@@ -4787,6 +4789,7 @@ $as_echo_n "checking build system compil
 # remove anything that might look like compiler output to our "||" expression
 rm -f conftest* a.out b.out a.exe a_out.exe
 cat >conftest.c <<EOF
+#include <stdlib.h>
 int
 main ()
 {
@@ -4837,6 +4840,7 @@ else
   gmp_cv_prog_exeext_for_build="$EXEEXT"
 else
   cat >conftest.c <<EOF
+#include <stdlib.h>
 int
 main ()
 {
