--- a/include/linux/netdevice.h
+++ b/include/linux/netdevice.h
@@ -3422,6 +3422,7 @@ static inline void unregister_netdevice(
 
 int netdev_refcnt_read(const struct net_device *dev);
 void free_netdev(struct net_device *dev);
+void init_dummy_netdev(struct net_device *dev);
 
 struct net_device *netdev_get_xmit_slave(struct net_device *dev,
 					 struct sk_buff *skb,
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -11476,8 +11476,13 @@ EXPORT_SYMBOL(register_netdevice);
  * The setup steps dummy netdevs need which normal netdevs get by going
  * through register_netdevice().
  */
-static void init_dummy_netdev(struct net_device *dev)
+void init_dummy_netdev(struct net_device *dev)
 {
+	memset(dev, 0, sizeof(struct net_device));
+	INIT_LIST_HEAD(&dev->napi_list);
+	mutex_init(&dev->lock);
+	dev_net_set(dev, &init_net);
+
 	/* make sure we BUG if trying to hit standard
 	 * register/unregister code path
 	 */
@@ -11492,6 +11497,7 @@ static void init_dummy_netdev(struct net
 	 * its refcount.
 	 */
 }
+EXPORT_SYMBOL_GPL(init_dummy_netdev);
 
 /**
  *	register_netdev	- register a network device
