--- a/include/linux/netdevice.h
+++ b/include/linux/netdevice.h
@@ -3392,6 +3392,7 @@ static inline void unregister_netdevice(
 
 int netdev_refcnt_read(const struct net_device *dev);
 void free_netdev(struct net_device *dev);
+void init_dummy_netdev(struct net_device *dev);
 
 struct net_device *netdev_get_xmit_slave(struct net_device *dev,
 					 struct sk_buff *skb,
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -11436,7 +11436,7 @@ EXPORT_SYMBOL(register_netdevice);
  * The setup steps dummy netdevs need which normal netdevs get by going
  * through register_netdevice().
  */
-static void init_dummy_netdev(struct net_device *dev)
+void init_dummy_netdev(struct net_device *dev)
 {
 	/* make sure we BUG if trying to hit standard
 	 * register/unregister code path
@@ -11452,6 +11452,7 @@ static void init_dummy_netdev(struct net
 	 * its refcount.
 	 */
 }
+EXPORT_SYMBOL_GPL(init_dummy_netdev);
 
 /**
  *	register_netdev	- register a network device
