--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -55,6 +55,7 @@
 
 /* task_struct member predeclarations (sorted alphabetically): */
 struct audit_context;
+struct backing_dev_info;
 struct bio_list;
 struct blk_plug;
 struct bpf_local_storage;
@@ -1290,6 +1291,8 @@ struct task_struct {
 
 	/* VM state: */
 	struct reclaim_state		*reclaim_state;
+	
+	struct backing_dev_info		*backing_dev_info;
 
 	struct io_context		*io_context;
 
--- a/fs/ext4/file.c
+++ b/fs/ext4/file.c
@@ -296,7 +296,9 @@ static ssize_t ext4_buffered_write_iter(
 	if (ret <= 0)
 		goto out;
 
+	current->backing_dev_info = inode_to_bdi(inode);
 	ret = generic_perform_write(iocb, from);
+	current->backing_dev_info = NULL;
 
 out:
 	inode_unlock(inode);
