--- a/include/net/dropreason.h
+++ b/include/net/dropreason.h
@@ -18,6 +18,12 @@ enum skb_drop_reason_subsys {
 	SKB_DROP_REASON_SUBSYS_MAC80211_UNUSABLE,
 
 	/**
+	 * @SKB_DROP_REASON_SUBSYS_MAC80211_MONITOR: mac80211 drop reasons
+	 * for frames still going to monitor, see net/mac80211/drop.h
+	 */
+	SKB_DROP_REASON_SUBSYS_MAC80211_MONITOR,
+
+	/**
 	 * @SKB_DROP_REASON_SUBSYS_OPENVSWITCH: openvswitch drop reasons,
 	 * see net/openvswitch/drop.h
 	 */
