--- a/configfiles/shorewall.conf
+++ b/configfiles/shorewall.conf
@@ -33,7 +33,7 @@ FIREWALL=
 #			       L O G G I N G
 ###############################################################################
 
-LOG_LEVEL="info"
+LOG_LEVEL="warn"
 
 BLACKLIST_LOG_LEVEL=
 
@@ -103,7 +103,7 @@ RESTOREFILE=restore
 
 SHOREWALL_SHELL=/bin/sh
 
-SUBSYSLOCK=/var/lock/subsys/shorewall
+SUBSYSLOCK=/var/lock/shorewall
 
 TC=
 
