--- a/drivers/gpu/drm/i915/gt/uc/intel_uc.c
+++ b/drivers/gpu/drm/i915/gt/uc/intel_uc.c
@@ -27,13 +27,13 @@ static void uc_expand_default_options(st
 
 	/* Don't enable GuC/HuC on pre-Gen12 */
 	if (GRAPHICS_VER(i915) < 12) {
-		i915->params.enable_guc = 0;
+		i915->params.enable_guc = 3;
 		return;
 	}
 
 	/* Don't enable GuC/HuC on older Gen12 platforms */
 	if (IS_TIGERLAKE(i915) || IS_ROCKETLAKE(i915)) {
-		i915->params.enable_guc = 0;
+		i915->params.enable_guc = 3;
 		return;
 	}
 
